/*! For license information please see main.53955aac.js.LICENSE.txt */
(()=>{"use strict";var e={956:(e,n,t)=>{var o=t(277),a={childContextTypes:!0,contextType:!0,contextTypes:!0,defaultProps:!0,displayName:!0,getDefaultProps:!0,getDerivedStateFromError:!0,getDerivedStateFromProps:!0,mixins:!0,propTypes:!0,type:!0},i={name:!0,length:!0,prototype:!0,caller:!0,callee:!0,arguments:!0,arity:!0},r={$$typeof:!0,compare:!0,defaultProps:!0,displayName:!0,propTypes:!0,type:!0},s={};function l(e){return o.isMemo(e)?r:s[e.$$typeof]||a}s[o.ForwardRef]={$$typeof:!0,render:!0,defaultProps:!0,displayName:!0,propTypes:!0},s[o.Memo]=r;var c=Object.defineProperty,d=Object.getOwnPropertyNames,u=Object.getOwnPropertySymbols,h=Object.getOwnPropertyDescriptor,p=Object.getPrototypeOf,m=Object.prototype;e.exports=function e(n,t,o){if("string"!==typeof t){if(m){var a=p(t);a&&a!==m&&e(n,a,o)}var r=d(t);u&&(r=r.concat(u(t)));for(var s=l(n),f=l(t),g=0;g<r.length;++g){var y=r[g];if(!i[y]&&(!o||!o[y])&&(!f||!f[y])&&(!s||!s[y])){var x=h(t,y);try{c(n,y,x)}catch(b){}}}}return n}},401:(e,n,t)=>{var o=t(998),a=t(588);function i(e){for(var n="https://reactjs.org/docs/error-decoder.html?invariant="+e,t=1;t<arguments.length;t++)n+="&args[]="+encodeURIComponent(arguments[t]);return"Minified React error #"+e+"; visit "+n+" for the full message or use the non-minified dev environment for full errors and additional helpful warnings."}var r=new Set,s={};function l(e,n){c(e,n),c(e+"Capture",n)}function c(e,n){for(s[e]=n,e=0;e<n.length;e++)r.add(n[e])}var d=!("undefined"===typeof window||"undefined"===typeof window.document||"undefined"===typeof window.document.createElement),u=Object.prototype.hasOwnProperty,h=/^[:A-Z_a-z\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u02FF\u0370-\u037D\u037F-\u1FFF\u200C-\u200D\u2070-\u218F\u2C00-\u2FEF\u3001-\uD7FF\uF900-\uFDCF\uFDF0-\uFFFD][:A-Z_a-z\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u02FF\u0370-\u037D\u037F-\u1FFF\u200C-\u200D\u2070-\u218F\u2C00-\u2FEF\u3001-\uD7FF\uF900-\uFDCF\uFDF0-\uFFFD\-.0-9\u00B7\u0300-\u036F\u203F-\u2040]*$/,p={},m={};function f(e,n,t,o,a,i,r){this.acceptsBooleans=2===n||3===n||4===n,this.attributeName=o,this.attributeNamespace=a,this.mustUseProperty=t,this.propertyName=e,this.type=n,this.sanitizeURL=i,this.removeEmptyString=r}var g={};"children dangerouslySetInnerHTML defaultValue defaultChecked innerHTML suppressContentEditableWarning suppressHydrationWarning style".split(" ").forEach((function(e){g[e]=new f(e,0,!1,e,null,!1,!1)})),[["acceptCharset","accept-charset"],["className","class"],["htmlFor","for"],["httpEquiv","http-equiv"]].forEach((function(e){var n=e[0];g[n]=new f(n,1,!1,e[1],null,!1,!1)})),["contentEditable","draggable","spellCheck","value"].forEach((function(e){g[e]=new f(e,2,!1,e.toLowerCase(),null,!1,!1)})),["autoReverse","externalResourcesRequired","focusable","preserveAlpha"].forEach((function(e){g[e]=new f(e,2,!1,e,null,!1,!1)})),"allowFullScreen async autoFocus autoPlay controls default defer disabled disablePictureInPicture disableRemotePlayback formNoValidate hidden loop noModule noValidate open playsInline readOnly required reversed scoped seamless itemScope".split(" ").forEach((function(e){g[e]=new f(e,3,!1,e.toLowerCase(),null,!1,!1)})),["checked","multiple","muted","selected"].forEach((function(e){g[e]=new f(e,3,!0,e,null,!1,!1)})),["capture","download"].forEach((function(e){g[e]=new f(e,4,!1,e,null,!1,!1)})),["cols","rows","size","span"].forEach((function(e){g[e]=new f(e,6,!1,e,null,!1,!1)})),["rowSpan","start"].forEach((function(e){g[e]=new f(e,5,!1,e.toLowerCase(),null,!1,!1)}));var y=/[\-:]([a-z])/g;function x(e){return e[1].toUpperCase()}function b(e,n,t,o){var a=g.hasOwnProperty(n)?g[n]:null;(null!==a?0!==a.type:o||!(2<n.length)||"o"!==n[0]&&"O"!==n[0]||"n"!==n[1]&&"N"!==n[1])&&(function(e,n,t,o){if(null===n||"undefined"===typeof n||function(e,n,t,o){if(null!==t&&0===t.type)return!1;switch(typeof n){case"function":case"symbol":return!0;case"boolean":return!o&&(null!==t?!t.acceptsBooleans:"data-"!==(e=e.toLowerCase().slice(0,5))&&"aria-"!==e);default:return!1}}(e,n,t,o))return!0;if(o)return!1;if(null!==t)switch(t.type){case 3:return!n;case 4:return!1===n;case 5:return isNaN(n);case 6:return isNaN(n)||1>n}return!1}(n,t,a,o)&&(t=null),o||null===a?function(e){return!!u.call(m,e)||!u.call(p,e)&&(h.test(e)?m[e]=!0:(p[e]=!0,!1))}(n)&&(null===t?e.removeAttribute(n):e.setAttribute(n,""+t)):a.mustUseProperty?e[a.propertyName]=null===t?3!==a.type&&"":t:(n=a.attributeName,o=a.attributeNamespace,null===t?e.removeAttribute(n):(t=3===(a=a.type)||4===a&&!0===t?"":""+t,o?e.setAttributeNS(o,n,t):e.setAttribute(n,t))))}"accent-height alignment-baseline arabic-form baseline-shift cap-height clip-path clip-rule color-interpolation color-interpolation-filters color-profile color-rendering dominant-baseline enable-background fill-opacity fill-rule flood-color flood-opacity font-family font-size font-size-adjust font-stretch font-style font-variant font-weight glyph-name glyph-orientation-horizontal glyph-orientation-vertical horiz-adv-x horiz-origin-x image-rendering letter-spacing lighting-color marker-end marker-mid marker-start overline-position overline-thickness paint-order panose-1 pointer-events rendering-intent shape-rendering stop-color stop-opacity strikethrough-position strikethrough-thickness stroke-dasharray stroke-dashoffset stroke-linecap stroke-linejoin stroke-miterlimit stroke-opacity stroke-width text-anchor text-decoration text-rendering underline-position underline-thickness unicode-bidi unicode-range units-per-em v-alphabetic v-hanging v-ideographic v-mathematical vector-effect vert-adv-y vert-origin-x vert-origin-y word-spacing writing-mode xmlns:xlink x-height".split(" ").forEach((function(e){var n=e.replace(y,x);g[n]=new f(n,1,!1,e,null,!1,!1)})),"xlink:actuate xlink:arcrole xlink:role xlink:show xlink:title xlink:type".split(" ").forEach((function(e){var n=e.replace(y,x);g[n]=new f(n,1,!1,e,"http://www.w3.org/1999/xlink",!1,!1)})),["xml:base","xml:lang","xml:space"].forEach((function(e){var n=e.replace(y,x);g[n]=new f(n,1,!1,e,"http://www.w3.org/XML/1998/namespace",!1,!1)})),["tabIndex","crossOrigin"].forEach((function(e){g[e]=new f(e,1,!1,e.toLowerCase(),null,!1,!1)})),g.xlinkHref=new f("xlinkHref",1,!1,"xlink:href","http://www.w3.org/1999/xlink",!0,!1),["src","href","action","formAction"].forEach((function(e){g[e]=new f(e,1,!1,e.toLowerCase(),null,!0,!0)}));var v=o.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED,j=Symbol.for("react.element"),C=Symbol.for("react.portal"),w=Symbol.for("react.fragment"),S=Symbol.for("react.strict_mode"),k=Symbol.for("react.profiler"),N=Symbol.for("react.provider"),_=Symbol.for("react.context"),T=Symbol.for("react.forward_ref"),I=Symbol.for("react.suspense"),E=Symbol.for("react.suspense_list"),A=Symbol.for("react.memo"),B=Symbol.for("react.lazy");Symbol.for("react.scope"),Symbol.for("react.debug_trace_mode");var R=Symbol.for("react.offscreen");Symbol.for("react.legacy_hidden"),Symbol.for("react.cache"),Symbol.for("react.tracing_marker");var F=Symbol.iterator;function P(e){return null===e||"object"!==typeof e?null:"function"===typeof(e=F&&e[F]||e["@@iterator"])?e:null}var D,O=Object.assign;function M(e){if(void 0===D)try{throw Error()}catch(t){var n=t.stack.trim().match(/\n( *(at )?)/);D=n&&n[1]||""}return"\n"+D+e}var L=!1;function G(e,n){if(!e||L)return"";L=!0;var t=Error.prepareStackTrace;Error.prepareStackTrace=void 0;try{if(n)if(n=function(){throw Error()},Object.defineProperty(n.prototype,"props",{set:function(){throw Error()}}),"object"===typeof Reflect&&Reflect.construct){try{Reflect.construct(n,[])}catch(c){var o=c}Reflect.construct(e,[],n)}else{try{n.call()}catch(c){o=c}e.call(n.prototype)}else{try{throw Error()}catch(c){o=c}e()}}catch(c){if(c&&o&&"string"===typeof c.stack){for(var a=c.stack.split("\n"),i=o.stack.split("\n"),r=a.length-1,s=i.length-1;1<=r&&0<=s&&a[r]!==i[s];)s--;for(;1<=r&&0<=s;r--,s--)if(a[r]!==i[s]){if(1!==r||1!==s)do{if(r--,0>--s||a[r]!==i[s]){var l="\n"+a[r].replace(" at new "," at ");return e.displayName&&l.includes("<anonymous>")&&(l=l.replace("<anonymous>",e.displayName)),l}}while(1<=r&&0<=s);break}}}finally{L=!1,Error.prepareStackTrace=t}return(e=e?e.displayName||e.name:"")?M(e):""}function V(e){switch(e.tag){case 5:return M(e.type);case 16:return M("Lazy");case 13:return M("Suspense");case 19:return M("SuspenseList");case 0:case 2:case 15:return e=G(e.type,!1);case 11:return e=G(e.type.render,!1);case 1:return e=G(e.type,!0);default:return""}}function q(e){if(null==e)return null;if("function"===typeof e)return e.displayName||e.name||null;if("string"===typeof e)return e;switch(e){case w:return"Fragment";case C:return"Portal";case k:return"Profiler";case S:return"StrictMode";case I:return"Suspense";case E:return"SuspenseList"}if("object"===typeof e)switch(e.$$typeof){case _:return(e.displayName||"Context")+".Consumer";case N:return(e._context.displayName||"Context")+".Provider";case T:var n=e.render;return(e=e.displayName)||(e=""!==(e=n.displayName||n.name||"")?"ForwardRef("+e+")":"ForwardRef"),e;case A:return null!==(n=e.displayName||null)?n:q(e.type)||"Memo";case B:n=e._payload,e=e._init;try{return q(e(n))}catch(t){}}return null}function H(e){var n=e.type;switch(e.tag){case 24:return"Cache";case 9:return(n.displayName||"Context")+".Consumer";case 10:return(n._context.displayName||"Context")+".Provider";case 18:return"DehydratedFragment";case 11:return e=(e=n.render).displayName||e.name||"",n.displayName||(""!==e?"ForwardRef("+e+")":"ForwardRef");case 7:return"Fragment";case 5:return n;case 4:return"Portal";case 3:return"Root";case 6:return"Text";case 16:return q(n);case 8:return n===S?"StrictMode":"Mode";case 22:return"Offscreen";case 12:return"Profiler";case 21:return"Scope";case 13:return"Suspense";case 19:return"SuspenseList";case 25:return"TracingMarker";case 1:case 0:case 17:case 2:case 14:case 15:if("function"===typeof n)return n.displayName||n.name||null;if("string"===typeof n)return n}return null}function z(e){switch(typeof e){case"boolean":case"number":case"string":case"undefined":case"object":return e;default:return""}}function U(e){var n=e.type;return(e=e.nodeName)&&"input"===e.toLowerCase()&&("checkbox"===n||"radio"===n)}function Q(e){e._valueTracker||(e._valueTracker=function(e){var n=U(e)?"checked":"value",t=Object.getOwnPropertyDescriptor(e.constructor.prototype,n),o=""+e[n];if(!e.hasOwnProperty(n)&&"undefined"!==typeof t&&"function"===typeof t.get&&"function"===typeof t.set){var a=t.get,i=t.set;return Object.defineProperty(e,n,{configurable:!0,get:function(){return a.call(this)},set:function(e){o=""+e,i.call(this,e)}}),Object.defineProperty(e,n,{enumerable:t.enumerable}),{getValue:function(){return o},setValue:function(e){o=""+e},stopTracking:function(){e._valueTracker=null,delete e[n]}}}}(e))}function W(e){if(!e)return!1;var n=e._valueTracker;if(!n)return!0;var t=n.getValue(),o="";return e&&(o=U(e)?e.checked?"true":"false":e.value),(e=o)!==t&&(n.setValue(e),!0)}function K(e){if("undefined"===typeof(e=e||("undefined"!==typeof document?document:void 0)))return null;try{return e.activeElement||e.body}catch(n){return e.body}}function $(e,n){var t=n.checked;return O({},n,{defaultChecked:void 0,defaultValue:void 0,value:void 0,checked:null!=t?t:e._wrapperState.initialChecked})}function Y(e,n){var t=null==n.defaultValue?"":n.defaultValue,o=null!=n.checked?n.checked:n.defaultChecked;t=z(null!=n.value?n.value:t),e._wrapperState={initialChecked:o,initialValue:t,controlled:"checkbox"===n.type||"radio"===n.type?null!=n.checked:null!=n.value}}function X(e,n){null!=(n=n.checked)&&b(e,"checked",n,!1)}function Z(e,n){X(e,n);var t=z(n.value),o=n.type;if(null!=t)"number"===o?(0===t&&""===e.value||e.value!=t)&&(e.value=""+t):e.value!==""+t&&(e.value=""+t);else if("submit"===o||"reset"===o)return void e.removeAttribute("value");n.hasOwnProperty("value")?ee(e,n.type,t):n.hasOwnProperty("defaultValue")&&ee(e,n.type,z(n.defaultValue)),null==n.checked&&null!=n.defaultChecked&&(e.defaultChecked=!!n.defaultChecked)}function J(e,n,t){if(n.hasOwnProperty("value")||n.hasOwnProperty("defaultValue")){var o=n.type;if(!("submit"!==o&&"reset"!==o||void 0!==n.value&&null!==n.value))return;n=""+e._wrapperState.initialValue,t||n===e.value||(e.value=n),e.defaultValue=n}""!==(t=e.name)&&(e.name=""),e.defaultChecked=!!e._wrapperState.initialChecked,""!==t&&(e.name=t)}function ee(e,n,t){"number"===n&&K(e.ownerDocument)===e||(null==t?e.defaultValue=""+e._wrapperState.initialValue:e.defaultValue!==""+t&&(e.defaultValue=""+t))}var ne=Array.isArray;function te(e,n,t,o){if(e=e.options,n){n={};for(var a=0;a<t.length;a++)n["$"+t[a]]=!0;for(t=0;t<e.length;t++)a=n.hasOwnProperty("$"+e[t].value),e[t].selected!==a&&(e[t].selected=a),a&&o&&(e[t].defaultSelected=!0)}else{for(t=""+z(t),n=null,a=0;a<e.length;a++){if(e[a].value===t)return e[a].selected=!0,void(o&&(e[a].defaultSelected=!0));null!==n||e[a].disabled||(n=e[a])}null!==n&&(n.selected=!0)}}function oe(e,n){if(null!=n.dangerouslySetInnerHTML)throw Error(i(91));return O({},n,{value:void 0,defaultValue:void 0,children:""+e._wrapperState.initialValue})}function ae(e,n){var t=n.value;if(null==t){if(t=n.children,n=n.defaultValue,null!=t){if(null!=n)throw Error(i(92));if(ne(t)){if(1<t.length)throw Error(i(93));t=t[0]}n=t}null==n&&(n=""),t=n}e._wrapperState={initialValue:z(t)}}function ie(e,n){var t=z(n.value),o=z(n.defaultValue);null!=t&&((t=""+t)!==e.value&&(e.value=t),null==n.defaultValue&&e.defaultValue!==t&&(e.defaultValue=t)),null!=o&&(e.defaultValue=""+o)}function re(e){var n=e.textContent;n===e._wrapperState.initialValue&&""!==n&&null!==n&&(e.value=n)}function se(e){switch(e){case"svg":return"http://www.w3.org/2000/svg";case"math":return"http://www.w3.org/1998/Math/MathML";default:return"http://www.w3.org/1999/xhtml"}}function le(e,n){return null==e||"http://www.w3.org/1999/xhtml"===e?se(n):"http://www.w3.org/2000/svg"===e&&"foreignObject"===n?"http://www.w3.org/1999/xhtml":e}var ce,de,ue=(de=function(e,n){if("http://www.w3.org/2000/svg"!==e.namespaceURI||"innerHTML"in e)e.innerHTML=n;else{for((ce=ce||document.createElement("div")).innerHTML="<svg>"+n.valueOf().toString()+"</svg>",n=ce.firstChild;e.firstChild;)e.removeChild(e.firstChild);for(;n.firstChild;)e.appendChild(n.firstChild)}},"undefined"!==typeof MSApp&&MSApp.execUnsafeLocalFunction?function(e,n,t,o){MSApp.execUnsafeLocalFunction((function(){return de(e,n)}))}:de);function he(e,n){if(n){var t=e.firstChild;if(t&&t===e.lastChild&&3===t.nodeType)return void(t.nodeValue=n)}e.textContent=n}var pe={animationIterationCount:!0,aspectRatio:!0,borderImageOutset:!0,borderImageSlice:!0,borderImageWidth:!0,boxFlex:!0,boxFlexGroup:!0,boxOrdinalGroup:!0,columnCount:!0,columns:!0,flex:!0,flexGrow:!0,flexPositive:!0,flexShrink:!0,flexNegative:!0,flexOrder:!0,gridArea:!0,gridRow:!0,gridRowEnd:!0,gridRowSpan:!0,gridRowStart:!0,gridColumn:!0,gridColumnEnd:!0,gridColumnSpan:!0,gridColumnStart:!0,fontWeight:!0,lineClamp:!0,lineHeight:!0,opacity:!0,order:!0,orphans:!0,tabSize:!0,widows:!0,zIndex:!0,zoom:!0,fillOpacity:!0,floodOpacity:!0,stopOpacity:!0,strokeDasharray:!0,strokeDashoffset:!0,strokeMiterlimit:!0,strokeOpacity:!0,strokeWidth:!0},me=["Webkit","ms","Moz","O"];function fe(e,n,t){return null==n||"boolean"===typeof n||""===n?"":t||"number"!==typeof n||0===n||pe.hasOwnProperty(e)&&pe[e]?(""+n).trim():n+"px"}function ge(e,n){for(var t in e=e.style,n)if(n.hasOwnProperty(t)){var o=0===t.indexOf("--"),a=fe(t,n[t],o);"float"===t&&(t="cssFloat"),o?e.setProperty(t,a):e[t]=a}}Object.keys(pe).forEach((function(e){me.forEach((function(n){n=n+e.charAt(0).toUpperCase()+e.substring(1),pe[n]=pe[e]}))}));var ye=O({menuitem:!0},{area:!0,base:!0,br:!0,col:!0,embed:!0,hr:!0,img:!0,input:!0,keygen:!0,link:!0,meta:!0,param:!0,source:!0,track:!0,wbr:!0});function xe(e,n){if(n){if(ye[e]&&(null!=n.children||null!=n.dangerouslySetInnerHTML))throw Error(i(137,e));if(null!=n.dangerouslySetInnerHTML){if(null!=n.children)throw Error(i(60));if("object"!==typeof n.dangerouslySetInnerHTML||!("__html"in n.dangerouslySetInnerHTML))throw Error(i(61))}if(null!=n.style&&"object"!==typeof n.style)throw Error(i(62))}}function be(e,n){if(-1===e.indexOf("-"))return"string"===typeof n.is;switch(e){case"annotation-xml":case"color-profile":case"font-face":case"font-face-src":case"font-face-uri":case"font-face-format":case"font-face-name":case"missing-glyph":return!1;default:return!0}}var ve=null;function je(e){return(e=e.target||e.srcElement||window).correspondingUseElement&&(e=e.correspondingUseElement),3===e.nodeType?e.parentNode:e}var Ce=null,we=null,Se=null;function ke(e){if(e=ja(e)){if("function"!==typeof Ce)throw Error(i(280));var n=e.stateNode;n&&(n=wa(n),Ce(e.stateNode,e.type,n))}}function Ne(e){we?Se?Se.push(e):Se=[e]:we=e}function _e(){if(we){var e=we,n=Se;if(Se=we=null,ke(e),n)for(e=0;e<n.length;e++)ke(n[e])}}function Ie(e,n){return e(n)}function Ee(){}var Ae=!1;function Be(e,n,t){if(Ae)return e(n,t);Ae=!0;try{return Ie(e,n,t)}finally{Ae=!1,(null!==we||null!==Se)&&(Ee(),_e())}}function Re(e,n){var t=e.stateNode;if(null===t)return null;var o=wa(t);if(null===o)return null;t=o[n];e:switch(n){case"onClick":case"onClickCapture":case"onDoubleClick":case"onDoubleClickCapture":case"onMouseDown":case"onMouseDownCapture":case"onMouseMove":case"onMouseMoveCapture":case"onMouseUp":case"onMouseUpCapture":case"onMouseEnter":(o=!o.disabled)||(o=!("button"===(e=e.type)||"input"===e||"select"===e||"textarea"===e)),e=!o;break e;default:e=!1}if(e)return null;if(t&&"function"!==typeof t)throw Error(i(231,n,typeof t));return t}var Fe=!1;if(d)try{var Pe={};Object.defineProperty(Pe,"passive",{get:function(){Fe=!0}}),window.addEventListener("test",Pe,Pe),window.removeEventListener("test",Pe,Pe)}catch(de){Fe=!1}function De(e,n,t,o,a,i,r,s,l){var c=Array.prototype.slice.call(arguments,3);try{n.apply(t,c)}catch(d){this.onError(d)}}var Oe=!1,Me=null,Le=!1,Ge=null,Ve={onError:function(e){Oe=!0,Me=e}};function qe(e,n,t,o,a,i,r,s,l){Oe=!1,Me=null,De.apply(Ve,arguments)}function He(e){var n=e,t=e;if(e.alternate)for(;n.return;)n=n.return;else{e=n;do{0!==(4098&(n=e).flags)&&(t=n.return),e=n.return}while(e)}return 3===n.tag?t:null}function ze(e){if(13===e.tag){var n=e.memoizedState;if(null===n&&(null!==(e=e.alternate)&&(n=e.memoizedState)),null!==n)return n.dehydrated}return null}function Ue(e){if(He(e)!==e)throw Error(i(188))}function Qe(e){return null!==(e=function(e){var n=e.alternate;if(!n){if(null===(n=He(e)))throw Error(i(188));return n!==e?null:e}for(var t=e,o=n;;){var a=t.return;if(null===a)break;var r=a.alternate;if(null===r){if(null!==(o=a.return)){t=o;continue}break}if(a.child===r.child){for(r=a.child;r;){if(r===t)return Ue(a),e;if(r===o)return Ue(a),n;r=r.sibling}throw Error(i(188))}if(t.return!==o.return)t=a,o=r;else{for(var s=!1,l=a.child;l;){if(l===t){s=!0,t=a,o=r;break}if(l===o){s=!0,o=a,t=r;break}l=l.sibling}if(!s){for(l=r.child;l;){if(l===t){s=!0,t=r,o=a;break}if(l===o){s=!0,o=r,t=a;break}l=l.sibling}if(!s)throw Error(i(189))}}if(t.alternate!==o)throw Error(i(190))}if(3!==t.tag)throw Error(i(188));return t.stateNode.current===t?e:n}(e))?We(e):null}function We(e){if(5===e.tag||6===e.tag)return e;for(e=e.child;null!==e;){var n=We(e);if(null!==n)return n;e=e.sibling}return null}var Ke=a.unstable_scheduleCallback,$e=a.unstable_cancelCallback,Ye=a.unstable_shouldYield,Xe=a.unstable_requestPaint,Ze=a.unstable_now,Je=a.unstable_getCurrentPriorityLevel,en=a.unstable_ImmediatePriority,nn=a.unstable_UserBlockingPriority,tn=a.unstable_NormalPriority,on=a.unstable_LowPriority,an=a.unstable_IdlePriority,rn=null,sn=null;var ln=Math.clz32?Math.clz32:function(e){return e>>>=0,0===e?32:31-(cn(e)/dn|0)|0},cn=Math.log,dn=Math.LN2;var un=64,hn=4194304;function pn(e){switch(e&-e){case 1:return 1;case 2:return 2;case 4:return 4;case 8:return 8;case 16:return 16;case 32:return 32;case 64:case 128:case 256:case 512:case 1024:case 2048:case 4096:case 8192:case 16384:case 32768:case 65536:case 131072:case 262144:case 524288:case 1048576:case 2097152:return 4194240&e;case 4194304:case 8388608:case 16777216:case 33554432:case 67108864:return 130023424&e;case 134217728:return 134217728;case 268435456:return 268435456;case 536870912:return 536870912;case 1073741824:return 1073741824;default:return e}}function mn(e,n){var t=e.pendingLanes;if(0===t)return 0;var o=0,a=e.suspendedLanes,i=e.pingedLanes,r=268435455&t;if(0!==r){var s=r&~a;0!==s?o=pn(s):0!==(i&=r)&&(o=pn(i))}else 0!==(r=t&~a)?o=pn(r):0!==i&&(o=pn(i));if(0===o)return 0;if(0!==n&&n!==o&&0===(n&a)&&((a=o&-o)>=(i=n&-n)||16===a&&0!==(4194240&i)))return n;if(0!==(4&o)&&(o|=16&t),0!==(n=e.entangledLanes))for(e=e.entanglements,n&=o;0<n;)a=1<<(t=31-ln(n)),o|=e[t],n&=~a;return o}function fn(e,n){switch(e){case 1:case 2:case 4:return n+250;case 8:case 16:case 32:case 64:case 128:case 256:case 512:case 1024:case 2048:case 4096:case 8192:case 16384:case 32768:case 65536:case 131072:case 262144:case 524288:case 1048576:case 2097152:return n+5e3;default:return-1}}function gn(e){return 0!==(e=-1073741825&e.pendingLanes)?e:1073741824&e?1073741824:0}function yn(){var e=un;return 0===(4194240&(un<<=1))&&(un=64),e}function xn(e){for(var n=[],t=0;31>t;t++)n.push(e);return n}function bn(e,n,t){e.pendingLanes|=n,536870912!==n&&(e.suspendedLanes=0,e.pingedLanes=0),(e=e.eventTimes)[n=31-ln(n)]=t}function vn(e,n){var t=e.entangledLanes|=n;for(e=e.entanglements;t;){var o=31-ln(t),a=1<<o;a&n|e[o]&n&&(e[o]|=n),t&=~a}}var jn=0;function Cn(e){return 1<(e&=-e)?4<e?0!==(268435455&e)?16:536870912:4:1}var wn,Sn,kn,Nn,_n,Tn=!1,In=[],En=null,An=null,Bn=null,Rn=new Map,Fn=new Map,Pn=[],Dn="mousedown mouseup touchcancel touchend touchstart auxclick dblclick pointercancel pointerdown pointerup dragend dragstart drop compositionend compositionstart keydown keypress keyup input textInput copy cut paste click change contextmenu reset submit".split(" ");function On(e,n){switch(e){case"focusin":case"focusout":En=null;break;case"dragenter":case"dragleave":An=null;break;case"mouseover":case"mouseout":Bn=null;break;case"pointerover":case"pointerout":Rn.delete(n.pointerId);break;case"gotpointercapture":case"lostpointercapture":Fn.delete(n.pointerId)}}function Mn(e,n,t,o,a,i){return null===e||e.nativeEvent!==i?(e={blockedOn:n,domEventName:t,eventSystemFlags:o,nativeEvent:i,targetContainers:[a]},null!==n&&(null!==(n=ja(n))&&Sn(n)),e):(e.eventSystemFlags|=o,n=e.targetContainers,null!==a&&-1===n.indexOf(a)&&n.push(a),e)}function Ln(e){var n=va(e.target);if(null!==n){var t=He(n);if(null!==t)if(13===(n=t.tag)){if(null!==(n=ze(t)))return e.blockedOn=n,void _n(e.priority,(function(){kn(t)}))}else if(3===n&&t.stateNode.current.memoizedState.isDehydrated)return void(e.blockedOn=3===t.tag?t.stateNode.containerInfo:null)}e.blockedOn=null}function Gn(e){if(null!==e.blockedOn)return!1;for(var n=e.targetContainers;0<n.length;){var t=Xn(e.domEventName,e.eventSystemFlags,n[0],e.nativeEvent);if(null!==t)return null!==(n=ja(t))&&Sn(n),e.blockedOn=t,!1;var o=new(t=e.nativeEvent).constructor(t.type,t);ve=o,t.target.dispatchEvent(o),ve=null,n.shift()}return!0}function Vn(e,n,t){Gn(e)&&t.delete(n)}function qn(){Tn=!1,null!==En&&Gn(En)&&(En=null),null!==An&&Gn(An)&&(An=null),null!==Bn&&Gn(Bn)&&(Bn=null),Rn.forEach(Vn),Fn.forEach(Vn)}function Hn(e,n){e.blockedOn===n&&(e.blockedOn=null,Tn||(Tn=!0,a.unstable_scheduleCallback(a.unstable_NormalPriority,qn)))}function zn(e){function n(n){return Hn(n,e)}if(0<In.length){Hn(In[0],e);for(var t=1;t<In.length;t++){var o=In[t];o.blockedOn===e&&(o.blockedOn=null)}}for(null!==En&&Hn(En,e),null!==An&&Hn(An,e),null!==Bn&&Hn(Bn,e),Rn.forEach(n),Fn.forEach(n),t=0;t<Pn.length;t++)(o=Pn[t]).blockedOn===e&&(o.blockedOn=null);for(;0<Pn.length&&null===(t=Pn[0]).blockedOn;)Ln(t),null===t.blockedOn&&Pn.shift()}var Un=v.ReactCurrentBatchConfig,Qn=!0;function Wn(e,n,t,o){var a=jn,i=Un.transition;Un.transition=null;try{jn=1,$n(e,n,t,o)}finally{jn=a,Un.transition=i}}function Kn(e,n,t,o){var a=jn,i=Un.transition;Un.transition=null;try{jn=4,$n(e,n,t,o)}finally{jn=a,Un.transition=i}}function $n(e,n,t,o){if(Qn){var a=Xn(e,n,t,o);if(null===a)Qo(e,n,o,Yn,t),On(e,o);else if(function(e,n,t,o,a){switch(n){case"focusin":return En=Mn(En,e,n,t,o,a),!0;case"dragenter":return An=Mn(An,e,n,t,o,a),!0;case"mouseover":return Bn=Mn(Bn,e,n,t,o,a),!0;case"pointerover":var i=a.pointerId;return Rn.set(i,Mn(Rn.get(i)||null,e,n,t,o,a)),!0;case"gotpointercapture":return i=a.pointerId,Fn.set(i,Mn(Fn.get(i)||null,e,n,t,o,a)),!0}return!1}(a,e,n,t,o))o.stopPropagation();else if(On(e,o),4&n&&-1<Dn.indexOf(e)){for(;null!==a;){var i=ja(a);if(null!==i&&wn(i),null===(i=Xn(e,n,t,o))&&Qo(e,n,o,Yn,t),i===a)break;a=i}null!==a&&o.stopPropagation()}else Qo(e,n,o,null,t)}}var Yn=null;function Xn(e,n,t,o){if(Yn=null,null!==(e=va(e=je(o))))if(null===(n=He(e)))e=null;else if(13===(t=n.tag)){if(null!==(e=ze(n)))return e;e=null}else if(3===t){if(n.stateNode.current.memoizedState.isDehydrated)return 3===n.tag?n.stateNode.containerInfo:null;e=null}else n!==e&&(e=null);return Yn=e,null}function Zn(e){switch(e){case"cancel":case"click":case"close":case"contextmenu":case"copy":case"cut":case"auxclick":case"dblclick":case"dragend":case"dragstart":case"drop":case"focusin":case"focusout":case"input":case"invalid":case"keydown":case"keypress":case"keyup":case"mousedown":case"mouseup":case"paste":case"pause":case"play":case"pointercancel":case"pointerdown":case"pointerup":case"ratechange":case"reset":case"resize":case"seeked":case"submit":case"touchcancel":case"touchend":case"touchstart":case"volumechange":case"change":case"selectionchange":case"textInput":case"compositionstart":case"compositionend":case"compositionupdate":case"beforeblur":case"afterblur":case"beforeinput":case"blur":case"fullscreenchange":case"focus":case"hashchange":case"popstate":case"select":case"selectstart":return 1;case"drag":case"dragenter":case"dragexit":case"dragleave":case"dragover":case"mousemove":case"mouseout":case"mouseover":case"pointermove":case"pointerout":case"pointerover":case"scroll":case"toggle":case"touchmove":case"wheel":case"mouseenter":case"mouseleave":case"pointerenter":case"pointerleave":return 4;case"message":switch(Je()){case en:return 1;case nn:return 4;case tn:case on:return 16;case an:return 536870912;default:return 16}default:return 16}}var Jn=null,et=null,nt=null;function tt(){if(nt)return nt;var e,n,t=et,o=t.length,a="value"in Jn?Jn.value:Jn.textContent,i=a.length;for(e=0;e<o&&t[e]===a[e];e++);var r=o-e;for(n=1;n<=r&&t[o-n]===a[i-n];n++);return nt=a.slice(e,1<n?1-n:void 0)}function ot(e){var n=e.keyCode;return"charCode"in e?0===(e=e.charCode)&&13===n&&(e=13):e=n,10===e&&(e=13),32<=e||13===e?e:0}function at(){return!0}function it(){return!1}function rt(e){function n(n,t,o,a,i){for(var r in this._reactName=n,this._targetInst=o,this.type=t,this.nativeEvent=a,this.target=i,this.currentTarget=null,e)e.hasOwnProperty(r)&&(n=e[r],this[r]=n?n(a):a[r]);return this.isDefaultPrevented=(null!=a.defaultPrevented?a.defaultPrevented:!1===a.returnValue)?at:it,this.isPropagationStopped=it,this}return O(n.prototype,{preventDefault:function(){this.defaultPrevented=!0;var e=this.nativeEvent;e&&(e.preventDefault?e.preventDefault():"unknown"!==typeof e.returnValue&&(e.returnValue=!1),this.isDefaultPrevented=at)},stopPropagation:function(){var e=this.nativeEvent;e&&(e.stopPropagation?e.stopPropagation():"unknown"!==typeof e.cancelBubble&&(e.cancelBubble=!0),this.isPropagationStopped=at)},persist:function(){},isPersistent:at}),n}var st,lt,ct,dt={eventPhase:0,bubbles:0,cancelable:0,timeStamp:function(e){return e.timeStamp||Date.now()},defaultPrevented:0,isTrusted:0},ut=rt(dt),ht=O({},dt,{view:0,detail:0}),pt=rt(ht),mt=O({},ht,{screenX:0,screenY:0,clientX:0,clientY:0,pageX:0,pageY:0,ctrlKey:0,shiftKey:0,altKey:0,metaKey:0,getModifierState:Nt,button:0,buttons:0,relatedTarget:function(e){return void 0===e.relatedTarget?e.fromElement===e.srcElement?e.toElement:e.fromElement:e.relatedTarget},movementX:function(e){return"movementX"in e?e.movementX:(e!==ct&&(ct&&"mousemove"===e.type?(st=e.screenX-ct.screenX,lt=e.screenY-ct.screenY):lt=st=0,ct=e),st)},movementY:function(e){return"movementY"in e?e.movementY:lt}}),ft=rt(mt),gt=rt(O({},mt,{dataTransfer:0})),yt=rt(O({},ht,{relatedTarget:0})),xt=rt(O({},dt,{animationName:0,elapsedTime:0,pseudoElement:0})),bt=O({},dt,{clipboardData:function(e){return"clipboardData"in e?e.clipboardData:window.clipboardData}}),vt=rt(bt),jt=rt(O({},dt,{data:0})),Ct={Esc:"Escape",Spacebar:" ",Left:"ArrowLeft",Up:"ArrowUp",Right:"ArrowRight",Down:"ArrowDown",Del:"Delete",Win:"OS",Menu:"ContextMenu",Apps:"ContextMenu",Scroll:"ScrollLock",MozPrintableKey:"Unidentified"},wt={8:"Backspace",9:"Tab",12:"Clear",13:"Enter",16:"Shift",17:"Control",18:"Alt",19:"Pause",20:"CapsLock",27:"Escape",32:" ",33:"PageUp",34:"PageDown",35:"End",36:"Home",37:"ArrowLeft",38:"ArrowUp",39:"ArrowRight",40:"ArrowDown",45:"Insert",46:"Delete",112:"F1",113:"F2",114:"F3",115:"F4",116:"F5",117:"F6",118:"F7",119:"F8",120:"F9",121:"F10",122:"F11",123:"F12",144:"NumLock",145:"ScrollLock",224:"Meta"},St={Alt:"altKey",Control:"ctrlKey",Meta:"metaKey",Shift:"shiftKey"};function kt(e){var n=this.nativeEvent;return n.getModifierState?n.getModifierState(e):!!(e=St[e])&&!!n[e]}function Nt(){return kt}var _t=O({},ht,{key:function(e){if(e.key){var n=Ct[e.key]||e.key;if("Unidentified"!==n)return n}return"keypress"===e.type?13===(e=ot(e))?"Enter":String.fromCharCode(e):"keydown"===e.type||"keyup"===e.type?wt[e.keyCode]||"Unidentified":""},code:0,location:0,ctrlKey:0,shiftKey:0,altKey:0,metaKey:0,repeat:0,locale:0,getModifierState:Nt,charCode:function(e){return"keypress"===e.type?ot(e):0},keyCode:function(e){return"keydown"===e.type||"keyup"===e.type?e.keyCode:0},which:function(e){return"keypress"===e.type?ot(e):"keydown"===e.type||"keyup"===e.type?e.keyCode:0}}),Tt=rt(_t),It=rt(O({},mt,{pointerId:0,width:0,height:0,pressure:0,tangentialPressure:0,tiltX:0,tiltY:0,twist:0,pointerType:0,isPrimary:0})),Et=rt(O({},ht,{touches:0,targetTouches:0,changedTouches:0,altKey:0,metaKey:0,ctrlKey:0,shiftKey:0,getModifierState:Nt})),At=rt(O({},dt,{propertyName:0,elapsedTime:0,pseudoElement:0})),Bt=O({},mt,{deltaX:function(e){return"deltaX"in e?e.deltaX:"wheelDeltaX"in e?-e.wheelDeltaX:0},deltaY:function(e){return"deltaY"in e?e.deltaY:"wheelDeltaY"in e?-e.wheelDeltaY:"wheelDelta"in e?-e.wheelDelta:0},deltaZ:0,deltaMode:0}),Rt=rt(Bt),Ft=[9,13,27,32],Pt=d&&"CompositionEvent"in window,Dt=null;d&&"documentMode"in document&&(Dt=document.documentMode);var Ot=d&&"TextEvent"in window&&!Dt,Mt=d&&(!Pt||Dt&&8<Dt&&11>=Dt),Lt=String.fromCharCode(32),Gt=!1;function Vt(e,n){switch(e){case"keyup":return-1!==Ft.indexOf(n.keyCode);case"keydown":return 229!==n.keyCode;case"keypress":case"mousedown":case"focusout":return!0;default:return!1}}function qt(e){return"object"===typeof(e=e.detail)&&"data"in e?e.data:null}var Ht=!1;var zt={color:!0,date:!0,datetime:!0,"datetime-local":!0,email:!0,month:!0,number:!0,password:!0,range:!0,search:!0,tel:!0,text:!0,time:!0,url:!0,week:!0};function Ut(e){var n=e&&e.nodeName&&e.nodeName.toLowerCase();return"input"===n?!!zt[e.type]:"textarea"===n}function Qt(e,n,t,o){Ne(o),0<(n=Ko(n,"onChange")).length&&(t=new ut("onChange","change",null,t,o),e.push({event:t,listeners:n}))}var Wt=null,Kt=null;function $t(e){Go(e,0)}function Yt(e){if(W(Ca(e)))return e}function Xt(e,n){if("change"===e)return n}var Zt=!1;if(d){var Jt;if(d){var eo="oninput"in document;if(!eo){var no=document.createElement("div");no.setAttribute("oninput","return;"),eo="function"===typeof no.oninput}Jt=eo}else Jt=!1;Zt=Jt&&(!document.documentMode||9<document.documentMode)}function to(){Wt&&(Wt.detachEvent("onpropertychange",oo),Kt=Wt=null)}function oo(e){if("value"===e.propertyName&&Yt(Kt)){var n=[];Qt(n,Kt,e,je(e)),Be($t,n)}}function ao(e,n,t){"focusin"===e?(to(),Kt=t,(Wt=n).attachEvent("onpropertychange",oo)):"focusout"===e&&to()}function io(e){if("selectionchange"===e||"keyup"===e||"keydown"===e)return Yt(Kt)}function ro(e,n){if("click"===e)return Yt(n)}function so(e,n){if("input"===e||"change"===e)return Yt(n)}var lo="function"===typeof Object.is?Object.is:function(e,n){return e===n&&(0!==e||1/e===1/n)||e!==e&&n!==n};function co(e,n){if(lo(e,n))return!0;if("object"!==typeof e||null===e||"object"!==typeof n||null===n)return!1;var t=Object.keys(e),o=Object.keys(n);if(t.length!==o.length)return!1;for(o=0;o<t.length;o++){var a=t[o];if(!u.call(n,a)||!lo(e[a],n[a]))return!1}return!0}function uo(e){for(;e&&e.firstChild;)e=e.firstChild;return e}function ho(e,n){var t,o=uo(e);for(e=0;o;){if(3===o.nodeType){if(t=e+o.textContent.length,e<=n&&t>=n)return{node:o,offset:n-e};e=t}e:{for(;o;){if(o.nextSibling){o=o.nextSibling;break e}o=o.parentNode}o=void 0}o=uo(o)}}function po(e,n){return!(!e||!n)&&(e===n||(!e||3!==e.nodeType)&&(n&&3===n.nodeType?po(e,n.parentNode):"contains"in e?e.contains(n):!!e.compareDocumentPosition&&!!(16&e.compareDocumentPosition(n))))}function mo(){for(var e=window,n=K();n instanceof e.HTMLIFrameElement;){try{var t="string"===typeof n.contentWindow.location.href}catch(o){t=!1}if(!t)break;n=K((e=n.contentWindow).document)}return n}function fo(e){var n=e&&e.nodeName&&e.nodeName.toLowerCase();return n&&("input"===n&&("text"===e.type||"search"===e.type||"tel"===e.type||"url"===e.type||"password"===e.type)||"textarea"===n||"true"===e.contentEditable)}function go(e){var n=mo(),t=e.focusedElem,o=e.selectionRange;if(n!==t&&t&&t.ownerDocument&&po(t.ownerDocument.documentElement,t)){if(null!==o&&fo(t))if(n=o.start,void 0===(e=o.end)&&(e=n),"selectionStart"in t)t.selectionStart=n,t.selectionEnd=Math.min(e,t.value.length);else if((e=(n=t.ownerDocument||document)&&n.defaultView||window).getSelection){e=e.getSelection();var a=t.textContent.length,i=Math.min(o.start,a);o=void 0===o.end?i:Math.min(o.end,a),!e.extend&&i>o&&(a=o,o=i,i=a),a=ho(t,i);var r=ho(t,o);a&&r&&(1!==e.rangeCount||e.anchorNode!==a.node||e.anchorOffset!==a.offset||e.focusNode!==r.node||e.focusOffset!==r.offset)&&((n=n.createRange()).setStart(a.node,a.offset),e.removeAllRanges(),i>o?(e.addRange(n),e.extend(r.node,r.offset)):(n.setEnd(r.node,r.offset),e.addRange(n)))}for(n=[],e=t;e=e.parentNode;)1===e.nodeType&&n.push({element:e,left:e.scrollLeft,top:e.scrollTop});for("function"===typeof t.focus&&t.focus(),t=0;t<n.length;t++)(e=n[t]).element.scrollLeft=e.left,e.element.scrollTop=e.top}}var yo=d&&"documentMode"in document&&11>=document.documentMode,xo=null,bo=null,vo=null,jo=!1;function Co(e,n,t){var o=t.window===t?t.document:9===t.nodeType?t:t.ownerDocument;jo||null==xo||xo!==K(o)||("selectionStart"in(o=xo)&&fo(o)?o={start:o.selectionStart,end:o.selectionEnd}:o={anchorNode:(o=(o.ownerDocument&&o.ownerDocument.defaultView||window).getSelection()).anchorNode,anchorOffset:o.anchorOffset,focusNode:o.focusNode,focusOffset:o.focusOffset},vo&&co(vo,o)||(vo=o,0<(o=Ko(bo,"onSelect")).length&&(n=new ut("onSelect","select",null,n,t),e.push({event:n,listeners:o}),n.target=xo)))}function wo(e,n){var t={};return t[e.toLowerCase()]=n.toLowerCase(),t["Webkit"+e]="webkit"+n,t["Moz"+e]="moz"+n,t}var So={animationend:wo("Animation","AnimationEnd"),animationiteration:wo("Animation","AnimationIteration"),animationstart:wo("Animation","AnimationStart"),transitionend:wo("Transition","TransitionEnd")},ko={},No={};function _o(e){if(ko[e])return ko[e];if(!So[e])return e;var n,t=So[e];for(n in t)if(t.hasOwnProperty(n)&&n in No)return ko[e]=t[n];return e}d&&(No=document.createElement("div").style,"AnimationEvent"in window||(delete So.animationend.animation,delete So.animationiteration.animation,delete So.animationstart.animation),"TransitionEvent"in window||delete So.transitionend.transition);var To=_o("animationend"),Io=_o("animationiteration"),Eo=_o("animationstart"),Ao=_o("transitionend"),Bo=new Map,Ro="abort auxClick cancel canPlay canPlayThrough click close contextMenu copy cut drag dragEnd dragEnter dragExit dragLeave dragOver dragStart drop durationChange emptied encrypted ended error gotPointerCapture input invalid keyDown keyPress keyUp load loadedData loadedMetadata loadStart lostPointerCapture mouseDown mouseMove mouseOut mouseOver mouseUp paste pause play playing pointerCancel pointerDown pointerMove pointerOut pointerOver pointerUp progress rateChange reset resize seeked seeking stalled submit suspend timeUpdate touchCancel touchEnd touchStart volumeChange scroll toggle touchMove waiting wheel".split(" ");function Fo(e,n){Bo.set(e,n),l(n,[e])}for(var Po=0;Po<Ro.length;Po++){var Do=Ro[Po];Fo(Do.toLowerCase(),"on"+(Do[0].toUpperCase()+Do.slice(1)))}Fo(To,"onAnimationEnd"),Fo(Io,"onAnimationIteration"),Fo(Eo,"onAnimationStart"),Fo("dblclick","onDoubleClick"),Fo("focusin","onFocus"),Fo("focusout","onBlur"),Fo(Ao,"onTransitionEnd"),c("onMouseEnter",["mouseout","mouseover"]),c("onMouseLeave",["mouseout","mouseover"]),c("onPointerEnter",["pointerout","pointerover"]),c("onPointerLeave",["pointerout","pointerover"]),l("onChange","change click focusin focusout input keydown keyup selectionchange".split(" ")),l("onSelect","focusout contextmenu dragend focusin keydown keyup mousedown mouseup selectionchange".split(" ")),l("onBeforeInput",["compositionend","keypress","textInput","paste"]),l("onCompositionEnd","compositionend focusout keydown keypress keyup mousedown".split(" ")),l("onCompositionStart","compositionstart focusout keydown keypress keyup mousedown".split(" ")),l("onCompositionUpdate","compositionupdate focusout keydown keypress keyup mousedown".split(" "));var Oo="abort canplay canplaythrough durationchange emptied encrypted ended error loadeddata loadedmetadata loadstart pause play playing progress ratechange resize seeked seeking stalled suspend timeupdate volumechange waiting".split(" "),Mo=new Set("cancel close invalid load scroll toggle".split(" ").concat(Oo));function Lo(e,n,t){var o=e.type||"unknown-event";e.currentTarget=t,function(e,n,t,o,a,r,s,l,c){if(qe.apply(this,arguments),Oe){if(!Oe)throw Error(i(198));var d=Me;Oe=!1,Me=null,Le||(Le=!0,Ge=d)}}(o,n,void 0,e),e.currentTarget=null}function Go(e,n){n=0!==(4&n);for(var t=0;t<e.length;t++){var o=e[t],a=o.event;o=o.listeners;e:{var i=void 0;if(n)for(var r=o.length-1;0<=r;r--){var s=o[r],l=s.instance,c=s.currentTarget;if(s=s.listener,l!==i&&a.isPropagationStopped())break e;Lo(a,s,c),i=l}else for(r=0;r<o.length;r++){if(l=(s=o[r]).instance,c=s.currentTarget,s=s.listener,l!==i&&a.isPropagationStopped())break e;Lo(a,s,c),i=l}}}if(Le)throw e=Ge,Le=!1,Ge=null,e}function Vo(e,n){var t=n[ya];void 0===t&&(t=n[ya]=new Set);var o=e+"__bubble";t.has(o)||(Uo(n,e,2,!1),t.add(o))}function qo(e,n,t){var o=0;n&&(o|=4),Uo(t,e,o,n)}var Ho="_reactListening"+Math.random().toString(36).slice(2);function zo(e){if(!e[Ho]){e[Ho]=!0,r.forEach((function(n){"selectionchange"!==n&&(Mo.has(n)||qo(n,!1,e),qo(n,!0,e))}));var n=9===e.nodeType?e:e.ownerDocument;null===n||n[Ho]||(n[Ho]=!0,qo("selectionchange",!1,n))}}function Uo(e,n,t,o){switch(Zn(n)){case 1:var a=Wn;break;case 4:a=Kn;break;default:a=$n}t=a.bind(null,n,t,e),a=void 0,!Fe||"touchstart"!==n&&"touchmove"!==n&&"wheel"!==n||(a=!0),o?void 0!==a?e.addEventListener(n,t,{capture:!0,passive:a}):e.addEventListener(n,t,!0):void 0!==a?e.addEventListener(n,t,{passive:a}):e.addEventListener(n,t,!1)}function Qo(e,n,t,o,a){var i=o;if(0===(1&n)&&0===(2&n)&&null!==o)e:for(;;){if(null===o)return;var r=o.tag;if(3===r||4===r){var s=o.stateNode.containerInfo;if(s===a||8===s.nodeType&&s.parentNode===a)break;if(4===r)for(r=o.return;null!==r;){var l=r.tag;if((3===l||4===l)&&((l=r.stateNode.containerInfo)===a||8===l.nodeType&&l.parentNode===a))return;r=r.return}for(;null!==s;){if(null===(r=va(s)))return;if(5===(l=r.tag)||6===l){o=i=r;continue e}s=s.parentNode}}o=o.return}Be((function(){var o=i,a=je(t),r=[];e:{var s=Bo.get(e);if(void 0!==s){var l=ut,c=e;switch(e){case"keypress":if(0===ot(t))break e;case"keydown":case"keyup":l=Tt;break;case"focusin":c="focus",l=yt;break;case"focusout":c="blur",l=yt;break;case"beforeblur":case"afterblur":l=yt;break;case"click":if(2===t.button)break e;case"auxclick":case"dblclick":case"mousedown":case"mousemove":case"mouseup":case"mouseout":case"mouseover":case"contextmenu":l=ft;break;case"drag":case"dragend":case"dragenter":case"dragexit":case"dragleave":case"dragover":case"dragstart":case"drop":l=gt;break;case"touchcancel":case"touchend":case"touchmove":case"touchstart":l=Et;break;case To:case Io:case Eo:l=xt;break;case Ao:l=At;break;case"scroll":l=pt;break;case"wheel":l=Rt;break;case"copy":case"cut":case"paste":l=vt;break;case"gotpointercapture":case"lostpointercapture":case"pointercancel":case"pointerdown":case"pointermove":case"pointerout":case"pointerover":case"pointerup":l=It}var d=0!==(4&n),u=!d&&"scroll"===e,h=d?null!==s?s+"Capture":null:s;d=[];for(var p,m=o;null!==m;){var f=(p=m).stateNode;if(5===p.tag&&null!==f&&(p=f,null!==h&&(null!=(f=Re(m,h))&&d.push(Wo(m,f,p)))),u)break;m=m.return}0<d.length&&(s=new l(s,c,null,t,a),r.push({event:s,listeners:d}))}}if(0===(7&n)){if(l="mouseout"===e||"pointerout"===e,(!(s="mouseover"===e||"pointerover"===e)||t===ve||!(c=t.relatedTarget||t.fromElement)||!va(c)&&!c[ga])&&(l||s)&&(s=a.window===a?a:(s=a.ownerDocument)?s.defaultView||s.parentWindow:window,l?(l=o,null!==(c=(c=t.relatedTarget||t.toElement)?va(c):null)&&(c!==(u=He(c))||5!==c.tag&&6!==c.tag)&&(c=null)):(l=null,c=o),l!==c)){if(d=ft,f="onMouseLeave",h="onMouseEnter",m="mouse","pointerout"!==e&&"pointerover"!==e||(d=It,f="onPointerLeave",h="onPointerEnter",m="pointer"),u=null==l?s:Ca(l),p=null==c?s:Ca(c),(s=new d(f,m+"leave",l,t,a)).target=u,s.relatedTarget=p,f=null,va(a)===o&&((d=new d(h,m+"enter",c,t,a)).target=p,d.relatedTarget=u,f=d),u=f,l&&c)e:{for(h=c,m=0,p=d=l;p;p=$o(p))m++;for(p=0,f=h;f;f=$o(f))p++;for(;0<m-p;)d=$o(d),m--;for(;0<p-m;)h=$o(h),p--;for(;m--;){if(d===h||null!==h&&d===h.alternate)break e;d=$o(d),h=$o(h)}d=null}else d=null;null!==l&&Yo(r,s,l,d,!1),null!==c&&null!==u&&Yo(r,u,c,d,!0)}if("select"===(l=(s=o?Ca(o):window).nodeName&&s.nodeName.toLowerCase())||"input"===l&&"file"===s.type)var g=Xt;else if(Ut(s))if(Zt)g=so;else{g=io;var y=ao}else(l=s.nodeName)&&"input"===l.toLowerCase()&&("checkbox"===s.type||"radio"===s.type)&&(g=ro);switch(g&&(g=g(e,o))?Qt(r,g,t,a):(y&&y(e,s,o),"focusout"===e&&(y=s._wrapperState)&&y.controlled&&"number"===s.type&&ee(s,"number",s.value)),y=o?Ca(o):window,e){case"focusin":(Ut(y)||"true"===y.contentEditable)&&(xo=y,bo=o,vo=null);break;case"focusout":vo=bo=xo=null;break;case"mousedown":jo=!0;break;case"contextmenu":case"mouseup":case"dragend":jo=!1,Co(r,t,a);break;case"selectionchange":if(yo)break;case"keydown":case"keyup":Co(r,t,a)}var x;if(Pt)e:{switch(e){case"compositionstart":var b="onCompositionStart";break e;case"compositionend":b="onCompositionEnd";break e;case"compositionupdate":b="onCompositionUpdate";break e}b=void 0}else Ht?Vt(e,t)&&(b="onCompositionEnd"):"keydown"===e&&229===t.keyCode&&(b="onCompositionStart");b&&(Mt&&"ko"!==t.locale&&(Ht||"onCompositionStart"!==b?"onCompositionEnd"===b&&Ht&&(x=tt()):(et="value"in(Jn=a)?Jn.value:Jn.textContent,Ht=!0)),0<(y=Ko(o,b)).length&&(b=new jt(b,e,null,t,a),r.push({event:b,listeners:y}),x?b.data=x:null!==(x=qt(t))&&(b.data=x))),(x=Ot?function(e,n){switch(e){case"compositionend":return qt(n);case"keypress":return 32!==n.which?null:(Gt=!0,Lt);case"textInput":return(e=n.data)===Lt&&Gt?null:e;default:return null}}(e,t):function(e,n){if(Ht)return"compositionend"===e||!Pt&&Vt(e,n)?(e=tt(),nt=et=Jn=null,Ht=!1,e):null;switch(e){case"paste":default:return null;case"keypress":if(!(n.ctrlKey||n.altKey||n.metaKey)||n.ctrlKey&&n.altKey){if(n.char&&1<n.char.length)return n.char;if(n.which)return String.fromCharCode(n.which)}return null;case"compositionend":return Mt&&"ko"!==n.locale?null:n.data}}(e,t))&&(0<(o=Ko(o,"onBeforeInput")).length&&(a=new jt("onBeforeInput","beforeinput",null,t,a),r.push({event:a,listeners:o}),a.data=x))}Go(r,n)}))}function Wo(e,n,t){return{instance:e,listener:n,currentTarget:t}}function Ko(e,n){for(var t=n+"Capture",o=[];null!==e;){var a=e,i=a.stateNode;5===a.tag&&null!==i&&(a=i,null!=(i=Re(e,t))&&o.unshift(Wo(e,i,a)),null!=(i=Re(e,n))&&o.push(Wo(e,i,a))),e=e.return}return o}function $o(e){if(null===e)return null;do{e=e.return}while(e&&5!==e.tag);return e||null}function Yo(e,n,t,o,a){for(var i=n._reactName,r=[];null!==t&&t!==o;){var s=t,l=s.alternate,c=s.stateNode;if(null!==l&&l===o)break;5===s.tag&&null!==c&&(s=c,a?null!=(l=Re(t,i))&&r.unshift(Wo(t,l,s)):a||null!=(l=Re(t,i))&&r.push(Wo(t,l,s))),t=t.return}0!==r.length&&e.push({event:n,listeners:r})}var Xo=/\r\n?/g,Zo=/\u0000|\uFFFD/g;function Jo(e){return("string"===typeof e?e:""+e).replace(Xo,"\n").replace(Zo,"")}function ea(e,n,t){if(n=Jo(n),Jo(e)!==n&&t)throw Error(i(425))}function na(){}var ta=null,oa=null;function aa(e,n){return"textarea"===e||"noscript"===e||"string"===typeof n.children||"number"===typeof n.children||"object"===typeof n.dangerouslySetInnerHTML&&null!==n.dangerouslySetInnerHTML&&null!=n.dangerouslySetInnerHTML.__html}var ia="function"===typeof setTimeout?setTimeout:void 0,ra="function"===typeof clearTimeout?clearTimeout:void 0,sa="function"===typeof Promise?Promise:void 0,la="function"===typeof queueMicrotask?queueMicrotask:"undefined"!==typeof sa?function(e){return sa.resolve(null).then(e).catch(ca)}:ia;function ca(e){setTimeout((function(){throw e}))}function da(e,n){var t=n,o=0;do{var a=t.nextSibling;if(e.removeChild(t),a&&8===a.nodeType)if("/$"===(t=a.data)){if(0===o)return e.removeChild(a),void zn(n);o--}else"$"!==t&&"$?"!==t&&"$!"!==t||o++;t=a}while(t);zn(n)}function ua(e){for(;null!=e;e=e.nextSibling){var n=e.nodeType;if(1===n||3===n)break;if(8===n){if("$"===(n=e.data)||"$!"===n||"$?"===n)break;if("/$"===n)return null}}return e}function ha(e){e=e.previousSibling;for(var n=0;e;){if(8===e.nodeType){var t=e.data;if("$"===t||"$!"===t||"$?"===t){if(0===n)return e;n--}else"/$"===t&&n++}e=e.previousSibling}return null}var pa=Math.random().toString(36).slice(2),ma="__reactFiber$"+pa,fa="__reactProps$"+pa,ga="__reactContainer$"+pa,ya="__reactEvents$"+pa,xa="__reactListeners$"+pa,ba="__reactHandles$"+pa;function va(e){var n=e[ma];if(n)return n;for(var t=e.parentNode;t;){if(n=t[ga]||t[ma]){if(t=n.alternate,null!==n.child||null!==t&&null!==t.child)for(e=ha(e);null!==e;){if(t=e[ma])return t;e=ha(e)}return n}t=(e=t).parentNode}return null}function ja(e){return!(e=e[ma]||e[ga])||5!==e.tag&&6!==e.tag&&13!==e.tag&&3!==e.tag?null:e}function Ca(e){if(5===e.tag||6===e.tag)return e.stateNode;throw Error(i(33))}function wa(e){return e[fa]||null}var Sa=[],ka=-1;function Na(e){return{current:e}}function _a(e){0>ka||(e.current=Sa[ka],Sa[ka]=null,ka--)}function Ta(e,n){ka++,Sa[ka]=e.current,e.current=n}var Ia={},Ea=Na(Ia),Aa=Na(!1),Ba=Ia;function Ra(e,n){var t=e.type.contextTypes;if(!t)return Ia;var o=e.stateNode;if(o&&o.__reactInternalMemoizedUnmaskedChildContext===n)return o.__reactInternalMemoizedMaskedChildContext;var a,i={};for(a in t)i[a]=n[a];return o&&((e=e.stateNode).__reactInternalMemoizedUnmaskedChildContext=n,e.__reactInternalMemoizedMaskedChildContext=i),i}function Fa(e){return null!==(e=e.childContextTypes)&&void 0!==e}function Pa(){_a(Aa),_a(Ea)}function Da(e,n,t){if(Ea.current!==Ia)throw Error(i(168));Ta(Ea,n),Ta(Aa,t)}function Oa(e,n,t){var o=e.stateNode;if(n=n.childContextTypes,"function"!==typeof o.getChildContext)return t;for(var a in o=o.getChildContext())if(!(a in n))throw Error(i(108,H(e)||"Unknown",a));return O({},t,o)}function Ma(e){return e=(e=e.stateNode)&&e.__reactInternalMemoizedMergedChildContext||Ia,Ba=Ea.current,Ta(Ea,e),Ta(Aa,Aa.current),!0}function La(e,n,t){var o=e.stateNode;if(!o)throw Error(i(169));t?(e=Oa(e,n,Ba),o.__reactInternalMemoizedMergedChildContext=e,_a(Aa),_a(Ea),Ta(Ea,e)):_a(Aa),Ta(Aa,t)}var Ga=null,Va=!1,qa=!1;function Ha(e){null===Ga?Ga=[e]:Ga.push(e)}function za(){if(!qa&&null!==Ga){qa=!0;var e=0,n=jn;try{var t=Ga;for(jn=1;e<t.length;e++){var o=t[e];do{o=o(!0)}while(null!==o)}Ga=null,Va=!1}catch(a){throw null!==Ga&&(Ga=Ga.slice(e+1)),Ke(en,za),a}finally{jn=n,qa=!1}}return null}var Ua=[],Qa=0,Wa=null,Ka=0,$a=[],Ya=0,Xa=null,Za=1,Ja="";function ei(e,n){Ua[Qa++]=Ka,Ua[Qa++]=Wa,Wa=e,Ka=n}function ni(e,n,t){$a[Ya++]=Za,$a[Ya++]=Ja,$a[Ya++]=Xa,Xa=e;var o=Za;e=Ja;var a=32-ln(o)-1;o&=~(1<<a),t+=1;var i=32-ln(n)+a;if(30<i){var r=a-a%5;i=(o&(1<<r)-1).toString(32),o>>=r,a-=r,Za=1<<32-ln(n)+a|t<<a|o,Ja=i+e}else Za=1<<i|t<<a|o,Ja=e}function ti(e){null!==e.return&&(ei(e,1),ni(e,1,0))}function oi(e){for(;e===Wa;)Wa=Ua[--Qa],Ua[Qa]=null,Ka=Ua[--Qa],Ua[Qa]=null;for(;e===Xa;)Xa=$a[--Ya],$a[Ya]=null,Ja=$a[--Ya],$a[Ya]=null,Za=$a[--Ya],$a[Ya]=null}var ai=null,ii=null,ri=!1,si=null;function li(e,n){var t=Bc(5,null,null,0);t.elementType="DELETED",t.stateNode=n,t.return=e,null===(n=e.deletions)?(e.deletions=[t],e.flags|=16):n.push(t)}function ci(e,n){switch(e.tag){case 5:var t=e.type;return null!==(n=1!==n.nodeType||t.toLowerCase()!==n.nodeName.toLowerCase()?null:n)&&(e.stateNode=n,ai=e,ii=ua(n.firstChild),!0);case 6:return null!==(n=""===e.pendingProps||3!==n.nodeType?null:n)&&(e.stateNode=n,ai=e,ii=null,!0);case 13:return null!==(n=8!==n.nodeType?null:n)&&(t=null!==Xa?{id:Za,overflow:Ja}:null,e.memoizedState={dehydrated:n,treeContext:t,retryLane:1073741824},(t=Bc(18,null,null,0)).stateNode=n,t.return=e,e.child=t,ai=e,ii=null,!0);default:return!1}}function di(e){return 0!==(1&e.mode)&&0===(128&e.flags)}function ui(e){if(ri){var n=ii;if(n){var t=n;if(!ci(e,n)){if(di(e))throw Error(i(418));n=ua(t.nextSibling);var o=ai;n&&ci(e,n)?li(o,t):(e.flags=-4097&e.flags|2,ri=!1,ai=e)}}else{if(di(e))throw Error(i(418));e.flags=-4097&e.flags|2,ri=!1,ai=e}}}function hi(e){for(e=e.return;null!==e&&5!==e.tag&&3!==e.tag&&13!==e.tag;)e=e.return;ai=e}function pi(e){if(e!==ai)return!1;if(!ri)return hi(e),ri=!0,!1;var n;if((n=3!==e.tag)&&!(n=5!==e.tag)&&(n="head"!==(n=e.type)&&"body"!==n&&!aa(e.type,e.memoizedProps)),n&&(n=ii)){if(di(e))throw mi(),Error(i(418));for(;n;)li(e,n),n=ua(n.nextSibling)}if(hi(e),13===e.tag){if(!(e=null!==(e=e.memoizedState)?e.dehydrated:null))throw Error(i(317));e:{for(e=e.nextSibling,n=0;e;){if(8===e.nodeType){var t=e.data;if("/$"===t){if(0===n){ii=ua(e.nextSibling);break e}n--}else"$"!==t&&"$!"!==t&&"$?"!==t||n++}e=e.nextSibling}ii=null}}else ii=ai?ua(e.stateNode.nextSibling):null;return!0}function mi(){for(var e=ii;e;)e=ua(e.nextSibling)}function fi(){ii=ai=null,ri=!1}function gi(e){null===si?si=[e]:si.push(e)}var yi=v.ReactCurrentBatchConfig;function xi(e,n,t){if(null!==(e=t.ref)&&"function"!==typeof e&&"object"!==typeof e){if(t._owner){if(t=t._owner){if(1!==t.tag)throw Error(i(309));var o=t.stateNode}if(!o)throw Error(i(147,e));var a=o,r=""+e;return null!==n&&null!==n.ref&&"function"===typeof n.ref&&n.ref._stringRef===r?n.ref:(n=function(e){var n=a.refs;null===e?delete n[r]:n[r]=e},n._stringRef=r,n)}if("string"!==typeof e)throw Error(i(284));if(!t._owner)throw Error(i(290,e))}return e}function bi(e,n){throw e=Object.prototype.toString.call(n),Error(i(31,"[object Object]"===e?"object with keys {"+Object.keys(n).join(", ")+"}":e))}function vi(e){return(0,e._init)(e._payload)}function ji(e){function n(n,t){if(e){var o=n.deletions;null===o?(n.deletions=[t],n.flags|=16):o.push(t)}}function t(t,o){if(!e)return null;for(;null!==o;)n(t,o),o=o.sibling;return null}function o(e,n){for(e=new Map;null!==n;)null!==n.key?e.set(n.key,n):e.set(n.index,n),n=n.sibling;return e}function a(e,n){return(e=Fc(e,n)).index=0,e.sibling=null,e}function r(n,t,o){return n.index=o,e?null!==(o=n.alternate)?(o=o.index)<t?(n.flags|=2,t):o:(n.flags|=2,t):(n.flags|=1048576,t)}function s(n){return e&&null===n.alternate&&(n.flags|=2),n}function l(e,n,t,o){return null===n||6!==n.tag?((n=Mc(t,e.mode,o)).return=e,n):((n=a(n,t)).return=e,n)}function c(e,n,t,o){var i=t.type;return i===w?u(e,n,t.props.children,o,t.key):null!==n&&(n.elementType===i||"object"===typeof i&&null!==i&&i.$$typeof===B&&vi(i)===n.type)?((o=a(n,t.props)).ref=xi(e,n,t),o.return=e,o):((o=Pc(t.type,t.key,t.props,null,e.mode,o)).ref=xi(e,n,t),o.return=e,o)}function d(e,n,t,o){return null===n||4!==n.tag||n.stateNode.containerInfo!==t.containerInfo||n.stateNode.implementation!==t.implementation?((n=Lc(t,e.mode,o)).return=e,n):((n=a(n,t.children||[])).return=e,n)}function u(e,n,t,o,i){return null===n||7!==n.tag?((n=Dc(t,e.mode,o,i)).return=e,n):((n=a(n,t)).return=e,n)}function h(e,n,t){if("string"===typeof n&&""!==n||"number"===typeof n)return(n=Mc(""+n,e.mode,t)).return=e,n;if("object"===typeof n&&null!==n){switch(n.$$typeof){case j:return(t=Pc(n.type,n.key,n.props,null,e.mode,t)).ref=xi(e,null,n),t.return=e,t;case C:return(n=Lc(n,e.mode,t)).return=e,n;case B:return h(e,(0,n._init)(n._payload),t)}if(ne(n)||P(n))return(n=Dc(n,e.mode,t,null)).return=e,n;bi(e,n)}return null}function p(e,n,t,o){var a=null!==n?n.key:null;if("string"===typeof t&&""!==t||"number"===typeof t)return null!==a?null:l(e,n,""+t,o);if("object"===typeof t&&null!==t){switch(t.$$typeof){case j:return t.key===a?c(e,n,t,o):null;case C:return t.key===a?d(e,n,t,o):null;case B:return p(e,n,(a=t._init)(t._payload),o)}if(ne(t)||P(t))return null!==a?null:u(e,n,t,o,null);bi(e,t)}return null}function m(e,n,t,o,a){if("string"===typeof o&&""!==o||"number"===typeof o)return l(n,e=e.get(t)||null,""+o,a);if("object"===typeof o&&null!==o){switch(o.$$typeof){case j:return c(n,e=e.get(null===o.key?t:o.key)||null,o,a);case C:return d(n,e=e.get(null===o.key?t:o.key)||null,o,a);case B:return m(e,n,t,(0,o._init)(o._payload),a)}if(ne(o)||P(o))return u(n,e=e.get(t)||null,o,a,null);bi(n,o)}return null}function f(a,i,s,l){for(var c=null,d=null,u=i,f=i=0,g=null;null!==u&&f<s.length;f++){u.index>f?(g=u,u=null):g=u.sibling;var y=p(a,u,s[f],l);if(null===y){null===u&&(u=g);break}e&&u&&null===y.alternate&&n(a,u),i=r(y,i,f),null===d?c=y:d.sibling=y,d=y,u=g}if(f===s.length)return t(a,u),ri&&ei(a,f),c;if(null===u){for(;f<s.length;f++)null!==(u=h(a,s[f],l))&&(i=r(u,i,f),null===d?c=u:d.sibling=u,d=u);return ri&&ei(a,f),c}for(u=o(a,u);f<s.length;f++)null!==(g=m(u,a,f,s[f],l))&&(e&&null!==g.alternate&&u.delete(null===g.key?f:g.key),i=r(g,i,f),null===d?c=g:d.sibling=g,d=g);return e&&u.forEach((function(e){return n(a,e)})),ri&&ei(a,f),c}function g(a,s,l,c){var d=P(l);if("function"!==typeof d)throw Error(i(150));if(null==(l=d.call(l)))throw Error(i(151));for(var u=d=null,f=s,g=s=0,y=null,x=l.next();null!==f&&!x.done;g++,x=l.next()){f.index>g?(y=f,f=null):y=f.sibling;var b=p(a,f,x.value,c);if(null===b){null===f&&(f=y);break}e&&f&&null===b.alternate&&n(a,f),s=r(b,s,g),null===u?d=b:u.sibling=b,u=b,f=y}if(x.done)return t(a,f),ri&&ei(a,g),d;if(null===f){for(;!x.done;g++,x=l.next())null!==(x=h(a,x.value,c))&&(s=r(x,s,g),null===u?d=x:u.sibling=x,u=x);return ri&&ei(a,g),d}for(f=o(a,f);!x.done;g++,x=l.next())null!==(x=m(f,a,g,x.value,c))&&(e&&null!==x.alternate&&f.delete(null===x.key?g:x.key),s=r(x,s,g),null===u?d=x:u.sibling=x,u=x);return e&&f.forEach((function(e){return n(a,e)})),ri&&ei(a,g),d}return function e(o,i,r,l){if("object"===typeof r&&null!==r&&r.type===w&&null===r.key&&(r=r.props.children),"object"===typeof r&&null!==r){switch(r.$$typeof){case j:e:{for(var c=r.key,d=i;null!==d;){if(d.key===c){if((c=r.type)===w){if(7===d.tag){t(o,d.sibling),(i=a(d,r.props.children)).return=o,o=i;break e}}else if(d.elementType===c||"object"===typeof c&&null!==c&&c.$$typeof===B&&vi(c)===d.type){t(o,d.sibling),(i=a(d,r.props)).ref=xi(o,d,r),i.return=o,o=i;break e}t(o,d);break}n(o,d),d=d.sibling}r.type===w?((i=Dc(r.props.children,o.mode,l,r.key)).return=o,o=i):((l=Pc(r.type,r.key,r.props,null,o.mode,l)).ref=xi(o,i,r),l.return=o,o=l)}return s(o);case C:e:{for(d=r.key;null!==i;){if(i.key===d){if(4===i.tag&&i.stateNode.containerInfo===r.containerInfo&&i.stateNode.implementation===r.implementation){t(o,i.sibling),(i=a(i,r.children||[])).return=o,o=i;break e}t(o,i);break}n(o,i),i=i.sibling}(i=Lc(r,o.mode,l)).return=o,o=i}return s(o);case B:return e(o,i,(d=r._init)(r._payload),l)}if(ne(r))return f(o,i,r,l);if(P(r))return g(o,i,r,l);bi(o,r)}return"string"===typeof r&&""!==r||"number"===typeof r?(r=""+r,null!==i&&6===i.tag?(t(o,i.sibling),(i=a(i,r)).return=o,o=i):(t(o,i),(i=Mc(r,o.mode,l)).return=o,o=i),s(o)):t(o,i)}}var Ci=ji(!0),wi=ji(!1),Si=Na(null),ki=null,Ni=null,_i=null;function Ti(){_i=Ni=ki=null}function Ii(e){var n=Si.current;_a(Si),e._currentValue=n}function Ei(e,n,t){for(;null!==e;){var o=e.alternate;if((e.childLanes&n)!==n?(e.childLanes|=n,null!==o&&(o.childLanes|=n)):null!==o&&(o.childLanes&n)!==n&&(o.childLanes|=n),e===t)break;e=e.return}}function Ai(e,n){ki=e,_i=Ni=null,null!==(e=e.dependencies)&&null!==e.firstContext&&(0!==(e.lanes&n)&&(vs=!0),e.firstContext=null)}function Bi(e){var n=e._currentValue;if(_i!==e)if(e={context:e,memoizedValue:n,next:null},null===Ni){if(null===ki)throw Error(i(308));Ni=e,ki.dependencies={lanes:0,firstContext:e}}else Ni=Ni.next=e;return n}var Ri=null;function Fi(e){null===Ri?Ri=[e]:Ri.push(e)}function Pi(e,n,t,o){var a=n.interleaved;return null===a?(t.next=t,Fi(n)):(t.next=a.next,a.next=t),n.interleaved=t,Di(e,o)}function Di(e,n){e.lanes|=n;var t=e.alternate;for(null!==t&&(t.lanes|=n),t=e,e=e.return;null!==e;)e.childLanes|=n,null!==(t=e.alternate)&&(t.childLanes|=n),t=e,e=e.return;return 3===t.tag?t.stateNode:null}var Oi=!1;function Mi(e){e.updateQueue={baseState:e.memoizedState,firstBaseUpdate:null,lastBaseUpdate:null,shared:{pending:null,interleaved:null,lanes:0},effects:null}}function Li(e,n){e=e.updateQueue,n.updateQueue===e&&(n.updateQueue={baseState:e.baseState,firstBaseUpdate:e.firstBaseUpdate,lastBaseUpdate:e.lastBaseUpdate,shared:e.shared,effects:e.effects})}function Gi(e,n){return{eventTime:e,lane:n,tag:0,payload:null,callback:null,next:null}}function Vi(e,n,t){var o=e.updateQueue;if(null===o)return null;if(o=o.shared,0!==(2&Il)){var a=o.pending;return null===a?n.next=n:(n.next=a.next,a.next=n),o.pending=n,Di(e,t)}return null===(a=o.interleaved)?(n.next=n,Fi(o)):(n.next=a.next,a.next=n),o.interleaved=n,Di(e,t)}function qi(e,n,t){if(null!==(n=n.updateQueue)&&(n=n.shared,0!==(4194240&t))){var o=n.lanes;t|=o&=e.pendingLanes,n.lanes=t,vn(e,t)}}function Hi(e,n){var t=e.updateQueue,o=e.alternate;if(null!==o&&t===(o=o.updateQueue)){var a=null,i=null;if(null!==(t=t.firstBaseUpdate)){do{var r={eventTime:t.eventTime,lane:t.lane,tag:t.tag,payload:t.payload,callback:t.callback,next:null};null===i?a=i=r:i=i.next=r,t=t.next}while(null!==t);null===i?a=i=n:i=i.next=n}else a=i=n;return t={baseState:o.baseState,firstBaseUpdate:a,lastBaseUpdate:i,shared:o.shared,effects:o.effects},void(e.updateQueue=t)}null===(e=t.lastBaseUpdate)?t.firstBaseUpdate=n:e.next=n,t.lastBaseUpdate=n}function zi(e,n,t,o){var a=e.updateQueue;Oi=!1;var i=a.firstBaseUpdate,r=a.lastBaseUpdate,s=a.shared.pending;if(null!==s){a.shared.pending=null;var l=s,c=l.next;l.next=null,null===r?i=c:r.next=c,r=l;var d=e.alternate;null!==d&&((s=(d=d.updateQueue).lastBaseUpdate)!==r&&(null===s?d.firstBaseUpdate=c:s.next=c,d.lastBaseUpdate=l))}if(null!==i){var u=a.baseState;for(r=0,d=c=l=null,s=i;;){var h=s.lane,p=s.eventTime;if((o&h)===h){null!==d&&(d=d.next={eventTime:p,lane:0,tag:s.tag,payload:s.payload,callback:s.callback,next:null});e:{var m=e,f=s;switch(h=n,p=t,f.tag){case 1:if("function"===typeof(m=f.payload)){u=m.call(p,u,h);break e}u=m;break e;case 3:m.flags=-65537&m.flags|128;case 0:if(null===(h="function"===typeof(m=f.payload)?m.call(p,u,h):m)||void 0===h)break e;u=O({},u,h);break e;case 2:Oi=!0}}null!==s.callback&&0!==s.lane&&(e.flags|=64,null===(h=a.effects)?a.effects=[s]:h.push(s))}else p={eventTime:p,lane:h,tag:s.tag,payload:s.payload,callback:s.callback,next:null},null===d?(c=d=p,l=u):d=d.next=p,r|=h;if(null===(s=s.next)){if(null===(s=a.shared.pending))break;s=(h=s).next,h.next=null,a.lastBaseUpdate=h,a.shared.pending=null}}if(null===d&&(l=u),a.baseState=l,a.firstBaseUpdate=c,a.lastBaseUpdate=d,null!==(n=a.shared.interleaved)){a=n;do{r|=a.lane,a=a.next}while(a!==n)}else null===i&&(a.shared.lanes=0);Ol|=r,e.lanes=r,e.memoizedState=u}}function Ui(e,n,t){if(e=n.effects,n.effects=null,null!==e)for(n=0;n<e.length;n++){var o=e[n],a=o.callback;if(null!==a){if(o.callback=null,o=t,"function"!==typeof a)throw Error(i(191,a));a.call(o)}}}var Qi={},Wi=Na(Qi),Ki=Na(Qi),$i=Na(Qi);function Yi(e){if(e===Qi)throw Error(i(174));return e}function Xi(e,n){switch(Ta($i,n),Ta(Ki,e),Ta(Wi,Qi),e=n.nodeType){case 9:case 11:n=(n=n.documentElement)?n.namespaceURI:le(null,"");break;default:n=le(n=(e=8===e?n.parentNode:n).namespaceURI||null,e=e.tagName)}_a(Wi),Ta(Wi,n)}function Zi(){_a(Wi),_a(Ki),_a($i)}function Ji(e){Yi($i.current);var n=Yi(Wi.current),t=le(n,e.type);n!==t&&(Ta(Ki,e),Ta(Wi,t))}function er(e){Ki.current===e&&(_a(Wi),_a(Ki))}var nr=Na(0);function tr(e){for(var n=e;null!==n;){if(13===n.tag){var t=n.memoizedState;if(null!==t&&(null===(t=t.dehydrated)||"$?"===t.data||"$!"===t.data))return n}else if(19===n.tag&&void 0!==n.memoizedProps.revealOrder){if(0!==(128&n.flags))return n}else if(null!==n.child){n.child.return=n,n=n.child;continue}if(n===e)break;for(;null===n.sibling;){if(null===n.return||n.return===e)return null;n=n.return}n.sibling.return=n.return,n=n.sibling}return null}var or=[];function ar(){for(var e=0;e<or.length;e++)or[e]._workInProgressVersionPrimary=null;or.length=0}var ir=v.ReactCurrentDispatcher,rr=v.ReactCurrentBatchConfig,sr=0,lr=null,cr=null,dr=null,ur=!1,hr=!1,pr=0,mr=0;function fr(){throw Error(i(321))}function gr(e,n){if(null===n)return!1;for(var t=0;t<n.length&&t<e.length;t++)if(!lo(e[t],n[t]))return!1;return!0}function yr(e,n,t,o,a,r){if(sr=r,lr=n,n.memoizedState=null,n.updateQueue=null,n.lanes=0,ir.current=null===e||null===e.memoizedState?es:ns,e=t(o,a),hr){r=0;do{if(hr=!1,pr=0,25<=r)throw Error(i(301));r+=1,dr=cr=null,n.updateQueue=null,ir.current=ts,e=t(o,a)}while(hr)}if(ir.current=Jr,n=null!==cr&&null!==cr.next,sr=0,dr=cr=lr=null,ur=!1,n)throw Error(i(300));return e}function xr(){var e=0!==pr;return pr=0,e}function br(){var e={memoizedState:null,baseState:null,baseQueue:null,queue:null,next:null};return null===dr?lr.memoizedState=dr=e:dr=dr.next=e,dr}function vr(){if(null===cr){var e=lr.alternate;e=null!==e?e.memoizedState:null}else e=cr.next;var n=null===dr?lr.memoizedState:dr.next;if(null!==n)dr=n,cr=e;else{if(null===e)throw Error(i(310));e={memoizedState:(cr=e).memoizedState,baseState:cr.baseState,baseQueue:cr.baseQueue,queue:cr.queue,next:null},null===dr?lr.memoizedState=dr=e:dr=dr.next=e}return dr}function jr(e,n){return"function"===typeof n?n(e):n}function Cr(e){var n=vr(),t=n.queue;if(null===t)throw Error(i(311));t.lastRenderedReducer=e;var o=cr,a=o.baseQueue,r=t.pending;if(null!==r){if(null!==a){var s=a.next;a.next=r.next,r.next=s}o.baseQueue=a=r,t.pending=null}if(null!==a){r=a.next,o=o.baseState;var l=s=null,c=null,d=r;do{var u=d.lane;if((sr&u)===u)null!==c&&(c=c.next={lane:0,action:d.action,hasEagerState:d.hasEagerState,eagerState:d.eagerState,next:null}),o=d.hasEagerState?d.eagerState:e(o,d.action);else{var h={lane:u,action:d.action,hasEagerState:d.hasEagerState,eagerState:d.eagerState,next:null};null===c?(l=c=h,s=o):c=c.next=h,lr.lanes|=u,Ol|=u}d=d.next}while(null!==d&&d!==r);null===c?s=o:c.next=l,lo(o,n.memoizedState)||(vs=!0),n.memoizedState=o,n.baseState=s,n.baseQueue=c,t.lastRenderedState=o}if(null!==(e=t.interleaved)){a=e;do{r=a.lane,lr.lanes|=r,Ol|=r,a=a.next}while(a!==e)}else null===a&&(t.lanes=0);return[n.memoizedState,t.dispatch]}function wr(e){var n=vr(),t=n.queue;if(null===t)throw Error(i(311));t.lastRenderedReducer=e;var o=t.dispatch,a=t.pending,r=n.memoizedState;if(null!==a){t.pending=null;var s=a=a.next;do{r=e(r,s.action),s=s.next}while(s!==a);lo(r,n.memoizedState)||(vs=!0),n.memoizedState=r,null===n.baseQueue&&(n.baseState=r),t.lastRenderedState=r}return[r,o]}function Sr(){}function kr(e,n){var t=lr,o=vr(),a=n(),r=!lo(o.memoizedState,a);if(r&&(o.memoizedState=a,vs=!0),o=o.queue,Or(Tr.bind(null,t,o,e),[e]),o.getSnapshot!==n||r||null!==dr&&1&dr.memoizedState.tag){if(t.flags|=2048,Br(9,_r.bind(null,t,o,a,n),void 0,null),null===El)throw Error(i(349));0!==(30&sr)||Nr(t,n,a)}return a}function Nr(e,n,t){e.flags|=16384,e={getSnapshot:n,value:t},null===(n=lr.updateQueue)?(n={lastEffect:null,stores:null},lr.updateQueue=n,n.stores=[e]):null===(t=n.stores)?n.stores=[e]:t.push(e)}function _r(e,n,t,o){n.value=t,n.getSnapshot=o,Ir(n)&&Er(e)}function Tr(e,n,t){return t((function(){Ir(n)&&Er(e)}))}function Ir(e){var n=e.getSnapshot;e=e.value;try{var t=n();return!lo(e,t)}catch(o){return!0}}function Er(e){var n=Di(e,1);null!==n&&oc(n,e,1,-1)}function Ar(e){var n=br();return"function"===typeof e&&(e=e()),n.memoizedState=n.baseState=e,e={pending:null,interleaved:null,lanes:0,dispatch:null,lastRenderedReducer:jr,lastRenderedState:e},n.queue=e,e=e.dispatch=$r.bind(null,lr,e),[n.memoizedState,e]}function Br(e,n,t,o){return e={tag:e,create:n,destroy:t,deps:o,next:null},null===(n=lr.updateQueue)?(n={lastEffect:null,stores:null},lr.updateQueue=n,n.lastEffect=e.next=e):null===(t=n.lastEffect)?n.lastEffect=e.next=e:(o=t.next,t.next=e,e.next=o,n.lastEffect=e),e}function Rr(){return vr().memoizedState}function Fr(e,n,t,o){var a=br();lr.flags|=e,a.memoizedState=Br(1|n,t,void 0,void 0===o?null:o)}function Pr(e,n,t,o){var a=vr();o=void 0===o?null:o;var i=void 0;if(null!==cr){var r=cr.memoizedState;if(i=r.destroy,null!==o&&gr(o,r.deps))return void(a.memoizedState=Br(n,t,i,o))}lr.flags|=e,a.memoizedState=Br(1|n,t,i,o)}function Dr(e,n){return Fr(8390656,8,e,n)}function Or(e,n){return Pr(2048,8,e,n)}function Mr(e,n){return Pr(4,2,e,n)}function Lr(e,n){return Pr(4,4,e,n)}function Gr(e,n){return"function"===typeof n?(e=e(),n(e),function(){n(null)}):null!==n&&void 0!==n?(e=e(),n.current=e,function(){n.current=null}):void 0}function Vr(e,n,t){return t=null!==t&&void 0!==t?t.concat([e]):null,Pr(4,4,Gr.bind(null,n,e),t)}function qr(){}function Hr(e,n){var t=vr();n=void 0===n?null:n;var o=t.memoizedState;return null!==o&&null!==n&&gr(n,o[1])?o[0]:(t.memoizedState=[e,n],e)}function zr(e,n){var t=vr();n=void 0===n?null:n;var o=t.memoizedState;return null!==o&&null!==n&&gr(n,o[1])?o[0]:(e=e(),t.memoizedState=[e,n],e)}function Ur(e,n,t){return 0===(21&sr)?(e.baseState&&(e.baseState=!1,vs=!0),e.memoizedState=t):(lo(t,n)||(t=yn(),lr.lanes|=t,Ol|=t,e.baseState=!0),n)}function Qr(e,n){var t=jn;jn=0!==t&&4>t?t:4,e(!0);var o=rr.transition;rr.transition={};try{e(!1),n()}finally{jn=t,rr.transition=o}}function Wr(){return vr().memoizedState}function Kr(e,n,t){var o=tc(e);if(t={lane:o,action:t,hasEagerState:!1,eagerState:null,next:null},Yr(e))Xr(n,t);else if(null!==(t=Pi(e,n,t,o))){oc(t,e,o,nc()),Zr(t,n,o)}}function $r(e,n,t){var o=tc(e),a={lane:o,action:t,hasEagerState:!1,eagerState:null,next:null};if(Yr(e))Xr(n,a);else{var i=e.alternate;if(0===e.lanes&&(null===i||0===i.lanes)&&null!==(i=n.lastRenderedReducer))try{var r=n.lastRenderedState,s=i(r,t);if(a.hasEagerState=!0,a.eagerState=s,lo(s,r)){var l=n.interleaved;return null===l?(a.next=a,Fi(n)):(a.next=l.next,l.next=a),void(n.interleaved=a)}}catch(c){}null!==(t=Pi(e,n,a,o))&&(oc(t,e,o,a=nc()),Zr(t,n,o))}}function Yr(e){var n=e.alternate;return e===lr||null!==n&&n===lr}function Xr(e,n){hr=ur=!0;var t=e.pending;null===t?n.next=n:(n.next=t.next,t.next=n),e.pending=n}function Zr(e,n,t){if(0!==(4194240&t)){var o=n.lanes;t|=o&=e.pendingLanes,n.lanes=t,vn(e,t)}}var Jr={readContext:Bi,useCallback:fr,useContext:fr,useEffect:fr,useImperativeHandle:fr,useInsertionEffect:fr,useLayoutEffect:fr,useMemo:fr,useReducer:fr,useRef:fr,useState:fr,useDebugValue:fr,useDeferredValue:fr,useTransition:fr,useMutableSource:fr,useSyncExternalStore:fr,useId:fr,unstable_isNewReconciler:!1},es={readContext:Bi,useCallback:function(e,n){return br().memoizedState=[e,void 0===n?null:n],e},useContext:Bi,useEffect:Dr,useImperativeHandle:function(e,n,t){return t=null!==t&&void 0!==t?t.concat([e]):null,Fr(4194308,4,Gr.bind(null,n,e),t)},useLayoutEffect:function(e,n){return Fr(4194308,4,e,n)},useInsertionEffect:function(e,n){return Fr(4,2,e,n)},useMemo:function(e,n){var t=br();return n=void 0===n?null:n,e=e(),t.memoizedState=[e,n],e},useReducer:function(e,n,t){var o=br();return n=void 0!==t?t(n):n,o.memoizedState=o.baseState=n,e={pending:null,interleaved:null,lanes:0,dispatch:null,lastRenderedReducer:e,lastRenderedState:n},o.queue=e,e=e.dispatch=Kr.bind(null,lr,e),[o.memoizedState,e]},useRef:function(e){return e={current:e},br().memoizedState=e},useState:Ar,useDebugValue:qr,useDeferredValue:function(e){return br().memoizedState=e},useTransition:function(){var e=Ar(!1),n=e[0];return e=Qr.bind(null,e[1]),br().memoizedState=e,[n,e]},useMutableSource:function(){},useSyncExternalStore:function(e,n,t){var o=lr,a=br();if(ri){if(void 0===t)throw Error(i(407));t=t()}else{if(t=n(),null===El)throw Error(i(349));0!==(30&sr)||Nr(o,n,t)}a.memoizedState=t;var r={value:t,getSnapshot:n};return a.queue=r,Dr(Tr.bind(null,o,r,e),[e]),o.flags|=2048,Br(9,_r.bind(null,o,r,t,n),void 0,null),t},useId:function(){var e=br(),n=El.identifierPrefix;if(ri){var t=Ja;n=":"+n+"R"+(t=(Za&~(1<<32-ln(Za)-1)).toString(32)+t),0<(t=pr++)&&(n+="H"+t.toString(32)),n+=":"}else n=":"+n+"r"+(t=mr++).toString(32)+":";return e.memoizedState=n},unstable_isNewReconciler:!1},ns={readContext:Bi,useCallback:Hr,useContext:Bi,useEffect:Or,useImperativeHandle:Vr,useInsertionEffect:Mr,useLayoutEffect:Lr,useMemo:zr,useReducer:Cr,useRef:Rr,useState:function(){return Cr(jr)},useDebugValue:qr,useDeferredValue:function(e){return Ur(vr(),cr.memoizedState,e)},useTransition:function(){return[Cr(jr)[0],vr().memoizedState]},useMutableSource:Sr,useSyncExternalStore:kr,useId:Wr,unstable_isNewReconciler:!1},ts={readContext:Bi,useCallback:Hr,useContext:Bi,useEffect:Or,useImperativeHandle:Vr,useInsertionEffect:Mr,useLayoutEffect:Lr,useMemo:zr,useReducer:wr,useRef:Rr,useState:function(){return wr(jr)},useDebugValue:qr,useDeferredValue:function(e){var n=vr();return null===cr?n.memoizedState=e:Ur(n,cr.memoizedState,e)},useTransition:function(){return[wr(jr)[0],vr().memoizedState]},useMutableSource:Sr,useSyncExternalStore:kr,useId:Wr,unstable_isNewReconciler:!1};function os(e,n){if(e&&e.defaultProps){for(var t in n=O({},n),e=e.defaultProps)void 0===n[t]&&(n[t]=e[t]);return n}return n}function as(e,n,t,o){t=null===(t=t(o,n=e.memoizedState))||void 0===t?n:O({},n,t),e.memoizedState=t,0===e.lanes&&(e.updateQueue.baseState=t)}var is={isMounted:function(e){return!!(e=e._reactInternals)&&He(e)===e},enqueueSetState:function(e,n,t){e=e._reactInternals;var o=nc(),a=tc(e),i=Gi(o,a);i.payload=n,void 0!==t&&null!==t&&(i.callback=t),null!==(n=Vi(e,i,a))&&(oc(n,e,a,o),qi(n,e,a))},enqueueReplaceState:function(e,n,t){e=e._reactInternals;var o=nc(),a=tc(e),i=Gi(o,a);i.tag=1,i.payload=n,void 0!==t&&null!==t&&(i.callback=t),null!==(n=Vi(e,i,a))&&(oc(n,e,a,o),qi(n,e,a))},enqueueForceUpdate:function(e,n){e=e._reactInternals;var t=nc(),o=tc(e),a=Gi(t,o);a.tag=2,void 0!==n&&null!==n&&(a.callback=n),null!==(n=Vi(e,a,o))&&(oc(n,e,o,t),qi(n,e,o))}};function rs(e,n,t,o,a,i,r){return"function"===typeof(e=e.stateNode).shouldComponentUpdate?e.shouldComponentUpdate(o,i,r):!n.prototype||!n.prototype.isPureReactComponent||(!co(t,o)||!co(a,i))}function ss(e,n,t){var o=!1,a=Ia,i=n.contextType;return"object"===typeof i&&null!==i?i=Bi(i):(a=Fa(n)?Ba:Ea.current,i=(o=null!==(o=n.contextTypes)&&void 0!==o)?Ra(e,a):Ia),n=new n(t,i),e.memoizedState=null!==n.state&&void 0!==n.state?n.state:null,n.updater=is,e.stateNode=n,n._reactInternals=e,o&&((e=e.stateNode).__reactInternalMemoizedUnmaskedChildContext=a,e.__reactInternalMemoizedMaskedChildContext=i),n}function ls(e,n,t,o){e=n.state,"function"===typeof n.componentWillReceiveProps&&n.componentWillReceiveProps(t,o),"function"===typeof n.UNSAFE_componentWillReceiveProps&&n.UNSAFE_componentWillReceiveProps(t,o),n.state!==e&&is.enqueueReplaceState(n,n.state,null)}function cs(e,n,t,o){var a=e.stateNode;a.props=t,a.state=e.memoizedState,a.refs={},Mi(e);var i=n.contextType;"object"===typeof i&&null!==i?a.context=Bi(i):(i=Fa(n)?Ba:Ea.current,a.context=Ra(e,i)),a.state=e.memoizedState,"function"===typeof(i=n.getDerivedStateFromProps)&&(as(e,n,i,t),a.state=e.memoizedState),"function"===typeof n.getDerivedStateFromProps||"function"===typeof a.getSnapshotBeforeUpdate||"function"!==typeof a.UNSAFE_componentWillMount&&"function"!==typeof a.componentWillMount||(n=a.state,"function"===typeof a.componentWillMount&&a.componentWillMount(),"function"===typeof a.UNSAFE_componentWillMount&&a.UNSAFE_componentWillMount(),n!==a.state&&is.enqueueReplaceState(a,a.state,null),zi(e,t,a,o),a.state=e.memoizedState),"function"===typeof a.componentDidMount&&(e.flags|=4194308)}function ds(e,n){try{var t="",o=n;do{t+=V(o),o=o.return}while(o);var a=t}catch(i){a="\nError generating stack: "+i.message+"\n"+i.stack}return{value:e,source:n,stack:a,digest:null}}function us(e,n,t){return{value:e,source:null,stack:null!=t?t:null,digest:null!=n?n:null}}function hs(e,n){try{console.error(n.value)}catch(t){setTimeout((function(){throw t}))}}var ps="function"===typeof WeakMap?WeakMap:Map;function ms(e,n,t){(t=Gi(-1,t)).tag=3,t.payload={element:null};var o=n.value;return t.callback=function(){Ul||(Ul=!0,Ql=o),hs(0,n)},t}function fs(e,n,t){(t=Gi(-1,t)).tag=3;var o=e.type.getDerivedStateFromError;if("function"===typeof o){var a=n.value;t.payload=function(){return o(a)},t.callback=function(){hs(0,n)}}var i=e.stateNode;return null!==i&&"function"===typeof i.componentDidCatch&&(t.callback=function(){hs(0,n),"function"!==typeof o&&(null===Wl?Wl=new Set([this]):Wl.add(this));var e=n.stack;this.componentDidCatch(n.value,{componentStack:null!==e?e:""})}),t}function gs(e,n,t){var o=e.pingCache;if(null===o){o=e.pingCache=new ps;var a=new Set;o.set(n,a)}else void 0===(a=o.get(n))&&(a=new Set,o.set(n,a));a.has(t)||(a.add(t),e=Nc.bind(null,e,n,t),n.then(e,e))}function ys(e){do{var n;if((n=13===e.tag)&&(n=null===(n=e.memoizedState)||null!==n.dehydrated),n)return e;e=e.return}while(null!==e);return null}function xs(e,n,t,o,a){return 0===(1&e.mode)?(e===n?e.flags|=65536:(e.flags|=128,t.flags|=131072,t.flags&=-52805,1===t.tag&&(null===t.alternate?t.tag=17:((n=Gi(-1,1)).tag=2,Vi(t,n,1))),t.lanes|=1),e):(e.flags|=65536,e.lanes=a,e)}var bs=v.ReactCurrentOwner,vs=!1;function js(e,n,t,o){n.child=null===e?wi(n,null,t,o):Ci(n,e.child,t,o)}function Cs(e,n,t,o,a){t=t.render;var i=n.ref;return Ai(n,a),o=yr(e,n,t,o,i,a),t=xr(),null===e||vs?(ri&&t&&ti(n),n.flags|=1,js(e,n,o,a),n.child):(n.updateQueue=e.updateQueue,n.flags&=-2053,e.lanes&=~a,Us(e,n,a))}function ws(e,n,t,o,a){if(null===e){var i=t.type;return"function"!==typeof i||Rc(i)||void 0!==i.defaultProps||null!==t.compare||void 0!==t.defaultProps?((e=Pc(t.type,null,o,n,n.mode,a)).ref=n.ref,e.return=n,n.child=e):(n.tag=15,n.type=i,Ss(e,n,i,o,a))}if(i=e.child,0===(e.lanes&a)){var r=i.memoizedProps;if((t=null!==(t=t.compare)?t:co)(r,o)&&e.ref===n.ref)return Us(e,n,a)}return n.flags|=1,(e=Fc(i,o)).ref=n.ref,e.return=n,n.child=e}function Ss(e,n,t,o,a){if(null!==e){var i=e.memoizedProps;if(co(i,o)&&e.ref===n.ref){if(vs=!1,n.pendingProps=o=i,0===(e.lanes&a))return n.lanes=e.lanes,Us(e,n,a);0!==(131072&e.flags)&&(vs=!0)}}return _s(e,n,t,o,a)}function ks(e,n,t){var o=n.pendingProps,a=o.children,i=null!==e?e.memoizedState:null;if("hidden"===o.mode)if(0===(1&n.mode))n.memoizedState={baseLanes:0,cachePool:null,transitions:null},Ta(Fl,Rl),Rl|=t;else{if(0===(1073741824&t))return e=null!==i?i.baseLanes|t:t,n.lanes=n.childLanes=1073741824,n.memoizedState={baseLanes:e,cachePool:null,transitions:null},n.updateQueue=null,Ta(Fl,Rl),Rl|=e,null;n.memoizedState={baseLanes:0,cachePool:null,transitions:null},o=null!==i?i.baseLanes:t,Ta(Fl,Rl),Rl|=o}else null!==i?(o=i.baseLanes|t,n.memoizedState=null):o=t,Ta(Fl,Rl),Rl|=o;return js(e,n,a,t),n.child}function Ns(e,n){var t=n.ref;(null===e&&null!==t||null!==e&&e.ref!==t)&&(n.flags|=512,n.flags|=2097152)}function _s(e,n,t,o,a){var i=Fa(t)?Ba:Ea.current;return i=Ra(n,i),Ai(n,a),t=yr(e,n,t,o,i,a),o=xr(),null===e||vs?(ri&&o&&ti(n),n.flags|=1,js(e,n,t,a),n.child):(n.updateQueue=e.updateQueue,n.flags&=-2053,e.lanes&=~a,Us(e,n,a))}function Ts(e,n,t,o,a){if(Fa(t)){var i=!0;Ma(n)}else i=!1;if(Ai(n,a),null===n.stateNode)zs(e,n),ss(n,t,o),cs(n,t,o,a),o=!0;else if(null===e){var r=n.stateNode,s=n.memoizedProps;r.props=s;var l=r.context,c=t.contextType;"object"===typeof c&&null!==c?c=Bi(c):c=Ra(n,c=Fa(t)?Ba:Ea.current);var d=t.getDerivedStateFromProps,u="function"===typeof d||"function"===typeof r.getSnapshotBeforeUpdate;u||"function"!==typeof r.UNSAFE_componentWillReceiveProps&&"function"!==typeof r.componentWillReceiveProps||(s!==o||l!==c)&&ls(n,r,o,c),Oi=!1;var h=n.memoizedState;r.state=h,zi(n,o,r,a),l=n.memoizedState,s!==o||h!==l||Aa.current||Oi?("function"===typeof d&&(as(n,t,d,o),l=n.memoizedState),(s=Oi||rs(n,t,s,o,h,l,c))?(u||"function"!==typeof r.UNSAFE_componentWillMount&&"function"!==typeof r.componentWillMount||("function"===typeof r.componentWillMount&&r.componentWillMount(),"function"===typeof r.UNSAFE_componentWillMount&&r.UNSAFE_componentWillMount()),"function"===typeof r.componentDidMount&&(n.flags|=4194308)):("function"===typeof r.componentDidMount&&(n.flags|=4194308),n.memoizedProps=o,n.memoizedState=l),r.props=o,r.state=l,r.context=c,o=s):("function"===typeof r.componentDidMount&&(n.flags|=4194308),o=!1)}else{r=n.stateNode,Li(e,n),s=n.memoizedProps,c=n.type===n.elementType?s:os(n.type,s),r.props=c,u=n.pendingProps,h=r.context,"object"===typeof(l=t.contextType)&&null!==l?l=Bi(l):l=Ra(n,l=Fa(t)?Ba:Ea.current);var p=t.getDerivedStateFromProps;(d="function"===typeof p||"function"===typeof r.getSnapshotBeforeUpdate)||"function"!==typeof r.UNSAFE_componentWillReceiveProps&&"function"!==typeof r.componentWillReceiveProps||(s!==u||h!==l)&&ls(n,r,o,l),Oi=!1,h=n.memoizedState,r.state=h,zi(n,o,r,a);var m=n.memoizedState;s!==u||h!==m||Aa.current||Oi?("function"===typeof p&&(as(n,t,p,o),m=n.memoizedState),(c=Oi||rs(n,t,c,o,h,m,l)||!1)?(d||"function"!==typeof r.UNSAFE_componentWillUpdate&&"function"!==typeof r.componentWillUpdate||("function"===typeof r.componentWillUpdate&&r.componentWillUpdate(o,m,l),"function"===typeof r.UNSAFE_componentWillUpdate&&r.UNSAFE_componentWillUpdate(o,m,l)),"function"===typeof r.componentDidUpdate&&(n.flags|=4),"function"===typeof r.getSnapshotBeforeUpdate&&(n.flags|=1024)):("function"!==typeof r.componentDidUpdate||s===e.memoizedProps&&h===e.memoizedState||(n.flags|=4),"function"!==typeof r.getSnapshotBeforeUpdate||s===e.memoizedProps&&h===e.memoizedState||(n.flags|=1024),n.memoizedProps=o,n.memoizedState=m),r.props=o,r.state=m,r.context=l,o=c):("function"!==typeof r.componentDidUpdate||s===e.memoizedProps&&h===e.memoizedState||(n.flags|=4),"function"!==typeof r.getSnapshotBeforeUpdate||s===e.memoizedProps&&h===e.memoizedState||(n.flags|=1024),o=!1)}return Is(e,n,t,o,i,a)}function Is(e,n,t,o,a,i){Ns(e,n);var r=0!==(128&n.flags);if(!o&&!r)return a&&La(n,t,!1),Us(e,n,i);o=n.stateNode,bs.current=n;var s=r&&"function"!==typeof t.getDerivedStateFromError?null:o.render();return n.flags|=1,null!==e&&r?(n.child=Ci(n,e.child,null,i),n.child=Ci(n,null,s,i)):js(e,n,s,i),n.memoizedState=o.state,a&&La(n,t,!0),n.child}function Es(e){var n=e.stateNode;n.pendingContext?Da(0,n.pendingContext,n.pendingContext!==n.context):n.context&&Da(0,n.context,!1),Xi(e,n.containerInfo)}function As(e,n,t,o,a){return fi(),gi(a),n.flags|=256,js(e,n,t,o),n.child}var Bs,Rs,Fs,Ps,Ds={dehydrated:null,treeContext:null,retryLane:0};function Os(e){return{baseLanes:e,cachePool:null,transitions:null}}function Ms(e,n,t){var o,a=n.pendingProps,r=nr.current,s=!1,l=0!==(128&n.flags);if((o=l)||(o=(null===e||null!==e.memoizedState)&&0!==(2&r)),o?(s=!0,n.flags&=-129):null!==e&&null===e.memoizedState||(r|=1),Ta(nr,1&r),null===e)return ui(n),null!==(e=n.memoizedState)&&null!==(e=e.dehydrated)?(0===(1&n.mode)?n.lanes=1:"$!"===e.data?n.lanes=8:n.lanes=1073741824,null):(l=a.children,e=a.fallback,s?(a=n.mode,s=n.child,l={mode:"hidden",children:l},0===(1&a)&&null!==s?(s.childLanes=0,s.pendingProps=l):s=Oc(l,a,0,null),e=Dc(e,a,t,null),s.return=n,e.return=n,s.sibling=e,n.child=s,n.child.memoizedState=Os(t),n.memoizedState=Ds,e):Ls(n,l));if(null!==(r=e.memoizedState)&&null!==(o=r.dehydrated))return function(e,n,t,o,a,r,s){if(t)return 256&n.flags?(n.flags&=-257,Gs(e,n,s,o=us(Error(i(422))))):null!==n.memoizedState?(n.child=e.child,n.flags|=128,null):(r=o.fallback,a=n.mode,o=Oc({mode:"visible",children:o.children},a,0,null),(r=Dc(r,a,s,null)).flags|=2,o.return=n,r.return=n,o.sibling=r,n.child=o,0!==(1&n.mode)&&Ci(n,e.child,null,s),n.child.memoizedState=Os(s),n.memoizedState=Ds,r);if(0===(1&n.mode))return Gs(e,n,s,null);if("$!"===a.data){if(o=a.nextSibling&&a.nextSibling.dataset)var l=o.dgst;return o=l,Gs(e,n,s,o=us(r=Error(i(419)),o,void 0))}if(l=0!==(s&e.childLanes),vs||l){if(null!==(o=El)){switch(s&-s){case 4:a=2;break;case 16:a=8;break;case 64:case 128:case 256:case 512:case 1024:case 2048:case 4096:case 8192:case 16384:case 32768:case 65536:case 131072:case 262144:case 524288:case 1048576:case 2097152:case 4194304:case 8388608:case 16777216:case 33554432:case 67108864:a=32;break;case 536870912:a=268435456;break;default:a=0}0!==(a=0!==(a&(o.suspendedLanes|s))?0:a)&&a!==r.retryLane&&(r.retryLane=a,Di(e,a),oc(o,e,a,-1))}return gc(),Gs(e,n,s,o=us(Error(i(421))))}return"$?"===a.data?(n.flags|=128,n.child=e.child,n=Tc.bind(null,e),a._reactRetry=n,null):(e=r.treeContext,ii=ua(a.nextSibling),ai=n,ri=!0,si=null,null!==e&&($a[Ya++]=Za,$a[Ya++]=Ja,$a[Ya++]=Xa,Za=e.id,Ja=e.overflow,Xa=n),n=Ls(n,o.children),n.flags|=4096,n)}(e,n,l,a,o,r,t);if(s){s=a.fallback,l=n.mode,o=(r=e.child).sibling;var c={mode:"hidden",children:a.children};return 0===(1&l)&&n.child!==r?((a=n.child).childLanes=0,a.pendingProps=c,n.deletions=null):(a=Fc(r,c)).subtreeFlags=14680064&r.subtreeFlags,null!==o?s=Fc(o,s):(s=Dc(s,l,t,null)).flags|=2,s.return=n,a.return=n,a.sibling=s,n.child=a,a=s,s=n.child,l=null===(l=e.child.memoizedState)?Os(t):{baseLanes:l.baseLanes|t,cachePool:null,transitions:l.transitions},s.memoizedState=l,s.childLanes=e.childLanes&~t,n.memoizedState=Ds,a}return e=(s=e.child).sibling,a=Fc(s,{mode:"visible",children:a.children}),0===(1&n.mode)&&(a.lanes=t),a.return=n,a.sibling=null,null!==e&&(null===(t=n.deletions)?(n.deletions=[e],n.flags|=16):t.push(e)),n.child=a,n.memoizedState=null,a}function Ls(e,n){return(n=Oc({mode:"visible",children:n},e.mode,0,null)).return=e,e.child=n}function Gs(e,n,t,o){return null!==o&&gi(o),Ci(n,e.child,null,t),(e=Ls(n,n.pendingProps.children)).flags|=2,n.memoizedState=null,e}function Vs(e,n,t){e.lanes|=n;var o=e.alternate;null!==o&&(o.lanes|=n),Ei(e.return,n,t)}function qs(e,n,t,o,a){var i=e.memoizedState;null===i?e.memoizedState={isBackwards:n,rendering:null,renderingStartTime:0,last:o,tail:t,tailMode:a}:(i.isBackwards=n,i.rendering=null,i.renderingStartTime=0,i.last=o,i.tail=t,i.tailMode=a)}function Hs(e,n,t){var o=n.pendingProps,a=o.revealOrder,i=o.tail;if(js(e,n,o.children,t),0!==(2&(o=nr.current)))o=1&o|2,n.flags|=128;else{if(null!==e&&0!==(128&e.flags))e:for(e=n.child;null!==e;){if(13===e.tag)null!==e.memoizedState&&Vs(e,t,n);else if(19===e.tag)Vs(e,t,n);else if(null!==e.child){e.child.return=e,e=e.child;continue}if(e===n)break e;for(;null===e.sibling;){if(null===e.return||e.return===n)break e;e=e.return}e.sibling.return=e.return,e=e.sibling}o&=1}if(Ta(nr,o),0===(1&n.mode))n.memoizedState=null;else switch(a){case"forwards":for(t=n.child,a=null;null!==t;)null!==(e=t.alternate)&&null===tr(e)&&(a=t),t=t.sibling;null===(t=a)?(a=n.child,n.child=null):(a=t.sibling,t.sibling=null),qs(n,!1,a,t,i);break;case"backwards":for(t=null,a=n.child,n.child=null;null!==a;){if(null!==(e=a.alternate)&&null===tr(e)){n.child=a;break}e=a.sibling,a.sibling=t,t=a,a=e}qs(n,!0,t,null,i);break;case"together":qs(n,!1,null,null,void 0);break;default:n.memoizedState=null}return n.child}function zs(e,n){0===(1&n.mode)&&null!==e&&(e.alternate=null,n.alternate=null,n.flags|=2)}function Us(e,n,t){if(null!==e&&(n.dependencies=e.dependencies),Ol|=n.lanes,0===(t&n.childLanes))return null;if(null!==e&&n.child!==e.child)throw Error(i(153));if(null!==n.child){for(t=Fc(e=n.child,e.pendingProps),n.child=t,t.return=n;null!==e.sibling;)e=e.sibling,(t=t.sibling=Fc(e,e.pendingProps)).return=n;t.sibling=null}return n.child}function Qs(e,n){if(!ri)switch(e.tailMode){case"hidden":n=e.tail;for(var t=null;null!==n;)null!==n.alternate&&(t=n),n=n.sibling;null===t?e.tail=null:t.sibling=null;break;case"collapsed":t=e.tail;for(var o=null;null!==t;)null!==t.alternate&&(o=t),t=t.sibling;null===o?n||null===e.tail?e.tail=null:e.tail.sibling=null:o.sibling=null}}function Ws(e){var n=null!==e.alternate&&e.alternate.child===e.child,t=0,o=0;if(n)for(var a=e.child;null!==a;)t|=a.lanes|a.childLanes,o|=14680064&a.subtreeFlags,o|=14680064&a.flags,a.return=e,a=a.sibling;else for(a=e.child;null!==a;)t|=a.lanes|a.childLanes,o|=a.subtreeFlags,o|=a.flags,a.return=e,a=a.sibling;return e.subtreeFlags|=o,e.childLanes=t,n}function Ks(e,n,t){var o=n.pendingProps;switch(oi(n),n.tag){case 2:case 16:case 15:case 0:case 11:case 7:case 8:case 12:case 9:case 14:return Ws(n),null;case 1:case 17:return Fa(n.type)&&Pa(),Ws(n),null;case 3:return o=n.stateNode,Zi(),_a(Aa),_a(Ea),ar(),o.pendingContext&&(o.context=o.pendingContext,o.pendingContext=null),null!==e&&null!==e.child||(pi(n)?n.flags|=4:null===e||e.memoizedState.isDehydrated&&0===(256&n.flags)||(n.flags|=1024,null!==si&&(sc(si),si=null))),Rs(e,n),Ws(n),null;case 5:er(n);var a=Yi($i.current);if(t=n.type,null!==e&&null!=n.stateNode)Fs(e,n,t,o,a),e.ref!==n.ref&&(n.flags|=512,n.flags|=2097152);else{if(!o){if(null===n.stateNode)throw Error(i(166));return Ws(n),null}if(e=Yi(Wi.current),pi(n)){o=n.stateNode,t=n.type;var r=n.memoizedProps;switch(o[ma]=n,o[fa]=r,e=0!==(1&n.mode),t){case"dialog":Vo("cancel",o),Vo("close",o);break;case"iframe":case"object":case"embed":Vo("load",o);break;case"video":case"audio":for(a=0;a<Oo.length;a++)Vo(Oo[a],o);break;case"source":Vo("error",o);break;case"img":case"image":case"link":Vo("error",o),Vo("load",o);break;case"details":Vo("toggle",o);break;case"input":Y(o,r),Vo("invalid",o);break;case"select":o._wrapperState={wasMultiple:!!r.multiple},Vo("invalid",o);break;case"textarea":ae(o,r),Vo("invalid",o)}for(var l in xe(t,r),a=null,r)if(r.hasOwnProperty(l)){var c=r[l];"children"===l?"string"===typeof c?o.textContent!==c&&(!0!==r.suppressHydrationWarning&&ea(o.textContent,c,e),a=["children",c]):"number"===typeof c&&o.textContent!==""+c&&(!0!==r.suppressHydrationWarning&&ea(o.textContent,c,e),a=["children",""+c]):s.hasOwnProperty(l)&&null!=c&&"onScroll"===l&&Vo("scroll",o)}switch(t){case"input":Q(o),J(o,r,!0);break;case"textarea":Q(o),re(o);break;case"select":case"option":break;default:"function"===typeof r.onClick&&(o.onclick=na)}o=a,n.updateQueue=o,null!==o&&(n.flags|=4)}else{l=9===a.nodeType?a:a.ownerDocument,"http://www.w3.org/1999/xhtml"===e&&(e=se(t)),"http://www.w3.org/1999/xhtml"===e?"script"===t?((e=l.createElement("div")).innerHTML="<script><\/script>",e=e.removeChild(e.firstChild)):"string"===typeof o.is?e=l.createElement(t,{is:o.is}):(e=l.createElement(t),"select"===t&&(l=e,o.multiple?l.multiple=!0:o.size&&(l.size=o.size))):e=l.createElementNS(e,t),e[ma]=n,e[fa]=o,Bs(e,n,!1,!1),n.stateNode=e;e:{switch(l=be(t,o),t){case"dialog":Vo("cancel",e),Vo("close",e),a=o;break;case"iframe":case"object":case"embed":Vo("load",e),a=o;break;case"video":case"audio":for(a=0;a<Oo.length;a++)Vo(Oo[a],e);a=o;break;case"source":Vo("error",e),a=o;break;case"img":case"image":case"link":Vo("error",e),Vo("load",e),a=o;break;case"details":Vo("toggle",e),a=o;break;case"input":Y(e,o),a=$(e,o),Vo("invalid",e);break;case"option":default:a=o;break;case"select":e._wrapperState={wasMultiple:!!o.multiple},a=O({},o,{value:void 0}),Vo("invalid",e);break;case"textarea":ae(e,o),a=oe(e,o),Vo("invalid",e)}for(r in xe(t,a),c=a)if(c.hasOwnProperty(r)){var d=c[r];"style"===r?ge(e,d):"dangerouslySetInnerHTML"===r?null!=(d=d?d.__html:void 0)&&ue(e,d):"children"===r?"string"===typeof d?("textarea"!==t||""!==d)&&he(e,d):"number"===typeof d&&he(e,""+d):"suppressContentEditableWarning"!==r&&"suppressHydrationWarning"!==r&&"autoFocus"!==r&&(s.hasOwnProperty(r)?null!=d&&"onScroll"===r&&Vo("scroll",e):null!=d&&b(e,r,d,l))}switch(t){case"input":Q(e),J(e,o,!1);break;case"textarea":Q(e),re(e);break;case"option":null!=o.value&&e.setAttribute("value",""+z(o.value));break;case"select":e.multiple=!!o.multiple,null!=(r=o.value)?te(e,!!o.multiple,r,!1):null!=o.defaultValue&&te(e,!!o.multiple,o.defaultValue,!0);break;default:"function"===typeof a.onClick&&(e.onclick=na)}switch(t){case"button":case"input":case"select":case"textarea":o=!!o.autoFocus;break e;case"img":o=!0;break e;default:o=!1}}o&&(n.flags|=4)}null!==n.ref&&(n.flags|=512,n.flags|=2097152)}return Ws(n),null;case 6:if(e&&null!=n.stateNode)Ps(e,n,e.memoizedProps,o);else{if("string"!==typeof o&&null===n.stateNode)throw Error(i(166));if(t=Yi($i.current),Yi(Wi.current),pi(n)){if(o=n.stateNode,t=n.memoizedProps,o[ma]=n,(r=o.nodeValue!==t)&&null!==(e=ai))switch(e.tag){case 3:ea(o.nodeValue,t,0!==(1&e.mode));break;case 5:!0!==e.memoizedProps.suppressHydrationWarning&&ea(o.nodeValue,t,0!==(1&e.mode))}r&&(n.flags|=4)}else(o=(9===t.nodeType?t:t.ownerDocument).createTextNode(o))[ma]=n,n.stateNode=o}return Ws(n),null;case 13:if(_a(nr),o=n.memoizedState,null===e||null!==e.memoizedState&&null!==e.memoizedState.dehydrated){if(ri&&null!==ii&&0!==(1&n.mode)&&0===(128&n.flags))mi(),fi(),n.flags|=98560,r=!1;else if(r=pi(n),null!==o&&null!==o.dehydrated){if(null===e){if(!r)throw Error(i(318));if(!(r=null!==(r=n.memoizedState)?r.dehydrated:null))throw Error(i(317));r[ma]=n}else fi(),0===(128&n.flags)&&(n.memoizedState=null),n.flags|=4;Ws(n),r=!1}else null!==si&&(sc(si),si=null),r=!0;if(!r)return 65536&n.flags?n:null}return 0!==(128&n.flags)?(n.lanes=t,n):((o=null!==o)!==(null!==e&&null!==e.memoizedState)&&o&&(n.child.flags|=8192,0!==(1&n.mode)&&(null===e||0!==(1&nr.current)?0===Pl&&(Pl=3):gc())),null!==n.updateQueue&&(n.flags|=4),Ws(n),null);case 4:return Zi(),Rs(e,n),null===e&&zo(n.stateNode.containerInfo),Ws(n),null;case 10:return Ii(n.type._context),Ws(n),null;case 19:if(_a(nr),null===(r=n.memoizedState))return Ws(n),null;if(o=0!==(128&n.flags),null===(l=r.rendering))if(o)Qs(r,!1);else{if(0!==Pl||null!==e&&0!==(128&e.flags))for(e=n.child;null!==e;){if(null!==(l=tr(e))){for(n.flags|=128,Qs(r,!1),null!==(o=l.updateQueue)&&(n.updateQueue=o,n.flags|=4),n.subtreeFlags=0,o=t,t=n.child;null!==t;)e=o,(r=t).flags&=14680066,null===(l=r.alternate)?(r.childLanes=0,r.lanes=e,r.child=null,r.subtreeFlags=0,r.memoizedProps=null,r.memoizedState=null,r.updateQueue=null,r.dependencies=null,r.stateNode=null):(r.childLanes=l.childLanes,r.lanes=l.lanes,r.child=l.child,r.subtreeFlags=0,r.deletions=null,r.memoizedProps=l.memoizedProps,r.memoizedState=l.memoizedState,r.updateQueue=l.updateQueue,r.type=l.type,e=l.dependencies,r.dependencies=null===e?null:{lanes:e.lanes,firstContext:e.firstContext}),t=t.sibling;return Ta(nr,1&nr.current|2),n.child}e=e.sibling}null!==r.tail&&Ze()>Hl&&(n.flags|=128,o=!0,Qs(r,!1),n.lanes=4194304)}else{if(!o)if(null!==(e=tr(l))){if(n.flags|=128,o=!0,null!==(t=e.updateQueue)&&(n.updateQueue=t,n.flags|=4),Qs(r,!0),null===r.tail&&"hidden"===r.tailMode&&!l.alternate&&!ri)return Ws(n),null}else 2*Ze()-r.renderingStartTime>Hl&&1073741824!==t&&(n.flags|=128,o=!0,Qs(r,!1),n.lanes=4194304);r.isBackwards?(l.sibling=n.child,n.child=l):(null!==(t=r.last)?t.sibling=l:n.child=l,r.last=l)}return null!==r.tail?(n=r.tail,r.rendering=n,r.tail=n.sibling,r.renderingStartTime=Ze(),n.sibling=null,t=nr.current,Ta(nr,o?1&t|2:1&t),n):(Ws(n),null);case 22:case 23:return hc(),o=null!==n.memoizedState,null!==e&&null!==e.memoizedState!==o&&(n.flags|=8192),o&&0!==(1&n.mode)?0!==(1073741824&Rl)&&(Ws(n),6&n.subtreeFlags&&(n.flags|=8192)):Ws(n),null;case 24:case 25:return null}throw Error(i(156,n.tag))}function $s(e,n){switch(oi(n),n.tag){case 1:return Fa(n.type)&&Pa(),65536&(e=n.flags)?(n.flags=-65537&e|128,n):null;case 3:return Zi(),_a(Aa),_a(Ea),ar(),0!==(65536&(e=n.flags))&&0===(128&e)?(n.flags=-65537&e|128,n):null;case 5:return er(n),null;case 13:if(_a(nr),null!==(e=n.memoizedState)&&null!==e.dehydrated){if(null===n.alternate)throw Error(i(340));fi()}return 65536&(e=n.flags)?(n.flags=-65537&e|128,n):null;case 19:return _a(nr),null;case 4:return Zi(),null;case 10:return Ii(n.type._context),null;case 22:case 23:return hc(),null;default:return null}}Bs=function(e,n){for(var t=n.child;null!==t;){if(5===t.tag||6===t.tag)e.appendChild(t.stateNode);else if(4!==t.tag&&null!==t.child){t.child.return=t,t=t.child;continue}if(t===n)break;for(;null===t.sibling;){if(null===t.return||t.return===n)return;t=t.return}t.sibling.return=t.return,t=t.sibling}},Rs=function(){},Fs=function(e,n,t,o){var a=e.memoizedProps;if(a!==o){e=n.stateNode,Yi(Wi.current);var i,r=null;switch(t){case"input":a=$(e,a),o=$(e,o),r=[];break;case"select":a=O({},a,{value:void 0}),o=O({},o,{value:void 0}),r=[];break;case"textarea":a=oe(e,a),o=oe(e,o),r=[];break;default:"function"!==typeof a.onClick&&"function"===typeof o.onClick&&(e.onclick=na)}for(d in xe(t,o),t=null,a)if(!o.hasOwnProperty(d)&&a.hasOwnProperty(d)&&null!=a[d])if("style"===d){var l=a[d];for(i in l)l.hasOwnProperty(i)&&(t||(t={}),t[i]="")}else"dangerouslySetInnerHTML"!==d&&"children"!==d&&"suppressContentEditableWarning"!==d&&"suppressHydrationWarning"!==d&&"autoFocus"!==d&&(s.hasOwnProperty(d)?r||(r=[]):(r=r||[]).push(d,null));for(d in o){var c=o[d];if(l=null!=a?a[d]:void 0,o.hasOwnProperty(d)&&c!==l&&(null!=c||null!=l))if("style"===d)if(l){for(i in l)!l.hasOwnProperty(i)||c&&c.hasOwnProperty(i)||(t||(t={}),t[i]="");for(i in c)c.hasOwnProperty(i)&&l[i]!==c[i]&&(t||(t={}),t[i]=c[i])}else t||(r||(r=[]),r.push(d,t)),t=c;else"dangerouslySetInnerHTML"===d?(c=c?c.__html:void 0,l=l?l.__html:void 0,null!=c&&l!==c&&(r=r||[]).push(d,c)):"children"===d?"string"!==typeof c&&"number"!==typeof c||(r=r||[]).push(d,""+c):"suppressContentEditableWarning"!==d&&"suppressHydrationWarning"!==d&&(s.hasOwnProperty(d)?(null!=c&&"onScroll"===d&&Vo("scroll",e),r||l===c||(r=[])):(r=r||[]).push(d,c))}t&&(r=r||[]).push("style",t);var d=r;(n.updateQueue=d)&&(n.flags|=4)}},Ps=function(e,n,t,o){t!==o&&(n.flags|=4)};var Ys=!1,Xs=!1,Zs="function"===typeof WeakSet?WeakSet:Set,Js=null;function el(e,n){var t=e.ref;if(null!==t)if("function"===typeof t)try{t(null)}catch(o){kc(e,n,o)}else t.current=null}function nl(e,n,t){try{t()}catch(o){kc(e,n,o)}}var tl=!1;function ol(e,n,t){var o=n.updateQueue;if(null!==(o=null!==o?o.lastEffect:null)){var a=o=o.next;do{if((a.tag&e)===e){var i=a.destroy;a.destroy=void 0,void 0!==i&&nl(n,t,i)}a=a.next}while(a!==o)}}function al(e,n){if(null!==(n=null!==(n=n.updateQueue)?n.lastEffect:null)){var t=n=n.next;do{if((t.tag&e)===e){var o=t.create;t.destroy=o()}t=t.next}while(t!==n)}}function il(e){var n=e.ref;if(null!==n){var t=e.stateNode;e.tag,e=t,"function"===typeof n?n(e):n.current=e}}function rl(e){var n=e.alternate;null!==n&&(e.alternate=null,rl(n)),e.child=null,e.deletions=null,e.sibling=null,5===e.tag&&(null!==(n=e.stateNode)&&(delete n[ma],delete n[fa],delete n[ya],delete n[xa],delete n[ba])),e.stateNode=null,e.return=null,e.dependencies=null,e.memoizedProps=null,e.memoizedState=null,e.pendingProps=null,e.stateNode=null,e.updateQueue=null}function sl(e){return 5===e.tag||3===e.tag||4===e.tag}function ll(e){e:for(;;){for(;null===e.sibling;){if(null===e.return||sl(e.return))return null;e=e.return}for(e.sibling.return=e.return,e=e.sibling;5!==e.tag&&6!==e.tag&&18!==e.tag;){if(2&e.flags)continue e;if(null===e.child||4===e.tag)continue e;e.child.return=e,e=e.child}if(!(2&e.flags))return e.stateNode}}function cl(e,n,t){var o=e.tag;if(5===o||6===o)e=e.stateNode,n?8===t.nodeType?t.parentNode.insertBefore(e,n):t.insertBefore(e,n):(8===t.nodeType?(n=t.parentNode).insertBefore(e,t):(n=t).appendChild(e),null!==(t=t._reactRootContainer)&&void 0!==t||null!==n.onclick||(n.onclick=na));else if(4!==o&&null!==(e=e.child))for(cl(e,n,t),e=e.sibling;null!==e;)cl(e,n,t),e=e.sibling}function dl(e,n,t){var o=e.tag;if(5===o||6===o)e=e.stateNode,n?t.insertBefore(e,n):t.appendChild(e);else if(4!==o&&null!==(e=e.child))for(dl(e,n,t),e=e.sibling;null!==e;)dl(e,n,t),e=e.sibling}var ul=null,hl=!1;function pl(e,n,t){for(t=t.child;null!==t;)ml(e,n,t),t=t.sibling}function ml(e,n,t){if(sn&&"function"===typeof sn.onCommitFiberUnmount)try{sn.onCommitFiberUnmount(rn,t)}catch(s){}switch(t.tag){case 5:Xs||el(t,n);case 6:var o=ul,a=hl;ul=null,pl(e,n,t),hl=a,null!==(ul=o)&&(hl?(e=ul,t=t.stateNode,8===e.nodeType?e.parentNode.removeChild(t):e.removeChild(t)):ul.removeChild(t.stateNode));break;case 18:null!==ul&&(hl?(e=ul,t=t.stateNode,8===e.nodeType?da(e.parentNode,t):1===e.nodeType&&da(e,t),zn(e)):da(ul,t.stateNode));break;case 4:o=ul,a=hl,ul=t.stateNode.containerInfo,hl=!0,pl(e,n,t),ul=o,hl=a;break;case 0:case 11:case 14:case 15:if(!Xs&&(null!==(o=t.updateQueue)&&null!==(o=o.lastEffect))){a=o=o.next;do{var i=a,r=i.destroy;i=i.tag,void 0!==r&&(0!==(2&i)||0!==(4&i))&&nl(t,n,r),a=a.next}while(a!==o)}pl(e,n,t);break;case 1:if(!Xs&&(el(t,n),"function"===typeof(o=t.stateNode).componentWillUnmount))try{o.props=t.memoizedProps,o.state=t.memoizedState,o.componentWillUnmount()}catch(s){kc(t,n,s)}pl(e,n,t);break;case 21:pl(e,n,t);break;case 22:1&t.mode?(Xs=(o=Xs)||null!==t.memoizedState,pl(e,n,t),Xs=o):pl(e,n,t);break;default:pl(e,n,t)}}function fl(e){var n=e.updateQueue;if(null!==n){e.updateQueue=null;var t=e.stateNode;null===t&&(t=e.stateNode=new Zs),n.forEach((function(n){var o=Ic.bind(null,e,n);t.has(n)||(t.add(n),n.then(o,o))}))}}function gl(e,n){var t=n.deletions;if(null!==t)for(var o=0;o<t.length;o++){var a=t[o];try{var r=e,s=n,l=s;e:for(;null!==l;){switch(l.tag){case 5:ul=l.stateNode,hl=!1;break e;case 3:case 4:ul=l.stateNode.containerInfo,hl=!0;break e}l=l.return}if(null===ul)throw Error(i(160));ml(r,s,a),ul=null,hl=!1;var c=a.alternate;null!==c&&(c.return=null),a.return=null}catch(d){kc(a,n,d)}}if(12854&n.subtreeFlags)for(n=n.child;null!==n;)yl(n,e),n=n.sibling}function yl(e,n){var t=e.alternate,o=e.flags;switch(e.tag){case 0:case 11:case 14:case 15:if(gl(n,e),xl(e),4&o){try{ol(3,e,e.return),al(3,e)}catch(g){kc(e,e.return,g)}try{ol(5,e,e.return)}catch(g){kc(e,e.return,g)}}break;case 1:gl(n,e),xl(e),512&o&&null!==t&&el(t,t.return);break;case 5:if(gl(n,e),xl(e),512&o&&null!==t&&el(t,t.return),32&e.flags){var a=e.stateNode;try{he(a,"")}catch(g){kc(e,e.return,g)}}if(4&o&&null!=(a=e.stateNode)){var r=e.memoizedProps,s=null!==t?t.memoizedProps:r,l=e.type,c=e.updateQueue;if(e.updateQueue=null,null!==c)try{"input"===l&&"radio"===r.type&&null!=r.name&&X(a,r),be(l,s);var d=be(l,r);for(s=0;s<c.length;s+=2){var u=c[s],h=c[s+1];"style"===u?ge(a,h):"dangerouslySetInnerHTML"===u?ue(a,h):"children"===u?he(a,h):b(a,u,h,d)}switch(l){case"input":Z(a,r);break;case"textarea":ie(a,r);break;case"select":var p=a._wrapperState.wasMultiple;a._wrapperState.wasMultiple=!!r.multiple;var m=r.value;null!=m?te(a,!!r.multiple,m,!1):p!==!!r.multiple&&(null!=r.defaultValue?te(a,!!r.multiple,r.defaultValue,!0):te(a,!!r.multiple,r.multiple?[]:"",!1))}a[fa]=r}catch(g){kc(e,e.return,g)}}break;case 6:if(gl(n,e),xl(e),4&o){if(null===e.stateNode)throw Error(i(162));a=e.stateNode,r=e.memoizedProps;try{a.nodeValue=r}catch(g){kc(e,e.return,g)}}break;case 3:if(gl(n,e),xl(e),4&o&&null!==t&&t.memoizedState.isDehydrated)try{zn(n.containerInfo)}catch(g){kc(e,e.return,g)}break;case 4:default:gl(n,e),xl(e);break;case 13:gl(n,e),xl(e),8192&(a=e.child).flags&&(r=null!==a.memoizedState,a.stateNode.isHidden=r,!r||null!==a.alternate&&null!==a.alternate.memoizedState||(ql=Ze())),4&o&&fl(e);break;case 22:if(u=null!==t&&null!==t.memoizedState,1&e.mode?(Xs=(d=Xs)||u,gl(n,e),Xs=d):gl(n,e),xl(e),8192&o){if(d=null!==e.memoizedState,(e.stateNode.isHidden=d)&&!u&&0!==(1&e.mode))for(Js=e,u=e.child;null!==u;){for(h=Js=u;null!==Js;){switch(m=(p=Js).child,p.tag){case 0:case 11:case 14:case 15:ol(4,p,p.return);break;case 1:el(p,p.return);var f=p.stateNode;if("function"===typeof f.componentWillUnmount){o=p,t=p.return;try{n=o,f.props=n.memoizedProps,f.state=n.memoizedState,f.componentWillUnmount()}catch(g){kc(o,t,g)}}break;case 5:el(p,p.return);break;case 22:if(null!==p.memoizedState){Cl(h);continue}}null!==m?(m.return=p,Js=m):Cl(h)}u=u.sibling}e:for(u=null,h=e;;){if(5===h.tag){if(null===u){u=h;try{a=h.stateNode,d?"function"===typeof(r=a.style).setProperty?r.setProperty("display","none","important"):r.display="none":(l=h.stateNode,s=void 0!==(c=h.memoizedProps.style)&&null!==c&&c.hasOwnProperty("display")?c.display:null,l.style.display=fe("display",s))}catch(g){kc(e,e.return,g)}}}else if(6===h.tag){if(null===u)try{h.stateNode.nodeValue=d?"":h.memoizedProps}catch(g){kc(e,e.return,g)}}else if((22!==h.tag&&23!==h.tag||null===h.memoizedState||h===e)&&null!==h.child){h.child.return=h,h=h.child;continue}if(h===e)break e;for(;null===h.sibling;){if(null===h.return||h.return===e)break e;u===h&&(u=null),h=h.return}u===h&&(u=null),h.sibling.return=h.return,h=h.sibling}}break;case 19:gl(n,e),xl(e),4&o&&fl(e);case 21:}}function xl(e){var n=e.flags;if(2&n){try{e:{for(var t=e.return;null!==t;){if(sl(t)){var o=t;break e}t=t.return}throw Error(i(160))}switch(o.tag){case 5:var a=o.stateNode;32&o.flags&&(he(a,""),o.flags&=-33),dl(e,ll(e),a);break;case 3:case 4:var r=o.stateNode.containerInfo;cl(e,ll(e),r);break;default:throw Error(i(161))}}catch(s){kc(e,e.return,s)}e.flags&=-3}4096&n&&(e.flags&=-4097)}function bl(e,n,t){Js=e,vl(e,n,t)}function vl(e,n,t){for(var o=0!==(1&e.mode);null!==Js;){var a=Js,i=a.child;if(22===a.tag&&o){var r=null!==a.memoizedState||Ys;if(!r){var s=a.alternate,l=null!==s&&null!==s.memoizedState||Xs;s=Ys;var c=Xs;if(Ys=r,(Xs=l)&&!c)for(Js=a;null!==Js;)l=(r=Js).child,22===r.tag&&null!==r.memoizedState?wl(a):null!==l?(l.return=r,Js=l):wl(a);for(;null!==i;)Js=i,vl(i,n,t),i=i.sibling;Js=a,Ys=s,Xs=c}jl(e)}else 0!==(8772&a.subtreeFlags)&&null!==i?(i.return=a,Js=i):jl(e)}}function jl(e){for(;null!==Js;){var n=Js;if(0!==(8772&n.flags)){var t=n.alternate;try{if(0!==(8772&n.flags))switch(n.tag){case 0:case 11:case 15:Xs||al(5,n);break;case 1:var o=n.stateNode;if(4&n.flags&&!Xs)if(null===t)o.componentDidMount();else{var a=n.elementType===n.type?t.memoizedProps:os(n.type,t.memoizedProps);o.componentDidUpdate(a,t.memoizedState,o.__reactInternalSnapshotBeforeUpdate)}var r=n.updateQueue;null!==r&&Ui(n,r,o);break;case 3:var s=n.updateQueue;if(null!==s){if(t=null,null!==n.child)switch(n.child.tag){case 5:case 1:t=n.child.stateNode}Ui(n,s,t)}break;case 5:var l=n.stateNode;if(null===t&&4&n.flags){t=l;var c=n.memoizedProps;switch(n.type){case"button":case"input":case"select":case"textarea":c.autoFocus&&t.focus();break;case"img":c.src&&(t.src=c.src)}}break;case 6:case 4:case 12:case 19:case 17:case 21:case 22:case 23:case 25:break;case 13:if(null===n.memoizedState){var d=n.alternate;if(null!==d){var u=d.memoizedState;if(null!==u){var h=u.dehydrated;null!==h&&zn(h)}}}break;default:throw Error(i(163))}Xs||512&n.flags&&il(n)}catch(Te){kc(n,n.return,Te)}}if(n===e){Js=null;break}if(null!==(t=n.sibling)){t.return=n.return,Js=t;break}Js=n.return}}function Cl(e){for(;null!==Js;){var n=Js;if(n===e){Js=null;break}var t=n.sibling;if(null!==t){t.return=n.return,Js=t;break}Js=n.return}}function wl(e){for(;null!==Js;){var n=Js;try{switch(n.tag){case 0:case 11:case 15:var t=n.return;try{al(4,n)}catch(l){kc(n,t,l)}break;case 1:var o=n.stateNode;if("function"===typeof o.componentDidMount){var a=n.return;try{o.componentDidMount()}catch(l){kc(n,a,l)}}var i=n.return;try{il(n)}catch(l){kc(n,i,l)}break;case 5:var r=n.return;try{il(n)}catch(l){kc(n,r,l)}}}catch(l){kc(n,n.return,l)}if(n===e){Js=null;break}var s=n.sibling;if(null!==s){s.return=n.return,Js=s;break}Js=n.return}}var Sl,kl=Math.ceil,Nl=v.ReactCurrentDispatcher,_l=v.ReactCurrentOwner,Tl=v.ReactCurrentBatchConfig,Il=0,El=null,Al=null,Bl=0,Rl=0,Fl=Na(0),Pl=0,Dl=null,Ol=0,Ml=0,Ll=0,Gl=null,Vl=null,ql=0,Hl=1/0,zl=null,Ul=!1,Ql=null,Wl=null,Kl=!1,$l=null,Yl=0,Xl=0,Zl=null,Jl=-1,ec=0;function nc(){return 0!==(6&Il)?Ze():-1!==Jl?Jl:Jl=Ze()}function tc(e){return 0===(1&e.mode)?1:0!==(2&Il)&&0!==Bl?Bl&-Bl:null!==yi.transition?(0===ec&&(ec=yn()),ec):0!==(e=jn)?e:e=void 0===(e=window.event)?16:Zn(e.type)}function oc(e,n,t,o){if(50<Xl)throw Xl=0,Zl=null,Error(i(185));bn(e,t,o),0!==(2&Il)&&e===El||(e===El&&(0===(2&Il)&&(Ml|=t),4===Pl&&lc(e,Bl)),ac(e,o),1===t&&0===Il&&0===(1&n.mode)&&(Hl=Ze()+500,Va&&za()))}function ac(e,n){var t=e.callbackNode;!function(e,n){for(var t=e.suspendedLanes,o=e.pingedLanes,a=e.expirationTimes,i=e.pendingLanes;0<i;){var r=31-ln(i),s=1<<r,l=a[r];-1===l?0!==(s&t)&&0===(s&o)||(a[r]=fn(s,n)):l<=n&&(e.expiredLanes|=s),i&=~s}}(e,n);var o=mn(e,e===El?Bl:0);if(0===o)null!==t&&$e(t),e.callbackNode=null,e.callbackPriority=0;else if(n=o&-o,e.callbackPriority!==n){if(null!=t&&$e(t),1===n)0===e.tag?function(e){Va=!0,Ha(e)}(cc.bind(null,e)):Ha(cc.bind(null,e)),la((function(){0===(6&Il)&&za()})),t=null;else{switch(Cn(o)){case 1:t=en;break;case 4:t=nn;break;case 16:default:t=tn;break;case 536870912:t=an}t=Ec(t,ic.bind(null,e))}e.callbackPriority=n,e.callbackNode=t}}function ic(e,n){if(Jl=-1,ec=0,0!==(6&Il))throw Error(i(327));var t=e.callbackNode;if(wc()&&e.callbackNode!==t)return null;var o=mn(e,e===El?Bl:0);if(0===o)return null;if(0!==(30&o)||0!==(o&e.expiredLanes)||n)n=yc(e,o);else{n=o;var a=Il;Il|=2;var r=fc();for(El===e&&Bl===n||(zl=null,Hl=Ze()+500,pc(e,n));;)try{bc();break}catch(l){mc(e,l)}Ti(),Nl.current=r,Il=a,null!==Al?n=0:(El=null,Bl=0,n=Pl)}if(0!==n){if(2===n&&(0!==(a=gn(e))&&(o=a,n=rc(e,a))),1===n)throw t=Dl,pc(e,0),lc(e,o),ac(e,Ze()),t;if(6===n)lc(e,o);else{if(a=e.current.alternate,0===(30&o)&&!function(e){for(var n=e;;){if(16384&n.flags){var t=n.updateQueue;if(null!==t&&null!==(t=t.stores))for(var o=0;o<t.length;o++){var a=t[o],i=a.getSnapshot;a=a.value;try{if(!lo(i(),a))return!1}catch(s){return!1}}}if(t=n.child,16384&n.subtreeFlags&&null!==t)t.return=n,n=t;else{if(n===e)break;for(;null===n.sibling;){if(null===n.return||n.return===e)return!0;n=n.return}n.sibling.return=n.return,n=n.sibling}}return!0}(a)&&(2===(n=yc(e,o))&&(0!==(r=gn(e))&&(o=r,n=rc(e,r))),1===n))throw t=Dl,pc(e,0),lc(e,o),ac(e,Ze()),t;switch(e.finishedWork=a,e.finishedLanes=o,n){case 0:case 1:throw Error(i(345));case 2:case 5:Cc(e,Vl,zl);break;case 3:if(lc(e,o),(130023424&o)===o&&10<(n=ql+500-Ze())){if(0!==mn(e,0))break;if(((a=e.suspendedLanes)&o)!==o){nc(),e.pingedLanes|=e.suspendedLanes&a;break}e.timeoutHandle=ia(Cc.bind(null,e,Vl,zl),n);break}Cc(e,Vl,zl);break;case 4:if(lc(e,o),(4194240&o)===o)break;for(n=e.eventTimes,a=-1;0<o;){var s=31-ln(o);r=1<<s,(s=n[s])>a&&(a=s),o&=~r}if(o=a,10<(o=(120>(o=Ze()-o)?120:480>o?480:1080>o?1080:1920>o?1920:3e3>o?3e3:4320>o?4320:1960*kl(o/1960))-o)){e.timeoutHandle=ia(Cc.bind(null,e,Vl,zl),o);break}Cc(e,Vl,zl);break;default:throw Error(i(329))}}}return ac(e,Ze()),e.callbackNode===t?ic.bind(null,e):null}function rc(e,n){var t=Gl;return e.current.memoizedState.isDehydrated&&(pc(e,n).flags|=256),2!==(e=yc(e,n))&&(n=Vl,Vl=t,null!==n&&sc(n)),e}function sc(e){null===Vl?Vl=e:Vl.push.apply(Vl,e)}function lc(e,n){for(n&=~Ll,n&=~Ml,e.suspendedLanes|=n,e.pingedLanes&=~n,e=e.expirationTimes;0<n;){var t=31-ln(n),o=1<<t;e[t]=-1,n&=~o}}function cc(e){if(0!==(6&Il))throw Error(i(327));wc();var n=mn(e,0);if(0===(1&n))return ac(e,Ze()),null;var t=yc(e,n);if(0!==e.tag&&2===t){var o=gn(e);0!==o&&(n=o,t=rc(e,o))}if(1===t)throw t=Dl,pc(e,0),lc(e,n),ac(e,Ze()),t;if(6===t)throw Error(i(345));return e.finishedWork=e.current.alternate,e.finishedLanes=n,Cc(e,Vl,zl),ac(e,Ze()),null}function dc(e,n){var t=Il;Il|=1;try{return e(n)}finally{0===(Il=t)&&(Hl=Ze()+500,Va&&za())}}function uc(e){null!==$l&&0===$l.tag&&0===(6&Il)&&wc();var n=Il;Il|=1;var t=Tl.transition,o=jn;try{if(Tl.transition=null,jn=1,e)return e()}finally{jn=o,Tl.transition=t,0===(6&(Il=n))&&za()}}function hc(){Rl=Fl.current,_a(Fl)}function pc(e,n){e.finishedWork=null,e.finishedLanes=0;var t=e.timeoutHandle;if(-1!==t&&(e.timeoutHandle=-1,ra(t)),null!==Al)for(t=Al.return;null!==t;){var o=t;switch(oi(o),o.tag){case 1:null!==(o=o.type.childContextTypes)&&void 0!==o&&Pa();break;case 3:Zi(),_a(Aa),_a(Ea),ar();break;case 5:er(o);break;case 4:Zi();break;case 13:case 19:_a(nr);break;case 10:Ii(o.type._context);break;case 22:case 23:hc()}t=t.return}if(El=e,Al=e=Fc(e.current,null),Bl=Rl=n,Pl=0,Dl=null,Ll=Ml=Ol=0,Vl=Gl=null,null!==Ri){for(n=0;n<Ri.length;n++)if(null!==(o=(t=Ri[n]).interleaved)){t.interleaved=null;var a=o.next,i=t.pending;if(null!==i){var r=i.next;i.next=a,o.next=r}t.pending=o}Ri=null}return e}function mc(e,n){for(;;){var t=Al;try{if(Ti(),ir.current=Jr,ur){for(var o=lr.memoizedState;null!==o;){var a=o.queue;null!==a&&(a.pending=null),o=o.next}ur=!1}if(sr=0,dr=cr=lr=null,hr=!1,pr=0,_l.current=null,null===t||null===t.return){Pl=1,Dl=n,Al=null;break}e:{var r=e,s=t.return,l=t,c=n;if(n=Bl,l.flags|=32768,null!==c&&"object"===typeof c&&"function"===typeof c.then){var d=c,u=l,h=u.tag;if(0===(1&u.mode)&&(0===h||11===h||15===h)){var p=u.alternate;p?(u.updateQueue=p.updateQueue,u.memoizedState=p.memoizedState,u.lanes=p.lanes):(u.updateQueue=null,u.memoizedState=null)}var m=ys(s);if(null!==m){m.flags&=-257,xs(m,s,l,0,n),1&m.mode&&gs(r,d,n),c=d;var f=(n=m).updateQueue;if(null===f){var g=new Set;g.add(c),n.updateQueue=g}else f.add(c);break e}if(0===(1&n)){gs(r,d,n),gc();break e}c=Error(i(426))}else if(ri&&1&l.mode){var y=ys(s);if(null!==y){0===(65536&y.flags)&&(y.flags|=256),xs(y,s,l,0,n),gi(ds(c,l));break e}}r=c=ds(c,l),4!==Pl&&(Pl=2),null===Gl?Gl=[r]:Gl.push(r),r=s;do{switch(r.tag){case 3:r.flags|=65536,n&=-n,r.lanes|=n,Hi(r,ms(0,c,n));break e;case 1:l=c;var x=r.type,b=r.stateNode;if(0===(128&r.flags)&&("function"===typeof x.getDerivedStateFromError||null!==b&&"function"===typeof b.componentDidCatch&&(null===Wl||!Wl.has(b)))){r.flags|=65536,n&=-n,r.lanes|=n,Hi(r,fs(r,l,n));break e}}r=r.return}while(null!==r)}jc(t)}catch(v){n=v,Al===t&&null!==t&&(Al=t=t.return);continue}break}}function fc(){var e=Nl.current;return Nl.current=Jr,null===e?Jr:e}function gc(){0!==Pl&&3!==Pl&&2!==Pl||(Pl=4),null===El||0===(268435455&Ol)&&0===(268435455&Ml)||lc(El,Bl)}function yc(e,n){var t=Il;Il|=2;var o=fc();for(El===e&&Bl===n||(zl=null,pc(e,n));;)try{xc();break}catch(a){mc(e,a)}if(Ti(),Il=t,Nl.current=o,null!==Al)throw Error(i(261));return El=null,Bl=0,Pl}function xc(){for(;null!==Al;)vc(Al)}function bc(){for(;null!==Al&&!Ye();)vc(Al)}function vc(e){var n=Sl(e.alternate,e,Rl);e.memoizedProps=e.pendingProps,null===n?jc(e):Al=n,_l.current=null}function jc(e){var n=e;do{var t=n.alternate;if(e=n.return,0===(32768&n.flags)){if(null!==(t=Ks(t,n,Rl)))return void(Al=t)}else{if(null!==(t=$s(t,n)))return t.flags&=32767,void(Al=t);if(null===e)return Pl=6,void(Al=null);e.flags|=32768,e.subtreeFlags=0,e.deletions=null}if(null!==(n=n.sibling))return void(Al=n);Al=n=e}while(null!==n);0===Pl&&(Pl=5)}function Cc(e,n,t){var o=jn,a=Tl.transition;try{Tl.transition=null,jn=1,function(e,n,t,o){do{wc()}while(null!==$l);if(0!==(6&Il))throw Error(i(327));t=e.finishedWork;var a=e.finishedLanes;if(null===t)return null;if(e.finishedWork=null,e.finishedLanes=0,t===e.current)throw Error(i(177));e.callbackNode=null,e.callbackPriority=0;var r=t.lanes|t.childLanes;if(function(e,n){var t=e.pendingLanes&~n;e.pendingLanes=n,e.suspendedLanes=0,e.pingedLanes=0,e.expiredLanes&=n,e.mutableReadLanes&=n,e.entangledLanes&=n,n=e.entanglements;var o=e.eventTimes;for(e=e.expirationTimes;0<t;){var a=31-ln(t),i=1<<a;n[a]=0,o[a]=-1,e[a]=-1,t&=~i}}(e,r),e===El&&(Al=El=null,Bl=0),0===(2064&t.subtreeFlags)&&0===(2064&t.flags)||Kl||(Kl=!0,Ec(tn,(function(){return wc(),null}))),r=0!==(15990&t.flags),0!==(15990&t.subtreeFlags)||r){r=Tl.transition,Tl.transition=null;var s=jn;jn=1;var l=Il;Il|=4,_l.current=null,function(e,n){if(ta=Qn,fo(e=mo())){if("selectionStart"in e)var t={start:e.selectionStart,end:e.selectionEnd};else e:{var o=(t=(t=e.ownerDocument)&&t.defaultView||window).getSelection&&t.getSelection();if(o&&0!==o.rangeCount){t=o.anchorNode;var a=o.anchorOffset,r=o.focusNode;o=o.focusOffset;try{t.nodeType,r.nodeType}catch(j){t=null;break e}var s=0,l=-1,c=-1,d=0,u=0,h=e,p=null;n:for(;;){for(var m;h!==t||0!==a&&3!==h.nodeType||(l=s+a),h!==r||0!==o&&3!==h.nodeType||(c=s+o),3===h.nodeType&&(s+=h.nodeValue.length),null!==(m=h.firstChild);)p=h,h=m;for(;;){if(h===e)break n;if(p===t&&++d===a&&(l=s),p===r&&++u===o&&(c=s),null!==(m=h.nextSibling))break;p=(h=p).parentNode}h=m}t=-1===l||-1===c?null:{start:l,end:c}}else t=null}t=t||{start:0,end:0}}else t=null;for(oa={focusedElem:e,selectionRange:t},Qn=!1,Js=n;null!==Js;)if(e=(n=Js).child,0!==(1028&n.subtreeFlags)&&null!==e)e.return=n,Js=e;else for(;null!==Js;){n=Js;try{var f=n.alternate;if(0!==(1024&n.flags))switch(n.tag){case 0:case 11:case 15:case 5:case 6:case 4:case 17:break;case 1:if(null!==f){var g=f.memoizedProps,y=f.memoizedState,x=n.stateNode,b=x.getSnapshotBeforeUpdate(n.elementType===n.type?g:os(n.type,g),y);x.__reactInternalSnapshotBeforeUpdate=b}break;case 3:var v=n.stateNode.containerInfo;1===v.nodeType?v.textContent="":9===v.nodeType&&v.documentElement&&v.removeChild(v.documentElement);break;default:throw Error(i(163))}}catch(j){kc(n,n.return,j)}if(null!==(e=n.sibling)){e.return=n.return,Js=e;break}Js=n.return}f=tl,tl=!1}(e,t),yl(t,e),go(oa),Qn=!!ta,oa=ta=null,e.current=t,bl(t,e,a),Xe(),Il=l,jn=s,Tl.transition=r}else e.current=t;if(Kl&&(Kl=!1,$l=e,Yl=a),r=e.pendingLanes,0===r&&(Wl=null),function(e){if(sn&&"function"===typeof sn.onCommitFiberRoot)try{sn.onCommitFiberRoot(rn,e,void 0,128===(128&e.current.flags))}catch(n){}}(t.stateNode),ac(e,Ze()),null!==n)for(o=e.onRecoverableError,t=0;t<n.length;t++)a=n[t],o(a.value,{componentStack:a.stack,digest:a.digest});if(Ul)throw Ul=!1,e=Ql,Ql=null,e;0!==(1&Yl)&&0!==e.tag&&wc(),r=e.pendingLanes,0!==(1&r)?e===Zl?Xl++:(Xl=0,Zl=e):Xl=0,za()}(e,n,t,o)}finally{Tl.transition=a,jn=o}return null}function wc(){if(null!==$l){var e=Cn(Yl),n=Tl.transition,t=jn;try{if(Tl.transition=null,jn=16>e?16:e,null===$l)var o=!1;else{if(e=$l,$l=null,Yl=0,0!==(6&Il))throw Error(i(331));var a=Il;for(Il|=4,Js=e.current;null!==Js;){var r=Js,s=r.child;if(0!==(16&Js.flags)){var l=r.deletions;if(null!==l){for(var c=0;c<l.length;c++){var d=l[c];for(Js=d;null!==Js;){var u=Js;switch(u.tag){case 0:case 11:case 15:ol(8,u,r)}var h=u.child;if(null!==h)h.return=u,Js=h;else for(;null!==Js;){var p=(u=Js).sibling,m=u.return;if(rl(u),u===d){Js=null;break}if(null!==p){p.return=m,Js=p;break}Js=m}}}var f=r.alternate;if(null!==f){var g=f.child;if(null!==g){f.child=null;do{var y=g.sibling;g.sibling=null,g=y}while(null!==g)}}Js=r}}if(0!==(2064&r.subtreeFlags)&&null!==s)s.return=r,Js=s;else e:for(;null!==Js;){if(0!==(2048&(r=Js).flags))switch(r.tag){case 0:case 11:case 15:ol(9,r,r.return)}var x=r.sibling;if(null!==x){x.return=r.return,Js=x;break e}Js=r.return}}var b=e.current;for(Js=b;null!==Js;){var v=(s=Js).child;if(0!==(2064&s.subtreeFlags)&&null!==v)v.return=s,Js=v;else e:for(s=b;null!==Js;){if(0!==(2048&(l=Js).flags))try{switch(l.tag){case 0:case 11:case 15:al(9,l)}}catch(C){kc(l,l.return,C)}if(l===s){Js=null;break e}var j=l.sibling;if(null!==j){j.return=l.return,Js=j;break e}Js=l.return}}if(Il=a,za(),sn&&"function"===typeof sn.onPostCommitFiberRoot)try{sn.onPostCommitFiberRoot(rn,e)}catch(C){}o=!0}return o}finally{jn=t,Tl.transition=n}}return!1}function Sc(e,n,t){e=Vi(e,n=ms(0,n=ds(t,n),1),1),n=nc(),null!==e&&(bn(e,1,n),ac(e,n))}function kc(e,n,t){if(3===e.tag)Sc(e,e,t);else for(;null!==n;){if(3===n.tag){Sc(n,e,t);break}if(1===n.tag){var o=n.stateNode;if("function"===typeof n.type.getDerivedStateFromError||"function"===typeof o.componentDidCatch&&(null===Wl||!Wl.has(o))){n=Vi(n,e=fs(n,e=ds(t,e),1),1),e=nc(),null!==n&&(bn(n,1,e),ac(n,e));break}}n=n.return}}function Nc(e,n,t){var o=e.pingCache;null!==o&&o.delete(n),n=nc(),e.pingedLanes|=e.suspendedLanes&t,El===e&&(Bl&t)===t&&(4===Pl||3===Pl&&(130023424&Bl)===Bl&&500>Ze()-ql?pc(e,0):Ll|=t),ac(e,n)}function _c(e,n){0===n&&(0===(1&e.mode)?n=1:(n=hn,0===(130023424&(hn<<=1))&&(hn=4194304)));var t=nc();null!==(e=Di(e,n))&&(bn(e,n,t),ac(e,t))}function Tc(e){var n=e.memoizedState,t=0;null!==n&&(t=n.retryLane),_c(e,t)}function Ic(e,n){var t=0;switch(e.tag){case 13:var o=e.stateNode,a=e.memoizedState;null!==a&&(t=a.retryLane);break;case 19:o=e.stateNode;break;default:throw Error(i(314))}null!==o&&o.delete(n),_c(e,t)}function Ec(e,n){return Ke(e,n)}function Ac(e,n,t,o){this.tag=e,this.key=t,this.sibling=this.child=this.return=this.stateNode=this.type=this.elementType=null,this.index=0,this.ref=null,this.pendingProps=n,this.dependencies=this.memoizedState=this.updateQueue=this.memoizedProps=null,this.mode=o,this.subtreeFlags=this.flags=0,this.deletions=null,this.childLanes=this.lanes=0,this.alternate=null}function Bc(e,n,t,o){return new Ac(e,n,t,o)}function Rc(e){return!(!(e=e.prototype)||!e.isReactComponent)}function Fc(e,n){var t=e.alternate;return null===t?((t=Bc(e.tag,n,e.key,e.mode)).elementType=e.elementType,t.type=e.type,t.stateNode=e.stateNode,t.alternate=e,e.alternate=t):(t.pendingProps=n,t.type=e.type,t.flags=0,t.subtreeFlags=0,t.deletions=null),t.flags=14680064&e.flags,t.childLanes=e.childLanes,t.lanes=e.lanes,t.child=e.child,t.memoizedProps=e.memoizedProps,t.memoizedState=e.memoizedState,t.updateQueue=e.updateQueue,n=e.dependencies,t.dependencies=null===n?null:{lanes:n.lanes,firstContext:n.firstContext},t.sibling=e.sibling,t.index=e.index,t.ref=e.ref,t}function Pc(e,n,t,o,a,r){var s=2;if(o=e,"function"===typeof e)Rc(e)&&(s=1);else if("string"===typeof e)s=5;else e:switch(e){case w:return Dc(t.children,a,r,n);case S:s=8,a|=8;break;case k:return(e=Bc(12,t,n,2|a)).elementType=k,e.lanes=r,e;case I:return(e=Bc(13,t,n,a)).elementType=I,e.lanes=r,e;case E:return(e=Bc(19,t,n,a)).elementType=E,e.lanes=r,e;case R:return Oc(t,a,r,n);default:if("object"===typeof e&&null!==e)switch(e.$$typeof){case N:s=10;break e;case _:s=9;break e;case T:s=11;break e;case A:s=14;break e;case B:s=16,o=null;break e}throw Error(i(130,null==e?e:typeof e,""))}return(n=Bc(s,t,n,a)).elementType=e,n.type=o,n.lanes=r,n}function Dc(e,n,t,o){return(e=Bc(7,e,o,n)).lanes=t,e}function Oc(e,n,t,o){return(e=Bc(22,e,o,n)).elementType=R,e.lanes=t,e.stateNode={isHidden:!1},e}function Mc(e,n,t){return(e=Bc(6,e,null,n)).lanes=t,e}function Lc(e,n,t){return(n=Bc(4,null!==e.children?e.children:[],e.key,n)).lanes=t,n.stateNode={containerInfo:e.containerInfo,pendingChildren:null,implementation:e.implementation},n}function Gc(e,n,t,o,a){this.tag=n,this.containerInfo=e,this.finishedWork=this.pingCache=this.current=this.pendingChildren=null,this.timeoutHandle=-1,this.callbackNode=this.pendingContext=this.context=null,this.callbackPriority=0,this.eventTimes=xn(0),this.expirationTimes=xn(-1),this.entangledLanes=this.finishedLanes=this.mutableReadLanes=this.expiredLanes=this.pingedLanes=this.suspendedLanes=this.pendingLanes=0,this.entanglements=xn(0),this.identifierPrefix=o,this.onRecoverableError=a,this.mutableSourceEagerHydrationData=null}function Vc(e,n,t,o,a,i,r,s,l){return e=new Gc(e,n,t,s,l),1===n?(n=1,!0===i&&(n|=8)):n=0,i=Bc(3,null,null,n),e.current=i,i.stateNode=e,i.memoizedState={element:o,isDehydrated:t,cache:null,transitions:null,pendingSuspenseBoundaries:null},Mi(i),e}function qc(e){if(!e)return Ia;e:{if(He(e=e._reactInternals)!==e||1!==e.tag)throw Error(i(170));var n=e;do{switch(n.tag){case 3:n=n.stateNode.context;break e;case 1:if(Fa(n.type)){n=n.stateNode.__reactInternalMemoizedMergedChildContext;break e}}n=n.return}while(null!==n);throw Error(i(171))}if(1===e.tag){var t=e.type;if(Fa(t))return Oa(e,t,n)}return n}function Hc(e,n,t,o,a,i,r,s,l){return(e=Vc(t,o,!0,e,0,i,0,s,l)).context=qc(null),t=e.current,(i=Gi(o=nc(),a=tc(t))).callback=void 0!==n&&null!==n?n:null,Vi(t,i,a),e.current.lanes=a,bn(e,a,o),ac(e,o),e}function zc(e,n,t,o){var a=n.current,i=nc(),r=tc(a);return t=qc(t),null===n.context?n.context=t:n.pendingContext=t,(n=Gi(i,r)).payload={element:e},null!==(o=void 0===o?null:o)&&(n.callback=o),null!==(e=Vi(a,n,r))&&(oc(e,a,r,i),qi(e,a,r)),r}function Uc(e){return(e=e.current).child?(e.child.tag,e.child.stateNode):null}function Qc(e,n){if(null!==(e=e.memoizedState)&&null!==e.dehydrated){var t=e.retryLane;e.retryLane=0!==t&&t<n?t:n}}function Wc(e,n){Qc(e,n),(e=e.alternate)&&Qc(e,n)}Sl=function(e,n,t){if(null!==e)if(e.memoizedProps!==n.pendingProps||Aa.current)vs=!0;else{if(0===(e.lanes&t)&&0===(128&n.flags))return vs=!1,function(e,n,t){switch(n.tag){case 3:Es(n),fi();break;case 5:Ji(n);break;case 1:Fa(n.type)&&Ma(n);break;case 4:Xi(n,n.stateNode.containerInfo);break;case 10:var o=n.type._context,a=n.memoizedProps.value;Ta(Si,o._currentValue),o._currentValue=a;break;case 13:if(null!==(o=n.memoizedState))return null!==o.dehydrated?(Ta(nr,1&nr.current),n.flags|=128,null):0!==(t&n.child.childLanes)?Ms(e,n,t):(Ta(nr,1&nr.current),null!==(e=Us(e,n,t))?e.sibling:null);Ta(nr,1&nr.current);break;case 19:if(o=0!==(t&n.childLanes),0!==(128&e.flags)){if(o)return Hs(e,n,t);n.flags|=128}if(null!==(a=n.memoizedState)&&(a.rendering=null,a.tail=null,a.lastEffect=null),Ta(nr,nr.current),o)break;return null;case 22:case 23:return n.lanes=0,ks(e,n,t)}return Us(e,n,t)}(e,n,t);vs=0!==(131072&e.flags)}else vs=!1,ri&&0!==(1048576&n.flags)&&ni(n,Ka,n.index);switch(n.lanes=0,n.tag){case 2:var o=n.type;zs(e,n),e=n.pendingProps;var a=Ra(n,Ea.current);Ai(n,t),a=yr(null,n,o,e,a,t);var r=xr();return n.flags|=1,"object"===typeof a&&null!==a&&"function"===typeof a.render&&void 0===a.$$typeof?(n.tag=1,n.memoizedState=null,n.updateQueue=null,Fa(o)?(r=!0,Ma(n)):r=!1,n.memoizedState=null!==a.state&&void 0!==a.state?a.state:null,Mi(n),a.updater=is,n.stateNode=a,a._reactInternals=n,cs(n,o,e,t),n=Is(null,n,o,!0,r,t)):(n.tag=0,ri&&r&&ti(n),js(null,n,a,t),n=n.child),n;case 16:o=n.elementType;e:{switch(zs(e,n),e=n.pendingProps,o=(a=o._init)(o._payload),n.type=o,a=n.tag=function(e){if("function"===typeof e)return Rc(e)?1:0;if(void 0!==e&&null!==e){if((e=e.$$typeof)===T)return 11;if(e===A)return 14}return 2}(o),e=os(o,e),a){case 0:n=_s(null,n,o,e,t);break e;case 1:n=Ts(null,n,o,e,t);break e;case 11:n=Cs(null,n,o,e,t);break e;case 14:n=ws(null,n,o,os(o.type,e),t);break e}throw Error(i(306,o,""))}return n;case 0:return o=n.type,a=n.pendingProps,_s(e,n,o,a=n.elementType===o?a:os(o,a),t);case 1:return o=n.type,a=n.pendingProps,Ts(e,n,o,a=n.elementType===o?a:os(o,a),t);case 3:e:{if(Es(n),null===e)throw Error(i(387));o=n.pendingProps,a=(r=n.memoizedState).element,Li(e,n),zi(n,o,null,t);var s=n.memoizedState;if(o=s.element,r.isDehydrated){if(r={element:o,isDehydrated:!1,cache:s.cache,pendingSuspenseBoundaries:s.pendingSuspenseBoundaries,transitions:s.transitions},n.updateQueue.baseState=r,n.memoizedState=r,256&n.flags){n=As(e,n,o,t,a=ds(Error(i(423)),n));break e}if(o!==a){n=As(e,n,o,t,a=ds(Error(i(424)),n));break e}for(ii=ua(n.stateNode.containerInfo.firstChild),ai=n,ri=!0,si=null,t=wi(n,null,o,t),n.child=t;t;)t.flags=-3&t.flags|4096,t=t.sibling}else{if(fi(),o===a){n=Us(e,n,t);break e}js(e,n,o,t)}n=n.child}return n;case 5:return Ji(n),null===e&&ui(n),o=n.type,a=n.pendingProps,r=null!==e?e.memoizedProps:null,s=a.children,aa(o,a)?s=null:null!==r&&aa(o,r)&&(n.flags|=32),Ns(e,n),js(e,n,s,t),n.child;case 6:return null===e&&ui(n),null;case 13:return Ms(e,n,t);case 4:return Xi(n,n.stateNode.containerInfo),o=n.pendingProps,null===e?n.child=Ci(n,null,o,t):js(e,n,o,t),n.child;case 11:return o=n.type,a=n.pendingProps,Cs(e,n,o,a=n.elementType===o?a:os(o,a),t);case 7:return js(e,n,n.pendingProps,t),n.child;case 8:case 12:return js(e,n,n.pendingProps.children,t),n.child;case 10:e:{if(o=n.type._context,a=n.pendingProps,r=n.memoizedProps,s=a.value,Ta(Si,o._currentValue),o._currentValue=s,null!==r)if(lo(r.value,s)){if(r.children===a.children&&!Aa.current){n=Us(e,n,t);break e}}else for(null!==(r=n.child)&&(r.return=n);null!==r;){var l=r.dependencies;if(null!==l){s=r.child;for(var c=l.firstContext;null!==c;){if(c.context===o){if(1===r.tag){(c=Gi(-1,t&-t)).tag=2;var d=r.updateQueue;if(null!==d){var u=(d=d.shared).pending;null===u?c.next=c:(c.next=u.next,u.next=c),d.pending=c}}r.lanes|=t,null!==(c=r.alternate)&&(c.lanes|=t),Ei(r.return,t,n),l.lanes|=t;break}c=c.next}}else if(10===r.tag)s=r.type===n.type?null:r.child;else if(18===r.tag){if(null===(s=r.return))throw Error(i(341));s.lanes|=t,null!==(l=s.alternate)&&(l.lanes|=t),Ei(s,t,n),s=r.sibling}else s=r.child;if(null!==s)s.return=r;else for(s=r;null!==s;){if(s===n){s=null;break}if(null!==(r=s.sibling)){r.return=s.return,s=r;break}s=s.return}r=s}js(e,n,a.children,t),n=n.child}return n;case 9:return a=n.type,o=n.pendingProps.children,Ai(n,t),o=o(a=Bi(a)),n.flags|=1,js(e,n,o,t),n.child;case 14:return a=os(o=n.type,n.pendingProps),ws(e,n,o,a=os(o.type,a),t);case 15:return Ss(e,n,n.type,n.pendingProps,t);case 17:return o=n.type,a=n.pendingProps,a=n.elementType===o?a:os(o,a),zs(e,n),n.tag=1,Fa(o)?(e=!0,Ma(n)):e=!1,Ai(n,t),ss(n,o,a),cs(n,o,a,t),Is(null,n,o,!0,e,t);case 19:return Hs(e,n,t);case 22:return ks(e,n,t)}throw Error(i(156,n.tag))};var Kc="function"===typeof reportError?reportError:function(e){console.error(e)};function $c(e){this._internalRoot=e}function Yc(e){this._internalRoot=e}function Xc(e){return!(!e||1!==e.nodeType&&9!==e.nodeType&&11!==e.nodeType)}function Zc(e){return!(!e||1!==e.nodeType&&9!==e.nodeType&&11!==e.nodeType&&(8!==e.nodeType||" react-mount-point-unstable "!==e.nodeValue))}function Jc(){}function ed(e,n,t,o,a){var i=t._reactRootContainer;if(i){var r=i;if("function"===typeof a){var s=a;a=function(){var e=Uc(r);s.call(e)}}zc(n,r,e,a)}else r=function(e,n,t,o,a){if(a){if("function"===typeof o){var i=o;o=function(){var e=Uc(r);i.call(e)}}var r=Hc(n,o,e,0,null,!1,0,"",Jc);return e._reactRootContainer=r,e[ga]=r.current,zo(8===e.nodeType?e.parentNode:e),uc(),r}for(;a=e.lastChild;)e.removeChild(a);if("function"===typeof o){var s=o;o=function(){var e=Uc(l);s.call(e)}}var l=Vc(e,0,!1,null,0,!1,0,"",Jc);return e._reactRootContainer=l,e[ga]=l.current,zo(8===e.nodeType?e.parentNode:e),uc((function(){zc(n,l,t,o)})),l}(t,n,e,a,o);return Uc(r)}Yc.prototype.render=$c.prototype.render=function(e){var n=this._internalRoot;if(null===n)throw Error(i(409));zc(e,n,null,null)},Yc.prototype.unmount=$c.prototype.unmount=function(){var e=this._internalRoot;if(null!==e){this._internalRoot=null;var n=e.containerInfo;uc((function(){zc(null,e,null,null)})),n[ga]=null}},Yc.prototype.unstable_scheduleHydration=function(e){if(e){var n=Nn();e={blockedOn:null,target:e,priority:n};for(var t=0;t<Pn.length&&0!==n&&n<Pn[t].priority;t++);Pn.splice(t,0,e),0===t&&Ln(e)}},wn=function(e){switch(e.tag){case 3:var n=e.stateNode;if(n.current.memoizedState.isDehydrated){var t=pn(n.pendingLanes);0!==t&&(vn(n,1|t),ac(n,Ze()),0===(6&Il)&&(Hl=Ze()+500,za()))}break;case 13:uc((function(){var n=Di(e,1);if(null!==n){var t=nc();oc(n,e,1,t)}})),Wc(e,1)}},Sn=function(e){if(13===e.tag){var n=Di(e,134217728);if(null!==n)oc(n,e,134217728,nc());Wc(e,134217728)}},kn=function(e){if(13===e.tag){var n=tc(e),t=Di(e,n);if(null!==t)oc(t,e,n,nc());Wc(e,n)}},Nn=function(){return jn},_n=function(e,n){var t=jn;try{return jn=e,n()}finally{jn=t}},Ce=function(e,n,t){switch(n){case"input":if(Z(e,t),n=t.name,"radio"===t.type&&null!=n){for(t=e;t.parentNode;)t=t.parentNode;for(t=t.querySelectorAll("input[name="+JSON.stringify(""+n)+'][type="radio"]'),n=0;n<t.length;n++){var o=t[n];if(o!==e&&o.form===e.form){var a=wa(o);if(!a)throw Error(i(90));W(o),Z(o,a)}}}break;case"textarea":ie(e,t);break;case"select":null!=(n=t.value)&&te(e,!!t.multiple,n,!1)}},Ie=dc,Ee=uc;var nd={usingClientEntryPoint:!1,Events:[ja,Ca,wa,Ne,_e,dc]},td={findFiberByHostInstance:va,bundleType:0,version:"18.3.1",rendererPackageName:"react-dom"},od={bundleType:td.bundleType,version:td.version,rendererPackageName:td.rendererPackageName,rendererConfig:td.rendererConfig,overrideHookState:null,overrideHookStateDeletePath:null,overrideHookStateRenamePath:null,overrideProps:null,overridePropsDeletePath:null,overridePropsRenamePath:null,setErrorHandler:null,setSuspenseHandler:null,scheduleUpdate:null,currentDispatcherRef:v.ReactCurrentDispatcher,findHostInstanceByFiber:function(e){return null===(e=Qe(e))?null:e.stateNode},findFiberByHostInstance:td.findFiberByHostInstance||function(){return null},findHostInstancesForRefresh:null,scheduleRefresh:null,scheduleRoot:null,setRefreshHandler:null,getCurrentFiber:null,reconcilerVersion:"18.3.1-next-f1338f8080-20240426"};if("undefined"!==typeof __REACT_DEVTOOLS_GLOBAL_HOOK__){var ad=__REACT_DEVTOOLS_GLOBAL_HOOK__;if(!ad.isDisabled&&ad.supportsFiber)try{rn=ad.inject(od),sn=ad}catch(de){}}n.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED=nd,n.createPortal=function(e,n){var t=2<arguments.length&&void 0!==arguments[2]?arguments[2]:null;if(!Xc(n))throw Error(i(200));return function(e,n,t){var o=3<arguments.length&&void 0!==arguments[3]?arguments[3]:null;return{$$typeof:C,key:null==o?null:""+o,children:e,containerInfo:n,implementation:t}}(e,n,null,t)},n.createRoot=function(e,n){if(!Xc(e))throw Error(i(299));var t=!1,o="",a=Kc;return null!==n&&void 0!==n&&(!0===n.unstable_strictMode&&(t=!0),void 0!==n.identifierPrefix&&(o=n.identifierPrefix),void 0!==n.onRecoverableError&&(a=n.onRecoverableError)),n=Vc(e,1,!1,null,0,t,0,o,a),e[ga]=n.current,zo(8===e.nodeType?e.parentNode:e),new $c(n)},n.findDOMNode=function(e){if(null==e)return null;if(1===e.nodeType)return e;var n=e._reactInternals;if(void 0===n){if("function"===typeof e.render)throw Error(i(188));throw e=Object.keys(e).join(","),Error(i(268,e))}return e=null===(e=Qe(n))?null:e.stateNode},n.flushSync=function(e){return uc(e)},n.hydrate=function(e,n,t){if(!Zc(n))throw Error(i(200));return ed(null,e,n,!0,t)},n.hydrateRoot=function(e,n,t){if(!Xc(e))throw Error(i(405));var o=null!=t&&t.hydratedSources||null,a=!1,r="",s=Kc;if(null!==t&&void 0!==t&&(!0===t.unstable_strictMode&&(a=!0),void 0!==t.identifierPrefix&&(r=t.identifierPrefix),void 0!==t.onRecoverableError&&(s=t.onRecoverableError)),n=Hc(n,null,e,1,null!=t?t:null,a,0,r,s),e[ga]=n.current,zo(e),o)for(e=0;e<o.length;e++)a=(a=(t=o[e])._getVersion)(t._source),null==n.mutableSourceEagerHydrationData?n.mutableSourceEagerHydrationData=[t,a]:n.mutableSourceEagerHydrationData.push(t,a);return new Yc(n)},n.render=function(e,n,t){if(!Zc(n))throw Error(i(200));return ed(null,e,n,!1,t)},n.unmountComponentAtNode=function(e){if(!Zc(e))throw Error(i(40));return!!e._reactRootContainer&&(uc((function(){ed(null,null,e,!1,(function(){e._reactRootContainer=null,e[ga]=null}))})),!0)},n.unstable_batchedUpdates=dc,n.unstable_renderSubtreeIntoContainer=function(e,n,t,o){if(!Zc(t))throw Error(i(200));if(null==e||void 0===e._reactInternals)throw Error(i(38));return ed(e,n,t,!1,o)},n.version="18.3.1-next-f1338f8080-20240426"},896:(e,n,t)=>{var o=t(407);n.createRoot=o.createRoot,n.hydrateRoot=o.hydrateRoot},407:(e,n,t)=>{!function e(){if("undefined"!==typeof __REACT_DEVTOOLS_GLOBAL_HOOK__&&"function"===typeof __REACT_DEVTOOLS_GLOBAL_HOOK__.checkDCE)try{__REACT_DEVTOOLS_GLOBAL_HOOK__.checkDCE(e)}catch(n){console.error(n)}}(),e.exports=t(401)},113:(e,n)=>{var t="function"===typeof Symbol&&Symbol.for,o=t?Symbol.for("react.element"):60103,a=t?Symbol.for("react.portal"):60106,i=t?Symbol.for("react.fragment"):60107,r=t?Symbol.for("react.strict_mode"):60108,s=t?Symbol.for("react.profiler"):60114,l=t?Symbol.for("react.provider"):60109,c=t?Symbol.for("react.context"):60110,d=t?Symbol.for("react.async_mode"):60111,u=t?Symbol.for("react.concurrent_mode"):60111,h=t?Symbol.for("react.forward_ref"):60112,p=t?Symbol.for("react.suspense"):60113,m=t?Symbol.for("react.suspense_list"):60120,f=t?Symbol.for("react.memo"):60115,g=t?Symbol.for("react.lazy"):60116,y=t?Symbol.for("react.block"):60121,x=t?Symbol.for("react.fundamental"):60117,b=t?Symbol.for("react.responder"):60118,v=t?Symbol.for("react.scope"):60119;function j(e){if("object"===typeof e&&null!==e){var n=e.$$typeof;switch(n){case o:switch(e=e.type){case d:case u:case i:case s:case r:case p:return e;default:switch(e=e&&e.$$typeof){case c:case h:case g:case f:case l:return e;default:return n}}case a:return n}}}function C(e){return j(e)===u}n.AsyncMode=d,n.ConcurrentMode=u,n.ContextConsumer=c,n.ContextProvider=l,n.Element=o,n.ForwardRef=h,n.Fragment=i,n.Lazy=g,n.Memo=f,n.Portal=a,n.Profiler=s,n.StrictMode=r,n.Suspense=p,n.isAsyncMode=function(e){return C(e)||j(e)===d},n.isConcurrentMode=C,n.isContextConsumer=function(e){return j(e)===c},n.isContextProvider=function(e){return j(e)===l},n.isElement=function(e){return"object"===typeof e&&null!==e&&e.$$typeof===o},n.isForwardRef=function(e){return j(e)===h},n.isFragment=function(e){return j(e)===i},n.isLazy=function(e){return j(e)===g},n.isMemo=function(e){return j(e)===f},n.isPortal=function(e){return j(e)===a},n.isProfiler=function(e){return j(e)===s},n.isStrictMode=function(e){return j(e)===r},n.isSuspense=function(e){return j(e)===p},n.isValidElementType=function(e){return"string"===typeof e||"function"===typeof e||e===i||e===u||e===s||e===r||e===p||e===m||"object"===typeof e&&null!==e&&(e.$$typeof===g||e.$$typeof===f||e.$$typeof===l||e.$$typeof===c||e.$$typeof===h||e.$$typeof===x||e.$$typeof===b||e.$$typeof===v||e.$$typeof===y)},n.typeOf=j},277:(e,n,t)=>{e.exports=t(113)},617:(e,n,t)=>{var o=t(998),a=Symbol.for("react.element"),i=Symbol.for("react.fragment"),r=Object.prototype.hasOwnProperty,s=o.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner,l={key:!0,ref:!0,__self:!0,__source:!0};function c(e,n,t){var o,i={},c=null,d=null;for(o in void 0!==t&&(c=""+t),void 0!==n.key&&(c=""+n.key),void 0!==n.ref&&(d=n.ref),n)r.call(n,o)&&!l.hasOwnProperty(o)&&(i[o]=n[o]);if(e&&e.defaultProps)for(o in n=e.defaultProps)void 0===i[o]&&(i[o]=n[o]);return{$$typeof:a,type:e,key:c,ref:d,props:i,_owner:s.current}}n.Fragment=i,n.jsx=c,n.jsxs=c},121:(e,n)=>{var t=Symbol.for("react.element"),o=Symbol.for("react.portal"),a=Symbol.for("react.fragment"),i=Symbol.for("react.strict_mode"),r=Symbol.for("react.profiler"),s=Symbol.for("react.provider"),l=Symbol.for("react.context"),c=Symbol.for("react.forward_ref"),d=Symbol.for("react.suspense"),u=Symbol.for("react.memo"),h=Symbol.for("react.lazy"),p=Symbol.iterator;var m={isMounted:function(){return!1},enqueueForceUpdate:function(){},enqueueReplaceState:function(){},enqueueSetState:function(){}},f=Object.assign,g={};function y(e,n,t){this.props=e,this.context=n,this.refs=g,this.updater=t||m}function x(){}function b(e,n,t){this.props=e,this.context=n,this.refs=g,this.updater=t||m}y.prototype.isReactComponent={},y.prototype.setState=function(e,n){if("object"!==typeof e&&"function"!==typeof e&&null!=e)throw Error("setState(...): takes an object of state variables to update or a function which returns an object of state variables.");this.updater.enqueueSetState(this,e,n,"setState")},y.prototype.forceUpdate=function(e){this.updater.enqueueForceUpdate(this,e,"forceUpdate")},x.prototype=y.prototype;var v=b.prototype=new x;v.constructor=b,f(v,y.prototype),v.isPureReactComponent=!0;var j=Array.isArray,C=Object.prototype.hasOwnProperty,w={current:null},S={key:!0,ref:!0,__self:!0,__source:!0};function k(e,n,o){var a,i={},r=null,s=null;if(null!=n)for(a in void 0!==n.ref&&(s=n.ref),void 0!==n.key&&(r=""+n.key),n)C.call(n,a)&&!S.hasOwnProperty(a)&&(i[a]=n[a]);var l=arguments.length-2;if(1===l)i.children=o;else if(1<l){for(var c=Array(l),d=0;d<l;d++)c[d]=arguments[d+2];i.children=c}if(e&&e.defaultProps)for(a in l=e.defaultProps)void 0===i[a]&&(i[a]=l[a]);return{$$typeof:t,type:e,key:r,ref:s,props:i,_owner:w.current}}function N(e){return"object"===typeof e&&null!==e&&e.$$typeof===t}var _=/\/+/g;function T(e,n){return"object"===typeof e&&null!==e&&null!=e.key?function(e){var n={"=":"=0",":":"=2"};return"$"+e.replace(/[=:]/g,(function(e){return n[e]}))}(""+e.key):n.toString(36)}function I(e,n,a,i,r){var s=typeof e;"undefined"!==s&&"boolean"!==s||(e=null);var l=!1;if(null===e)l=!0;else switch(s){case"string":case"number":l=!0;break;case"object":switch(e.$$typeof){case t:case o:l=!0}}if(l)return r=r(l=e),e=""===i?"."+T(l,0):i,j(r)?(a="",null!=e&&(a=e.replace(_,"$&/")+"/"),I(r,n,a,"",(function(e){return e}))):null!=r&&(N(r)&&(r=function(e,n){return{$$typeof:t,type:e.type,key:n,ref:e.ref,props:e.props,_owner:e._owner}}(r,a+(!r.key||l&&l.key===r.key?"":(""+r.key).replace(_,"$&/")+"/")+e)),n.push(r)),1;if(l=0,i=""===i?".":i+":",j(e))for(var c=0;c<e.length;c++){var d=i+T(s=e[c],c);l+=I(s,n,a,d,r)}else if(d=function(e){return null===e||"object"!==typeof e?null:"function"===typeof(e=p&&e[p]||e["@@iterator"])?e:null}(e),"function"===typeof d)for(e=d.call(e),c=0;!(s=e.next()).done;)l+=I(s=s.value,n,a,d=i+T(s,c++),r);else if("object"===s)throw n=String(e),Error("Objects are not valid as a React child (found: "+("[object Object]"===n?"object with keys {"+Object.keys(e).join(", ")+"}":n)+"). If you meant to render a collection of children, use an array instead.");return l}function E(e,n,t){if(null==e)return e;var o=[],a=0;return I(e,o,"","",(function(e){return n.call(t,e,a++)})),o}function A(e){if(-1===e._status){var n=e._result;(n=n()).then((function(n){0!==e._status&&-1!==e._status||(e._status=1,e._result=n)}),(function(n){0!==e._status&&-1!==e._status||(e._status=2,e._result=n)})),-1===e._status&&(e._status=0,e._result=n)}if(1===e._status)return e._result.default;throw e._result}var B={current:null},R={transition:null},F={ReactCurrentDispatcher:B,ReactCurrentBatchConfig:R,ReactCurrentOwner:w};function P(){throw Error("act(...) is not supported in production builds of React.")}n.Children={map:E,forEach:function(e,n,t){E(e,(function(){n.apply(this,arguments)}),t)},count:function(e){var n=0;return E(e,(function(){n++})),n},toArray:function(e){return E(e,(function(e){return e}))||[]},only:function(e){if(!N(e))throw Error("React.Children.only expected to receive a single React element child.");return e}},n.Component=y,n.Fragment=a,n.Profiler=r,n.PureComponent=b,n.StrictMode=i,n.Suspense=d,n.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED=F,n.act=P,n.cloneElement=function(e,n,o){if(null===e||void 0===e)throw Error("React.cloneElement(...): The argument must be a React element, but you passed "+e+".");var a=f({},e.props),i=e.key,r=e.ref,s=e._owner;if(null!=n){if(void 0!==n.ref&&(r=n.ref,s=w.current),void 0!==n.key&&(i=""+n.key),e.type&&e.type.defaultProps)var l=e.type.defaultProps;for(c in n)C.call(n,c)&&!S.hasOwnProperty(c)&&(a[c]=void 0===n[c]&&void 0!==l?l[c]:n[c])}var c=arguments.length-2;if(1===c)a.children=o;else if(1<c){l=Array(c);for(var d=0;d<c;d++)l[d]=arguments[d+2];a.children=l}return{$$typeof:t,type:e.type,key:i,ref:r,props:a,_owner:s}},n.createContext=function(e){return(e={$$typeof:l,_currentValue:e,_currentValue2:e,_threadCount:0,Provider:null,Consumer:null,_defaultValue:null,_globalName:null}).Provider={$$typeof:s,_context:e},e.Consumer=e},n.createElement=k,n.createFactory=function(e){var n=k.bind(null,e);return n.type=e,n},n.createRef=function(){return{current:null}},n.forwardRef=function(e){return{$$typeof:c,render:e}},n.isValidElement=N,n.lazy=function(e){return{$$typeof:h,_payload:{_status:-1,_result:e},_init:A}},n.memo=function(e,n){return{$$typeof:u,type:e,compare:void 0===n?null:n}},n.startTransition=function(e){var n=R.transition;R.transition={};try{e()}finally{R.transition=n}},n.unstable_act=P,n.useCallback=function(e,n){return B.current.useCallback(e,n)},n.useContext=function(e){return B.current.useContext(e)},n.useDebugValue=function(){},n.useDeferredValue=function(e){return B.current.useDeferredValue(e)},n.useEffect=function(e,n){return B.current.useEffect(e,n)},n.useId=function(){return B.current.useId()},n.useImperativeHandle=function(e,n,t){return B.current.useImperativeHandle(e,n,t)},n.useInsertionEffect=function(e,n){return B.current.useInsertionEffect(e,n)},n.useLayoutEffect=function(e,n){return B.current.useLayoutEffect(e,n)},n.useMemo=function(e,n){return B.current.useMemo(e,n)},n.useReducer=function(e,n,t){return B.current.useReducer(e,n,t)},n.useRef=function(e){return B.current.useRef(e)},n.useState=function(e){return B.current.useState(e)},n.useSyncExternalStore=function(e,n,t){return B.current.useSyncExternalStore(e,n,t)},n.useTransition=function(){return B.current.useTransition()},n.version="18.3.1"},998:(e,n,t)=>{e.exports=t(121)},446:(e,n,t)=>{e.exports=t(617)},345:(e,n)=>{function t(e,n){var t=e.length;e.push(n);e:for(;0<t;){var o=t-1>>>1,a=e[o];if(!(0<i(a,n)))break e;e[o]=n,e[t]=a,t=o}}function o(e){return 0===e.length?null:e[0]}function a(e){if(0===e.length)return null;var n=e[0],t=e.pop();if(t!==n){e[0]=t;e:for(var o=0,a=e.length,r=a>>>1;o<r;){var s=2*(o+1)-1,l=e[s],c=s+1,d=e[c];if(0>i(l,t))c<a&&0>i(d,l)?(e[o]=d,e[c]=t,o=c):(e[o]=l,e[s]=t,o=s);else{if(!(c<a&&0>i(d,t)))break e;e[o]=d,e[c]=t,o=c}}}return n}function i(e,n){var t=e.sortIndex-n.sortIndex;return 0!==t?t:e.id-n.id}if("object"===typeof performance&&"function"===typeof performance.now){var r=performance;n.unstable_now=function(){return r.now()}}else{var s=Date,l=s.now();n.unstable_now=function(){return s.now()-l}}var c=[],d=[],u=1,h=null,p=3,m=!1,f=!1,g=!1,y="function"===typeof setTimeout?setTimeout:null,x="function"===typeof clearTimeout?clearTimeout:null,b="undefined"!==typeof setImmediate?setImmediate:null;function v(e){for(var n=o(d);null!==n;){if(null===n.callback)a(d);else{if(!(n.startTime<=e))break;a(d),n.sortIndex=n.expirationTime,t(c,n)}n=o(d)}}function j(e){if(g=!1,v(e),!f)if(null!==o(c))f=!0,R(C);else{var n=o(d);null!==n&&F(j,n.startTime-e)}}function C(e,t){f=!1,g&&(g=!1,x(N),N=-1),m=!0;var i=p;try{for(v(t),h=o(c);null!==h&&(!(h.expirationTime>t)||e&&!I());){var r=h.callback;if("function"===typeof r){h.callback=null,p=h.priorityLevel;var s=r(h.expirationTime<=t);t=n.unstable_now(),"function"===typeof s?h.callback=s:h===o(c)&&a(c),v(t)}else a(c);h=o(c)}if(null!==h)var l=!0;else{var u=o(d);null!==u&&F(j,u.startTime-t),l=!1}return l}finally{h=null,p=i,m=!1}}"undefined"!==typeof navigator&&void 0!==navigator.scheduling&&void 0!==navigator.scheduling.isInputPending&&navigator.scheduling.isInputPending.bind(navigator.scheduling);var w,S=!1,k=null,N=-1,_=5,T=-1;function I(){return!(n.unstable_now()-T<_)}function E(){if(null!==k){var e=n.unstable_now();T=e;var t=!0;try{t=k(!0,e)}finally{t?w():(S=!1,k=null)}}else S=!1}if("function"===typeof b)w=function(){b(E)};else if("undefined"!==typeof MessageChannel){var A=new MessageChannel,B=A.port2;A.port1.onmessage=E,w=function(){B.postMessage(null)}}else w=function(){y(E,0)};function R(e){k=e,S||(S=!0,w())}function F(e,t){N=y((function(){e(n.unstable_now())}),t)}n.unstable_IdlePriority=5,n.unstable_ImmediatePriority=1,n.unstable_LowPriority=4,n.unstable_NormalPriority=3,n.unstable_Profiling=null,n.unstable_UserBlockingPriority=2,n.unstable_cancelCallback=function(e){e.callback=null},n.unstable_continueExecution=function(){f||m||(f=!0,R(C))},n.unstable_forceFrameRate=function(e){0>e||125<e?console.error("forceFrameRate takes a positive int between 0 and 125, forcing frame rates higher than 125 fps is not supported"):_=0<e?Math.floor(1e3/e):5},n.unstable_getCurrentPriorityLevel=function(){return p},n.unstable_getFirstCallbackNode=function(){return o(c)},n.unstable_next=function(e){switch(p){case 1:case 2:case 3:var n=3;break;default:n=p}var t=p;p=n;try{return e()}finally{p=t}},n.unstable_pauseExecution=function(){},n.unstable_requestPaint=function(){},n.unstable_runWithPriority=function(e,n){switch(e){case 1:case 2:case 3:case 4:case 5:break;default:e=3}var t=p;p=e;try{return n()}finally{p=t}},n.unstable_scheduleCallback=function(e,a,i){var r=n.unstable_now();switch("object"===typeof i&&null!==i?i="number"===typeof(i=i.delay)&&0<i?r+i:r:i=r,e){case 1:var s=-1;break;case 2:s=250;break;case 5:s=1073741823;break;case 4:s=1e4;break;default:s=5e3}return e={id:u++,callback:a,priorityLevel:e,startTime:i,expirationTime:s=i+s,sortIndex:-1},i>r?(e.sortIndex=i,t(d,e),null===o(c)&&e===o(d)&&(g?(x(N),N=-1):g=!0,F(j,i-r))):(e.sortIndex=s,t(c,e),f||m||(f=!0,R(C))),e},n.unstable_shouldYield=I,n.unstable_wrapCallback=function(e){var n=p;return function(){var t=p;p=n;try{return e.apply(this,arguments)}finally{p=t}}}},588:(e,n,t)=>{e.exports=t(345)}},n={};function t(o){var a=n[o];if(void 0!==a)return a.exports;var i=n[o]={exports:{}};return e[o](i,i.exports,t),i.exports}t.m=e,(()=>{var e,n=Object.getPrototypeOf?e=>Object.getPrototypeOf(e):e=>e.__proto__;t.t=function(o,a){if(1&a&&(o=this(o)),8&a)return o;if("object"===typeof o&&o){if(4&a&&o.__esModule)return o;if(16&a&&"function"===typeof o.then)return o}var i=Object.create(null);t.r(i);var r={};e=e||[null,n({}),n([]),n(n)];for(var s=2&a&&o;"object"==typeof s&&!~e.indexOf(s);s=n(s))Object.getOwnPropertyNames(s).forEach((e=>r[e]=()=>o[e]));return r.default=()=>o,t.d(i,r),i}})(),t.d=(e,n)=>{for(var o in n)t.o(n,o)&&!t.o(e,o)&&Object.defineProperty(e,o,{enumerable:!0,get:n[o]})},t.f={},t.e=e=>Promise.all(Object.keys(t.f).reduce(((n,o)=>(t.f[o](e,n),n)),[])),t.u=e=>"static/js/"+e+".43678a5e.chunk.js",t.miniCssF=e=>{},t.o=(e,n)=>Object.prototype.hasOwnProperty.call(e,n),(()=>{var e={},n="docu-app:";t.l=(o,a,i,r)=>{if(e[o])e[o].push(a);else{var s,l;if(void 0!==i)for(var c=document.getElementsByTagName("script"),d=0;d<c.length;d++){var u=c[d];if(u.getAttribute("src")==o||u.getAttribute("data-webpack")==n+i){s=u;break}}s||(l=!0,(s=document.createElement("script")).charset="utf-8",s.timeout=120,t.nc&&s.setAttribute("nonce",t.nc),s.setAttribute("data-webpack",n+i),s.src=o),e[o]=[a];var h=(n,t)=>{s.onerror=s.onload=null,clearTimeout(p);var a=e[o];if(delete e[o],s.parentNode&&s.parentNode.removeChild(s),a&&a.forEach((e=>e(t))),n)return n(t)},p=setTimeout(h.bind(null,void 0,{type:"timeout",target:s}),12e4);s.onerror=h.bind(null,s.onerror),s.onload=h.bind(null,s.onload),l&&document.head.appendChild(s)}}})(),t.r=e=>{"undefined"!==typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})},t.p="/beacon-ri-documentation/",(()=>{var e={792:0};t.f.j=(n,o)=>{var a=t.o(e,n)?e[n]:void 0;if(0!==a)if(a)o.push(a[2]);else{var i=new Promise(((t,o)=>a=e[n]=[t,o]));o.push(a[2]=i);var r=t.p+t.u(n),s=new Error;t.l(r,(o=>{if(t.o(e,n)&&(0!==(a=e[n])&&(e[n]=void 0),a)){var i=o&&("load"===o.type?"missing":o.type),r=o&&o.target&&o.target.src;s.message="Loading chunk "+n+" failed.\n("+i+": "+r+")",s.name="ChunkLoadError",s.type=i,s.request=r,a[1](s)}}),"chunk-"+n,n)}};var n=(n,o)=>{var a,i,r=o[0],s=o[1],l=o[2],c=0;if(r.some((n=>0!==e[n]))){for(a in s)t.o(s,a)&&(t.m[a]=s[a]);if(l)l(t)}for(n&&n(o);c<r.length;c++)i=r[c],t.o(e,i)&&e[i]&&e[i][0](),e[i]=0},o=self.webpackChunkdocu_app=self.webpackChunkdocu_app||[];o.forEach(n.bind(null,0)),o.push=n.bind(null,o.push.bind(o))})();var o,a=t(998),i=t.t(a,2),r=t(896),s=t(407),l=t.t(s,2);function c(){return c=Object.assign?Object.assign.bind():function(e){for(var n=1;n<arguments.length;n++){var t=arguments[n];for(var o in t)Object.prototype.hasOwnProperty.call(t,o)&&(e[o]=t[o])}return e},c.apply(this,arguments)}!function(e){e.Pop="POP",e.Push="PUSH",e.Replace="REPLACE"}(o||(o={}));const d="popstate";function u(e,n){if(!1===e||null===e||"undefined"===typeof e)throw new Error(n)}function h(e,n){if(!e){"undefined"!==typeof console&&console.warn(n);try{throw new Error(n)}catch(t){}}}function p(e,n){return{usr:e.state,key:e.key,idx:n}}function m(e,n,t,o){return void 0===t&&(t=null),c({pathname:"string"===typeof e?e:e.pathname,search:"",hash:""},"string"===typeof n?g(n):n,{state:t,key:n&&n.key||o||Math.random().toString(36).substr(2,8)})}function f(e){let{pathname:n="/",search:t="",hash:o=""}=e;return t&&"?"!==t&&(n+="?"===t.charAt(0)?t:"?"+t),o&&"#"!==o&&(n+="#"===o.charAt(0)?o:"#"+o),n}function g(e){let n={};if(e){let t=e.indexOf("#");t>=0&&(n.hash=e.substr(t),e=e.substr(0,t));let o=e.indexOf("?");o>=0&&(n.search=e.substr(o),e=e.substr(0,o)),e&&(n.pathname=e)}return n}function y(e,n,t,a){void 0===a&&(a={});let{window:i=document.defaultView,v5Compat:r=!1}=a,s=i.history,l=o.Pop,h=null,g=y();function y(){return(s.state||{idx:null}).idx}function x(){l=o.Pop;let e=y(),n=null==e?null:e-g;g=e,h&&h({action:l,location:v.location,delta:n})}function b(e){let n="null"!==i.location.origin?i.location.origin:i.location.href,t="string"===typeof e?e:f(e);return t=t.replace(/ $/,"%20"),u(n,"No window.location.(origin|href) available to create URL for href: "+t),new URL(t,n)}null==g&&(g=0,s.replaceState(c({},s.state,{idx:g}),""));let v={get action(){return l},get location(){return e(i,s)},listen(e){if(h)throw new Error("A history only accepts one active listener");return i.addEventListener(d,x),h=e,()=>{i.removeEventListener(d,x),h=null}},createHref:e=>n(i,e),createURL:b,encodeLocation(e){let n=b(e);return{pathname:n.pathname,search:n.search,hash:n.hash}},push:function(e,n){l=o.Push;let a=m(v.location,e,n);t&&t(a,e),g=y()+1;let c=p(a,g),d=v.createHref(a);try{s.pushState(c,"",d)}catch(u){if(u instanceof DOMException&&"DataCloneError"===u.name)throw u;i.location.assign(d)}r&&h&&h({action:l,location:v.location,delta:1})},replace:function(e,n){l=o.Replace;let a=m(v.location,e,n);t&&t(a,e),g=y();let i=p(a,g),c=v.createHref(a);s.replaceState(i,"",c),r&&h&&h({action:l,location:v.location,delta:0})},go:e=>s.go(e)};return v}var x;!function(e){e.data="data",e.deferred="deferred",e.redirect="redirect",e.error="error"}(x||(x={}));new Set(["lazy","caseSensitive","path","id","index","children"]);function b(e,n,t){return void 0===t&&(t="/"),v(e,n,t,!1)}function v(e,n,t,o){let a=F(("string"===typeof n?g(n):n).pathname||"/",t);if(null==a)return null;let i=j(e);!function(e){e.sort(((e,n)=>e.score!==n.score?n.score-e.score:function(e,n){let t=e.length===n.length&&e.slice(0,-1).every(((e,t)=>e===n[t]));return t?e[e.length-1]-n[n.length-1]:0}(e.routesMeta.map((e=>e.childrenIndex)),n.routesMeta.map((e=>e.childrenIndex)))))}(i);let r=null;for(let s=0;null==r&&s<i.length;++s){let e=R(a);r=A(i[s],e,o)}return r}function j(e,n,t,o){void 0===n&&(n=[]),void 0===t&&(t=[]),void 0===o&&(o="");let a=(e,a,i)=>{let r={relativePath:void 0===i?e.path||"":i,caseSensitive:!0===e.caseSensitive,childrenIndex:a,route:e};r.relativePath.startsWith("/")&&(u(r.relativePath.startsWith(o),'Absolute route path "'+r.relativePath+'" nested under path "'+o+'" is not valid. An absolute child route path must start with the combined path of all its parent routes.'),r.relativePath=r.relativePath.slice(o.length));let s=L([o,r.relativePath]),l=t.concat(r);e.children&&e.children.length>0&&(u(!0!==e.index,'Index routes must not have child routes. Please remove all child routes from route path "'+s+'".'),j(e.children,n,l,s)),(null!=e.path||e.index)&&n.push({path:s,score:E(s,e.index),routesMeta:l})};return e.forEach(((e,n)=>{var t;if(""!==e.path&&null!=(t=e.path)&&t.includes("?"))for(let o of C(e.path))a(e,n,o);else a(e,n)})),n}function C(e){let n=e.split("/");if(0===n.length)return[];let[t,...o]=n,a=t.endsWith("?"),i=t.replace(/\?$/,"");if(0===o.length)return a?[i,""]:[i];let r=C(o.join("/")),s=[];return s.push(...r.map((e=>""===e?i:[i,e].join("/")))),a&&s.push(...r),s.map((n=>e.startsWith("/")&&""===n?"/":n))}const w=/^:[\w-]+$/,S=3,k=2,N=1,_=10,T=-2,I=e=>"*"===e;function E(e,n){let t=e.split("/"),o=t.length;return t.some(I)&&(o+=T),n&&(o+=k),t.filter((e=>!I(e))).reduce(((e,n)=>e+(w.test(n)?S:""===n?N:_)),o)}function A(e,n,t){void 0===t&&(t=!1);let{routesMeta:o}=e,a={},i="/",r=[];for(let s=0;s<o.length;++s){let e=o[s],l=s===o.length-1,c="/"===i?n:n.slice(i.length)||"/",d=B({path:e.relativePath,caseSensitive:e.caseSensitive,end:l},c),u=e.route;if(!d&&l&&t&&!o[o.length-1].route.index&&(d=B({path:e.relativePath,caseSensitive:e.caseSensitive,end:!1},c)),!d)return null;Object.assign(a,d.params),r.push({params:a,pathname:L([i,d.pathname]),pathnameBase:G(L([i,d.pathnameBase])),route:u}),"/"!==d.pathnameBase&&(i=L([i,d.pathnameBase]))}return r}function B(e,n){"string"===typeof e&&(e={path:e,caseSensitive:!1,end:!0});let[t,o]=function(e,n,t){void 0===n&&(n=!1);void 0===t&&(t=!0);h("*"===e||!e.endsWith("*")||e.endsWith("/*"),'Route path "'+e+'" will be treated as if it were "'+e.replace(/\*$/,"/*")+'" because the `*` character must always follow a `/` in the pattern. To get rid of this warning, please change the route path to "'+e.replace(/\*$/,"/*")+'".');let o=[],a="^"+e.replace(/\/*\*?$/,"").replace(/^\/*/,"/").replace(/[\\.*+^${}|()[\]]/g,"\\$&").replace(/\/:([\w-]+)(\?)?/g,((e,n,t)=>(o.push({paramName:n,isOptional:null!=t}),t?"/?([^\\/]+)?":"/([^\\/]+)")));e.endsWith("*")?(o.push({paramName:"*"}),a+="*"===e||"/*"===e?"(.*)$":"(?:\\/(.+)|\\/*)$"):t?a+="\\/*$":""!==e&&"/"!==e&&(a+="(?:(?=\\/|$))");let i=new RegExp(a,n?void 0:"i");return[i,o]}(e.path,e.caseSensitive,e.end),a=n.match(t);if(!a)return null;let i=a[0],r=i.replace(/(.)\/+$/,"$1"),s=a.slice(1);return{params:o.reduce(((e,n,t)=>{let{paramName:o,isOptional:a}=n;if("*"===o){let e=s[t]||"";r=i.slice(0,i.length-e.length).replace(/(.)\/+$/,"$1")}const l=s[t];return e[o]=a&&!l?void 0:(l||"").replace(/%2F/g,"/"),e}),{}),pathname:i,pathnameBase:r,pattern:e}}function R(e){try{return e.split("/").map((e=>decodeURIComponent(e).replace(/\//g,"%2F"))).join("/")}catch(n){return h(!1,'The URL path "'+e+'" could not be decoded because it is is a malformed URL segment. This is probably due to a bad percent encoding ('+n+")."),e}}function F(e,n){if("/"===n)return e;if(!e.toLowerCase().startsWith(n.toLowerCase()))return null;let t=n.endsWith("/")?n.length-1:n.length,o=e.charAt(t);return o&&"/"!==o?null:e.slice(t)||"/"}function P(e,n,t,o){return"Cannot include a '"+e+"' character in a manually specified `to."+n+"` field ["+JSON.stringify(o)+"].  Please separate it out to the `to."+t+'` field. Alternatively you may provide the full path as a string in <Link to="..."> and the router will parse it for you.'}function D(e){return e.filter(((e,n)=>0===n||e.route.path&&e.route.path.length>0))}function O(e,n){let t=D(e);return n?t.map(((n,t)=>t===e.length-1?n.pathname:n.pathnameBase)):t.map((e=>e.pathnameBase))}function M(e,n,t,o){let a;void 0===o&&(o=!1),"string"===typeof e?a=g(e):(a=c({},e),u(!a.pathname||!a.pathname.includes("?"),P("?","pathname","search",a)),u(!a.pathname||!a.pathname.includes("#"),P("#","pathname","hash",a)),u(!a.search||!a.search.includes("#"),P("#","search","hash",a)));let i,r=""===e||""===a.pathname,s=r?"/":a.pathname;if(null==s)i=t;else{let e=n.length-1;if(!o&&s.startsWith("..")){let n=s.split("/");for(;".."===n[0];)n.shift(),e-=1;a.pathname=n.join("/")}i=e>=0?n[e]:"/"}let l=function(e,n){void 0===n&&(n="/");let{pathname:t,search:o="",hash:a=""}="string"===typeof e?g(e):e,i=t?t.startsWith("/")?t:function(e,n){let t=n.replace(/\/+$/,"").split("/");return e.split("/").forEach((e=>{".."===e?t.length>1&&t.pop():"."!==e&&t.push(e)})),t.length>1?t.join("/"):"/"}(t,n):n;return{pathname:i,search:V(o),hash:q(a)}}(a,i),d=s&&"/"!==s&&s.endsWith("/"),h=(r||"."===s)&&t.endsWith("/");return l.pathname.endsWith("/")||!d&&!h||(l.pathname+="/"),l}const L=e=>e.join("/").replace(/\/\/+/g,"/"),G=e=>e.replace(/\/+$/,"").replace(/^\/*/,"/"),V=e=>e&&"?"!==e?e.startsWith("?")?e:"?"+e:"",q=e=>e&&"#"!==e?e.startsWith("#")?e:"#"+e:"";Error;function H(e){return null!=e&&"number"===typeof e.status&&"string"===typeof e.statusText&&"boolean"===typeof e.internal&&"data"in e}const z=["post","put","patch","delete"],U=(new Set(z),["get",...z]);new Set(U),new Set([301,302,303,307,308]),new Set([307,308]);Symbol("deferred");function Q(){return Q=Object.assign?Object.assign.bind():function(e){for(var n=1;n<arguments.length;n++){var t=arguments[n];for(var o in t)Object.prototype.hasOwnProperty.call(t,o)&&(e[o]=t[o])}return e},Q.apply(this,arguments)}const W=a.createContext(null);const K=a.createContext(null);const $=a.createContext(null);const Y=a.createContext(null);const X=a.createContext({outlet:null,matches:[],isDataRoute:!1});const Z=a.createContext(null);function J(){return null!=a.useContext(Y)}function ee(){return J()||u(!1),a.useContext(Y).location}function ne(e){a.useContext($).static||a.useLayoutEffect(e)}function te(){let{isDataRoute:e}=a.useContext(X);return e?function(){let{router:e}=ue(ce.UseNavigateStable),n=pe(de.UseNavigateStable),t=a.useRef(!1);return ne((()=>{t.current=!0})),a.useCallback((function(o,a){void 0===a&&(a={}),t.current&&("number"===typeof o?e.navigate(o):e.navigate(o,Q({fromRouteId:n},a)))}),[e,n])}():function(){J()||u(!1);let e=a.useContext(W),{basename:n,future:t,navigator:o}=a.useContext($),{matches:i}=a.useContext(X),{pathname:r}=ee(),s=JSON.stringify(O(i,t.v7_relativeSplatPath)),l=a.useRef(!1);return ne((()=>{l.current=!0})),a.useCallback((function(t,a){if(void 0===a&&(a={}),!l.current)return;if("number"===typeof t)return void o.go(t);let i=M(t,JSON.parse(s),r,"path"===a.relative);null==e&&"/"!==n&&(i.pathname="/"===i.pathname?n:L([n,i.pathname])),(a.replace?o.replace:o.push)(i,a.state,a)}),[n,o,s,r,e])}()}function oe(e,n,t,i){J()||u(!1);let{navigator:r}=a.useContext($),{matches:s}=a.useContext(X),l=s[s.length-1],c=l?l.params:{},d=(l&&l.pathname,l?l.pathnameBase:"/");l&&l.route;let h,p=ee();if(n){var m;let e="string"===typeof n?g(n):n;"/"===d||(null==(m=e.pathname)?void 0:m.startsWith(d))||u(!1),h=e}else h=p;let f=h.pathname||"/",y=f;if("/"!==d){let e=d.replace(/^\//,"").split("/");y="/"+f.replace(/^\//,"").split("/").slice(e.length).join("/")}let x=b(e,{pathname:y});let v=le(x&&x.map((e=>Object.assign({},e,{params:Object.assign({},c,e.params),pathname:L([d,r.encodeLocation?r.encodeLocation(e.pathname).pathname:e.pathname]),pathnameBase:"/"===e.pathnameBase?d:L([d,r.encodeLocation?r.encodeLocation(e.pathnameBase).pathname:e.pathnameBase])}))),s,t,i);return n&&v?a.createElement(Y.Provider,{value:{location:Q({pathname:"/",search:"",hash:"",state:null,key:"default"},h),navigationType:o.Pop}},v):v}function ae(){let e=function(){var e;let n=a.useContext(Z),t=he(de.UseRouteError),o=pe(de.UseRouteError);if(void 0!==n)return n;return null==(e=t.errors)?void 0:e[o]}(),n=H(e)?e.status+" "+e.statusText:e instanceof Error?e.message:JSON.stringify(e),t=e instanceof Error?e.stack:null,o="rgba(200,200,200, 0.5)",i={padding:"0.5rem",backgroundColor:o};return a.createElement(a.Fragment,null,a.createElement("h2",null,"Unexpected Application Error!"),a.createElement("h3",{style:{fontStyle:"italic"}},n),t?a.createElement("pre",{style:i},t):null,null)}const ie=a.createElement(ae,null);class re extends a.Component{constructor(e){super(e),this.state={location:e.location,revalidation:e.revalidation,error:e.error}}static getDerivedStateFromError(e){return{error:e}}static getDerivedStateFromProps(e,n){return n.location!==e.location||"idle"!==n.revalidation&&"idle"===e.revalidation?{error:e.error,location:e.location,revalidation:e.revalidation}:{error:void 0!==e.error?e.error:n.error,location:n.location,revalidation:e.revalidation||n.revalidation}}componentDidCatch(e,n){console.error("React Router caught the following error during render",e,n)}render(){return void 0!==this.state.error?a.createElement(X.Provider,{value:this.props.routeContext},a.createElement(Z.Provider,{value:this.state.error,children:this.props.component})):this.props.children}}function se(e){let{routeContext:n,match:t,children:o}=e,i=a.useContext(W);return i&&i.static&&i.staticContext&&(t.route.errorElement||t.route.ErrorBoundary)&&(i.staticContext._deepestRenderedBoundaryId=t.route.id),a.createElement(X.Provider,{value:n},o)}function le(e,n,t,o){var i;if(void 0===n&&(n=[]),void 0===t&&(t=null),void 0===o&&(o=null),null==e){var r;if(null==(r=t)||!r.errors)return null;e=t.matches}let s=e,l=null==(i=t)?void 0:i.errors;if(null!=l){let e=s.findIndex((e=>e.route.id&&void 0!==(null==l?void 0:l[e.route.id])));e>=0||u(!1),s=s.slice(0,Math.min(s.length,e+1))}let c=!1,d=-1;if(t&&o&&o.v7_partialHydration)for(let a=0;a<s.length;a++){let e=s[a];if((e.route.HydrateFallback||e.route.hydrateFallbackElement)&&(d=a),e.route.id){let{loaderData:n,errors:o}=t,a=e.route.loader&&void 0===n[e.route.id]&&(!o||void 0===o[e.route.id]);if(e.route.lazy||a){c=!0,s=d>=0?s.slice(0,d+1):[s[0]];break}}}return s.reduceRight(((e,o,i)=>{let r,u=!1,h=null,p=null;var m;t&&(r=l&&o.route.id?l[o.route.id]:void 0,h=o.route.errorElement||ie,c&&(d<0&&0===i?(m="route-fallback",!1||me[m]||(me[m]=!0),u=!0,p=null):d===i&&(u=!0,p=o.route.hydrateFallbackElement||null)));let f=n.concat(s.slice(0,i+1)),g=()=>{let n;return n=r?h:u?p:o.route.Component?a.createElement(o.route.Component,null):o.route.element?o.route.element:e,a.createElement(se,{match:o,routeContext:{outlet:e,matches:f,isDataRoute:null!=t},children:n})};return t&&(o.route.ErrorBoundary||o.route.errorElement||0===i)?a.createElement(re,{location:t.location,revalidation:t.revalidation,component:h,error:r,children:g(),routeContext:{outlet:null,matches:f,isDataRoute:!0}}):g()}),null)}var ce=function(e){return e.UseBlocker="useBlocker",e.UseRevalidator="useRevalidator",e.UseNavigateStable="useNavigate",e}(ce||{}),de=function(e){return e.UseBlocker="useBlocker",e.UseLoaderData="useLoaderData",e.UseActionData="useActionData",e.UseRouteError="useRouteError",e.UseNavigation="useNavigation",e.UseRouteLoaderData="useRouteLoaderData",e.UseMatches="useMatches",e.UseRevalidator="useRevalidator",e.UseNavigateStable="useNavigate",e.UseRouteId="useRouteId",e}(de||{});function ue(e){let n=a.useContext(W);return n||u(!1),n}function he(e){let n=a.useContext(K);return n||u(!1),n}function pe(e){let n=function(e){let n=a.useContext(X);return n||u(!1),n}(),t=n.matches[n.matches.length-1];return t.route.id||u(!1),t.route.id}const me={};i.startTransition;function fe(e){u(!1)}function ge(e){let{basename:n="/",children:t=null,location:i,navigationType:r=o.Pop,navigator:s,static:l=!1,future:c}=e;J()&&u(!1);let d=n.replace(/^\/*/,"/"),h=a.useMemo((()=>({basename:d,navigator:s,static:l,future:Q({v7_relativeSplatPath:!1},c)})),[d,c,s,l]);"string"===typeof i&&(i=g(i));let{pathname:p="/",search:m="",hash:f="",state:y=null,key:x="default"}=i,b=a.useMemo((()=>{let e=F(p,d);return null==e?null:{location:{pathname:e,search:m,hash:f,state:y,key:x},navigationType:r}}),[d,p,m,f,y,x,r]);return null==b?null:a.createElement($.Provider,{value:h},a.createElement(Y.Provider,{children:t,value:b}))}function ye(e){let{children:n,location:t}=e;return oe(xe(n),t)}new Promise((()=>{}));a.Component;function xe(e,n){void 0===n&&(n=[]);let t=[];return a.Children.forEach(e,((e,o)=>{if(!a.isValidElement(e))return;let i=[...n,o];if(e.type===a.Fragment)return void t.push.apply(t,xe(e.props.children,i));e.type!==fe&&u(!1),e.props.index&&e.props.children&&u(!1);let r={id:e.props.id||i.join("-"),caseSensitive:e.props.caseSensitive,element:e.props.element,Component:e.props.Component,index:e.props.index,path:e.props.path,loader:e.props.loader,action:e.props.action,errorElement:e.props.errorElement,ErrorBoundary:e.props.ErrorBoundary,hasErrorBoundary:null!=e.props.ErrorBoundary||null!=e.props.errorElement,shouldRevalidate:e.props.shouldRevalidate,handle:e.props.handle,lazy:e.props.lazy};e.props.children&&(r.children=xe(e.props.children,i)),t.push(r)})),t}new Set(["application/x-www-form-urlencoded","multipart/form-data","text/plain"]);try{window.__reactRouterVersion="6"}catch(kc){}new Map;const be=i.startTransition;l.flushSync,i.useId;function ve(e){let{basename:n,children:t,future:o,window:i}=e,r=a.useRef();var s;null==r.current&&(r.current=(void 0===(s={window:i,v5Compat:!0})&&(s={}),y((function(e,n){let{pathname:t,search:o,hash:a}=e.location;return m("",{pathname:t,search:o,hash:a},n.state&&n.state.usr||null,n.state&&n.state.key||"default")}),(function(e,n){return"string"===typeof n?n:f(n)}),null,s)));let l=r.current,[c,d]=a.useState({action:l.action,location:l.location}),{v7_startTransition:u}=o||{},h=a.useCallback((e=>{u&&be?be((()=>d(e))):d(e)}),[d,u]);return a.useLayoutEffect((()=>l.listen(h)),[l,h]),a.createElement(ge,{basename:n,children:t,location:c.location,navigationType:c.action,navigator:l,future:o})}"undefined"!==typeof window&&"undefined"!==typeof window.document&&window.document.createElement;var je,Ce;(function(e){e.UseScrollRestoration="useScrollRestoration",e.UseSubmit="useSubmit",e.UseSubmitFetcher="useSubmitFetcher",e.UseFetcher="useFetcher",e.useViewTransitionState="useViewTransitionState"})(je||(je={})),function(e){e.UseFetcher="useFetcher",e.UseFetchers="useFetchers",e.UseScrollRestoration="useScrollRestoration"}(Ce||(Ce={}));var we=t(446);const Se=(0,a.createContext)(void 0),ke=e=>{let{children:n}=e;const[t,o]=(0,a.useState)([]);return(0,we.jsx)(Se.Provider,{value:{searchableContent:t,addSearchableContent:(e,n,t)=>{o((o=>[...o,{route:e,text:n,title:t}]))}},children:n})},Ne=()=>{const e=(0,a.useContext)(Se);if(!e)throw new Error("useSearchableContent must be used within a SearchableContentProvider");return e},_e=e=>{let{onSearch:n}=e;const{searchableContent:t}=Ne(),[o,i]=(0,a.useState)(""),[r,s]=(0,a.useState)(!1),l=(e,n)=>{if(!n.trim())return e;const t=new RegExp("(".concat(n,")"),"gi");return e.split(t).map(((e,t)=>e.toLowerCase()===n.toLowerCase()?(0,we.jsx)("span",{className:"highlight",children:e},t):e))},c=te(),d=o?t.filter((e=>e.text.toLowerCase().includes(o.toLowerCase()))).map((e=>{const n=e.text.toLowerCase().indexOf(o.toLowerCase()),t=n+o.length,a="".concat(e.text.substring(Math.max(0,n-20),n)).concat(o).concat(e.text.substring(t,Math.min(e.text.length,t+20)));return{...e,previewText:a}})):[];return(0,we.jsxs)("div",{className:"containerNav",children:[(0,we.jsx)("h1",{className:"main-title",children:"Beacon v2 RI / PI Documentation"}),(0,we.jsx)("div",{className:"divOptionsInput",children:(0,we.jsxs)("div",{className:"inputContainer",children:[(0,we.jsxs)("div",{className:"inputDiv",children:[(0,we.jsx)("img",{src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/Zoom-in.png?raw=true",className:"searchIcon",alt:"Search Icon"}),(0,we.jsx)("input",{className:"inputSearch",type:"text",placeholder:"Search keywords",value:o,onChange:e=>{const t=e.target.value;console.log("User Typed:",t),i(t),n(t),s(""!==t)}}),o&&(0,we.jsx)("button",{className:"clearButton",onClick:()=>{i(""),n(""),s(!1)},children:"\xd7"})]}),r&&d.length>0&&(0,we.jsx)("div",{className:"searchResults",children:d.map(((e,n)=>(0,we.jsxs)("div",{className:"searchResultItem",onClick:()=>{return n=e.route,c(n),s(!1),void window.scrollTo({top:0,behavior:"smooth"});var n},children:[(0,we.jsx)("div",{className:"titleResult",children:(0,we.jsx)("strong",{children:e.title})}),(0,we.jsxs)("div",{className:"previewTextResult",children:["\u2022 ",l(e.previewText,o),"..."]})]},n)))})]})})]})};function Te(e){var n,t,o="";if("string"==typeof e||"number"==typeof e)o+=e;else if("object"==typeof e)if(Array.isArray(e)){var a=e.length;for(n=0;n<a;n++)e[n]&&(t=Te(e[n]))&&(o&&(o+=" "),o+=t)}else for(t in e)e[t]&&(o&&(o+=" "),o+=t);return o}const Ie=function(){for(var e,n,t=0,o="",a=arguments.length;t<a;t++)(e=arguments[t])&&(n=Te(e))&&(o&&(o+=" "),o+=n);return o};function Ee(){return Ee=Object.assign?Object.assign.bind():function(e){for(var n=1;n<arguments.length;n++){var t=arguments[n];for(var o in t)({}).hasOwnProperty.call(t,o)&&(e[o]=t[o])}return e},Ee.apply(null,arguments)}var Ae=function(){function e(e){var n=this;this._insertTag=function(e){var t;t=0===n.tags.length?n.insertionPoint?n.insertionPoint.nextSibling:n.prepend?n.container.firstChild:n.before:n.tags[n.tags.length-1].nextSibling,n.container.insertBefore(e,t),n.tags.push(e)},this.isSpeedy=void 0===e.speedy||e.speedy,this.tags=[],this.ctr=0,this.nonce=e.nonce,this.key=e.key,this.container=e.container,this.prepend=e.prepend,this.insertionPoint=e.insertionPoint,this.before=null}var n=e.prototype;return n.hydrate=function(e){e.forEach(this._insertTag)},n.insert=function(e){this.ctr%(this.isSpeedy?65e3:1)===0&&this._insertTag(function(e){var n=document.createElement("style");return n.setAttribute("data-emotion",e.key),void 0!==e.nonce&&n.setAttribute("nonce",e.nonce),n.appendChild(document.createTextNode("")),n.setAttribute("data-s",""),n}(this));var n=this.tags[this.tags.length-1];if(this.isSpeedy){var t=function(e){if(e.sheet)return e.sheet;for(var n=0;n<document.styleSheets.length;n++)if(document.styleSheets[n].ownerNode===e)return document.styleSheets[n]}(n);try{t.insertRule(e,t.cssRules.length)}catch(kc){}}else n.appendChild(document.createTextNode(e));this.ctr++},n.flush=function(){this.tags.forEach((function(e){var n;return null==(n=e.parentNode)?void 0:n.removeChild(e)})),this.tags=[],this.ctr=0},e}(),Be=Math.abs,Re=String.fromCharCode,Fe=Object.assign;function Pe(e){return e.trim()}function De(e,n,t){return e.replace(n,t)}function Oe(e,n){return e.indexOf(n)}function Me(e,n){return 0|e.charCodeAt(n)}function Le(e,n,t){return e.slice(n,t)}function Ge(e){return e.length}function Ve(e){return e.length}function qe(e,n){return n.push(e),e}var He=1,ze=1,Ue=0,Qe=0,We=0,Ke="";function $e(e,n,t,o,a,i,r){return{value:e,root:n,parent:t,type:o,props:a,children:i,line:He,column:ze,length:r,return:""}}function Ye(e,n){return Fe($e("",null,null,"",null,null,0),e,{length:-e.length},n)}function Xe(){return We=Qe>0?Me(Ke,--Qe):0,ze--,10===We&&(ze=1,He--),We}function Ze(){return We=Qe<Ue?Me(Ke,Qe++):0,ze++,10===We&&(ze=1,He++),We}function Je(){return Me(Ke,Qe)}function en(){return Qe}function nn(e,n){return Le(Ke,e,n)}function tn(e){switch(e){case 0:case 9:case 10:case 13:case 32:return 5;case 33:case 43:case 44:case 47:case 62:case 64:case 126:case 59:case 123:case 125:return 4;case 58:return 3;case 34:case 39:case 40:case 91:return 2;case 41:case 93:return 1}return 0}function on(e){return He=ze=1,Ue=Ge(Ke=e),Qe=0,[]}function an(e){return Ke="",e}function rn(e){return Pe(nn(Qe-1,cn(91===e?e+2:40===e?e+1:e)))}function sn(e){for(;(We=Je())&&We<33;)Ze();return tn(e)>2||tn(We)>3?"":" "}function ln(e,n){for(;--n&&Ze()&&!(We<48||We>102||We>57&&We<65||We>70&&We<97););return nn(e,en()+(n<6&&32==Je()&&32==Ze()))}function cn(e){for(;Ze();)switch(We){case e:return Qe;case 34:case 39:34!==e&&39!==e&&cn(We);break;case 40:41===e&&cn(e);break;case 92:Ze()}return Qe}function dn(e,n){for(;Ze()&&e+We!==57&&(e+We!==84||47!==Je()););return"/*"+nn(n,Qe-1)+"*"+Re(47===e?e:Ze())}function un(e){for(;!tn(Je());)Ze();return nn(e,Qe)}var hn="-ms-",pn="-moz-",mn="-webkit-",fn="comm",gn="rule",yn="decl",xn="@keyframes";function bn(e,n){for(var t="",o=Ve(e),a=0;a<o;a++)t+=n(e[a],a,e,n)||"";return t}function vn(e,n,t,o){switch(e.type){case"@layer":if(e.children.length)break;case"@import":case yn:return e.return=e.return||e.value;case fn:return"";case xn:return e.return=e.value+"{"+bn(e.children,o)+"}";case gn:e.value=e.props.join(",")}return Ge(t=bn(e.children,o))?e.return=e.value+"{"+t+"}":""}function jn(e){return an(Cn("",null,null,null,[""],e=on(e),0,[0],e))}function Cn(e,n,t,o,a,i,r,s,l){for(var c=0,d=0,u=r,h=0,p=0,m=0,f=1,g=1,y=1,x=0,b="",v=a,j=i,C=o,w=b;g;)switch(m=x,x=Ze()){case 40:if(108!=m&&58==Me(w,u-1)){-1!=Oe(w+=De(rn(x),"&","&\f"),"&\f")&&(y=-1);break}case 34:case 39:case 91:w+=rn(x);break;case 9:case 10:case 13:case 32:w+=sn(m);break;case 92:w+=ln(en()-1,7);continue;case 47:switch(Je()){case 42:case 47:qe(Sn(dn(Ze(),en()),n,t),l);break;default:w+="/"}break;case 123*f:s[c++]=Ge(w)*y;case 125*f:case 59:case 0:switch(x){case 0:case 125:g=0;case 59+d:-1==y&&(w=De(w,/\f/g,"")),p>0&&Ge(w)-u&&qe(p>32?kn(w+";",o,t,u-1):kn(De(w," ","")+";",o,t,u-2),l);break;case 59:w+=";";default:if(qe(C=wn(w,n,t,c,d,a,s,b,v=[],j=[],u),i),123===x)if(0===d)Cn(w,n,C,C,v,i,u,s,j);else switch(99===h&&110===Me(w,3)?100:h){case 100:case 108:case 109:case 115:Cn(e,C,C,o&&qe(wn(e,C,C,0,0,a,s,b,a,v=[],u),j),a,j,u,s,o?v:j);break;default:Cn(w,C,C,C,[""],j,0,s,j)}}c=d=p=0,f=y=1,b=w="",u=r;break;case 58:u=1+Ge(w),p=m;default:if(f<1)if(123==x)--f;else if(125==x&&0==f++&&125==Xe())continue;switch(w+=Re(x),x*f){case 38:y=d>0?1:(w+="\f",-1);break;case 44:s[c++]=(Ge(w)-1)*y,y=1;break;case 64:45===Je()&&(w+=rn(Ze())),h=Je(),d=u=Ge(b=w+=un(en())),x++;break;case 45:45===m&&2==Ge(w)&&(f=0)}}return i}function wn(e,n,t,o,a,i,r,s,l,c,d){for(var u=a-1,h=0===a?i:[""],p=Ve(h),m=0,f=0,g=0;m<o;++m)for(var y=0,x=Le(e,u+1,u=Be(f=r[m])),b=e;y<p;++y)(b=Pe(f>0?h[y]+" "+x:De(x,/&\f/g,h[y])))&&(l[g++]=b);return $e(e,n,t,0===a?gn:s,l,c,d)}function Sn(e,n,t){return $e(e,n,t,fn,Re(We),Le(e,2,-2),0)}function kn(e,n,t,o){return $e(e,n,t,yn,Le(e,0,o),Le(e,o+1,-1),o)}var Nn=function(e,n,t){for(var o=0,a=0;o=a,a=Je(),38===o&&12===a&&(n[t]=1),!tn(a);)Ze();return nn(e,Qe)},_n=function(e,n){return an(function(e,n){var t=-1,o=44;do{switch(tn(o)){case 0:38===o&&12===Je()&&(n[t]=1),e[t]+=Nn(Qe-1,n,t);break;case 2:e[t]+=rn(o);break;case 4:if(44===o){e[++t]=58===Je()?"&\f":"",n[t]=e[t].length;break}default:e[t]+=Re(o)}}while(o=Ze());return e}(on(e),n))},Tn=new WeakMap,In=function(e){if("rule"===e.type&&e.parent&&!(e.length<1)){for(var n=e.value,t=e.parent,o=e.column===t.column&&e.line===t.line;"rule"!==t.type;)if(!(t=t.parent))return;if((1!==e.props.length||58===n.charCodeAt(0)||Tn.get(t))&&!o){Tn.set(e,!0);for(var a=[],i=_n(n,a),r=t.props,s=0,l=0;s<i.length;s++)for(var c=0;c<r.length;c++,l++)e.props[l]=a[s]?i[s].replace(/&\f/g,r[c]):r[c]+" "+i[s]}}},En=function(e){if("decl"===e.type){var n=e.value;108===n.charCodeAt(0)&&98===n.charCodeAt(2)&&(e.return="",e.value="")}};function An(e,n){switch(function(e,n){return 45^Me(e,0)?(((n<<2^Me(e,0))<<2^Me(e,1))<<2^Me(e,2))<<2^Me(e,3):0}(e,n)){case 5103:return mn+"print-"+e+e;case 5737:case 4201:case 3177:case 3433:case 1641:case 4457:case 2921:case 5572:case 6356:case 5844:case 3191:case 6645:case 3005:case 6391:case 5879:case 5623:case 6135:case 4599:case 4855:case 4215:case 6389:case 5109:case 5365:case 5621:case 3829:return mn+e+e;case 5349:case 4246:case 4810:case 6968:case 2756:return mn+e+pn+e+hn+e+e;case 6828:case 4268:return mn+e+hn+e+e;case 6165:return mn+e+hn+"flex-"+e+e;case 5187:return mn+e+De(e,/(\w+).+(:[^]+)/,mn+"box-$1$2"+hn+"flex-$1$2")+e;case 5443:return mn+e+hn+"flex-item-"+De(e,/flex-|-self/,"")+e;case 4675:return mn+e+hn+"flex-line-pack"+De(e,/align-content|flex-|-self/,"")+e;case 5548:return mn+e+hn+De(e,"shrink","negative")+e;case 5292:return mn+e+hn+De(e,"basis","preferred-size")+e;case 6060:return mn+"box-"+De(e,"-grow","")+mn+e+hn+De(e,"grow","positive")+e;case 4554:return mn+De(e,/([^-])(transform)/g,"$1"+mn+"$2")+e;case 6187:return De(De(De(e,/(zoom-|grab)/,mn+"$1"),/(image-set)/,mn+"$1"),e,"")+e;case 5495:case 3959:return De(e,/(image-set\([^]*)/,mn+"$1$`$1");case 4968:return De(De(e,/(.+:)(flex-)?(.*)/,mn+"box-pack:$3"+hn+"flex-pack:$3"),/s.+-b[^;]+/,"justify")+mn+e+e;case 4095:case 3583:case 4068:case 2532:return De(e,/(.+)-inline(.+)/,mn+"$1$2")+e;case 8116:case 7059:case 5753:case 5535:case 5445:case 5701:case 4933:case 4677:case 5533:case 5789:case 5021:case 4765:if(Ge(e)-1-n>6)switch(Me(e,n+1)){case 109:if(45!==Me(e,n+4))break;case 102:return De(e,/(.+:)(.+)-([^]+)/,"$1"+mn+"$2-$3$1"+pn+(108==Me(e,n+3)?"$3":"$2-$3"))+e;case 115:return~Oe(e,"stretch")?An(De(e,"stretch","fill-available"),n)+e:e}break;case 4949:if(115!==Me(e,n+1))break;case 6444:switch(Me(e,Ge(e)-3-(~Oe(e,"!important")&&10))){case 107:return De(e,":",":"+mn)+e;case 101:return De(e,/(.+:)([^;!]+)(;|!.+)?/,"$1"+mn+(45===Me(e,14)?"inline-":"")+"box$3$1"+mn+"$2$3$1"+hn+"$2box$3")+e}break;case 5936:switch(Me(e,n+11)){case 114:return mn+e+hn+De(e,/[svh]\w+-[tblr]{2}/,"tb")+e;case 108:return mn+e+hn+De(e,/[svh]\w+-[tblr]{2}/,"tb-rl")+e;case 45:return mn+e+hn+De(e,/[svh]\w+-[tblr]{2}/,"lr")+e}return mn+e+hn+e+e}return e}var Bn=[function(e,n,t,o){if(e.length>-1&&!e.return)switch(e.type){case yn:e.return=An(e.value,e.length);break;case xn:return bn([Ye(e,{value:De(e.value,"@","@"+mn)})],o);case gn:if(e.length)return function(e,n){return e.map(n).join("")}(e.props,(function(n){switch(function(e,n){return(e=n.exec(e))?e[0]:e}(n,/(::plac\w+|:read-\w+)/)){case":read-only":case":read-write":return bn([Ye(e,{props:[De(n,/:(read-\w+)/,":-moz-$1")]})],o);case"::placeholder":return bn([Ye(e,{props:[De(n,/:(plac\w+)/,":"+mn+"input-$1")]}),Ye(e,{props:[De(n,/:(plac\w+)/,":-moz-$1")]}),Ye(e,{props:[De(n,/:(plac\w+)/,hn+"input-$1")]})],o)}return""}))}}],Rn=function(e){var n=e.key;if("css"===n){var t=document.querySelectorAll("style[data-emotion]:not([data-s])");Array.prototype.forEach.call(t,(function(e){-1!==e.getAttribute("data-emotion").indexOf(" ")&&(document.head.appendChild(e),e.setAttribute("data-s",""))}))}var o,a,i=e.stylisPlugins||Bn,r={},s=[];o=e.container||document.head,Array.prototype.forEach.call(document.querySelectorAll('style[data-emotion^="'+n+' "]'),(function(e){for(var n=e.getAttribute("data-emotion").split(" "),t=1;t<n.length;t++)r[n[t]]=!0;s.push(e)}));var l,c,d=[vn,(c=function(e){l.insert(e)},function(e){e.root||(e=e.return)&&c(e)})],u=function(e){var n=Ve(e);return function(t,o,a,i){for(var r="",s=0;s<n;s++)r+=e[s](t,o,a,i)||"";return r}}([In,En].concat(i,d));a=function(e,n,t,o){l=t,function(e){bn(jn(e),u)}(e?e+"{"+n.styles+"}":n.styles),o&&(h.inserted[n.name]=!0)};var h={key:n,sheet:new Ae({key:n,container:o,nonce:e.nonce,speedy:e.speedy,prepend:e.prepend,insertionPoint:e.insertionPoint}),nonce:e.nonce,inserted:r,registered:{},insert:a};return h.sheet.hydrate(s),h};function Fn(e,n,t){var o="";return t.split(" ").forEach((function(t){void 0!==e[t]?n.push(e[t]+";"):t&&(o+=t+" ")})),o}var Pn=function(e,n,t){var o=e.key+"-"+n.name;!1===t&&void 0===e.registered[o]&&(e.registered[o]=n.styles)},Dn=function(e,n,t){Pn(e,n,t);var o=e.key+"-"+n.name;if(void 0===e.inserted[n.name]){var a=n;do{e.insert(n===a?"."+o:"",a,e.sheet,!0),a=a.next}while(void 0!==a)}};var On={animationIterationCount:1,aspectRatio:1,borderImageOutset:1,borderImageSlice:1,borderImageWidth:1,boxFlex:1,boxFlexGroup:1,boxOrdinalGroup:1,columnCount:1,columns:1,flex:1,flexGrow:1,flexPositive:1,flexShrink:1,flexNegative:1,flexOrder:1,gridRow:1,gridRowEnd:1,gridRowSpan:1,gridRowStart:1,gridColumn:1,gridColumnEnd:1,gridColumnSpan:1,gridColumnStart:1,msGridRow:1,msGridRowSpan:1,msGridColumn:1,msGridColumnSpan:1,fontWeight:1,lineHeight:1,opacity:1,order:1,orphans:1,scale:1,tabSize:1,widows:1,zIndex:1,zoom:1,WebkitLineClamp:1,fillOpacity:1,floodOpacity:1,stopOpacity:1,strokeDasharray:1,strokeDashoffset:1,strokeMiterlimit:1,strokeOpacity:1,strokeWidth:1};function Mn(e){var n=Object.create(null);return function(t){return void 0===n[t]&&(n[t]=e(t)),n[t]}}var Ln=!1,Gn=/[A-Z]|^ms/g,Vn=/_EMO_([^_]+?)_([^]*?)_EMO_/g,qn=function(e){return 45===e.charCodeAt(1)},Hn=function(e){return null!=e&&"boolean"!==typeof e},zn=Mn((function(e){return qn(e)?e:e.replace(Gn,"-$&").toLowerCase()})),Un=function(e,n){switch(e){case"animation":case"animationName":if("string"===typeof n)return n.replace(Vn,(function(e,n,t){return Kn={name:n,styles:t,next:Kn},n}))}return 1===On[e]||qn(e)||"number"!==typeof n||0===n?n:n+"px"},Qn="Component selectors can only be used in conjunction with @emotion/babel-plugin, the swc Emotion plugin, or another Emotion-aware compiler transform.";function Wn(e,n,t){if(null==t)return"";var o=t;if(void 0!==o.__emotion_styles)return o;switch(typeof t){case"boolean":return"";case"object":var a=t;if(1===a.anim)return Kn={name:a.name,styles:a.styles,next:Kn},a.name;var i=t;if(void 0!==i.styles){var r=i.next;if(void 0!==r)for(;void 0!==r;)Kn={name:r.name,styles:r.styles,next:Kn},r=r.next;return i.styles+";"}return function(e,n,t){var o="";if(Array.isArray(t))for(var a=0;a<t.length;a++)o+=Wn(e,n,t[a])+";";else for(var i in t){var r=t[i];if("object"!==typeof r){var s=r;null!=n&&void 0!==n[s]?o+=i+"{"+n[s]+"}":Hn(s)&&(o+=zn(i)+":"+Un(i,s)+";")}else{if("NO_COMPONENT_SELECTOR"===i&&Ln)throw new Error(Qn);if(!Array.isArray(r)||"string"!==typeof r[0]||null!=n&&void 0!==n[r[0]]){var l=Wn(e,n,r);switch(i){case"animation":case"animationName":o+=zn(i)+":"+l+";";break;default:o+=i+"{"+l+"}"}}else for(var c=0;c<r.length;c++)Hn(r[c])&&(o+=zn(i)+":"+Un(i,r[c])+";")}}return o}(e,n,t);case"function":if(void 0!==e){var s=Kn,l=t(e);return Kn=s,Wn(e,n,l)}}var c=t;if(null==n)return c;var d=n[c];return void 0!==d?d:c}var Kn,$n=/label:\s*([^\s;{]+)\s*(;|$)/g;function Yn(e,n,t){if(1===e.length&&"object"===typeof e[0]&&null!==e[0]&&void 0!==e[0].styles)return e[0];var o=!0,a="";Kn=void 0;var i=e[0];null==i||void 0===i.raw?(o=!1,a+=Wn(t,n,i)):a+=i[0];for(var r=1;r<e.length;r++){if(a+=Wn(t,n,e[r]),o)a+=i[r]}$n.lastIndex=0;for(var s,l="";null!==(s=$n.exec(a));)l+="-"+s[1];var c=function(e){for(var n,t=0,o=0,a=e.length;a>=4;++o,a-=4)n=1540483477*(65535&(n=255&e.charCodeAt(o)|(255&e.charCodeAt(++o))<<8|(255&e.charCodeAt(++o))<<16|(255&e.charCodeAt(++o))<<24))+(59797*(n>>>16)<<16),t=1540483477*(65535&(n^=n>>>24))+(59797*(n>>>16)<<16)^1540483477*(65535&t)+(59797*(t>>>16)<<16);switch(a){case 3:t^=(255&e.charCodeAt(o+2))<<16;case 2:t^=(255&e.charCodeAt(o+1))<<8;case 1:t=1540483477*(65535&(t^=255&e.charCodeAt(o)))+(59797*(t>>>16)<<16)}return(((t=1540483477*(65535&(t^=t>>>13))+(59797*(t>>>16)<<16))^t>>>15)>>>0).toString(36)}(a)+l;return{name:c,styles:a,next:Kn}}var Xn=!!i.useInsertionEffect&&i.useInsertionEffect,Zn=Xn||function(e){return e()},Jn=(Xn||a.useLayoutEffect,a.createContext("undefined"!==typeof HTMLElement?Rn({key:"css"}):null)),et=(Jn.Provider,function(e){return(0,a.forwardRef)((function(n,t){var o=(0,a.useContext)(Jn);return e(n,o,t)}))}),nt=a.createContext({});var tt={}.hasOwnProperty,ot="__EMOTION_TYPE_PLEASE_DO_NOT_USE__",at=function(e){var n=e.cache,t=e.serialized,o=e.isStringTag;return Pn(n,t,o),Zn((function(){return Dn(n,t,o)})),null},it=et((function(e,n,t){var o=e.css;"string"===typeof o&&void 0!==n.registered[o]&&(o=n.registered[o]);var i=e[ot],r=[o],s="";"string"===typeof e.className?s=Fn(n.registered,r,e.className):null!=e.className&&(s=e.className+" ");var l=Yn(r,void 0,a.useContext(nt));s+=n.key+"-"+l.name;var c={};for(var d in e)tt.call(e,d)&&"css"!==d&&d!==ot&&(c[d]=e[d]);return c.className=s,t&&(c.ref=t),a.createElement(a.Fragment,null,a.createElement(at,{cache:n,serialized:l,isStringTag:"string"===typeof i}),a.createElement(i,c))})),rt=/^((children|dangerouslySetInnerHTML|key|ref|autoFocus|defaultValue|defaultChecked|innerHTML|suppressContentEditableWarning|suppressHydrationWarning|valueLink|abbr|accept|acceptCharset|accessKey|action|allow|allowUserMedia|allowPaymentRequest|allowFullScreen|allowTransparency|alt|async|autoComplete|autoPlay|capture|cellPadding|cellSpacing|challenge|charSet|checked|cite|classID|className|cols|colSpan|content|contentEditable|contextMenu|controls|controlsList|coords|crossOrigin|data|dateTime|decoding|default|defer|dir|disabled|disablePictureInPicture|disableRemotePlayback|download|draggable|encType|enterKeyHint|fetchpriority|fetchPriority|form|formAction|formEncType|formMethod|formNoValidate|formTarget|frameBorder|headers|height|hidden|high|href|hrefLang|htmlFor|httpEquiv|id|inputMode|integrity|is|keyParams|keyType|kind|label|lang|list|loading|loop|low|marginHeight|marginWidth|max|maxLength|media|mediaGroup|method|min|minLength|multiple|muted|name|nonce|noValidate|open|optimum|pattern|placeholder|playsInline|poster|preload|profile|radioGroup|readOnly|referrerPolicy|rel|required|reversed|role|rows|rowSpan|sandbox|scope|scoped|scrolling|seamless|selected|shape|size|sizes|slot|span|spellCheck|src|srcDoc|srcLang|srcSet|start|step|style|summary|tabIndex|target|title|translate|type|useMap|value|width|wmode|wrap|about|datatype|inlist|prefix|property|resource|typeof|vocab|autoCapitalize|autoCorrect|autoSave|color|incremental|fallback|inert|itemProp|itemScope|itemType|itemID|itemRef|on|option|results|security|unselectable|accentHeight|accumulate|additive|alignmentBaseline|allowReorder|alphabetic|amplitude|arabicForm|ascent|attributeName|attributeType|autoReverse|azimuth|baseFrequency|baselineShift|baseProfile|bbox|begin|bias|by|calcMode|capHeight|clip|clipPathUnits|clipPath|clipRule|colorInterpolation|colorInterpolationFilters|colorProfile|colorRendering|contentScriptType|contentStyleType|cursor|cx|cy|d|decelerate|descent|diffuseConstant|direction|display|divisor|dominantBaseline|dur|dx|dy|edgeMode|elevation|enableBackground|end|exponent|externalResourcesRequired|fill|fillOpacity|fillRule|filter|filterRes|filterUnits|floodColor|floodOpacity|focusable|fontFamily|fontSize|fontSizeAdjust|fontStretch|fontStyle|fontVariant|fontWeight|format|from|fr|fx|fy|g1|g2|glyphName|glyphOrientationHorizontal|glyphOrientationVertical|glyphRef|gradientTransform|gradientUnits|hanging|horizAdvX|horizOriginX|ideographic|imageRendering|in|in2|intercept|k|k1|k2|k3|k4|kernelMatrix|kernelUnitLength|kerning|keyPoints|keySplines|keyTimes|lengthAdjust|letterSpacing|lightingColor|limitingConeAngle|local|markerEnd|markerMid|markerStart|markerHeight|markerUnits|markerWidth|mask|maskContentUnits|maskUnits|mathematical|mode|numOctaves|offset|opacity|operator|order|orient|orientation|origin|overflow|overlinePosition|overlineThickness|panose1|paintOrder|pathLength|patternContentUnits|patternTransform|patternUnits|pointerEvents|points|pointsAtX|pointsAtY|pointsAtZ|preserveAlpha|preserveAspectRatio|primitiveUnits|r|radius|refX|refY|renderingIntent|repeatCount|repeatDur|requiredExtensions|requiredFeatures|restart|result|rotate|rx|ry|scale|seed|shapeRendering|slope|spacing|specularConstant|specularExponent|speed|spreadMethod|startOffset|stdDeviation|stemh|stemv|stitchTiles|stopColor|stopOpacity|strikethroughPosition|strikethroughThickness|string|stroke|strokeDasharray|strokeDashoffset|strokeLinecap|strokeLinejoin|strokeMiterlimit|strokeOpacity|strokeWidth|surfaceScale|systemLanguage|tableValues|targetX|targetY|textAnchor|textDecoration|textRendering|textLength|to|transform|u1|u2|underlinePosition|underlineThickness|unicode|unicodeBidi|unicodeRange|unitsPerEm|vAlphabetic|vHanging|vIdeographic|vMathematical|values|vectorEffect|version|vertAdvY|vertOriginX|vertOriginY|viewBox|viewTarget|visibility|widths|wordSpacing|writingMode|x|xHeight|x1|x2|xChannelSelector|xlinkActuate|xlinkArcrole|xlinkHref|xlinkRole|xlinkShow|xlinkTitle|xlinkType|xmlBase|xmlns|xmlnsXlink|xmlLang|xmlSpace|y|y1|y2|yChannelSelector|z|zoomAndPan|for|class|autofocus)|(([Dd][Aa][Tt][Aa]|[Aa][Rr][Ii][Aa]|x)-.*))$/,st=Mn((function(e){return rt.test(e)||111===e.charCodeAt(0)&&110===e.charCodeAt(1)&&e.charCodeAt(2)<91})),lt=function(e){return"theme"!==e},ct=function(e){return"string"===typeof e&&e.charCodeAt(0)>96?st:lt},dt=function(e,n,t){var o;if(n){var a=n.shouldForwardProp;o=e.__emotion_forwardProp&&a?function(n){return e.__emotion_forwardProp(n)&&a(n)}:a}return"function"!==typeof o&&t&&(o=e.__emotion_forwardProp),o},ut=function(e){var n=e.cache,t=e.serialized,o=e.isStringTag;return Pn(n,t,o),Zn((function(){return Dn(n,t,o)})),null},ht=function e(n,t){var o,i,r=n.__emotion_real===n,s=r&&n.__emotion_base||n;void 0!==t&&(o=t.label,i=t.target);var l=dt(n,t,r),c=l||ct(s),d=!c("as");return function(){var u=arguments,h=r&&void 0!==n.__emotion_styles?n.__emotion_styles.slice(0):[];if(void 0!==o&&h.push("label:"+o+";"),null==u[0]||void 0===u[0].raw)h.push.apply(h,u);else{var p=u[0];h.push(p[0]);for(var m=u.length,f=1;f<m;f++)h.push(u[f],p[f])}var g=et((function(e,n,t){var o=d&&e.as||s,r="",u=[],p=e;if(null==e.theme){for(var m in p={},e)p[m]=e[m];p.theme=a.useContext(nt)}"string"===typeof e.className?r=Fn(n.registered,u,e.className):null!=e.className&&(r=e.className+" ");var f=Yn(h.concat(u),n.registered,p);r+=n.key+"-"+f.name,void 0!==i&&(r+=" "+i);var g=d&&void 0===l?ct(o):c,y={};for(var x in e)d&&"as"===x||g(x)&&(y[x]=e[x]);return y.className=r,t&&(y.ref=t),a.createElement(a.Fragment,null,a.createElement(ut,{cache:n,serialized:f,isStringTag:"string"===typeof o}),a.createElement(o,y))}));return g.displayName=void 0!==o?o:"Styled("+("string"===typeof s?s:s.displayName||s.name||"Component")+")",g.defaultProps=n.defaultProps,g.__emotion_real=g,g.__emotion_base=s,g.__emotion_styles=h,g.__emotion_forwardProp=l,Object.defineProperty(g,"toString",{value:function(){return"."+i}}),g.withComponent=function(n,o){return e(n,Ee({},t,o,{shouldForwardProp:dt(g,o,!0)})).apply(void 0,h)},g}}.bind(null);function pt(e,n){return ht(e,n)}["a","abbr","address","area","article","aside","audio","b","base","bdi","bdo","big","blockquote","body","br","button","canvas","caption","cite","code","col","colgroup","data","datalist","dd","del","details","dfn","dialog","div","dl","dt","em","embed","fieldset","figcaption","figure","footer","form","h1","h2","h3","h4","h5","h6","head","header","hgroup","hr","html","i","iframe","img","input","ins","kbd","keygen","label","legend","li","link","main","map","mark","marquee","menu","menuitem","meta","meter","nav","noscript","object","ol","optgroup","option","output","p","param","picture","pre","progress","q","rp","rt","ruby","s","samp","script","section","select","small","source","span","strong","style","sub","summary","sup","table","tbody","td","textarea","tfoot","th","thead","time","title","tr","track","u","ul","var","video","wbr","circle","clipPath","defs","ellipse","foreignObject","g","image","line","linearGradient","mask","path","pattern","polygon","polyline","radialGradient","rect","stop","svg","text","tspan"].forEach((function(e){ht[e]=ht(e)}));const mt=[];function ft(e){return mt[0]=e,Yn(mt)}function gt(e){const n=new URL("https://mui.com/production-error/?code=".concat(e));for(var t=arguments.length,o=new Array(t>1?t-1:0),a=1;a<t;a++)o[a-1]=arguments[a];return o.forEach((e=>n.searchParams.append("args[]",e))),"Minified MUI error #".concat(e,"; visit ").concat(n," for the full message.")}function yt(e){if("string"!==typeof e)throw new Error(gt(7));return e.charAt(0).toUpperCase()+e.slice(1)}function xt(e){if("object"!==typeof e||null===e)return!1;const n=Object.getPrototypeOf(e);return(null===n||n===Object.prototype||null===Object.getPrototypeOf(n))&&!(Symbol.toStringTag in e)&&!(Symbol.iterator in e)}function bt(e){if(a.isValidElement(e)||!xt(e))return e;const n={};return Object.keys(e).forEach((t=>{n[t]=bt(e[t])})),n}function vt(e,n){let t=arguments.length>2&&void 0!==arguments[2]?arguments[2]:{clone:!0};const o=t.clone?{...e}:e;return xt(e)&&xt(n)&&Object.keys(n).forEach((i=>{a.isValidElement(n[i])?o[i]=n[i]:xt(n[i])&&Object.prototype.hasOwnProperty.call(e,i)&&xt(e[i])?o[i]=vt(e[i],n[i],t):t.clone?o[i]=xt(n[i])?bt(n[i]):n[i]:o[i]=n[i]})),o}const jt=function(e,n){return n?vt(e,n,{clone:!1}):e};const Ct={xs:0,sm:600,md:900,lg:1200,xl:1536},wt={keys:["xs","sm","md","lg","xl"],up:e=>"@media (min-width:".concat(Ct[e],"px)")},St={containerQueries:e=>({up:n=>{let t="number"===typeof n?n:Ct[n]||n;return"number"===typeof t&&(t="".concat(t,"px")),e?"@container ".concat(e," (min-width:").concat(t,")"):"@container (min-width:".concat(t,")")}})};function kt(e,n,t){const o=e.theme||{};if(Array.isArray(n)){const e=o.breakpoints||wt;return n.reduce(((o,a,i)=>(o[e.up(e.keys[i])]=t(n[i]),o)),{})}if("object"===typeof n){const e=o.breakpoints||wt;return Object.keys(n).reduce(((a,i)=>{if(function(e,n){return"@"===n||n.startsWith("@")&&(e.some((e=>n.startsWith("@".concat(e))))||!!n.match(/^@\d/))}(e.keys,i)){const e=function(e,n){const t=n.match(/^@([^/]+)?\/?(.+)?$/);if(!t)return null;const[,o,a]=t,i=Number.isNaN(+o)?o||0:+o;return e.containerQueries(a).up(i)}(o.containerQueries?o:St,i);e&&(a[e]=t(n[i],i))}else if(Object.keys(e.values||Ct).includes(i)){a[e.up(i)]=t(n[i],i)}else{const e=i;a[e]=n[e]}return a}),{})}return t(n)}function Nt(){var e;let n=arguments.length>0&&void 0!==arguments[0]?arguments[0]:{};return(null===(e=n.keys)||void 0===e?void 0:e.reduce(((e,t)=>(e[n.up(t)]={},e)),{}))||{}}function _t(e,n){return e.reduce(((e,n)=>{const t=e[n];return(!t||0===Object.keys(t).length)&&delete e[n],e}),n)}function Tt(e,n){let t=!(arguments.length>2&&void 0!==arguments[2])||arguments[2];if(!n||"string"!==typeof n)return null;if(e&&e.vars&&t){const t="vars.".concat(n).split(".").reduce(((e,n)=>e&&e[n]?e[n]:null),e);if(null!=t)return t}return n.split(".").reduce(((e,n)=>e&&null!=e[n]?e[n]:null),e)}function It(e,n,t){let o,a=arguments.length>3&&void 0!==arguments[3]?arguments[3]:t;return o="function"===typeof e?e(t):Array.isArray(e)?e[t]||a:Tt(e,t)||a,n&&(o=n(o,a,e)),o}const Et=function(e){const{prop:n,cssProperty:t=e.prop,themeKey:o,transform:a}=e,i=e=>{if(null==e[n])return null;const i=e[n],r=Tt(e.theme,o)||{};return kt(e,i,(e=>{let o=It(r,a,e);return e===o&&"string"===typeof e&&(o=It(r,a,"".concat(n).concat("default"===e?"":yt(e)),e)),!1===t?o:{[t]:o}}))};return i.propTypes={},i.filterProps=[n],i};const At={m:"margin",p:"padding"},Bt={t:"Top",r:"Right",b:"Bottom",l:"Left",x:["Left","Right"],y:["Top","Bottom"]},Rt={marginX:"mx",marginY:"my",paddingX:"px",paddingY:"py"},Ft=function(e){const n={};return t=>(void 0===n[t]&&(n[t]=e(t)),n[t])}((e=>{if(e.length>2){if(!Rt[e])return[e];e=Rt[e]}const[n,t]=e.split(""),o=At[n],a=Bt[t]||"";return Array.isArray(a)?a.map((e=>o+e)):[o+a]})),Pt=["m","mt","mr","mb","ml","mx","my","margin","marginTop","marginRight","marginBottom","marginLeft","marginX","marginY","marginInline","marginInlineStart","marginInlineEnd","marginBlock","marginBlockStart","marginBlockEnd"],Dt=["p","pt","pr","pb","pl","px","py","padding","paddingTop","paddingRight","paddingBottom","paddingLeft","paddingX","paddingY","paddingInline","paddingInlineStart","paddingInlineEnd","paddingBlock","paddingBlockStart","paddingBlockEnd"],Ot=[...Pt,...Dt];function Mt(e,n,t,o){var a;const i=null!==(a=Tt(e,n,!0))&&void 0!==a?a:t;return"number"===typeof i||"string"===typeof i?e=>"string"===typeof e?e:"string"===typeof i?"calc(".concat(e," * ").concat(i,")"):i*e:Array.isArray(i)?e=>{if("string"===typeof e)return e;const n=Math.abs(e);const t=i[n];return e>=0?t:"number"===typeof t?-t:"-".concat(t)}:"function"===typeof i?i:()=>{}}function Lt(e){return Mt(e,"spacing",8)}function Gt(e,n){return"string"===typeof n||null==n?n:e(n)}function Vt(e,n,t,o){if(!n.includes(t))return null;const a=function(e,n){return t=>e.reduce(((e,o)=>(e[o]=Gt(n,t),e)),{})}(Ft(t),o);return kt(e,e[t],a)}function qt(e,n){const t=Lt(e.theme);return Object.keys(e).map((o=>Vt(e,n,o,t))).reduce(jt,{})}function Ht(e){return qt(e,Pt)}function zt(e){return qt(e,Dt)}function Ut(e){return qt(e,Ot)}Ht.propTypes={},Ht.filterProps=Pt,zt.propTypes={},zt.filterProps=Dt,Ut.propTypes={},Ut.filterProps=Ot;const Qt=function(){for(var e=arguments.length,n=new Array(e),t=0;t<e;t++)n[t]=arguments[t];const o=n.reduce(((e,n)=>(n.filterProps.forEach((t=>{e[t]=n})),e)),{}),a=e=>Object.keys(e).reduce(((n,t)=>o[t]?jt(n,o[t](e)):n),{});return a.propTypes={},a.filterProps=n.reduce(((e,n)=>e.concat(n.filterProps)),[]),a};function Wt(e){return"number"!==typeof e?e:"".concat(e,"px solid")}function Kt(e,n){return Et({prop:e,themeKey:"borders",transform:n})}const $t=Kt("border",Wt),Yt=Kt("borderTop",Wt),Xt=Kt("borderRight",Wt),Zt=Kt("borderBottom",Wt),Jt=Kt("borderLeft",Wt),eo=Kt("borderColor"),no=Kt("borderTopColor"),to=Kt("borderRightColor"),oo=Kt("borderBottomColor"),ao=Kt("borderLeftColor"),io=Kt("outline",Wt),ro=Kt("outlineColor"),so=e=>{if(void 0!==e.borderRadius&&null!==e.borderRadius){const n=Mt(e.theme,"shape.borderRadius",4),t=e=>({borderRadius:Gt(n,e)});return kt(e,e.borderRadius,t)}return null};so.propTypes={},so.filterProps=["borderRadius"];Qt($t,Yt,Xt,Zt,Jt,eo,no,to,oo,ao,so,io,ro);const lo=e=>{if(void 0!==e.gap&&null!==e.gap){const n=Mt(e.theme,"spacing",8),t=e=>({gap:Gt(n,e)});return kt(e,e.gap,t)}return null};lo.propTypes={},lo.filterProps=["gap"];const co=e=>{if(void 0!==e.columnGap&&null!==e.columnGap){const n=Mt(e.theme,"spacing",8),t=e=>({columnGap:Gt(n,e)});return kt(e,e.columnGap,t)}return null};co.propTypes={},co.filterProps=["columnGap"];const uo=e=>{if(void 0!==e.rowGap&&null!==e.rowGap){const n=Mt(e.theme,"spacing",8),t=e=>({rowGap:Gt(n,e)});return kt(e,e.rowGap,t)}return null};uo.propTypes={},uo.filterProps=["rowGap"];Qt(lo,co,uo,Et({prop:"gridColumn"}),Et({prop:"gridRow"}),Et({prop:"gridAutoFlow"}),Et({prop:"gridAutoColumns"}),Et({prop:"gridAutoRows"}),Et({prop:"gridTemplateColumns"}),Et({prop:"gridTemplateRows"}),Et({prop:"gridTemplateAreas"}),Et({prop:"gridArea"}));function ho(e,n){return"grey"===n?n:e}Qt(Et({prop:"color",themeKey:"palette",transform:ho}),Et({prop:"bgcolor",cssProperty:"backgroundColor",themeKey:"palette",transform:ho}),Et({prop:"backgroundColor",themeKey:"palette",transform:ho}));function po(e){return e<=1&&0!==e?"".concat(100*e,"%"):e}const mo=Et({prop:"width",transform:po}),fo=e=>{if(void 0!==e.maxWidth&&null!==e.maxWidth){const n=n=>{var t,o;const a=(null===(t=e.theme)||void 0===t||null===(t=t.breakpoints)||void 0===t||null===(t=t.values)||void 0===t?void 0:t[n])||Ct[n];return a?"px"!==(null===(o=e.theme)||void 0===o||null===(o=o.breakpoints)||void 0===o?void 0:o.unit)?{maxWidth:"".concat(a).concat(e.theme.breakpoints.unit)}:{maxWidth:a}:{maxWidth:po(n)}};return kt(e,e.maxWidth,n)}return null};fo.filterProps=["maxWidth"];const go=Et({prop:"minWidth",transform:po}),yo=Et({prop:"height",transform:po}),xo=Et({prop:"maxHeight",transform:po}),bo=Et({prop:"minHeight",transform:po}),vo=(Et({prop:"size",cssProperty:"width",transform:po}),Et({prop:"size",cssProperty:"height",transform:po}),Qt(mo,fo,go,yo,xo,bo,Et({prop:"boxSizing"})),{border:{themeKey:"borders",transform:Wt},borderTop:{themeKey:"borders",transform:Wt},borderRight:{themeKey:"borders",transform:Wt},borderBottom:{themeKey:"borders",transform:Wt},borderLeft:{themeKey:"borders",transform:Wt},borderColor:{themeKey:"palette"},borderTopColor:{themeKey:"palette"},borderRightColor:{themeKey:"palette"},borderBottomColor:{themeKey:"palette"},borderLeftColor:{themeKey:"palette"},outline:{themeKey:"borders",transform:Wt},outlineColor:{themeKey:"palette"},borderRadius:{themeKey:"shape.borderRadius",style:so},color:{themeKey:"palette",transform:ho},bgcolor:{themeKey:"palette",cssProperty:"backgroundColor",transform:ho},backgroundColor:{themeKey:"palette",transform:ho},p:{style:zt},pt:{style:zt},pr:{style:zt},pb:{style:zt},pl:{style:zt},px:{style:zt},py:{style:zt},padding:{style:zt},paddingTop:{style:zt},paddingRight:{style:zt},paddingBottom:{style:zt},paddingLeft:{style:zt},paddingX:{style:zt},paddingY:{style:zt},paddingInline:{style:zt},paddingInlineStart:{style:zt},paddingInlineEnd:{style:zt},paddingBlock:{style:zt},paddingBlockStart:{style:zt},paddingBlockEnd:{style:zt},m:{style:Ht},mt:{style:Ht},mr:{style:Ht},mb:{style:Ht},ml:{style:Ht},mx:{style:Ht},my:{style:Ht},margin:{style:Ht},marginTop:{style:Ht},marginRight:{style:Ht},marginBottom:{style:Ht},marginLeft:{style:Ht},marginX:{style:Ht},marginY:{style:Ht},marginInline:{style:Ht},marginInlineStart:{style:Ht},marginInlineEnd:{style:Ht},marginBlock:{style:Ht},marginBlockStart:{style:Ht},marginBlockEnd:{style:Ht},displayPrint:{cssProperty:!1,transform:e=>({"@media print":{display:e}})},display:{},overflow:{},textOverflow:{},visibility:{},whiteSpace:{},flexBasis:{},flexDirection:{},flexWrap:{},justifyContent:{},alignItems:{},alignContent:{},order:{},flex:{},flexGrow:{},flexShrink:{},alignSelf:{},justifyItems:{},justifySelf:{},gap:{style:lo},rowGap:{style:uo},columnGap:{style:co},gridColumn:{},gridRow:{},gridAutoFlow:{},gridAutoColumns:{},gridAutoRows:{},gridTemplateColumns:{},gridTemplateRows:{},gridTemplateAreas:{},gridArea:{},position:{},zIndex:{themeKey:"zIndex"},top:{},right:{},bottom:{},left:{},boxShadow:{themeKey:"shadows"},width:{transform:po},maxWidth:{style:fo},minWidth:{transform:po},height:{transform:po},maxHeight:{transform:po},minHeight:{transform:po},boxSizing:{},font:{themeKey:"font"},fontFamily:{themeKey:"typography"},fontSize:{themeKey:"typography"},fontStyle:{themeKey:"typography"},fontWeight:{themeKey:"typography"},letterSpacing:{},textTransform:{},lineHeight:{},textAlign:{},typography:{cssProperty:!1,themeKey:"typography"}});const jo=function(){function e(e,n,t,o){const a={[e]:n,theme:t},i=o[e];if(!i)return{[e]:n};const{cssProperty:r=e,themeKey:s,transform:l,style:c}=i;if(null==n)return null;if("typography"===s&&"inherit"===n)return{[e]:n};const d=Tt(t,s)||{};if(c)return c(a);return kt(a,n,(n=>{let t=It(d,l,n);return n===t&&"string"===typeof n&&(t=It(d,l,"".concat(e).concat("default"===n?"":yt(n)),n)),!1===r?t:{[r]:t}}))}return function n(t){var o;const{sx:a,theme:i={}}=t||{};if(!a)return null;const r=null!==(o=i.unstable_sxConfig)&&void 0!==o?o:vo;function s(t){let o=t;if("function"===typeof t)o=t(i);else if("object"!==typeof t)return t;if(!o)return null;const a=Nt(i.breakpoints),s=Object.keys(a);let l=a;return Object.keys(o).forEach((t=>{const a=function(e,n){return"function"===typeof e?e(n):e}(o[t],i);if(null!==a&&void 0!==a)if("object"===typeof a)if(r[t])l=jt(l,e(t,a,i,r));else{const e=kt({theme:i},a,(e=>({[t]:e})));!function(){for(var e=arguments.length,n=new Array(e),t=0;t<e;t++)n[t]=arguments[t];const o=n.reduce(((e,n)=>e.concat(Object.keys(n))),[]),a=new Set(o);return n.every((e=>a.size===Object.keys(e).length))}(e,a)?l=jt(l,e):l[t]=n({sx:a,theme:i})}else l=jt(l,e(t,a,i,r))})),function(e,n){if(!e.containerQueries)return n;const t=Object.keys(n).filter((e=>e.startsWith("@container"))).sort(((e,n)=>{var t,o;const a=/min-width:\s*([0-9.]+)/;return+((null===(t=e.match(a))||void 0===t?void 0:t[1])||0)-+((null===(o=n.match(a))||void 0===o?void 0:o[1])||0)}));return t.length?t.reduce(((e,t)=>{const o=n[t];return delete e[t],e[t]=o,e}),{...n}):n}(i,_t(s,l))}return Array.isArray(a)?a.map(s):s(a)}}();jo.filterProps=["sx"];const Co=jo,wo=e=>{var n,t;const o={systemProps:{},otherProps:{}},a=null!==(n=null===e||void 0===e||null===(t=e.theme)||void 0===t?void 0:t.unstable_sxConfig)&&void 0!==n?n:vo;return Object.keys(e).forEach((n=>{a[n]?o.systemProps[n]=e[n]:o.otherProps[n]=e[n]})),o};function So(e){const{sx:n,...t}=e,{systemProps:o,otherProps:a}=wo(t);let i;return i=Array.isArray(n)?[o,...n]:"function"===typeof n?function(){const e=n(...arguments);return xt(e)?{...o,...e}:o}:{...o,...n},{...a,sx:i}}const ko=e=>{const n=Object.keys(e).map((n=>({key:n,val:e[n]})))||[];return n.sort(((e,n)=>e.val-n.val)),n.reduce(((e,n)=>({...e,[n.key]:n.val})),{})};const No={borderRadius:4};function _o(){let e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:8,n=arguments.length>1&&void 0!==arguments[1]?arguments[1]:Lt({spacing:e});if(e.mui)return e;const t=function(){for(var e=arguments.length,t=new Array(e),o=0;o<e;o++)t[o]=arguments[o];return(0===t.length?[1]:t).map((e=>{const t=n(e);return"number"===typeof t?"".concat(t,"px"):t})).join(" ")};return t.mui=!0,t}function To(e,n){const t=this;if(t.vars){var o;if(null===(o=t.colorSchemes)||void 0===o||!o[e]||"function"!==typeof t.getColorSchemeSelector)return{};let a=t.getColorSchemeSelector(e);return"&"===a?n:((a.includes("data-")||a.includes("."))&&(a="*:where(".concat(a.replace(/\s*&$/,""),") &")),{[a]:n})}return t.palette.mode===e?n:{}}const Io=function(){let e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:{};const{breakpoints:n={},palette:t={},spacing:o,shape:a={},...i}=e,r=function(e){const{values:n={xs:0,sm:600,md:900,lg:1200,xl:1536},unit:t="px",step:o=5,...a}=e,i=ko(n),r=Object.keys(i);function s(e){const o="number"===typeof n[e]?n[e]:e;return"@media (min-width:".concat(o).concat(t,")")}function l(e){const a="number"===typeof n[e]?n[e]:e;return"@media (max-width:".concat(a-o/100).concat(t,")")}function c(e,a){const i=r.indexOf(a);return"@media (min-width:".concat("number"===typeof n[e]?n[e]:e).concat(t,") and ")+"(max-width:".concat((-1!==i&&"number"===typeof n[r[i]]?n[r[i]]:a)-o/100).concat(t,")")}return{keys:r,values:i,up:s,down:l,between:c,only:function(e){return r.indexOf(e)+1<r.length?c(e,r[r.indexOf(e)+1]):s(e)},not:function(e){const n=r.indexOf(e);return 0===n?s(r[1]):n===r.length-1?l(r[n]):c(e,r[r.indexOf(e)+1]).replace("@media","@media not all and")},unit:t,...a}}(n);let s=vt({breakpoints:r,direction:"ltr",components:{},palette:{mode:"light",...t},spacing:_o(o),shape:{...No,...a}},i);s=function(e){const n=(e,n)=>e.replace("@media",n?"@container ".concat(n):"@container");function t(t,o){t.up=function(){return n(e.breakpoints.up(...arguments),o)},t.down=function(){return n(e.breakpoints.down(...arguments),o)},t.between=function(){return n(e.breakpoints.between(...arguments),o)},t.only=function(){return n(e.breakpoints.only(...arguments),o)},t.not=function(){const t=n(e.breakpoints.not(...arguments),o);return t.includes("not all and")?t.replace("not all and ","").replace("min-width:","width<").replace("max-width:","width>").replace("and","or"):t}}const o={},a=e=>(t(o,e),o);return t(a),{...e,containerQueries:a}}(s),s.applyStyles=To;for(var l=arguments.length,c=new Array(l>1?l-1:0),d=1;d<l;d++)c[d-1]=arguments[d];return s=c.reduce(((e,n)=>vt(e,n)),s),s.unstable_sxConfig={...vo,...null===i||void 0===i?void 0:i.unstable_sxConfig},s.unstable_sx=function(e){return Co({sx:e,theme:this})},s};const Eo=function(){let e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:null;const n=a.useContext(nt);return n&&(t=n,0!==Object.keys(t).length)?n:e;var t},Ao=Io();const Bo=function(){return Eo(arguments.length>0&&void 0!==arguments[0]?arguments[0]:Ao)};const Ro=e=>e,Fo=(()=>{let e=Ro;return{configure(n){e=n},generate:n=>e(n),reset(){e=Ro}}})();const Po=function(e){let n=arguments.length>1&&void 0!==arguments[1]?arguments[1]:Number.MIN_SAFE_INTEGER,t=arguments.length>2&&void 0!==arguments[2]?arguments[2]:Number.MAX_SAFE_INTEGER;return Math.max(n,Math.min(e,t))};function Do(e){return Po(e,arguments.length>1&&void 0!==arguments[1]?arguments[1]:0,arguments.length>2&&void 0!==arguments[2]?arguments[2]:1)}function Oo(e){if(e.type)return e;if("#"===e.charAt(0))return Oo(function(e){e=e.slice(1);const n=new RegExp(".{1,".concat(e.length>=6?2:1,"}"),"g");let t=e.match(n);return t&&1===t[0].length&&(t=t.map((e=>e+e))),t?"rgb".concat(4===t.length?"a":"","(").concat(t.map(((e,n)=>n<3?parseInt(e,16):Math.round(parseInt(e,16)/255*1e3)/1e3)).join(", "),")"):""}(e));const n=e.indexOf("("),t=e.substring(0,n);if(!["rgb","rgba","hsl","hsla","color"].includes(t))throw new Error(gt(9,e));let o,a=e.substring(n+1,e.length-1);if("color"===t){if(a=a.split(" "),o=a.shift(),4===a.length&&"/"===a[3].charAt(0)&&(a[3]=a[3].slice(1)),!["srgb","display-p3","a98-rgb","prophoto-rgb","rec-2020"].includes(o))throw new Error(gt(10,o))}else a=a.split(",");return a=a.map((e=>parseFloat(e))),{type:t,values:a,colorSpace:o}}const Mo=(e,n)=>{try{return(e=>{const n=Oo(e);return n.values.slice(0,3).map(((e,t)=>n.type.includes("hsl")&&0!==t?"".concat(e,"%"):e)).join(" ")})(e)}catch(t){return e}};function Lo(e){const{type:n,colorSpace:t}=e;let{values:o}=e;return n.includes("rgb")?o=o.map(((e,n)=>n<3?parseInt(e,10):e)):n.includes("hsl")&&(o[1]="".concat(o[1],"%"),o[2]="".concat(o[2],"%")),o=n.includes("color")?"".concat(t," ").concat(o.join(" ")):"".concat(o.join(", ")),"".concat(n,"(").concat(o,")")}function Go(e){e=Oo(e);const{values:n}=e,t=n[0],o=n[1]/100,a=n[2]/100,i=o*Math.min(a,1-a),r=function(e){let n=arguments.length>1&&void 0!==arguments[1]?arguments[1]:(e+t/30)%12;return a-i*Math.max(Math.min(n-3,9-n,1),-1)};let s="rgb";const l=[Math.round(255*r(0)),Math.round(255*r(8)),Math.round(255*r(4))];return"hsla"===e.type&&(s+="a",l.push(n[3])),Lo({type:s,values:l})}function Vo(e){let n="hsl"===(e=Oo(e)).type||"hsla"===e.type?Oo(Go(e)).values:e.values;return n=n.map((n=>("color"!==e.type&&(n/=255),n<=.03928?n/12.92:((n+.055)/1.055)**2.4))),Number((.2126*n[0]+.7152*n[1]+.0722*n[2]).toFixed(3))}function qo(e,n){return e=Oo(e),n=Do(n),"rgb"!==e.type&&"hsl"!==e.type||(e.type+="a"),"color"===e.type?e.values[3]="/".concat(n):e.values[3]=n,Lo(e)}function Ho(e,n,t){try{return qo(e,n)}catch(o){return e}}function zo(e,n){if(e=Oo(e),n=Do(n),e.type.includes("hsl"))e.values[2]*=1-n;else if(e.type.includes("rgb")||e.type.includes("color"))for(let t=0;t<3;t+=1)e.values[t]*=1-n;return Lo(e)}function Uo(e,n,t){try{return zo(e,n)}catch(o){return e}}function Qo(e,n){if(e=Oo(e),n=Do(n),e.type.includes("hsl"))e.values[2]+=(100-e.values[2])*n;else if(e.type.includes("rgb"))for(let t=0;t<3;t+=1)e.values[t]+=(255-e.values[t])*n;else if(e.type.includes("color"))for(let t=0;t<3;t+=1)e.values[t]+=(1-e.values[t])*n;return Lo(e)}function Wo(e,n,t){try{return Qo(e,n)}catch(o){return e}}function Ko(e,n,t){try{return function(e){let n=arguments.length>1&&void 0!==arguments[1]?arguments[1]:.15;return Vo(e)>.5?zo(e,n):Qo(e,n)}(e,n)}catch(o){return e}}const $o={black:"#000",white:"#fff"},Yo={50:"#fafafa",100:"#f5f5f5",200:"#eeeeee",300:"#e0e0e0",400:"#bdbdbd",500:"#9e9e9e",600:"#757575",700:"#616161",800:"#424242",900:"#212121",A100:"#f5f5f5",A200:"#eeeeee",A400:"#bdbdbd",A700:"#616161"},Xo={50:"#f3e5f5",100:"#e1bee7",200:"#ce93d8",300:"#ba68c8",400:"#ab47bc",500:"#9c27b0",600:"#8e24aa",700:"#7b1fa2",800:"#6a1b9a",900:"#4a148c",A100:"#ea80fc",A200:"#e040fb",A400:"#d500f9",A700:"#aa00ff"},Zo={50:"#ffebee",100:"#ffcdd2",200:"#ef9a9a",300:"#e57373",400:"#ef5350",500:"#f44336",600:"#e53935",700:"#d32f2f",800:"#c62828",900:"#b71c1c",A100:"#ff8a80",A200:"#ff5252",A400:"#ff1744",A700:"#d50000"},Jo={50:"#fff3e0",100:"#ffe0b2",200:"#ffcc80",300:"#ffb74d",400:"#ffa726",500:"#ff9800",600:"#fb8c00",700:"#f57c00",800:"#ef6c00",900:"#e65100",A100:"#ffd180",A200:"#ffab40",A400:"#ff9100",A700:"#ff6d00"},ea={50:"#e3f2fd",100:"#bbdefb",200:"#90caf9",300:"#64b5f6",400:"#42a5f5",500:"#2196f3",600:"#1e88e5",700:"#1976d2",800:"#1565c0",900:"#0d47a1",A100:"#82b1ff",A200:"#448aff",A400:"#2979ff",A700:"#2962ff"},na={50:"#e1f5fe",100:"#b3e5fc",200:"#81d4fa",300:"#4fc3f7",400:"#29b6f6",500:"#03a9f4",600:"#039be5",700:"#0288d1",800:"#0277bd",900:"#01579b",A100:"#80d8ff",A200:"#40c4ff",A400:"#00b0ff",A700:"#0091ea"},ta={50:"#e8f5e9",100:"#c8e6c9",200:"#a5d6a7",300:"#81c784",400:"#66bb6a",500:"#4caf50",600:"#43a047",700:"#388e3c",800:"#2e7d32",900:"#1b5e20",A100:"#b9f6ca",A200:"#69f0ae",A400:"#00e676",A700:"#00c853"};function oa(){return{text:{primary:"rgba(0, 0, 0, 0.87)",secondary:"rgba(0, 0, 0, 0.6)",disabled:"rgba(0, 0, 0, 0.38)"},divider:"rgba(0, 0, 0, 0.12)",background:{paper:$o.white,default:$o.white},action:{active:"rgba(0, 0, 0, 0.54)",hover:"rgba(0, 0, 0, 0.04)",hoverOpacity:.04,selected:"rgba(0, 0, 0, 0.08)",selectedOpacity:.08,disabled:"rgba(0, 0, 0, 0.26)",disabledBackground:"rgba(0, 0, 0, 0.12)",disabledOpacity:.38,focus:"rgba(0, 0, 0, 0.12)",focusOpacity:.12,activatedOpacity:.12}}}const aa=oa();function ia(){return{text:{primary:$o.white,secondary:"rgba(255, 255, 255, 0.7)",disabled:"rgba(255, 255, 255, 0.5)",icon:"rgba(255, 255, 255, 0.5)"},divider:"rgba(255, 255, 255, 0.12)",background:{paper:"#121212",default:"#121212"},action:{active:$o.white,hover:"rgba(255, 255, 255, 0.08)",hoverOpacity:.08,selected:"rgba(255, 255, 255, 0.16)",selectedOpacity:.16,disabled:"rgba(255, 255, 255, 0.3)",disabledBackground:"rgba(255, 255, 255, 0.12)",disabledOpacity:.38,focus:"rgba(255, 255, 255, 0.12)",focusOpacity:.12,activatedOpacity:.24}}}const ra=ia();function sa(e,n,t,o){const a=o.light||o,i=o.dark||1.5*o;e[n]||(e.hasOwnProperty(t)?e[n]=e[t]:"light"===n?e.light=Qo(e.main,a):"dark"===n&&(e.dark=zo(e.main,i)))}function la(e){const{mode:n="light",contrastThreshold:t=3,tonalOffset:o=.2,...a}=e,i=e.primary||function(){return"dark"===(arguments.length>0&&void 0!==arguments[0]?arguments[0]:"light")?{main:ea[200],light:ea[50],dark:ea[400]}:{main:ea[700],light:ea[400],dark:ea[800]}}(n),r=e.secondary||function(){return"dark"===(arguments.length>0&&void 0!==arguments[0]?arguments[0]:"light")?{main:Xo[200],light:Xo[50],dark:Xo[400]}:{main:Xo[500],light:Xo[300],dark:Xo[700]}}(n),s=e.error||function(){return"dark"===(arguments.length>0&&void 0!==arguments[0]?arguments[0]:"light")?{main:Zo[500],light:Zo[300],dark:Zo[700]}:{main:Zo[700],light:Zo[400],dark:Zo[800]}}(n),l=e.info||function(){return"dark"===(arguments.length>0&&void 0!==arguments[0]?arguments[0]:"light")?{main:na[400],light:na[300],dark:na[700]}:{main:na[700],light:na[500],dark:na[900]}}(n),c=e.success||function(){return"dark"===(arguments.length>0&&void 0!==arguments[0]?arguments[0]:"light")?{main:ta[400],light:ta[300],dark:ta[700]}:{main:ta[800],light:ta[500],dark:ta[900]}}(n),d=e.warning||function(){return"dark"===(arguments.length>0&&void 0!==arguments[0]?arguments[0]:"light")?{main:Jo[400],light:Jo[300],dark:Jo[700]}:{main:"#ed6c02",light:Jo[500],dark:Jo[900]}}(n);function u(e){const n=function(e,n){const t=Vo(e),o=Vo(n);return(Math.max(t,o)+.05)/(Math.min(t,o)+.05)}(e,ra.text.primary)>=t?ra.text.primary:aa.text.primary;return n}const h=e=>{let{color:n,name:t,mainShade:a=500,lightShade:i=300,darkShade:r=700}=e;if(n={...n},!n.main&&n[a]&&(n.main=n[a]),!n.hasOwnProperty("main"))throw new Error(gt(11,t?" (".concat(t,")"):"",a));if("string"!==typeof n.main)throw new Error(gt(12,t?" (".concat(t,")"):"",JSON.stringify(n.main)));return sa(n,"light",i,o),sa(n,"dark",r,o),n.contrastText||(n.contrastText=u(n.main)),n};let p;"light"===n?p=oa():"dark"===n&&(p=ia());return vt({common:{...$o},mode:n,primary:h({color:i,name:"primary"}),secondary:h({color:r,name:"secondary",mainShade:"A400",lightShade:"A200",darkShade:"A700"}),error:h({color:s,name:"error"}),warning:h({color:d,name:"warning"}),info:h({color:l,name:"info"}),success:h({color:c,name:"success"}),grey:Yo,contrastThreshold:t,getContrastText:u,augmentColor:h,tonalOffset:o,...p},a)}function ca(){let e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:"";function n(){for(var t=arguments.length,o=new Array(t),a=0;a<t;a++)o[a]=arguments[a];if(!o.length)return"";const i=o[0];return"string"!==typeof i||i.match(/(#|\(|\)|(-?(\d*\.)?\d+)(px|em|%|ex|ch|rem|vw|vh|vmin|vmax|cm|mm|in|pt|pc))|^(-?(\d*\.)?\d+)$|(\d+ \d+ \d+)/)?", ".concat(i):", var(--".concat(e?"".concat(e,"-"):"").concat(i).concat(n(...o.slice(1)),")")}return function(t){for(var o=arguments.length,a=new Array(o>1?o-1:0),i=1;i<o;i++)a[i-1]=arguments[i];return"var(--".concat(e?"".concat(e,"-"):"").concat(t).concat(n(...a),")")}}function da(e){const n={};return Object.entries(e).forEach((e=>{const[t,o]=e;"object"===typeof o&&(n[t]="".concat(o.fontStyle?"".concat(o.fontStyle," "):"").concat(o.fontVariant?"".concat(o.fontVariant," "):"").concat(o.fontWeight?"".concat(o.fontWeight," "):"").concat(o.fontStretch?"".concat(o.fontStretch," "):"").concat(o.fontSize||"").concat(o.lineHeight?"/".concat(o.lineHeight," "):"").concat(o.fontFamily||""))})),n}const ua=function(e,n,t){let o=arguments.length>3&&void 0!==arguments[3]?arguments[3]:[],a=e;n.forEach(((e,i)=>{i===n.length-1?Array.isArray(a)?a[Number(e)]=t:a&&"object"===typeof a&&(a[e]=t):a&&"object"===typeof a&&(a[e]||(a[e]=o.includes(e)?[]:{}),a=a[e])}))},ha=(e,n,t)=>{!function e(o){let a=arguments.length>1&&void 0!==arguments[1]?arguments[1]:[],i=arguments.length>2&&void 0!==arguments[2]?arguments[2]:[];Object.entries(o).forEach((o=>{let[r,s]=o;(!t||t&&!t([...a,r]))&&void 0!==s&&null!==s&&("object"===typeof s&&Object.keys(s).length>0?e(s,[...a,r],Array.isArray(s)?[...i,r]:i):n([...a,r],s,i))}))}(e)},pa=(e,n)=>{if("number"===typeof n){if(["lineHeight","fontWeight","opacity","zIndex"].some((n=>e.includes(n))))return n;return e[e.length-1].toLowerCase().includes("opacity")?n:"".concat(n,"px")}return n};function ma(e,n){const{prefix:t,shouldSkipGeneratingVar:o}=n||{},a={},i={},r={};return ha(e,((e,n,s)=>{if(("string"===typeof n||"number"===typeof n)&&(!o||!o(e,n))){const o="--".concat(t?"".concat(t,"-"):"").concat(e.join("-")),l=pa(e,n);Object.assign(a,{[o]:l}),ua(i,e,"var(".concat(o,")"),s),ua(r,e,"var(".concat(o,", ").concat(l,")"),s)}}),(e=>"vars"===e[0])),{css:a,vars:i,varsWithDefaults:r}}const fa=function(e){let n=arguments.length>1&&void 0!==arguments[1]?arguments[1]:{};const{getSelector:t=g,disableCssColorScheme:o,colorSchemeSelector:a}=n,{colorSchemes:i={},components:r,defaultColorScheme:s="light",...l}=e,{vars:c,css:d,varsWithDefaults:u}=ma(l,n);let h=u;const p={},{[s]:m,...f}=i;if(Object.entries(f||{}).forEach((e=>{let[t,o]=e;const{vars:a,css:i,varsWithDefaults:r}=ma(o,n);h=vt(h,r),p[t]={css:i,vars:a}})),m){const{css:e,vars:t,varsWithDefaults:o}=ma(m,n);h=vt(h,o),p[s]={css:e,vars:t}}function g(n,t){let o=a;if("class"===a&&(o=".%s"),"data"===a&&(o="[data-%s]"),null!==a&&void 0!==a&&a.startsWith("data-")&&!a.includes("%s")&&(o="[".concat(a,'="%s"]')),n){if("media"===o){var r;if(e.defaultColorScheme===n)return":root";const o=(null===(r=i[n])||void 0===r||null===(r=r.palette)||void 0===r?void 0:r.mode)||n;return{["@media (prefers-color-scheme: ".concat(o,")")]:{":root":t}}}if(o)return e.defaultColorScheme===n?":root, ".concat(o.replace("%s",String(n))):o.replace("%s",String(n))}return":root"}return{vars:h,generateThemeVars:()=>{let e={...c};return Object.entries(p).forEach((n=>{let[,{vars:t}]=n;e=vt(e,t)})),e},generateStyleSheets:()=>{const n=[],a=e.defaultColorScheme||"light";function r(e,t){Object.keys(t).length&&n.push("string"===typeof e?{[e]:{...t}}:e)}r(t(void 0,{...d}),d);const{[a]:s,...l}=p;if(s){var c;const{css:e}=s,n=null===(c=i[a])||void 0===c||null===(c=c.palette)||void 0===c?void 0:c.mode,l=!o&&n?{colorScheme:n,...e}:{...e};r(t(a,{...l}),l)}return Object.entries(l).forEach((e=>{var n;let[a,{css:s}]=e;const l=null===(n=i[a])||void 0===n||null===(n=n.palette)||void 0===n?void 0:n.mode,c=!o&&l?{colorScheme:l,...s}:{...s};r(t(a,{...c}),c)})),n}}};function ga(e,n){return{toolbar:{minHeight:56,[e.up("xs")]:{"@media (orientation: landscape)":{minHeight:48}},[e.up("sm")]:{minHeight:64}},...n}}const ya={textTransform:"uppercase"},xa='"Roboto", "Helvetica", "Arial", sans-serif';function ba(e,n){const{fontFamily:t=xa,fontSize:o=14,fontWeightLight:a=300,fontWeightRegular:i=400,fontWeightMedium:r=500,fontWeightBold:s=700,htmlFontSize:l=16,allVariants:c,pxToRem:d,...u}="function"===typeof n?n(e):n;const h=o/14,p=d||(e=>"".concat(e/l*h,"rem")),m=(e,n,o,a,i)=>{return{fontFamily:t,fontWeight:e,fontSize:p(n),lineHeight:o,...t===xa?{letterSpacing:"".concat((r=a/n,Math.round(1e5*r)/1e5),"em")}:{},...i,...c};var r},f={h1:m(a,96,1.167,-1.5),h2:m(a,60,1.2,-.5),h3:m(i,48,1.167,0),h4:m(i,34,1.235,.25),h5:m(i,24,1.334,0),h6:m(r,20,1.6,.15),subtitle1:m(i,16,1.75,.15),subtitle2:m(r,14,1.57,.1),body1:m(i,16,1.5,.15),body2:m(i,14,1.43,.15),button:m(r,14,1.75,.4,ya),caption:m(i,12,1.66,.4),overline:m(i,12,2.66,1,ya),inherit:{fontFamily:"inherit",fontWeight:"inherit",fontSize:"inherit",lineHeight:"inherit",letterSpacing:"inherit"}};return vt({htmlFontSize:l,pxToRem:p,fontFamily:t,fontSize:o,fontWeightLight:a,fontWeightRegular:i,fontWeightMedium:r,fontWeightBold:s,...f},u,{clone:!1})}function va(){return["".concat(arguments.length<=0?void 0:arguments[0],"px ").concat(arguments.length<=1?void 0:arguments[1],"px ").concat(arguments.length<=2?void 0:arguments[2],"px ").concat(arguments.length<=3?void 0:arguments[3],"px rgba(0,0,0,").concat(.2,")"),"".concat(arguments.length<=4?void 0:arguments[4],"px ").concat(arguments.length<=5?void 0:arguments[5],"px ").concat(arguments.length<=6?void 0:arguments[6],"px ").concat(arguments.length<=7?void 0:arguments[7],"px rgba(0,0,0,").concat(.14,")"),"".concat(arguments.length<=8?void 0:arguments[8],"px ").concat(arguments.length<=9?void 0:arguments[9],"px ").concat(arguments.length<=10?void 0:arguments[10],"px ").concat(arguments.length<=11?void 0:arguments[11],"px rgba(0,0,0,").concat(.12,")")].join(",")}const ja=["none",va(0,2,1,-1,0,1,1,0,0,1,3,0),va(0,3,1,-2,0,2,2,0,0,1,5,0),va(0,3,3,-2,0,3,4,0,0,1,8,0),va(0,2,4,-1,0,4,5,0,0,1,10,0),va(0,3,5,-1,0,5,8,0,0,1,14,0),va(0,3,5,-1,0,6,10,0,0,1,18,0),va(0,4,5,-2,0,7,10,1,0,2,16,1),va(0,5,5,-3,0,8,10,1,0,3,14,2),va(0,5,6,-3,0,9,12,1,0,3,16,2),va(0,6,6,-3,0,10,14,1,0,4,18,3),va(0,6,7,-4,0,11,15,1,0,4,20,3),va(0,7,8,-4,0,12,17,2,0,5,22,4),va(0,7,8,-4,0,13,19,2,0,5,24,4),va(0,7,9,-4,0,14,21,2,0,5,26,4),va(0,8,9,-5,0,15,22,2,0,6,28,5),va(0,8,10,-5,0,16,24,2,0,6,30,5),va(0,8,11,-5,0,17,26,2,0,6,32,5),va(0,9,11,-5,0,18,28,2,0,7,34,6),va(0,9,12,-6,0,19,29,2,0,7,36,6),va(0,10,13,-6,0,20,31,3,0,8,38,7),va(0,10,13,-6,0,21,33,3,0,8,40,7),va(0,10,14,-6,0,22,35,3,0,8,42,7),va(0,11,14,-7,0,23,36,3,0,9,44,8),va(0,11,15,-7,0,24,38,3,0,9,46,8)],Ca={easeInOut:"cubic-bezier(0.4, 0, 0.2, 1)",easeOut:"cubic-bezier(0.0, 0, 0.2, 1)",easeIn:"cubic-bezier(0.4, 0, 1, 1)",sharp:"cubic-bezier(0.4, 0, 0.6, 1)"},wa={shortest:150,shorter:200,short:250,standard:300,complex:375,enteringScreen:225,leavingScreen:195};function Sa(e){return"".concat(Math.round(e),"ms")}function ka(e){if(!e)return 0;const n=e/36;return Math.min(Math.round(10*(4+15*n**.25+n/5)),3e3)}function Na(e){const n={...Ca,...e.easing},t={...wa,...e.duration};return{getAutoHeightDuration:ka,create:function(){let e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:["all"],o=arguments.length>1&&void 0!==arguments[1]?arguments[1]:{};const{duration:a=t.standard,easing:i=n.easeInOut,delay:r=0,...s}=o;return(Array.isArray(e)?e:[e]).map((e=>"".concat(e," ").concat("string"===typeof a?a:Sa(a)," ").concat(i," ").concat("string"===typeof r?r:Sa(r)))).join(",")},...e,easing:n,duration:t}}const _a={mobileStepper:1e3,fab:1050,speedDial:1050,appBar:1100,drawer:1200,modal:1300,snackbar:1400,tooltip:1500};function Ta(){const e={...arguments.length>0&&void 0!==arguments[0]?arguments[0]:{}};return function e(n){const t=Object.entries(n);for(let a=0;a<t.length;a++){const[i,r]=t[a];!xt(o=r)&&"undefined"!==typeof o&&"string"!==typeof o&&"boolean"!==typeof o&&"number"!==typeof o&&!Array.isArray(o)||i.startsWith("unstable_")?delete n[i]:xt(r)&&(n[i]={...r},e(n[i]))}var o}(e),"import { unstable_createBreakpoints as createBreakpoints, createTransitions } from '@mui/material/styles';\n\nconst theme = ".concat(JSON.stringify(e,null,2),";\n\ntheme.breakpoints = createBreakpoints(theme.breakpoints || {});\ntheme.transitions = createTransitions(theme.transitions || {});\n\nexport default theme;")}function Ia(){let e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:{};const{breakpoints:n,mixins:t={},spacing:o,palette:a={},transitions:i={},typography:r={},shape:s,...l}=e;if(e.vars)throw new Error(gt(20));const c=la(a),d=Io(e);let u=vt(d,{mixins:ga(d.breakpoints,t),palette:c,shadows:ja.slice(),typography:ba(c,r),transitions:Na(i),zIndex:{..._a}});u=vt(u,l);for(var h=arguments.length,p=new Array(h>1?h-1:0),m=1;m<h;m++)p[m-1]=arguments[m];return u=p.reduce(((e,n)=>vt(e,n)),u),u.unstable_sxConfig={...vo,...null===l||void 0===l?void 0:l.unstable_sxConfig},u.unstable_sx=function(e){return Co({sx:e,theme:this})},u.toRuntimeSource=Ta,u}const Ea=Ia;function Aa(e){let n;return n=e<1?5.11916*e**2:4.5*Math.log(e+1)+2,Math.round(10*n)/1e3}const Ba=[...Array(25)].map(((e,n)=>{if(0===n)return"none";const t=Aa(n);return"linear-gradient(rgba(255 255 255 / ".concat(t,"), rgba(255 255 255 / ").concat(t,"))")}));function Ra(e){return{inputPlaceholder:"dark"===e?.5:.42,inputUnderline:"dark"===e?.7:.42,switchTrackDisabled:"dark"===e?.2:.12,switchTrack:"dark"===e?.3:.38}}function Fa(e){return"dark"===e?Ba:[]}function Pa(e){var n;return!!e[0].match(/(cssVarPrefix|colorSchemeSelector|rootSelector|typography|mixins|breakpoints|direction|transitions)/)||!!e[0].match(/sxConfig$/)||"palette"===e[0]&&!(null===(n=e[1])||void 0===n||!n.match(/(mode|contrastThreshold|tonalOffset)/))}const Da=e=>[...[...Array(25)].map(((n,t)=>"--".concat(e?"".concat(e,"-"):"","overlays-").concat(t))),"--".concat(e?"".concat(e,"-"):"","palette-AppBar-darkBg"),"--".concat(e?"".concat(e,"-"):"","palette-AppBar-darkColor")],Oa=e=>(n,t)=>{const o=e.rootSelector||":root",a=e.colorSchemeSelector;let i=a;if("class"===a&&(i=".%s"),"data"===a&&(i="[data-%s]"),null!==a&&void 0!==a&&a.startsWith("data-")&&!a.includes("%s")&&(i="[".concat(a,'="%s"]')),e.defaultColorScheme===n){if("dark"===n){const a={};return Da(e.cssVarPrefix).forEach((e=>{a[e]=t[e],delete t[e]})),"media"===i?{[o]:t,"@media (prefers-color-scheme: dark)":{[o]:a}}:i?{[i.replace("%s",n)]:a,["".concat(o,", ").concat(i.replace("%s",n))]:t}:{[o]:{...t,...a}}}if(i&&"media"!==i)return"".concat(o,", ").concat(i.replace("%s",String(n)))}else if(n){if("media"===i)return{["@media (prefers-color-scheme: ".concat(String(n),")")]:{[o]:t}};if(i)return i.replace("%s",String(n))}return o};function Ma(e,n,t){!e[n]&&t&&(e[n]=t)}function La(e){return"string"===typeof e&&e.startsWith("hsl")?Go(e):e}function Ga(e,n){"".concat(n,"Channel")in e||(e["".concat(n,"Channel")]=Mo(La(e[n]),"MUI: Can't create `palette.".concat(n,"Channel` because `palette.").concat(n,"` is not one of these formats: #nnn, #nnnnnn, rgb(), rgba(), hsl(), hsla(), color().")+"\n"+"To suppress this warning, you need to explicitly provide the `palette.".concat(n,'Channel` as a string (in rgb format, for example "12 12 12") or undefined if you want to remove the channel token.')))}const Va=e=>{try{return e()}catch(n){}},qa=function(){return ca(arguments.length>0&&void 0!==arguments[0]?arguments[0]:"mui")};function Ha(e,n,t,o){var a,i,r;if(!n)return;n=!0===n?{}:n;const s="dark"===o?"dark":"light";var l;if(!t)return void(e[o]=function(e){const{palette:n={mode:"light"},opacity:t,overlays:o,...a}=e,i=la(n);return{palette:i,opacity:{...Ra(i.mode),...t},overlays:o||Fa(i.mode),...a}}({...n,palette:{mode:s,...null===(l=n)||void 0===l?void 0:l.palette}}));const{palette:c,...d}=Ea({...t,palette:{mode:s,...null===(a=n)||void 0===a?void 0:a.palette}});return e[o]={...n,palette:c,opacity:{...Ra(s),...null===(i=n)||void 0===i?void 0:i.opacity},overlays:(null===(r=n)||void 0===r?void 0:r.overlays)||Fa(s)},d}function za(e,n,t){e.colorSchemes&&t&&(e.colorSchemes[n]={...!0!==t&&t,palette:la({...!0===t?{}:t.palette,mode:n})})}function Ua(){let e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:{};const{palette:n,cssVariables:t=!1,colorSchemes:o=(n?void 0:{light:!0}),defaultColorScheme:a=(null===n||void 0===n?void 0:n.mode),...i}=e,r=a||"light",s=null===o||void 0===o?void 0:o[r],l={...o,...n?{[r]:{..."boolean"!==typeof s&&s,palette:n}}:void 0};for(var c=arguments.length,d=new Array(c>1?c-1:0),u=1;u<c;u++)d[u-1]=arguments[u];if(!1===t){if(!("colorSchemes"in e))return Ea(e,...d);let t=n;"palette"in e||l[r]&&(!0!==l[r]?t=l[r].palette:"dark"===r&&(t={mode:"dark"}));const o=Ea({...e,palette:t},...d);return o.defaultColorScheme=r,o.colorSchemes=l,"light"===o.palette.mode&&(o.colorSchemes.light={...!0!==l.light&&l.light,palette:o.palette},za(o,"dark",l.dark)),"dark"===o.palette.mode&&(o.colorSchemes.dark={...!0!==l.dark&&l.dark,palette:o.palette},za(o,"light",l.light)),o}return n||"light"in l||"light"!==r||(l.light=!0),function(){let e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:{};const{colorSchemes:n={light:!0},defaultColorScheme:t,disableCssColorScheme:o=!1,cssVarPrefix:a="mui",shouldSkipGeneratingVar:i=Pa,colorSchemeSelector:r=(n.light&&n.dark?"media":void 0),rootSelector:s=":root",...l}=e,c=Object.keys(n)[0],d=t||(n.light&&"light"!==c?"light":c),u=qa(a),{[d]:h,light:p,dark:m,...f}=n,g={...f};let y=h;if(("dark"===d&&!("dark"in n)||"light"===d&&!("light"in n))&&(y=!0),!y)throw new Error(gt(21,d));const x=Ha(g,y,l,d);p&&!g.light&&Ha(g,p,void 0,"light"),m&&!g.dark&&Ha(g,m,void 0,"dark");let b={defaultColorScheme:d,...x,cssVarPrefix:a,colorSchemeSelector:r,rootSelector:s,getCssVar:u,colorSchemes:g,font:{...da(x.typography),...x.font},spacing:(v=l.spacing,"number"===typeof v?"".concat(v,"px"):"string"===typeof v||"function"===typeof v||Array.isArray(v)?v:"8px")};var v;Object.keys(b.colorSchemes).forEach((e=>{const n=b.colorSchemes[e].palette,t=e=>{const t=e.split("-"),o=t[1],a=t[2];return u(e,n[o][a])};var o;if("light"===n.mode&&(Ma(n.common,"background","#fff"),Ma(n.common,"onBackground","#000")),"dark"===n.mode&&(Ma(n.common,"background","#000"),Ma(n.common,"onBackground","#fff")),o=n,["Alert","AppBar","Avatar","Button","Chip","FilledInput","LinearProgress","Skeleton","Slider","SnackbarContent","SpeedDialAction","StepConnector","StepContent","Switch","TableCell","Tooltip"].forEach((e=>{o[e]||(o[e]={})})),"light"===n.mode){Ma(n.Alert,"errorColor",Uo(n.error.light,.6)),Ma(n.Alert,"infoColor",Uo(n.info.light,.6)),Ma(n.Alert,"successColor",Uo(n.success.light,.6)),Ma(n.Alert,"warningColor",Uo(n.warning.light,.6)),Ma(n.Alert,"errorFilledBg",t("palette-error-main")),Ma(n.Alert,"infoFilledBg",t("palette-info-main")),Ma(n.Alert,"successFilledBg",t("palette-success-main")),Ma(n.Alert,"warningFilledBg",t("palette-warning-main")),Ma(n.Alert,"errorFilledColor",Va((()=>n.getContrastText(n.error.main)))),Ma(n.Alert,"infoFilledColor",Va((()=>n.getContrastText(n.info.main)))),Ma(n.Alert,"successFilledColor",Va((()=>n.getContrastText(n.success.main)))),Ma(n.Alert,"warningFilledColor",Va((()=>n.getContrastText(n.warning.main)))),Ma(n.Alert,"errorStandardBg",Wo(n.error.light,.9)),Ma(n.Alert,"infoStandardBg",Wo(n.info.light,.9)),Ma(n.Alert,"successStandardBg",Wo(n.success.light,.9)),Ma(n.Alert,"warningStandardBg",Wo(n.warning.light,.9)),Ma(n.Alert,"errorIconColor",t("palette-error-main")),Ma(n.Alert,"infoIconColor",t("palette-info-main")),Ma(n.Alert,"successIconColor",t("palette-success-main")),Ma(n.Alert,"warningIconColor",t("palette-warning-main")),Ma(n.AppBar,"defaultBg",t("palette-grey-100")),Ma(n.Avatar,"defaultBg",t("palette-grey-400")),Ma(n.Button,"inheritContainedBg",t("palette-grey-300")),Ma(n.Button,"inheritContainedHoverBg",t("palette-grey-A100")),Ma(n.Chip,"defaultBorder",t("palette-grey-400")),Ma(n.Chip,"defaultAvatarColor",t("palette-grey-700")),Ma(n.Chip,"defaultIconColor",t("palette-grey-700")),Ma(n.FilledInput,"bg","rgba(0, 0, 0, 0.06)"),Ma(n.FilledInput,"hoverBg","rgba(0, 0, 0, 0.09)"),Ma(n.FilledInput,"disabledBg","rgba(0, 0, 0, 0.12)"),Ma(n.LinearProgress,"primaryBg",Wo(n.primary.main,.62)),Ma(n.LinearProgress,"secondaryBg",Wo(n.secondary.main,.62)),Ma(n.LinearProgress,"errorBg",Wo(n.error.main,.62)),Ma(n.LinearProgress,"infoBg",Wo(n.info.main,.62)),Ma(n.LinearProgress,"successBg",Wo(n.success.main,.62)),Ma(n.LinearProgress,"warningBg",Wo(n.warning.main,.62)),Ma(n.Skeleton,"bg","rgba(".concat(t("palette-text-primaryChannel")," / 0.11)")),Ma(n.Slider,"primaryTrack",Wo(n.primary.main,.62)),Ma(n.Slider,"secondaryTrack",Wo(n.secondary.main,.62)),Ma(n.Slider,"errorTrack",Wo(n.error.main,.62)),Ma(n.Slider,"infoTrack",Wo(n.info.main,.62)),Ma(n.Slider,"successTrack",Wo(n.success.main,.62)),Ma(n.Slider,"warningTrack",Wo(n.warning.main,.62));const e=Ko(n.background.default,.8);Ma(n.SnackbarContent,"bg",e),Ma(n.SnackbarContent,"color",Va((()=>n.getContrastText(e)))),Ma(n.SpeedDialAction,"fabHoverBg",Ko(n.background.paper,.15)),Ma(n.StepConnector,"border",t("palette-grey-400")),Ma(n.StepContent,"border",t("palette-grey-400")),Ma(n.Switch,"defaultColor",t("palette-common-white")),Ma(n.Switch,"defaultDisabledColor",t("palette-grey-100")),Ma(n.Switch,"primaryDisabledColor",Wo(n.primary.main,.62)),Ma(n.Switch,"secondaryDisabledColor",Wo(n.secondary.main,.62)),Ma(n.Switch,"errorDisabledColor",Wo(n.error.main,.62)),Ma(n.Switch,"infoDisabledColor",Wo(n.info.main,.62)),Ma(n.Switch,"successDisabledColor",Wo(n.success.main,.62)),Ma(n.Switch,"warningDisabledColor",Wo(n.warning.main,.62)),Ma(n.TableCell,"border",Wo(Ho(n.divider,1),.88)),Ma(n.Tooltip,"bg",Ho(n.grey[700],.92))}if("dark"===n.mode){Ma(n.Alert,"errorColor",Wo(n.error.light,.6)),Ma(n.Alert,"infoColor",Wo(n.info.light,.6)),Ma(n.Alert,"successColor",Wo(n.success.light,.6)),Ma(n.Alert,"warningColor",Wo(n.warning.light,.6)),Ma(n.Alert,"errorFilledBg",t("palette-error-dark")),Ma(n.Alert,"infoFilledBg",t("palette-info-dark")),Ma(n.Alert,"successFilledBg",t("palette-success-dark")),Ma(n.Alert,"warningFilledBg",t("palette-warning-dark")),Ma(n.Alert,"errorFilledColor",Va((()=>n.getContrastText(n.error.dark)))),Ma(n.Alert,"infoFilledColor",Va((()=>n.getContrastText(n.info.dark)))),Ma(n.Alert,"successFilledColor",Va((()=>n.getContrastText(n.success.dark)))),Ma(n.Alert,"warningFilledColor",Va((()=>n.getContrastText(n.warning.dark)))),Ma(n.Alert,"errorStandardBg",Uo(n.error.light,.9)),Ma(n.Alert,"infoStandardBg",Uo(n.info.light,.9)),Ma(n.Alert,"successStandardBg",Uo(n.success.light,.9)),Ma(n.Alert,"warningStandardBg",Uo(n.warning.light,.9)),Ma(n.Alert,"errorIconColor",t("palette-error-main")),Ma(n.Alert,"infoIconColor",t("palette-info-main")),Ma(n.Alert,"successIconColor",t("palette-success-main")),Ma(n.Alert,"warningIconColor",t("palette-warning-main")),Ma(n.AppBar,"defaultBg",t("palette-grey-900")),Ma(n.AppBar,"darkBg",t("palette-background-paper")),Ma(n.AppBar,"darkColor",t("palette-text-primary")),Ma(n.Avatar,"defaultBg",t("palette-grey-600")),Ma(n.Button,"inheritContainedBg",t("palette-grey-800")),Ma(n.Button,"inheritContainedHoverBg",t("palette-grey-700")),Ma(n.Chip,"defaultBorder",t("palette-grey-700")),Ma(n.Chip,"defaultAvatarColor",t("palette-grey-300")),Ma(n.Chip,"defaultIconColor",t("palette-grey-300")),Ma(n.FilledInput,"bg","rgba(255, 255, 255, 0.09)"),Ma(n.FilledInput,"hoverBg","rgba(255, 255, 255, 0.13)"),Ma(n.FilledInput,"disabledBg","rgba(255, 255, 255, 0.12)"),Ma(n.LinearProgress,"primaryBg",Uo(n.primary.main,.5)),Ma(n.LinearProgress,"secondaryBg",Uo(n.secondary.main,.5)),Ma(n.LinearProgress,"errorBg",Uo(n.error.main,.5)),Ma(n.LinearProgress,"infoBg",Uo(n.info.main,.5)),Ma(n.LinearProgress,"successBg",Uo(n.success.main,.5)),Ma(n.LinearProgress,"warningBg",Uo(n.warning.main,.5)),Ma(n.Skeleton,"bg","rgba(".concat(t("palette-text-primaryChannel")," / 0.13)")),Ma(n.Slider,"primaryTrack",Uo(n.primary.main,.5)),Ma(n.Slider,"secondaryTrack",Uo(n.secondary.main,.5)),Ma(n.Slider,"errorTrack",Uo(n.error.main,.5)),Ma(n.Slider,"infoTrack",Uo(n.info.main,.5)),Ma(n.Slider,"successTrack",Uo(n.success.main,.5)),Ma(n.Slider,"warningTrack",Uo(n.warning.main,.5));const e=Ko(n.background.default,.98);Ma(n.SnackbarContent,"bg",e),Ma(n.SnackbarContent,"color",Va((()=>n.getContrastText(e)))),Ma(n.SpeedDialAction,"fabHoverBg",Ko(n.background.paper,.15)),Ma(n.StepConnector,"border",t("palette-grey-600")),Ma(n.StepContent,"border",t("palette-grey-600")),Ma(n.Switch,"defaultColor",t("palette-grey-300")),Ma(n.Switch,"defaultDisabledColor",t("palette-grey-600")),Ma(n.Switch,"primaryDisabledColor",Uo(n.primary.main,.55)),Ma(n.Switch,"secondaryDisabledColor",Uo(n.secondary.main,.55)),Ma(n.Switch,"errorDisabledColor",Uo(n.error.main,.55)),Ma(n.Switch,"infoDisabledColor",Uo(n.info.main,.55)),Ma(n.Switch,"successDisabledColor",Uo(n.success.main,.55)),Ma(n.Switch,"warningDisabledColor",Uo(n.warning.main,.55)),Ma(n.TableCell,"border",Uo(Ho(n.divider,1),.68)),Ma(n.Tooltip,"bg",Ho(n.grey[700],.92))}Ga(n.background,"default"),Ga(n.background,"paper"),Ga(n.common,"background"),Ga(n.common,"onBackground"),Ga(n,"divider"),Object.keys(n).forEach((e=>{const t=n[e];"tonalOffset"!==e&&t&&"object"===typeof t&&(t.main&&Ma(n[e],"mainChannel",Mo(La(t.main))),t.light&&Ma(n[e],"lightChannel",Mo(La(t.light))),t.dark&&Ma(n[e],"darkChannel",Mo(La(t.dark))),t.contrastText&&Ma(n[e],"contrastTextChannel",Mo(La(t.contrastText))),"text"===e&&(Ga(n[e],"primary"),Ga(n[e],"secondary")),"action"===e&&(t.active&&Ga(n[e],"active"),t.selected&&Ga(n[e],"selected")))}))}));for(var j=arguments.length,C=new Array(j>1?j-1:0),w=1;w<j;w++)C[w-1]=arguments[w];b=C.reduce(((e,n)=>vt(e,n)),b);const S={prefix:a,disableCssColorScheme:o,shouldSkipGeneratingVar:i,getSelector:Oa(b)},{vars:k,generateThemeVars:N,generateStyleSheets:_}=fa(b,S);return b.vars=k,Object.entries(b.colorSchemes[b.defaultColorScheme]).forEach((e=>{let[n,t]=e;b[n]=t})),b.generateThemeVars=N,b.generateStyleSheets=_,b.generateSpacing=function(){return _o(l.spacing,Lt(this))},b.getColorSchemeSelector=function(e){return function(n){return"media"===e?"@media (prefers-color-scheme: ".concat(n,")"):e?e.startsWith("data-")&&!e.includes("%s")?"[".concat(e,'="').concat(n,'"] &'):"class"===e?".".concat(n," &"):"data"===e?"[data-".concat(n,"] &"):"".concat(e.replace("%s",n)," &"):"&"}}(r),b.spacing=b.generateSpacing(),b.shouldSkipGeneratingVar=i,b.unstable_sxConfig={...vo,...null===l||void 0===l?void 0:l.unstable_sxConfig},b.unstable_sx=function(e){return Co({sx:e,theme:this})},b.toRuntimeSource=Ta,b}({...i,colorSchemes:l,defaultColorScheme:r,..."boolean"!==typeof t&&t},...d)}const Qa="$$material",Wa={active:"active",checked:"checked",completed:"completed",disabled:"disabled",error:"error",expanded:"expanded",focused:"focused",focusVisible:"focusVisible",open:"open",readOnly:"readOnly",required:"required",selected:"selected"};function Ka(e,n){let t=arguments.length>2&&void 0!==arguments[2]?arguments[2]:"Mui";const o=Wa[n];return o?"".concat(t,"-").concat(o):"".concat(Fo.generate(e),"-").concat(n)}function $a(e,n){let t=arguments.length>2&&void 0!==arguments[2]?arguments[2]:"Mui";const o={};return n.forEach((n=>{o[n]=Ka(e,n,t)})),o}const Ya=$a("MuiBox",["root"]),Xa=Ua(),Za=function(){let e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:{};const{themeId:n,defaultTheme:t,defaultClassName:o="MuiBox-root",generateClassName:i}=e,r=pt("div",{shouldForwardProp:e=>"theme"!==e&&"sx"!==e&&"as"!==e})(Co);return a.forwardRef((function(e,a){const s=Bo(t),{className:l,component:c="div",...d}=So(e);return(0,we.jsx)(r,{as:c,ref:a,className:Ie(l,i?i(o):o),theme:n&&s[n]||s,...d})}))}({themeId:Qa,defaultTheme:Xa,defaultClassName:Ya.root,generateClassName:Fo.generate}),Ja=Za;function ei(e,n){let t=arguments.length>2&&void 0!==arguments[2]?arguments[2]:void 0;const o={};for(const a in e){const i=e[a];let r="",s=!0;for(let e=0;e<i.length;e+=1){const o=i[e];o&&(r+=(!0===s?"":" ")+n(o),s=!1,t&&t[o]&&(r+=" "+t[o]))}o[a]=r}return o}function ni(e){const{variants:n,...t}=e,o={variants:n,style:ft(t),isProcessed:!0};return o.style===t||n&&n.forEach((e=>{"function"!==typeof e.style&&(e.style=ft(e.style))})),o}const ti=Io();function oi(e){return"ownerState"!==e&&"theme"!==e&&"sx"!==e&&"as"!==e}function ai(e){return e?(n,t)=>t[e]:null}function ii(e,n){const t="function"===typeof n?n(e):n;if(Array.isArray(t))return t.flatMap((n=>ii(e,n)));if(Array.isArray(null===t||void 0===t?void 0:t.variants)){let n;if(t.isProcessed)n=t.style;else{const{variants:e,...o}=t;n=o}return ri(e,t.variants,[n])}return null!==t&&void 0!==t&&t.isProcessed?t.style:t}function ri(e,n){let t,o=arguments.length>2&&void 0!==arguments[2]?arguments[2]:[];e:for(let s=0;s<n.length;s+=1){const l=n[s];var a,i;if("function"===typeof l.props){if(null!==(a=t)&&void 0!==a||(t={...e,...e.ownerState,ownerState:e.ownerState}),!l.props(t))continue}else for(const n in l.props){var r;if(e[n]!==l.props[n]&&(null===(r=e.ownerState)||void 0===r?void 0:r[n])!==l.props[n])continue e}if("function"===typeof l.style)null!==(i=t)&&void 0!==i||(t={...e,...e.ownerState,ownerState:e.ownerState}),o.push(l.style(t));else o.push(l.style)}return o}function si(e,n){}function li(e){return e?e.charAt(0).toLowerCase()+e.slice(1):e}const ci=Ua();const di=function(e){return"ownerState"!==e&&"theme"!==e&&"sx"!==e&&"as"!==e},ui=e=>di(e)&&"classes"!==e,hi=function(){let e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:{};const{themeId:n,defaultTheme:t=ti,rootShouldForwardProp:o=oi,slotShouldForwardProp:a=oi}=e;function i(e){!function(e,n,t){e.theme=function(e){for(const n in e)return!1;return!0}(e.theme)?t:e.theme[n]||e.theme}(e,n,t)}return function(e){let n=arguments.length>1&&void 0!==arguments[1]?arguments[1]:{};!function(e,n){Array.isArray(e.__emotion_styles)&&(e.__emotion_styles=n(e.__emotion_styles))}(e,(e=>e.filter((e=>e!==Co))));const{name:t,slot:r,skipVariantsResolver:s,skipSx:l,overridesResolver:c=ai(li(r)),...d}=n,u=void 0!==s?s:r&&"Root"!==r&&"root"!==r||!1,h=l||!1;let p=oi;"Root"===r||"root"===r?p=o:r?p=a:function(e){return"string"===typeof e&&e.charCodeAt(0)>96}(e)&&(p=void 0);const m=pt(e,{shouldForwardProp:p,label:si(t,r),...d}),f=e=>{if("function"===typeof e&&e.__emotion_real!==e)return function(n){return ii(n,e)};if(xt(e)){const n=ni(e);return n.variants?function(e){return ii(e,n)}:n.style}return e},g=function(){const n=[];for(var o=arguments.length,a=new Array(o),r=0;r<o;r++)a[r]=arguments[r];const s=a.map(f),l=[];if(n.push(i),t&&c&&l.push((function(e){var n;const o=null===(n=e.theme.components)||void 0===n||null===(n=n[t])||void 0===n?void 0:n.styleOverrides;if(!o)return null;const a={};for(const t in o)a[t]=ii(e,o[t]);return c(e,a)})),t&&!u&&l.push((function(e){var n;const o=e.theme,a=null===o||void 0===o||null===(n=o.components)||void 0===n||null===(n=n[t])||void 0===n?void 0:n.variants;return a?ri(e,a):null})),h||l.push(Co),Array.isArray(s[0])){const e=s.shift(),t=new Array(n.length).fill(""),o=new Array(l.length).fill("");let a;a=[...t,...e,...o],a.raw=[...t,...e.raw,...o],n.unshift(a)}const d=[...n,...s,...l],p=m(...d);return e.muiName&&(p.muiName=e.muiName),p};return m.withConfig&&(g.withConfig=m.withConfig),g}}({themeId:Qa,defaultTheme:ci,rootShouldForwardProp:ui}),pi=hi,mi={theme:void 0};const fi=function(e){let n,t;return function(o){let a=n;return void 0!==a&&o.theme===t||(mi.theme=o.theme,a=ni(e(mi)),n=a,t=o.theme),a}};function gi(e,n){const t={...n};for(const o in e)if(Object.prototype.hasOwnProperty.call(e,o)){const a=o;if("components"===a||"slots"===a)t[a]={...e[a],...t[a]};else if("componentsProps"===a||"slotProps"===a){const o=e[a],i=n[a];if(i)if(o){t[a]={...i};for(const e in o)if(Object.prototype.hasOwnProperty.call(o,e)){const n=e;t[a][n]=gi(o[n],i[n])}}else t[a]=i;else t[a]=o||{}}else void 0===t[a]&&(t[a]=e[a])}return t}const yi=a.createContext(void 0);function xi(e){let{props:n,name:t}=e;return function(e){const{theme:n,name:t,props:o}=e;if(!n||!n.components||!n.components[t])return o;const a=n.components[t];return a.defaultProps?gi(a.defaultProps,o):a.styleOverrides||a.variants?o:gi(a,o)}({props:n,name:t,theme:{components:a.useContext(yi)}})}function bi(e){return xi(e)}function vi(e){return Ka("MuiToolbar",e)}$a("MuiToolbar",["root","gutters","regular","dense"]);const ji=pi("div",{name:"MuiToolbar",slot:"Root",overridesResolver:(e,n)=>{const{ownerState:t}=e;return[n.root,!t.disableGutters&&n.gutters,n[t.variant]]}})(fi((e=>{let{theme:n}=e;return{position:"relative",display:"flex",alignItems:"center",variants:[{props:e=>{let{ownerState:n}=e;return!n.disableGutters},style:{paddingLeft:n.spacing(2),paddingRight:n.spacing(2),[n.breakpoints.up("sm")]:{paddingLeft:n.spacing(3),paddingRight:n.spacing(3)}}},{props:{variant:"dense"},style:{minHeight:48}},{props:{variant:"regular"},style:n.mixins.toolbar}]}}))),Ci=a.forwardRef((function(e,n){const t=bi({props:e,name:"MuiToolbar"}),{className:o,component:a="div",disableGutters:i=!1,variant:r="regular",...s}=t,l={...t,component:a,disableGutters:i,variant:r},c=(e=>{const{classes:n,disableGutters:t,variant:o}=e;return ei({root:["root",!t&&"gutters",o]},vi,n)})(l);return(0,we.jsx)(ji,{as:a,className:Ie(c.root,o),ref:n,ownerState:l,...s})}));const wi=a.createContext({});function Si(e){return Ka("MuiList",e)}$a("MuiList",["root","padding","dense","subheader"]);const ki=pi("ul",{name:"MuiList",slot:"Root",overridesResolver:(e,n)=>{const{ownerState:t}=e;return[n.root,!t.disablePadding&&n.padding,t.dense&&n.dense,t.subheader&&n.subheader]}})({listStyle:"none",margin:0,padding:0,position:"relative",variants:[{props:e=>{let{ownerState:n}=e;return!n.disablePadding},style:{paddingTop:8,paddingBottom:8}},{props:e=>{let{ownerState:n}=e;return n.subheader},style:{paddingTop:0}}]}),Ni=a.forwardRef((function(e,n){const t=bi({props:e,name:"MuiList"}),{children:o,className:i,component:r="ul",dense:s=!1,disablePadding:l=!1,subheader:c,...d}=t,u=a.useMemo((()=>({dense:s})),[s]),h={...t,component:r,dense:s,disablePadding:l},p=(e=>{const{classes:n,disablePadding:t,dense:o,subheader:a}=e;return ei({root:["root",!t&&"padding",o&&"dense",a&&"subheader"]},Si,n)})(h);return(0,we.jsx)(wi.Provider,{value:u,children:(0,we.jsxs)(ki,{as:r,className:Ie(p.root,i),ref:n,ownerState:h,...d,children:[c,o]})})}));function _i(e){try{return e.matches(":focus-visible")}catch(n){0}return!1}function Ti(e,n){"function"===typeof e?e(n):e&&(e.current=n)}function Ii(){for(var e=arguments.length,n=new Array(e),t=0;t<e;t++)n[t]=arguments[t];return a.useMemo((()=>n.every((e=>null==e))?null:e=>{n.forEach((n=>{Ti(n,e)}))}),n)}const Ei=Ii,Ai="undefined"!==typeof window?a.useLayoutEffect:a.useEffect;const Bi=function(e){const n=a.useRef(e);return Ai((()=>{n.current=e})),a.useRef((function(){return(0,n.current)(...arguments)})).current},Ri=Bi;function Fi(e){return Fi="function"==typeof Symbol&&"symbol"==typeof Symbol.iterator?function(e){return typeof e}:function(e){return e&&"function"==typeof Symbol&&e.constructor===Symbol&&e!==Symbol.prototype?"symbol":typeof e},Fi(e)}function Pi(e){var n=function(e,n){if("object"!=Fi(e)||!e)return e;var t=e[Symbol.toPrimitive];if(void 0!==t){var o=t.call(e,n||"default");if("object"!=Fi(o))return o;throw new TypeError("@@toPrimitive must return a primitive value.")}return("string"===n?String:Number)(e)}(e,"string");return"symbol"==Fi(n)?n:n+""}function Di(e,n,t){return(n=Pi(n))in e?Object.defineProperty(e,n,{value:t,enumerable:!0,configurable:!0,writable:!0}):e[n]=t,e}const Oi={};function Mi(e,n){const t=a.useRef(Oi);return t.current===Oi&&(t.current=e(n)),t}class Li{static create(){return new Li}static use(){const e=Mi(Li.create).current,[n,t]=a.useState(!1);return e.shouldMount=n,e.setShouldMount=t,a.useEffect(e.mountEffect,[n]),e}constructor(){Di(this,"mountEffect",(()=>{this.shouldMount&&!this.didMount&&null!==this.ref.current&&(this.didMount=!0,this.mounted.resolve())})),this.ref={current:null},this.mounted=null,this.didMount=!1,this.shouldMount=!1,this.setShouldMount=null}mount(){return this.mounted||(this.mounted=function(){let e,n;const t=new Promise(((t,o)=>{e=t,n=o}));return t.resolve=e,t.reject=n,t}(),this.shouldMount=!0,this.setShouldMount(this.shouldMount)),this.mounted}start(){for(var e=arguments.length,n=new Array(e),t=0;t<e;t++)n[t]=arguments[t];this.mount().then((()=>{var e;return null===(e=this.ref.current)||void 0===e?void 0:e.start(...n)}))}stop(){for(var e=arguments.length,n=new Array(e),t=0;t<e;t++)n[t]=arguments[t];this.mount().then((()=>{var e;return null===(e=this.ref.current)||void 0===e?void 0:e.stop(...n)}))}pulsate(){for(var e=arguments.length,n=new Array(e),t=0;t<e;t++)n[t]=arguments[t];this.mount().then((()=>{var e;return null===(e=this.ref.current)||void 0===e?void 0:e.pulsate(...n)}))}}function Gi(e,n){return n||(n=e.slice(0)),Object.freeze(Object.defineProperties(e,{raw:{value:Object.freeze(n)}}))}function Vi(e,n){if(null==e)return{};var t={};for(var o in e)if({}.hasOwnProperty.call(e,o)){if(n.includes(o))continue;t[o]=e[o]}return t}function qi(e,n){return qi=Object.setPrototypeOf?Object.setPrototypeOf.bind():function(e,n){return e.__proto__=n,e},qi(e,n)}function Hi(e,n){e.prototype=Object.create(n.prototype),e.prototype.constructor=e,qi(e,n)}const zi=a.createContext(null);function Ui(e,n){var t=Object.create(null);return e&&a.Children.map(e,(function(e){return e})).forEach((function(e){t[e.key]=function(e){return n&&(0,a.isValidElement)(e)?n(e):e}(e)})),t}function Qi(e,n,t){return null!=t[n]?t[n]:e.props[n]}function Wi(e,n,t){var o=Ui(e.children),i=function(e,n){function t(t){return t in n?n[t]:e[t]}e=e||{},n=n||{};var o,a=Object.create(null),i=[];for(var r in e)r in n?i.length&&(a[r]=i,i=[]):i.push(r);var s={};for(var l in n){if(a[l])for(o=0;o<a[l].length;o++){var c=a[l][o];s[a[l][o]]=t(c)}s[l]=t(l)}for(o=0;o<i.length;o++)s[i[o]]=t(i[o]);return s}(n,o);return Object.keys(i).forEach((function(r){var s=i[r];if((0,a.isValidElement)(s)){var l=r in n,c=r in o,d=n[r],u=(0,a.isValidElement)(d)&&!d.props.in;!c||l&&!u?c||!l||u?c&&l&&(0,a.isValidElement)(d)&&(i[r]=(0,a.cloneElement)(s,{onExited:t.bind(null,s),in:d.props.in,exit:Qi(s,"exit",e),enter:Qi(s,"enter",e)})):i[r]=(0,a.cloneElement)(s,{in:!1}):i[r]=(0,a.cloneElement)(s,{onExited:t.bind(null,s),in:!0,exit:Qi(s,"exit",e),enter:Qi(s,"enter",e)})}})),i}var Ki=Object.values||function(e){return Object.keys(e).map((function(n){return e[n]}))},$i=function(e){function n(n,t){var o,a=(o=e.call(this,n,t)||this).handleExited.bind(function(e){if(void 0===e)throw new ReferenceError("this hasn't been initialised - super() hasn't been called");return e}(o));return o.state={contextValue:{isMounting:!0},handleExited:a,firstRender:!0},o}Hi(n,e);var t=n.prototype;return t.componentDidMount=function(){this.mounted=!0,this.setState({contextValue:{isMounting:!1}})},t.componentWillUnmount=function(){this.mounted=!1},n.getDerivedStateFromProps=function(e,n){var t,o,i=n.children,r=n.handleExited;return{children:n.firstRender?(t=e,o=r,Ui(t.children,(function(e){return(0,a.cloneElement)(e,{onExited:o.bind(null,e),in:!0,appear:Qi(e,"appear",t),enter:Qi(e,"enter",t),exit:Qi(e,"exit",t)})}))):Wi(e,i,r),firstRender:!1}},t.handleExited=function(e,n){var t=Ui(this.props.children);e.key in t||(e.props.onExited&&e.props.onExited(n),this.mounted&&this.setState((function(n){var t=Ee({},n.children);return delete t[e.key],{children:t}})))},t.render=function(){var e=this.props,n=e.component,t=e.childFactory,o=Vi(e,["component","childFactory"]),i=this.state.contextValue,r=Ki(this.state.children).map(t);return delete o.appear,delete o.enter,delete o.exit,null===n?a.createElement(zi.Provider,{value:i},r):a.createElement(zi.Provider,{value:i},a.createElement(n,o,r))},n}(a.Component);$i.propTypes={},$i.defaultProps={component:"div",childFactory:function(e){return e}};const Yi=$i,Xi=[];class Zi{constructor(){Di(this,"currentId",null),Di(this,"clear",(()=>{null!==this.currentId&&(clearTimeout(this.currentId),this.currentId=null)})),Di(this,"disposeEffect",(()=>this.clear))}static create(){return new Zi}start(e,n){this.clear(),this.currentId=setTimeout((()=>{this.currentId=null,n()}),e)}}function Ji(){const e=Mi(Zi.create).current;var n;return n=e.disposeEffect,a.useEffect(n,Xi),e}t(956);var er=function(e,n){var t=arguments;if(null==n||!tt.call(n,"css"))return a.createElement.apply(void 0,t);var o=t.length,i=new Array(o);i[0]=it,i[1]=function(e,n){var t={};for(var o in n)tt.call(n,o)&&(t[o]=n[o]);return t[ot]=e,t}(e,n);for(var r=2;r<o;r++)i[r]=t[r];return a.createElement.apply(null,i)};!function(e){var n;n||(n=e.JSX||(e.JSX={}))}(er||(er={}));function nr(){for(var e=arguments.length,n=new Array(e),t=0;t<e;t++)n[t]=arguments[t];return Yn(n)}function tr(){var e=nr.apply(void 0,arguments),n="animation-"+e.name;return{name:n,styles:"@keyframes "+n+"{"+e.styles+"}",anim:1,toString:function(){return"_EMO_"+this.name+"_"+this.styles+"_EMO_"}}}const or=function(e){const{className:n,classes:t,pulsate:o=!1,rippleX:i,rippleY:r,rippleSize:s,in:l,onExited:c,timeout:d}=e,[u,h]=a.useState(!1),p=Ie(n,t.ripple,t.rippleVisible,o&&t.ripplePulsate),m={width:s,height:s,top:-s/2+r,left:-s/2+i},f=Ie(t.child,u&&t.childLeaving,o&&t.childPulsate);return l||u||h(!0),a.useEffect((()=>{if(!l&&null!=c){const e=setTimeout(c,d);return()=>{clearTimeout(e)}}}),[c,l,d]),(0,we.jsx)("span",{className:p,style:m,children:(0,we.jsx)("span",{className:f})})};const ar=$a("MuiTouchRipple",["root","ripple","rippleVisible","ripplePulsate","child","childLeaving","childPulsate"]);var ir,rr,sr,lr;const cr=tr(ir||(ir=Gi(["\n  0% {\n    transform: scale(0);\n    opacity: 0.1;\n  }\n\n  100% {\n    transform: scale(1);\n    opacity: 0.3;\n  }\n"]))),dr=tr(rr||(rr=Gi(["\n  0% {\n    opacity: 1;\n  }\n\n  100% {\n    opacity: 0;\n  }\n"]))),ur=tr(sr||(sr=Gi(["\n  0% {\n    transform: scale(1);\n  }\n\n  50% {\n    transform: scale(0.92);\n  }\n\n  100% {\n    transform: scale(1);\n  }\n"]))),hr=pi("span",{name:"MuiTouchRipple",slot:"Root"})({overflow:"hidden",pointerEvents:"none",position:"absolute",zIndex:0,top:0,right:0,bottom:0,left:0,borderRadius:"inherit"}),pr=pi(or,{name:"MuiTouchRipple",slot:"Ripple"})(lr||(lr=Gi(["\n  opacity: 0;\n  position: absolute;\n\n  &."," {\n    opacity: 0.3;\n    transform: scale(1);\n    animation-name: ",";\n    animation-duration: ","ms;\n    animation-timing-function: ",";\n  }\n\n  &."," {\n    animation-duration: ","ms;\n  }\n\n  & ."," {\n    opacity: 1;\n    display: block;\n    width: 100%;\n    height: 100%;\n    border-radius: 50%;\n    background-color: currentColor;\n  }\n\n  & ."," {\n    opacity: 0;\n    animation-name: ",";\n    animation-duration: ","ms;\n    animation-timing-function: ",";\n  }\n\n  & ."," {\n    position: absolute;\n    /* @noflip */\n    left: 0px;\n    top: 0;\n    animation-name: ",";\n    animation-duration: 2500ms;\n    animation-timing-function: ",";\n    animation-iteration-count: infinite;\n    animation-delay: 200ms;\n  }\n"])),ar.rippleVisible,cr,550,(e=>{let{theme:n}=e;return n.transitions.easing.easeInOut}),ar.ripplePulsate,(e=>{let{theme:n}=e;return n.transitions.duration.shorter}),ar.child,ar.childLeaving,dr,550,(e=>{let{theme:n}=e;return n.transitions.easing.easeInOut}),ar.childPulsate,ur,(e=>{let{theme:n}=e;return n.transitions.easing.easeInOut})),mr=a.forwardRef((function(e,n){const t=bi({props:e,name:"MuiTouchRipple"}),{center:o=!1,classes:i={},className:r,...s}=t,[l,c]=a.useState([]),d=a.useRef(0),u=a.useRef(null);a.useEffect((()=>{u.current&&(u.current(),u.current=null)}),[l]);const h=a.useRef(!1),p=Ji(),m=a.useRef(null),f=a.useRef(null),g=a.useCallback((e=>{const{pulsate:n,rippleX:t,rippleY:o,rippleSize:a,cb:r}=e;c((e=>[...e,(0,we.jsx)(pr,{classes:{ripple:Ie(i.ripple,ar.ripple),rippleVisible:Ie(i.rippleVisible,ar.rippleVisible),ripplePulsate:Ie(i.ripplePulsate,ar.ripplePulsate),child:Ie(i.child,ar.child),childLeaving:Ie(i.childLeaving,ar.childLeaving),childPulsate:Ie(i.childPulsate,ar.childPulsate)},timeout:550,pulsate:n,rippleX:t,rippleY:o,rippleSize:a},d.current)])),d.current+=1,u.current=r}),[i]),y=a.useCallback((function(){let e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:{},n=arguments.length>1&&void 0!==arguments[1]?arguments[1]:{},t=arguments.length>2&&void 0!==arguments[2]?arguments[2]:()=>{};const{pulsate:a=!1,center:i=o||n.pulsate,fakeElement:r=!1}=n;if("mousedown"===(null===e||void 0===e?void 0:e.type)&&h.current)return void(h.current=!1);"touchstart"===(null===e||void 0===e?void 0:e.type)&&(h.current=!0);const s=r?null:f.current,l=s?s.getBoundingClientRect():{width:0,height:0,left:0,top:0};let c,d,u;if(i||void 0===e||0===e.clientX&&0===e.clientY||!e.clientX&&!e.touches)c=Math.round(l.width/2),d=Math.round(l.height/2);else{const{clientX:n,clientY:t}=e.touches&&e.touches.length>0?e.touches[0]:e;c=Math.round(n-l.left),d=Math.round(t-l.top)}if(i)u=Math.sqrt((2*l.width**2+l.height**2)/3),u%2===0&&(u+=1);else{const e=2*Math.max(Math.abs((s?s.clientWidth:0)-c),c)+2,n=2*Math.max(Math.abs((s?s.clientHeight:0)-d),d)+2;u=Math.sqrt(e**2+n**2)}null!==e&&void 0!==e&&e.touches?null===m.current&&(m.current=()=>{g({pulsate:a,rippleX:c,rippleY:d,rippleSize:u,cb:t})},p.start(80,(()=>{m.current&&(m.current(),m.current=null)}))):g({pulsate:a,rippleX:c,rippleY:d,rippleSize:u,cb:t})}),[o,g,p]),x=a.useCallback((()=>{y({},{pulsate:!0})}),[y]),b=a.useCallback(((e,n)=>{if(p.clear(),"touchend"===(null===e||void 0===e?void 0:e.type)&&m.current)return m.current(),m.current=null,void p.start(0,(()=>{b(e,n)}));m.current=null,c((e=>e.length>0?e.slice(1):e)),u.current=n}),[p]);return a.useImperativeHandle(n,(()=>({pulsate:x,start:y,stop:b})),[x,y,b]),(0,we.jsx)(hr,{className:Ie(ar.root,i.root,r),ref:f,...s,children:(0,we.jsx)(Yi,{component:null,exit:!0,children:l})})}));function fr(e){return Ka("MuiButtonBase",e)}const gr=$a("MuiButtonBase",["root","disabled","focusVisible"]),yr=pi("button",{name:"MuiButtonBase",slot:"Root",overridesResolver:(e,n)=>n.root})({display:"inline-flex",alignItems:"center",justifyContent:"center",position:"relative",boxSizing:"border-box",WebkitTapHighlightColor:"transparent",backgroundColor:"transparent",outline:0,border:0,margin:0,borderRadius:0,padding:0,cursor:"pointer",userSelect:"none",verticalAlign:"middle",MozAppearance:"none",WebkitAppearance:"none",textDecoration:"none",color:"inherit","&::-moz-focus-inner":{borderStyle:"none"},["&.".concat(gr.disabled)]:{pointerEvents:"none",cursor:"default"},"@media print":{colorAdjust:"exact"}});function xr(e,n,t){let o=arguments.length>3&&void 0!==arguments[3]&&arguments[3];return Ri((a=>(t&&t(a),o||e[n](a),!0)))}const br=a.forwardRef((function(e,n){const t=bi({props:e,name:"MuiButtonBase"}),{action:o,centerRipple:i=!1,children:r,className:s,component:l="button",disabled:c=!1,disableRipple:d=!1,disableTouchRipple:u=!1,focusRipple:h=!1,focusVisibleClassName:p,LinkComponent:m="a",onBlur:f,onClick:g,onContextMenu:y,onDragLeave:x,onFocus:b,onFocusVisible:v,onKeyDown:j,onKeyUp:C,onMouseDown:w,onMouseLeave:S,onMouseUp:k,onTouchEnd:N,onTouchMove:_,onTouchStart:T,tabIndex:I=0,TouchRippleProps:E,touchRippleRef:A,type:B,...R}=t,F=a.useRef(null),P=Li.use(),D=Ei(P.ref,A),[O,M]=a.useState(!1);c&&O&&M(!1),a.useImperativeHandle(o,(()=>({focusVisible:()=>{M(!0),F.current.focus()}})),[]);const L=P.shouldMount&&!d&&!c;a.useEffect((()=>{O&&h&&!d&&P.pulsate()}),[d,h,O,P]);const G=xr(P,"start",w,u),V=xr(P,"stop",y,u),q=xr(P,"stop",x,u),H=xr(P,"stop",k,u),z=xr(P,"stop",(e=>{O&&e.preventDefault(),S&&S(e)}),u),U=xr(P,"start",T,u),Q=xr(P,"stop",N,u),W=xr(P,"stop",_,u),K=xr(P,"stop",(e=>{_i(e.target)||M(!1),f&&f(e)}),!1),$=Ri((e=>{F.current||(F.current=e.currentTarget),_i(e.target)&&(M(!0),v&&v(e)),b&&b(e)})),Y=()=>{const e=F.current;return l&&"button"!==l&&!("A"===e.tagName&&e.href)},X=Ri((e=>{h&&!e.repeat&&O&&" "===e.key&&P.stop(e,(()=>{P.start(e)})),e.target===e.currentTarget&&Y()&&" "===e.key&&e.preventDefault(),j&&j(e),e.target===e.currentTarget&&Y()&&"Enter"===e.key&&!c&&(e.preventDefault(),g&&g(e))})),Z=Ri((e=>{h&&" "===e.key&&O&&!e.defaultPrevented&&P.stop(e,(()=>{P.pulsate(e)})),C&&C(e),g&&e.target===e.currentTarget&&Y()&&" "===e.key&&!e.defaultPrevented&&g(e)}));let J=l;"button"===J&&(R.href||R.to)&&(J=m);const ee={};"button"===J?(ee.type=void 0===B?"button":B,ee.disabled=c):(R.href||R.to||(ee.role="button"),c&&(ee["aria-disabled"]=c));const ne=Ei(n,F),te={...t,centerRipple:i,component:l,disabled:c,disableRipple:d,disableTouchRipple:u,focusRipple:h,tabIndex:I,focusVisible:O},oe=(e=>{const{disabled:n,focusVisible:t,focusVisibleClassName:o,classes:a}=e,i=ei({root:["root",n&&"disabled",t&&"focusVisible"]},fr,a);return t&&o&&(i.root+=" ".concat(o)),i})(te);return(0,we.jsxs)(yr,{as:J,className:Ie(oe.root,s),ownerState:te,onBlur:K,onClick:g,onContextMenu:V,onFocus:$,onKeyDown:X,onKeyUp:Z,onMouseDown:G,onMouseLeave:z,onMouseUp:H,onDragLeave:q,onTouchEnd:Q,onTouchMove:W,onTouchStart:U,ref:ne,tabIndex:c?-1:I,type:B,...ee,...R,children:[r,L?(0,we.jsx)(mr,{ref:D,center:i,...E}):null]})})),vr=Ai;function jr(e){return Ka("MuiListItemButton",e)}const Cr=$a("MuiListItemButton",["root","focusVisible","dense","alignItemsFlexStart","disabled","divider","gutters","selected"]),wr=pi(br,{shouldForwardProp:e=>ui(e)||"classes"===e,name:"MuiListItemButton",slot:"Root",overridesResolver:(e,n)=>{const{ownerState:t}=e;return[n.root,t.dense&&n.dense,"flex-start"===t.alignItems&&n.alignItemsFlexStart,t.divider&&n.divider,!t.disableGutters&&n.gutters]}})(fi((e=>{let{theme:n}=e;return{display:"flex",flexGrow:1,justifyContent:"flex-start",alignItems:"center",position:"relative",textDecoration:"none",minWidth:0,boxSizing:"border-box",textAlign:"left",paddingTop:8,paddingBottom:8,transition:n.transitions.create("background-color",{duration:n.transitions.duration.shortest}),"&:hover":{textDecoration:"none",backgroundColor:(n.vars||n).palette.action.hover,"@media (hover: none)":{backgroundColor:"transparent"}},["&.".concat(Cr.selected)]:{backgroundColor:n.vars?"rgba(".concat(n.vars.palette.primary.mainChannel," / ").concat(n.vars.palette.action.selectedOpacity,")"):qo(n.palette.primary.main,n.palette.action.selectedOpacity),["&.".concat(Cr.focusVisible)]:{backgroundColor:n.vars?"rgba(".concat(n.vars.palette.primary.mainChannel," / calc(").concat(n.vars.palette.action.selectedOpacity," + ").concat(n.vars.palette.action.focusOpacity,"))"):qo(n.palette.primary.main,n.palette.action.selectedOpacity+n.palette.action.focusOpacity)}},["&.".concat(Cr.selected,":hover")]:{backgroundColor:n.vars?"rgba(".concat(n.vars.palette.primary.mainChannel," / calc(").concat(n.vars.palette.action.selectedOpacity," + ").concat(n.vars.palette.action.hoverOpacity,"))"):qo(n.palette.primary.main,n.palette.action.selectedOpacity+n.palette.action.hoverOpacity),"@media (hover: none)":{backgroundColor:n.vars?"rgba(".concat(n.vars.palette.primary.mainChannel," / ").concat(n.vars.palette.action.selectedOpacity,")"):qo(n.palette.primary.main,n.palette.action.selectedOpacity)}},["&.".concat(Cr.focusVisible)]:{backgroundColor:(n.vars||n).palette.action.focus},["&.".concat(Cr.disabled)]:{opacity:(n.vars||n).palette.action.disabledOpacity},variants:[{props:e=>{let{ownerState:n}=e;return n.divider},style:{borderBottom:"1px solid ".concat((n.vars||n).palette.divider),backgroundClip:"padding-box"}},{props:{alignItems:"flex-start"},style:{alignItems:"flex-start"}},{props:e=>{let{ownerState:n}=e;return!n.disableGutters},style:{paddingLeft:16,paddingRight:16}},{props:e=>{let{ownerState:n}=e;return n.dense},style:{paddingTop:4,paddingBottom:4}}]}}))),Sr=a.forwardRef((function(e,n){const t=bi({props:e,name:"MuiListItemButton"}),{alignItems:o="center",autoFocus:i=!1,component:r="div",children:s,dense:l=!1,disableGutters:c=!1,divider:d=!1,focusVisibleClassName:u,selected:h=!1,className:p,...m}=t,f=a.useContext(wi),g=a.useMemo((()=>({dense:l||f.dense||!1,alignItems:o,disableGutters:c})),[o,f.dense,l,c]),y=a.useRef(null);vr((()=>{i&&y.current&&y.current.focus()}),[i]);const x={...t,alignItems:o,dense:g.dense,disableGutters:c,divider:d,selected:h},b=(e=>{const{alignItems:n,classes:t,dense:o,disabled:a,disableGutters:i,divider:r,selected:s}=e,l=ei({root:["root",o&&"dense",!i&&"gutters",r&&"divider",a&&"disabled","flex-start"===n&&"alignItemsFlexStart",s&&"selected"]},jr,t);return{...t,...l}})(x),v=Ei(y,n);return(0,we.jsx)(wi.Provider,{value:g,children:(0,we.jsx)(wr,{ref:v,href:m.href||m.to,component:(m.href||m.to)&&"div"===r?"button":r,focusVisibleClassName:Ie(b.focusVisible,u),ownerState:x,className:Ie(b.root,p),...m,classes:b,children:s})})}));function kr(e){return Ka("MuiTypography",e)}const Nr=$a("MuiTypography",["root","h1","h2","h3","h4","h5","h6","subtitle1","subtitle2","body1","body2","inherit","button","caption","overline","alignLeft","alignRight","alignCenter","alignJustify","noWrap","gutterBottom","paragraph"]);const _r=yt;function Tr(){let e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:[];return n=>{let[,t]=n;return t&&function(e){let n=arguments.length>1&&void 0!==arguments[1]?arguments[1]:[];if(!function(e){return"string"===typeof e.main}(e))return!1;for(const t of n)if(!e.hasOwnProperty(t)||"string"!==typeof e[t])return!1;return!0}(t,e)}}const Ir={primary:!0,secondary:!0,error:!0,info:!0,success:!0,warning:!0,textPrimary:!0,textSecondary:!0,textDisabled:!0},Er=So,Ar=pi("span",{name:"MuiTypography",slot:"Root",overridesResolver:(e,n)=>{const{ownerState:t}=e;return[n.root,t.variant&&n[t.variant],"inherit"!==t.align&&n["align".concat(_r(t.align))],t.noWrap&&n.noWrap,t.gutterBottom&&n.gutterBottom,t.paragraph&&n.paragraph]}})(fi((e=>{var n;let{theme:t}=e;return{margin:0,variants:[{props:{variant:"inherit"},style:{font:"inherit",lineHeight:"inherit",letterSpacing:"inherit"}},...Object.entries(t.typography).filter((e=>{let[n,t]=e;return"inherit"!==n&&t&&"object"===typeof t})).map((e=>{let[n,t]=e;return{props:{variant:n},style:t}})),...Object.entries(t.palette).filter(Tr()).map((e=>{let[n]=e;return{props:{color:n},style:{color:(t.vars||t).palette[n].main}}})),...Object.entries((null===(n=t.palette)||void 0===n?void 0:n.text)||{}).filter((e=>{let[,n]=e;return"string"===typeof n})).map((e=>{let[n]=e;return{props:{color:"text".concat(_r(n))},style:{color:(t.vars||t).palette.text[n]}}})),{props:e=>{let{ownerState:n}=e;return"inherit"!==n.align},style:{textAlign:"var(--Typography-textAlign)"}},{props:e=>{let{ownerState:n}=e;return n.noWrap},style:{overflow:"hidden",textOverflow:"ellipsis",whiteSpace:"nowrap"}},{props:e=>{let{ownerState:n}=e;return n.gutterBottom},style:{marginBottom:"0.35em"}},{props:e=>{let{ownerState:n}=e;return n.paragraph},style:{marginBottom:16}}]}}))),Br={h1:"h1",h2:"h2",h3:"h3",h4:"h4",h5:"h5",h6:"h6",subtitle1:"h6",subtitle2:"h6",body1:"p",body2:"p",inherit:"p"},Rr=a.forwardRef((function(e,n){const{color:t,...o}=bi({props:e,name:"MuiTypography"}),a=Er({...o,...!Ir[t]&&{color:t}}),{align:i="inherit",className:r,component:s,gutterBottom:l=!1,noWrap:c=!1,paragraph:d=!1,variant:u="body1",variantMapping:h=Br,...p}=a,m={...a,align:i,color:t,className:r,component:s,gutterBottom:l,noWrap:c,paragraph:d,variant:u,variantMapping:h},f=s||(d?"p":h[u]||Br[u])||"span",g=(e=>{const{align:n,gutterBottom:t,noWrap:o,paragraph:a,variant:i,classes:r}=e;return ei({root:["root",i,"inherit"!==e.align&&"align".concat(_r(n)),t&&"gutterBottom",o&&"noWrap",a&&"paragraph"]},kr,r)})(m);return(0,we.jsx)(Ar,{as:f,ref:n,className:Ie(g.root,r),...p,ownerState:m,style:{..."inherit"!==i&&{"--Typography-textAlign":i},...p.style}})})),Fr=Rr;function Pr(e){return Ka("MuiListItemText",e)}const Dr=$a("MuiListItemText",["root","multiline","dense","inset","primary","secondary"]);const Or=function(e){return"string"===typeof e};const Mr=function(e,n,t){return void 0===e||Or(e)?n:{...n,ownerState:{...n.ownerState,...t}}};const Lr=function(e,n,t){return"function"===typeof e?e(n,t):e};const Gr=function(e){let n=arguments.length>1&&void 0!==arguments[1]?arguments[1]:[];if(void 0===e)return{};const t={};return Object.keys(e).filter((t=>t.match(/^on[A-Z]/)&&"function"===typeof e[t]&&!n.includes(t))).forEach((n=>{t[n]=e[n]})),t};const Vr=function(e){if(void 0===e)return{};const n={};return Object.keys(e).filter((n=>!(n.match(/^on[A-Z]/)&&"function"===typeof e[n]))).forEach((t=>{n[t]=e[t]})),n};const qr=function(e){const{getSlotProps:n,additionalProps:t,externalSlotProps:o,externalForwardedProps:a,className:i}=e;if(!n){const e=Ie(null===t||void 0===t?void 0:t.className,i,null===a||void 0===a?void 0:a.className,null===o||void 0===o?void 0:o.className),n={...null===t||void 0===t?void 0:t.style,...null===a||void 0===a?void 0:a.style,...null===o||void 0===o?void 0:o.style},r={...t,...a,...o};return e.length>0&&(r.className=e),Object.keys(n).length>0&&(r.style=n),{props:r,internalRef:void 0}}const r=Gr({...a,...o}),s=Vr(o),l=Vr(a),c=n(r),d=Ie(null===c||void 0===c?void 0:c.className,null===t||void 0===t?void 0:t.className,i,null===a||void 0===a?void 0:a.className,null===o||void 0===o?void 0:o.className),u={...null===c||void 0===c?void 0:c.style,...null===t||void 0===t?void 0:t.style,...null===a||void 0===a?void 0:a.style,...null===o||void 0===o?void 0:o.style},h={...c,...t,...l,...s};return d.length>0&&(h.className=d),Object.keys(u).length>0&&(h.style=u),{props:h,internalRef:c.ref}};function Hr(e,n){const{className:t,elementType:o,ownerState:a,externalForwardedProps:i,internalForwardedProps:r,shouldForwardComponentProp:s=!1,...l}=n,{component:c,slots:d={[e]:void 0},slotProps:u={[e]:void 0},...h}=i,p=d[e]||o,m=Lr(u[e],a),{props:{component:f,...g},internalRef:y}=qr({className:t,...l,externalForwardedProps:"root"===e?h:void 0,externalSlotProps:m}),x=Ii(y,null===m||void 0===m?void 0:m.ref,n.ref),b="root"===e?f||c:f;return[p,Mr(p,{..."root"===e&&!c&&!d[e]&&r,..."root"!==e&&!d[e]&&r,...g,...b&&!s&&{as:b},...b&&s&&{component:b},ref:x},a)]}const zr=pi("div",{name:"MuiListItemText",slot:"Root",overridesResolver:(e,n)=>{const{ownerState:t}=e;return[{["& .".concat(Dr.primary)]:n.primary},{["& .".concat(Dr.secondary)]:n.secondary},n.root,t.inset&&n.inset,t.primary&&t.secondary&&n.multiline,t.dense&&n.dense]}})({flex:"1 1 auto",minWidth:0,marginTop:4,marginBottom:4,[".".concat(Nr.root,":where(& .").concat(Dr.primary,")")]:{display:"block"},[".".concat(Nr.root,":where(& .").concat(Dr.secondary,")")]:{display:"block"},variants:[{props:e=>{let{ownerState:n}=e;return n.primary&&n.secondary},style:{marginTop:6,marginBottom:6}},{props:e=>{let{ownerState:n}=e;return n.inset},style:{paddingLeft:56}}]}),Ur=a.forwardRef((function(e,n){const t=bi({props:e,name:"MuiListItemText"}),{children:o,className:i,disableTypography:r=!1,inset:s=!1,primary:l,primaryTypographyProps:c,secondary:d,secondaryTypographyProps:u,slots:h={},slotProps:p={},...m}=t,{dense:f}=a.useContext(wi);let g=null!=l?l:o,y=d;const x={...t,disableTypography:r,inset:s,primary:!!g,secondary:!!y,dense:f},b=(e=>{const{classes:n,inset:t,primary:o,secondary:a,dense:i}=e;return ei({root:["root",t&&"inset",i&&"dense",o&&a&&"multiline"],primary:["primary"],secondary:["secondary"]},Pr,n)})(x),v={slots:h,slotProps:{primary:c,secondary:u,...p}},[j,C]=Hr("primary",{className:b.primary,elementType:Fr,externalForwardedProps:v,ownerState:x}),[w,S]=Hr("secondary",{className:b.secondary,elementType:Fr,externalForwardedProps:v,ownerState:x});return null==g||g.type===Fr||r||(g=(0,we.jsx)(j,{variant:f?"body2":"body1",component:null!==C&&void 0!==C&&C.variant?void 0:"span",...C,children:g})),null==y||y.type===Fr||r||(y=(0,we.jsx)(w,{variant:"body2",color:"textSecondary",...S,children:y})),(0,we.jsxs)(zr,{className:Ie(b.root,i),ownerState:x,ref:n,...m,children:[g,y]})})),Qr=!1;var Wr="unmounted",Kr="exited",$r="entering",Yr="entered",Xr="exiting",Zr=function(e){function n(n,t){var o;o=e.call(this,n,t)||this;var a,i=t&&!t.isMounting?n.enter:n.appear;return o.appearStatus=null,n.in?i?(a=Kr,o.appearStatus=$r):a=Yr:a=n.unmountOnExit||n.mountOnEnter?Wr:Kr,o.state={status:a},o.nextCallback=null,o}Hi(n,e),n.getDerivedStateFromProps=function(e,n){return e.in&&n.status===Wr?{status:Kr}:null};var t=n.prototype;return t.componentDidMount=function(){this.updateStatus(!0,this.appearStatus)},t.componentDidUpdate=function(e){var n=null;if(e!==this.props){var t=this.state.status;this.props.in?t!==$r&&t!==Yr&&(n=$r):t!==$r&&t!==Yr||(n=Xr)}this.updateStatus(!1,n)},t.componentWillUnmount=function(){this.cancelNextCallback()},t.getTimeouts=function(){var e,n,t,o=this.props.timeout;return e=n=t=o,null!=o&&"number"!==typeof o&&(e=o.exit,n=o.enter,t=void 0!==o.appear?o.appear:n),{exit:e,enter:n,appear:t}},t.updateStatus=function(e,n){if(void 0===e&&(e=!1),null!==n)if(this.cancelNextCallback(),n===$r){if(this.props.unmountOnExit||this.props.mountOnEnter){var t=this.props.nodeRef?this.props.nodeRef.current:s.findDOMNode(this);t&&function(e){e.scrollTop}(t)}this.performEnter(e)}else this.performExit();else this.props.unmountOnExit&&this.state.status===Kr&&this.setState({status:Wr})},t.performEnter=function(e){var n=this,t=this.props.enter,o=this.context?this.context.isMounting:e,a=this.props.nodeRef?[o]:[s.findDOMNode(this),o],i=a[0],r=a[1],l=this.getTimeouts(),c=o?l.appear:l.enter;!e&&!t||Qr?this.safeSetState({status:Yr},(function(){n.props.onEntered(i)})):(this.props.onEnter(i,r),this.safeSetState({status:$r},(function(){n.props.onEntering(i,r),n.onTransitionEnd(c,(function(){n.safeSetState({status:Yr},(function(){n.props.onEntered(i,r)}))}))})))},t.performExit=function(){var e=this,n=this.props.exit,t=this.getTimeouts(),o=this.props.nodeRef?void 0:s.findDOMNode(this);n&&!Qr?(this.props.onExit(o),this.safeSetState({status:Xr},(function(){e.props.onExiting(o),e.onTransitionEnd(t.exit,(function(){e.safeSetState({status:Kr},(function(){e.props.onExited(o)}))}))}))):this.safeSetState({status:Kr},(function(){e.props.onExited(o)}))},t.cancelNextCallback=function(){null!==this.nextCallback&&(this.nextCallback.cancel(),this.nextCallback=null)},t.safeSetState=function(e,n){n=this.setNextCallback(n),this.setState(e,n)},t.setNextCallback=function(e){var n=this,t=!0;return this.nextCallback=function(o){t&&(t=!1,n.nextCallback=null,e(o))},this.nextCallback.cancel=function(){t=!1},this.nextCallback},t.onTransitionEnd=function(e,n){this.setNextCallback(n);var t=this.props.nodeRef?this.props.nodeRef.current:s.findDOMNode(this),o=null==e&&!this.props.addEndListener;if(t&&!o){if(this.props.addEndListener){var a=this.props.nodeRef?[this.nextCallback]:[t,this.nextCallback],i=a[0],r=a[1];this.props.addEndListener(i,r)}null!=e&&setTimeout(this.nextCallback,e)}else setTimeout(this.nextCallback,0)},t.render=function(){var e=this.state.status;if(e===Wr)return null;var n=this.props,t=n.children,o=(n.in,n.mountOnEnter,n.unmountOnExit,n.appear,n.enter,n.exit,n.timeout,n.addEndListener,n.onEnter,n.onEntering,n.onEntered,n.onExit,n.onExiting,n.onExited,n.nodeRef,Vi(n,["children","in","mountOnEnter","unmountOnExit","appear","enter","exit","timeout","addEndListener","onEnter","onEntering","onEntered","onExit","onExiting","onExited","nodeRef"]));return a.createElement(zi.Provider,{value:null},"function"===typeof t?t(e,o):a.cloneElement(a.Children.only(t),o))},n}(a.Component);function Jr(){}Zr.contextType=zi,Zr.propTypes={},Zr.defaultProps={in:!1,mountOnEnter:!1,unmountOnExit:!1,appear:!1,enter:!0,exit:!0,onEnter:Jr,onEntering:Jr,onEntered:Jr,onExit:Jr,onExiting:Jr,onExited:Jr},Zr.UNMOUNTED=Wr,Zr.EXITED=Kr,Zr.ENTERING=$r,Zr.ENTERED=Yr,Zr.EXITING=Xr;const es=Zr;function ns(){const e=Bo(ci);return e[Qa]||e}const ts=e=>e.scrollTop;function os(e,n){var t,o;const{timeout:a,easing:i,style:r={}}=e;return{duration:null!==(t=r.transitionDuration)&&void 0!==t?t:"number"===typeof a?a:a[n.mode]||0,easing:null!==(o=r.transitionTimingFunction)&&void 0!==o?o:"object"===typeof i?i[n.mode]:i,delay:r.transitionDelay}}function as(e){return Ka("MuiCollapse",e)}$a("MuiCollapse",["root","horizontal","vertical","entered","hidden","wrapper","wrapperInner"]);const is=pi("div",{name:"MuiCollapse",slot:"Root",overridesResolver:(e,n)=>{const{ownerState:t}=e;return[n.root,n[t.orientation],"entered"===t.state&&n.entered,"exited"===t.state&&!t.in&&"0px"===t.collapsedSize&&n.hidden]}})(fi((e=>{let{theme:n}=e;return{height:0,overflow:"hidden",transition:n.transitions.create("height"),variants:[{props:{orientation:"horizontal"},style:{height:"auto",width:0,transition:n.transitions.create("width")}},{props:{state:"entered"},style:{height:"auto",overflow:"visible"}},{props:{state:"entered",orientation:"horizontal"},style:{width:"auto"}},{props:e=>{let{ownerState:n}=e;return"exited"===n.state&&!n.in&&"0px"===n.collapsedSize},style:{visibility:"hidden"}}]}}))),rs=pi("div",{name:"MuiCollapse",slot:"Wrapper",overridesResolver:(e,n)=>n.wrapper})({display:"flex",width:"100%",variants:[{props:{orientation:"horizontal"},style:{width:"auto",height:"100%"}}]}),ss=pi("div",{name:"MuiCollapse",slot:"WrapperInner",overridesResolver:(e,n)=>n.wrapperInner})({width:"100%",variants:[{props:{orientation:"horizontal"},style:{width:"auto",height:"100%"}}]}),ls=a.forwardRef((function(e,n){const t=bi({props:e,name:"MuiCollapse"}),{addEndListener:o,children:i,className:r,collapsedSize:s="0px",component:l,easing:c,in:d,onEnter:u,onEntered:h,onEntering:p,onExit:m,onExited:f,onExiting:g,orientation:y="vertical",style:x,timeout:b=wa.standard,TransitionComponent:v=es,...j}=t,C={...t,orientation:y,collapsedSize:s},w=(e=>{const{orientation:n,classes:t}=e;return ei({root:["root","".concat(n)],entered:["entered"],hidden:["hidden"],wrapper:["wrapper","".concat(n)],wrapperInner:["wrapperInner","".concat(n)]},as,t)})(C),S=ns(),k=Ji(),N=a.useRef(null),_=a.useRef(),T="number"===typeof s?"".concat(s,"px"):s,I="horizontal"===y,E=I?"width":"height",A=a.useRef(null),B=Ei(n,A),R=e=>n=>{if(e){const t=A.current;void 0===n?e(t):e(t,n)}},F=()=>N.current?N.current[I?"clientWidth":"clientHeight"]:0,P=R(((e,n)=>{N.current&&I&&(N.current.style.position="absolute"),e.style[E]=T,u&&u(e,n)})),D=R(((e,n)=>{const t=F();N.current&&I&&(N.current.style.position="");const{duration:o,easing:a}=os({style:x,timeout:b,easing:c},{mode:"enter"});if("auto"===b){const n=S.transitions.getAutoHeightDuration(t);e.style.transitionDuration="".concat(n,"ms"),_.current=n}else e.style.transitionDuration="string"===typeof o?o:"".concat(o,"ms");e.style[E]="".concat(t,"px"),e.style.transitionTimingFunction=a,p&&p(e,n)})),O=R(((e,n)=>{e.style[E]="auto",h&&h(e,n)})),M=R((e=>{e.style[E]="".concat(F(),"px"),m&&m(e)})),L=R(f),G=R((e=>{const n=F(),{duration:t,easing:o}=os({style:x,timeout:b,easing:c},{mode:"exit"});if("auto"===b){const t=S.transitions.getAutoHeightDuration(n);e.style.transitionDuration="".concat(t,"ms"),_.current=t}else e.style.transitionDuration="string"===typeof t?t:"".concat(t,"ms");e.style[E]=T,e.style.transitionTimingFunction=o,g&&g(e)}));return(0,we.jsx)(v,{in:d,onEnter:P,onEntered:O,onEntering:D,onExit:M,onExited:L,onExiting:G,addEndListener:e=>{"auto"===b&&k.start(_.current||0,e),o&&o(A.current,e)},nodeRef:A,timeout:"auto"===b?null:b,...j,children:(e,n)=>{let{ownerState:t,...o}=n;return(0,we.jsx)(is,{as:l,className:Ie(w.root,r,{entered:w.entered,exited:!d&&"0px"===T&&w.hidden}[e]),style:{[I?"minWidth":"minHeight"]:T,...x},ref:B,ownerState:{...C,state:e},...o,children:(0,we.jsx)(rs,{ownerState:{...C,state:e},className:w.wrapper,ref:N,children:(0,we.jsx)(ss,{ownerState:{...C,state:e},className:w.wrapperInner,children:i})})})}})}));ls&&(ls.muiSupportAuto=!0);const cs=ls;let ds=0;const us={...i}.useId;const hs=function(e){if(void 0!==us){const n=us();return null!==e&&void 0!==e?e:n}return function(e){const[n,t]=a.useState(e),o=e||n;return a.useEffect((()=>{null==n&&(ds+=1,t("mui-".concat(ds)))}),[n]),o}(e)};function ps(e){return Ka("MuiCircularProgress",e)}$a("MuiCircularProgress",["root","determinate","indeterminate","colorPrimary","colorSecondary","svg","circle","circleDeterminate","circleIndeterminate","circleDisableShrink"]);var ms,fs,gs,ys;const xs=44,bs=tr(ms||(ms=Gi(["\n  0% {\n    transform: rotate(0deg);\n  }\n\n  100% {\n    transform: rotate(360deg);\n  }\n"]))),vs=tr(fs||(fs=Gi(["\n  0% {\n    stroke-dasharray: 1px, 200px;\n    stroke-dashoffset: 0;\n  }\n\n  50% {\n    stroke-dasharray: 100px, 200px;\n    stroke-dashoffset: -15px;\n  }\n\n  100% {\n    stroke-dasharray: 1px, 200px;\n    stroke-dashoffset: -126px;\n  }\n"]))),js="string"!==typeof bs?nr(gs||(gs=Gi(["\n        animation: "," 1.4s linear infinite;\n      "])),bs):null,Cs="string"!==typeof vs?nr(ys||(ys=Gi(["\n        animation: "," 1.4s ease-in-out infinite;\n      "])),vs):null,ws=pi("span",{name:"MuiCircularProgress",slot:"Root",overridesResolver:(e,n)=>{const{ownerState:t}=e;return[n.root,n[t.variant],n["color".concat(_r(t.color))]]}})(fi((e=>{let{theme:n}=e;return{display:"inline-block",variants:[{props:{variant:"determinate"},style:{transition:n.transitions.create("transform")}},{props:{variant:"indeterminate"},style:js||{animation:"".concat(bs," 1.4s linear infinite")}},...Object.entries(n.palette).filter(Tr()).map((e=>{let[t]=e;return{props:{color:t},style:{color:(n.vars||n).palette[t].main}}}))]}}))),Ss=pi("svg",{name:"MuiCircularProgress",slot:"Svg",overridesResolver:(e,n)=>n.svg})({display:"block"}),ks=pi("circle",{name:"MuiCircularProgress",slot:"Circle",overridesResolver:(e,n)=>{const{ownerState:t}=e;return[n.circle,n["circle".concat(_r(t.variant))],t.disableShrink&&n.circleDisableShrink]}})(fi((e=>{let{theme:n}=e;return{stroke:"currentColor",variants:[{props:{variant:"determinate"},style:{transition:n.transitions.create("stroke-dashoffset")}},{props:{variant:"indeterminate"},style:{strokeDasharray:"80px, 200px",strokeDashoffset:0}},{props:e=>{let{ownerState:n}=e;return"indeterminate"===n.variant&&!n.disableShrink},style:Cs||{animation:"".concat(vs," 1.4s ease-in-out infinite")}}]}}))),Ns=a.forwardRef((function(e,n){const t=bi({props:e,name:"MuiCircularProgress"}),{className:o,color:a="primary",disableShrink:i=!1,size:r=40,style:s,thickness:l=3.6,value:c=0,variant:d="indeterminate",...u}=t,h={...t,color:a,disableShrink:i,size:r,thickness:l,value:c,variant:d},p=(e=>{const{classes:n,variant:t,color:o,disableShrink:a}=e;return ei({root:["root",t,"color".concat(_r(o))],svg:["svg"],circle:["circle","circle".concat(_r(t)),a&&"circleDisableShrink"]},ps,n)})(h),m={},f={},g={};if("determinate"===d){const e=2*Math.PI*((xs-l)/2);m.strokeDasharray=e.toFixed(3),g["aria-valuenow"]=Math.round(c),m.strokeDashoffset="".concat(((100-c)/100*e).toFixed(3),"px"),f.transform="rotate(-90deg)"}return(0,we.jsx)(ws,{className:Ie(p.root,o),style:{width:r,height:r,...f,...s},ownerState:h,ref:n,role:"progressbar",...g,...u,children:(0,we.jsx)(Ss,{className:p.svg,ownerState:h,viewBox:"".concat(22," ").concat(22," ").concat(xs," ").concat(xs),children:(0,we.jsx)(ks,{className:p.circle,style:m,ownerState:h,cx:xs,cy:xs,r:(xs-l)/2,fill:"none",strokeWidth:l})})})})),_s=Ns;function Ts(e){return Ka("MuiIconButton",e)}const Is=$a("MuiIconButton",["root","disabled","colorInherit","colorPrimary","colorSecondary","colorError","colorInfo","colorSuccess","colorWarning","edgeStart","edgeEnd","sizeSmall","sizeMedium","sizeLarge","loading","loadingIndicator"]),Es=pi(br,{name:"MuiIconButton",slot:"Root",overridesResolver:(e,n)=>{const{ownerState:t}=e;return[n.root,t.loading&&n.loading,"default"!==t.color&&n["color".concat(_r(t.color))],t.edge&&n["edge".concat(_r(t.edge))],n["size".concat(_r(t.size))]]}})(fi((e=>{let{theme:n}=e;return{textAlign:"center",flex:"0 0 auto",fontSize:n.typography.pxToRem(24),padding:8,borderRadius:"50%",color:(n.vars||n).palette.action.active,transition:n.transitions.create("background-color",{duration:n.transitions.duration.shortest}),variants:[{props:e=>!e.disableRipple,style:{"--IconButton-hoverBg":n.vars?"rgba(".concat(n.vars.palette.action.activeChannel," / ").concat(n.vars.palette.action.hoverOpacity,")"):qo(n.palette.action.active,n.palette.action.hoverOpacity),"&:hover":{backgroundColor:"var(--IconButton-hoverBg)","@media (hover: none)":{backgroundColor:"transparent"}}}},{props:{edge:"start"},style:{marginLeft:-12}},{props:{edge:"start",size:"small"},style:{marginLeft:-3}},{props:{edge:"end"},style:{marginRight:-12}},{props:{edge:"end",size:"small"},style:{marginRight:-3}}]}})),fi((e=>{let{theme:n}=e;return{variants:[{props:{color:"inherit"},style:{color:"inherit"}},...Object.entries(n.palette).filter(Tr()).map((e=>{let[t]=e;return{props:{color:t},style:{color:(n.vars||n).palette[t].main}}})),...Object.entries(n.palette).filter(Tr()).map((e=>{let[t]=e;return{props:{color:t},style:{"--IconButton-hoverBg":n.vars?"rgba(".concat((n.vars||n).palette[t].mainChannel," / ").concat(n.vars.palette.action.hoverOpacity,")"):qo((n.vars||n).palette[t].main,n.palette.action.hoverOpacity)}}})),{props:{size:"small"},style:{padding:5,fontSize:n.typography.pxToRem(18)}},{props:{size:"large"},style:{padding:12,fontSize:n.typography.pxToRem(28)}}],["&.".concat(Is.disabled)]:{backgroundColor:"transparent",color:(n.vars||n).palette.action.disabled},["&.".concat(Is.loading)]:{color:"transparent"}}}))),As=pi("span",{name:"MuiIconButton",slot:"LoadingIndicator",overridesResolver:(e,n)=>n.loadingIndicator})((e=>{let{theme:n}=e;return{display:"none",position:"absolute",visibility:"visible",top:"50%",left:"50%",transform:"translate(-50%, -50%)",color:(n.vars||n).palette.action.disabled,variants:[{props:{loading:!0},style:{display:"flex"}}]}})),Bs=a.forwardRef((function(e,n){const t=bi({props:e,name:"MuiIconButton"}),{edge:o=!1,children:a,className:i,color:r="default",disabled:s=!1,disableFocusRipple:l=!1,size:c="medium",id:d,loading:u=!1,loadingIndicator:h,...p}=t,m=hs(d),f=null!==h&&void 0!==h?h:(0,we.jsx)(_s,{"aria-labelledby":m,color:"inherit",size:16}),g={...t,edge:o,color:r,disabled:s,disableFocusRipple:l,loading:u,loadingIndicator:f,size:c},y=(e=>{const{classes:n,disabled:t,color:o,edge:a,size:i,loading:r}=e;return ei({root:["root",r&&"loading",t&&"disabled","default"!==o&&"color".concat(_r(o)),a&&"edge".concat(_r(a)),"size".concat(_r(i))],loadingIndicator:["loadingIndicator"]},Ts,n)})(g);return(0,we.jsxs)(Es,{id:m,className:Ie(y.root,i),centerRipple:!0,focusRipple:!l,disabled:s||u,ref:n,...p,ownerState:g,children:[(0,we.jsx)(As,{className:y.loadingIndicator,ownerState:g,children:u&&f}),a]})})),Rs=Bs,Fs=a.createContext();function Ps(e){var n;return parseInt(a.version,10)>=19?(null===e||void 0===e||null===(n=e.props)||void 0===n?void 0:n.ref)||null:(null===e||void 0===e?void 0:e.ref)||null}function Ds(e){return e&&e.ownerDocument||document}const Os=["input","select","textarea","a[href]","button","[tabindex]","audio[controls]","video[controls]",'[contenteditable]:not([contenteditable="false"])'].join(",");function Ms(e){const n=[],t=[];return Array.from(e.querySelectorAll(Os)).forEach(((e,o)=>{const a=function(e){const n=parseInt(e.getAttribute("tabindex")||"",10);return Number.isNaN(n)?"true"===e.contentEditable||("AUDIO"===e.nodeName||"VIDEO"===e.nodeName||"DETAILS"===e.nodeName)&&null===e.getAttribute("tabindex")?0:e.tabIndex:n}(e);-1!==a&&function(e){return!(e.disabled||"INPUT"===e.tagName&&"hidden"===e.type||function(e){if("INPUT"!==e.tagName||"radio"!==e.type)return!1;if(!e.name)return!1;const n=n=>e.ownerDocument.querySelector('input[type="radio"]'.concat(n));let t=n('[name="'.concat(e.name,'"]:checked'));return t||(t=n('[name="'.concat(e.name,'"]'))),t!==e}(e))}(e)&&(0===a?n.push(e):t.push({documentOrder:o,tabIndex:a,node:e}))})),t.sort(((e,n)=>e.tabIndex===n.tabIndex?e.documentOrder-n.documentOrder:e.tabIndex-n.tabIndex)).map((e=>e.node)).concat(n)}function Ls(){return!0}const Gs=function(e){const{children:n,disableAutoFocus:t=!1,disableEnforceFocus:o=!1,disableRestoreFocus:i=!1,getTabbable:r=Ms,isEnabled:s=Ls,open:l}=e,c=a.useRef(!1),d=a.useRef(null),u=a.useRef(null),h=a.useRef(null),p=a.useRef(null),m=a.useRef(!1),f=a.useRef(null),g=Ii(Ps(n),f),y=a.useRef(null);a.useEffect((()=>{l&&f.current&&(m.current=!t)}),[t,l]),a.useEffect((()=>{if(!l||!f.current)return;const e=Ds(f.current);return f.current.contains(e.activeElement)||(f.current.hasAttribute("tabIndex")||f.current.setAttribute("tabIndex","-1"),m.current&&f.current.focus()),()=>{i||(h.current&&h.current.focus&&(c.current=!0,h.current.focus()),h.current=null)}}),[l]),a.useEffect((()=>{if(!l||!f.current)return;const e=Ds(f.current),n=n=>{y.current=n,!o&&s()&&"Tab"===n.key&&e.activeElement===f.current&&n.shiftKey&&(c.current=!0,u.current&&u.current.focus())},t=()=>{const n=f.current;if(null===n)return;if(!e.hasFocus()||!s()||c.current)return void(c.current=!1);if(n.contains(e.activeElement))return;if(o&&e.activeElement!==d.current&&e.activeElement!==u.current)return;if(e.activeElement!==p.current)p.current=null;else if(null!==p.current)return;if(!m.current)return;let t=[];if(e.activeElement!==d.current&&e.activeElement!==u.current||(t=r(f.current)),t.length>0){var a,i;const e=Boolean((null===(a=y.current)||void 0===a?void 0:a.shiftKey)&&"Tab"===(null===(i=y.current)||void 0===i?void 0:i.key)),n=t[0],o=t[t.length-1];"string"!==typeof n&&"string"!==typeof o&&(e?o.focus():n.focus())}else n.focus()};e.addEventListener("focusin",t),e.addEventListener("keydown",n,!0);const a=setInterval((()=>{e.activeElement&&"BODY"===e.activeElement.tagName&&t()}),50);return()=>{clearInterval(a),e.removeEventListener("focusin",t),e.removeEventListener("keydown",n,!0)}}),[t,o,i,s,l,r]);const x=e=>{null===h.current&&(h.current=e.relatedTarget),m.current=!0};return(0,we.jsxs)(a.Fragment,{children:[(0,we.jsx)("div",{tabIndex:l?0:-1,onFocus:x,ref:d,"data-testid":"sentinelStart"}),a.cloneElement(n,{ref:g,onFocus:e=>{null===h.current&&(h.current=e.relatedTarget),m.current=!0,p.current=e.target;const t=n.props.onFocus;t&&t(e)}}),(0,we.jsx)("div",{tabIndex:l?0:-1,onFocus:x,ref:u,"data-testid":"sentinelEnd"})]})};const Vs=a.forwardRef((function(e,n){const{children:t,container:o,disablePortal:i=!1}=e,[r,l]=a.useState(null),c=Ii(a.isValidElement(t)?Ps(t):null,n);if(Ai((()=>{i||l(function(e){return"function"===typeof e?e():e}(o)||document.body)}),[o,i]),Ai((()=>{if(r&&!i)return Ti(n,r),()=>{Ti(n,null)}}),[n,r,i]),i){if(a.isValidElement(t)){const e={ref:c};return a.cloneElement(t,e)}return t}return r?s.createPortal(t,r):r})),qs={entering:{opacity:1},entered:{opacity:1}},Hs=a.forwardRef((function(e,n){const t=ns(),o={enter:t.transitions.duration.enteringScreen,exit:t.transitions.duration.leavingScreen},{addEndListener:i,appear:r=!0,children:s,easing:l,in:c,onEnter:d,onEntered:u,onEntering:h,onExit:p,onExited:m,onExiting:f,style:g,timeout:y=o,TransitionComponent:x=es,...b}=e,v=a.useRef(null),j=Ei(v,Ps(s),n),C=e=>n=>{if(e){const t=v.current;void 0===n?e(t):e(t,n)}},w=C(h),S=C(((e,n)=>{ts(e);const o=os({style:g,timeout:y,easing:l},{mode:"enter"});e.style.webkitTransition=t.transitions.create("opacity",o),e.style.transition=t.transitions.create("opacity",o),d&&d(e,n)})),k=C(u),N=C(f),_=C((e=>{const n=os({style:g,timeout:y,easing:l},{mode:"exit"});e.style.webkitTransition=t.transitions.create("opacity",n),e.style.transition=t.transitions.create("opacity",n),p&&p(e)})),T=C(m);return(0,we.jsx)(x,{appear:r,in:c,nodeRef:v,onEnter:S,onEntered:k,onEntering:w,onExit:_,onExited:T,onExiting:N,addEndListener:e=>{i&&i(v.current,e)},timeout:y,...b,children:(e,n)=>{let{ownerState:t,...o}=n;return a.cloneElement(s,{style:{opacity:0,visibility:"exited"!==e||c?void 0:"hidden",...qs[e],...g,...s.props.style},ref:j,...o})}})})),zs=Hs;function Us(e){return Ka("MuiBackdrop",e)}$a("MuiBackdrop",["root","invisible"]);const Qs=pi("div",{name:"MuiBackdrop",slot:"Root",overridesResolver:(e,n)=>{const{ownerState:t}=e;return[n.root,t.invisible&&n.invisible]}})({position:"fixed",display:"flex",alignItems:"center",justifyContent:"center",right:0,bottom:0,top:0,left:0,backgroundColor:"rgba(0, 0, 0, 0.5)",WebkitTapHighlightColor:"transparent",variants:[{props:{invisible:!0},style:{backgroundColor:"transparent"}}]}),Ws=a.forwardRef((function(e,n){const t=bi({props:e,name:"MuiBackdrop"}),{children:o,className:a,component:i="div",invisible:r=!1,open:s,components:l={},componentsProps:c={},slotProps:d={},slots:u={},TransitionComponent:h,transitionDuration:p,...m}=t,f={...t,component:i,invisible:r},g=(e=>{const{classes:n,invisible:t}=e;return ei({root:["root",t&&"invisible"]},Us,n)})(f),y={slots:{transition:h,root:l.Root,...u},slotProps:{...c,...d}},[x,b]=Hr("root",{elementType:Qs,externalForwardedProps:y,className:Ie(g.root,a),ownerState:f}),[v,j]=Hr("transition",{elementType:zs,externalForwardedProps:y,ownerState:f});return(0,we.jsx)(v,{in:s,timeout:p,...m,...j,children:(0,we.jsx)(x,{"aria-hidden":!0,...b,classes:g,ref:n,children:o})})}));function Ks(){for(var e=arguments.length,n=new Array(e),t=0;t<e;t++)n[t]=arguments[t];return n.reduce(((e,n)=>null==n?e:function(){for(var t=arguments.length,o=new Array(t),a=0;a<t;a++)o[a]=arguments[a];e.apply(this,o),n.apply(this,o)}),(()=>{}))}function $s(e){return Ds(e).defaultView||window}function Ys(e,n){n?e.setAttribute("aria-hidden","true"):e.removeAttribute("aria-hidden")}function Xs(e){return parseInt($s(e).getComputedStyle(e).paddingRight,10)||0}function Zs(e,n,t,o,a){const i=[n,t,...o];[].forEach.call(e.children,(e=>{const n=!i.includes(e),t=!function(e){const n=["TEMPLATE","SCRIPT","STYLE","LINK","MAP","META","NOSCRIPT","PICTURE","COL","COLGROUP","PARAM","SLOT","SOURCE","TRACK"].includes(e.tagName),t="INPUT"===e.tagName&&"hidden"===e.getAttribute("type");return n||t}(e);n&&t&&Ys(e,a)}))}function Js(e,n){let t=-1;return e.some(((e,o)=>!!n(e)&&(t=o,!0))),t}function el(e,n){const t=[],o=e.container;if(!n.disableScrollLock){if(function(e){const n=Ds(e);return n.body===e?$s(e).innerWidth>n.documentElement.clientWidth:e.scrollHeight>e.clientHeight}(o)){const e=function(){let e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:window;const n=e.document.documentElement.clientWidth;return e.innerWidth-n}($s(o));t.push({value:o.style.paddingRight,property:"padding-right",el:o}),o.style.paddingRight="".concat(Xs(o)+e,"px");const n=Ds(o).querySelectorAll(".mui-fixed");[].forEach.call(n,(n=>{t.push({value:n.style.paddingRight,property:"padding-right",el:n}),n.style.paddingRight="".concat(Xs(n)+e,"px")}))}let e;if(o.parentNode instanceof DocumentFragment)e=Ds(o).body;else{const n=o.parentElement,t=$s(o);e="HTML"===(null===n||void 0===n?void 0:n.nodeName)&&"scroll"===t.getComputedStyle(n).overflowY?n:o}t.push({value:e.style.overflow,property:"overflow",el:e},{value:e.style.overflowX,property:"overflow-x",el:e},{value:e.style.overflowY,property:"overflow-y",el:e}),e.style.overflow="hidden"}return()=>{t.forEach((e=>{let{value:n,el:t,property:o}=e;n?t.style.setProperty(o,n):t.style.removeProperty(o)}))}}const nl=()=>{},tl=new class{constructor(){this.modals=[],this.containers=[]}add(e,n){let t=this.modals.indexOf(e);if(-1!==t)return t;t=this.modals.length,this.modals.push(e),e.modalRef&&Ys(e.modalRef,!1);const o=function(e){const n=[];return[].forEach.call(e.children,(e=>{"true"===e.getAttribute("aria-hidden")&&n.push(e)})),n}(n);Zs(n,e.mount,e.modalRef,o,!0);const a=Js(this.containers,(e=>e.container===n));return-1!==a?(this.containers[a].modals.push(e),t):(this.containers.push({modals:[e],container:n,restore:null,hiddenSiblings:o}),t)}mount(e,n){const t=Js(this.containers,(n=>n.modals.includes(e))),o=this.containers[t];o.restore||(o.restore=el(o,n))}remove(e){let n=!(arguments.length>1&&void 0!==arguments[1])||arguments[1];const t=this.modals.indexOf(e);if(-1===t)return t;const o=Js(this.containers,(n=>n.modals.includes(e))),a=this.containers[o];if(a.modals.splice(a.modals.indexOf(e),1),this.modals.splice(t,1),0===a.modals.length)a.restore&&a.restore(),e.modalRef&&Ys(e.modalRef,n),Zs(a.container,e.mount,e.modalRef,a.hiddenSiblings,!1),this.containers.splice(o,1);else{const e=a.modals[a.modals.length-1];e.modalRef&&Ys(e.modalRef,!1)}return t}isTopModal(e){return this.modals.length>0&&this.modals[this.modals.length-1]===e}};const ol=function(e){const{container:n,disableEscapeKeyDown:t=!1,disableScrollLock:o=!1,closeAfterTransition:i=!1,onTransitionEnter:r,onTransitionExited:s,children:l,onClose:c,open:d,rootRef:u}=e,h=a.useRef({}),p=a.useRef(null),m=a.useRef(null),f=Ii(m,u),[g,y]=a.useState(!d),x=function(e){return!!e&&e.props.hasOwnProperty("in")}(l);let b=!0;"false"!==e["aria-hidden"]&&!1!==e["aria-hidden"]||(b=!1);const v=()=>(h.current.modalRef=m.current,h.current.mount=p.current,h.current),j=()=>{tl.mount(v(),{disableScrollLock:o}),m.current&&(m.current.scrollTop=0)},C=Bi((()=>{const e=function(e){return"function"===typeof e?e():e}(n)||Ds(p.current).body;tl.add(v(),e),m.current&&j()})),w=()=>tl.isTopModal(v()),S=Bi((e=>{p.current=e,e&&(d&&w()?j():m.current&&Ys(m.current,b))})),k=a.useCallback((()=>{tl.remove(v(),b)}),[b]);a.useEffect((()=>()=>{k()}),[k]),a.useEffect((()=>{d?C():x&&i||k()}),[d,k,x,i,C]);const N=e=>n=>{var o;null===(o=e.onKeyDown)||void 0===o||o.call(e,n),"Escape"===n.key&&229!==n.which&&w()&&(t||(n.stopPropagation(),c&&c(n,"escapeKeyDown")))},_=e=>n=>{var t;null===(t=e.onClick)||void 0===t||t.call(e,n),n.target===n.currentTarget&&c&&c(n,"backdropClick")};return{getRootProps:function(){let n=arguments.length>0&&void 0!==arguments[0]?arguments[0]:{};const t=Gr(e);delete t.onTransitionEnter,delete t.onTransitionExited;const o={...t,...n};return{role:"presentation",...o,onKeyDown:N(o),ref:f}},getBackdropProps:function(){const e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:{};return{"aria-hidden":!0,...e,onClick:_(e),open:d}},getTransitionProps:()=>{var e,n;return{onEnter:Ks((()=>{y(!1),r&&r()}),null!==(e=null===l||void 0===l?void 0:l.props.onEnter)&&void 0!==e?e:nl),onExited:Ks((()=>{y(!0),s&&s(),i&&k()}),null!==(n=null===l||void 0===l?void 0:l.props.onExited)&&void 0!==n?n:nl)}},rootRef:f,portalRef:S,isTopModal:w,exited:g,hasTransition:x}};function al(e){return Ka("MuiModal",e)}$a("MuiModal",["root","hidden","backdrop"]);const il=pi("div",{name:"MuiModal",slot:"Root",overridesResolver:(e,n)=>{const{ownerState:t}=e;return[n.root,!t.open&&t.exited&&n.hidden]}})(fi((e=>{let{theme:n}=e;return{position:"fixed",zIndex:(n.vars||n).zIndex.modal,right:0,bottom:0,top:0,left:0,variants:[{props:e=>{let{ownerState:n}=e;return!n.open&&n.exited},style:{visibility:"hidden"}}]}}))),rl=pi(Ws,{name:"MuiModal",slot:"Backdrop",overridesResolver:(e,n)=>n.backdrop})({zIndex:-1}),sl=a.forwardRef((function(e,n){const t=bi({name:"MuiModal",props:e}),{BackdropComponent:o=rl,BackdropProps:i,classes:r,className:s,closeAfterTransition:l=!1,children:c,container:d,component:u,components:h={},componentsProps:p={},disableAutoFocus:m=!1,disableEnforceFocus:f=!1,disableEscapeKeyDown:g=!1,disablePortal:y=!1,disableRestoreFocus:x=!1,disableScrollLock:b=!1,hideBackdrop:v=!1,keepMounted:j=!1,onBackdropClick:C,onClose:w,onTransitionEnter:S,onTransitionExited:k,open:N,slotProps:_={},slots:T={},theme:I,...E}=t,A={...t,closeAfterTransition:l,disableAutoFocus:m,disableEnforceFocus:f,disableEscapeKeyDown:g,disablePortal:y,disableRestoreFocus:x,disableScrollLock:b,hideBackdrop:v,keepMounted:j},{getRootProps:B,getBackdropProps:R,getTransitionProps:F,portalRef:P,isTopModal:D,exited:O,hasTransition:M}=ol({...A,rootRef:n}),L={...A,exited:O},G=(e=>{const{open:n,exited:t,classes:o}=e;return ei({root:["root",!n&&t&&"hidden"],backdrop:["backdrop"]},al,o)})(L),V={};if(void 0===c.props.tabIndex&&(V.tabIndex="-1"),M){const{onEnter:e,onExited:n}=F();V.onEnter=e,V.onExited=n}const q={...E,slots:{root:h.Root,backdrop:h.Backdrop,...T},slotProps:{...p,..._}},[H,z]=Hr("root",{elementType:il,externalForwardedProps:q,getSlotProps:B,additionalProps:{ref:n,as:u},ownerState:L,className:Ie(s,null===G||void 0===G?void 0:G.root,!L.open&&L.exited&&(null===G||void 0===G?void 0:G.hidden))}),[U,Q]=Hr("backdrop",{elementType:o,externalForwardedProps:q,additionalProps:i,getSlotProps:e=>R({...e,onClick:n=>{C&&C(n),null!==e&&void 0!==e&&e.onClick&&e.onClick(n)}}),className:Ie(null===i||void 0===i?void 0:i.className,null===G||void 0===G?void 0:G.backdrop),ownerState:L}),W=Ei(null===i||void 0===i?void 0:i.ref,Q.ref);return j||N||M&&!O?(0,we.jsx)(Vs,{ref:P,container:d,disablePortal:y,children:(0,we.jsxs)(H,{...z,children:[!v&&o?(0,we.jsx)(U,{...Q,ref:W}):null,(0,we.jsx)(Gs,{disableEnforceFocus:f,disableAutoFocus:m,disableRestoreFocus:x,isEnabled:D,open:N,children:a.cloneElement(c,V)})]})}):null})),ll=sl;const cl=function(e){let n,t=arguments.length>1&&void 0!==arguments[1]?arguments[1]:166;function o(){for(var o=arguments.length,a=new Array(o),i=0;i<o;i++)a[i]=arguments[i];clearTimeout(n),n=setTimeout((()=>{e.apply(this,a)}),t)}return o.clear=()=>{clearTimeout(n)},o},dl=$s;function ul(e,n,t){var o;const a=function(e,n,t){const o=n.getBoundingClientRect(),a=t&&t.getBoundingClientRect(),i=dl(n);let r;if(n.fakeTransform)r=n.fakeTransform;else{const e=i.getComputedStyle(n);r=e.getPropertyValue("-webkit-transform")||e.getPropertyValue("transform")}let s=0,l=0;if(r&&"none"!==r&&"string"===typeof r){const e=r.split("(")[1].split(")")[0].split(",");s=parseInt(e[4],10),l=parseInt(e[5],10)}return"left"===e?"translateX(".concat(a?a.right+s-o.left:i.innerWidth+s-o.left,"px)"):"right"===e?"translateX(-".concat(a?o.right-a.left-s:o.left+o.width-s,"px)"):"up"===e?"translateY(".concat(a?a.bottom+l-o.top:i.innerHeight+l-o.top,"px)"):"translateY(-".concat(a?o.top-a.top+o.height-l:o.top+o.height-l,"px)")}(e,n,"function"===typeof(o=t)?o():o);a&&(n.style.webkitTransform=a,n.style.transform=a)}const hl=a.forwardRef((function(e,n){const t=ns(),o={enter:t.transitions.easing.easeOut,exit:t.transitions.easing.sharp},i={enter:t.transitions.duration.enteringScreen,exit:t.transitions.duration.leavingScreen},{addEndListener:r,appear:s=!0,children:l,container:c,direction:d="down",easing:u=o,in:h,onEnter:p,onEntered:m,onEntering:f,onExit:g,onExited:y,onExiting:x,style:b,timeout:v=i,TransitionComponent:j=es,...C}=e,w=a.useRef(null),S=Ei(Ps(l),w,n),k=e=>n=>{e&&(void 0===n?e(w.current):e(w.current,n))},N=k(((e,n)=>{ul(d,e,c),ts(e),p&&p(e,n)})),_=k(((e,n)=>{const o=os({timeout:v,style:b,easing:u},{mode:"enter"});e.style.webkitTransition=t.transitions.create("-webkit-transform",{...o}),e.style.transition=t.transitions.create("transform",{...o}),e.style.webkitTransform="none",e.style.transform="none",f&&f(e,n)})),T=k(m),I=k(x),E=k((e=>{const n=os({timeout:v,style:b,easing:u},{mode:"exit"});e.style.webkitTransition=t.transitions.create("-webkit-transform",n),e.style.transition=t.transitions.create("transform",n),ul(d,e,c),g&&g(e)})),A=k((e=>{e.style.webkitTransition="",e.style.transition="",y&&y(e)})),B=a.useCallback((()=>{w.current&&ul(d,w.current,c)}),[d,c]);return a.useEffect((()=>{if(h||"down"===d||"right"===d)return;const e=cl((()=>{w.current&&ul(d,w.current,c)})),n=dl(w.current);return n.addEventListener("resize",e),()=>{e.clear(),n.removeEventListener("resize",e)}}),[d,h,c]),a.useEffect((()=>{h||B()}),[h,B]),(0,we.jsx)(j,{nodeRef:w,onEnter:N,onEntered:T,onEntering:_,onExit:E,onExited:A,onExiting:I,addEndListener:e=>{r&&r(w.current,e)},appear:s,in:h,timeout:v,...C,children:(e,n)=>{let{ownerState:t,...o}=n;return a.cloneElement(l,{ref:S,style:{visibility:"exited"!==e||h?void 0:"hidden",...b,...l.props.style},...o})}})})),pl=hl;function ml(e){return Ka("MuiPaper",e)}$a("MuiPaper",["root","rounded","outlined","elevation","elevation0","elevation1","elevation2","elevation3","elevation4","elevation5","elevation6","elevation7","elevation8","elevation9","elevation10","elevation11","elevation12","elevation13","elevation14","elevation15","elevation16","elevation17","elevation18","elevation19","elevation20","elevation21","elevation22","elevation23","elevation24"]);const fl=pi("div",{name:"MuiPaper",slot:"Root",overridesResolver:(e,n)=>{const{ownerState:t}=e;return[n.root,n[t.variant],!t.square&&n.rounded,"elevation"===t.variant&&n["elevation".concat(t.elevation)]]}})(fi((e=>{let{theme:n}=e;return{backgroundColor:(n.vars||n).palette.background.paper,color:(n.vars||n).palette.text.primary,transition:n.transitions.create("box-shadow"),variants:[{props:e=>{let{ownerState:n}=e;return!n.square},style:{borderRadius:n.shape.borderRadius}},{props:{variant:"outlined"},style:{border:"1px solid ".concat((n.vars||n).palette.divider)}},{props:{variant:"elevation"},style:{boxShadow:"var(--Paper-shadow)",backgroundImage:"var(--Paper-overlay)"}}]}}))),gl=a.forwardRef((function(e,n){var t;const o=bi({props:e,name:"MuiPaper"}),a=ns(),{className:i,component:r="div",elevation:s=1,square:l=!1,variant:c="elevation",...d}=o,u={...o,component:r,elevation:s,square:l,variant:c},h=(e=>{const{square:n,elevation:t,variant:o,classes:a}=e;return ei({root:["root",o,!n&&"rounded","elevation"===o&&"elevation".concat(t)]},ml,a)})(u);return(0,we.jsx)(fl,{as:r,ownerState:u,className:Ie(h.root,i),ref:n,...d,style:{..."elevation"===c&&{"--Paper-shadow":(a.vars||a).shadows[s],...a.vars&&{"--Paper-overlay":null===(t=a.vars.overlays)||void 0===t?void 0:t[s]},...!a.vars&&"dark"===a.palette.mode&&{"--Paper-overlay":"linear-gradient(".concat(qo("#fff",Aa(s)),", ").concat(qo("#fff",Aa(s)),")")}},...d.style}})}));function yl(e){return Ka("MuiDrawer",e)}$a("MuiDrawer",["root","docked","paper","paperAnchorLeft","paperAnchorRight","paperAnchorTop","paperAnchorBottom","paperAnchorDockedLeft","paperAnchorDockedRight","paperAnchorDockedTop","paperAnchorDockedBottom","modal"]);const xl=(e,n)=>{const{ownerState:t}=e;return[n.root,("permanent"===t.variant||"persistent"===t.variant)&&n.docked,n.modal]},bl=pi(ll,{name:"MuiDrawer",slot:"Root",overridesResolver:xl})(fi((e=>{let{theme:n}=e;return{zIndex:(n.vars||n).zIndex.drawer}}))),vl=pi("div",{shouldForwardProp:ui,name:"MuiDrawer",slot:"Docked",skipVariantsResolver:!1,overridesResolver:xl})({flex:"0 0 auto"}),jl=pi(gl,{name:"MuiDrawer",slot:"Paper",overridesResolver:(e,n)=>{const{ownerState:t}=e;return[n.paper,n["paperAnchor".concat(_r(t.anchor))],"temporary"!==t.variant&&n["paperAnchorDocked".concat(_r(t.anchor))]]}})(fi((e=>{let{theme:n}=e;return{overflowY:"auto",display:"flex",flexDirection:"column",height:"100%",flex:"1 0 auto",zIndex:(n.vars||n).zIndex.drawer,WebkitOverflowScrolling:"touch",position:"fixed",top:0,outline:0,variants:[{props:{anchor:"left"},style:{left:0}},{props:{anchor:"top"},style:{top:0,left:0,right:0,height:"auto",maxHeight:"100%"}},{props:{anchor:"right"},style:{right:0}},{props:{anchor:"bottom"},style:{top:"auto",left:0,bottom:0,right:0,height:"auto",maxHeight:"100%"}},{props:e=>{let{ownerState:n}=e;return"left"===n.anchor&&"temporary"!==n.variant},style:{borderRight:"1px solid ".concat((n.vars||n).palette.divider)}},{props:e=>{let{ownerState:n}=e;return"top"===n.anchor&&"temporary"!==n.variant},style:{borderBottom:"1px solid ".concat((n.vars||n).palette.divider)}},{props:e=>{let{ownerState:n}=e;return"right"===n.anchor&&"temporary"!==n.variant},style:{borderLeft:"1px solid ".concat((n.vars||n).palette.divider)}},{props:e=>{let{ownerState:n}=e;return"bottom"===n.anchor&&"temporary"!==n.variant},style:{borderTop:"1px solid ".concat((n.vars||n).palette.divider)}}]}}))),Cl={left:"right",right:"left",top:"down",bottom:"up"};const wl=a.forwardRef((function(e,n){const t=bi({props:e,name:"MuiDrawer"}),o=ns(),i=(()=>{const e=a.useContext(Fs);return null!==e&&void 0!==e&&e})(),r={enter:o.transitions.duration.enteringScreen,exit:o.transitions.duration.leavingScreen},{anchor:s="left",BackdropProps:l,children:c,className:d,elevation:u=16,hideBackdrop:h=!1,ModalProps:{BackdropProps:p,...m}={},onClose:f,open:g=!1,PaperProps:y={},SlideProps:x,TransitionComponent:b=pl,transitionDuration:v=r,variant:j="temporary",...C}=t,w=a.useRef(!1);a.useEffect((()=>{w.current=!0}),[]);const S=function(e,n){let{direction:t}=e;return"rtl"===t&&function(e){return["left","right"].includes(e)}(n)?Cl[n]:n}({direction:i?"rtl":"ltr"},s),k=s,N={...t,anchor:k,elevation:u,open:g,variant:j,...C},_=(e=>{const{classes:n,anchor:t,variant:o}=e;return ei({root:["root"],docked:[("permanent"===o||"persistent"===o)&&"docked"],modal:["modal"],paper:["paper","paperAnchor".concat(_r(t)),"temporary"!==o&&"paperAnchorDocked".concat(_r(t))]},yl,n)})(N),T=(0,we.jsx)(jl,{elevation:"temporary"===j?u:0,square:!0,...y,className:Ie(_.paper,y.className),ownerState:N,children:c});if("permanent"===j)return(0,we.jsx)(vl,{className:Ie(_.root,_.docked,d),ownerState:N,ref:n,...C,children:T});const I=(0,we.jsx)(b,{in:g,direction:Cl[S],timeout:v,appear:w.current,...x,children:T});return"persistent"===j?(0,we.jsx)(vl,{className:Ie(_.root,_.docked,d),ownerState:N,ref:n,...C,children:I}):(0,we.jsx)(bl,{BackdropProps:{...l,...p,transitionDuration:v},className:Ie(_.root,_.modal,d),open:g,ownerState:N,onClose:f,hideBackdrop:h,ref:n,...C,...m,children:I})}));function Sl(e){return Ka("MuiSvgIcon",e)}$a("MuiSvgIcon",["root","colorPrimary","colorSecondary","colorAction","colorError","colorDisabled","fontSizeInherit","fontSizeSmall","fontSizeMedium","fontSizeLarge"]);const kl=pi("svg",{name:"MuiSvgIcon",slot:"Root",overridesResolver:(e,n)=>{const{ownerState:t}=e;return[n.root,"inherit"!==t.color&&n["color".concat(_r(t.color))],n["fontSize".concat(_r(t.fontSize))]]}})(fi((e=>{var n,t,o,a,i,r,s,l,c,d,u,h,p,m,f;let{theme:g}=e;return{userSelect:"none",width:"1em",height:"1em",display:"inline-block",flexShrink:0,transition:null===(n=g.transitions)||void 0===n||null===(t=n.create)||void 0===t?void 0:t.call(n,"fill",{duration:null===(o=(null!==(a=g.vars)&&void 0!==a?a:g).transitions)||void 0===o||null===(o=o.duration)||void 0===o?void 0:o.shorter}),variants:[{props:e=>!e.hasSvgAsChild,style:{fill:"currentColor"}},{props:{fontSize:"inherit"},style:{fontSize:"inherit"}},{props:{fontSize:"small"},style:{fontSize:(null===(i=g.typography)||void 0===i||null===(r=i.pxToRem)||void 0===r?void 0:r.call(i,20))||"1.25rem"}},{props:{fontSize:"medium"},style:{fontSize:(null===(s=g.typography)||void 0===s||null===(l=s.pxToRem)||void 0===l?void 0:l.call(s,24))||"1.5rem"}},{props:{fontSize:"large"},style:{fontSize:(null===(c=g.typography)||void 0===c||null===(d=c.pxToRem)||void 0===d?void 0:d.call(c,35))||"2.1875rem"}},...Object.entries((null!==(u=g.vars)&&void 0!==u?u:g).palette).filter((e=>{let[,n]=e;return n&&n.main})).map((e=>{var n,t;let[o]=e;return{props:{color:o},style:{color:null===(n=(null!==(t=g.vars)&&void 0!==t?t:g).palette)||void 0===n||null===(n=n[o])||void 0===n?void 0:n.main}}})),{props:{color:"action"},style:{color:null===(h=(null!==(p=g.vars)&&void 0!==p?p:g).palette)||void 0===h||null===(h=h.action)||void 0===h?void 0:h.active}},{props:{color:"disabled"},style:{color:null===(m=(null!==(f=g.vars)&&void 0!==f?f:g).palette)||void 0===m||null===(m=m.action)||void 0===m?void 0:m.disabled}},{props:{color:"inherit"},style:{color:void 0}}]}}))),Nl=a.forwardRef((function(e,n){const t=bi({props:e,name:"MuiSvgIcon"}),{children:o,className:i,color:r="inherit",component:s="svg",fontSize:l="medium",htmlColor:c,inheritViewBox:d=!1,titleAccess:u,viewBox:h="0 0 24 24",...p}=t,m=a.isValidElement(o)&&"svg"===o.type,f={...t,color:r,component:s,fontSize:l,instanceFontSize:e.fontSize,inheritViewBox:d,viewBox:h,hasSvgAsChild:m},g={};d||(g.viewBox=h);const y=(e=>{const{color:n,fontSize:t,classes:o}=e;return ei({root:["root","inherit"!==n&&"color".concat(_r(n)),"fontSize".concat(_r(t))]},Sl,o)})(f);return(0,we.jsxs)(kl,{as:s,className:Ie(y.root,i),focusable:"false",color:c,"aria-hidden":!u||void 0,role:u?"img":void 0,ref:n,...g,...p,...m&&o.props,ownerState:f,children:[m?o.props.children:o,u?(0,we.jsx)("title",{children:u}):null]})}));Nl.muiName="SvgIcon";const _l=Nl;function Tl(e,n){function t(t,o){return(0,we.jsx)(_l,{"data-testid":"".concat(n,"Icon"),ref:o,...t,children:e})}return t.muiName=_l.muiName,a.memo(a.forwardRef(t))}const Il=Tl((0,we.jsx)("path",{d:"M3 18h18v-2H3zm0-5h18v-2H3zm0-7v2h18V6z"}),"Menu"),El=Tl((0,we.jsx)("path",{d:"m12 8-6 6 1.41 1.41L12 10.83l4.59 4.58L18 14z"}),"ExpandLess"),Al=Tl((0,we.jsx)("path",{d:"M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"}),"ExpandMore"),Bl=Tl((0,we.jsx)("path",{d:"M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"}),"Close");function Rl(e,n,t,o,i){const[r,s]=a.useState((()=>i&&t?t(e).matches:o?o(e).matches:n));return Ai((()=>{if(!t)return;const n=t(e),o=()=>{s(n.matches)};return o(),n.addEventListener("change",o),()=>{n.removeEventListener("change",o)}}),[e,t]),r}const Fl={...i}.useSyncExternalStore;function Pl(e,n,t,o,i){const r=a.useCallback((()=>n),[n]),s=a.useMemo((()=>{if(i&&t)return()=>t(e).matches;if(null!==o){const{matches:n}=o(e);return()=>n}return r}),[r,e,o,i,t]),[l,c]=a.useMemo((()=>{if(null===t)return[r,()=>()=>{}];const n=t(e);return[()=>n.matches,e=>(n.addEventListener("change",e),()=>{n.removeEventListener("change",e)})]}),[r,t,e]);return Fl(c,l,s)}function Dl(){let e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:{};const{themeId:n}=e;return function(e){let t=arguments.length>1&&void 0!==arguments[1]?arguments[1]:{},o=Eo();o&&n&&(o=o[n]||o);const a="undefined"!==typeof window&&"undefined"!==typeof window.matchMedia,{defaultMatches:i=!1,matchMedia:r=(a?window.matchMedia:null),ssrMatchMedia:s=null,noSsr:l=!1}=function(e){const{theme:n,name:t,props:o}=e;return n&&n.components&&n.components[t]&&n.components[t].defaultProps?gi(n.components[t].defaultProps,o):o}({name:"MuiUseMediaQuery",props:t,theme:o});let c="function"===typeof e?e(o):e;c=c.replace(/^@media( ?)/m,"");return(void 0!==Fl?Pl:Rl)(c,i,r,s,l)}}Dl();const Ol=Dl({themeId:Qa}),Ml=240,Ll=180;function Gl(e){let{menuItems:n,subMenuItems:t}=e;const o=te(),i=ee(),[r,s]=(0,a.useState)(null),[l,c]=(0,a.useState)(null),[d,u]=(0,a.useState)(!1),h=Ol("(max-width: 816px)"),p=()=>{u(!d)},m=(0,we.jsxs)(Ja,{children:[(0,we.jsx)(Ci,{children:(0,we.jsx)("a",{href:"https://www.crg.eu/",target:"_blank",rel:"noreferrer",children:(0,we.jsx)("img",{className:"CRGLogo",src:"./crg_logo_white.svg",alt:"CRG Logo"})})}),(0,we.jsx)(Ja,{sx:{overflow:"auto"},children:(0,we.jsx)(Ni,{sx:{paddingTop:h?"10px":"30px"},children:n.map((e=>{var n;return(0,we.jsxs)(Ja,{children:[(0,we.jsxs)(Sr,{onClick:()=>(e=>{if("Introduction"===e)return o("/"),window.scrollTo({top:0,behavior:"smooth"}),c(e),void s(null);const n=t[e];if(n)if(r===e)s(null),c(null);else{let t=n[0],a="/".concat(t.toLowerCase().replace(/ /g,"-"));"Beacon 2 PI API"===e&&"Automated Deployment"===t&&(a="/pi-automated-deployment"),"Beacon 2 RI API"===e&&"Automated Deployment"===t&&(a="/automated-deployment"),o(a),window.scrollTo({top:0,behavior:"smooth"}),s(e),c(t)}else{const n="/".concat(e.toLowerCase().replace(/ /g,"-"));o(n),window.scrollTo({top:0,behavior:"smooth"}),c(e),s(null)}})(e),sx:{bgcolor:l===e||r===e?"#4A88B1":"inherit",color:l===e||r===e?"white":"inherit","&:hover":{bgcolor:"#4A88B1",color:"white"}},children:[(0,we.jsx)(Ur,{primaryTypographyProps:{variant:h?"body2":"body1",fontWeight:"bold",color:"inherit"},primary:e}),t[e]?r===e?(0,we.jsx)(El,{}):(0,we.jsx)(Al,{}):null]}),t[e]&&(0,we.jsx)(cs,{in:r===e,timeout:"auto",unmountOnExit:!0,children:null===(n=t[e])||void 0===n?void 0:n.map((n=>(0,we.jsx)(Sr,{onClick:()=>((e,n)=>{let t=e.toLowerCase().replace(/ /g,"-");"Configuration"===e&&(t="Beacon 2 RI API"===n?"api-configuration":"configuration"),"Automated Deployment"===e&&(t="Beacon 2 PI API"===n?"pi-automated-deployment":"automated-deployment"),"Manual Deployment"===e&&(t="Beacon 2 PI API"===n?"pi-manual-deployment":"manual-deployment"),"Querying the API"===e&&(t="Beacon 2 PI API"===n?"pi-querying-the-api":"querying-the-api"),c(e),o("/".concat(t))})(n,e),sx:{color:l===n||i.pathname==="/".concat(n.toLowerCase().replace(/ /g,"-"))?"#185177":"#4A88B1",bgcolor:l===n||i.pathname==="/".concat(n.toLowerCase().replace(/ /g,"-"))?"white":"#E5ECF3","&:hover":{bgcolor:"white",color:"#185177"}},children:(0,we.jsx)(Fr,{variant:h?"body2":"body1",fontWeight:"bold",sx:{color:"inherit",p:.2},children:n})},n)))})]},e)}))})})]});return(0,we.jsxs)(Ja,{sx:{display:"flex",position:"relative",top:"-40px"},children:[(0,we.jsxs)(Ci,{sx:{position:"relative"},children:[h&&(0,we.jsx)(Rs,{"aria-label":"close drawer",onClick:p,sx:{position:"fixed",top:"10px",left:"10px",color:"white",zIndex:1500,backgroundColor:"rgba(0, 0, 0, 0.2)"},children:d?(0,we.jsx)(Bl,{}):(0,we.jsx)(Il,{})}),(0,we.jsx)(Fr,{variant:"h6",noWrap:!0})]}),(0,we.jsx)(wl,{variant:h?"temporary":"permanent",open:!h||d,onClose:p,sx:{width:h?Ll:Ml,flexShrink:0,"& .MuiDrawer-paper":{width:h?Ll:Ml,bgcolor:"#185177",color:"white",border:"none",zIndex:1200}},children:m})]})}const Vl=function(){return(0,we.jsxs)("div",{className:"landingPage",children:[(0,we.jsxs)("div",{children:[" ",(0,we.jsx)("a",{href:"/",className:"no-undeline",children:(0,we.jsx)("h2",{children:"Documentation"})}),(0,we.jsx)("h3",{children:"Beacon v2 Reference/Production Implementation Documentation"}),(0,we.jsxs)("p",{children:[(0,we.jsx)("b",{children:"Beacon v2 Reference Implementation (B2RI) "}),"and"," ",(0,we.jsx)("b",{children:"Beacon v2 Production Implementation (B2PI) "})," are tools that provide a REST API and enable queries of genomic and phenotypic data in compliance with Beacon v2 standard.",(0,we.jsx)("br",{}),"Below is an overview of the schema process of B2RI/B2PI, which involves transforming raw data into a Beacon v2-compliant .json format, uploading it to the API and querying the API through a user-friendly interface.",(0,we.jsx)("br",{}),"The process relies on the following four processes:"]}),(0,we.jsxs)("li",{children:[(0,we.jsx)("b",{children:"Beacon RI Tools v2."})," This tool enables conversion of .csv/.vcf files to .json files in a Beacon Friendly Format (BFF). Note that Beacon RI Tools v2 are designed to work with both Beacon 2 RI API or Beacon 2 PI API. Specifically, it includes all the features that are needed for generating Beacon v2 PI API compatible data."]}),(0,we.jsxs)("li",{children:[(0,we.jsx)("b",{children:"Beacon RI API / PI API."})," The API connects to a MongoDB database that stores the .json files, allowing one to query the data following the Beacon v2 standards. Beacon v2 PI is an upgraded version of Beacon v2 RI API, developed for production environments and recommended for improved performance and user experience."]}),(0,we.jsxs)("li",{children:[(0,we.jsx)("b",{children:"Beacon Verifier v2."})," A tool that checks that the data in the API is compliant with Beacon v2 and provides a verification report."]}),(0,we.jsxs)("li",{children:[(0,we.jsx)("b",{children:"Beacon User Interface."})," The user-friendly interface for querying the API without needing to worry about the underlying standards. It supports intuitive queries and enables authentication through LifeScience or Keycloak."]}),(0,we.jsx)("p",{children:"Regarding security, the API always validates authentication through the chosen identity provider, following the globally accepted OIDC and Oauth protocols."})]}),(0,we.jsx)("img",{className:"relationship-elements",alt:"Relationship-elements",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/scheme_RI.svg?raw=true"}),(0,we.jsx)("h2",{className:"h2-subtitle",children:"What brings the new Beacon v2 Production Implementation?"}),(0,we.jsx)("p",{children:"As stated before, Beacon v2 PI API is a beacon instance for production environments. The steps that this production instance follows are the next ones shown in the next flowchart:"}),(0,we.jsx)("img",{className:"relationship-elements",alt:"Relationship-elements",src:"./scheme_PI.svg"}),(0,we.jsxs)("p",{children:["The process to deploy and inject data is the same than in Beacon v2 RI API (data comes from Beacon RI Tools v2) but this instance is optimized and tested, having the next upgrades from Beacon v2 RI API:",(0,we.jsx)("li",{children:"Handlers of the endpoints are classes, not functions."}),(0,we.jsx)("li",{children:"Unit testing has been developed for the application, starting with 108 unit tests that cover 4000 lines of code approximately (100%)."}),(0,we.jsxs)("li",{children:["Concurrency testing has been applied for this new beacon instance, showing results of responses for more than 3 million genomic variants splitted in different datasets in less than 100 millisecs, for a total of 1000 requests made by 10 users per second at the same time."," "]}),(0,we.jsx)("li",{children:"Linking ids to a dataset in a yaml file is not needed anymore. "}),(0,we.jsxs)("li",{children:["A couple more indexes for mongoDB have been applied, that, in addition to the restructuration of the code, have improved the quickness of the responses."," "]}),(0,we.jsxs)("li",{children:["Authentication/Authorization is now applied as a decorator, not as a different container."," "]}),(0,we.jsx)("li",{children:"LOGS now show more relevant information about the different processes (from request to response) including transaction id, the time of execution of each function and the initial call and the return call."}),(0,we.jsxs)("li",{children:["Exceptions now are raised from the lower layer to the top layer, with information and status for the origin of the exception."," "]}),(0,we.jsxs)("li",{children:["Architecture of the code is not dependent on a particular database, meaning that different types of databases (and more than one) can be potentially applied to this instance (although now only MongoDB is the one developed)."," "]}),(0,we.jsx)("li",{children:"Parameters are sanitized. "})," ",(0,we.jsxs)("li",{children:["Users can manage what entry types want their beacon to show by editing a manage conf file inside source (Endpoint Mapper)."," "]})]}),(0,we.jsxs)("div",{className:"collaborators-div",children:[(0,we.jsx)("h3",{children:"Collaborators"}),(0,we.jsxs)("div",{className:"collaborators-images",children:[(0,we.jsx)("img",{className:"collaborators-img",alt:"Europe",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/europe_logo.png?raw=true"}),(0,we.jsx)("img",{className:"collaborators-img",alt:"CaixaBank",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/caixabank_logo.png?raw=true"}),(0,we.jsx)("img",{className:"collaborators-img",alt:"EGA",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/ega_logo.png?raw=true"}),(0,we.jsx)("img",{className:"collaborators-img",alt:"CRG",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/crg_logo.png?raw=true"}),(0,we.jsx)("img",{className:"collaborators-img",alt:"ELIXIR",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/elixir_logo.png?raw=true"}),(0,we.jsx)("img",{className:"collaborators-img",alt:"EMBL-EBI",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/embl-ebi_logo.png?raw=true"})]})]}),(0,we.jsx)("footer",{className:"footer",children:(0,we.jsx)("span",{className:"footer-text",children:"\xa9 Copyright 2025, B2RI/PI Documentation Contributors"})}),(0,we.jsx)("br",{}),(0,we.jsx)("br",{})]})};const ql=t.p+"static/media/copy-symbol.e5f79483cb3d662c42bb7c6b7aacef44.svg",Hl=()=>{const e=ee(),[n,t]=(0,a.useState)([]),[o,i]=(0,a.useState)(null);(0,a.useEffect)((()=>{const e=Array.from(document.querySelectorAll("h1[id],h2[id], h3[id], h4[id]")).map((e=>({id:e.id,text:e.textContent||""})));t(e)}),[e.pathname]);return 0===n.length?null:(0,we.jsxs)("div",{className:"onThisPageContainer sidebarColumn",children:[(0,we.jsx)("h3",{className:"onThisPageTitle",children:"On this page"}),(0,we.jsx)("ul",{children:n.map((e=>(0,we.jsx)("li",{children:(0,we.jsx)("a",{href:"#".concat(e.id),className:"onThisPageLink ".concat(o===e.id?"active":""),onClick:()=>{return n=e.id,void i(n);var n},children:e.text})},e.id)))})]})},zl=()=>{const[e,n]=(0,a.useState)({}),t=e=>{const t={"cloning-repository":"git clone https://github.com/EGA-archive/beacon2-ri-api.git","create-network":"docker network create my-app-network","cd-deploy":"cd deploy && docker-compose up -d --build","copy-files":["docker cp /path/to/analyses.json rimongo:tmp/analyses.json","docker cp /path/to/biosamples.json rimongo:tmp/biosamples.json","docker cp /path/to/cohorts.json rimongo:tmp/cohorts.json","docker cp /path/to/datasets.json rimongo:tmp/datasets.json","docker cp /path/to/genomicVariations.json rimongo:tmp/genomicVariations.json","docker cp /path/to/individuals.json rimongo:tmp/individuals.json","docker cp /path/to/runs.json rimongo:tmp/runs.json"].join("\n"),"data-injection":['docker exec rimongo mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /tmp/datasets.json --collection datasets','docker exec rimongo mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /tmp/analyses.json --collection analyses','docker exec rimongo mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /tmp/biosamples.json --collection biosamples','docker exec rimongo mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /tmp/cohorts.json --collection cohorts','docker exec rimongo mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /tmp/genomicVariations.json --collection genomicVariations','docker exec rimongo mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /tmp/individuals.json --collection individuals','docker exec rimongo mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /tmp/runs.json --collection runs'].join("\n"),"data-indexing":"docker exec beacon python beacon/reindex.py"}[e];t&&navigator.clipboard.writeText(t).then((()=>{n((n=>({...n,[e]:!0}))),setTimeout((()=>n((n=>({...n,[e]:!1})))),1500)})).catch((e=>console.log(e)))};return(0,we.jsxs)("div",{className:"deploymentContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/manual-deployment",className:"no-undeline",children:"Beacon 2 RI API"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/manual-deployment",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:" Manual Deployment"})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h3",{children:"Beacon 2 RI API"}),(0,we.jsx)("h1",{children:"Manual Deployment"}),(0,we.jsx)("h2",{id:"cloning-repository",children:"Cloning the repository"}),(0,we.jsx)("p",{children:"First of all, start by cloning the GitHub repository in your system."}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["git clone"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-api.git",children:"https://github.com/EGA-archive/beacon2-ri-api.git"})]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("cloning-repository"),children:e["cloning-repository"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h2",{id:"create-network",children:"Creating the containers"}),(0,we.jsx)("p",{children:"Create the external docker network needed before building the containers."}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker network create my-app-network"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("create-network"),children:e["create-network"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"Make sure the next list of ports are free of use in your system:"}),(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"27017 \u2192 mongo"}),(0,we.jsx)("li",{children:"8081 \u2192 mongo-express"}),(0,we.jsx)("li",{children:"5050 \u2192 beacon"}),(0,we.jsx)("li",{children:"5051 \u2192 permissions"}),(0,we.jsx)("li",{children:"8010 \u2192 permissions UI"}),(0,we.jsx)("li",{children:"8080 \u2192 Keycloak"}),(0,we.jsx)("li",{children:"9991 \u2192 Keycloak SSL"}),(0,we.jsx)("li",{children:"3000 \u2192 Beacon UI"})]}),(0,we.jsx)("p",{children:"Light up the containers from the deploy folder."}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"cd deploy && docker-compose up -d --build"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("cd-deploy"),children:e["cd-deploy"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"If the containers are built correctly:"}),(0,we.jsxs)("ul",{children:[(0,we.jsxs)("li",{children:["The API will run in"," ",(0,we.jsx)("a",{href:"http://localhost:5050/api",target:"_blank",rel:"noopener noreferrer",children:"http://localhost:5050/api"}),"."]}),(0,we.jsxs)("li",{children:["The Beacon UI will run in"," ",(0,we.jsx)("a",{href:"http://localhost:3000",target:"_blank",rel:"noopener noreferrer",children:"http://localhost:3000"}),"."]}),(0,we.jsxs)("li",{children:["The mongo-express UI will run in"," ",(0,we.jsx)("a",{href:"http://localhost:8081",target:"_blank",rel:"noopener noreferrer",children:"http://localhost:8081"}),"."]}),(0,we.jsxs)("li",{children:["The Keycloak UI will run in"," ",(0,we.jsx)("a",{href:"http://localhost:8080/auth",target:"_blank",rel:"noopener noreferrer",children:"http://localhost:8080/auth"}),"."]})]}),(0,we.jsx)("h2",{id:"data-injection",children:"Data injection"}),(0,we.jsx)("p",{children:"Copy all the available BFF files to the mongo container using the following commands:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["docker cp /path/to/analyses.json rimongo:tmp/analyses.json\n","docker cp /path/to/biosamples.json rimongo:tmp/biosamples.json\n","docker cp /path/to/cohorts.json rimongo:tmp/cohorts.json\n","docker cp /path/to/datasets.json rimongo:tmp/datasets.json\n","docker cp /path/to/genomicVariations.json rimongo:tmp/genomicVariations.json\n","docker cp /path/to/individuals.json rimongo:tmp/individuals.json\n","docker cp /path/to/runs.json rimongo:tmp/runs.json"]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("copy-files"),children:e["copy-files"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:"Not all the above commands are needed to run a beacon. A beacon can run without data or data for a single collection (e.g. only individuals)."})]}),(0,we.jsx)("p",{children:"Now execute the insertion of the files copied into the mongo database."}),(0,we.jsx)("div",{className:"codeSnippet codeSnippet-mongodb",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:['docker exec rimongo mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /tmp/datasets.json --collection datasets\n','docker exec rimongo mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /tmp/analyses.json --collection analyses\n','docker exec rimongo mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /tmp/biosamples.json --collection biosamples\n','docker exec rimongo mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /tmp/cohorts.json --collection cohorts\n','docker exec rimongo mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /tmp/genomicVariations.json --collection genomicVariations\n','docker exec rimongo mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /tmp/individuals.json --collection individuals\n','docker exec rimongo mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /tmp/runs.json --collection runs']}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("data-injection"),children:e["data-injection"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol copySymbol-custom",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h2",{id:"data-indexing",children:"Data indexing"}),(0,we.jsx)("p",{children:"For the API to respond fast to the queries, you have to index your database. You can create the necessary indexes by running the next script:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker exec beacon python beacon/reindex.py"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("data-indexing"),children:e["data-indexing"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:"You will need to run this script each time you inject new data."})]})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]})},Ul=(e,n,t)=>{const{addSearchableContent:o,searchableContent:i}=Ne();(0,a.useEffect)((()=>{i.some((t=>t.route===e&&t.text===n))||o(e,n,t)}),[o,e,n,t,i])},Ql=()=>(Ul("/","Documentation\nBeacon v2 Reference/Production Implementation Documentation\n\n\nBeacon v2 Reference Implementation (B2RI) and Beacon v2 Production Implementation (B2PI) are tools that provide a REST API and enable queries of genomic and phenotypic data in compliance with Beacon v2 standard.\n\nBelow is an overview of the schema process of B2RI/B2PI, which involves transforming raw data into a Beacon v2-compliant .json format, uploading it to the API and querying the API through a user-friendly interface.\n\nThe process relies on the following four processes:\n\nBeacon RI Tools v2. This tool enables conversion of .csv/.vcf files to .json files in a Beacon Friendly Format (BFF). Note that Beacon RI Tools v2 are designed to work with both Beacon 2 RI API or Beacon 2 PI API. Specifically, it includes all the features that are needed for generating Beacon v2 PI API compatible data.\nBeacon RI API / PI API. The API connects to a MongoDB database that stores the .json files, allowing one to query the data following the Beacon v2 standards. Beacon v2 PI is an upgraded version of Beacon v2 RI API, developed for production environments and recommended for improved performance and user experience.\nBeacon Verifier v2. A tool that checks that the data in the API is compliant with Beacon v2 and provides a verification report.\nBeacon User Interface. The user-friendly interface for querying the API without needing to worry about the underlying standards. It supports intuitive queries and enables authentication through LifeScience or Keycloak.\nRegarding security, the API always validates authentication through the chosen identity provider, following the globally accepted OIDC and Oauth protocols.\n\nAs stated before, Beacon v2 PI API is a beacon instance for production environments. The steps that this production instance follows are the next ones shown in the next flowchart:\n\nThe process to deploy and inject data is the same than in Beacon v2 RI API (data comes from Beacon RI Tools v2) but this instance is optimized and tested, having the next upgrades from Beacon v2 RI API:\n\nHandlers of the endpoints are classes, not functions.\nUnit testing has been developed for the application, starting with 108 unit tests that cover 4000 lines of code approximately (100%).\nConcurrency testing has been applied for this new beacon instance, showing results of responses for more than 3 million genomic variants splitted in different datasets in less than 100 millisecs, for a total of 1000 requests made by 10 users per second at the same time.\nLinking ids to a dataset in a yaml file is not needed anymore.\nA couple more indexes for mongoDB have been applied, that, in addition to the restructuration of the code, have improved the quickness of the responses.\nAuthentication/Authorization is now applied as a decorator, not as a different container.\nLOGS now show more relevant information about the different processes (from request to response) including transaction id, the time of execution of each function and the initial call and the return call.\nExceptions now are raised from the lower layer to the top layer, with information and status for the origin of the exception.\nArchitecture of the code is not dependent on a particular database, meaning that different types of databases (and more than one) can be potentially applied to this instance (although now only MongoDB is the one developed).\nParameters are sanitized.\nUsers can manage what entry types want their beacon to show by editing a manage conf file inside source (Endpoint Mapper).\n\nCollaborators\nEurope\nCaixaBank\nEGA\nCRG\nELIXIR\nEMBL-EBI\n\xa9 Copyright 2025, B2RI Documentation Contributors","Introduction"),Ul("/pi-automated-deployment","Documentation Beacon 2 PI API Automated Deployment\nBeacon 2 Production Implementation API\nPrerequisites\nYou should have installed:\nDocker\nDocker Compose\nData from Beacon RI Tools v2. Please, bear in mind that the datasetId for your records must match the id for the dataset in the /datasets entry type.\n\nAutomated Deployment\nThis deployment relies on a Bash script and a Makefile and it is only compatible with MongoDB database deployments.\n\nCloning the repository\nFirst of all, start by cloning the GitHub repository in your system.\n\ngit clone https://github.com/EGA-archive/beacon2-pi-api.git\n\nMake sure the next list of ports are free of use in your system:\n\n27017 \u2192 mongo\n5050 \u2192 beacon\nExecute start script from root\nTo quickly deploy your Beacon instance and load initial data, run the following command from the root of your project:\n\nbash mongostart.sh\n\nThis script will:\nDeploy the Beacon API\nInsert default Beacon-formatted data into your MongoDB instance, using the files located at: /beacon2-pi-api/beacon/connections/mongo/data\n\n\nWant to insert your own data?\nSimply replace the contents of the data folder with your custom Beacon Friendly Format (BFF) files before running the script.\n\nIf the operation is successful, you will have a beacon up and running at http://localhost:5050/api.\n\nNext Steps: Customize Your Beacon\nMake your Beacon your own by following these next steps:\nEdit your instance\u2019s metadata. Update the /info endpoint with your organization's name, description, version, and contact details.\nManage dataset permissions. Control which datasets are public or require authentication.\nEnable advanced filtering\nFound more setting options in Configuration","Automated Deployment"),Ul("/pi-manual-deployment",'Beacon 2 PI API - Manual Deployment\n\nCloning the Repository\nFirst of all, start by cloning the GitHub repository in your system.\n\n```bash\ngit clone https://github.com/EGA-archive/beacon2-pi-api.git\n```\n\nCreating the Containers\nMake sure the next list of ports are free of use in your system:\n- 27017 \u2192 mongo\n- 8081 \u2192 mongo-express (optional)\n- 5050 \u2192 beacon\n- 8080 \u2192 Keycloak (optional)\n- 9991 \u2192 Keycloak SSL (optional)\n\nLight up the containers from the deploy folder.\n\n```bash\ndocker compose up -d --build\n```\n\nIf you wish to light up only some of the services, not all, just write the name of the services in the `docker-compose.yml` file right after the command (e.g., for having the mandatory services only):\n\n```bash\ndocker compose up -d \u2013build beaconprod db\n```\n\nIf the containers are built correctly:\n- The API will run in [http://localhost:5050/api](http://localhost:5050/api).\n- The mongo-express UI will run in [http://localhost:8081](http://localhost:8081).\n- The Keycloak UI will run in [http://localhost:8080/auth](http://localhost:8080/auth).\n\nData Injection\nCopy all the available BFF files you have to the [data folder for mongo database](https://github.com/EGA-archive/beacon2-pi-api/tree/main/beacon/connections/mongo/data).\n\n**Note:** A beacon can run without data or data for a single collection (e.g., only individuals).\n\nNow execute the insertion of the files copied into the mongo database.\n\n```bash\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/datasets.json --collection datasets\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/individuals.json --collection individuals\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/cohorts.json --collection cohorts\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/analyses.json --collection analyses\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/biosamples.json --collection biosamples\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/genomicVariations.json --collection genomicVariations\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/runs.json --collection runs\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/targets.json --collection targets\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/caseLevelData.json --collection caseLevelData\n```\n\nData Indexing\nFor the API to respond fast to the queries, you have to index your database. You can create the necessary indexes by running the next script:\n\n```bash\ndocker exec beaconprod python -m beacon.connections.mongo.reindex\n```\n\n**Note:** You will need to run this script each time you inject new data.',"Manual Deployment"),Ul("/filtering-terms",'Documentationat Beacon 2 PI API Filtering Terms\nBeacon 2 Production Implementation API\nFiltering Terms\nFiltering terms are metadata fields that allow users to query your Beacon more precisely \u2014 for example, by filtering results based on sequencing strategy (libraryStrategy), tissue type, disease, or other structured attributes. These terms often rely on ontologies to ensure consistent vocabulary across datasets.\n\nThis section explains how to extract, add, and enhance filtering terms for your Beacon instance.\n\nAutomatically Extract Filtering Terms\nIf your data collections (e.g., runs, biosamples, etc.) already contain structured metadata using ontology terms (like NCIT, UBERON, EFO...), you can extract filtering terms automatically.\nThis will populate the /filteringTerms endpoint of your Beacon, enabling more advanced queries.\n\ndocker exec beaconprod python beacon/connections/mongo/extract_filtering_terms.py\n\nManually Add Filtering Terms (optional)\nIf you want to enable filtering for fields that aren\u2019t auto-detected or don\u2019t use ontologies (e.g., simple alphanumeric fields like libraryStrategy), you can add them manually.\n\nExecute the following command:\n\ndb.filtering_terms.insertMany([\n  {\n     "type": "alphanumeric",\n     "id": "libraryStrategy",\n     "scope": ["runs"]\n  }\n])\n\nField descriptions:\ntype: alphanumeric indicates this is a plain text value, not an ontology term.\nid: The metadata field name.\nscope: The collection where this field is used (e.g., runs, biosamples, analyses).\n\nEnhance Ontology Filtering with Descendants and semantic similarities (optional)\nTo make ontology-based filtering more powerful and user-friendly, you can enable two features:\nDescendant terms: Automatically include all subcategories of a given ontology term (e.g., querying for \u201ccancer\u201d also returns \u201clung cancer\u201d, \u201cbreast cancer\u201d, etc.).\nid: The metadata field name.\nSemantic similarity: Enable matching to related ontology terms based on meaning, not just hierarchy. This allows for more flexible queries when users don\u2019t know the exact term used in the data.\n\nExample\nIf your data uses the term \u201cglioblastoma\u201d, but a user searches for \u201cbrain tumor\u201d, semantic similarity can help bridge the gap \u2014 even if \u201cbrain tumor\u201d isn\u2019t a direct parent of \u201cglioblastoma\u201d.\n\nHow to Enable\nAdd your .obo ontology files\nPlace ontology files into the ontologies/ folder in your Beacon instance. The filename must match the ontology prefix in lowercase (e.g., ncit.obo, uberon.obo, etc.).\n\nRun the script to enhance filtering terms\ndocker exec beaconprod python beacon/connections/mongo/get_descendants.py\n\nThis script:\nAdds descendant terms from the ontologies to each filtering term.\nComputes semantic similarities based on the ontology structure, enriching your Beacon\u2019s ability to match related terms.',"Filtering Terms"),Ul("/configuration","Documentation > 2 PI API > Configuration\n\nBeacon 2 Production Implementation API\n\nConfiguration\n\nBeacon2 PI API has two folders where a user can configure their beacons: conf for general configuration of the beacon and source for configuring the model of the beacon: e.g. entry types and the databases related to each entry type.\n\nEditing your beacon information\n\nTo show correctly your beacon\u2019s information you will need to edit both conf.py files from beacon and deploy folders.\nTo do so, edit the following variables:\n\nbeacon_id = 'org.ega-archive.beacon-ri-demo'\nbeacon_name = 'Beacon Reference Implementation demo'\napi_version = 'v2.0.0'\nuri = 'https://beacon-apis-demo.ega-archive.org/api/'\norg_id = 'EGA'\norg_name = 'European Genome-Phenome Archive (EGA)'\norg_description = 'The European Genome-phenome Archive (EGA) is a service for permanent archiving and sharing of all types of personally identifiable genetic and phenotypic data resulting from biomedical research projects.'\norg_adress = 'C/ Dr. Aiguader, 88 PRBB Building 08003 Barcelona, Spain'\norg_welcome_url = 'https://ega-archive.org/'\norg_contact_url = 'mailto:beacon.ega@crg.eu'\norg_logo_url = 'https://legacy.ega-archive.org/images/logo.png'\norg_info = ''\ndescription = \"This Beacon is based on synthetic data hosted at the EGA. The dataset contains 2504 samples including genetic data based on 1K Genomes data, and 76 individual attributes and phenotypic data derived from UKBiobank.\"\nversion = 'v2.0'\nwelcome_url = 'https://beacon.ega-archive.org/'\nalternative_url = 'https://beacon-apis-demo.ega-archive.org/api/'\ncreate_datetime = '2021-11-29T12:00:00.000000'\nupdate_datetime = ''\n\nNote that this is the info that will be shown in the /info endpoint for your beacon.\n\nManaging dataset permissions\n\nThere are 3 possible levels of beacon security for a dataset: public, registered and controlled.\n\nA public dataset is a dataset that will be returned in a beacon query without an authentication token.\nA registered dataset is a dataset that will be shown after a user sends a valid token (in other words, is logged in).\nA controlled dataset is a dataset that needs a user to send a valid token for authentication and the user needs to be allowed to query that dataset.\n\nEdit the .yml files inside permissions/datasets\n\nFor a public dataset, edit the public_datasets.yml file inside permissions folder and add the dataset with the exact name it appears in the id from its pertinent record in datasets mongo collection:\n\npublic_datasets:\n\ndataset_id\nFor a registered dataset, edit the registered_datasets.yml file inside permissions folder and add the dataset with the exact name it appears in the id from its pertinent record in datasets mongo collection:\n\nregistered_datasets:\n\ndataset_id\nFor a controlled dataset, edit the controlled_datasets.yml file inside permissions folder and add the dataset inside the username you wish to give permissions with the exact name for the dataset as it appears in the id from its pertinent record in datasets mongo collection and the exact name the user has in Keycloak:\n\nusername:\n\ndataset_id\nSupplying AAI credentials for your IDP\n\nFor making any IDP (based on OpenID and Oauthv2) work with beacon, you will need to add the client ID and client secret for each IDP in a new file you have to create inside auth/idp_providers folder (for each IDP). This file must have a name with an .env extension (e.g. something.env) and needs to have the following variables:\n\nCLIENT_ID=beacon\nCLIENT_SECRET='b26ca0f9-1137-4bee-b453-ee51eefbe7ba'\nUSER_INFO='http://idp:8080/auth/realms/Beacon/protocol/openid-connect/userinfo'\nINTROSPECTION='http://idp:8080/auth/realms/Beacon/protocol/openid-connect/token/introspect'\nISSUER='http://localhost:8080/auth/realms/Beacon'\nJWKS_URL='http://idp:8080/auth/realms/Beacon/protocol/openid-connect/certs'\n\nAfter that, make sure you build your beacon container again:\n\ndocker-compose up -d --build beaconprod\n\nHandling CORS\n\nTo avoid CORS using beacon and the frontend or a third-party authorization site like Keycloak, you will have to include all these URLs inside variable cors_urls in beacon/conf/conf.py file.\n\ncors_urls = [\"http://localhost:3000\",\"https://cancer-beacon-demo.ega-archive.org\", \"https://beacon-network-demo2.ega-archive.org\", \"https://beacon.ega-archive.org\"]\n\nSetting the default granularity\n\nBeacon PI can only limit the granularity for all the beacon at the moment. If you want to limit the granularity for all users across all datasets, you can point it in beacon/conf/conf.py, by editing the next variables:\n\nmax_beacon_granularity = \"record\"\n\nAdding beacon handovers\n\nA handover can be added to the general beacon response or to the response per dataset.\n\nIn any of the cases, you have to modify the beacon/utils/handovers.py file and you can write your handover there following the Beacon v2 spec response for handovers, e.g.\n\nhandover_1= {\n\"note\": \"Description of the handover\",\n\"url\": \"Link for the handover\",\n\"handoverType\": {\n\"id\": \"NCIT:C189151\",\n\"label\": \"Study Data Repository\"\n}\n}\n\nConfiguring your entry types\n\nIf you go to the source folder inside beacon folder, you will find a file called manage.py that you will need to edit in order to tell the API what is implemented in your beacon for each entry type and in what database you have the data related to each entry_type.\n\nIn order to do this, you will just need to put a True or False response to what granularity you have implemented per each entryType. See the next example:\n\ng_variants={ 'granularity': {\n'boolean': True,\n'count': True,\n'record': True },\n'singleEntryUrl': True,\n'endpoints': {'analysis': True, 'biosample': True, 'individual': True, 'run': True },\n'testMode': True,\n'database': 'mongo' }\n\nSetting the logs\n\nIn this production instance of beacon, you can set the level of the logs you want to output and where do you want to see the output in.\n\nFor setting the level of the logs, specify one amongst NOTSET, DEBUG, ERROR, INFO in the variable level (after logging.):\n\nlevel=logging.NOTSET\n\nTLS configuration\n\nTo enable TLS for the Beacon API set beacon_server_crt and beacon_server_key to the full path of the server certificate and server key in beacon/conf/conf.py file.\n\nTLS secured MongoDB\n\nEdit the file beacon/connections/mongo/conf.py and set database_certificate to the full path to the client certificate. If a private CA is used also set the database_cafile to the full path to the CA certificate.\n\nThe MongoDB client certificate should be in the combined PEM format client.key + \"\\n\" + client.crt\n","Configuration"),Ul("/pi-querying-the-api",'Beacon RI accepts two types of request methods: GET and POST.\n\nGET Method\n\nGET method is a bit more limited as this method only accepts request parameters, three generic parameters and ontology filters.\n\nQuery Parameters\nThese are the defined set of parameters (key-value pair) attached to the end of a URL used to provide additional information to a web server when making requests.\n\nGeneric parameters\n\nrequestedSchema\tstring\nExample: requestedSchema=ga4gh-service-info-v1.0\n\nskip\tinteger (Skip) >= 0\nDefault: 0\nExample: skip=0\n\nlimit\tinteger (Limit) >= 0\nDefault: 10\nMax: 100\nExample: limit=10\n\nRequest parameters\n\n\tstart\tinteger <int64> >= 0 (or comma separated integers)\n\tExample: start=16050074\n\nend\tinteger <int64> >= 0 (or comma separated integers)\nExample: end=16052080\n\n\tassemblyId\tstring\n\tExample: assemblyId=GRCh38 \n\nreferenceName\t<int64> >= 0\nExample: referenceName=22\n\nreferenceBases\tstring\nExample: referenceBases=G\n\nalternateBases\tstring\nExample: alternateBases=G\n\nvariantMinLength\tinteger <int64> >= 0\nExample: variantMinLength=5\n\nvariantMaxLength\tinteger <int64> >= 0\nExample: variantMaxLength=2\n\ngeneId\t\tstring\nExample: geneId=CCNL2\n\naminoacidChange\tstring\nExample: aminoacidChange=p.Thr130Met\n\nvariantType\tstring\nExample: variantType=SNP\n\ngenomicAlleleShortForm\tstring\nExample: genomicAlleleShortForm=NC_000022.11:g.16050075A>G\n\nclinicalRelevance\tstring\nExample: clinicalRelevance=pathogenic\n\nFilters\n\t\n\tfilters\tstring\n\tExample: filters=NCIT:C16576,NCIT:C42331\n\nGET query examples\n\nBeacon Sequence Query\n\nhttp://localhost:5050/api/g_variants?referenceName=22&start=7577120&referenceBases=G&alternateBases=A\n\nBeacon GeneId Query\n\nhttp://localhost:5050/api/g_variants?geneId=CHR_START-DUXAP8&variantMaxLength=2&variantType=INDEL\n\nBeacon Bracket Query\n\nhttp://localhost:5050/api/g_variants?referenceName=22&start=16050074,16050076&end=16050090&variantType=SNP\n\n\nPOST Method\n\nIn POST method requests, the endpoints accept all generic parameters and all the filters, apart from the request parameters.\n\nQuery Parameters\n\nGeneric parameters\n\nincludeResultsetResponses\tstring (IncludeResultsetResponses)\nDefault: "HIT"\nEnum: "ALL" "HIT" "MISS" "NONE"\n\npagination\tobject (Pagination)\nItems: \nskip\tinteger (Skip) >= 0\nDefault: 0\nExample: skip=0\n\nlimit\tinteger (Limit) >= 0\nDefault: 10\nMax: 100\nExample: limit=10\n\nrequestedGranularity\tstring (Granularity)\nDefault: "boolean"\nEnum: "boolean" "count" "record"\n\ntestMode\t\nboolean (TestMode)\nDefault: false\n\nRequest parameters\n\n\tstart\tinteger <int64> >= 0 (or comma separated integers)\n\tExample: start=16050074\n\nend\tinteger <int64> >= 0 (or comma separated integers)\nExample: end=16052080\n\n\tassemblyId\tstring\n\tExample: assemblyId=GRCh38 \n\nreferenceName\t<int64> >= 0\nExample: referenceName=22\n\nreferenceBases\tstring\nExample: referenceBases=G\n\nalternateBases\tstring\nExample: alternateBases=G\n\nvariantMinLength\tinteger <int64> >= 0\nExample: variantMinLength=5\n\nvariantMaxLength\tinteger <int64> >= 0\nExample: variantMaxLength=2\n\ngeneId\t\tstring\nExample: geneId=CCNL2\n\naminoacidChange\tstring\nExample: aminoacidChange=p.Thr130Met\n\nvariantType\tstring\nExample: variantType=SNP\n\ngenomicAlleleShortForm\tstring\nExample: genomicAlleleShortForm=NC_000022.11:g.16050075A>G\n\nclinicalRelevance\tstring\nExample: clinicalRelevance=pathogenic\n\nFilters\n\t\n\tfilters\tArray of OntologyFilter (object) or AlphanumericFilter (object) or CustomFilter (object) (Filtering Term Element)\n\tExample:\n       "filters": [    {\n        "id": "geographicOrigin",\n        "operator": "!",\n        "value": "England"\n    }]\n\nPOST query examples\n\nGenomic query\n\ncurl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n\t"meta": {\n    \t"apiVersion": "2.0"\n\t},\n\t"query": {\n    \t"requestParameters": {\n\t"alternateBases": "G" ,\n\t"referenceBases": "A" ,\n"start": [ 16050074 ],\n        \t"end": [ 16050568 ],\n     \t\t"variantType": "SNP"\n    \t},\n    \t"filters": [],\n    \t"includeResultsetResponses": "HIT",\n    \t"pagination": {\n        \t"skip": 0,\n        \t"limit": 10\n    \t},\n    \t"testMode": false,\n    \t"requestedGranularity": "record"\n\t}\n}\' \\\n  http://localhost:5050/api/g_variants\n\n\nSimple CURIE based filters query\n\ncurl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n\t"meta": {\n    \t"apiVersion": "2.0"\n\t},\n\t"query": {\n    \t"requestParameters": {\n    \t},\n    \t"filters": [\n{"id": "NCIT:C20197", "scope": "individual"}, {"id": "NCIT:C42331", "scope": "individual"}],\n    \t"includeResultsetResponses": "HIT",\n    \t"pagination": {\n        \t"skip": 0,\n        \t"limit": 10\n    \t},\n    \t"testMode": false,\n    \t"requestedGranularity": "record"\n\t}\n}\' \\\n  http://localhost:5050/api/individuals\n\nNumerical value query\n\ncurl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n\t"meta": {\n    \t"apiVersion": "2.0"\n\t},\n\t"query": {\n    \t"filters": [{\n    \t"id": "Weight",\n    \t"operator": ">",\n    \t"value": "75"\n\t}],\n    \t"includeResultsetResponses": "HIT",\n    \t"pagination": {\n        \t"skip": 0,\n        \t"limit": 10\n    \t},\n    \t"testMode": false,\n    \t"requestedGranularity": "record"\n\t}\n}\' \\\n  http://localhost:5050/api/individuals\n\n\n\nAlphanumerical value query (Exact value)\n\ncurl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n\t"meta": {\n    \t"apiVersion": "2.0"\n\t},\n\t"query": {\n    \t"filters": [\t{\n    \t"id": "geographicOrigin",  \n    \t"operator": "=",\n    \t"value": "England"    \n\t}],\n    \t"includeResultsetResponses": "HIT",\n    \t"pagination": {\n        \t"skip": 0,\n        \t"limit": 10\n    \t},\n    \t"testMode": false,\n    \t"requestedGranularity": "record"\n\t}\n}\' \\\n  http://localhost:5050/api/individuals\n\nAlphanumerical value query (Like value)\n\ncurl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n\t"meta": {\n    \t"apiVersion": "2.0"\n\t},\n\t"query": {\n    \t"filters": [\t{\n    \t"id": "geographicOrigin",\n    \t"operator": "=",\n    \t"value": "%land%"\n\t}],\n    \t"includeResultsetResponses": "HIT",\n    \t"pagination": {\n        \t"skip": 0,\n        \t"limit": 10\n    \t},\n    \t"testMode": false,\n    \t"requestedGranularity": "record"\n\t}\n}\' \\\n  http://localhost:5050/api/individuals\n\n\nAlphanumerical value query (NOT value)\n\ncurl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n\t"meta": {\n    \t"apiVersion": "2.0"\n\t},\n\t"query": {\n    \t"filters": [\t{\n    \t"id": "geographicOrigin",\n    \t"operator": "!",\n    \t"value": "%land%"\n\t}],\n    \t"includeResultsetResponses": "HIT",\n    \t"pagination": {\n        \t"skip": 0,\n        \t"limit": 10\n    \t},\n    \t"testMode": false,\n    \t"requestedGranularity": "record"\n\t}\n}\' \\\n  http://localhost:5050/api/individuals\n\n\n\n\nSemantic similarity query\n\ncurl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n\t"meta": {\n    \t"apiVersion": "2.0"\n\t},\n\t"query": {\n    \t"filters": [    \t{\n    \t"id": "NCIT:C16576",\n    \t"similarity": "high"\n\t}],\n    \t"includeResultsetResponses": "HIT",\n    \t"pagination": {\n        \t"skip": 0,\n        \t"limit": 10\n    \t},\n    \t"testMode": false,\n    \t"requestedGranularity": "record"\n\t}\n}\' \\\n  http://localhost:5050/api/individuals\n\nCustom filter query\n\ncurl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n\t"meta": {\n    \t"apiVersion": "2.0"\n\t},\n\t"query": {\n    \t"filters": [    \t{\n    \t"id": "sampleOriginType:blood"\n\t}],\n    \t"includeResultsetResponses": "HIT",\n    \t"pagination": {\n        \t"skip": 0,\n        \t"limit": 10\n    \t},\n    \t"testMode": false,\n    \t"requestedGranularity": "record"\n\t}\n}\' \\\n  http://localhost:5050/api/biosamples\n\nJoin genomic query\n\ncurl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n\t"meta": {\n    \t"apiVersion": "2.0"\n\t},\n\t"query": {\n    \t"requestParameters": {\n\t"alternateBases": "G" ,\n\t"referenceBases": "A" ,\n"start": [ 16050074 ],\n        \t"end": [ 16050568 ],\n     \t\t"variantType": "SNP"\n    \t},\n    \t"filters": [\n{"id": "NCIT:C20197", "scope": "individual"}, {"id": "NCIT:C42331", "scope": "individual"}],\n    \t"includeResultsetResponses": "HIT",\n    \t"pagination": {\n        \t"skip": 0,\n        \t"limit": 100\n    \t},\n    \t"testMode": false,\n    \t"requestedGranularity": "record"\n\t}\n}\' \\\n  http://localhost:5050/api/g_variants\n\n\n\n\n\nJoin individuals query\n\ncurl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n\t"meta": {\n    \t"apiVersion": "2.0"\n\t},\n\t"query": {\n    \t"requestParameters": {\n\t"alternateBases": "G" ,\n\t"referenceBases": "A" ,\n"start": [ 16050004 ],\n        \t"end": [ 16950115 ],\n        \t"variantType": "SNP"\n    \t},\n    \t"filters": [\n{"id": "NCIT:C20197"}, {"id": "NCIT:C42331"}],\n    \t"includeResultsetResponses": "HIT",\n    \t"pagination": {\n        \t"skip": 0,\n        \t"limit": 10\n    \t},\n    \t"testMode": false,\n    \t"requestedGranularity": "record"\n\t}\n}\' \\\n  http://localhost:5050/api/individuals\n\nQuery with access token\n\ncurl \\\n  -H \'Content-Type: application/json\' \\\n  -H \'Authorization: Bearer <access_token>\' \\\n  -X POST \\\n  -d \'{\n\t"meta": {\n    \t"apiVersion": "2.0"\n\t},\n\t"query": {\n"requestParameters": {    \t},\n \n    \t"filters": [\t{\n    \t"id": "geographicOrigin",\n    \t"operator": "=",\n    \t"value": "%land%"\n\t}],\n    \t"includeResultsetResponses": "HIT",\n    \t"pagination": {\n        \t"skip": 0,\n        \t"limit": 10\n    \t},\n    \t"testMode": false,\n    \t"requestedGranularity": "record"\n\t}\n}\' \\\n  http://localhost:5050/api/individuals\n\n\nList of endpoints\n\nEndpoints that accept all query parameters\n\n/api/g_variants\n/api/individuals\n\nEndpoints that only accept filters and generic parameters\n\n/api/analyses\n/api/biosamples\n/api/cohorts\n/api/datasets\n/api/runs\n\nEndpoints that only accept generic parameters\n\n/api/filtering_terms\n/api/analyses/{id}\n/api/analyses/{id]/g_variants\n/api/analyses/filtering_terms\n/api/biosamples/{id}\n/api/biosamples/{id}/analyses\n/api/biosamples/{id}/g_variants\n/api/biosamples/{id}/runs\n/api/biosamples/filtering_terms\n/api/cohorts/{id}\n/api/cohorts/{id}/analyses\n/api/cohorts/{id}/biosamples\n/api/cohorts/{id}/g_variants\n/api/cohorts/{id}/individuals\n/api/cohorts/{id}/runs\n/api/cohorts/filtering_terms\n/api/datasets/{id}\n/api/datasets/{id}/analyses\n/api/datasets/{id}/biosamples\n/api/datasets/{id}/g_variants\n/api/datasets/{id}/individuals\n/api/datasets/{id}/runs\n/api/datasets/filtering_terms\n/api/g_variants/{id}\n/api/g_variants/{id}/analyses\n/api/g_variants/{id}/biosamples\n/api/g_variants/{id}/individuals\n/api/g_variants/{id}/runs\n/api/g_variants/filtering_terms\n/api/individuals/{id}\n/api/individuals/{id}/analyses\n/api/individuals/{id}/biosamples\n/api/individuals/{id}/g_variants\n/api/individuals/{id}/runs\n/api/individuals/filtering_terms\n/api/runs/{id}\n/api/runs/{id}/analyses\n/api/runs/{id}/g_variants\n/api/runs/filtering_terms\n\nEndpoints that don\u2019t accept query parameters\n\n/api\n/api/info\n/api/service-info\n/api/configuration\n/api/entry_types\n/api/map\n\n\n\n',"Querying the API"),Ul("/automated-deployment","Documentation\n\nBeacon 2 RI API\n\nAutomated Deployment\n\nThis deployment just uses a bash script and a Makefile which won\u2019t deploy the idp authentication service. If you wish to build the idp service you will have to go to Manual Deployment.\n\nCloning the repository\n\nFirst of all, start by cloning the GitHub repository in your system.\n\ngit clone https://github.com/EGA-archive/beacon2-ri-api.git\n\nMake sure the next list of ports are free of use in your system:\n\n27017 \u2192 mongo\n5050 \u2192 beacon\n5051 \u2192 permissions\nExecute start script from root\n\nbash start.sh\n\nIf the operation is successful, you will have a beacon up and running at:\nhttp://localhost:5050/api","Automated Deployment"),Ul("/manual-deployment",'Documentation\n\nBeacon 2 RI API\n\nManual Deployment\n\nCloning the repository\n\nFirst of all, start by cloning the GitHub repository in your system.\n\ngit clone https://github.com/EGA-archive/beacon2-ri-api.git\n\nCreating the containers\n\nCreate the external docker network needed before building the containers.\n\ndocker network create my-app-network\n\nMake sure the next list of ports are free for use in your system:\n\n27017 \u2192 mongo\n8081 \u2192 mongo-express\n5050 \u2192 beacon\n5051 \u2192 permissions\n8010 \u2192 permissions UI\n8080 \u2192 Keycloak\n9991 \u2192 Keycloak SSL\n3000 \u2192 Beacon UI\nLight up the containers from the deploy folder.\n\ncd deploy && docker-compose up -d --build\n\nIf the containers are built correctly:\n\nThe API will run in http://localhost:5050/api\nThe Beacon UI will run in http://localhost:3000\nThe mongo-express UI will run in http://localhost:8081\nThe Keycloak UI will run in http://localhost:8080/auth\nData injection\n\nCopy all the available BFF files to the mongo container using the following commands:\n\nbash\nCopy\nEdit\ndocker cp /path/to/analyses.json rimongo:tmp/analyses.json\ndocker cp /path/to/biosamples.json rimongo:tmp/biosamples.json\ndocker cp /path/to/cohorts.json rimongo:tmp/cohorts.json\ndocker cp /path/to/datasets.json rimongo:tmp/datasets.json\ndocker cp /path/to/genomicVariations.json rimongo:tmp/genomicVariations.json\ndocker cp /path/to/individuals.json rimongo:tmp/individuals.json\ndocker cp /path/to/runs.json rimongo:tmp/runs.json\nNot all the above commands are needed to run a beacon. A beacon can run without data or with data for a single collection (e.g., only individuals).\n\nNow execute the insertion of the copied files into the mongo database.\n\nbash\nCopy\nEdit\ndocker exec rimongo mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /tmp/datasets.json --collection datasets\ndocker exec rimongo mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /tmp/analyses.json --collection analyses\ndocker exec rimongo mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /tmp/biosamples.json --collection biosamples\ndocker exec rimongo mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /tmp/cohorts.json --collection cohorts\ndocker exec rimongo mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /tmp/genomicVariations.json --collection genomicVariations\ndocker exec rimongo mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /tmp/individuals.json --collection individuals\ndocker exec rimongo mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /tmp/runs.json --collection runs\nData indexing\n\nFor the API to respond quickly to queries, you have to index your database. You can create the necessary indexes by running the following script:\n\ndocker exec beacon python beacon/reindex.py\n\nYou will need to run this script each time you inject new data.',"Manual Deployment"),Ul("/data-linking",'Linking ids to dataset and cohort\n\nIn case your data is splitted by dataset or cohort, you will need to tell beacon which biosample ids and which individual ids belong to which dataset.\n\nEdit the file datasets.yml that is inside beacon/request folder and write the dataset with an array of all the ids (for individuals and biosamples together).\n\ndataset_name: [\u201cindividualId1\u201d, \u201cindividualId2\u201d, \u201cbiosampleId1\u201d, \u201cbiosampleId2\u201d]\n\ncohort_name: [\u201cindividualId1\u201d, \u201cindividualId2\u201d, \u201cbiosampleId1\u201d, \u201cbiosampleId2\u201d]\n\nNote that the dataset_name and cohort_name have to be identical as how is spelled the id for the dataset or cohort in the BFF record.\n\nExtract filtering terms\n\nTo automatically fill in the filtering terms endpoint and be able to apply the ontologies that are found inside your data to query your beacon, please execute the next script:\n\ndocker exec beacon python beacon/db/extract_filtering_terms.py\n\nManually adding filtering terms (optional)\n\nTo manually add filtering terms to your beacon, you will need to add them in your mongodb, by executing the next command, for example, for an alphanumeric term that applies to the libraryStrategy field for the runs collection:\n\ndb.filtering_terms.insertMany([{ "type": "alphanumeric", "id": "libraryStrategy", "scope": [ "runs"] }])\n\nAfter that, you will need to specify that this alphanumeric term is not related to an ontology by editing the conf.py inside beacon and deploy folder in the alphanumeric_terms array:\n\nalphanumeric_terms = [\'libraryStrategy\', \'molecularAttributes.geneIds\', \'diseases.ageOfOnset.iso8601duration\']\n\nGet descendant terms (optional)\n\nIf you have extracted filtering terms. in case you wish to add information about the descendant terms and similarity found for each ontology and be able to use the similarity and descendant terms as beacon options, please execute the next script:\n\ndocker exec beacon python beacon/db/get_descendants.py.\n',"Data Linking"),Ul("/api-configuration","Managing dataset permissions\n\nThere are 3 possible levels of beacon security for a dataset: public, registered and controlled.\nA public dataset is a dataset that will be returned in a beacon query without an authentication token. \nA registered dataset is a dataset that will be shown after a user sends a valid token (in other words, is logged in).\nA controlled dataset is a dataset that needs a user to send a valid token for authentication and the user needs to be allowed to query that dataset.\nThere are two possible methods to administer these permissions for your datasets in beacon:\n\nMethod 1: Edit the .yml files\n\nFor a public dataset, edit the public_datasets.yml file inside permissions folder and add the dataset with the exact name it appears in the id from its pertinent record in datasets mongo collection:\n\npublic_datasets:\n- dataset_id\n\nFor a registered dataset, edit the registered_datasets.yml file inside permissions folder and add the dataset with the exact name it appears in the id from its pertinent record in datasets mongo collection:\n\nregistered_datasets:\n- dataset_id\n\nFor a controlled dataset, edit the controlled_datasets.yml file inside permissions folder and add the dataset inside the username you wish to give permissions with the exact name for the dataset as it appears in the id from its pertinent record in datasets mongo collection and the exact name the user has in Keycloak:\n\nusername:\n- dataset_id\n\nMethod 2: Launching the beacon admin UI\n\nCreate an .env file inside permissions/permissions-ui/web folder and add the next variables with their values:\n\nSECRET_KEY=\"your_permissions_ui_secret_key\"\nOIDC_RP_CLIENT_ID='your_client_id'\nOIDC_RP_CLIENT_SECRET='your_client_secret'\n\nExecute the next command to launch the UI:\n\ndocker exec beacon-permissions bash permissions/permissions-ui/start.sh\n\nAfter that, open the ui in port http://localhost:8010 and navigate through public, registered and controlled datasets to assign permissions for the datasets you have inside your mongo database.\n\nSupplying AAI credentials\n\nFor making Keycloak or LifeSciende AAI work with beacon, you will need to add the client ID and client secret for each IDP in a new file you have to create inside permissions folder. This file must be called .env and needs to have the following variables:\n\nLSAAI_CLIENT_ID='your_lsaai_client_id'\nLSAAI_CLIENT_SECRET='your_lsaai_client_secret'\nKEYCLOAK_CLIENT_ID='your_keycloak_client_id'\nKEYCLOAK_CLIENT_SECRET='your_keycloak_client_secret'\n\nNote that you only need to fill variables for the AAI method you want to enable.\n\nAfter that, make sure you build your beacon container again:\n\ncd deploy && docker-compose up -d --build beacon\n\nHandling CORS\n\nTo avoid CORS using beacon and the frontend or a third-party authorization site like Keycloak, you will have to include all these URLs inside __main__.py in the beacon folder, including them in the CORS middleware,\n\nmiddlewares=[web.normalize_path_middleware(), middlewares.error_middleware, cors_middleware(origins=[\"your_URL\"...\n\n and CORS routes lists:\n\n   for route in list(beacon.router.routes()):\n        cors.add(route, {\n        \"your_URL\":\n            aiohttp_cors.ResourceOptions(allow_credentials=True,\n            expose_headers=\"*\",\n            allow_methods=(\"POST\", \"PATCH\", \"GET\", \"OPTIONS\"),\n            allow_headers=DEFAULT_ALLOW_HEADERS),\n\nEditing your beacon information\n\nTo show correctly your beacon\u2019s information you will need to edit both conf.py files from beacon and deploy folders.\nTo do so, edit the following variables:\n\nbeacon_id = 'org.ega-archive.beacon-ri-demo'  # ID of the Beacon\nbeacon_name = 'Beacon Reference Implementation demo'  # Name of the Beacon service\napi_version = 'v2.0.0'  # Version of the Beacon implementation\nuri = 'https://beacon-apis-demo.ega-archive.org/api/'\norg_id = 'EGA'  # Id of the organization\norg_name = 'European Genome-Phenome Archive (EGA)'  # Full name\norg_description = ('The European Genome-phenome Archive (EGA) '\n                   'is a service for permanent archiving and sharing '\n                   'of all types of personally identifiable genetic '\n                   'and phenotypic data resulting from biomedical research projects.')\norg_adress = ('C/ Dr. Aiguader, 88'\n              'PRBB Building'\n              '08003 Barcelona, Spain')\norg_welcome_url = 'https://ega-archive.org/'\norg_contact_url = 'mailto:beacon.ega@crg.eu'\norg_logo_url = 'https://legacy.ega-archive.org/images/logo.png'\norg_info = ''\ndescription = r\"This Beacon is based on synthetic data hosted at the <a href='https://ega-archive.org/datasets/EGAD00001003338'>EGA</a>. The dataset contains 2504 samples including genetic data based on 1K Genomes data, and 76 individual attributes and phenotypic data derived from UKBiobank.\"\nversion = 'v2.0'\nwelcome_url = 'https://beacon.ega-archive.org/'\nalternative_url = 'https://beacon-apis-demo.ega-archive.org/api/'\ncreate_datetime = '2021-11-29T12:00:00.000000'\nupdate_datetime = ''\n\nNote that this is the info that will be shown in the /info endpoint for your beacon.\n\nLimiting the granularity\n\nBeacon RI can limit the granularity per user or for all the beacon.\nIf you want to limit the granularity for all the users to a maximum, you can point it in conf.py for both beacon and deploy folders, by editing the next variables:\n\ndefault_beacon_granularity = \"record\"\nmax_beacon_granularity = \"record\"\n\nIf you want to limit the granularity per user, you can do this editing the response_type.yml file inside beacon/request folder, adding the username and the maximum granularity you want to allow this user to have:\n\ndummy_user:\n- record\n\nAdding beacon handovers\n\nYou can specify what are the handovers for your beacon by adding them in the array of your conf.py file inside beacon folder:\n\nbeacon_handovers = [\n    {\n        'handoverType': {\n            'id': 'CUSTOM:000001',\n            'label': 'Project description'\n        },\n        'note': 'Project description',\n        'url': 'https://www.nist.gov/programs-projects/genome-bottle'\n    }\n]\n\n\n\n\n","Configuration"),Ul("/querying-the-api",'Beacon RI accepts two types of request methods: GET and POST.\n\nGET Method\n\nGET method is a bit more limited as this method only accepts request parameters, three generic parameters and ontology filters.\n\nQuery Parameters\nThese are the defined set of parameters (key-value pair) attached to the end of a URL used to provide additional information to a web server when making requests.\n\nGeneric parameters\n\nrequestedSchema\tstring\nExample: requestedSchema=ga4gh-service-info-v1.0\n\nskip\tinteger (Skip) >= 0\nDefault: 0\nExample: skip=0\n\nlimit\tinteger (Limit) >= 0\nDefault: 10\nMax: 100\nExample: limit=10\n\nRequest parameters\n\n\tstart\tinteger <int64> >= 0 (or comma separated integers)\n\tExample: start=16050074\n\nend\tinteger <int64> >= 0 (or comma separated integers)\nExample: end=16052080\n\n\tassemblyId\tstring\n\tExample: assemblyId=GRCh38 \n\nreferenceName\t<int64> >= 0\nExample: referenceName=22\n\nreferenceBases\tstring\nExample: referenceBases=G\n\nalternateBases\tstring\nExample: alternateBases=G\n\nvariantMinLength\tinteger <int64> >= 0\nExample: variantMinLength=5\n\nvariantMaxLength\tinteger <int64> >= 0\nExample: variantMaxLength=2\n\ngeneId\t\tstring\nExample: geneId=CCNL2\n\naminoacidChange\tstring\nExample: aminoacidChange=p.Thr130Met\n\nvariantType\tstring\nExample: variantType=SNP\n\ngenomicAlleleShortForm\tstring\nExample: genomicAlleleShortForm=NC_000022.11:g.16050075A>G\n\nclinicalRelevance\tstring\nExample: clinicalRelevance=pathogenic\n\nFilters\n\t\n\tfilters\tstring \n\tExample: filters=NCIT:C16576,NCIT:C42331\n\nGET query examples\n\nBeacon Sequence Query\n\nhttp://localhost:5050/api/g_variants?referenceName=22&start=7577120&referenceBases=G&alternateBases=A\n\nBeacon GeneId Query\n\nhttp://localhost:5050/api/g_variants?geneId=CHR_START-DUXAP8&variantMaxLength=2&variantType=INDEL\n\nBeacon Bracket Query\n\nhttp://localhost:5050/api/g_variants?referenceName=22&start=16050074,16050076&end=16050090&variantType=SNP\n\n\nPOST Method\n\nIn POST method requests, the endpoints accept all generic parameters and all the filters, apart from the request parameters.\n\nQuery Parameters\n\nGeneric parameters\n\nincludeResultsetResponses\tstring (IncludeResultsetResponses)\nDefault: "HIT"\nEnum: "ALL" "HIT" "MISS" "NONE"\n\npagination\tobject (Pagination)\nItems: \nskip\tinteger (Skip) >= 0\nDefault: 0\nExample: skip=0\n\nlimit\tinteger (Limit) >= 0\nDefault: 10\nMax: 100\nExample: limit=10\n\nrequestedGranularity\tstring (Granularity)\nDefault: "boolean"\nEnum: "boolean" "count" "record"\n\ntestMode\t\nboolean (TestMode)\nDefault: false\n\nRequest parameters\n\n\tstart\tinteger <int64> >= 0 (or comma separated integers)\n\tExample: start=16050074\n\nend\tinteger <int64> >= 0 (or comma separated integers)\nExample: end=16052080\n\n\tassemblyId\tstring\n\tExample: assemblyId=GRCh38 \n\nreferenceName\t<int64> >= 0\nExample: referenceName=22\n\nreferenceBases\tstring\nExample: referenceBases=G\n\nalternateBases\tstring\nExample: alternateBases=G\n\nvariantMinLength\tinteger <int64> >= 0\nExample: variantMinLength=5\n\nvariantMaxLength\tinteger <int64> >= 0\nExample: variantMaxLength=2\n\ngeneId\t\tstring\nExample: geneId=CCNL2\n\naminoacidChange\tstring\nExample: aminoacidChange=p.Thr130Met\n\nvariantType\tstring\nExample: variantType=SNP\n\ngenomicAlleleShortForm\tstring\nExample: genomicAlleleShortForm=NC_000022.11:g.16050075A>G\n\nclinicalRelevance\tstring\nExample: clinicalRelevance=pathogenic\n\nFilters\n\t\n\tfilters\tArray of OntologyFilter (object) or AlphanumericFilter (object) or CustomFilter (object) (Filtering Term Element)\n\tExample:\n       "filters": [    {\n        "id": "geographicOrigin",\n        "operator": "!",\n        "value": "England"\n    }]\n\nPOST query examples\n\nGenomic query\n\ncurl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n\t"meta": {\n    \t"apiVersion": "2.0"\n\t},\n\t"query": {\n    \t"requestParameters": {\n\t"alternateBases": "G" ,\n\t"referenceBases": "A" ,\n"start": [ 16050074 ],\n        \t"end": [ 16050568 ],\n     \t\t"variantType": "SNP"\n    \t},\n    \t"filters": [],\n    \t"includeResultsetResponses": "HIT",\n    \t"pagination": {\n        \t"skip": 0,\n        \t"limit": 10\n    \t},\n    \t"testMode": false,\n    \t"requestedGranularity": "record"\n\t}\n}\' \\\n  http://localhost:5050/api/g_variants\n\n\nSimple CURIE based filters query\n\ncurl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n\t"meta": {\n    \t"apiVersion": "2.0"\n\t},\n\t"query": {\n    \t"requestParameters": {\n    \t},\n    \t"filters": [\n{"id": "NCIT:C20197", "scope": "individual"}, {"id": "NCIT:C42331", "scope": "individual"}],\n    \t"includeResultsetResponses": "HIT",\n    \t"pagination": {\n        \t"skip": 0,\n        \t"limit": 10\n    \t},\n    \t"testMode": false,\n    \t"requestedGranularity": "record"\n\t}\n}\' \\\n  http://localhost:5050/api/individuals\n\nNumerical value query\n\ncurl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n\t"meta": {\n    \t"apiVersion": "2.0"\n\t},\n\t"query": {\n    \t"filters": [{\n    \t"id": "Weight",\n    \t"operator": ">",\n    \t"value": "75"\n\t}],\n    \t"includeResultsetResponses": "HIT",\n    \t"pagination": {\n        \t"skip": 0,\n        \t"limit": 10\n    \t},\n    \t"testMode": false,\n    \t"requestedGranularity": "record"\n\t}\n}\' \\\n  http://localhost:5050/api/individuals\n\n\n\nAlphanumerical value query (Exact value)\n\ncurl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n\t"meta": {\n    \t"apiVersion": "2.0"\n\t},\n\t"query": {\n    \t"filters": [\t{\n    \t"id": "geographicOrigin",  \n    \t"operator": "=",\n    \t"value": "England"    \n\t}],\n    \t"includeResultsetResponses": "HIT",\n    \t"pagination": {\n        \t"skip": 0,\n        \t"limit": 10\n    \t},\n    \t"testMode": false,\n    \t"requestedGranularity": "record"\n\t}\n}\' \\\n  http://localhost:5050/api/individuals\n\nAlphanumerical value query (Like value)\n\ncurl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n\t"meta": {\n    \t"apiVersion": "2.0"\n\t},\n\t"query": {\n    \t"filters": [\t{\n    \t"id": "geographicOrigin",\n    \t"operator": "=",\n    \t"value": "%land%"\n\t}],\n    \t"includeResultsetResponses": "HIT",\n    \t"pagination": {\n        \t"skip": 0,\n        \t"limit": 10\n    \t},\n    \t"testMode": false,\n    \t"requestedGranularity": "record"\n\t}\n}\' \\\n  http://localhost:5050/api/individuals\n\n\nAlphanumerical value query (NOT value)\n\ncurl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n\t"meta": {\n    \t"apiVersion": "2.0"\n\t},\n\t"query": {\n    \t"filters": [\t{\n    \t"id": "geographicOrigin",\n    \t"operator": "!",\n    \t"value": "%land%"\n\t}],\n    \t"includeResultsetResponses": "HIT",\n    \t"pagination": {\n        \t"skip": 0,\n        \t"limit": 10\n    \t},\n    \t"testMode": false,\n    \t"requestedGranularity": "record"\n\t}\n}\' \\\n  http://localhost:5050/api/individuals\n\n\n\n\nSemantic similarity query\n\ncurl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n\t"meta": {\n    \t"apiVersion": "2.0"\n\t},\n\t"query": {\n    \t"filters": [    \t{\n    \t"id": "NCIT:C16576",\n    \t"similarity": "high"\n\t}],\n    \t"includeResultsetResponses": "HIT",\n    \t"pagination": {\n        \t"skip": 0,\n        \t"limit": 10\n    \t},\n    \t"testMode": false,\n    \t"requestedGranularity": "record"\n\t}\n}\' \\\n  http://localhost:5050/api/individuals\n\nCustom filter query\n\ncurl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n\t"meta": {\n    \t"apiVersion": "2.0"\n\t},\n\t"query": {\n    \t"filters": [    \t{\n    \t"id": "sampleOriginType:blood"\n\t}],\n    \t"includeResultsetResponses": "HIT",\n    \t"pagination": {\n        \t"skip": 0,\n        \t"limit": 10\n    \t},\n    \t"testMode": false,\n    \t"requestedGranularity": "record"\n\t}\n}\' \\\n  http://localhost:5050/api/biosamples\n\nJoin genomic query\n\ncurl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n\t"meta": {\n    \t"apiVersion": "2.0"\n\t},\n\t"query": {\n    \t"requestParameters": {\n\t"alternateBases": "G" ,\n\t"referenceBases": "A" ,\n"start": [ 16050074 ],\n        \t"end": [ 16050568 ],\n     \t\t"variantType": "SNP"\n    \t},\n    \t"filters": [\n{"id": "NCIT:C20197", "scope": "individual"}, {"id": "NCIT:C42331", "scope": "individual"}],\n    \t"includeResultsetResponses": "HIT",\n    \t"pagination": {\n        \t"skip": 0,\n        \t"limit": 100\n    \t},\n    \t"testMode": false,\n    \t"requestedGranularity": "record"\n\t}\n}\' \\\n  http://localhost:5050/api/g_variants\n\n\n\n\n\nJoin individuals query\n\ncurl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n\t"meta": {\n    \t"apiVersion": "2.0"\n\t},\n\t"query": {\n    \t"requestParameters": {\n\t"alternateBases": "G" ,\n\t"referenceBases": "A" ,\n"start": [ 16050004 ],\n        \t"end": [ 16950115 ],\n        \t"variantType": "SNP"\n    \t},\n    \t"filters": [\n{"id": "NCIT:C20197"}, {"id": "NCIT:C42331"}],\n    \t"includeResultsetResponses": "HIT",\n    \t"pagination": {\n        \t"skip": 0,\n        \t"limit": 10\n    \t},\n    \t"testMode": false,\n    \t"requestedGranularity": "record"\n\t}\n}\' \\\n  http://localhost:5050/api/individuals\n\nQuery with access token\n\ncurl \\\n  -H \'Content-Type: application/json\' \\\n  -H \'Authorization: Bearer <access_token>\' \\\n  -X POST \\\n  -d \'{\n\t"meta": {\n    \t"apiVersion": "2.0"\n\t},\n\t"query": {\n"requestParameters": {    \t},\n \n    \t"filters": [\t{\n    \t"id": "geographicOrigin",\n    \t"operator": "=",\n    \t"value": "%land%"\n\t}],\n    \t"includeResultsetResponses": "HIT",\n    \t"pagination": {\n        \t"skip": 0,\n        \t"limit": 10\n    \t},\n    \t"testMode": false,\n    \t"requestedGranularity": "record"\n\t}\n}\' \\\n  http://localhost:5050/api/individuals\n\n\nList of endpoints\n\nEndpoints that accept all query parameters\n\n/api/g_variants\n/api/individuals\n\nEndpoints that only accept filters and generic parameters\n\n/api/analyses\n/api/biosamples\n/api/cohorts\n/api/datasets\n/api/runs\n\nEndpoints that only accept generic parameters\n\n/api/filtering_terms\n/api/analyses/{id}\n/api/analyses/{id]/g_variants\n/api/analyses/filtering_terms\n/api/biosamples/{id}\n/api/biosamples/{id}/analyses\n/api/biosamples/{id}/g_variants\n/api/biosamples/{id}/runs\n/api/biosamples/filtering_terms\n/api/cohorts/{id}\n/api/cohorts/{id}/analyses\n/api/cohorts/{id}/biosamples\n/api/cohorts/{id}/g_variants\n/api/cohorts/{id}/individuals\n/api/cohorts/{id}/runs\n/api/cohorts/filtering_terms\n/api/datasets/{id}\n/api/datasets/{id}/analyses\n/api/datasets/{id}/biosamples\n/api/datasets/{id}/g_variants\n/api/datasets/{id}/individuals\n/api/datasets/{id}/runs\n/api/datasets/filtering_terms\n/api/g_variants/{id}\n/api/g_variants/{id}/analyses\n/api/g_variants/{id}/biosamples\n/api/g_variants/{id}/individuals\n/api/g_variants/{id}/runs\n/api/g_variants/filtering_terms\n/api/individuals/{id}\n/api/individuals/{id}/analyses\n/api/individuals/{id}/biosamples\n/api/individuals/{id}/g_variants\n/api/individuals/{id}/runs\n/api/individuals/filtering_terms\n/api/runs/{id}\n/api/runs/{id}/analyses\n/api/runs/{id}/g_variants\n/api/runs/filtering_terms\n\nEndpoints that don\u2019t accept query parameters\n\n/api\n/api/info\n/api/service-info\n/api/configuration\n/api/entry_types\n/api/map\n\n\n\n',"Querying the API"),Ul("/starting-guide","Documentation > Beacon RI Tools v2 > Starting Guide\n\nBeacon RI Tools v2\nStarting Guide\n\nThe Beacon RI Tools v2 is a set of tools written in Python available in the following GitHub repository:\nhttps://github.com/EGA-archive/beacon-data-tools\n\nNotice that Beacon RI Tools v2 are meant to work with either Beacon 2 RI API or Beacon 2 PI API. Furthermore, it specifically comes with all the features that are needed for generating Beacon v2 PI API compatible data.\n\nThe main goal of Beacon RI Tools v2 is to obtain the Beacon Friendly Format (BFF) files, which are JSON files following Beacon v2 official specifications. These files have the correct format and structure to be injected into a Beacon v2 MongoDB database.\n\nDiscover how to set up a Beacon v2 with MongoDB and learn to inject BFF files by visiting the official Beacon v2 PI API:\nhttps://github.com/EGA-archive/beacon2-ri-api \u2014 where you can download your own Beacon v2 for free!\n\nThe Beacon data model consists of 7 collections:\n- Analyses\n- Biosamples\n- Cohorts\n- Datasets\n- GenomicVariations\n- Individuals\n- Runs\n\nTherefore, the Beacon RI Tools v2 allow the user to create the different 7 BFF options.\n\nThe 6 metadata BFF files are created populating a CSV file with a defined structure that follows the Beacon data model. However, the genomicVariations collection can be created either from a CSV or from a VCF file, minimizing conversion time.\n\nThe initial step, since the Beacon RI Tools v2 cannot convert the data from any source, it is to populate the CSV templates that are built following the 7 collections of the Beacon data model. This enables the Beacon RI Tools v2 to convert from CSV to BFF. Note that you can find the CSV templates in the templates folder of the repository:\nhttps://github.com/EGA-archive/beacon-data-tools/tree/main/csv/templates\n\nInstallation Guide\n\nFirst of all, clone or download the Beacon RI Tools v2 repository to your computer:\n\ngit clone https://github.com/EGA-archive/beacon-data-tools.git\n\nTo light up the container with Beacon RI Tools v2, execute the docker-compose command inside the root folder of the repository:\n\ndocker compose up -d --build\n\nOnce the container is up and running, you can start using Beacon RI Tools v2, congratulations!","Starting Guide"),Ul("/configuration-file","Documentation Beacon RI Tools v2 Configuration File\nBeacon RI Tools v2\nConfiguration File\nIn order to use Beacon RI Tools v2 one needs to modify a configuration file with the desired files and options for the different collections.\n\nYou will find conf.py inside the conf directory. Inside this file, you will find the following information:\n\nInput and Output files config parameters\ncsv_filename='csv/examples/cohorts.csv' output_docs_folder='output_docs/CINECA_dataset/'\n\nGeneric config parameters\nThe first part of this configuration only concerns the conversion of CSV to BFF. It is used only when you are converting from CSV.\n\nThe csv_filename variable sets where the CSV file is and from where the script will read the data. Bear in mind that the CSV must be inside the folder csv, and you can use subfolders within it. This CSV file needs to have the headers written as you can find in the files inside templates. Note that any header that has a different name from the ones that appear inside the templates will not be read by the Beacon Data Tools.\n\nThe output_docs_folder sets the folder where your final BFF (JSON files) will be saved once execution of beacon tools finishes. This folder is mandatory to be always inside output_docs, so only the subdirectory inside output_docscan be modified in this path.\n\nVCF conversion config parameters\nallele_counts=False\nreference_genome='GRCh37' # Choose one between NCBI36, GRCh37, GRCh38\ndatasetId='COVID_pop11_fin_2'\ncase_level_data=False\nexact_heterozygosity=False\nnum_rows=15000000\nverbosity=False\n\nThe second part of the configuration file pertains to the VCF to BFF conversion. This only needs to be used in case you are using a VCF as a source for the genomic variants collection.\n\nThe num_variants is the variable you need to write in case you are executing the VCF conversion (genomicVariations_vcf.py). This will tell the script how many variants will be read and converted from the file(s).\n\nThe allele_frequency field lets you set a threshold for the allele frequency (AF) of the variants you want to convert from the VCF file. If you set it at 0.5, all the variants with 0.5 or less AF will be converted to BFF. 1 is the default value (all variants will be converted).\n\nThe allele_counts now is not implemented yet, just leave it as False.\n\nThe reference_genome is the reference genome the tool will use to map the position of the chromosomes. Make sure to select the same version as the one used to generate your data.\n\nThe datasetId, case_level_data, and exact_zygosity parameters are only applicable in the Beacon Production Implementation environment.\n\nThe datasetId needs to match the id of your datasets.csv or datasets.json file. This will add a datasetId field in every record to match the record with the dataset it belongs to.\n\nWhen converting a VCF file to BFF, you can use two optional boolean parameters to include sample-level information about each variant: case_level_data and exact_zygosity.\n\ncase_level_data will link variants to biosamples. If set to True, this option enables mapping each variant to the specific biosamples (i.e., samples) that carry it. This allows downstream queries to return more granular, sample-level results instead of just listing variants in general.\n\n\nImportant: For this mapping to work correctly, you must:\nHave genotype (GT) information available in the VCF.\nProvide a valid biosamples schema in your BFF files.\n\nBe sure to read the Case-level data conversion section for detailed guidance on setting this up.\n\nexact_zygosity will refine biosample classification. If set to True This parameter classifies each biosample based on its genotype:\nHomozygous for the reference allele\nHeterozygous\nHomozygous for the alternate allele\nThis adds precision to your data and can help users interpret zygosity in their queries.\n\n\nIf case_level_data is set to False, make sure to also setexact_zygosity to False \u2014 it won\u2019t have any effect otherwise.\n\nThe num_rows  are the aproximate calculation you expect for the total of variants in each vcf there are. Make sure this is greater than the total variants expected. It was automatically calculated before but it was very slow sometimes to calculate all the variants number in a VCF.\n\nThe verbosity will give streaming logs with the reason why a variant has been skipped to be inserted. Recommendation is to leave this as False.","Configuration File"),Ul("/conversion-from-csv-to-bff",'Documentation Beacon RI Tools v2 Conversion from CSV to BFF\nBeacon RI Tools v2\nConversion from CSV to BFF\nChoosing Column Headers\nEach type of entry (e.g., analyses, biosamples, individuals, etc.) has its own corresponding CSV template located in the csv-templates folder. These templates list all the valid column headers you can use.\n\nFor example, in biosamples.csv, one of the headers you\'ll see is:\n\nbiosampleStatus|id\n\nThis naming convention follows the GA4GH Beacon v2 specification. The | (pipe) character separates different levels of a nested field: the part on the left represents the parent field, and the part on the right is the nested element within it. In this case:\n\nbiosampleStatus is an object field.\nid is one of its subfields, representing an ontology identifier.\nAccording to the Beacon v2 specification, biosampleStatus is defined as an ontology term describing the classification of the sample (e.g., whether it\u2019s a reference or abnormal sample). Here\u2019s a simplified excerpt from the spec:\n\n"biosampleStatus": {\n  "description": "Classification of the sample using EFO ontology terms.",\n  "examples": [\n    { "id": "EFO:0009654", "label": "reference sample" },\n    { "id": "EFO:0009655", "label": "abnormal sample" },\n    { "id": "EFO:0009656", "label": "neoplastic sample" }\n    // ...\n  ]\n}\n\nSo, in your CSV file, the biosampleStatus|id column should contain ontology IDs like:\n\nbiosampleStatus|id\nEFO:0009654\nEFO:0009655\n\nOptionally, you can include the label for each ontology ID in a second column called biosampleStatus|label, like so:\n\nbiosampleStatus|id,biosampleStatus|label\nEFO:0009654,reference sample\nEFO:0009655,abnormal sample\n\nEach row corresponds to a different biosample, and both the id and label must match the values defined in the ontology.\n\nFilling in Data\nTo convert metadata or phenotypic/clinical data into BFF (Beacon File Format), you need to create one CSV file per collection type (e.g., individuals, biosamples, analyses, etc.). Each file should contain records formatted according to the column headers, which map directly to fields in the Beacon v2 schema.\n\nEach row in a CSV file represents a new and independent document in the final output. Templates and examples for each collection type can be found in the csv folder of the repository.\n\nGeneral Rules for Filling In the CSV Files\nMultiple values in the same document:\nTo include multiple values in a single field (e.g. multiple IDs, multiple measures), separate them using the pipe symbol (|).\nExample:\n\nid\nHG00001|HG00002\n\nMake sure the values in related columns are in the same order:\n\nlabel\nlabelForHG00001|labelForHG00002\n\n\nFilling in info fields:\nThe info field is flexible and accepts free-form data, but must be written as JSON. This is also required for certain fields in the genomicVariations collection (e.g., copies or subjects).\nExample:\n\n{"info": "Some important information to be added."}\n\n\nRespect the column names:\nColumn headers must match exactly those defined in the template files. The conversion script will only recognize correctly spelled headers. Mistyped headers will be ignored or cause errors.\n\n\nOptional vs. required fields:\nNot all columns need to be filled in. Some are optional or context-specific. However, certain fields are required depending on the collection type (see below). If a required field is missing or misfilled, an error will be raised. Only columns with actual data will be included in the final output \u2014 unused columns can be removed if preferred.\n\nRequired Fields per Collection\nCollection\tRequired Fields\nAnalyses\tid, analysisDate, pipelineName\nBiosamples\tid, biosampleStatus, sampleOriginType\nCohorts\tid, name, cohortType\nDatasets\tid, name\nIndividuals\tid, sex\nRuns\tid, biosampleId, runDate\nNote symbol\nFor more information about each field, refer to the default schemas and the Beacon v2 specification.\n\nAdditional Notes\nThe id field in the Biosamples collection must match the sample IDs in your VCF header.\nThe id field in the Individuals collection does not have to match VCF sample IDs. But, if it doesn\u2019t, you must fill in the individualId field in the Biosamples collection to link individuals and samples.\nThe datasetId in the Datasets collection must match the datasetId defined in your conf.py file (this applies only to production Beacon deployments).\nYou are not required to submit CSV files for all collections. If you do not have data for a given collection, you can omit its CSV file \u2014 the Beacon tools will not raise any errors.\nConversion from CSV to BFF\nBefore converting your CSV files, ensure that the conf.py file (located in the conf folder) is correctly configured to read the appropriate input files and settings.\n\nConvert All CSVs at Once\nTo convert all CSV files at once, make sure they are saved in the csv folder (refer to the Configuration File section for details).\n\nThen run the following command from the root of the project:\n\ndocker exec -it ri-tools python convert_csvTObff.py\n\nThis script will process all available CSVs and generate one JSON file per collection in the output_docs folder.\n\nConvert a Specific Collection\nIf you prefer to convert only one collection at a time, you can use one of the following commands depending on the type of data:\n\ndocker exec -it ri-tools python analyses_csv.py\n\ndocker exec -it ri-tools python biosamples_csv.py\n\ndocker exec -it ri-tools python cohorts_csv.py\n\ndocker exec -it ri-tools python datasets_csv.py\n\ndocker exec -it ri-tools python genomicVariations_csv.py\n\ndocker exec -it ri-tools python individuals_csv.py\n\ndocker exec -it ri-tools python runs_csv.py\n\nEach script will generate a corresponding JSON file (e.g., cohorts.json) inside the output_docs folder.\n\nThese JSON documents follow the Beacon File Format (BFF) and are ready to be imported into a MongoDB database for use with a GA4GH Beacon v2 instance.\n\nTo learn how to load these JSON files into your Beacon, refer to the documentation for the Beacon v2 RI API and the Beacon v2 PI API.',"Conversion from CSV to BFF"),Ul("/conversion-from-vcf-to-bff",'Documentationa Beacon RI Tools v2 Conversion from VCF to BFF\nBeacon RI Tools v2\nConversion from VCF to BFF\nReading your VCF\nBeacon RI Tools v2 will read the different columns for your variants in the VCF and place them inside the Legacy Variation schema for the Beacon v2 Spec.\n\nFirst of all, Beacon RI Tools v2 will read the basic VCF parameters that come inside their columns:\nCHROM: will add the chromosome number in the genomicHGVSId field of the record.\nPOS: Will add the coordinates in start and end fields of the record in 0-typed annotation.\nID: will add the identification of the variant in the variantType field (SNP, INDEL, etc...)\nREF: will fill the referenceBases field in the record of the variant.\nALT: will fill the alternateBases field in the record of the variant.\n\nBy default, Beacon RI Tools v2 supports VEP annotation, in case your VCF has the designated VEP row in the VCF header:\n\n##INFO=<ID=CSQ,Number=.,Type=String,Description="Consequence annotations from Ensembl VEP. Format: Allele|Consequence|IMPACT|SYMBOL|Gene|Feature_type|Feature|BIOTYPE|EXON|INTRON|HGVSc|HGVSp|cDNA_position|CDS_position|Protein_position|Amino_acids|Codons|Existing_variation|REF_ALLELE|UPLOADED_ALLELE|DISTANCE|STRAND|FLAGS|SYMBOL_SOURCE|HGNC_ID|CANONICAL|HGVS_OFFSET">\n\nSome of the fields will be parsed into BFF. Right now, the fields that will be read are:\n\nSymbol \u2192 molecularAttributes|geneIds\nUploaded_Allele \u2192 variation|variantType\nHGVSp \u2192 molecularAttributes|aminoacidChanges\nConsequence \u2192 molecularAttributes|molecularEffects|label\nAdditionally, for filling in the required fields, the INFO column will read the next entries:\n\nVT \u2192 variation|variantType (in case VCF is not VEP annotated)\nAF \u2192 frequencyInPopulations| frequencies| alleleFrequency\nAN \u2192 frequencyInPopulations| frequencies| alleleNumber\nAC \u2192 frequencyInPopulations| frequencies| alleleCount\nAC_Hom \u2192 frequencyInPopulations| frequencies| alleleCountHomozygous\nAC_Het \u2192 frequencyInPopulations| frequencies| alleleCountHeterozygous\nEND \u2192 variation| location| interval| end| value (in case END column is not filled in)\nOn the other hand, if your VCF doesn\u2019t have VEP annotations or you want to use your own customized annotations, you can do that by editing the files that are located in this GitHub repository. The files that you have to modify are populations.json and template.json.\nThe populations.json will allow you to add how you annotated all the allele frequency related entries:\n\n{ \n  "numberOfPopulations": 1,\n  "source": "The Genome Aggregation Database (gnomAD)",\n  "sourceReference": "gnomad.broadinstitute.org/",\n  "populations": [{\n            "population": "Total",\n            "alleleFrequency": "AF",\n            "alleleCount": "AC",\n            "alleleCountHomozygous": "AC_hom",\n            "alleleCountHeterozygous": "AC_het",\n            "alleleNumber": "AN" }]\n  }\n\nTip: If numberOfPopulations is greater than 1, you have to add as many populations you have in the populations array, while if populations is 0, then, no allele frequency will be read from this pipeline.\n\nThe template.json file will allow you to map the annotations entries related to the variant type, the aminoacid change, the gene Id or the molecular effects in your vcf:\n\n{\n    "template": false,\n    "variantType": "VT",\n    "aminoacidChange": "HGVSp",\n    "geneId": "SYMBOL",\n    "molecularEffects": "CONSEQUENCE"\n}\n\nTip: If you want to activate this pipeline, change the template variable to true. If you activate this template, this will override the VEP annotations.\n\nPlease, keep in mind that multiallelic variants need to be split onto separate rows in the VCF.\n\nVariant data conversion\nIf you do not want to fill the CSV file for the genomicVariations collection or you already have your data in the VCF format, you can convert directly from VCF to BFF.\n\nTo convert data from a VCF file to BFF (json), the VCF must be compressed and indexed (.vcf.gz + .vcf.gz.tbi). Beacon RI Tools v2 will read all the VCF files inside the files_to_read folder. You can convert one or multiple VCF files at a time.\n\nTo execute the conversion, use the next command:\n\ndocker exec -it ri-tools python genomicVariations_vcf.py\n\nThis command will do the conversion from VCF to BFF and will load the final BFF documents into a mongoDB inside a container. This is done for memory size usage.\n\nAfter that, if needed, export your documents from the mongoDB to your machine as a BFF file (json) using two possible commands.\n\nThe first command will delete an internal "_id" for each record that is generated by MongoDB:\ndocker exec ri-tools-mongo mongoexport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --collection genomicVariations | sed \'/"_id":/s/"_id":[^,]*,//g\' > genomicVariations.json\n\n\nThe second command will keep the "_id" entries generated by MongoDB. Note that this ID is not part of the specifications of the Beacon and will not affect your data and Beacon, you can keep it if you want:\ndocker exec ri-tools-mongo mongoexport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --collection genomicVariations > genomicVariations.json\n\nThis will generate the final BFF file (json) for the genomicVariations collection using the VCF format as the source. Bear in mind that this time, the file will be saved in the directory you are located, so if you want to save it in the output_docs folder, add it in the path of the mongoexport (e.g. > output_docs/genomicVariations.json).\n\n\nIf you are using the ri-tools image directly into Beacon v2 PI API or Beacon v2 RI API you won\u2019t need to mongoexport the .json files unless you specifically need them for any other purpose, as the MongoDB instance for the API will already be filled with the variants.\n\nAs it has already been mentioned, the variants read from the VCF are directly stored into the Beacon RI Tools v2 MongoDB.\n\nIf you need to do more conversions and you don\u2019t want to keep the variants inside it, you can remove them by using the next command:\n\ndocker exec ri-tools-mongo /bin/bash -c \'mongo beacon -u root -p example --authenticationDatabase admin --eval "db.genomicVariations.deleteMany()"\'\n\nAt this point, you should have your data ready to be injected into the Beacon v2 RI API.\n\nCase level data conversion\nIf you are converting with the paramater case_level_data to True, this will add data into two collections: targets and caseLevelData. If you need to export the variants to insert them in another mongoDB, you will need to export these two collections as well, by executing the next commands:\n\ndocker exec ri-tools-mongo mongoexport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --collection caseLevelData > caseLevelData.json\n\ndocker exec ri-tools-mongo mongoexport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --collection targets > targets.json\n\nBear in mind that if you are converting VCFs one by one for the same samples and not by batches (multiple VCFs at the same time), this will generate a target file every time. We recommend converting all VCFs for the same sampling (dataset) at the same time to avoid unnecessary target records to be created.\n\nCase level data conversion is only applicable in the Beacon Production Implementation environment, as it is specifically designed to handle detailed sample-level associations and structured genomic data storage optimized for production use.',"Conversion from VCF to BFF"),Ul("/conversion-from-phenopackets-to-bff",'Documentation Beacon RI Tools v2 Conversion from Phenopackets to BFF\n\nBeacon RI Tools v2\nConversion from Phenopackets to BFF\n\n\nThis option will convert a Phenopacket v2 into Beacon Friendly Format Biosamples and Individuals models while preserving as much information as possible.\nFor detailed information about the mapping of properties between Phenopackets and Beacon schemas, please refer to this spreadsheet: Beacon + Phenopackets Schemas.\nMost of the mapping between the Phenopackets and Beacon models was straightforward. However, some fields in the Phenopacket schema did not have a direct match in the Beacon schema.\nFields containing important metadata that could not be mapped to a specific Beacon property were stored in the additionalInformation (notes) field to preserve as much information as possible.\nSpecific Mappings\nIndividuals.diseases.notes:\nThe following Phenopacket fields related to diseases were saved in the notes field:\nresolution\nprimary_site\nlaterality\nexcluded\nIndividuals.info:\nThe file.individual_to_file_identifiers field from Phenopackets is saved in the info property of the Beacon individual.\nBiosamples.notes:\nThe biosamples.description field is stored in the notes field of the Beacon biosample.\nHandling collectionData\nThe biosamples.timeOfCollection property in Phenopackets supports various data types, including:\ngestationalAge: Measure of the age of a pregnancy\nAge: Age as an ISO8601 duration (e.g., P40Y10M05D)\nAgeRange: Age within a given range\nOntologyClass: Age as an ontology class\nTimestamp: Specific date and time\nTimeInterval: Time interval\nIn contrast, the Beacon schema defines biosamples.collectionDate as the "Date of biosample collection in ISO8601 format" and expects a simple string.\nTo maximize data retention, the tool converts the collectionData from Phenopackets to a string and stores it in the collectionDate property of Beacon. While this approach does not fully align with the intended collectionDate field usage, it ensures valuable data is not lost.\nConverting data from phenopackets\nTo convert a phenopacket into Biosamples and Individuals schemas first save the phenopacket in this folder: phenopackets-to-BFF.\nAnd then run:\ndocker exec phenopackets-to-BFF python working-w-phenopackets.py /usr/src/app/examples/phenopacket.json\nBear in mind that the path /usr/src/app/examples must remain unchanged.\nYou\'ll find your BFFs in the same folder where you saved the phenopacket, phenopackets-to-BFF.',"Conversion from Phenopackets to BFF"),Ul("/deployment","Documentation > Beacon UI > Deployment\nBeacon UI\nDeployment\nUse the deployment documentation (https://github.com/EGA-archive/beacon2-ri-api/blob/master/deploy/README.md)\nfor all the containers for Beacon to also deploy the user interface.\nYou will find it running at *http://localhost:3000.\n","Deployment"),Ul("/configuration-ui",'Documentation > Beacon UI > Configuration UI\n\nBeacon UI\nConfiguration\n\nPlease first create a .env file inside the frontend folder so that you can modify some variables. \nTake into account that the file below will not be copied to GitHub as it contains keys and \nfor security reasons it should be ignored:\n\n```\nREACT_APP_CLIENT_ID="ID of your LS Login"\nREACT_APP_CLIENT_SECRET="password of your LS Login"\nREACT_APP_KEYCLOAK_CLIENT_SECRET="password of your Keycloak login"\n```\n\nTip: for Life Science environment, please first create a user (\n[create a user](https://lifescience-ri.eu/ls-login/users/how-to-get-and-use-life-science-id.html)). \nAfter that you will need to register a service registry in order to be able to administrate your logins. \nPlease go [here](https://services.aai.lifescience-ri.eu/spreg/) and ask for a New Service - type OIDC.\n\nThen please edit the file [config.json](https://github.com/EGA-archive/beacon2-ri-api/blob/master/frontend/src/config.json). \nYou need to decide where you want the UI to point to when making requests. \nFind below an example:\n\n```\n{\n  "API_URL": "https://yourAPIdomain.com/beacon-network/v2.0.0",\n  "REDIRECT_URL": "https://yourUIdomain.com",\n  "KEYCLOAK_URL": "https://yourKEYCLOAKdomain.com"\n}\n```\n\nFinally, please include the URL of your User interface to the file beacon/_main_.py (line 103), \nso that it becomes part of the list of URLs accepted by CORS.\n\n**Note**: In the frontend folder you will find a file called .gitignore with the list of all files that need to be ignored.',"Configuration UI"),Ul("/querying-the-ui","Documentation > Beacon UI > Querying the UI\n\nBeacon UI\nQueries\n\nComing soon...\n","Querying the UI"),Ul("/network-deployment","Please run the following command in order to build the Docker images as defined in the docker-compose.yml file and to start the containers in detached mode (running in the background). The UI will be \n\ndocker-compose up -d \u2013build","Network Deployment"),Ul("/network-configuration-ui",'\n\nPlease first create a .env file inside the frontend folder so that you can modify some variables. Take into account that the file below will not be copied to GitHub as it contains keys and for security reasons it should ignored:\n\n\n\n\nREACT_APP_CLIENT_ID=\'ID of your LS Login\'\nREACT_APP_CLIENT_SECRET=\'password of your LS Login\'\nREACT_APP_KEYCLOAK_CLIENT_SECRET=\'password of your Keycloak login\'\nREACT_APP_KEYCLOAK_CLIENT_ID=\'ID of your Keycloak login\'\nREACT_APP_KEYCLOAK_CLIENT_REALM=\'REALM of your Keycloak login\'\n\nYou will need to have created your Life Science and Keycloak environments before. \n\n\nNote: for Life Science environment, please first create a user . After that you will need to register a service registry in order to be able to administrate your logins. \nPlease go here and ask for a New Service - type OIDC -.\n\nThen please edit the file config.json, which can be found inside the folder frontend/src. You need to decide where you want the UI to point to when making requests. Find below an example:\n\n\n\n\n{\n  "API_URL": "https://yourAPIdomain.com/beacon-network/v2.0.0",\n  "REDIRECT_URL": "https://yourUIdomain.com",\n  "KEYCLOAK_URL": "https://yourKEYCLOAKdomain.com"\n}\n\nNote that in the root folder (frontend) you will find a file called .gitignore with the list of all files that need to be ignored.',"Network Configuration UI"),Ul("/network-querying-the-ui","Documentation > Beacon Network UI > Querying the UI\n\nNetwork UI\nQueries\nComing soon...\n","Network Querying the UI"),Ul("/use-case:-rare-diseases-data",'Documentationarrow rightTutorialsarrow rightUse Case: Rare Diseases Data\nUse Case : Inserting Rare Disease Synthetic Dataset into a beacon.\nIntroduction\nFor this example we are going to use the synthetic dataset, hosted at the European Genome-Phenome Archive, EGAD00001008392.\n\nThe Rare Disease Synthetic Dataset (EGAD00001008392), contains public human datasets for the study of rare diseases. The use of public human genomic background combined with the in-silico insertion of real disease-causing variants enable a representative dataset for testing purposes without facing ethical and legal issues associated with the use of human sensitive data. This project aims to help develop technical implementations for rare disease data integration, analysis, discovery, and federated access.\n\nThis dataset contains 18 samples and different data types (FASTQs, BAMs and VCFs). To simplify the example we\u2019ll only insert data from the sample EGAN00003364605, case1_index, from which we have the VCFs for the 24 chromosomes, a BAM and its index and a phenopacket with phenotypic and clinical information.\n\nFor this example we\u2019ll include into our beacon, the variants from chromosome 21 VCF, the phenopacket and all the rest of information available using CSV templates.\n\nObject Type\tIdentifier\nRare Disease Synthetic Dataset\tEGAD00001008392\ncase1_index sample\tEGAN00003364605\ncase1_index fastq\tEGAR00003021166\nChromosome 21 VCF\tEGAZ00001744017\nPhenopacket\tEGAZ00001744003\nTable 1. EGA identifiers for all the synthetic data used. \nFor further information about how to use the public metadata API please visit the EGA documentation.\n\n\nBefore continuing with this tutorial, please make sure you\'ve reviewed 10 steps to create your first beacon. In that tutorial you\'ll find important information that will help you follow this use case theoretically and technically.\n\nDataset model creation \nDataset model is described as a group of variants or individuals that share something in common, in this case that they belong to the same repository.\n\nLet\u2019s start checking which information we\u2019ve got and which one can be included in the dataset model.\n\nEGA metadata about EGAD00001008392:\n\nMetadata field\tContent\nstable_id\tEGAD00001008392\nTitle\tRare Disease Synthetic Dataset\nDescription\tThe purpose of this project is to provide public human datasets for the study of rare diseases. The use of public human genomic background combined with the in-silico insertion of real disease-causing variants enable a representative dataset for testing purposes without facing ethical and legal issues associated with the use of human sensitive data. This project aims to help develop technical implementations for rare disease data integration, analysis, discovery, and federated access.\nDataset types\t{"Genomic variant calling", "Phenotype information", "Whole genome sequencing"}\nTechnologies\t{"Illumina HiSeq 2000"}\nNumber of samples\t18\nCreated at\t2021-12-01 11:53:53.000 +0100\nEdited at\t2025-04-14 23:18:43.985 +0200\nUse conditions\tData may be used freely for non-commercial research and educational activities.\nNow, let\u2019s take a look at the dataset beacon model. \n\nThe mapping between our data and the beacon model would look like: \n\nDataset Beacon\nModel Term\tProperty\nType\tEGA Metadata\nfield\tContent\nid\tstring\tstable_id\tEGAD00001008392\nName\tstring\tTitle\tRare Disease Synthetic Dataset\nDescription\tstring\tDescription\tThe purpose of this project is to provide public human datasets for the study of rare diseases. The use of public human genomic background combined with the in-silico insertion of real disease-causing variants enable a representative dataset for testing purposes without facing ethical and legal issues associated with the use of human sensitive data. This project aims to help develop technical implementations for rare disease data integration, analysis, discovery, and federated access.\nX\t\tDataset types\t{"Genomic variant calling", "Phenotype information", "Whole genome sequencing"}\nX\t\tTechnologies\t{"Illumina HiSeq 2000"}\nX\t\tNumber of samples\t18\ncreateDateTime\tstring\tCreated at\t2021-12-01 11:53:53.000 +0100\nupdateDateTime\tstring\tEdited at\t2025-04-14 23:18:43.985 +0200\ndataUseConditions \tduoDataUse\tUse conditions\tData may be used freely for non-commercial research and educational activities.\nThe first thing we should check it\u2019s if we have the mandatory information for the Dataset model, which is id and name. As we have, we can now fill in the CSV with all the information. \n\nWe can see that most of the metadata and properties have a direct match, except the dataUseConditions property. \nIn our metadata we have string but the property type is duoDataUse, so we\u2019ll have to find the DUO code that means Data may be used freely for non-commercial research and educational activities. We\u2019ll choose:\n\nTerm\tLabel\tDefinition\tVersion\nDUO:0000018\tnot for profit, non commercial use only\tThis data use modifier indicates that use of the data is limited to not-for-profit organizations and not-for-profit use, non-commercial use.\t2021-02-23\nIn the beacon model the term duoDataUse has four different properties: \nid\nlabel\ndescription\nmodifiers\nversion\n\nTo correctly include our information to the CSV we\u2019ll need to use the columns: \ndataUseConditions|duoDataUse|id\ndataUseConditions|duoDataUse|label\ndataUseConditions|duoDataUse|description\ndataUseConditions|duoDataUse|version\n\nTake a look to our final CSV.\n\nAnalysis model creation\nNow, let\u2019s focus on the information we\u2019ve got about EGAZ00001744017.\n\nMetadata field\tContent\nstable_id\tEGAZ00001744017\nTitle\tcase1_index VCF chr21\nDescription\tvariants were called using GATK3.6\nAnalysis type\tSEQUENCE VARIATION\nAnalysis file type\tVCF\nExperiment type\t{"Whole genome sequencing"}\nGenome\tGRCh37\nChromosomes\tCM000683.1\ncreation date\t2021-10-22\nAnd let\u2019s try to match as much as possible to the analysis model:\n\nDataset Beacon\nModel Term\tProperty\nType\tEGA Metadata\nfield\tContent\nid\tstring\tstable_id\tEGAZ00001744017\nvariantCaller\tstring\tDescription\tvariants were called using GATK3.6\nrunId\tstring\trun\tEGAR00003021166\nanalysisDate\tstring\tcreation date\t2021-10-22\nLet\u2019s remember that to include Analysis model into beacon the required properties are: \nid\nanalysisDate\npipelineName\n\nUnfortunately, we don\u2019t have any information about the pipelineName used, so we won\u2019t be able to fill in the analysis model.\n\nRun model creation\nThis is the information we have available regarding the procedures used for sequencing a biosample.\n\nMetadata field\tContent\nbiosampleID\tSAMEA10450712\nExperiment type\tWhole genome sequencing\nrun creation date\t2021-10-22\nTo fill in Run model we need at least the following properties:\nid\nbiosampleId\nrunDate\n\nIt seems we\u2019ll have enough data to include Runs into our beacon instance:\n\nDataset Beacon\nModel Term\tProperty\nType\tEGA Metadata\nfield\tContent\nid\tstring\tstable_id\tEGAR00003021166\nbiosampleId\tstring\tbiosampleID\tSAMEA10450712\nrunDate\tstring\trun creation\t2021-10-22\nlibraryStrategy\tstring\tExperiment type\tWhole genome sequencing\nBut there are other terms from Runs that we can fill with the metadata we have available now, for example, from Dataset:\n\nDataset Beacon\nModel Term\tProperty\nType\tEGA Metadata\nfield\tContent\nplatform\tstring\tTechnology\tIllumina\nplatformModel\tid,label\tTechnology\tIllumina HiSeq 2000\nTake a look at the CSV with all this information.\n\nIndividual model from phenopacket\nIn EGAZ00001744003 we have a phenopacket with the following information: \nPlease save it to a file named Case1C.json.\n\n{\n  "id": "P0007498",\n  "subject": {\n    "id": "P0007498",\n    "dateOfBirth": "2013-01-01T00:00:00Z",\n    "sex": "MALE"\n  },\n  "phenotypicFeatures": [\n    {\n      "excluded": true,\n      "type": {\n        "id": "HP:0001249",\n        "label": "Intellectual disability"\n      }\n    },\n    {\n      "type": {\n        "id": "HP:0000467",\n        "label": "Neck muscle weakness"\n      }\n    },\n    {\n      "type": {\n        "id": "HP:0001252",\n        "label": "Muscular hypotonia"\n      }\n    },\n    {\n      "type": {\n        "id": "HP:0001319",\n        "label": "Neonatal hypotonia"\n      }\n    },\n    {\n      "type": {\n        "id": "HP:0001374",\n        "label": "Congenital hip dislocation"\n      }\n    },\n    {\n      "type": {\n        "id": "HP:0002540",\n        "label": "Inability to walk"\n      }\n    },\n    {\n      "type": {\n        "id": "HP:0002783",\n        "label": "Recurrent lower respiratory tract infections"\n      }\n    },\n    {\n      "type": {\n        "id": "HP:0002804",\n        "label": "Arthrogryposis multiplex congenita"\n      }\n    },\n    {\n      "type": {\n        "id": "HP:0003202",\n        "label": "Skeletal muscle atrophy"\n      }\n    },\n    {\n      "type": {\n        "id": "HP:0005684",\n        "label": "Distal arthrogryposis"\n      }\n    },\n    {\n      "type": {\n        "id": "HP:0030319",\n        "label": "Weakness of facial musculature"\n      }\n    }\n  ],\n  "diseases": [\n    {\n      "term": {\n        "id": "Orphanet:97245",\n        "label": "Congenital myopathy"\n      }\n    },\n    {\n      "term": {\n        "id": "OMIM:117000",\n        "label": "CENTRAL CORE DISEASE OF MUSCLE"\n      }\n    }\n  ],\n  "meta_data": {\n    "created": "2021-04-21T09:32:38.488Z",\n    "resources": [\n      {\n        "id": "hp",\n        "name": "Human Phenotype Ontology",\n        "url": "http://purl.obolibrary.org/obo/hp.owl",\n        "version": "2020-12-07",\n        "namespacePrefix": "HP",\n        "iriPrefix": "http://purl.obolibrary.org/obo/HP_"\n      },\n      {\n        "id": "orphanet",\n        "name": "Orphanet Rare Disease Ontology",\n        "url": "http://orpha.net/ontology/ORDO_en_3.1.owl",\n        "version": "3.1",\n        "namespacePrefix": "Orphanet",\n        "iriPrefix": "http://www.orpha.net/ORDO/Orphanet_"\n      },\n      {\n        "id": "hgnc",\n        "name": "HUGO Gene Nomenclature Committee",\n        "url": "https://www.genenames.org",\n        "version": "2021-01-13",\n        "namespacePrefix": "HGNC",\n        "iriPrefix": "https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/"\n      },\n      {\n        "id": "mim",\n        "name": "Online Mendelian Inheritance in Man",\n        "url": "https://omim.org/",\n        "version": "2021-01-21",\n        "namespacePrefix": "OMIM",\n        "iriPrefix": "https://omim.org/entry/"\n      }\n    ]\n  }\n}\n\nPhenopackets v2 and the beacon schemas are only compatible for the biosamples and individual models. \n\nIn this example, this phenopacket contains only information for individuals, because the three mandatory terms for biosamples (id, biosampleStatus, sampleOriginType) are not present.\n\ngenomicVariations model from VCF\nBelow, you\u2019ll find a subset of 5 variants from EGAZ00001744017.\nPlease, save the following information into a Case1C_subset.vcf file:\n\n##fileformat=VCFv4.2\n##ALT=<ID=NON_REF,Description="Represents any possible alternative allele at this location">\n##FILTER=<ID=LowQual,Description="Low quality">\n##FORMAT=<ID=AD,Number=R,Type=Integer,Description="Allelic depths for the ref and alt alleles in the order listed">\n##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Approximate read depth (reads with MQ=255 or with bad mates are filtered)">\n##FORMAT=<ID=GQ,Number=1,Type=Integer,Description="Genotype Quality">\n##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">\n##FORMAT=<ID=MIN_DP,Number=1,Type=Integer,Description="Minimum DP observed within the GVCF block">\n##FORMAT=<ID=MQ0,Number=1,Type=Integer,Description="Number of Mapping Quality Zero Reads per sample">\n##FORMAT=<ID=PGT,Number=1,Type=String,Description="Physical phasing haplotype information, describing how the alternate alleles are phased in relation to one another">\n##FORMAT=<ID=PID,Number=1,Type=String,Description="Physical phasing ID information, where each unique ID within a given sample (but not across samples) connects records within a phasing group">\n##FORMAT=<ID=PL,Number=G,Type=Integer,Description="Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification">\n##FORMAT=<ID=SB,Number=4,Type=Integer,Description="Per-sample component statistics which comprise the Fisher\'s Exact Test to detect strand bias.">\n##GATKCommandLine.HaplotypeCaller=<ID=HaplotypeCaller,Version=3.6-0-g89b7209,Date="Thu Mar 23 15:53:02 CET 2017",Epoch=1490280782918,CommandLineOptions="analysis_type=HaplotypeCaller input_file=[/scratch/production/slaurie/PlatinumTrioForRD-Connect/HapMapTrio/gVCF/Case1C/E012882.bam] showFullBamList=false read_buffer_size=null read_filter=[BadCigar] disable_read_filter=[] intervals=[21] excludeIntervals=null interval_set_rule=UNION interval_merging=ALL interval_padding=0 reference_sequence=/project/production/Indexes/samtools/hsapiens.hs37d5.fasta nonDeterministicRandomSeed=false disableDithering=false maxRuntime=-1 maxRuntimeUnits=MINUTES downsampling_type=NONE downsample_to_fraction=null downsample_to_coverage=null baq=OFF baqGapOpenPenalty=40.0 refactor_NDN_cigar_string=false fix_misencoded_quality_scores=false allow_potentially_misencoded_quality_scores=false useOriginalQualities=false defaultBaseQualities=-1 performanceLog=null BQSR=null quantize_quals=0 static_quantized_quals=null round_down_quantized=false disable_indel_quals=false emit_original_quals=false preserve_qscores_less_than=6 globalQScorePrior=-1.0 validation_strictness=SILENT remove_program_records=false keep_program_records=false sample_rename_mapping_file=null unsafe=null disable_auto_index_creation_and_locking_when_reading_rods=false no_cmdline_in_header=false sites_only=false never_trim_vcf_format_field=true bcf=false bam_compression=null simplifyBAM=false disable_bam_indexing=false generate_md5=false num_threads=1 num_cpu_threads_per_data_thread=1 num_io_threads=0 monitorThreadEfficiency=false num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false generateShadowBCF=false variant_index_type=LINEAR variant_index_parameter=128000 reference_window_stop=0 phone_home= gatk_key=null tag=NA logging_level=INFO log_to_file=null help=false version=false likelihoodCalculationEngine=PairHMM heterogeneousKmerSizeResolution=COMBO_MIN dbsnp=(RodBinding name= source=UNBOUND) dontTrimActiveRegions=false maxDiscARExtension=25 maxGGAARExtension=300 paddingAroundIndels=150 paddingAroundSNPs=20 comp=[] annotation=[AlleleBalance, BaseCounts, BaseQualityRankSumTest, ChromosomeCounts, ClippingRankSumTest, Coverage, DepthPerAlleleBySample, DepthPerSampleHC, FisherStrand, GCContent, HaplotypeScore, HardyWeinberg, HomopolymerRun, ClippingRankSumTest, LikelihoodRankSumTest, LowMQ, MappingQualityRankSumTest, MappingQualityZero, MappingQualityZeroBySample, NBaseCount, QualByDepth, RMSMappingQuality, ReadPosRankSumTest, StrandBiasBySample, StrandOddsRatio, VariantType, StrandBiasBySample] excludeAnnotation=[ChromosomeCounts, FisherStrand, StrandOddsRatio, QualByDepth] group=[StandardAnnotation, StandardHCAnnotation] debug=false useFilteredReadsForAnnotations=false emitRefConfidence=GVCF bamOutput=null bamWriterType=CALLED_HAPLOTYPES emitDroppedReads=false disableOptimizations=false annotateNDA=false heterozygosity=0.001 indel_heterozygosity=1.25E-4 standard_min_confidence_threshold_for_calling=-0.0 standard_min_confidence_threshold_for_emitting=-0.0 max_alternate_alleles=6 max_num_PL_values=100 input_prior=[] sample_ploidy=2 genotyping_mode=DISCOVERY alleles=(RodBinding name= source=UNBOUND) contamination_fraction_to_filter=0.0 contamination_fraction_per_sample_file=null p_nonref_model=null exactcallslog=null output_mode=EMIT_VARIANTS_ONLY allSitePLs=true gcpHMM=10 pair_hmm_implementation=VECTOR_LOGLESS_CACHING pair_hmm_sub_implementation=ENABLE_ALL always_load_vector_logless_PairHMM_lib=false phredScaledGlobalReadMismappingRate=45 noFpga=false sample_name=null kmerSize=[10, 25] dontIncreaseKmerSizesForCycles=false allowNonUniqueKmersInRef=false numPruningSamples=1 recoverDanglingHeads=false doNotRecoverDanglingBranches=false minDanglingBranchLength=4 consensus=false maxNumHaplotypesInPopulation=128 errorCorrectKmers=false minPruning=2 debugGraphTransformations=false allowCyclesInKmerGraphToGeneratePaths=false graphOutput=null kmerLengthForReadErrorCorrection=25 minObservationsForKmerToBeSolid=20 GVCFGQBands=[20, 25, 30, 35, 40, 45, 50, 70, 90, 99] indelSizeToEliminateInRefModel=10 min_base_quality_score=10 includeUmappedReads=false useAllelesTrigger=false doNotRunPhysicalPhasing=false keepRG=null justDetermineActiveRegions=false dontGenotype=false dontUseSoftClippedBases=false captureAssemblyFailureBAM=false errorCorrectReads=false pcr_indel_model=CONSERVATIVE maxReadsInRegionPerSample=10000 minReadsPerAlignmentStart=10 mergeVariantsViaLD=false activityProfileOut=null activeRegionOut=null activeRegionIn=null activeRegionExtension=null forceActive=false activeRegionMaxSize=null bandPassSigma=null maxProbPropagationDistance=50 activeProbabilityThreshold=0.002 min_mapping_quality_score=20 filter_reads_with_N_cigar=false filter_mismatching_base_and_quals=false filter_bases_not_stored=false">\n##GVCFBlock0-20=minGQ=0(inclusive),maxGQ=20(exclusive)\n##GVCFBlock20-25=minGQ=20(inclusive),maxGQ=25(exclusive)\n##GVCFBlock25-30=minGQ=25(inclusive),maxGQ=30(exclusive)\n##GVCFBlock30-35=minGQ=30(inclusive),maxGQ=35(exclusive)\n##GVCFBlock35-40=minGQ=35(inclusive),maxGQ=40(exclusive)\n##GVCFBlock40-45=minGQ=40(inclusive),maxGQ=45(exclusive)\n##GVCFBlock45-50=minGQ=45(inclusive),maxGQ=50(exclusive)\n##GVCFBlock50-70=minGQ=50(inclusive),maxGQ=70(exclusive)\n##GVCFBlock70-90=minGQ=70(inclusive),maxGQ=90(exclusive)\n##GVCFBlock90-99=minGQ=90(inclusive),maxGQ=99(exclusive)\n##GVCFBlock99-2147483647=minGQ=99(inclusive),maxGQ=2147483647(exclusive)\n##INFO=<ID=ABHet,Number=1,Type=Float,Description="Allele Balance for heterozygous calls (ref/(ref+alt))">\n##INFO=<ID=ABHom,Number=1,Type=Float,Description="Allele Balance for homozygous calls (A/(A+O)) where A is the allele (ref or alt) and O is anything other">\n##INFO=<ID=BaseCounts,Number=4,Type=Integer,Description="Counts of each base">\n##INFO=<ID=BaseQRankSum,Number=1,Type=Float,Description="Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities">\n##INFO=<ID=ClippingRankSum,Number=1,Type=Float,Description="Z-score From Wilcoxon rank sum test of Alt vs. Ref number of hard clipped bases">\n##INFO=<ID=DP,Number=1,Type=Integer,Description="Approximate read depth; some reads may have been filtered">\n##INFO=<ID=DS,Number=0,Type=Flag,Description="Were any of the samples downsampled?">\n##INFO=<ID=END,Number=1,Type=Integer,Description="Stop position of the interval">\n##INFO=<ID=ExcessHet,Number=1,Type=Float,Description="Phred-scaled p-value for exact test of excess heterozygosity">\n##INFO=<ID=GC,Number=1,Type=Float,Description="GC content around the variant (see docs for window size details)">\n##INFO=<ID=HRun,Number=1,Type=Integer,Description="Largest Contiguous Homopolymer Run of Variant Allele In Either Direction">\n##INFO=<ID=HW,Number=1,Type=Float,Description="Phred-scaled p-value for Hardy-Weinberg violation">\n##INFO=<ID=HaplotypeScore,Number=1,Type=Float,Description="Consistency of the site with at most two segregating haplotypes">\n##INFO=<ID=InbreedingCoeff,Number=1,Type=Float,Description="Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation">\n##INFO=<ID=LikelihoodRankSum,Number=1,Type=Float,Description="Z-score from Wilcoxon rank sum test of Alt Vs. Ref haplotype likelihoods">\n##INFO=<ID=LowMQ,Number=3,Type=Float,Description="3-tuple: <fraction of reads with MQ=0>,<fraction of reads with MQ<=10>,<total number of reads>">\n##INFO=<ID=MLEAC,Number=A,Type=Integer,Description="Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed">\n##INFO=<ID=MLEAF,Number=A,Type=Float,Description="Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed">\n##INFO=<ID=MQ,Number=1,Type=Float,Description="RMS Mapping Quality">\n##INFO=<ID=MQ0,Number=1,Type=Integer,Description="Total Mapping Quality Zero Reads">\n##INFO=<ID=MQRankSum,Number=1,Type=Float,Description="Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities">\n##INFO=<ID=OND,Number=1,Type=Float,Description="Overall non-diploid ratio (alleles/(alleles+non-alleles))">\n##INFO=<ID=PercentNBase,Number=1,Type=Float,Description="Percentage of N bases in the pileup">\n##INFO=<ID=RAW_MQ,Number=1,Type=Float,Description="Raw data for RMS Mapping Quality">\n##INFO=<ID=ReadPosRankSum,Number=1,Type=Float,Description="Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias">\n##INFO=<ID=VariantType,Number=1,Type=String,Description="Variant type description">\n##contig=<ID=21,length=48129895>\n##reference=file:///project/production/Indexes/samtools/hsapiens.hs37d5.fasta\n#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tCase1C\n21\t9411635\t.\tC\tT\t49.77\t.\tBaseQRankSum=0.420;ClippingRankSum=0.000;DP=21;ExcessHet=3.0103;LikelihoodRankSum=-0.537;MLEAC=1,0;MLEAF=0.500,0.00;MQ0=0;MQRankSum=-1.786;RAW_MQ=42918.00;ReadPosRankSum=1.162\tGT:AD:DP:GQ:PL:SB\t0/1:17,4,0:21:78:78,0,544,129,556,685:8,9,4,0\n21\t9411732\t.\tT\tG\t823.77\t.\tDP=25;ExcessHet=3.0103;MLEAC=2,0;MLEAF=1.00,0.00;MQ0=0;RAW_MQ=67833.00\tGT:AD:DP:GQ:PL:SB\t1/1:0,25,0:25:75:852,75,0,852,75,852:0,0,14,11\n21\t9411931\t.\tG\tA\t104.77\t.\tBaseQRankSum=-0.873;ClippingRankSum=0.000;DP=27;ExcessHet=3.0103;LikelihoodRankSum=-0.375;MLEAC=1,0;MLEAF=0.500,0.00;MQ0=0;MQRankSum=-3.169;RAW_MQ=65235.00;ReadPosRankSum=0.657\tGT:AD:DP:GQ:PGT:PID:PL:SB\t0/1:22,5,0:27:99:0|1:9411931_G_A:133,0,874,199,889,1087:11,11,0,5\n21\t9411949\t.\tT\tC\t55.77\t.\tBaseQRankSum=1.914;ClippingRankSum=0.000;DP=29;ExcessHet=3.0103;LikelihoodRankSum=-0.063;MLEAC=1,0;MLEAF=0.500,0.00;MQ0=0;MQRankSum=-3.040;RAW_MQ=73991.00;ReadPosRankSum=-1.787\tGT:AD:DP:GQ:PGT:PID:PL:SB\t0/1:25,4,0:29:84:0|1:9411931_G_A:84,0,1132,168,1144,1312:12,13,0,4\n21\t9411965\t.\tT\tC\t0.01\t.\tBaseQRankSum=0.337;ClippingRankSum=0.000;DP=30;ExcessHet=3.0103;LikelihoodRankSum=-1.995;MLEAC=0,0;MLEAF=0.00,0.00;MQ0=0;MQRankSum=-2.589;RAW_MQ=90291.00;ReadPosRankSum=-2.080\tGT:AD:DP:GQ:PGT:PID:PL:SB\t0/0:28,2,0:30:0:0|1:9411931_G_A:0,0,1105,84,1111,1195:11,17,0,2\n\nAnd then, compress it:\n\nbgzip Case1C_subset.vcf\n\nConversion of CSV, Penopacket and VCF to BFF\nAt this point we have information ready to create 4 beacon models: \ndataset.csv\nrun.csv\nphenopacket v2\nVCF for genomicVariations\n\nNow, let\u2019s convert them to Beacon Friendly Format (BFF).\n\nInstalling Beacon 2 Production Implementation API and beacon2-ri-tools-v2\nBefore starting, you should have installed:\nDocker\nDocker Compose\n\nFirst, let\u2019s clone the beacon production implementation repository in your system:\n\ngit clone https://github.com/EGA-archive/beacon2-pi-api.git\n\nMake sure the next list of ports are free of use in your system:\n\n27017 \u2192 mongo\n5050 \u2192 beacon\nAnd now let\u2019s build all the necessary containers:\n\ndocker compose up -d --build \n\nBy cloning and building the Beacon 2 Production Implementation repository we have downloaded the tools that will deploy our beacon and the beacon2-ri-tools-v2, a toolkit that will allow you to populate your own beacon in the easiest way possible, by using CSVs and directly reading variant information from a VCF.\n\nLet\u2019s check if the building was correct. Run the following command:\n\ndocker ps\n\nThis command will show all the running containers in your computer. You should have built: \n\nImage\tNames\nghcr.io/ega-archive/beacon2-ri-tools-v2:latest\n\nri-tools\n\nghcr.io/ega-archive/phenopackets-to-bff:latest\n\nphenopackets-to-BFF\n\nmongo-express\n\nbeacon2-pi-api-mongo-express-1\n\nghcr.io/ega-archive/beacon2-ri-postgres-v2:latest\n\nidp-db\n\nbeacon2-pi-api-beaconprod\n\nbeaconprod\n\nSetting up conf.py\nThis will be the content that you find in beacon2-ri-tools-v2/conf/conf.py:\n\n#### Input and Output files config parameters ####\ncsv_folder = \'./csv/use_case_RD/\'\noutput_docs_folder=\'./output_docs/use_case_RD_output/\'\n\n#### VCF Conversion config parameters ####\nallele_counts=False # Variable still in test, leave it as False for now.\nreference_genome=\'GRCh37\' # Choose one between NCBI36, GRCh37, GRCh38\ndatasetId=\'EGAD00001008392\'\ncase_level_data=False\nexact_heterozygosity=False\nnum_rows=5\nverbosity=False # This variable, if True, will make the program run slower but give logs about all the skipped variants and the reason why.\n\nLet\u2019s start at the beginning of the file with the parameters for the CSV conversion.\n\nCreate a folder named use_case_RD in ./csv/  and move the dataset.csv and  run.csv there, this will be the CSV converted to BFF.  Please, make sure to end the path with a slash (/). \nCreate another folder where the output BFF will be saved. Let\u2019s name it use_case_RD_output inside the folder./output_docs/. \nLet\u2019s continue with the parameters for the VCF. \n\nFrom the header of the VCF we know that the variants belong to GRCh37. \ndatasetId must match the ID in the dataset.csv for the mapping of the variant and metadata to be correct. So, in our case we need to fill in datasetID as EGAD00001008392. \nThe case_level_data is a boolean parameter (True or False) which will relate your variants to the samples they belong to. As we don\u2019t have the biosamples schema, let\u2019s set it as False. \nThe exact_heterozygosity is a boolean parameter (True or False) that, in case case_level_data is True, then, it will classify the biosamples as being heterozygous for either the reference or the alternate allele. As in our case case_level_data is false, let\u2019s set exact_heterozygosity as False too. \nFill in num_rows with a 5, as we only have 5 variants in our VCF.\nConversion from CSV\nAs we have already moved our CSVs to the ./csv/use_case_RD/ we only need to run: \n\ndocker exec -it ri-tools python convert_csvTObff.py\n\nIn the folder ./output_docs/use_case_RD_output/ will find two json. These files contain our run and dataset information in Beacon Friendly Format. \n\nConversion from Phenopacket\nLet\u2019s save the Case1C_phenopacket.json in./beacon2-ri-tools-v2/phenopackets-to-BFF folder and then run:\n\ndocker exec phenopackets-to-BFF python working-w-phenopackets.py /usr/src/app/examples/case1c.json EGAD00001008392\n\n\nBear in mind that the path cannot be changed. Replace only the name of the phenopacket, once it has been copied in phenopackets-to-BFF folder.\n\nMake sure to add the same datasetID as in the dataset model. This will allow the correct link between the phenopackets information and the dataset it belongs to.\n\nIf everything works as expected your terminal output should look like: \n\nThe mandatory fields for BioSamples were not present in the phenopacket \nMandatory properties for Individual schema (id and sex) present in the phenopacket.\n-> Creating Individuals schema ...\n    - diseases added to Individuals\n    - phenotypicFeatures added to Individuals\n-> Creating Individuals schema - DONE\n+ BFFs Individuals JSON saved in:  \n/beacon2-pi-api/ri-tools/phenopackets-to-BFF/case1c.json\n\nYou\u2019ll find your phenopacket in  phenopackets-to-BFF.\n\nConversion from VCF\nFirst, let\u2019s move the VCF to the correct folder: \n\n./beacon2-pi-api/ri-tools/files/vcf/files_to_read/case1C_subset.vcf.gz\n\nAnd now, run the conversion command from VCF to BFF: \n\ndocker exec -it ri-tools python genomicVariations_vcf.py\n\nIf everything ran correctly, this will be your output: \n\nfiles/vcf/files_to_read/case1C_subset.vcf.gz\n100% | 5/5 [00:00<00:00, 6603.12it/s]\nSuccessfully inserted 5 records into beacon\nA total of 5 variants were processed\nA total of 0 variants were skipped\n\n\nIf the conversion of the variants is not working as expected modify the conf.py and set verbosity = True, this will give you information about which variants are being included and why.\n\nThe VCF was converted into BFF and saved in a MongoDB. Let\u2019s extract the variants to insert them into our beacon:\n\ndocker exec ri-tools-mongo mongoexport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --collection genomicVariations > ri-tools/output_docs/genomicVariations.json\n\nModify the output path so it works on your computer. In this case, we\u2019ve saved it with the BFFs converted from CSV.\n\nDeploying a Beacon 2 production implementation API\nNow that we have all the data ready let\u2019s start with the deployment. \nFirst, we\u2019ll make three important steps for the configuration of the beacon: \nCreate our own landing page\nAdd the right permission for our dataset \nMove our BFF to the correct folder\n\nTo know about further settings, please visit the Configuration section.\n\nPersonalize info entry point\nThe info entry point will be the landing page of your beacon, found at http://localhost:5050/api.\n\nTo modify the information shown you\u2019ll need to update the ./beacon2-pi-api/beacon/conf/conf.py\nOur conf.py would look something like:\n\nbeacon_id = \'org.ega-archive.beacon-ri-demo\'\n\nbeacon_name = \'Beacon Reference Implementation: Rare Diseases Use Case\u2019\n\napi_version = \'v2.0.0\'\n\nuri = \'https://beacon-apis-demo.ega-archive.org/api/\'\n\norg_id = \'EGA\'\n\norg_name = \'European Genome-Phenome Archive (EGA)\'\n\norg_description = \'The European Genome-phenome Archive (EGA) is a service for permanent archiving and sharing of all types of personally identifiable genetic and phenotypic data resulting from biomedical research projects.\'\n\norg_adress = \'C/ Dr. Aiguader, 88\nPRBB Building\n08003 Barcelona, Spain\'\n\norg_welcome_url = \'https://ega-archive.org/\'\n\norg_contact_url = \'mailto:beacon.ega@crg.eu\'\n\norg_logo_url = \'https://legacy.ega-archive.org/images/logo.png\'\n\norg_info = \'\'\n\ndescription = "This Beacon is based on synthetic data hosted at the EGA. It contains information from one sample from EGAD00001008392"\n\nversion = \'v2.0\'\n\nwelcome_url = \'https://ega-archive.org/datasets/EGAD00001008392\'\n\nalternative_url = \'https://beacon-apis-demo.ega-archive.org/api/\'\n\ncreate_datetime = \'2025-05-12T12:00:00.000000\'\n\nupdate_datetime = \'\'\n\nManaging dataset permissions\nThere are 3 possible levels of beacon security for a dataset: public, registered and controlled.\nA public dataset is a dataset that will be returned in a beacon query without an authentication token.\nA registered dataset is a dataset that will be shown after a user sends a valid token (in other words, is logged in).\nA controlled dataset is a dataset that needs a user to send a valid token for authentication and the user needs to be allowed to query that dataset.\n\nAs our data is synthetic and has no control access demand, we\u2019ll add our dataset as a public dataset into the beacon. \n\nFor this, first we\u2019ll need to edit the .yml inside ./beacon2-pi-api/beacon/permissions/datasets. As we have decided that EGAD00001008392 will be of public access let\u2019s edit public_datasets.yml:\n\npublic_datasets:\n- EGAD00001008392\n\nMove the BFF for the insertion\nLet\u2019s prepare the data for the insertion. Edit the following Makefile:\n/beacon/connections/mongo/Makefile\n\nAnd point the paths to your BFF files.\n\n\nBear in mind that if you don\u2019t modify the commands in the Makefile the files must be named as: \nanalyses.json\nbiosamples.json\ngenomicVariations.json\nindividuals.json\nruns.json\ncohorts.json\ndatasets.json\n\nDeploy a beacon instance\nFinally, simply run:\n\nbash mongostart.sh\n\nIf the operation is successful, you will have your beacon up and running at http://localhost:5050/api.\n\nRelationship-elements\nFiltering Terms\nFiltering terms are metadata fields that allow users to query your Beacon more precisely \u2014 for example, by filtering results based on sequencing strategy (libraryStrategy), tissue type, disease, or other structured attributes. These terms often rely on ontologies to ensure consistent vocabulary across datasets.\n\nAs we have some ontology fields, let\u2019s create the filtering terms for our beacon. We\u2019ll populate the /filteringTerms endpoint automatically by running:\n\ndocker exec beaconprod python beacon/connections/mongo/extract_filtering_terms.py\n\n\n If you want to add new filtering terms manually or enhance ontology filtering with descendants and semantic similarities, please visit the Filtering Terms section.\n\nAcknowledgments\nWe acknowledge the use of data from the RD-Connect Genome-Phenome Analysis Platform. This work was supported by the European Commission H2020 projects EJP-RD (grant number 825575) and B1MG (grant number 951724), as well as the Generalitat de Catalunya VEIS project (grant number 001-P-001647).\n\n',"Use Case: Rare Disease Data"),Ul("/create-your-beacon",'Documentationarrow rightTutorialsarrow rightCreate your Beacon\n10 steps to create your first beacon\nIntroduction\nThis guide will walk you through setting up a GA4GH Beacon v2 instance using the Production Implementation API. You\u2019ll deploy your own Beacon, load example data in Beacon Friendly Format (BFF), and explore how to query it through the API.\n\nAs defined by the Global Alliance for Genomics and Health (GA4GH), a Beacon v2 is an open standard for the discovery of genomic (and phenotypic and clinical) data in biomedical research and clinical applications. Its main goal is to facilitate the discovery - and potential retrieval - of genomic variants and biomedical data from different organisational and geographic locations.\n\nIf you are here it means you are interested in creating your own beacon!\n\n\nThis will be the steps you\'ll need to follow:\n\nInstalling Beacon 2 Production Implementation API\nConverting VCF to Genomic Variations Model\nFilling in the CSV Templates\nConfiguring conf.py\nConverting CSV/VCF to BFF\nDeploying your beacon instance\n1. Installing Beacon 2 Production Implementation API and beacon2-ri-tools-v2\nBefore getting started, make sure you have the following:\n\nDocker\nDocker Compose\ngit\nThis tutorial uses Beacon Friendly Format (BFF) files \u2013 standardized JSON documents that describe the beacon models. To create the BFF in the easiest way possible (directly from CSV, VCF or phenopacket) we\u2019ll use the Beacon2 ri tools v2, included in the beacon2-pi-api repository. \n\nFirst, let\u2019s clone the beacon production implementation repository in your system:\n\ngit clone https://github.com/EGA-archive/beacon2-pi-api.git\n\n\nEnsure port 21017 and 5050 are free for MongoDb and Beacon.\n\nAnd now let\u2019s build all the necessary containers: \n\ndocker compose up -d --build \n\nLet\u2019s check if the building was correct. Run the following command:\n\ndocker ps\n\nThis command will show all the running containers in your computer. You should have built: \n\nImage\tNames\nghcr.io/ega-archive/beacon2-ri-tools-v2:latest\n\nri-tools\n\nghcr.io/ega-archive/phenopackets-to-bff:latest\n\nphenopackets-to-BFF\n\nmongo-express\n\nbeacon2-pi-api-mongo-express-1\n\nghcr.io/ega-archive/beacon2-ri-postgres-v2:latest\n\nidp-db\n\nbeacon2-pi-api-beaconprod\n\nbeaconprod\n\n2. Understanding Beacon Data Models\nThere are several ways of populating a beacon. To introduce information into a beacon you need to populate one, some, or all of the beacon model entities.\n\nLet\u2019s recapitulate about what model entities the beacon allows: \n\nCollections (Datasets and Cohorts): groupings of variants or individuals that share something in common: e.g., who belong to the same repository (datasets) or study populations (cohorts).\nGenomic variations: unique genomic alterations, e.g., position in a genome, sequence alterations, type, etc. This model can be filled in directly with a VCF or creating a CSV.\nIndividuals: either patients or healthy controls whose details (including phenotypic and clinical) are stored in the repository.\nBiosamples: samples taken from individuals, including details of procedures, dates and times.\nAnalyses & Runs: details on (a) procedures used for sequencing a biosample (runs), and (b) bioinformatic procedures to identify variants (analyses).\n\nNote that none of these models are mandatory for a beacon to be functional, but the dataset model is mandatory for the API to be queryable. All the beacons will respond with boolean, count and full response. But the response type, hit, miss, all and none  will be available if the dataset model was included in the beacon. This is what these responses mean:\nHit: it returns the positive results of the query per dataset. \nMiss: it returns the datasets where the query did not have a hit. \nAll: it returns all the datasets in the beacon. \nNone: aggregated results (not splitted by dataset)\nIf you are interested in using them, make sure to include the datasets model into your beacon instance.\n\n\n3. Converting VCF to Genomic Variations Model beacon2-ri-tools-v2\nThe most simple way of converting variant information into the genomicVariations model is directly reading a VCF into Beacon Friendly Format (BFF). To convert directly from a VCF, copy all the files into the files_to_read folder and follow the instructions below to correctly fill in the conf.py.\n\n\nBear in mind that if your variant information is not in a VCF format you can also fill in the Genomic Variations CSV.\n\n4. Filling in the CSV Templates\nNow, let\u2019s see how to fill in the CSV that will be converted by the beacon r.i tools into the models.\n\nHere you\u2019ll find the CSV templates for all the models.\n\nTo correctly fill in all the properties you have two sources of information:\n\nThe JSON references schemas which will be really useful if you are used to working with JSON schemas.\nThe GA4GH Beacon v2 Documentation:\nCohorts\nDatasets\nBiosamples\ngenomicVariations\nIndividuals\nRuns\nAnalyses\nBeacon v2 aims to be a flexible tool, thus, again not all the fields are mandatory to be filled in. These are the fields that must be included in each model if you decide to include it:\n\nModel \u2192 Required fields\nAnalyses \u2192 id, analysisDate, pipelineName\nBiosamples \u2192 id, biosampleStatus, sampleOriginType\nCohorts \u2192 id, name, cohortType\nDatasets \u2192 id, name\nIndividuals \u2192 id, sex\nRuns \u2192 id, biosampleId, runDate\ngenomic Variations \u2192 variantInternalId, variation\n5. CSV Formatting Tips and Rules\nHere are some things to consider during your CSV creation:\n\nOnly \u201ccorrectly spelled\u201d fields (aka columns) will be inserted into the beacon. You can delete unused columns of your CSV, but don\u2019t modify the headers of the used ones.\nEvery new row will be appended to the final output as a new and independent document. e.g. If in dataset.csv you have two rows, you will be creating two independent datasets.\nTo add multiple information in the same row separate it by a pipe \'|\'. For instance, in genomicVariations, if several samples have the same variant you will need to fill in caseLevelData properties as below. Note that you can leave the pipe empty or write the condition multiple times. e.g:\n\ncaseLevelData | biosampleId      caseLevelData | phenotypicEffects | conditionId\nSAMPLE2 | SAMPLE3 | SAMPLE4      Alzheimer | |\n\nIn this case, this would mean that the variant in that row was found in sample 2, 3, and 4 and it\u2019s associated to Alzheimer.\nThe info field for each collection is very generic and can be filled with different data. You will need to fill the column data directly with JSON-type data. e.g:\n\n{"info": "This dataset contains synthetic metadata"}\n\nThe id field in Biosamples must match the sample IDs in the genomicVariations/VCF model. If these IDs don\u2019t match, the mapping with the variants and the biosamples won\u2019t be correctly performed.\nThe id in Individuals can match or not the ID in genomicVariations/VCF. If it doesn\u2019t match, for the mapping between individuals and variants to be correctly performed, the field individualsId must be filled in the Biosamples model, mapping the IDs in Individuals and the IDs in genomicVariations/VCF.\nThe datasetId in Datasets must match the datasetId field in the conf.py (this requirement applies only in the Beacon Production Implementation environment).\n6. Configuring conf.py\nNow that you\u2019ve already gathered all the necessary information, let\u2019s convert it into Beacon Friendly Format (BFF) using the beacon r.i tools.\n\nFirst, let\u2019s edit the ./beacon2-pi-api/ri-tools/conf/conf.py:\n\n#### Input and Output files config parameters ####\n\ncsv_folder = \'./csv/examples/\'\noutput_docs_folder=\'./output_docs/\'\n\n#### VCF Conversion config parameters ####\n\nallele_counts=False\nreference_genome=\'GRCh37\' # Choose one between NCBI36, GRCh37, GRCh38\ndatasetId=\'COVID_pop11_fin_2\'\ncase_level_data=False\nexact_heterozygosity=False\nnum_rows=1500000\nverbosity=False\n\nHow to fill in config.py?\nThe csv_folder variable sets the path of a folder containing all the CSVs that will be converted to BFF. Please, don\u2019t modify the path ./csv/.\nThe output_docs_folder variable sets the folder where your final .json files (BFF format) will be saved once execution of beacon tools finishes. This folder should always be located within the folder \'output_docs\', but subdirectories can be created inside it. e.g output_docs_folder=\'./output_docs/test1\'\nIf you\u2019ve chosen to read variants directly from a VCF instead of filling in the genomicVariations CSV you\u2019ll need to fill in the VCF Conversion config parameters:\n\nThe reference_genome is the reference genome the tool will use to map the position of the chromosomes.\nMake sure to select the same version as the one used to generate your data.\nThe datasetId needs to match the id of your datasets.csv or datasets.json file.\nThis will add a datasetId field in every record to match the record with the dataset it belongs to.\nThe case_level_data is a boolean parameter ( True or False) which will relate your variants to the samples they belong to.\nAs we don\u2019t have the biosamples schema, this mapping won\u2019t work. Let\u2019s set it up as False.\nThe exact_heterozygosity is a boolean parameter ( True or False) that, in case case_level_data is True, then it will classify the biosamples as being heterozygous for either the reference or the alternate allele.\nIn our case, as case_level_data is False, let\u2019s set this parameter as False too.\nThe num_rows are the approximate calculation you expect for the total of variants in each VCF there are.\nMake sure this is greater than the total variants expected.\nThe verbosity will give streaming logs with the reason why a variant has been skipped to be inserted.\nRecommendation is to leave this as False.\n7. Converting CSV/VCF to BFF\nTime for the conversion! \nBefore preceding ensure that: \nAll the CSVs that you\u2019ve created have been saved in the csv folder stated in the conf.py, e.g ./beacon2-pi-api/ri-tools/conf/conf.py/csv/your/path. \nAll the VCFs to be converted are saved in./beacon2-pi-api/ri-tools/files/vcf/files_to_read\n\nIf this is done, then run:\n\ndocker exec -it ri-tools python convert_csvTObff.py\n\nThe Beacon Friendly Format JSONs from CSVs will be generated in the output_docs folder, with the name of the collection followed by .json extension, e.g. biosamples.json.\nIf you have VCFs to convert, then run too:\n\ndocker exec -it ri-tools python genomicVariations_vcf.py\n\nAnd export it from the mongoDB to a JSON BFF with: \n\ndocker exec ri-tools mongoexport --jsonArray --uri "mongodb://\nroot:example@127.0.0.1:27017/beacon?authSource=admin" --collection\ngenomicVariations > ./beacon2-pi-api/ri-tools/output_docs//\ngenomicVariations.json\n\nAll these BFF JSONs will be used to populate a mongoDB for beacon usage.\nBefore moving to the actual installation and population of the beacon, ensure all the following steps have been accomplished:\nIf reading from VCFs, move the necessary files to files_to_read folder\nFill in CSV templates\nSet the correct paths in conf.py\nRun: docker exec -it ri-tools python convert_csvTObff.py\nIf reading from a VCF, run: docker exec -it ri-tools python genomicVariations_vcf.py\nFind your BFFs in the output_docs\n\n8. Deploying a Beacon 2 Production Implementation API\nTo deploy your first beacon we\u2019ll use the automated deployment that relies on the beacon2-pi-api repository that we cloned at the beginning of this tutorial.\nBy executing the automated deployment we\u2019ll be performing two key actions:\nThe deployment of your beacon instance\nThe initial insertion of data\n\nBy default, the data saved in the directory ./beacon2-pi-api/beacon/connections/mongo/data/test will be inserted.\nWant to insert your own data?\nMove your BFFs file to the /data/test directory.\nUpdate the path in the /beacon/connections/mongo/Makefile document.\n\nFrom the root of the repository, ./beacon2-pi-api, run:\n\nbash mongostart.sh\n\nIf the operation is successful, you will have a beacon up and running at http://localhost:5050/api.\n\nThis is what you should be seeing:\n\ndocker exec -it ri-tools python genomicVariations_vcf.py\n\nRelationship-elements\nWhen searching http://localhost:5050/api we land at the /info entrypoint of the API. To make your own landing page please visit Editing your beacon information webpage and modify the required files before deploying.\n\n9. Reviewing Inserted Entries\nNow, let\u2019s move through your beacon and see the data you have inserted. If you did not modify the content of the data folder the test BFFs will have been included.\n\nTo visit the inserted data you need to move through the different beacon endpoints:\n\nhttp://localhost:5050/api/datasets\nhttp://localhost:5050/api/analyses\nhttp://localhost:5050/api/runs\nhttp://localhost:5050/api/cohorts\nhttp://localhost:5050/api/biosamples\nhttp://localhost:5050/api/individuals\nhttp://localhost:5050/api/g_variants\n10. Loading more data\nFrom the cloned repository move to:\n\ncd ./beacon2-pi-api/beacon/connections/mongo/data\n\nAnd copy all the BFFs that you want to insert.\nThen, run:\n\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/datasets.json --collection datasets\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/individuals.json --collection individuals\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/cohorts.json --collection cohorts\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/analyses.json --collection analyses\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/biosamples.json --collection biosamples\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/genomicVariations.json --collection genomicVariations\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/runs.json --collection runs\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/targets.json --collection targets\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/caseLevelData.json --collection caseLevelData\n\nNote that you will only need to run the command for the collections that you have new data. \n\nEach time data is imported into the beacon, indexes need to be created for the queries to run smoothly. Do it by running:\n\ndocker exec beaconprod python /beacon/connections/mongo/reindex.py',"Create Your Beacon"),Ul("/resources","Documentation\n\nResources\n\nExplore more about Beacon v2 with the following resources:\n\n- Beacon v2 webpage: https://genomebeacons.org/\n- Beacon v2 specification documentation: https://docs.genomebeacons.org/\n- Beacon v2 API on GitHub: https://github.com/ga4gh-beacon/beacon-v2/\n- Beacon Verifier (verify the configuration of your implementation): https://beacon-verifier-demo.ega-archive.org/\n- Add your Beacon implementation to our Beacon registry: https://docs.google.com/forms/d/e/1FAIpQLScORwEVRAlsa8qe9SerKZLGy6qjphApjsHXC8-EcaOrUpW8tw/viewform\n","Resources"),null),Wl=()=>{const[e,n]=(0,a.useState)({}),t=e=>{const t="cloning-repository"===e?"git clone https://github.com/EGA-archive/beacon2-ri-api.git":"bash start.sh";navigator.clipboard.writeText(t).then((()=>{n((n=>({...n,[e]:!0}))),setTimeout((()=>n((n=>({...n,[e]:!1})))),1500)})).catch((e=>console.log(e)))};return(0,we.jsxs)("div",{className:"deploymentContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/automated-deployment",className:"no-undeline",children:"Beacon 2 RI API"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/automated-deployment",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:" Automated Deployment"})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h3",{children:"Beacon 2 RI API"}),(0,we.jsx)("h1",{children:"Automated Deployment"}),(0,we.jsx)("p",{children:"This deployment relies on a Bash script and a Makefile, which won\u2019t deploy the IDP authentication service. If you wish to build the IDP service you will have to go to Manual Deployment."}),(0,we.jsx)("h2",{id:"cloning-repository",children:"Cloning the repository"}),(0,we.jsx)("p",{children:"First of all, start by cloning the GitHub repository in your system."}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["git clone"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-api.git",children:"https://github.com/EGA-archive/beacon2-ri-api.git"})]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("cloning-repository"),children:e["cloning-repository"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{className:"paragraph-ports",children:"Make sure the next list of ports are free of use in your system:"}),(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"27017 \u2192 mongo"}),(0,we.jsx)("li",{children:"5050 \u2192 beacon"}),(0,we.jsx)("li",{children:"5051 \u2192 permissions"})]}),(0,we.jsx)("h2",{id:"execute-start-script",children:"Execute start script from root"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"bash start.sh"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("execute-start-script"),children:e["execute-start-script"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{className:"paragraph-final",children:["If the operation is successful, you will have a beacon up and running at"," ",(0,we.jsx)("a",{href:"http://localhost:5050/api",target:"_blank",rel:"noopener noreferrer",children:"http://localhost:5050/api"}),"."]})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]})},Kl=()=>{const[e,n]=(0,a.useState)({}),t=e=>{const t={"linking-ids":'dataset_name: ["individualId1", "individualId2", "biosampleId1", "biosampleId2"]\ncohort_name: ["individualId1", "individualId2", "biosampleId1", "biosampleId2"]',"extract-filtering-terms":"docker exec beacon python beacon/db/extract_filtering_terms.py","manual-filtering-terms":'db.filtering_terms.insertMany([{ "type": "alphanumeric", "id": "libraryStrategy", "scope": [ "runs"] }])',"conf-py":"alphanumeric_terms = ['libraryStrategy', 'molecularAttributes.geneIds', 'diseases.ageOfOnset.iso8601duration']","get-descendants":"docker exec beacon python beacon/db/get_descendants.py"}[e];t&&navigator.clipboard.writeText(t).then((()=>{n((n=>({...n,[e]:!0}))),setTimeout((()=>n((n=>({...n,[e]:!1})))),1500)})).catch((e=>console.log(e)))};return(0,we.jsxs)("div",{className:"dataLinkingContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/data-linking",className:"no-undeline",children:"Beacon 2 RI API"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/data-linking",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:"Data Linking"})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h3",{children:"Beacon 2 RI API"}),(0,we.jsx)("h1",{children:"Data Linking"}),(0,we.jsx)("h2",{id:"linking-ids",children:"Linking ids to dataset and cohort"}),(0,we.jsx)("p",{children:"In case your data is splitted by dataset or cohort, you will need to tell beacon which biosample ids and which individual ids belong to which dataset."}),(0,we.jsx)("p",{children:"Edit the file datasets.yml that is inside beacon/request folder and write the dataset with an array of all the ids (for individuals and biosamples together)."}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:[(0,we.jsx)("div",{children:"dataset_name: [\u201cindividualId1\u201d, \u201cindividualId2\u201d, \u201cbiosampleId1\u201d, \u201cbiosampleId2\u201d]"}),(0,we.jsx)("br",{}),(0,we.jsx)("div",{children:"cohort_name: [\u201cindividualId1\u201d, \u201cindividualId2\u201d, \u201cbiosampleId1\u201d, \u201cbiosampleId2\u201d]"})]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("linking-ids"),children:e["linking-ids"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:"Note that the dataset_name and cohort_name have to be identical as how is spelled the id for the dataset or cohort in the BFF record."})]}),(0,we.jsx)("h2",{id:"extract-filtering-terms",children:"Extract filtering terms"}),(0,we.jsx)("p",{children:"To automatically fill in the filtering terms endpoint and be able to apply the ontologies that are found inside your data to query your beacon, please execute the next script:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker exec beacon python beacon/db/extract_filtering_terms.py"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("extract-filtering-terms"),children:e["extract-filtering-terms"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("h2",{id:"manual-filtering-terms",children:["Manually adding filtering terms"," ",(0,we.jsx)("span",{className:"optional",children:"(optional)"})]}),(0,we.jsx)("p",{children:"To manually add filtering terms to your beacon, you will need to add them in your mongodb, by executing the next command, for example, for an alphanumeric term that applies to the libraryStrategy field for the runs collection:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["db.filtering_terms.insertMany([","{",' "type": "alphanumeric", "id": "libraryStrategy", "scope": [ "runs"] ',"}","])"]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("manual-filtering-terms"),children:e["manual-filtering-terms"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"After that, you will need to specify that this alphanumeric term is not related to an ontology by editing the conf.py inside beacon and deploy folder in the alphanumeric_terms array:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"alphanumeric_terms = ['libraryStrategy', 'molecularAttributes.geneIds', 'diseases.ageOfOnset.iso8601duration']"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("conf-py"),children:e["conf-py"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("h2",{id:"get-descendants",children:["Get descendant terms ",(0,we.jsx)("span",{className:"optional",children:"(optional)"})]}),(0,we.jsx)("p",{children:"If you have extracted filtering terms. in case you wish to add information about the descendant terms and similarity found for each ontology and be able to use the similarity and descendant terms as beacon options, please execute the next script:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker exec beacon python beacon/db/get_descendants.py"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("get-descendants"),children:e["get-descendants"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]})},$l=()=>{const[e,n]=(0,a.useState)({}),t=e=>{const t={"method-1-public":"public_datasets:\n- dataset_id","method-1-registered":"registered_datasets:\n- dataset_id","method-1-controlled":"username:\n- dataset_id","method-2-env":"SECRET_KEY=\"your_permissions_ui_secret_key\"\nOIDC_RP_CLIENT_ID='your_client_id'\nOIDC_RP_CLIENT_SECRET='your_client_secret'","method-2-start":"docker exec beacon-permissions bash permissions/permissions-ui/start.sh","aai-env":"LSAAI_CLIENT_ID='your_lsaai_client_id'\nLSAAI_CLIENT_SECRET='your_lsaai_client_secret'\nKEYCLOAK_CLIENT_ID='your_keycloak_client_id'\nKEYCLOAK_CLIENT_SECRET='your_keycloak_client_secret'","beacon-rebuild":"cd deploy && docker-compose up -d --build beacon","cors-main":'middlewares=[web.normalize_path_middleware(), middlewares.error_middleware, cors_middleware(origins=["your_URL"...',"cors-routes":'for route in list(beacon.router.routes()):\n        cors.add(route, {\n        "your_URL":\n            aiohttp_cors.ResourceOptions(allow_credentials=True,\n            expose_headers="*",\n            allow_methods=("POST", "PATCH", "GET", "OPTIONS"),\n            allow_headers=DEFAULT_ALLOW_HEADERS)',"beacon-info":"beacon_id = 'org.ega-archive.beacon-ri-demo'\nbeacon_name = 'Beacon Reference Implementation demo'\napi_version = 'v2.0.0'\nuri = 'https://beacon-apis-demo.ega-archive.org/api/'\norg_id = 'EGA'\norg_name = 'European Genome-Phenome Archive (EGA)'\norg_description = 'The European Genome-phenome Archive (EGA) is a service for permanent archiving and sharing of all types of personally identifiable genetic and phenotypic data resulting from biomedical research projects.'\norg_adress = 'C/ Dr. Aiguader, 88\nPRBB Building\n08003 Barcelona, Spain'\norg_welcome_url = 'https://ega-archive.org/'\norg_contact_url = 'mailto:beacon.ega@crg.eu'\norg_logo_url = 'https://legacy.ega-archive.org/images/logo.png'\norg_info = ''\ndescription = \"This Beacon is based on synthetic data hosted at the <a href='https://ega-archive.org/datasets/EGAD00001003338'>EGA</a>. The dataset contains 2504 samples including genetic data based on 1K Genomes data, and 76 individual attributes and phenotypic data derived from UKBiobank.\"\nversion = 'v2.0'\nwelcome_url = 'https://beacon.ega-archive.org/'\nalternative_url = 'https://beacon-apis-demo.ega-archive.org/api/'\ncreate_datetime = '2021-11-29T12:00:00.000000'\nupdate_datetime = ''","granularity-conf":'default_beacon_granularity = "record"\nmax_beacon_granularity = "record"',"granularity-response":"dummy_user:\n- record","beacon-handovers":"beacon_handovers = [\n    {\n        handoverType: {\n            id: 'CUSTOM:000001',\n            label: 'Project description'\n        },\n        note: 'Project description',\n        url: 'https://www.nist.gov/programs-projects/genome-bottle'\n    }\n]"}[e];t&&navigator.clipboard.writeText(t).then((()=>{n((n=>({...n,[e]:!0}))),setTimeout((()=>n((n=>({...n,[e]:!1})))),1500)})).catch((e=>console.log(e)))};return(0,we.jsxs)("div",{className:"apiConfigContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/api-configuration",className:"no-undeline",children:"Beacon 2 RI API"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/api-configuration",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:"Configuration"})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h3",{children:"Beacon 2 RI API"}),(0,we.jsx)("h1",{children:"Configuration"}),(0,we.jsx)("h2",{id:"managing-dataset-permissions",children:"Managing dataset permissions"}),(0,we.jsx)("p",{children:"There are 3 possible levels of beacon security for a dataset: public, registered and controlled."}),(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"A public dataset is a dataset that will be returned in a beacon query without an authentication token.\xa0"}),(0,we.jsx)("li",{children:"A registered dataset is a dataset that will be shown after a user sends a valid token (in other words, is logged in)."}),(0,we.jsx)("li",{children:"A controlled dataset is a dataset that needs a user to send a valid token for authentication and the user needs to be allowed to query that dataset."})]}),(0,we.jsx)("br",{}),(0,we.jsx)("p",{children:"There are two possible methods to administer these permissions for your datasets in beacon:"}),(0,we.jsx)("h4",{id:"method-1",children:"Method 1: Edit the .yml files"}),(0,we.jsxs)("p",{children:["For a ",(0,we.jsx)("b",{children:"public dataset"}),", edit the public_datasets.yml file inside permissions folder and add the dataset with the exact name it appears in the id from its pertinent record in datasets mongo collection:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["public_datasets:",(0,we.jsx)("br",{}),"-dataset_id"]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("method-1-public"),children:e["method-1-public"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{children:["For a ",(0,we.jsx)("b",{children:"registered dataset"}),", edit the registered_datasets.yml file inside permissions folder and add the dataset with the exact name it appears in the id from its pertinent record in datasets mongo collection:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["registered_datasets: ",(0,we.jsx)("br",{})," -dataset_id"]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("method-1-registered"),children:e["method-1-registered"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{children:["For a ",(0,we.jsx)("b",{children:"controlled dataset"}),", edit the controlled_datasets.yml file inside permissions folder and add the dataset inside the username you wish to give permissions with the exact name for the dataset as it appears in the id from its pertinent record in datasets mongo collection and the exact name the user has in Keycloak:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["username: ",(0,we.jsx)("br",{}),"-dataset_id"]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("method-1-controlled"),children:e["method-1-controlled"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("br",{}),(0,we.jsx)("h4",{id:"method-2",children:"Method 2: Launching the beacon admin UI"}),(0,we.jsx)("p",{children:"Create an .env file inside permissions/permissions-ui/web folder and add the next variables with their values:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:['SECRET_KEY="your_permissions_ui_secret_key"',(0,we.jsx)("br",{}),"OIDC_RP_CLIENT_ID='your_client_id'",(0,we.jsx)("br",{}),"OIDC_RP_CLIENT_SECRET='your_client_secret'"]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("method-2-env"),children:e["method-2-env"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"Execute the next command to launch the UI:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker exec beacon-permissions bash permissions/permissions-ui/start.sh"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("method-2-start"),children:e["method-2-start"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:(0,we.jsxs)("p",{children:["After that, open the UI in port"," ",(0,we.jsx)("a",{href:"http://localhost:8010",target:"_blank",rel:"noopener noreferrer",children:"http://localhost:8010"})," ","and navigate through public, registered, and controlled datasets to assign permissions for the datasets you have inside your Mongo database."]})}),(0,we.jsx)("h2",{id:"supplying-aai-credentials",children:"Supplying AAI credentials"}),(0,we.jsxs)("p",{children:["For making ",(0,we.jsx)("b",{children:"Keycloak"})," or ",(0,we.jsx)("b",{children:"LifeSciende AAI"})," work with beacon, you will need to add the client ID and client secret for each IDP in a new file you have to create inside permissions folder. This file must be called .env and needs to have the following variables:"]}),(0,we.jsxs)("div",{className:"codeSnippet",children:[(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["LSAAI_CLIENT_ID='your_lsaai_client_id' ",(0,we.jsx)("br",{}),"LSAAI_CLIENT_SECRET='your_lsaai_client_secret' ",(0,we.jsx)("br",{}),"KEYCLOAK_CLIENT_ID='your_keycloak_client_id' ",(0,we.jsx)("br",{}),"KEYCLOAK_CLIENT_SECRET='your_keycloak_client_secret'"]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("aai-env"),children:e["aai-env"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:"Note that you only need to fill variables for the AAI method you want to enable."})]})]}),(0,we.jsx)("p",{children:"After that, make sure you build your beacon container again:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"cd deploy && docker-compose up -d --build beacon"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("beacon-rebuild"),children:e["beacon-rebuild"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h2",{id:"handling-cors",children:"Handling CORS"}),(0,we.jsx)("p",{children:"To avoid CORS using beacon and the frontend or a third-party authorization site like Keycloak, you will have to include all these URLs inside __main__.py in the beacon folder, including them in the CORS middleware."}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'middlewares=[web.normalize_path_middleware(), middlewares.error_middleware, cors_middleware(origins=["your_URL"...'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("cors-main"),children:e["cors-main"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"and CORS routes lists:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["for route in list(beacon.router.routes()):",(0,we.jsx)("br",{}),"\xa0\xa0\xa0\xa0cors.add(route, {",(0,we.jsx)("br",{}),'\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0"your_URL": {',(0,we.jsx)("br",{}),"\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0aiohttp_cors.ResourceOptions(allow_credentials=True,",(0,we.jsx)("br",{}),'\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0expose_headers="*",',(0,we.jsx)("br",{}),'\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0allow_methods=("POST", "PATCH", "GET", "OPTIONS"),',(0,we.jsx)("br",{}),"\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0\xa0allow_headers=DEFAULT_ALLOW_HEADERS)} \xa0\xa0\xa0\xa0})"]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("your-text-here"),children:e["your-text-here"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h2",{id:"editing-beacon-info",children:"Editing your beacon information"}),(0,we.jsxs)("p",{children:["To show correctly your beacon\u2019s information you will need to edit both conf.py files from beacon and deploy folders. ",(0,we.jsx)("br",{}),"To do so, edit the following variables:"]}),(0,we.jsxs)("div",{className:"codeSnippet",children:[(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["beacon_id = 'org.ega-archive.beacon-ri-demo' # ID of the Beacon",(0,we.jsx)("br",{}),"beacon_name = 'Beacon Reference Implementation demo' # Name of the Beacon service",(0,we.jsx)("br",{}),"api_version = 'v2.0.0'  # Version of the Beacon implementation",(0,we.jsx)("br",{}),"uri = 'https://beacon-apis-demo.ega-archive.org/api/'\n","org_id = 'EGA' # Id of the organization",(0,we.jsx)("br",{}),"org_name = 'European Genome-Phenome Archive (EGA)' # Full name",(0,we.jsx)("br",{}),(0,we.jsx)(we.Fragment,{children:"org_description = ('The European Genome-phenome Archive (EGA)'\n            'is a service for permanent archiving and sharing' \n            'of all types of personally identifiable genetic' \n            'and phenotypic data resulting from biomedical research projects.')\n"}),"org_adress = ('C/ Dr. Aiguader, 88'\n          'PRBB Building'\n          '08003 Barcelona, Spain')\n","org_welcome_url = 'https://ega-archive.org/'\n","org_contact_url = 'mailto:beacon.ega@crg.eu'\n","org_logo_url = 'https://legacy.ega-archive.org/images/logo.png'\n","org_info = ''\n","description = \"This Beacon is based on synthetic data hosted at the <a href='https://ega-archive.org/datasets/EGAD00001003338'>EGA</a>. The dataset contains 2504 samples including genetic data based on 1K Genomes data, and 76 individual attributes and phenotypic data derived from UKBiobank.\"\n","version = 'v2.0'\n","welcome_url = 'https://beacon.ega-archive.org/'\n","alternative_url = 'https://beacon-apis-demo.ega-archive.org/api/'\n","create_datetime = '2021-11-29T12:00:00.000000'\n","update_datetime = ''"]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("beacon-info"),children:e["beacon-info"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:"Note that this is the info that will be shown in the /info endpoint for your beacon."})]})]}),(0,we.jsx)("h2",{id:"limiting-granularity",children:"Limiting the granularity"}),(0,we.jsxs)("p",{children:["Beacon RI can limit the granularity per user or for all the beacon.",(0,we.jsx)("br",{})," ",(0,we.jsx)("br",{})," If you want to"," ",(0,we.jsx)("b",{children:"limit the granularity for all the users to a maximum"}),", you can point it in conf.py for both beacon and deploy folders, by editing the next variables:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:['default_beacon_granularity = "record" ',(0,we.jsx)("br",{}),'max_beacon_granularity = "record"']}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("granularity-conf"),children:e["granularity-conf"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{children:["If you want to ",(0,we.jsx)("b",{children:"limit the granularity per user"}),", you can do this editing the response_type.yml file inside beacon/request folder, adding the username and the maximum granularity you want to allow this user to have:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["dummy_user:",(0,we.jsx)("br",{}),"-record"]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("granularity-response"),children:e["granularity-response"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h2",{id:"adding-beacon-handovers",children:"Adding beacon handovers"}),(0,we.jsx)("p",{children:"You can specify what are the handovers for your beacon by adding them in the array of your conf.py file inside beacon folder:"}),(0,we.jsx)("div",{className:"codeSnippet custom-codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["beacon_handovers = [",(0,we.jsx)("br",{}),"\xa0\xa0{",(0,we.jsx)("br",{}),"\xa0\xa0\xa0\xa0'handoverType': {",(0,we.jsx)("br",{}),"\xa0\xa0\xa0\xa0\xa0\xa0'id': 'CUSTOM:000001',",(0,we.jsx)("br",{}),"\xa0\xa0\xa0\xa0\xa0\xa0'label': 'Project description'",(0,we.jsx)("br",{}),"\xa0\xa0\xa0\xa0},",(0,we.jsx)("br",{}),"\xa0\xa0\xa0\xa0'note': 'Project description',",(0,we.jsx)("br",{}),"\xa0\xa0\xa0\xa0'url': 'https://www.nist.gov/programs-projects/genome-bottle'",(0,we.jsx)("br",{}),"\xa0\xa0}",(0,we.jsx)("br",{}),"]"]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("beacon-handovers"),children:e["beacon-handovers"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]})},Yl=()=>{const[e,n]=(0,a.useState)({}),t=(e,t)=>{navigator.clipboard.writeText(e).then((()=>{n((e=>({...e,[t]:!0}))),setTimeout((()=>{n((e=>({...e,[t]:!1})))}),1500)})).catch((e=>console.error("Failed to copy text: ",e)))};return(0,we.jsxs)("div",{className:"queryingApiContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsxs)("a",{href:"/querying-the-api",className:"no-undeline",children:[" ","Beacon 2 RI API"]}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/querying-the-api",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:"Querying the API"})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h3",{children:"Beacon 2 RI API"}),(0,we.jsx)("h1",{children:"Querying the API"}),(0,we.jsx)("p",{children:"Beacon RI accepts two types of request methods: GET and POST."}),(0,we.jsxs)("div",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsxs)("div",{children:["You can find further introduction to Beacon parameters in\xa0",(0,we.jsx)("a",{href:"https://docs.genomebeacons.org/variant-queries/",target:"_blank",rel:"noopener noreferrer",children:"Beacon v2 Documentation"}),"."]})]}),(0,we.jsx)("h2",{id:"get-method",children:"GET Method"}),(0,we.jsx)("p",{children:"GET method is a bit more limited as this method only accepts request parameters, three generic parameters and ontology filters."}),(0,we.jsx)("h6",{children:"Query Parameters"}),(0,we.jsx)("p",{children:"These are the defined set of parameters (key-value pair) attached to the end of a URL used to provide additional information to a web server when making requests."}),(0,we.jsxs)("table",{children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:"Generic parameters"}),(0,we.jsx)("th",{children:"Type"}),(0,we.jsx)("th",{children:"Examples"})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 requestedSchema"}),(0,we.jsx)("td",{children:"String"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"requestedSchema=ga4gh-service-info-v1.0"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("requestedSchema=ga4gh-service-info-v1.0","requestedSchema"),children:e.requestedSchema?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 skip"}),(0,we.jsxs)("td",{children:["integer (Skip) \u2265 0",(0,we.jsx)("br",{}),"Default: 0"]}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"skip=0"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("skip=0","skip"),children:e.skip?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 limit"}),(0,we.jsxs)("td",{children:["integer (Limit) \u2265 0",(0,we.jsx)("br",{}),"Default: 10",(0,we.jsx)("br",{}),"Max: 100"]}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"limit=10"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("limit=10","limit"),children:e.limit?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]})]})]}),(0,we.jsxs)("table",{children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:"Request parameters"}),(0,we.jsx)("th",{children:"Type"}),(0,we.jsx)("th",{children:"Examples"})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 start"}),(0,we.jsxs)("td",{children:["integer <int64> \u2265 0",(0,we.jsx)("br",{}),"(or comma-separated integers)"]}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"start=16050074"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("start=16050074","start"),children:e.start?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 end"}),(0,we.jsxs)("td",{children:["integer <int64> \u2265 0",(0,we.jsx)("br",{}),"(or comma-separated integers)"]}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"end=16052080"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("end=16052080","end"),children:e.end?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 assemblyId"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"assemblyId=GRCh38"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("assemblyId=GRCh38","assemblyId"),children:e.assemblyId?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 referenceName"}),(0,we.jsx)("td",{children:"integer <int64> \u2265 0"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"referenceName=22"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("referenceName=22","referenceName"),children:e.referenceName?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 referenceBases"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"referenceBases=G"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("referenceBases=G","referenceBases"),children:e.referenceBases?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 alternateBases"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"alternateBases=A"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("alternateBases=A","alternateBases"),children:e.alternateBases?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 variantMinLength"}),(0,we.jsx)("td",{children:"integer <int64> \u2265 0"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"variantMinLength=5"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("variantMinLength=5","variantMinLength"),children:e.variantMinLength?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 variantMaxLength"}),(0,we.jsx)("td",{children:"integer <int64> \u2265 0"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"variantMaxLength=2"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("variantMaxLength=2","variantMaxLength"),children:e.variantMaxLength?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 geneId"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"geneId=CCNL2"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("geneId=CCNL2","geneId"),children:e.geneId?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 aminoacidChange"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"aminoacidChange=p.Thr130Met"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("aminoacidChange=p.Thr130Met","aminoacidChange"),children:e.aminoacidChange?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 variantType"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"variantType=SNP"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("variantType=SNP","variantType"),children:e.variantType?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 genomicAlleleShortForm"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"genomicAlleleShortForm=NC_000022.11:g.16050075A>G"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("genomicAlleleShortForm=NC_000022.11:g.16050075A&gt;G","genomicAlleleShortForm"),children:e.genomicAlleleShortForm?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 clinicalRelevance"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"clinicalRelevance=pathogenic"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("clinicalRelevance=pathogenic","clinicalRelevance"),children:e.clinicalRelevance?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]})]})]}),(0,we.jsxs)("table",{children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:"Filters"}),(0,we.jsx)("th",{children:"Type"}),(0,we.jsx)("th",{children:"Examples"})]})}),(0,we.jsx)("tbody",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 filters"}),(0,we.jsxs)("td",{children:["string",(0,we.jsx)("br",{}),"(comma-separated strings of ontologies)"]}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"filters=NCIT:C16576,NCIT:C42331"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("filters=NCIT:C16576,NCIT:C42331","filters"),children:e.filters?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]})})]}),(0,we.jsx)("h4",{id:"get-query-examples",children:"GET query examples"}),(0,we.jsx)("h5",{className:"cursive",children:"Beacon Sequence Query"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"http://localhost:5050/api/g_variants?referenceName=22&start=7577120&referenceBases=G&alternateBases=A"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("http://localhost:5050/api/g_variants?referenceName=22&start=7577120&referenceBases=G&alternateBases=A","beacon-sequence-query"),children:e["beacon-sequence-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Beacon GeneId Query"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"http://localhost:5050/api/g_variants?geneId=CHR_START-DUXAP8&variantMaxLength=2&variantType=INDEL"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("http://localhost:5050/api/g_variants?geneId=CHR_START-DUXAP8&variantMaxLength=2&variantType=INDEL","beacon-geneid-query"),children:e["beacon-geneid-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Beacon Bracket Query"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"http://localhost:5050/api/g_variants?referenceName=22&start=16050074,16050076&end=16050090&variantType=SNP"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("http://localhost:5050/api/g_variants?referenceName=22&start=16050074,16050076&end=16050090&variantType=SNP","beacon-bracket-query"),children:e["beacon-bracket-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h2",{id:"post-method",children:"POST Method"}),(0,we.jsx)("p",{children:"In POST method requests, the endpoints accept all generic parameters and all the filters, apart from the request parameters."}),(0,we.jsx)("h6",{className:"underline",children:"Query Parameters"}),(0,we.jsxs)("table",{children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:"Generic parameters"}),(0,we.jsx)("th",{children:"Type"}),(0,we.jsx)("th",{children:"Examples"})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 includeResultsetResponses"}),(0,we.jsxs)("td",{children:["string ",(0,we.jsx)("br",{}),'Default: "HIT"',(0,we.jsx)("br",{}),'Enum: "ALL" "HIT" "MISS" "NONE"']}),(0,we.jsx)("td",{})]}),(0,we.jsxs)("tr",{children:[(0,we.jsxs)("td",{children:["\u2022 pagination object (Pagination) Items:",(0,we.jsx)("li",{children:"skip"}),(0,we.jsx)("li",{children:"limit"})]}),(0,we.jsxs)("td",{children:[(0,we.jsx)(we.Fragment,{}),"integer (Skip) \u2265 0",(0,we.jsx)("br",{}),"Default: 0 ",(0,we.jsx)("br",{}),(0,we.jsx)("br",{}),"integer (Limit) \u2265 0",(0,we.jsx)("br",{}),"Default: 10",(0,we.jsx)("br",{}),"Max: 100"]}),(0,we.jsxs)("td",{children:[(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"skip=0"}),(0,we.jsx)("br",{})]}),(0,we.jsx)("br",{}),(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"limit=10"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("skip=0 limit=10","pagination"),children:e.pagination?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})]})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 requestedGranularity"}),(0,we.jsxs)("td",{children:["string (Granularity)",(0,we.jsx)("br",{}),'Default: "boolean"',(0,we.jsx)("br",{}),'Enum: "boolean" "count" "record"']}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:'requestedGranularity="record"'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('requestedGranularity="record"',"requestedGranularity"),children:e.requestedGranularity?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 testMode"}),(0,we.jsxs)("td",{children:["boolean (TestMode)",(0,we.jsx)("br",{}),"Default: false"]}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"testMode=false"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("testMode=false","testMode"),children:e.testMode?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]})]})]}),(0,we.jsxs)("table",{children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:"Request parameters"}),(0,we.jsx)("th",{children:"Type"}),(0,we.jsx)("th",{children:"Examples"})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 start"}),(0,we.jsxs)("td",{children:["integer <int64> \u2265 0",(0,we.jsx)("br",{}),"(or comma-separated integers)"]}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"start=16050074"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("start=16050074","start"),children:e.start?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 end"}),(0,we.jsxs)("td",{children:["integer <int64> \u2265 0",(0,we.jsx)("br",{}),"(or comma-separated integers)"]}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"end=16052080"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("end=16052080","end"),children:e.end?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 assemblyId"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"assemblyId=GRCh38"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("assemblyId=GRCh38","assemblyId"),children:e.assemblyId?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 referenceName"}),(0,we.jsx)("td",{children:"integer <int64> \u2265 0"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"referenceName=22"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("referenceName=22","referenceName"),children:e.referenceName?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 referenceBases"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"referenceBases=G"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("referenceBases=G","referenceBases"),children:e.referenceBases?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 alternateBases"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"alternateBases=A"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("alternateBases=A","alternateBases"),children:e.alternateBases?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 variantMinLength"}),(0,we.jsx)("td",{children:"integer <int64> \u2265 0"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"variantMinLength=5"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("variantMinLength=5","variantMinLength"),children:e.variantMinLength?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 variantMaxLength"}),(0,we.jsx)("td",{children:"integer <int64> \u2265 0"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"variantMaxLength=2"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("variantMaxLength=2","variantMaxLength"),children:e.variantMaxLength?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 geneId"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"geneId=CCNL2"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("geneId=CCNL2","geneId"),children:e.geneId?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 aminoacidChange"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"aminoacidChange=p.Thr130Met"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("aminoacidChange=p.Thr130Met","aminoacidChange"),children:e.aminoacidChange?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 variantType"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"variantType=SNP"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("variantType=SNP","variantType"),children:e.variantType?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 genomicAlleleShortForm"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"genomicAlleleShortForm=NC_000022.11:g.16050075A>G"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("genomicAlleleShortForm=NC_000022.11:g.16050075A&gt;G","genomicAlleleShortForm"),children:e.genomicAlleleShortForm?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 clinicalRelevance"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"clinicalRelevance=pathogenic"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("clinicalRelevance=pathogenic","clinicalRelevance"),children:e.clinicalRelevance?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]})]})]}),(0,we.jsxs)("table",{children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:"Filters"}),(0,we.jsx)("th",{children:"Type"}),(0,we.jsx)("th",{children:"Examples"})]})}),(0,we.jsx)("tbody",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 filters"}),(0,we.jsxs)("td",{children:["Array of OntologyFilter ",(0,we.jsx)("br",{}),"(object) or ",(0,we.jsx)("br",{}),"AlphanumericFilter (object) ",(0,we.jsx)("br",{}),"or CustomFilter (object) ",(0,we.jsx)("br",{}),"(Filtering Term Element)"]}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsxs)("code",{children:["filters: [",(0,we.jsx)("br",{}),"\xa0\xa0","{",(0,we.jsx)("br",{}),"\xa0\xa0\xa0\xa0",'"id": "geographicOrigin",',(0,we.jsx)("br",{}),"\xa0\xa0\xa0\xa0",'"operator": "!",',(0,we.jsx)("br",{}),"\xa0\xa0\xa0\xa0",'"value": "England"',(0,we.jsx)("br",{}),"\xa0\xa0","}",(0,we.jsx)("br",{}),"]"]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('"filters": [\n  {\n    "id": "geographicOrigin",\n    "operator": "!",\n    "value": "England"\n  }\n]',"filters-example"),children:e["filters-example"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]})})]}),(0,we.jsx)("h4",{className:"underline",id:"post-query-examples",children:"POST query examples"}),(0,we.jsx)("h5",{className:"cursive",children:"Genomic query"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": {\n        "apiVersion": "2.0"\n    },\n    "query": {\n        "requestParameters": {\n            "alternateBases": "G",\n            "referenceBases": "A",\n            "start": [16050074],\n            "end": [16050568],\n            "variantType": "SNP"\n        },\n        "filters": [],\n        "includeResultsetResponses": "HIT",\n        "pagination": {\n            "skip": 0,\n            "limit": 10\n        },\n        "testMode": false,\n        "requestedGranularity": "record"\n    }\n}\' \\\n  http://localhost:5050/api/g_variants'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": {\n        "apiVersion": "2.0"\n    },\n    "query": {\n        "requestParameters": {\n            "alternateBases": "G",\n            "referenceBases": "A",\n            "start": [16050074],\n            "end": [16050568],\n            "variantType": "SNP"\n        },\n        "filters": [],\n        "includeResultsetResponses": "HIT",\n        "pagination": {\n            "skip": 0,\n            "limit": 10\n        },\n        "testMode": false,\n        "requestedGranularity": "record"\n    }\n}\' \\\n  http://localhost:5050/api/g_variants',"genomic-query"),children:e["genomic-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("div",{children:[(0,we.jsx)("h5",{className:"cursive",children:"Simple CURIE based filters query"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": {\n      "apiVersion": "2.0"\n    },\n    "query": {\n      "requestParameters": {},\n      "filters": [\n        { "id": "NCIT:C20197", "scope": "individual" },\n        { "id": "NCIT:C42331", "scope": "individual" }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": {\n        "skip": 0,\n        "limit": 10\n      },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": {\n      "apiVersion": "2.0"\n    },\n    "query": {\n      "requestParameters": {},\n      "filters": [\n        { "id": "NCIT:C20197", "scope": "individual" },\n        { "id": "NCIT:C42331", "scope": "individual" }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": {\n        "skip": 0,\n        "limit": 10\n      },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals',"curie-query"),children:e["curie-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Numerical value query"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": {\n      "apiVersion": "2.0"\n    },\n    "query": {\n      "filters": [\n        {\n          "id": "Weight",\n          "operator": ">",\n          "value": "75"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": {\n        "skip": 0,\n        "limit": 10\n      },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": {\n      "apiVersion": "2.0"\n    },\n    "query": {\n      "filters": [\n        {\n          "id": "Weight",\n          "operator": ">",\n          "value": "75"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": {\n        "skip": 0,\n        "limit": 10\n      },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals',"numerical-query"),children:e["numerical-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Alphanumerical value query (Exact value)"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": {\n      "apiVersion": "2.0"\n    },\n    "query": {\n      "filters": [\n        {\n          "id": "geographicOrigin",\n          "operator": "=",\n          "value": "England"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": {\n        "skip": 0,\n        "limit": 10\n      },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": {\n      "apiVersion": "2.0"\n    },\n    "query": {\n      "filters": [\n        {\n          "id": "geographicOrigin",\n          "operator": "=",\n          "value": "England"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": {\n        "skip": 0,\n        "limit": 10\n      },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals',"exact-value-query"),children:e["exact-value-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Alphanumerical value query (Like value)"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": {\n      "apiVersion": "2.0"\n    },\n    "query": {\n      "filters": [\n        {\n          "id": "geographicOrigin",\n          "operator": "=",\n          "value": "England"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": {\n        "skip": 0,\n        "limit": 10\n      },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": {\n      "apiVersion": "2.0"\n    },\n    "query": {\n      "filters": [\n        {\n          "id": "geographicOrigin",\n          "operator": "=",\n          "value": "%land%"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": {\n        "skip": 0,\n        "limit": 10\n      },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals',"exact-value-query"),children:e["exact-value-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Alphanumerical value query (NOT value)"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": {\n      "apiVersion": "2.0"\n    },\n    "query": {\n      "filters": [\n        {\n          "id": "geographicOrigin",\n          "operator": "!",\n          "value": "%land%"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": {\n        "skip": 0,\n        "limit": 10\n      },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": {\n      "apiVersion": "2.0"\n    },\n    "query": {\n      "filters": [\n        {\n          "id": "geographicOrigin",\n          "operator": "!",\n          "value": "%land%"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": {\n        "skip": 0,\n        "limit": 10\n      },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals',"not-value-query"),children:e["not-value-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Semantic similarity query"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": {\n      "apiVersion": "2.0"\n    },\n    "query": {\n      "filters": [\n        {\n          "id": "NCIT:C16576",\n          "similarity": "high"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": {\n        "skip": 0,\n        "limit": 10\n      },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": {\n      "apiVersion": "2.0"\n    },\n    "query": {\n      "filters": [\n        {\n          "id": "NCIT:C16576",\n          "similarity": "high"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": {\n        "skip": 0,\n        "limit": 10\n      },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals',"similarity-query"),children:e["similarity-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Custom filter query"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n  "meta": { "apiVersion": "2.0" },\n  "query": {\n    "filters": [{"id": "sampleOriginType:blood"}],\n    "includeResultsetResponses": "HIT",\n    "pagination": { "skip": 0, "limit": 10 },\n    "testMode": false,\n    "requestedGranularity": "record"\n  }\n}\' \\\n  http://localhost:5050/api/biosamples'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n  "meta": { "apiVersion": "2.0" },\n  "query": {\n    "filters": [{"id": "sampleOriginType:blood"}],\n    "includeResultsetResponses": "HIT",\n    "pagination": { "skip": 0, "limit": 10 },\n    "testMode": false,\n    "requestedGranularity": "record"\n  }\n}\' \\\n  http://localhost:5050/api/biosamples',"custom-filter-query"),children:e["custom-filter-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Join genomic query"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": {\n      "apiVersion": "2.0"\n    },\n    "query": {\n      "requestParameters": {\n        "alternateBases": "G",\n        "referenceBases": "A",\n        "start": [16050074],\n        "end": [16050568],\n        "variantType": "SNP"\n      },\n      "filters": [\n        {\n          "id": "NCIT:C20197",\n          "scope": "individual"\n        },\n        {\n          "id": "NCIT:C42331",\n          "scope": "individual"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": {\n        "skip": 0,\n        "limit": 100\n      },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/g_variants'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": {\n      "apiVersion": "2.0"\n    },\n    "query": {\n      "requestParameters": {\n        "alternateBases": "G",\n        "referenceBases": "A",\n        "start": [16050074],\n        "end": [16050568],\n        "variantType": "SNP"\n      },\n      "filters": [\n        {\n          "id": "NCIT:C20197",\n          "scope": "individual"\n        },\n        {\n          "id": "NCIT:C42331",\n          "scope": "individual"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": {\n        "skip": 0,\n        "limit": 100\n      },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/g_variants',"join-genomic-query"),children:e["join-genomic-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Join individuals query"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": {\n      "apiVersion": "2.0"\n    },\n    "query": {\n      "requestParameters": {\n        "alternateBases": "G",\n        "referenceBases": "A",\n        "start": [16050004],\n        "end": [16950115],\n        "variantType": "SNP"\n      },\n      "filters": [\n        {\n          "id": "NCIT:C20197"\n        },\n        {\n          "id": "NCIT:C42331"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": {\n        "skip": 0,\n        "limit": 10\n      },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": {\n      "apiVersion": "2.0"\n    },\n    "query": {\n      "requestParameters": {\n        "alternateBases": "G",\n        "referenceBases": "A",\n        "start": [16050004],\n        "end": [16950115],\n        "variantType": "SNP"\n      },\n      "filters": [\n        {\n          "id": "NCIT:C20197"\n        },\n        {\n          "id": "NCIT:C42331"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": {\n        "skip": 0,\n        "limit": 10\n      },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals',"join-individuals-query"),children:e["join-individuals-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Query with access token"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'curl \\\n  -H \'Content-Type: application/json\' \\\n  -H \'Authorization: Bearer <access_token>\' \\\n  -X POST \\\n  -d \'{\n    "meta": {\n      "apiVersion": "2.0"\n    },\n    "query": {\n      "requestParameters": {},\n      "filters": [\n        {\n          "id": "geographicOrigin",\n          "operator": "=",\n          "value": "%land%"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": {\n        "skip": 0,\n        "limit": 10\n      },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('curl \\\n  -H \'Content-Type: application/json\' \\\n  -H \'Authorization: Bearer <access_token>\' \\\n  -X POST \\\n  -d \'{\n    "meta": {\n      "apiVersion": "2.0"\n    },\n    "query": {\n      "requestParameters": {},\n      "filters": [\n        {\n          "id": "geographicOrigin",\n          "operator": "=",\n          "value": "%land%"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": {\n        "skip": 0,\n        "limit": 10\n      },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals',"query-with-token"),children:e["query-with-token"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsx)("h6",{className:"underline",children:"List of endpoints"}),(0,we.jsx)("h5",{className:"cursive",children:"Endpoints that accept all query parameters"}),(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"/api/g_variants"}),(0,we.jsx)("li",{children:"/api/individuals"})]}),(0,we.jsx)("h5",{className:"cursive",children:"Endpoints that only accept filters and generic parameters"}),(0,we.jsxs)("ul",{className:"endpoint-list",children:[(0,we.jsx)("li",{children:"/api/analyses"}),(0,we.jsx)("li",{children:"/api/biosamples"}),(0,we.jsx)("li",{children:"/api/cohorts"}),(0,we.jsx)("li",{children:"/api/datasets"}),(0,we.jsx)("li",{children:"/api/runs"})]}),(0,we.jsx)("h5",{className:"cursive",children:"Endpoints that only accept generic parameters"}),(0,we.jsxs)("ul",{className:"endpoint-list",children:[(0,we.jsx)("li",{children:"/api/filtering_terms"}),(0,we.jsxs)("li",{children:["/api/analyses/","{id}"]}),(0,we.jsxs)("li",{children:["/api/analyses/","{id}","/g_variants"]}),(0,we.jsx)("li",{children:"/api/analyses/filtering_terms"}),(0,we.jsxs)("li",{children:["/api/biosamples/","{id}"]}),(0,we.jsxs)("li",{children:["/api/biosamples/","{id}","/analyses"]}),(0,we.jsxs)("li",{children:["/api/biosamples/","{id}","/g_variants"]}),(0,we.jsxs)("li",{children:["/api/biosamples/","{id}","/runs"]}),(0,we.jsx)("li",{children:"/api/biosamples/filtering_terms"}),(0,we.jsxs)("li",{children:["/api/cohorts/","{id}"]}),(0,we.jsxs)("li",{children:["/api/cohorts/","{id}","/analyses"]}),(0,we.jsxs)("li",{children:["/api/cohorts/","{id}","/biosamples"]}),(0,we.jsxs)("li",{children:["/api/cohorts/","{id}","/g_variants"]}),(0,we.jsxs)("li",{children:["/api/cohorts/","{id}","/individuals"]}),(0,we.jsxs)("li",{children:["/api/cohorts/","{id}","/runs"]}),(0,we.jsx)("li",{children:"/api/cohorts/filtering_terms"}),(0,we.jsxs)("li",{children:["/api/datasets/","{id}"]}),(0,we.jsxs)("li",{children:["/api/datasets/","{id}","/analyses"]}),(0,we.jsxs)("li",{children:["/api/datasets/","{id}","/biosamples"]}),(0,we.jsxs)("li",{children:["/api/datasets/","{id}","/g_variants"]}),(0,we.jsxs)("li",{children:["/api/datasets/","{id}","/individuals"]}),(0,we.jsxs)("li",{children:["/api/datasets/","{id}","/runs"]}),(0,we.jsx)("li",{children:"/api/datasets/filtering_terms"}),(0,we.jsxs)("li",{children:["/api/g_variants/","{id}"]}),(0,we.jsxs)("li",{children:["/api/g_variants/","{id}","/analyses"]}),(0,we.jsxs)("li",{children:["/api/g_variants/","{id}","/biosamples"]}),(0,we.jsxs)("li",{children:["/api/g_variants/","{id}","/individuals"]}),(0,we.jsxs)("li",{children:["/api/g_variants/","{id}","/runs"]}),(0,we.jsx)("li",{children:"/api/g_variants/filtering_terms"}),(0,we.jsxs)("li",{children:["/api/individuals/","{id}"]}),(0,we.jsxs)("li",{children:["/api/individuals/","{id}","/analyses"]}),(0,we.jsxs)("li",{children:["/api/individuals/","{id}","/biosamples"]}),(0,we.jsxs)("li",{children:["/api/individuals/","{id}","/g_variants"]}),(0,we.jsxs)("li",{children:["/api/individuals/","{id}","/runs"]}),(0,we.jsx)("li",{children:"/api/individuals/filtering_terms"}),(0,we.jsxs)("li",{children:["/api/runs/","{id}"]}),(0,we.jsxs)("li",{children:["/api/runs/","{id}","/analyses"]}),(0,we.jsxs)("li",{children:["/api/runs/","{id}","/g_variants"]}),(0,we.jsx)("li",{children:"/api/runs/filtering_terms"})]}),(0,we.jsx)("h5",{className:"cursive",children:"Endpoints that don\u2019t accept query parameters"}),(0,we.jsxs)("ul",{className:"endpoint-list",children:[(0,we.jsx)("li",{children:"/api"}),(0,we.jsx)("li",{children:"/api/info"}),(0,we.jsx)("li",{children:"/api/service-info"}),(0,we.jsx)("li",{children:"/api/configuration"}),(0,we.jsx)("li",{children:"/api/entry_types"}),(0,we.jsx)("li",{children:"/api/map"})]})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]})},Xl=()=>{const[e,n]=(0,a.useState)({}),t=(t,o)=>{navigator.clipboard.writeText(t).then((()=>{n({...e,[o]:!0}),setTimeout((()=>n({...e,[o]:!1})),2e3)}))};return(0,we.jsxs)("div",{className:"startingGuideContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/starting-guide",className:"no-undeline",children:"Beacon RI Tools v2"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/starting-guide",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:"Starting Guide"})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h3",{children:"Beacon RI Tools v2"}),(0,we.jsx)("h1",{children:"Tools Introduction"}),(0,we.jsxs)("p",{children:["The Beacon RI Tools v2 are a set of tools written in Python available in the following\xa0",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon-data-tools",target:"_blank",rel:"noopener noreferrer",children:"GitHub repository"}),"."]}),(0,we.jsx)("p",{children:"Notice that Beacon RI Tools v2 are meant to work with either Beacon 2 RI API or Beacon 2 PI API. Furthermore, it specifically comes with all the features that are needed for generating Beacon v2 PI API compatible data."}),(0,we.jsxs)("p",{children:["The main goal of Beacon RI Tools v2 is to obtain the"," ",(0,we.jsx)("b",{children:"Beacon Friendly Format (BFF) files "}),", which are JSON files following Beacon v2 official specifications. These files have the correct format and structure to be injected into a Beacon v2 MongoDB database."]}),(0,we.jsxs)("div",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:(0,we.jsxs)("p",{className:"note-paragraph",children:["Discover how to set up a Beacon v2 with MongoDB and learn to inject BFF files by visiting the official"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-api",target:"_blank",rel:"noopener noreferrer",children:"Beacon v2 PI API"}),"\u2014 where you can download your own Beacon v2 for free!"]})})]}),(0,we.jsxs)("p",{children:["The Beacon data model consists of 7 collections:",(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:" Analyses"}),(0,we.jsx)("li",{children:" Biosamples"}),(0,we.jsx)("li",{children:" Cohorts"}),(0,we.jsx)("li",{children:" Datasets"}),(0,we.jsx)("li",{children:" GenomicVariations"}),(0,we.jsx)("li",{children:" Individuals"}),(0,we.jsx)("li",{children:" Runs"})]})," "]}),(0,we.jsx)("p",{children:"Therefore the Beacon RI Tools v2 allow the user to create the different 7 BFF options."}),(0,we.jsx)("p",{children:"The 6 metadata BFF files are created populating a CSV file with a defined structure that follows the Beacon data model. However, the genomicVariations collection can be created either from a CSV or from a VCF file, minimising conversion time."}),(0,we.jsxs)("p",{children:["The initial step, since Beacon RI Tools v2 cannot convert the data from any source, is to populate the CSV templates that are built following the 7 collections of the Beacon data model. This enables the Beacon RI Tools v2 to convert from CSV to BFF. Note that you can find the CSV templates in the ","",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon-data-tools/tree/main/csv/templates",target:"_blank",rel:"noopener noreferrer",children:"templates folder"}),""," of the same repository."]}),(0,we.jsx)("h2",{id:"installation-guide",children:"Installation Guide"}),(0,we.jsx)("p",{children:"First of all, clone or download the Beacon RI Tools v2 repository to your computer:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["git clone"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon-data-tools.git",target:"_blank",rel:"noopener noreferrer",children:"https://github.com/EGA-archive/beacon-data-tools.git"})]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("git clone https://github.com/EGA-archive/beacon-data-tools.git","clone-repo"),children:e["clone-repo"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"To light up the container with Beacon RI Tools v2, execute the docker-compose command inside the root folder of the repository:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker compose up -d --build"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("docker compose up -d --build","docker-command"),children:e["docker-command"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"Once the container is up and running, you can start using Beacon RI Tools v2 , congratulations!"})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]})},Zl=()=>{const[e,n]=(0,a.useState)({}),t=(e,t)=>{navigator.clipboard.writeText(e).then((()=>{n((e=>({...e,[t]:!0}))),setTimeout((()=>{n((e=>({...e,[t]:!1})))}),1500)})).catch((e=>console.error("Failed to copy text: ",e)))};return(0,we.jsxs)("div",{className:"configFileToolsContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/configuration-file",className:"no-undeline",children:"Beacon RI Tools v2"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/configuration-file",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:"Configuration File"})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h3",{children:"Beacon RI Tools v2"}),(0,we.jsx)("h1",{children:"Configuration File"}),(0,we.jsx)("p",{children:"In order to use Beacon RI Tools v2 one needs to modify a configuration file with the desired files and options for the different collections."}),(0,we.jsxs)("p",{children:["You will find conf.py inside the"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon-data-tools/tree/main/conf",target:"_blank",rel:"noopener noreferrer",children:"conf directory"}),". Inside this file, you will find the following information:"]}),(0,we.jsx)("h2",{id:"input-output-files-config",children:"Input and Output files config parameters"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{id:"input-output-config",children:"csv_filename='csv/examples/cohorts.csv' output_docs_folder='output_docs/CINECA_dataset/'"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>{var e;const n=null===(e=document.getElementById("input-output-config"))||void 0===e?void 0:e.innerText;n&&t(n,"input-output-config")},children:e["input-output-config"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h2",{id:"generic-config-parameters",children:"Generic config parameters"}),(0,we.jsx)("p",{children:"The first part of this configuration only concerns the conversion of CSV to BFF. It is used only when you are converting from CSV."}),(0,we.jsxs)("p",{children:["The ",(0,we.jsx)("i",{children:"csv_filename"})," variable sets where the CSV file is and from where the script will read the data. Bear in mind that the CSV must be inside the folder ",(0,we.jsx)("i",{children:"csv"}),", and you can use subfolders within it. This CSV file needs to have the headers written as you can find in the files inside"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon-data-tools/tree/main/csv/templates",target:"_blank",rel:"noopener noreferrer",children:"templates"}),". Note that any header that has a different name from the ones that appear inside the templates will not be read by the Beacon Data Tools."]}),(0,we.jsxs)("p",{children:["The ",(0,we.jsx)("i",{children:"output_docs_folder"})," sets the folder where your final BFF (JSON files) will be saved once execution of beacon tools finishes. This folder is mandatory to be always inside ",(0,we.jsx)("i",{children:"output_docs"}),", so only the subdirectory inside ",(0,we.jsx)("i",{children:"output_docs"}),"can be modified in this path."]}),(0,we.jsx)("h2",{id:"VCF-conversion-config-parameters",children:"VCF conversion config parameters"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{id:"vcf-config",children:["allele_counts=False",(0,we.jsx)("br",{}),"reference_genome='GRCh37' # Choose one between NCBI36, GRCh37, GRCh38",(0,we.jsx)("br",{}),"datasetId='COVID_pop11_fin_2'",(0,we.jsx)("br",{}),"case_level_data=False",(0,we.jsx)("br",{}),"exact_heterozygosity=False",(0,we.jsx)("br",{}),"num_rows=15000000",(0,we.jsx)("br",{}),"verbosity=False"]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>{var e;const n=null===(e=document.getElementById("vcf-config"))||void 0===e?void 0:e.innerText;n&&t(n,"vcf-config")},children:e["vcf-config"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"The second part of the configuration file pertains to the VCF to BFF conversion. This only needs to be used in case you are using a VCF as a source for the genomic variants collection."}),(0,we.jsxs)("p",{children:["The\xa0",(0,we.jsx)("i",{children:"num_variants"}),"\xa0is the variable you need to write in case you are executing the VCF conversion"," ",(0,we.jsx)("i",{children:"(genomicVariations_vcf.py)"}),". This will tell the script how many variants will be read and converted from the file(s)."]}),(0,we.jsxs)("p",{children:["The ",(0,we.jsx)("i",{children:"allele_frequency"})," field lets you set a threshold for the allele frequency (AF) of the variants you want to convert from the VCF file. If you set it at 0.5, all the variants with 0.5 or less AF will be converted to BFF. 1 is the default value (all variants will be converted)."]}),(0,we.jsxs)("p",{children:["The\xa0",(0,we.jsx)("i",{children:"allele_counts"}),"\xa0now is not implemented yet, just leave it as False."]}),(0,we.jsxs)("p",{children:["The ",(0,we.jsx)("i",{children:"reference_genome"})," is the reference genome the tool will use to map the position of the chromosomes. Make sure to select the same version as the one used to generate your data."]}),(0,we.jsxs)("p",{children:["The ",(0,we.jsx)("i",{children:"datasetId"}),", ",(0,we.jsx)("i",{children:"case_level_data"}),", and"," ",(0,we.jsx)("i",{children:"exact_zygosity"})," parameters are ",(0,we.jsx)("b",{children:"only"})," applicable in the"," ",(0,we.jsx)("b",{children:"Beacon Production Implementation environment"}),"."]}),(0,we.jsxs)("p",{children:["The\xa0",(0,we.jsx)("i",{children:"datasetId"}),"\xa0needs to match the id of your datasets.csv or datasets.json file. This will add a datasetId field in every record to match the record with the dataset it belongs to."]}),(0,we.jsxs)("p",{children:["When converting a VCF file to BFF, you can use two optional boolean parameters to include sample-level information about each variant:"," ",(0,we.jsx)("i",{children:"case_level_data"})," and ",(0,we.jsx)("i",{children:"exact_zygosity"}),"."]}),(0,we.jsxs)("p",{children:[(0,we.jsx)("i",{children:"case_level_data"})," will link variants to biosamples. If set to True, this option enables mapping each variant to the specific biosamples (i.e., samples) that carry it. This allows downstream queries to return more granular, sample-level results instead of just listing variants in general."]}),(0,we.jsxs)("p",{className:"wider-note",children:[(0,we.jsx)("img",{className:"note-symbol-wider",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsxs)("div",{children:["Important: For this mapping to work correctly, you must:",(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"Have genotype (GT) information available in the VCF. "}),(0,we.jsx)("li",{children:"Provide a valid biosamples schema in your BFF files."})]})]})]}),(0,we.jsxs)("p",{children:["Be sure to read the Case-level"," ",(0,we.jsx)("a",{href:"conversion-from-vcf-to-bff#case-level-data-conversion",children:"data conversion section"})," ","for detailed guidance on setting this up."]}),(0,we.jsxs)("p",{children:[(0,we.jsx)("i",{children:"exact_zygosity "}),"will refine biosample classification. If set to True This parameter classifies each biosample based on its genotype:",(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"Homozygous for the reference allele"}),(0,we.jsx)("li",{children:"Heterozygous"}),(0,we.jsx)("li",{children:"Homozygous for the alternate allele"})]}),"This adds precision to your data and can help users interpret zygosity in their queries."]}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol-wider",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsxs)("div",{children:["If ",(0,we.jsx)("i",{children:"case_level_data"})," is set to False, make sure to also set",(0,we.jsx)("i",{children:"exact_zygosity "}),"to False \u2014 it won\u2019t have any effect otherwise."]})]}),(0,we.jsxs)("p",{children:["The\xa0",(0,we.jsx)("i",{children:"num_rows"}),"\xa0 are the aproximate calculation you expect for the total of variants in each vcf there are. Make sure this is greater than the total variants expected. It was automatically calculated before but it was very slow sometimes to calculate all the variants number in a VCF."]}),(0,we.jsxs)("p",{children:["The\xa0",(0,we.jsx)("i",{children:"verbosity"}),"\xa0will give streaming logs with the reason why a variant has been skipped to be inserted. Recommendation is to leave this as False."]}),(0,we.jsx)("br",{}),(0,we.jsx)("br",{})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]})},Jl=()=>{const[e,n]=(0,a.useState)({}),t=(e,t)=>{navigator.clipboard.writeText(t).then((()=>{n((n=>({...n,[e]:!0}))),setTimeout((()=>n((n=>({...n,[e]:!1})))),1500)})).catch((e=>console.log(e)))};return(0,we.jsxs)("div",{className:"conversionContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/conversion-from-csv-to-bff",className:"no-undeline",children:"Beacon RI Tools v2"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/conversion-from-csv-to-bff",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:"Conversion from CSV to BFF"})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h3",{children:"Beacon RI Tools v2"}),(0,we.jsx)("h1",{children:"Conversion from CSV to BFF"}),(0,we.jsx)("h1",{id:"choosing-column-headers",children:"Choosing Column Headers"}),(0,we.jsxs)("p",{children:["Each type of entry (e.g., analyses, biosamples, individuals, etc.) has its own corresponding CSV template located in the"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/tree/main/csv/templates",children:"csv-templates folder"}),". These templates list all the valid column headers you can use."]}),(0,we.jsxs)("p",{children:["For example, in ",(0,we.jsx)("em",{children:"biosamples.csv"}),", one of the headers you'll see is:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"biosampleStatus|id"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("biosample-status-id","biosampleStatus|id"),children:e["biosample-status-id"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"This naming convention follows the GA4GH Beacon v2 specification. The | (pipe) character separates different levels of a nested field: the part on the left represents the parent field, and the part on the right is the nested element within it. In this case:"}),(0,we.jsxs)("ul",{children:[(0,we.jsxs)("li",{children:[(0,we.jsx)("em",{children:"biosampleStatus"})," is an object field."]})," ",(0,we.jsxs)("li",{children:[(0,we.jsx)("em",{children:"id"})," is one of its subfields, representing an ontology identifier."]})]}),(0,we.jsx)("p",{children:"According to the Beacon v2 specification, biosampleStatus is defined as an ontology term describing the classification of the sample (e.g., whether it\u2019s a reference or abnormal sample). Here\u2019s a simplified excerpt from the spec:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'"biosampleStatus": {\n  "description": "Classification of the sample using EFO ontology terms.",\n  "examples": [\n    { "id": "EFO:0009654", "label": "reference sample" },\n    { "id": "EFO:0009655", "label": "abnormal sample" },\n    { "id": "EFO:0009656", "label": "neoplastic sample" }\n    // ...\n  ]\n}'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("biosample-status-id",'"biosampleStatus": {\n  "description": "Classification of the sample using EFO ontology terms.",\n  "examples": [\n    { "id": "EFO:0009654", "label": "reference sample" },\n    { "id": "EFO:0009655", "label": "abnormal sample" },\n    { "id": "EFO:0009656", "label": "neoplastic sample" }\n    // ...\n  ]\n}'),children:e["biosample-status-id"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"So, in your CSV file, the biosampleStatus|id column should contain ontology IDs like:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"biosampleStatus|id\nEFO:0009654\nEFO:0009655"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("biosample-status-id-values","biosampleStatus|id\nEFO:0009654\nEFO:0009655"),children:e["biosample-status-id-values"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"Optionally, you can include the label for each ontology ID in a second column called biosampleStatus|label, like so:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"biosampleStatus|id,biosampleStatus|label\nEFO:0009654,reference sample\nEFO:0009655,abnormal sample"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("biosample-status-id-label","biosampleStatus|id,biosampleStatus|label\nEFO:0009654,reference sample\nEFO:0009655,abnormal sample"),children:e["biosample-status-id-label"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"Each row corresponds to a different biosample, and both the id and label must match the values defined in the ontology."}),(0,we.jsx)("h1",{id:"filling-in-data",children:"Filling in Data"}),(0,we.jsx)("p",{children:"To convert metadata or phenotypic/clinical data into BFF (Beacon File Format), you need to create one CSV file per collection type (e.g., individuals, biosamples, analyses, etc.). Each file should contain records formatted according to the column headers, which map directly to fields in the Beacon v2 schema."}),(0,we.jsxs)("p",{children:["Each row in a CSV file represents a new and independent document in the final output. Templates and examples for each collection type can be found in the"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/tree/main/csv/examples",children:"csv folder"})," ","of the repository."]}),(0,we.jsx)("h6",{className:"underline",children:"General Rules for Filling In the CSV Files"}),(0,we.jsx)("ul",{children:(0,we.jsxs)("li",{children:[(0,we.jsx)("b",{children:"Multiple values in the same document"}),":"]})}),(0,we.jsxs)("p",{children:["To include multiple values in a single field (e.g. multiple IDs, multiple measures), separate them using the pipe symbol (|).",(0,we.jsx)("br",{}),"Example:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"id\nHG00001|HG00002"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("individual-id-values","id\nHG00001|HG00002"),children:e["individual-id-values"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"Make sure the values in related columns are in the same order:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"label\nlabelForHG00001|labelForHG00002"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("individual-label-values","label\nlabelForHG00001|labelForHG00002"),children:e["individual-label-values"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("br",{}),(0,we.jsx)("ul",{children:(0,we.jsxs)("li",{children:[(0,we.jsx)("b",{children:"Filling in info fields"}),":"]})}),(0,we.jsxs)("p",{children:["The info field is flexible and accepts free-form data, but must be written as JSON. This is also required for certain fields in the genomicVariations collection (e.g., copies or subjects).",(0,we.jsx)("br",{}),"Example:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'{"info": "Some important information to be added."}'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("info-snippet",'{"info": "Some important information to be added."}'),children:e["info-snippet"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("br",{}),(0,we.jsx)("ul",{children:(0,we.jsxs)("li",{children:[(0,we.jsx)("b",{children:"Respect the column names"}),":"]})}),(0,we.jsx)("p",{children:"Column headers must match exactly those defined in the template files. The conversion script will only recognize correctly spelled headers. Mistyped headers will be ignored or cause errors."}),(0,we.jsx)("br",{}),(0,we.jsx)("ul",{children:(0,we.jsxs)("li",{children:[(0,we.jsx)("b",{children:"Optional vs. required fields"}),":"]})}),(0,we.jsx)("p",{children:"Not all columns need to be filled in. Some are optional or context-specific. However, certain fields are required depending on the collection type (see below). If a required field is missing or misfilled, an error will be raised. Only columns with actual data will be included in the final output \u2014 unused columns can be removed if preferred."}),(0,we.jsx)("h4",{className:"no-underline",children:"Required Fields per Collection"}),(0,we.jsxs)("table",{className:"dockerTableSimple",children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Collection"})}),(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Required Fields"})})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Analyses"}),(0,we.jsx)("td",{children:"id, analysisDate, pipelineName"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Biosamples"}),(0,we.jsx)("td",{children:"id, biosampleStatus, sampleOriginType"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Cohorts"}),(0,we.jsx)("td",{children:"id, name, cohortType"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Datasets"}),(0,we.jsx)("td",{children:"id, name"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Individuals"}),(0,we.jsx)("td",{children:"id, sex"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Runs"}),(0,we.jsx)("td",{children:"id, biosampleId, runDate"})]})]})]}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsxs)("div",{children:["For more information about each field, refer to the"," ",(0,we.jsx)("a",{href:"https://github.com/ga4gh-beacon/beacon-v2/tree/main/models/json/beacon-v2-default-model",children:"default schemas"})," ","and the"," ",(0,we.jsx)("a",{href:"https://github.com/ga4gh-beacon/beacon-v2",children:"Beacon v2 specification"}),"."]})]}),(0,we.jsx)("h6",{className:"underline",children:"Additional Notes"}),(0,we.jsxs)("ul",{children:[(0,we.jsxs)("li",{children:["The ",(0,we.jsx)("em",{children:"id"})," field in the ",(0,we.jsx)("b",{children:"Biosamples"})," collection must match the sample IDs in your VCF header."]}),(0,we.jsxs)("li",{children:["The ",(0,we.jsx)("em",{children:"id"})," field in the ",(0,we.jsx)("b",{children:"Individuals"})," collection does"," ",(0,we.jsx)("em",{children:"not"})," have to match VCF sample IDs. But, if it doesn\u2019t, you must fill in the individualId field in the ",(0,we.jsx)("b",{children:"Biosamples"})," ","collection to link individuals and samples."]}),(0,we.jsxs)("li",{children:["The ",(0,we.jsx)("em",{children:"datasetId"})," in the ",(0,we.jsx)("b",{children:"Datasets"})," collection must match the ",(0,we.jsx)("em",{children:"datasetId"})," defined in your conf.py file (this applies only to production Beacon deployments)."]}),(0,we.jsx)("li",{children:"You are not required to submit CSV files for all collections. If you do not have data for a given collection, you can omit its CSV file \u2014 the Beacon tools will not raise any errors."})]}),(0,we.jsx)("h1",{id:"conversion-from-csv-to-bff",children:"Conversion from CSV to BFF"}),(0,we.jsxs)("p",{children:["Before converting your CSV files, ensure that the"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/blob/main/conf/conf.py",children:"conf.py"})," ","file (located in the"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/tree/main/conf",children:"conf folder"}),") is correctly configured to read the appropriate input files and settings."]}),(0,we.jsx)("h6",{className:"underline",children:"Convert All CSVs at Once"}),(0,we.jsxs)("p",{children:["To convert all CSV files at once, make sure they are saved in the"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/tree/main/csv",children:"csv folder"})," ","(refer to the"," ",(0,we.jsx)("a",{href:"https://b2ri-documentation-demo.ega-archive.org/api-configuration",children:"Configuration File"})," ","section for details)."]}),(0,we.jsx)("p",{children:"Then run the following command from the root of the project:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker exec -it ri-tools python convert_csvTObff.py"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("convertCSVtoBFF","docker exec -it ri-tools python convert_csvTObff.py"),children:e.convertCSVtoBFF?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{children:["This script will process all available CSVs and generate one JSON file per collection in the"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/tree/main/output_docs",children:"output_docs"})," ","folder."]}),(0,we.jsx)("h6",{className:"underline",children:"Convert a Specific Collection"}),(0,we.jsx)("p",{children:"If you prefer to convert only one collection at a time, you can use one of the following commands depending on the type of data:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker exec -it ri-tools python analyses_csv.py"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("analyses","docker exec -it ri-tools python analyses_csv.py"),children:e.analyses?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker exec -it ri-tools python biosamples_csv.py"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("biosamples","docker exec -it ri-tools python biosamples_csv.py"),children:e.biosamples?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker exec -it ri-tools python cohorts_csv.py"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("cohorts","docker exec -it ri-tools python cohorts_csv.py"),children:e.cohorts?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker exec -it ri-tools python datasets_csv.py"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("datasets","docker exec -it ri-tools python datasets_csv.py"),children:e.datasets?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker exec -it ri-tools python genomicVariations_csv.py"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("genomicVariations","docker exec -it ri-tools python genomicVariations_csv.py"),children:e.genomicVariations?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker exec -it ri-tools python individuals_csv.py"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("individuals","docker exec -it ri-tools python individuals_csv.py"),children:e.individuals?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker exec -it ri-tools python runs_csv.py"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("runs","docker exec -it ri-tools python runs_csv.py"),children:e.runs?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{children:["Each script will generate a corresponding JSON file (e.g., cohorts.json) inside the"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/tree/main/output_docs",children:"output_docs"})," ","folder."]}),(0,we.jsx)("p",{children:"These JSON documents follow the Beacon File Format (BFF) and are ready to be imported into a MongoDB database for use with a GA4GH Beacon v2 instance."}),(0,we.jsxs)("p",{children:["To learn how to load these JSON files into your Beacon, refer to the documentation for the"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-api",children:"Beacon v2 RI API"})," ","and the"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-pi-api",children:"Beacon v2 PI API"}),"."]})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]})},ec=()=>{const[e,n]=(0,a.useState)({}),t=(e,t)=>{navigator.clipboard.writeText(t).then((()=>{n((n=>({...n,[e]:!0}))),setTimeout((()=>n((n=>({...n,[e]:!1})))),1500)})).catch((e=>console.log(e)))};return(0,we.jsxs)("div",{className:"conversionContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/conversion-from-vcf-to-bff",className:"no-undeline",children:"Beacon RI Tools v2"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/conversion-from-vcf-to-bff",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:"Conversion from VCF to BFF"})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h3",{children:"Beacon RI Tools v2"}),(0,we.jsx)("h1",{children:"Conversion from VCF to BFF"}),(0,we.jsx)("h2",{id:"reading-VCF",children:"Reading your VCF"}),(0,we.jsx)("p",{children:"Beacon RI Tools v2 will read the different columns for your variants in the VCF and place them inside the Legacy Variation schema for the Beacon v2 Spec."}),(0,we.jsxs)("p",{children:["First of all, Beacon RI Tools v2 will read the basic VCF parameters that come inside their columns:",(0,we.jsxs)("ol",{children:[(0,we.jsx)("li",{children:"CHROM: will add the chromosome number in the genomicHGVSId field of the record."}),(0,we.jsx)("li",{children:"POS: Will add the coordinates in start and end fields of the record in 0-typed annotation."}),(0,we.jsx)("li",{children:"ID: will add the identification of the variant in the variantType field (SNP, INDEL, etc...)"}),(0,we.jsx)("li",{children:"REF: will fill the referenceBases field in the record of the variant."}),(0,we.jsx)("li",{children:"ALT: will fill the alternateBases field in the record of the variant."})]})]}),(0,we.jsxs)("p",{children:["By default, Beacon RI Tools v2 supports ",(0,we.jsx)("b",{children:"VEP annotation"}),", in case your VCF has the designated VEP row in the VCF header:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'##INFO=<ID=CSQ,Number=.,Type=String,Description="Consequence annotations from Ensembl VEP. Format: Allele|Consequence|IMPACT|SYMBOL|Gene|Feature_type|Feature|BIOTYPE|EXON|INTRON|HGVSc|HGVSp|cDNA_position|CDS_position|Protein_position|Amino_acids|Codons|Existing_variation|REF_ALLELE|UPLOADED_ALLELE|DISTANCE|STRAND|FLAGS|SYMBOL_SOURCE|HGNC_ID|CANONICAL|HGVS_OFFSET">'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("vcfInfo",'##INFO=<ID=CSQ,Number=.,Type=String,Description="Consequence annotations from Ensembl VEP. Format: Allele|Consequence|IMPACT|SYMBOL|Gene|Feature_type|Feature|BIOTYPE|EXON|INTRON|HGVSc|HGVSp|cDNA_position|CDS_position|Protein_position|Amino_acids|Codons|Existing_variation|REF_ALLELE|UPLOADED_ALLELE|DISTANCE|STRAND|FLAGS|SYMBOL_SOURCE|HGNC_ID|CANONICAL|HGVS_OFFSET">'),children:e.vcfInfo?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"Some of the fields will be parsed into BFF. Right now, the fields that will be read are:"}),(0,we.jsxs)("ul",{className:"list-no-bullets",children:[(0,we.jsx)("li",{children:"Symbol \u2192 molecularAttributes|geneIds"}),(0,we.jsx)("li",{children:"Uploaded_Allele \u2192 variation|variantType"}),(0,we.jsx)("li",{children:"HGVSp \u2192 molecularAttributes|aminoacidChanges"}),(0,we.jsx)("li",{children:"Consequence \u2192 molecularAttributes|molecularEffects|label"})]}),(0,we.jsx)("p",{children:"Additionally, for filling in the required fields, the INFO column will read the next entries:"}),(0,we.jsxs)("ul",{className:"list-no-bullets",children:[(0,we.jsx)("li",{children:"VT \u2192 variation|variantType (in case VCF is not VEP annotated)"}),(0,we.jsx)("li",{children:"AF \u2192 frequencyInPopulations| frequencies| alleleFrequency"}),(0,we.jsx)("li",{children:"AN \u2192 frequencyInPopulations| frequencies| alleleNumber"}),(0,we.jsx)("li",{children:"AC \u2192 frequencyInPopulations| frequencies| alleleCount"}),(0,we.jsx)("li",{children:"AC_Hom \u2192 frequencyInPopulations| frequencies| alleleCountHomozygous"}),(0,we.jsx)("li",{children:"AC_Het \u2192 frequencyInPopulations| frequencies| alleleCountHeterozygous"}),(0,we.jsx)("li",{children:"END \u2192 variation| location| interval| end| value (in case END column is not filled in)"})]}),(0,we.jsxs)("p",{children:["On the other hand, if your VCF doesn\u2019t have VEP annotations or you want to use your own customized annotations, you can do that by editing the files that are located in"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon-data-tools/tree/main/pipelines/default/templates",target:"_blank",rel:"noopener noreferrer",children:"this GitHub repository"}),". The files that you have to modify are"," ",(0,we.jsx)("strong",{children:"populations.json"})," and",(0,we.jsx)("strong",{children:" template.json"}),". ",(0,we.jsx)("br",{}),"The ",(0,we.jsx)("strong",{children:"populations.json"})," will allow you to add how you annotated all the allele frequency related entries:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'{ \n  "numberOfPopulations": 1,\n  "source": "The Genome Aggregation Database (gnomAD)",\n  "sourceReference": "gnomad.broadinstitute.org/",\n  "populations": [{\n            "population": "Total",\n            "alleleFrequency": "AF",\n            "alleleCount": "AC",\n            "alleleCountHomozygous": "AC_hom",\n            "alleleCountHeterozygous": "AC_het",\n            "alleleNumber": "AN" }]\n  }'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("genomic-data-extraction",'{\n    "numberOfPopulations": 1,\n    "source": "The Genome Aggregation Database (gnomAD)",\n    "sourceReference": "gnomad.broadinstitute.org/",\n    "populations": [\n        {\n            "population": "Total",\n            "alleleFrequency": "AF",\n            "alleleCount": "AC",\n            "alleleCountHomozygous": "AC_hom",\n            "alleleCountHeterozygous": "AC_het",\n            "alleleNumber": "AN"\n        }\n    ]\n}'),children:e["genomic-data-extraction"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:"Tip: If numberOfPopulations is greater than 1, you have to add as many populations you have in the populations array, while if populations is 0, then, no allele frequency will be read from this pipeline."})]}),(0,we.jsx)("p",{children:"The template.json file will allow you to map the annotations entries related to the variant type, the aminoacid change, the gene Id or the molecular effects in your vcf:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'{\n    "template": false,\n    "variantType": "VT",\n    "aminoacidChange": "HGVSp",\n    "geneId": "SYMBOL",\n    "molecularEffects": "CONSEQUENCE"\n}'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("variant-effects-json",'{\n    "template": false,\n    "variantType": "VT",\n    "aminoacidChange": "HGVSp",\n    "geneId": "SYMBOL",\n    "molecularEffects": "CONSEQUENCE"\n}'),children:e["variant-effects-json"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:"Tip: If you want to activate this pipeline, change the template variable to true. If you activate this template, this will override the VEP annotations."})]}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:"Please, keep in mind that multiallelic variants need to be split onto separate rows in the VCF."})]}),(0,we.jsx)("h2",{id:"variant-data-conversion",children:"Variant data conversion"}),(0,we.jsx)("p",{children:"If you do not want to fill the CSV file for the genomicVariations collection or you already have your data in the VCF format, you can convert directly from VCF to BFF."}),(0,we.jsxs)("p",{children:["To convert data from a VCF file to BFF (json), the VCF must be compressed and indexed (.vcf.gz + .vcf.gz.tbi). Beacon RI Tools v2 will read all the VCF files inside the"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon-data-tools/tree/main/files/vcf/files_to_read",target:"_blank",rel:"noopener noreferrer",children:"files_to_read folder"}),". You can convert one or multiple VCF files at a time."]}),(0,we.jsx)("p",{children:"To execute the conversion, use the next command:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker exec -it ri-tools python genomicVariations_vcf.py"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("genomicVariations_vcf","docker exec -it ri-tools python genomicVariations_vcf.py"),children:e.genomicVariations_vcf?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"This command will do the conversion from VCF to BFF and will load the final BFF documents into a mongoDB inside a container. This is done for memory size usage."}),(0,we.jsx)("p",{children:"After that, if needed, export your documents from the mongoDB to your machine as a BFF file (json) using two possible commands."}),(0,we.jsxs)("ul",{children:[(0,we.jsxs)("li",{children:["The first command will delete an internal ",(0,we.jsx)("i",{children:'"_id"'})," for each record that is generated by MongoDB:",(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'docker exec ri-tools mongoexport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --collection genomicVariations | sed \'/"_id":/s/"_id":[^,]*,//g\' > genomicVariations.json'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("exportWithoutId",'docker exec ri-tools mongoexport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --collection genomicVariations | sed \'/"_id":/s/"_id":[^,]*,//g\' > genomicVariations.json'),children:e.exportWithoutId?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsx)("br",{}),(0,we.jsxs)("li",{children:["The second command will keep the ",(0,we.jsx)("i",{children:'"_id"'})," entries generated by MongoDB. Note that this ID is not part of the specifications of the Beacon and will not affect your data and Beacon, you can keep it if you want:",(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'docker exec ri-tools mongoexport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --collection genomicVariations > genomicVariations.json'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("exportWithId",'docker exec ri-tools mongoexport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --collection genomicVariations > genomicVariations.json'),children:e.exportWithId?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]})]}),(0,we.jsxs)("p",{children:["This will generate the final BFF file (json) for the"," ",(0,we.jsx)("code",{children:"genomicVariations"})," collection using the VCF format as the source. Bear in mind that this time, the file will be saved in the directory you are located, so if you want to save it in the"," ",(0,we.jsx)("code",{children:"output_docs"})," folder, add it in the path of the"," ",(0,we.jsx)("code",{children:"mongoexport"})," (e.g."," ",(0,we.jsx)("code",{children:"> output_docs/genomicVariations.json"}),")."]}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:"If you are using the ri-tools image directly into Beacon v2 PI API or Beacon v2 RI API you won\u2019t need to mongoexport the .json files unless you specifically need them for any other purpose, as the MongoDB instance for the API will already be filled with the variants."})]}),(0,we.jsx)("p",{children:"As it has already been mentioned, the variants read from the VCF are directly stored into the Beacon RI Tools v2 MongoDB."}),(0,we.jsx)("p",{children:"If you need to do more conversions and you don\u2019t want to keep the variants inside it, you can remove them by using the next command:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["docker exec ri-tools-mongo /bin/bash -c 'mongo beacon -u root -p example --authenticationDatabase admin --eval \"db.genomicVariations.deleteMany(",")\"'"]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("deleteVariants","docker exec ri-tools-mongo /bin/bash -c 'mongo beacon -u root -p example --authenticationDatabase admin --eval \"db.genomicVariations.deleteMany({})\"'"),children:e.deleteVariants?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"At this point, you should have your data ready to be injected into the Beacon v2 RI API."}),(0,we.jsx)("h2",{id:"case-level-data-conversion",children:"Case level data conversion"}),(0,we.jsxs)("p",{children:["If you are converting with the paramater ",(0,we.jsx)("b",{children:"case_level_data"})," to True, this will add data into two collections: ",(0,we.jsx)("b",{children:"targets"})," ","and\xa0",(0,we.jsx)("b",{children:"caseLevelData"}),". If you need to export the variants to insert them in another mongoDB, you will need to export these two collections as well, by executing the next commands:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'docker exec ri-tools mongoexport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --collection caseLevelData > caseLevelData.json'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("exportCaseLevelData",'docker exec ri-tools mongoexport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --collection caseLevelData > caseLevelData.json'),children:e.exportCaseLevelData?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'docker exec ri-tools mongoexport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --collection targets > targets.json'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("exportTargets",'docker exec ri-tools mongoexport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --collection targets > targets.json'),children:e.exportTargets?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:"Bear in mind that if you are converting VCFs one by one for the same samples and not by batches (multiple VCFs at the same time), this will generate a target file every time. We recommend converting all VCFs for the same sampling (dataset) at the same time to avoid unnecessary target records to be created."})]}),(0,we.jsx)("p",{children:"Case level data conversion is only applicable in the Beacon Production Implementation environment, as it is specifically designed to handle detailed sample-level associations and structured genomic data storage optimized for production use."})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]})},nc=()=>(0,we.jsxs)("div",{className:"beaconUIDeploymentContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/deployment",className:"no-undeline",children:"Beacon UI"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/deployment",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:"Deployment"})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h3",{children:"Beacon UI"}),(0,we.jsx)("h1",{children:"Deployment"}),(0,we.jsxs)("p",{children:["Use the deployment"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-api/blob/master/deploy/README.md",target:"_blank",rel:"noopener noreferrer",children:"documentation"})," ","for all the containers for beacon to also deploy the user interface. You will find it running at ",(0,we.jsx)("em",{children:"http://localhost:3000"}),"."]})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]}),tc=()=>{const[e,n]=(0,a.useState)({}),t=(e,t)=>{navigator.clipboard.writeText(e).then((()=>{n((e=>({...e,[t]:!0}))),setTimeout((()=>{n((e=>({...e,[t]:!1})))}),1500)})).catch((e=>console.error("Failed to copy text: ",e)))};return(0,we.jsxs)("div",{className:"beaconUIConfigurationContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/configuration-ui",className:"no-undeline",children:"Beacon UI"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/configuration-ui",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:"Configuration UI"})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h3",{children:"Beacon UI"}),(0,we.jsx)("h1",{children:"Configuration"}),(0,we.jsxs)("p",{children:["Please first create a ",(0,we.jsx)("code",{children:".env"})," file inside the frontend folder so that you can modify some variables. Take into account that the file below will not be copied to GitHub as it contains keys and for security reasons it should be ignored:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:['REACT_APP_CLIENT_ID="ID of your LS Login" ',(0,we.jsx)("br",{}),'REACT_APP_CLIENT_SECRET="password of your LS Login" ',(0,we.jsx)("br",{}),'REACT_APP_KEYCLOAK_CLIENT_SECRET="password of your Keycloak login"']}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('REACT_APP_CLIENT_ID="ID of your LS Login"\nREACT_APP_CLIENT_SECRET="password of your LS Login"\nREACT_APP_KEYCLOAK_CLIENT_SECRET="password of your Keycloak login"',"env-config"),children:e["env-config"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{children:["Tip: for Life Science environment, please first create a user (",(0,we.jsx)("a",{href:"https://lifescience-ri.eu/ls-login/users/how-to-get-and-use-life-science-id.html",target:"_blank",rel:"noopener noreferrer",children:"create a user"}),"). After that you will need to register a service registry in order to be able to administrate your logins. Please go \xa0",(0,we.jsx)("a",{href:"https://services.aai.lifescience-ri.eu/spreg/",target:"_blank",rel:"noopener noreferrer",children:"here"}),"\xa0 and ask for a New Service - type OIDC."]}),(0,we.jsxs)("p",{children:["Then please edit the file \xa0",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-api/blob/master/frontend/src/config.json",target:"_blank",rel:"noopener noreferrer",children:"config.json"}),". You need to decide where you want the UI to point to when making requests. Find below an example:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["{"," ",(0,we.jsx)("br",{}),'\xa0\xa0"API_URL": "https://yourAPIdomain.com/beacon-network/v2.0.0", ',(0,we.jsx)("br",{}),'\xa0\xa0"REDIRECT_URL": "https://yourUIdomain.com", ',(0,we.jsx)("br",{}),'\xa0\xa0"KEYCLOAK_URL": "https://yourKEYCLOAKdomain.com"'," ",(0,we.jsx)("br",{}),"}"]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('{\n  "API_URL": "https://yourAPIdomain.com/beacon-network/v2.0.0",\n  "REDIRECT_URL": "https://yourUIdomain.com",\n  "KEYCLOAK_URL": "https://yourKEYCLOAKdomain.com"\n}',"config-json"),children:e["config-json"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"Finally, please include the URL of your User interface to the file beacon/_main_.py (line 103), so that it becomes part of the list of URLs accepted by CORS."}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:"Note that in the frontend folder you will find a file called .gitignore with the list of all files that need to be ignored."})]})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]})},oc=()=>(0,we.jsxs)("div",{className:"beaconUIQueriesContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/querying-the-ui",className:"no-undeline",children:"Beacon UI"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/querying-the-ui",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:"Querying the UI"})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h3",{children:"Beacon UI"}),(0,we.jsx)("h1",{children:"Queries"}),(0,we.jsx)("h4",{children:"Coming soon..."})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]}),ac=()=>{const[e,n]=(0,a.useState)({});return(0,we.jsxs)("div",{className:"networkUIDeploymentContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/network-deployment",className:"no-undeline",children:"Beacon Network UI"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/network-deployment",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:"Deployment"})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h3",{children:"Network UI"}),(0,we.jsx)("h1",{children:"Deployment"}),(0,we.jsxs)("p",{children:["Please run the following command in order to build the Docker images as defined in the ",(0,we.jsx)("code",{children:"docker-compose.yml"})," file and to start the containers in detached mode (running in the background):"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker-compose up -d --build"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>{return e="docker-compose up -d --build",t="docker-compose-command",void navigator.clipboard.writeText(e).then((()=>{n((e=>({...e,[t]:!0}))),setTimeout((()=>{n((e=>({...e,[t]:!1})))}),1500)})).catch((e=>console.error("Failed to copy text: ",e)));var e,t},children:e["docker-compose-command"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]})},ic=()=>{const[e,n]=(0,a.useState)({}),t=(e,t)=>{navigator.clipboard.writeText(e).then((()=>{n((e=>({...e,[t]:!0}))),setTimeout((()=>{n((e=>({...e,[t]:!1})))}),1500)})).catch((e=>console.error("Failed to copy text: ",e)))};return(0,we.jsxs)("div",{className:"networkUIConfigurationContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/network-configuration-ui",className:"no-undeline",children:"Beacon Network UI"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/network-configuration-ui",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:"Configuration UI"})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h3",{children:"Network UI"}),(0,we.jsx)("h1",{children:"Configuration"}),(0,we.jsxs)("p",{children:["Please first create a ",(0,we.jsx)("code",{children:".env"})," file inside the frontend folder so that you can modify some variables. Take into account that the file below will not be copied to GitHub as it contains keys and for security reasons it should be ignored:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["REACT_APP_CLIENT_ID='ID of your LS Login' ",(0,we.jsx)("br",{}),"REACT_APP_CLIENT_SECRET='password of your LS Login' ",(0,we.jsx)("br",{}),"REACT_APP_KEYCLOAK_CLIENT_SECRET='password of your Keycloak login' ",(0,we.jsx)("br",{}),"REACT_APP_KEYCLOAK_CLIENT_ID='ID of your Keycloak login' ",(0,we.jsx)("br",{}),"REACT_APP_KEYCLOAK_CLIENT_REALM='REALM of your Keycloak login'"]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("REACT_APP_CLIENT_ID='ID of your LS Login'\nREACT_APP_CLIENT_SECRET='password of your LS Login'\nREACT_APP_KEYCLOAK_CLIENT_SECRET='password of your Keycloak login'\nREACT_APP_KEYCLOAK_CLIENT_ID='ID of your Keycloak login'\nREACT_APP_KEYCLOAK_CLIENT_REALM='REALM of your Keycloak login'","env-config"),children:e["env-config"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"You will need to have created your Life Science and Keycloak environments before:"}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:(0,we.jsxs)("p",{className:"note-paragraph",children:["For Life Science environment, please first create a user. After that, you will need to register a service registry in order to be able to administrate your logins. Please go\xa0",(0,we.jsx)("a",{href:"https://services.aai.lifescience-ri.eu/spreg/",target:"_blank",rel:"noopener noreferrer",children:"here"}),"\xa0and ask for a New Service - type OIDC."]})})]}),(0,we.jsxs)("p",{children:["Then please edit the file\xa0",(0,we.jsx)("a",{href:"https://github.com/elixir-europe/beacon-network-ui/tree/main/frontend/src",target:"_blank",rel:"noopener noreferrer",children:"config.json"}),", which can be found inside the folder frontend/src. You need to decide where you want the UI to point to when making requests. Find below an example:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'{\n"API_URL": "https://yourAPIdomain.com/beacon-network/v2.0.0",\n"REDIRECT_URL": "https://yourUIdomain.com",\n"KEYCLOAK_URL": "https://yourKEYCLOAKdomain.com"\n}'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('{ "API_URL": "https://yourAPIdomain.com/beacon-network/v2.0.0",\n  "REDIRECT_URL": "https://yourUIdomain.com",\n  "KEYCLOAK_URL": "https://yourKEYCLOAKdomain.com"\n  }',"config-json"),children:e["config-json"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:"In the root folder (frontend) you will find a file called .gitignore with the list of all files that need to be ignored."})]})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]})},rc=()=>(0,we.jsxs)("div",{className:"networkUIQueriesContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/network-querying-the-ui",className:"no-undeline",children:"Beacon Network UI"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/network-querying-the-ui",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:"Querying the UI"})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h3",{children:"Network UI"}),(0,we.jsx)("h1",{children:"Queries"}),(0,we.jsx)("h4",{children:"Coming soon..."})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]}),sc=()=>{const[e,n]=(0,a.useState)({}),t=e=>{const t={"method-1-public":"public_datasets:\n- dataset_id","method-1-registered":"registered_datasets:\n- dataset_id","method-1-controlled":"username:\n- dataset_id","method-2-env":"SECRET_KEY=\"your_permissions_ui_secret_key\"\nOIDC_RP_CLIENT_ID='your_client_id'\nOIDC_RP_CLIENT_SECRET='your_client_secret'","method-2-start":"docker exec beacon-permissions bash permissions/permissions-ui/start.sh","aai-env":"LSAAI_CLIENT_ID='your_lsaai_client_id'\nLSAAI_CLIENT_SECRET='your_lsaai_client_secret'\nKEYCLOAK_CLIENT_ID='your_keycloak_client_id'\nKEYCLOAK_CLIENT_SECRET='your_keycloak_client_secret'","beacon-rebuild":"cd deploy && docker-compose up -d --build beacon","cors-main":'cors_urls = ["http://localhost:3000","https://cancer-beacon-demo.ega-archive.org", "https://beacon-network-demo2.ega-archive.org", "https://beacon.ega-archive.org"]',"cors-routes":'for route in list(beacon.router.routes()):\n        cors.add(route, {\n        "your_URL":\n            aiohttp_cors.ResourceOptions(allow_credentials=True,\n            expose_headers="*",\n            allow_methods=("POST", "PATCH", "GET", "OPTIONS"),\n            allow_headers=DEFAULT_ALLOW_HEADERS)',"beacon-info":"beacon_id = 'org.ega-archive.beacon-ri-demo'\nbeacon_name = 'Beacon Reference Implementation demo'\napi_version = 'v2.0.0'\nuri = 'https://beacon-apis-demo.ega-archive.org/api/'\norg_id = 'EGA'\norg_name = 'European Genome-Phenome Archive (EGA)'\norg_description = 'The European Genome-phenome Archive (EGA) is a service for permanent archiving and sharing of all types of personally identifiable genetic and phenotypic data resulting from biomedical research projects.'\norg_adress = 'C/ Dr. Aiguader, 88\nPRBB Building\n08003 Barcelona, Spain'\norg_welcome_url = 'https://ega-archive.org/'\norg_contact_url = 'mailto:beacon.ega@crg.eu'\norg_logo_url = 'https://legacy.ega-archive.org/images/logo.png'\norg_info = ''\ndescription = \"This Beacon is based on synthetic data hosted at the <a href='https://ega-archive.org/datasets/EGAD00001003338'>EGA</a>. The dataset contains 2504 samples including genetic data based on 1K Genomes data, and 76 individual attributes and phenotypic data derived from UKBiobank.\"\nversion = 'v2.0'\nwelcome_url = 'https://beacon.ega-archive.org/'\nalternative_url = 'https://beacon-apis-demo.ega-archive.org/api/'\ncreate_datetime = '2021-11-29T12:00:00.000000'\nupdate_datetime = ''","granularity-conf":'default_beacon_granularity = "record"\nmax_beacon_granularity = "record"',"granularity-response":"dummy_user:\n- record","beacon-handovers":"beacon_handovers = [\n    {\n        handoverType: {\n            id: 'CUSTOM:000001',\n            label: 'Project description'\n        },\n        note: 'Project description',\n        url: 'https://www.nist.gov/programs-projects/genome-bottle'\n    }\n]"}[e];t&&navigator.clipboard.writeText(t).then((()=>{n((n=>({...n,[e]:!0}))),setTimeout((()=>n((n=>({...n,[e]:!1})))),1500)})).catch((e=>console.log(e)))};return(0,we.jsxs)("div",{className:"apiConfigContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/configuration",className:"no-undeline",children:"Beacon 2 PI API"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/configuration",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:"Configuration"})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h3",{children:"Beacon 2 Production Implementation API"}),(0,we.jsx)("h1",{children:"Configuration"}),(0,we.jsx)("p",{children:"Beacon2 PI API has two folders where a user can configure their beacons: conf for general configuration of the beacon and source for configuring the model of the beacon: e.g. entry types and the databases related to each entry type."}),(0,we.jsx)("h2",{id:"editing-beacon-info",children:"Editing your beacon information"}),(0,we.jsxs)("p",{children:["To show correctly your beacon\u2019s information you will need to edit both conf.py files from beacon and deploy folders.",(0,we.jsx)("br",{}),"To do so, edit the following variables:"]}),(0,we.jsxs)("div",{className:"codeSnippet",children:[(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["beacon_id = 'org.ega-archive.beacon-ri-demo'\n",(0,we.jsx)("br",{}),"beacon_name = 'Beacon Reference Implementation demo'\n",(0,we.jsx)("br",{}),"api_version = 'v2.0.0'\n",(0,we.jsx)("br",{}),"uri = 'https://beacon-apis-demo.ega-archive.org/api/'\n",(0,we.jsx)("br",{}),"org_id = 'EGA'\n",(0,we.jsx)("br",{}),"org_name = 'European Genome-Phenome Archive (EGA)'\n",(0,we.jsx)("br",{}),"org_description = 'The European Genome-phenome Archive (EGA) is a service for permanent archiving and sharing of all types of personally identifiable genetic and phenotypic data resulting from biomedical research projects.'\n",(0,we.jsx)("br",{}),"org_adress = 'C/ Dr. Aiguader, 88\nPRBB Building\n08003 Barcelona, Spain'\n",(0,we.jsx)("br",{}),"org_welcome_url = 'https://ega-archive.org/'\n",(0,we.jsx)("br",{}),"org_contact_url = 'mailto:beacon.ega@crg.eu'\n",(0,we.jsx)("br",{}),"org_logo_url = 'https://legacy.ega-archive.org/images/logo.png'\n",(0,we.jsx)("br",{}),"org_info = ''\n",(0,we.jsx)("br",{}),(0,we.jsxs)("code",{children:['description = "This Beacon is based on synthetic data hosted at the ',(0,we.jsx)("a",{href:"https://ega-archive.org/datasets/EGAD00001003338",target:"_blank",rel:"noopener noreferrer",children:"EGA"}),'. The dataset contains 2504 samples including genetic data based on 1K Genomes data, and 76 individual attributes and phenotypic data derived from UKBiobank."']}),(0,we.jsx)("br",{}),"version = 'v2.0'\n",(0,we.jsx)("br",{}),"welcome_url = 'https://beacon.ega-archive.org/'\n",(0,we.jsx)("br",{}),"alternative_url = 'https://beacon-apis-demo.ega-archive.org/api/'\n",(0,we.jsx)("br",{}),"create_datetime = '2021-11-29T12:00:00.000000'\n",(0,we.jsx)("br",{}),"update_datetime = ''",(0,we.jsx)("br",{})]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("beacon-info"),children:e["beacon-info"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:"Note that this is the info that will be shown in the /info endpoint for your beacon."})]})]}),(0,we.jsx)("h2",{id:"managing-dataset-permissions",children:"Managing dataset permissions"}),(0,we.jsx)("p",{children:"There are 3 possible levels of beacon security for a dataset: public, registered and controlled."}),(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"A public dataset is a dataset that will be returned in a beacon query without an authentication token.\xa0"}),(0,we.jsx)("li",{children:"A registered dataset is a dataset that will be shown after a user sends a valid token (in other words, is logged in)."}),(0,we.jsx)("li",{children:"A controlled dataset is a dataset that needs a user to send a valid token for authentication and the user needs to be allowed to query that dataset."})]}),(0,we.jsx)("br",{}),(0,we.jsx)("h4",{id:"edit-yml-files",children:"Edit the .yml files inside permissions/datasets"}),(0,we.jsxs)("p",{children:["For a ",(0,we.jsx)("b",{children:"public dataset"}),", edit the public_datasets.yml file inside permissions folder and add the dataset with the exact name it appears in the id from its pertinent record in datasets mongo collection:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["public_datasets:",(0,we.jsx)("br",{}),"-dataset_id"]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("method-1-public"),children:e["method-1-public"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{children:["For a ",(0,we.jsx)("b",{children:"registered dataset"}),", edit the registered_datasets.yml file inside permissions folder and add the dataset with the exact name it appears in the id from its pertinent record in datasets mongo collection:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["registered_datasets: ",(0,we.jsx)("br",{})," -dataset_id"]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("method-1-registered"),children:e["method-1-registered"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{children:["For a ",(0,we.jsx)("b",{children:"controlled dataset"}),", edit the controlled_datasets.yml file inside permissions folder and add the dataset inside the username you wish to give permissions with the exact name for the dataset as it appears in the id from its pertinent record in datasets mongo collection and the exact name the user has in Keycloak:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["username: ",(0,we.jsx)("br",{}),"-dataset_id"]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("method-1-controlled"),children:e["method-1-controlled"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h2",{id:"supplying-aai-credentials",children:"Supplying AAI credentials for your Identity Provider (IDP)"}),(0,we.jsxs)("p",{children:["For making any IDP (based on OpenID and Oauthv2) work with beacon, you will need to add the ",(0,we.jsx)("b",{children:"client ID "}),"and ",(0,we.jsx)("b",{children:"client secret"})," ","for each IDP in a new file you have to create inside auth/idp_providers folder (for each IDP). This file must have a name with an .env extension (e.g. something.env) and needs to have the following variables:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["CLIENT_ID=beacon ",(0,we.jsx)("br",{}),"CLIENT_SECRET='b26ca0f9-1137-4bee-b453-ee51eefbe7ba' ",(0,we.jsx)("br",{}),"USER_INFO='http://idp:8080/auth/realms/Beacon/protocol/openid-connect/userinfo'"," ",(0,we.jsx)("br",{}),"INTROSPECTION='http://idp:8080/auth/realms/Beacon/protocol/openid-connect/token/introspect'"," ",(0,we.jsx)("br",{}),"ISSUER='http://localhost:8080/auth/realms/Beacon' ",(0,we.jsx)("br",{}),"JWKS_URL='http://idp:8080/auth/realms/Beacon/protocol/openid-connect/certs'"]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("aai-env"),children:e["aai-env"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"After that, make sure you build your beacon container again:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker-compose up -d --build beaconprod"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("beacon-rebuild"),children:e["beacon-rebuild"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h2",{id:"handling-cors",children:"Handling CORS"}),(0,we.jsxs)("p",{children:["To avoid CORS using beacon and the frontend or a third-party authorization site like Keycloak, you will have to include all these URLs inside variable cors_urls in ",(0,we.jsx)("b",{children:"beacon/conf/conf.py file"}),".",(0,we.jsx)("br",{}),"Example usage:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'cors_urls = ["http://localhost:3000","https://cancer-beacon-demo.ega-archive.org", "https://beacon-network-demo2.ega-archive.org", "https://beacon.ega-archive.org"]'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("cors-main"),children:e["cors-main"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h2",{id:"setting-default-granularity",children:"Setting the default granularity"}),(0,we.jsxs)("p",{children:["Beacon PI can only limit the granularity for all the beacon at the moment. If you want to limit the granularity for all users across all datasets, you can point it in ",(0,we.jsx)("b",{children:"becon/conf/conf.py"}),", by editing the next variables:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'max_beacon_granularity = "record"'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("granularity-conf"),children:e["granularity-conf"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h2",{id:"adding-beacon-handovers",children:"Adding beacon handovers"}),(0,we.jsxs)("p",{children:["A handover can be added to the general beacon response or to the response per dataset.",(0,we.jsx)("br",{}),"In any of the cases, you have to modify the"," ",(0,we.jsx)("b",{children:"beacon/utils/handovers.py"})," file and you can write your handover there following the Beacon v2 spec response for handovers, e.g."]}),(0,we.jsx)("div",{className:"codeSnippet custom-codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["handover_1= {",(0,we.jsx)("br",{}),'\xa0\xa0"note": "Description of the handover",',(0,we.jsx)("br",{}),'\xa0\xa0"url": "Link for the handover",',(0,we.jsx)("br",{}),'\xa0\xa0"handoverType": {',(0,we.jsx)("br",{}),'\xa0\xa0\xa0\xa0"id": "NCIT:C189151",',(0,we.jsx)("br",{}),'\xa0\xa0\xa0\xa0"label": "Study Data Repository" \xa0\xa0} }']}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("beacon-handovers"),children:e["beacon-handovers"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h4",{children:"General response"}),(0,we.jsx)("p",{children:"After doing that if you want to just show the handover in general (not per dataset) you will need to include it in the list_of_handovers array:"}),(0,we.jsx)("div",{className:"codeSnippet custom-codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"list_of_handovers=[handover_1]"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("beacon-handovers"),children:e["beacon-handovers"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h4",{children:"Handover per dataset"}),(0,we.jsx)("p",{children:"If you want to add the handover per dataset, then you will need to create first a dictionary per dataset with the handover and the dataset id, like this:"})," ",(0,we.jsx)("div",{className:"codeSnippet custom-codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'dataset1_handover = { "dataset": dataset1_id, "handover": handover_1 }'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("beacon-handovers"),children:e["beacon-handovers"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"And then you will need to add it in the next array:"})," ",(0,we.jsx)("div",{className:"codeSnippet custom-codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"list_of_handovers_per_dataset=[dataset1_handover]"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("beacon-handovers"),children:e["beacon-handovers"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h2",{id:"configuring-entry-types",children:"Configuring your entry types"}),(0,we.jsxs)("p",{children:["If you go to the source folder inside beacon folder, you will find a file called ",(0,we.jsx)("b",{children:"manage.py"})," that you will need to edit in order to tell the API what is implemented in your beacon for each entry type and in what database you have the data related to each entry_type."," ",(0,we.jsx)("br",{}),(0,we.jsx)("br",{}),"In order to do this, you will just need to put a True or False response to what granularity you have implemented per each entryType. See the next example:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"g_variants={ 'granularity': {\n                  'boolean': True,\n                  'count': True,\n                  'record': True },\n          'singleEntryUrl': True,\n          'endpoints': {'analysis': True, 'biosample': True, 'individual': True, 'run': True },\n          'testMode': True,\n          'database': 'mongo' }"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("granularity-conf"),children:e["granularity-conf"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{children:["If none of the granularities are True, then beacon will not look at the other variables (as entryType will be programmed as not implemented). If some of the granularities are True, then, singleEntryUrl will tell if a beacon has the id queries implemented for this entryType and for each endpoint inside endpoints, which cross query with this entryType and the id parameter is implemented as well. The ",(0,we.jsx)("b",{children:"testMode"})," is to point if the entryType can be queried using testMode and the database field will tell the users which database has the data for that particular entryType."]}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/europe_logo.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:"Bear in mind that the name of the database variable needs to be the exact same name that the name for the folder it corresponds to in beacon/connections."})]}),(0,we.jsx)("h2",{id:"setting-logs",children:"Setting the logs"}),(0,we.jsxs)("p",{children:["In this production instance of beacon, you can set the level of the logs you want to output and where do you want to see the output in."," ",(0,we.jsx)("br",{}),(0,we.jsx)("br",{}),"All the outputs will be streamed but if you define a path for a log file (file with extension .log) you will be able to have all the history of logs for your beacon saved in this file. ",(0,we.jsx)("br",{}),(0,we.jsx)("br",{}),"For setting the level of the logs, specify one amongst NOTSET, DEBUG, ERROR, INFO in the variable level (after logging.):"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"level=logging.NOTSET"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("granularity-conf"),children:e["granularity-conf"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"For setting the path to the file, define it in the variable log_file:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"log_file=None"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("granularity-conf"),children:e["granularity-conf"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/europe_logo.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:(0,we.jsxs)("p",{className:"note-paragraph",children:["To save the logs in a file out of docker, you will need to create a volume for the file. There is a default log file volume defined in the docker-compose, which is -",(0,we.jsx)("b",{children:"./beacon/logs/logs.log:/beacon/logs/logs.log "}),", hence no need to create any other volume if you use this same file for logs."]})})]}),(0,we.jsx)("h2",{id:"TLS-configuration",children:"TLS configuration"}),(0,we.jsxs)("p",{children:["To enable TLS for the Becaon API set beacon_server_crt and beacon_server_key to the full paht of the server certificate and server key in ",(0,we.jsx)("b",{children:"beacon/conf/conf.py"})," file."]}),(0,we.jsx)("h2",{id:"TLS-secured-MongoDB",children:"TLS secured MongoDB"}),(0,we.jsxs)("p",{children:["Edit the file\xa0",(0,we.jsx)("b",{children:"beacon/connections/mongo/conf.py "}),"\xa0and set\xa0database_certificate\xa0to the full path to the client certificate. If a private CA is used also set the\xa0database_cafile\xa0to the full path to the CA certificate.",(0,we.jsx)("li",{children:'The MongoDB client certificate should be in the combined PEM format\xa0client.key + "\\n" + client.crt'})]}),(0,we.jsx)("br",{}),(0,we.jsx)("br",{})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]})},lc=()=>{const[e,n]=(0,a.useState)({}),t=(e,t)=>{navigator.clipboard.writeText(e).then((()=>{n((e=>({...e,[t]:!0}))),setTimeout((()=>{n((e=>({...e,[t]:!1})))}),1500)})).catch(console.error)};return(0,we.jsxs)("div",{className:"queryingApiContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/pi-querying-the-api",className:"no-undeline",children:"Beacon 2 PI API"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/pi-querying-the-api",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:"Querying the API"})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h3",{children:"Beacon 2 Production Implementation API"}),(0,we.jsx)("h1",{children:"Querying the API"}),(0,we.jsx)("p",{children:"Beacon PI accepts two types of request methods: GET and POST."}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsxs)("div",{children:["You can find further introduction to Beacon parameters in\xa0",(0,we.jsxs)("a",{href:"https://docs.genomebeacons.org/variant-queries/",target:"_blank",rel:"noopener noreferrer",children:[" ","Beacon v2 Documentation"]}),"."]})]}),(0,we.jsx)("h2",{id:"pi-get-method",children:"GET Method"}),(0,we.jsx)("p",{children:"GET method is a bit more limited as this method only accepts request parameters, three generic parameters and ontology filters."}),(0,we.jsx)("h6",{children:"Query Parameters"}),(0,we.jsx)("p",{children:"These are the defined set of parameters (key-value pair) attached to the end of a URL used to provide additional information to a web server when making requests."}),(0,we.jsxs)("table",{children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:"Generic parameters"}),(0,we.jsx)("th",{children:"Type"}),(0,we.jsx)("th",{children:"Examples"})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 requestedSchema"}),(0,we.jsx)("td",{children:"String"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"requestedSchema=ga4gh-service-info-v1.0"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("requestedSchema=ga4gh-service-info-v1.0","requestedSchema"),children:e.requestedSchema?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 skip"}),(0,we.jsxs)("td",{children:["integer (Skip) \u2265 0",(0,we.jsx)("br",{}),"Default: 0"]}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"skip=0"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("skip=0","skip"),children:e.skip?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 limit"}),(0,we.jsxs)("td",{children:["integer (Limit) \u2265 0",(0,we.jsx)("br",{}),"Default: 10",(0,we.jsx)("br",{}),"Max: 100"]}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"limit=10"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("limit=10","limit"),children:e.limit?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]})]})]}),(0,we.jsxs)("table",{children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:"Request parameters"}),(0,we.jsx)("th",{children:"Type"}),(0,we.jsx)("th",{children:"Examples"})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 start"}),(0,we.jsxs)("td",{children:["integer <int64> \u2265 0",(0,we.jsx)("br",{}),"(or comma-separated integers)"]}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"start=16050074"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("start=16050074","start"),children:e.start?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 end"}),(0,we.jsxs)("td",{children:["integer <int64> \u2265 0",(0,we.jsx)("br",{}),"(or comma-separated integers)"]}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"end=16052080"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("end=16052080","end"),children:e.end?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 assemblyId"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"assemblyId=GRCh38"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("assemblyId=GRCh38","assemblyId"),children:e.assemblyId?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 referenceName"}),(0,we.jsx)("td",{children:"integer <int64> \u2265 0"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"referenceName=22"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("referenceName=22","referenceName"),children:e.referenceName?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 referenceBases"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"referenceBases=G"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("referenceBases=G","referenceBases"),children:e.referenceBases?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 alternateBases"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"alternateBases=A"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("alternateBases=A","alternateBases"),children:e.alternateBases?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 variantMinLength"}),(0,we.jsx)("td",{children:"integer <int64> \u2265 0"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"variantMinLength=5"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("variantMinLength=5","variantMinLength"),children:e.variantMinLength?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 variantMaxLength"}),(0,we.jsx)("td",{children:"integer <int64> \u2265 0"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"variantMaxLength=2"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("variantMaxLength=2","variantMaxLength"),children:e.variantMaxLength?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 geneId"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"geneId=CCNL2"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("geneId=CCNL2","geneId"),children:e.geneId?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 aminoacidChange"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"aminoacidChange=p.Thr130Met"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("aminoacidChange=p.Thr130Met","aminoacidChange"),children:e.aminoacidChange?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 variantType"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"variantType=SNP"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("variantType=SNP","variantType"),children:e.variantType?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 genomicAlleleShortForm"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"genomicAlleleShortForm=NC_000022.11:g.16050075A>G"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("genomicAlleleShortForm=NC_000022.11:g.16050075A&gt;G","genomicAlleleShortForm"),children:e.genomicAlleleShortForm?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 clinicalRelevance"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"clinicalRelevance=pathogenic"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("clinicalRelevance=pathogenic","clinicalRelevance"),children:e.clinicalRelevance?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]})]})]}),(0,we.jsxs)("table",{children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:"Filters"}),(0,we.jsx)("th",{children:"Type"}),(0,we.jsx)("th",{children:"Examples"})]})}),(0,we.jsx)("tbody",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 filters"}),(0,we.jsxs)("td",{children:["string",(0,we.jsx)("br",{}),"(comma-separated strings of ontologies)"]}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"filters=NCIT:C16576,NCIT:C42331"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("filters=NCIT:C16576,NCIT:C42331","filters"),children:e.filters?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]})})]}),(0,we.jsx)("h2",{id:"pi-get-examples",children:"GET query examples"}),(0,we.jsx)("h5",{className:"cursive",children:"Beacon Sequence Query"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"http://localhost:5050/api/g_variants?start=16050606&alternateBases=A&referenceBases=G&referenceName=22&assemblyId=GRCh37"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("http://localhost:5050/api/g_variants?start=16050606&alternateBases=A&referenceBases=G&referenceName=22&assemblyId=GRCh37","beacon-sequence-query"),children:e["beacon-sequence-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Beacon Range Query"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"http://localhost:5050/api/g_variants?start=16050606&end=16050607&alternateBases=A&referenceBases=G&referenceName=22&assemblyId=GRCh37"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("http://localhost:5050/api/g_variants?start=16050606&end=16050607&alternateBases=A&referenceBases=G&referenceName=22&assemblyId=GRCh37","beacon-range-query"),children:e["beacon-range-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Beacon GeneId Query"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"http://localhost:5050/api/g_variants?geneId=KRAS&variantType=INDEL"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("http://localhost:5050/api/g_variants?geneId=KRAS&variantType=INDEL","beacon-geneid-query"),children:e["beacon-geneid-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Beacon Bracket Query"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"http://localhost:5050/api/g_variants?referenceName=22&start=16050074,16050076&end=16050090,16050096&variantType=SNP&assemblyId=GRCh38"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("http://localhost:5050/api/g_variants?referenceName=22&start=16050074,16050076&end=16050090,16050096&variantType=SNP&assemblyId=GRCh38","beacon-bracket-query"),children:e["beacon-bracket-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Genomic Allele Query"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"http://localhost:5050/api/g_variants?genomicAlleleShortForm= NC_000021.10:g.9411383C>T"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("http://localhost:5050/api/g_variants?genomicAlleleShortForm= NC_000021.10:g.9411383C>T","genomic-allele-query"),children:e["genomic-allele-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Aminoacid Change Query"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"http://localhost:5050/api/g_variants?geneId=KRAS&aminoacidChange=Thr130Met"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("http://localhost:5050/api/g_variants?geneId=KRAS&aminoacidChange=Thr130Met","aminoacid-change-query"),children:e["aminoacid-change-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h2",{id:"pi-post-method",children:"POST Method"}),(0,we.jsx)("p",{children:"In POST method requests, the endpoints accept all generic parameters and all the filters, apart from the request parameters."}),(0,we.jsx)("h6",{className:"underline",children:"Query Parameters"}),(0,we.jsxs)("table",{className:"codeSnippet",children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:"Generic parameters"}),(0,we.jsx)("th",{children:"Type"}),(0,we.jsx)("th",{children:"Examples"})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 includeResultsetResponses"}),(0,we.jsxs)("td",{children:["string ",(0,we.jsx)("br",{}),'Default: "HIT"',(0,we.jsx)("br",{}),'Enum: "ALL" "HIT" "MISS" "NONE"']}),(0,we.jsx)("td",{})]}),(0,we.jsxs)("tr",{children:[(0,we.jsxs)("td",{children:["\u2022 pagination object (Pagination) Items:",(0,we.jsx)("li",{children:"skip"}),(0,we.jsx)("li",{children:"limit"})]}),(0,we.jsxs)("td",{children:[(0,we.jsx)(we.Fragment,{}),"integer (Skip) \u2265 0",(0,we.jsx)("br",{}),"Default: 0 ",(0,we.jsx)("br",{}),(0,we.jsx)("br",{}),"integer (Limit) \u2265 0",(0,we.jsx)("br",{}),"Default: 10",(0,we.jsx)("br",{}),"Max: 100"]}),(0,we.jsxs)("td",{children:[(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"skip=0"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("skip=0","pagination-skip"),children:e["pagination-skip"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]}),(0,we.jsx)("br",{}),(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"limit=10"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("limit=10","pagination-limit"),children:e["pagination-limit"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})]})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 requestedGranularity"}),(0,we.jsxs)("td",{children:["string (Granularity)",(0,we.jsx)("br",{}),'Default: "boolean"',(0,we.jsx)("br",{}),'Enum: "boolean" "count" "record"']}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:'requestedGranularity="record"'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('requestedGranularity="record"',"requestedGranularity"),children:e.requestedGranularity?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 testMode"}),(0,we.jsxs)("td",{children:["boolean (TestMode)",(0,we.jsx)("br",{}),"Default: false ",(0,we.jsx)("br",{})]}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"testMode=false"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("testMode=false","testMode"),children:e.testMode?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]})]})]}),(0,we.jsxs)("table",{children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:"Request parameters"}),(0,we.jsx)("th",{children:"Type"}),(0,we.jsx)("th",{children:"Examples"})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 start"}),(0,we.jsxs)("td",{children:["integer <int64> \u2265 0",(0,we.jsx)("br",{}),"(or comma-separated integers)"]}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"start=16050074"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("start=16050074","start"),children:e.start?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 end"}),(0,we.jsxs)("td",{children:["integer <int64> \u2265 0",(0,we.jsx)("br",{}),"(or comma-separated integers)"]}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"end=16052080"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("end=16052080","end"),children:e.end?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 assemblyId"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"assemblyId=GRCh38"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("assemblyId=GRCh38","assemblyId"),children:e.assemblyId?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 referenceName"}),(0,we.jsx)("td",{children:"integer <int64> \u2265 0"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"referenceName=22"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("referenceName=22","referenceName"),children:e.referenceName?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 referenceBases"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"referenceBases=G"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("referenceBases=G","referenceBases"),children:e.referenceBases?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 alternateBases"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"alternateBases=A"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("alternateBases=A","alternateBases"),children:e.alternateBases?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 variantMinLength"}),(0,we.jsx)("td",{children:"integer <int64> \u2265 0"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"variantMinLength=5"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("variantMinLength=5","variantMinLength"),children:e.variantMinLength?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 variantMaxLength"}),(0,we.jsx)("td",{children:"integer <int64> \u2265 0"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"variantMaxLength=2"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("variantMaxLength=2","variantMaxLength"),children:e.variantMaxLength?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 geneId"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"geneId=CCNL2"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("geneId=CCNL2","geneId"),children:e.geneId?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 aminoacidChange"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"aminoacidChange=p.Thr130Met"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("aminoacidChange=p.Thr130Met","aminoacidChange"),children:e.aminoacidChange?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 variantType"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"variantType=SNP"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("variantType=SNP","variantType"),children:e.variantType?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 genomicAlleleShortForm"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"genomicAlleleShortForm=NC_000022.11:g.16050075A>G"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("genomicAlleleShortForm=NC_000022.11:g.16050075A&gt;G","genomicAlleleShortForm"),children:e.genomicAlleleShortForm?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 clinicalRelevance"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsx)("code",{children:"clinicalRelevance=pathogenic"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("clinicalRelevance=pathogenic","clinicalRelevance"),children:e.clinicalRelevance?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]})]})]}),(0,we.jsxs)("table",{children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:"Filters"}),(0,we.jsx)("th",{children:"Type"}),(0,we.jsx)("th",{children:"Examples"})]})}),(0,we.jsx)("tbody",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"\u2022 filters"}),(0,we.jsxs)("td",{children:["Array of OntologyFilter ",(0,we.jsx)("br",{}),"(object) or ",(0,we.jsx)("br",{}),"AlphanumericFilter (object) ",(0,we.jsx)("br",{}),"or CustomFilter (object) ",(0,we.jsx)("br",{}),"(Filtering Term Element)"]}),(0,we.jsx)("td",{children:(0,we.jsxs)("div",{className:"codeSnippet-table",children:[(0,we.jsxs)("code",{children:["filters: [",(0,we.jsx)("br",{}),"\xa0\xa0","{",(0,we.jsx)("br",{}),"\xa0\xa0\xa0\xa0",'"id": "geographicOrigin",',(0,we.jsx)("br",{}),"\xa0\xa0\xa0\xa0",'"operator": "!",',(0,we.jsx)("br",{}),"\xa0\xa0\xa0\xa0",'"value": "England"',(0,we.jsx)("br",{}),"\xa0\xa0","}",(0,we.jsx)("br",{}),"]"]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('"filters": [\n  {\n    "id": "geographicOrigin",\n    "operator": "!",\n    "value": "England"\n  }\n]',"filters-example"),children:e["filters-example"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})})]})})]}),(0,we.jsx)("h2",{id:"pi-post-examples",children:"POST query examples"}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:"You can either make a query with a scope field for each of the filters or just not insert the scope in the filter object. If scope is not stated, then, the scope of the filter will be the same than the endpoint being queried."})]}),(0,we.jsx)("h5",{className:"cursive",children:"Beacon Range Query"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": { "apiVersion": "2.0" },\n    "query": {\n      "requestParameters": {"alternateBases": "G", "referenceBases": "A", "referenceName": "22", "start": [16050074], "end": [16050568], "variantType": "SNP", "assemblyId": "GRCh37"},\n      "filters": [],\n      "includeResultsetResponses": "HIT",\n      "pagination": { "skip": 0, "limit": 10 },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/g_variants'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": { "apiVersion": "2.0" },\n    "query": {\n      "requestParameters": {"alternateBases": "G", "referenceBases": "A", "referenceName": "22", "start": [16050074], "end": [16050568], "variantType": "SNP", "assemblyId": "GRCh37"},\n      "filters": [],\n      "includeResultsetResponses": "HIT",\n      "pagination": { "skip": 0, "limit": 10 },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/g_variants',"genomic-query"),children:e["genomic-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Simple CURIE based filters query"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": { "apiVersion": "2.0" },\n    "query": {\n      "requestParameters": {},\n      "filters": [ { "id": "NCIT:C20197", "scope": "individual" }, { "id": "NCIT:C42331", "scope": "individual" } ],\n      "includeResultsetResponses": "HIT",\n      "pagination": { "skip": 0, "limit": 10 },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": { "apiVersion": "2.0" },\n    "query": {\n      "requestParameters": {},\n      "filters": [ { "id": "NCIT:C20197", "scope": "individual" }, { "id": "NCIT:C42331", "scope": "individual" } ],\n      "includeResultsetResponses": "HIT",\n      "pagination": { "skip": 0, "limit": 10 },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals',"curie-query"),children:e["curie-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Numerical Age Query"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": { "apiVersion": "2.0" },\n    "query": {\n      "filters": [\n        {\n          "id": "diseases.ageOfOnset.iso8601duration",\n          "operator": ">",\n          "value": "75"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": { "skip": 0, "limit": 10 },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": { "apiVersion": "2.0" },\n    "query": {\n      "filters": [\n        {\n          "id": "diseases.ageOfOnset.iso8601duration",\n          "operator": ">",\n          "value": "75"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": { "skip": 0, "limit": 10 },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals',"numerical-query"),children:e["numerical-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Alphanumerical Value Query (Exact Value)"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": { "apiVersion": "2.0" },\n    "query": {\n      "filters": [\n        {\n          "id": "geographicOrigin",\n          "operator": "=",\n          "value": "England"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": { "skip": 0, "limit": 10 },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": { "apiVersion": "2.0" },\n    "query": {\n      "filters": [\n        {\n          "id": "geographicOrigin",\n          "operator": "=",\n          "value": "England"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": { "skip": 0, "limit": 10 },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals',"exact-value-query"),children:e["exact-value-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Alphanumerical Value Query (Like Value)"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": { "apiVersion": "2.0" },\n    "query": {\n      "filters": [\n        {\n          "id": "geographicOrigin",\n          "operator": "=",\n          "value": "%land%"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": { "skip": 0, "limit": 10 },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": { "apiVersion": "2.0" },\n    "query": {\n      "filters": [\n        {\n          "id": "geographicOrigin",\n          "operator": "=",\n          "value": "%land%"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": { "skip": 0, "limit": 10 },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals',"exact-value-query"),children:e["exact-value-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Alphanumerical Value Query (NOT Value)"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": { "apiVersion": "2.0" },\n    "query": {\n      "filters": [\n        {\n          "id": "geographicOrigin",\n          "operator": "!",\n          "value": "%land%"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": { "skip": 0, "limit": 10 },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": { "apiVersion": "2.0" },\n    "query": {\n      "filters": [\n        {\n          "id": "geographicOrigin",\n          "operator": "!",\n          "value": "%land%"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": { "skip": 0, "limit": 10 },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals',"not-value-query"),children:e["not-value-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Semantic Similarity Query"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": { "apiVersion": "2.0" },\n    "query": {\n      "filters": [\n        {\n          "id": "NCIT:C16576",\n          "similarity": "high"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": { "skip": 0, "limit": 10 },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n    "meta": { "apiVersion": "2.0" },\n    "query": {\n      "filters": [\n        {\n          "id": "NCIT:C16576",\n          "similarity": "high"\n        }\n      ],\n      "includeResultsetResponses": "HIT",\n      "pagination": { "skip": 0, "limit": 10 },\n      "testMode": false,\n      "requestedGranularity": "record"\n    }\n  }\' \\\n  http://localhost:5050/api/individuals',"similarity-query"),children:e["similarity-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Custom Filter Query"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n  "meta": { "apiVersion": "2.0" },\n  "query": {\n    "filters": [{"id": "sampleOriginType:blood", "scope": "biosample"}],\n    "includeResultsetResponses": "HIT",\n    "pagination": { "skip": 0, "limit": 10 },\n    "testMode": false,\n    "requestedGranularity": "record"\n  }\n}\' \\\n  http://localhost:5050/api/biosamples'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n  "meta": { "apiVersion": "2.0" },\n  "query": {\n    "filters": [{"id": "sampleOriginType:blood", "scope": "biosample"}],\n    "includeResultsetResponses": "HIT",\n    "pagination": { "skip": 0, "limit": 10 },\n    "testMode": false,\n    "requestedGranularity": "record"\n  }\n}\' \\\n  http://localhost:5050/api/biosamples',"custom-filter-query"),children:e["custom-filter-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Join Genomic Query"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n  "meta": { "apiVersion": "2.0" },\n  "query": {\n    "requestParameters": {\n      "alternateBases": "G",\n      "referenceBases": "A",\n      "referenceName": "22",\n      "start": [16050074],\n      "end": [16050568],\n      "variantType": "SNP",\n      "assemblyId": "GRCh37"\n    },\n    "filters": [\n      { "id": "NCIT:C20197", "scope": "individual" },\n      { "id": "NCIT:C42331", "scope": "individual" }\n    ],\n    "includeResultsetResponses": "HIT",\n    "pagination": { "skip": 0, "limit": 100 },\n    "testMode": false,\n    "requestedGranularity": "record"\n  }\n}\' \\\n  http://localhost:5050/api/g_variants'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n  "meta": { "apiVersion": "2.0" },\n  "query": {\n    "requestParameters": {\n      "alternateBases": "G",\n      "referenceBases": "A",\n      "referenceName": "22",\n      "start": [16050074],\n      "end": [16050568],\n      "variantType": "SNP",\n      "assemblyId": "GRCh37"\n    },\n    "filters": [\n      { "id": "NCIT:C20197", "scope": "individual" },\n      { "id": "NCIT:C42331", "scope": "individual" }\n    ],\n    "includeResultsetResponses": "HIT",\n    "pagination": { "skip": 0, "limit": 100 },\n    "testMode": false,\n    "requestedGranularity": "record"\n  }\n}\' \\\n  http://localhost:5050/api/g_variants',"join-genomic-query"),children:e["join-genomic-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Join Individuals Query"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n  "meta": { "apiVersion": "2.0" },\n  "query": {\n    "requestParameters": {\n      "alternateBases": "G",\n      "referenceBases": "A",\n      "referenceName": "22",\n      "start": [16050074],\n      "end": [16050568],\n      "variantType": "SNP",\n      "assemblyId": "GRCh37"\n    },\n    "filters": [\n      { "id": "NCIT:C20197", "scope": "individual" },\n      { "id": "NCIT:C42331", "scope": "individual" }\n    ],\n    "includeResultsetResponses": "HIT",\n    "pagination": { "skip": 0, "limit": 10 },\n    "testMode": false,\n    "requestedGranularity": "record"\n  }\n}\' \\\n  http://localhost:5050/api/individuals'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('curl \\\n  -H \'Content-Type: application/json\' \\\n  -X POST \\\n  -d \'{\n  "meta": { "apiVersion": "2.0" },\n  "query": {\n    "requestParameters": {\n      "alternateBases": "G",\n      "referenceBases": "A",\n      "referenceName": "22",\n      "start": [16050074],\n      "end": [16050568],\n      "variantType": "SNP",\n      "assemblyId": "GRCh37"\n    },\n    "filters": [\n      { "id": "NCIT:C20197", "scope": "individual" },\n      { "id": "NCIT:C42331", "scope": "individual" }\n    ],\n    "includeResultsetResponses": "HIT",\n    "pagination": { "skip": 0, "limit": 10 },\n    "testMode": false,\n    "requestedGranularity": "record"\n  }\n}\' \\\n  http://localhost:5050/api/individuals',"join-individuals-query"),children:e["join-individuals-query"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h5",{className:"cursive",children:"Query with Access Token"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'curl \\\n  -H \'Content-Type: application/json\' \\\n  -H \'Authorization: Bearer <access_token>\' \\\n  -X POST \\\n  -d \'{\n  "meta": { "apiVersion": "2.0" },\n  "query": {\n    "requestParameters": {},\n    "filters": [\n      { "id": "geographicOrigin", "operator": "=", "value": "%land%" }\n    ],\n    "includeResultsetResponses": "HIT",\n    "pagination": { "skip": 0, "limit": 10 },\n    "testMode": false,\n    "requestedGranularity": "record"\n  }\n}\' \\\n  http://localhost:5050/api/individuals'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t('curl \\\n  -H \'Content-Type: application/json\' \\\n  -H \'Authorization: Bearer <access_token>\' \\\n  -X POST \\\n  -d \'{\n  "meta": { "apiVersion": "2.0" },\n  "query": {\n    "requestParameters": {},\n    "filters": [\n      { "id": "geographicOrigin", "operator": "=", "value": "%land%" }\n    ],\n    "includeResultsetResponses": "HIT",\n    "pagination": { "skip": 0, "limit": 10 },\n    "testMode": false,\n    "requestedGranularity": "record"\n  }\n}\' \\\n  http://localhost:5050/api/individuals',"query-with-token"),children:e["query-with-token"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h6",{className:"underline",children:"List of endpoints"}),(0,we.jsx)("h5",{className:"cursive",children:"Endpoints that accept all query parameters"}),(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"/api/g_variants"}),(0,we.jsx)("li",{children:"/api/individuals"})]}),(0,we.jsx)("h5",{className:"cursive",children:"Endpoints that only accept filters and generic parameters"}),(0,we.jsxs)("ul",{className:"endpoint-list",children:[(0,we.jsx)("li",{children:"/api/analyses"}),(0,we.jsx)("li",{children:"/api/biosamples"}),(0,we.jsx)("li",{children:"/api/cohorts"}),(0,we.jsx)("li",{children:"/api/datasets"}),(0,we.jsx)("li",{children:"/api/runs"})]}),(0,we.jsx)("h5",{className:"cursive",children:"Endpoints that only accept generic parameters"}),(0,we.jsxs)("ul",{className:"endpoint-list",children:[(0,we.jsx)("li",{children:"/api/filtering_terms"}),(0,we.jsxs)("li",{children:["/api/analyses/","{id}"]}),(0,we.jsxs)("li",{children:["/api/analyses/","{id}","/g_variants"]}),(0,we.jsx)("li",{children:"/api/analyses/filtering_terms"}),(0,we.jsxs)("li",{children:["/api/biosamples/","{id}"]}),(0,we.jsxs)("li",{children:["/api/biosamples/","{id}","/analyses"]}),(0,we.jsxs)("li",{children:["/api/biosamples/","{id}","/g_variants"]}),(0,we.jsxs)("li",{children:["/api/biosamples/","{id}","/runs"]}),(0,we.jsx)("li",{children:"/api/biosamples/filtering_terms"}),(0,we.jsxs)("li",{children:["/api/cohorts/","{id}"]}),(0,we.jsxs)("li",{children:["/api/cohorts/","{id}","/analyses"]}),(0,we.jsxs)("li",{children:["/api/cohorts/","{id}","/biosamples"]}),(0,we.jsxs)("li",{children:["/api/cohorts/","{id}","/g_variants"]}),(0,we.jsxs)("li",{children:["/api/cohorts/","{id}","/individuals"]}),(0,we.jsxs)("li",{children:["/api/cohorts/","{id}","/runs"]}),(0,we.jsx)("li",{children:"/api/cohorts/filtering_terms"}),(0,we.jsxs)("li",{children:["/api/datasets/","{id}"]}),(0,we.jsxs)("li",{children:["/api/datasets/","{id}","/analyses"]}),(0,we.jsxs)("li",{children:["/api/datasets/","{id}","/biosamples"]}),(0,we.jsxs)("li",{children:["/api/datasets/","{id}","/g_variants"]}),(0,we.jsxs)("li",{children:["/api/datasets/","{id}","/individuals"]}),(0,we.jsxs)("li",{children:["/api/datasets/","{id}","/runs"]}),(0,we.jsx)("li",{children:"/api/datasets/filtering_terms"}),(0,we.jsxs)("li",{children:["/api/g_variants/","{id}"]}),(0,we.jsxs)("li",{children:["/api/g_variants/","{id}","/analyses"]}),(0,we.jsxs)("li",{children:["/api/g_variants/","{id}","/biosamples"]}),(0,we.jsxs)("li",{children:["/api/g_variants/","{id}","/individuals"]}),(0,we.jsxs)("li",{children:["/api/g_variants/","{id}","/runs"]}),(0,we.jsx)("li",{children:"/api/g_variants/filtering_terms"}),(0,we.jsxs)("li",{children:["/api/individuals/","{id}"]}),(0,we.jsxs)("li",{children:["/api/individuals/","{id}","/analyses"]}),(0,we.jsxs)("li",{children:["/api/individuals/","{id}","/biosamples"]}),(0,we.jsxs)("li",{children:["/api/individuals/","{id}","/g_variants"]}),(0,we.jsxs)("li",{children:["/api/individuals/","{id}","/runs"]}),(0,we.jsx)("li",{children:"/api/individuals/filtering_terms"}),(0,we.jsxs)("li",{children:["/api/runs/","{id}"]}),(0,we.jsxs)("li",{children:["/api/runs/","{id}","/analyses"]}),(0,we.jsxs)("li",{children:["/api/runs/","{id}","/g_variants"]}),(0,we.jsx)("li",{children:"/api/runs/filtering_terms"})]}),(0,we.jsx)("h5",{className:"cursive",children:"Endpoints that don\u2019t accept query parameters"}),(0,we.jsxs)("ul",{className:"endpoint-list",children:[(0,we.jsx)("li",{children:"/api"}),(0,we.jsx)("li",{children:"/api/info"}),(0,we.jsx)("li",{children:"/api/service-info"}),(0,we.jsx)("li",{children:"/api/configuration"}),(0,we.jsx)("li",{children:"/api/entry_types"}),(0,we.jsx)("li",{children:"/api/map"})]})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]})},cc=()=>{const[e,n]=(0,a.useState)({}),t=e=>{const t={"cloning-repository":"git clone https://github.com/EGA-archive/beacon2-pi-api.git","cd-deploy":"docker compose up -d --build","create-network":"docker compose up -d \u2013build beaconprod db","data-injection":['docker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/datasets.json --collection datasets','docker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/individuals.json --collection individuals','docker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/cohorts.json --collection cohorts','docker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/analyses.json --collection analyses','docker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/biosamples.json --collection biosamples','docker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/genomicVariations.json --collection genomicVariations','docker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/runs.json --collection runs','docker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/targets.json --collection targets','docker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/caseLevelData.json --collection caseLevelData'].join("\n"),"data-indexing":"docker exec beaconprod python -m beacon.connections.mongo.reindex"}[e];t&&navigator.clipboard.writeText(t).then((()=>{n((n=>({...n,[e]:!0}))),setTimeout((()=>n((n=>({...n,[e]:!1})))),1500)})).catch((e=>console.log(e)))};return(0,we.jsxs)("div",{className:"deploymentContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/pi-manual-deployment",className:"no-undeline",children:"Beacon 2 PI API"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/pi-manual-deployment",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:" Manual Deployment"})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h3",{children:"Beacon 2 Production Implementation API"}),(0,we.jsx)("h1",{children:"Manual Deployment"}),(0,we.jsx)("h2",{id:"cloning-repository",children:"Cloning the repository"}),(0,we.jsx)("p",{children:"First of all, start by cloning the GitHub repository in your system."}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["git clone"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-pi-api.git",children:"https://github.com/EGA-archive/beacon2-pi-api.git"})]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("cloning-repository"),children:e["cloning-repository"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h2",{id:"creating-the-containers",children:"Creating the containers"}),(0,we.jsx)("p",{children:"Make sure the next list of ports are free of use in your system:"}),(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"27017 \u2192 mongo"}),(0,we.jsx)("li",{children:"8081 \u2192 mongo-express (optional)"}),(0,we.jsx)("li",{children:"5050 \u2192 beacon"}),(0,we.jsx)("li",{children:"8080 \u2192 Keycloak (optional)"}),(0,we.jsx)("li",{children:"9991 \u2192 Keycloak SSL (optional)"})]}),(0,we.jsx)("p",{children:"Light up the containers from the deploy folder."}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker compose up -d --build"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("cd-deploy"),children:e["cd-deploy"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{children:["If you wish to light up only some of the services, not all, just write the name of the services in the ",(0,we.jsx)("b",{children:"docker-compose.yml"})," file right after the command (e.g. for having the mandatory services only):"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker compose up -d \u2013build beaconprod db"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("create-network"),children:e["create-network"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"If the containers are built correctly:"}),(0,we.jsxs)("ul",{children:[(0,we.jsxs)("li",{children:["The API will run in"," ",(0,we.jsx)("a",{href:"http://localhost:5050/api",target:"_blank",rel:"noopener noreferrer",children:"http://localhost:5050/api"}),"."]}),(0,we.jsxs)("li",{children:["The mongo-express UI will run in"," ",(0,we.jsx)("a",{href:"http://localhost:8081",target:"_blank",rel:"noopener noreferrer",children:"http://localhost:8081"}),"."]}),(0,we.jsxs)("li",{children:["The Keycloak UI will run in"," ",(0,we.jsx)("a",{href:"http://localhost:8080/auth",target:"_blank",rel:"noopener noreferrer",children:"http://localhost:8080/auth"}),"."]})]}),(0,we.jsx)("h2",{id:"data-injection",children:"Data injection"}),(0,we.jsxs)("p",{children:["Copy all the available BFF files you have to the"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-pi-api/tree/main/beacon/connections/mongo/data",target:"_blank",rel:"noopener noreferrer",children:"data folder for mongo database"}),"."]}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:"Tip: A beacon can run without data or data for a single collection (e.g. only individuals)."})]}),(0,we.jsx)("p",{children:"Now execute the insertion of the files copied into the mongo database."}),(0,we.jsx)("div",{className:"codeSnippet codeSnippet-mongodb",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:['docker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/datasets.json --collection datasets\n',(0,we.jsx)("br",{}),'docker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/individuals.json --collection individuals\n',(0,we.jsx)("br",{}),'docker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/cohorts.json --collection cohorts\n',(0,we.jsx)("br",{}),'docker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/analyses.json --collection analyses\n',(0,we.jsx)("br",{}),'docker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/biosamples.json --collection biosamples\n',(0,we.jsx)("br",{}),'docker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/genomicVariations.json --collection genomicVariations\n',(0,we.jsx)("br",{}),'docker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/runs.json --collection runs\n',(0,we.jsx)("br",{}),'docker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/targets.json --collection targets\n',(0,we.jsx)("br",{}),'docker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/caseLevelData.json --collection caseLevelData']}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("data-injection"),children:e["data-injection"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol copySymbol-custom",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h2",{id:"data-indexing",children:"Data indexing"}),(0,we.jsx)("p",{children:"For the API to respond fast to the queries, you have to index your database. You can create the necessary indexes by running the next script:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker exec beaconprod python -m beacon.connections.mongo.reindex"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("data-indexing"),children:e["data-indexing"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:"Note that you will need to run this script each time you inject new data."})]})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]})},dc=()=>{const[e,n]=(0,a.useState)({}),t=e=>{const t={"extract-terms":"docker exec beaconprod python beacon/connections/mongo/extract_filtering_terms.py","manual-filtering-terms":'db.filtering_terms.insertMany([\n  {\n    "type": "alphanumeric",\n    "id": "libraryStrategy",\n    "scope": ["runs"]\n  }\n])',"get-descendant-terms":"docker exec beaconprod python beacon/connections/mongo/get_descendants.py"};t[e]&&navigator.clipboard.writeText(t[e]).then((()=>{n((n=>({...n,[e]:!0}))),setTimeout((()=>{n((n=>({...n,[e]:!1})))}),1500)})).catch(console.error)};return(0,we.jsxs)("div",{className:"apiConfigContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/filtering-terms",className:"no-undeline",children:"Beacon 2 PI API"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/filtering-terms",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:"Filtering Terms"})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h3",{children:"Beacon 2 Production Implementation API"}),(0,we.jsx)("h1",{children:"Filtering Terms"}),(0,we.jsx)("p",{children:"Filtering terms are metadata fields that allow users to query your Beacon more precisely \u2014 for example, by filtering results based on sequencing strategy (libraryStrategy), tissue type, disease, or other structured attributes. These terms often rely on ontologies to ensure consistent vocabulary across datasets."}),(0,we.jsx)("p",{children:"This section explains how to extract, add, and enhance filtering terms for your Beacon instance."}),(0,we.jsx)("h2",{id:"automatically-extract-filtering-terms",children:"Automatically Extract Filtering Terms"}),(0,we.jsxs)("p",{children:["If your data collections (e.g., runs, biosamples, etc.) already contain structured metadata using ontology terms (like NCIT, UBERON, EFO...), you can extract filtering terms automatically.",(0,we.jsx)("br",{})," ","This will populate the ",(0,we.jsx)("em",{children:"/filteringTerms"})," endpoint of your Beacon, enabling more advanced queries."]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker exec beaconprod python beacon/connections/mongo/extract_filtering_terms.py"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("extract-terms"),children:e["extract-terms"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("h2",{id:"manually-add-filtering-terms",children:["Manually Add Filtering Terms"," ",(0,we.jsx)("span",{className:"optional",children:"(optional)"})]}),(0,we.jsxs)("p",{children:["If you want to enable filtering for fields that aren\u2019t auto-detected or don\u2019t use ontologies (e.g., simple alphanumeric fields like"," ",(0,we.jsx)("em",{children:"libraryStrategy"}),"), you can add them manually."]}),(0,we.jsx)("p",{children:"Execute the following command:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["db.filtering_terms.insertMany([","\n","  ","{","\n","    ",' "type": "alphanumeric",',"\n","    ",' "id": "libraryStrategy",',"\n","    ",' "scope": ["runs"]',"\n","  ","}","\n","])"]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("manual-filtering-terms"),children:e["manual-filtering-terms"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol copySymbol-custom",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{children:["Field descriptions:",(0,we.jsxs)("ul",{children:[(0,we.jsxs)("li",{children:[(0,we.jsx)("b",{children:"type"}),": alphanumeric indicates this is a plain text value, not an ontology term."]}),(0,we.jsxs)("li",{children:[(0,we.jsx)("b",{children:"id"}),": The metadata field name."]}),(0,we.jsxs)("li",{children:[(0,we.jsx)("b",{children:"scope"}),": The collection where this field is used (e.g., runs, biosamples, analyses)."]})]})]}),(0,we.jsxs)("h2",{id:"enhance-ontology-filtering-with-descendants",children:["Enhance Ontology Filtering with Descendants and semantic similarities ",(0,we.jsx)("span",{className:"optional",children:"(optional)"})]}),(0,we.jsxs)("p",{children:["To make ontology-based filtering more powerful and user-friendly, you can enable two features:",(0,we.jsxs)("ul",{children:[(0,we.jsxs)("li",{children:[(0,we.jsx)("b",{children:"Descendant terms"}),": Automatically include all subcategories of a given ontology term (e.g., querying for \u201ccancer\u201d also returns \u201clung cancer\u201d, \u201cbreast cancer\u201d, etc.)."]}),(0,we.jsxs)("li",{children:[(0,we.jsx)("b",{children:"id"}),": The metadata field name."]}),(0,we.jsxs)("li",{children:[(0,we.jsx)("b",{children:"Semantic similarity"}),": Enable matching to related ontology terms based on meaning, not just hierarchy. This allows for more flexible queries when users don\u2019t know the exact term used in the data."]})]})]}),(0,we.jsx)("h2",{children:"Example"}),(0,we.jsxs)("p",{children:["If your data uses the term ",(0,we.jsx)("b",{children:"\u201cglioblastoma\u201d"}),", but a user searches for ",(0,we.jsx)("b",{children:"\u201cbrain tumor\u201d"}),", semantic similarity can help bridge the gap \u2014 even if \u201cbrain tumor\u201d isn\u2019t a direct parent of \u201cglioblastoma\u201d."]}),(0,we.jsx)("h2",{children:"How to Enable"}),(0,we.jsxs)("ol",{children:[(0,we.jsxs)("li",{children:["Add your ",(0,we.jsx)("em",{children:".obo"})," ontology files"]}),(0,we.jsx)("p",{className:"lessMargin",children:"Place ontology files into the ontologies/ folder in your Beacon instance. The filename must match the ontology prefix in lowercase (e.g., ncit.obo, uberon.obo, etc.)."}),(0,we.jsx)("li",{children:"Run the script to enhance filtering terms"})]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker exec beaconprod python beacon/connections/mongo/get_descendants.py"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("get-descendant-terms"),children:e["get-descendant-terms"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{children:["This script:",(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"Adds descendant terms from the ontologies to each filtering term."}),(0,we.jsx)("li",{children:"Computes semantic similarities based on the ontology structure, enriching your Beacon\u2019s ability to match related terms."})]})]}),(0,we.jsx)("br",{}),(0,we.jsx)("br",{})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]})},uc=()=>{const[e,n]=(0,a.useState)({}),t=e=>{const t="cloning-repository"===e?"git clone https://github.com/EGA-archive/beacon2-pi-api.git":"bash mongostart.sh";navigator.clipboard.writeText(t).then((()=>{n((n=>({...n,[e]:!0}))),setTimeout((()=>n((n=>({...n,[e]:!1})))),1500)})).catch((e=>console.log(e)))};return(0,we.jsxs)("div",{className:"deploymentContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/pi-automated-deployment",className:"no-undeline",children:"Beacon 2 PI API"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/pi-automated-deployment",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:"Automated Deployment"})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h3",{children:"Beacon 2 Production Implementation API"}),(0,we.jsx)("h2",{id:"prerequisites",children:"Prerequisites"}),(0,we.jsxs)("p",{children:["You should have installed:",(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:(0,we.jsx)("a",{href:"https://docs.docker.com/engine/install/",target:"_blank",rel:"noopener noreferrer",children:"Docker"})}),(0,we.jsx)("li",{children:(0,we.jsx)("a",{href:"https://docs.docker.com/compose/install/",target:"_blank",rel:"noopener noreferrer",children:"Docker Compose"})}),(0,we.jsxs)("li",{children:[(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon-data-tools",target:"_blank",rel:"noopener noreferrer",children:"Data from Beacon RI Tools v2"}),". Please, bear in mind that the datasetId for your records must match the id for the dataset in the /datasets entry type."]})]})]}),(0,we.jsx)("h2",{id:"automated-deployment",children:"Automated Deployment"}),(0,we.jsx)("p",{children:"This deployment relies on a Bash script and a Makefile and it is only compatible with MongoDB database deployments."}),(0,we.jsx)("h2",{id:"cloning-repository",children:"Cloning the repository"}),(0,we.jsx)("p",{children:"First of all, start by cloning the GitHub repository in your system."}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["git clone"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-pi-api.git",children:"https://github.com/EGA-archive/beacon2-pi-api.git"})]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("cloning-repository"),children:e["cloning-repository"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{className:"paragraph-ports",children:"Make sure the next list of ports are free of use in your system:"}),(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"27017 \u2192 mongo"}),(0,we.jsx)("li",{children:"5050 \u2192 beacon"})]}),(0,we.jsx)("h2",{id:"execute-start-script",children:"Execute start script from root"}),(0,we.jsx)("p",{children:"To quickly deploy your Beacon instance and load initial data, run the following command from the root of your project:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"bash mongostart.sh"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("execute-start-script"),children:e["execute-start-script"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{children:["This script will:",(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"Deploy the Beacon API"}),(0,we.jsxs)("li",{children:["Insert default Beacon-formatted data into your MongoDB instance, using the files located at:"," ",(0,we.jsx)("em",{children:"/beacon2-pi-api/beacon/connections/mongo/data"})]})]})]}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsxs)("div",{children:["Want to insert your own data?",(0,we.jsx)("br",{}),"Simply replace the contents of the data folder with your custom Beacon Friendly Format (BFF) files before running the script."]})]}),(0,we.jsxs)("p",{className:"paragraph-final",children:["If the operation is successful, you will have a beacon up and running at"," ",(0,we.jsx)("a",{href:"http://localhost:5050/api",target:"_blank",rel:"noopener noreferrer",children:"http://localhost:5050/api"}),"."]}),(0,we.jsx)("h2",{id:"customize-your-beacon",children:"Next Steps: Customize Your Beacon"}),(0,we.jsxs)("p",{children:["Make your Beacon your own by following these next steps:",(0,we.jsxs)("ul",{children:[(0,we.jsxs)("li",{children:["Edit your instance\u2019s"," ",(0,we.jsx)("a",{href:"https://b2ri-documentation-demo.ega-archive.org/configuration#editing-beacon-info",children:"metadata"}),". Update the ",(0,we.jsx)("em",{children:"/info"})," endpoint with your organization's name, description, version, and contact details."]}),(0,we.jsxs)("li",{children:["Manage dataset"," ",(0,we.jsx)("a",{href:"https://b2ri-documentation-demo.ega-archive.org/configuration#managing-dataset-permissions",children:"permissions"}),". Control which datasets are public or require authentication."]}),(0,we.jsxs)("li",{children:["Enable advanced"," ",(0,we.jsxs)("a",{href:"https://b2ri-documentation-demo.ega-archive.org/filtering-terms#extract-terms",children:[" ","filtering"]})]}),(0,we.jsxs)("li",{children:["Found more setting options in"," ",(0,we.jsx)("a",{href:"https://b2ri-documentation-demo.ega-archive.org/configuration",children:"Configuration"})]})]})]})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]})},hc=()=>{const[e,n]=(0,a.useState)({});return(0,we.jsxs)("div",{className:"configFileToolsContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/conversion-from-phenopackets-to-bff",className:"no-undeline",children:"Beacon RI Tools v2"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/configuration-file",className:"no-undeline",children:(0,we.jsxs)("span",{className:"user-path-title",children:["Conversion from Phenopackets to BFF"," "]})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h3",{children:"Beacon RI Tools v2"}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsxs)("div",{children:["Please bear in mind that this is a beta version. If you encounter issues or have questions, feel free to open an issue in the"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/phenopackets-to-BFF/issues",target:"_blank",rel:"noopener noreferrer",children:"GitHub repository."})]})]}),(0,we.jsx)("h1",{children:"Conversion from Phenopackets to BFF "}),(0,we.jsxs)("p",{children:["This option will convert a ",(0,we.jsx)("b",{children:"Phenopacket v2"})," into"," ",(0,we.jsx)("b",{children:"Beacon Friendly Format"})," Biosamples and Individuals models while preserving as much information as possible."]}),(0,we.jsxs)("p",{children:["For detailed information about the mapping of properties between Phenopackets and Beacon schemas, please refer to this spreadsheet:"," ",(0,we.jsx)("a",{href:"https://docs.google.com/spreadsheets/d/1DfkV5BwXzOggDl53-ofi7obnHT4O7J-rtUrnhZH2BiE/edit?gid=474476020#gid=474476020)",target:"_blank",rel:"noopener noreferrer",children:"Beacon + Phenopackets Schemas."})]}),(0,we.jsxs)("p",{children:["Most of the mapping between the Phenopackets and Beacon models was straightforward. However, some fields in the Phenopacket schema did not have a direct match in the Beacon schema.",(0,we.jsx)("br",{}),"Fields containing important metadata that could not be mapped to a specific Beacon property were stored in the"," ",(0,we.jsx)("i",{children:"additionalInformation"})," (notes) field to preserve as much information as possible."]}),(0,we.jsx)("h2",{id:"specific-mappings",children:"Specific Mappings"}),(0,we.jsx)("h4",{children:"Individuals.diseases.notes"}),(0,we.jsxs)("p",{className:"lessMargin",children:["The ",(0,we.jsx)("i",{children:"file.individual_to_file_identifiers"})," field from Phenopackets is saved in the info property of the Beacon individual."]}),(0,we.jsx)("h4",{children:"Biosamples.notes"}),(0,we.jsxs)("p",{className:"lessMargin",children:["The ",(0,we.jsx)("i",{children:"biosamples.description"})," field is stored in the notes field of the Beacon biosample."]}),(0,we.jsx)("h4",{children:"Handling collectionData"}),(0,we.jsxs)("p",{className:"lessMargin",children:["The ",(0,we.jsx)("i",{children:"biosamples.timeOfCollection"})," property in Phenopackets supports various data types, including:",(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"gestationalAge: Measure of the age of a pregnancy"}),(0,we.jsx)("li",{children:"Age: Age as an ISO8601 duration (e.g., P40Y10M05D)"}),(0,we.jsx)("li",{children:"AgeRange: Age within a given range"}),(0,we.jsx)("li",{children:"OntologyClass: Age as an ontology class"}),(0,we.jsx)("li",{children:"Timestamp: Specific date and time"}),(0,we.jsx)("li",{children:"TimeInterval: Time interval"})]}),"In contrast, the Beacon schema defines"," ",(0,we.jsx)("i",{children:"biosamples.collectionDate"}),' as the "Date of biosample collection in ISO8601 format" and expects a simple string.',(0,we.jsx)("br",{}),"To maximize data retention, the tool converts the"," ",(0,we.jsx)("i",{children:"collectionData"})," from Phenopackets to a string and stores it in the ",(0,we.jsx)("i",{children:"collectionDate"})," property of Beacon. While this approach does not fully align with the intended ",(0,we.jsx)("i",{children:"collectionDate"})," field usage, it ensures valuable data is not lost."]}),(0,we.jsx)("h2",{id:"converting-data-from-phenopackets",children:"Converting data from phenopackets"}),(0,we.jsx)("p",{children:"To convert a phenopacket into Biosamples and Individuals schemas follow the following steps:"}),(0,we.jsxs)("ul",{children:[(0,we.jsxs)("li",{children:["Copy the phenopacket into this folder:"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/tree/main/phenopackets-to-BFF",target:"_blank",rel:"noopener noreferrer",children:"phenopackets-to-BFF."})]}),(0,we.jsx)("li",{children:"Modify and run the execution query:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{id:"phenopackets-to-BFF",children:"docker exec phenopackets-to-BFF python working-w-phenopackets.py /usr/src/app/examples/YOUR_PHENOPACKET_NAME.json"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>{var e;const t=null===(e=document.getElementById("phenopackets-to-BFF"))||void 0===e?void 0:e.innerText;var o,a;t&&(o=t,a="phenopackets-to-BFF",navigator.clipboard.writeText(o).then((()=>{n((e=>({...e,[a]:!0}))),setTimeout((()=>{n((e=>({...e,[a]:!1})))}),1500)})).catch((e=>console.error("Failed to copy text: ",e))))},children:e["phenopackets-to-BFF"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:"Bear in mind that the path **/usr/src/app/examples** must remain unchanged, but you need to add your phenopackets name to the query."})]}),(0,we.jsxs)("li",{children:["You'll find your BFFs in the same folder where you saved the phenopacket:"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/tree/main/phenopackets-to-BFF",target:"_blank",rel:"noopener noreferrer",children:"phenopackets-to-BFF."})]})]}),(0,we.jsx)("br",{}),(0,we.jsx)("br",{})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]})},pc=t.p+"static/media/qrcode.375e0c6c742eaa87792c.png",mc=()=>(0,we.jsxs)("div",{className:"networkUIQueriesContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/resources",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:"Resources"})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h1",{children:"Resources"}),(0,we.jsx)("h2",{children:"Explore more about Beacon v2 with the following resources:"}),(0,we.jsxs)("ul",{children:[(0,we.jsxs)("li",{children:["Beacon v2 webpage:"," ",(0,we.jsx)("a",{href:"https://genomebeacons.org/",target:"_blank",rel:"noopener noreferrer",children:"https://genomebeacons.org/"})]}),(0,we.jsxs)("li",{children:["Beacon v2 specification documentation:"," ",(0,we.jsx)("a",{href:"https://docs.genomebeacons.org/",target:"_blank",rel:"noopener noreferrer",children:"https://docs.genomebeacons.org/"})]}),(0,we.jsxs)("li",{children:["Beacon v2 API on GitHub:"," ",(0,we.jsx)("a",{href:"https://github.com/ga4gh-beacon/beacon-v2/",target:"_blank",rel:"noopener noreferrer",children:"https://github.com/ga4gh-beacon/beacon-v2/"})]}),(0,we.jsxs)("li",{children:["Beacon Verifier (verify the configuration of your implementation):"," ",(0,we.jsx)("a",{href:"https://beacon-verifier-demo.ega-archive.org/",target:"_blank",rel:"noopener noreferrer",children:"https://beacon-verifier-demo.ega-archive.org/"})]}),(0,we.jsxs)("li",{children:["Add your Beacon implementation to our Beacon registry:"," ",(0,we.jsx)("a",{href:"https://docs.google.com/forms/d/e/1FAIpQLScORwEVRAlsa8qe9SerKZLGy6qjphApjsHXC8-EcaOrUpW8tw/viewform",target:"_blank",rel:"noopener noreferrer",children:"https://docs.google.com/forms/d/e/1FAIpQLScORwEVRAlsa8qe9SerKZLGy6qjphApjsHXC8-EcaOrUpW8tw/viewform"})]}),(0,we.jsx)("img",{src:pc,className:"qrcode"})]})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]}),fc=()=>{const[e,n]=(0,a.useState)({}),t=e=>{const t={"bash-mongostart":"bash mongostart.sh","extract-filtering-terms":"docker exec beaconprod python beacon/connections/mongo/extract_filtering_terms.py","public-datasets":"public_datasets:\n- EGAD00001008392","beacon-ri-metadata":"beacon_id = 'org.ega-archive.beacon-ri-demo'\n\nbeacon_name = 'Beacon Reference Implementation: Rare Diseases Use Case\u2019\n\napi_version = 'v2.0.0'\n\nuri = 'https://beacon-apis-demo.ega-archive.org/api/'\n\norg_id = 'EGA'\n\norg_name = 'European Genome-Phenome Archive (EGA)'\n\norg_description = 'The European Genome-phenome Archive (EGA) is a service for permanent archiving and sharing of all types of personally identifiable genetic and phenotypic data resulting from biomedical research projects.'\n\norg_adress = 'C/ Dr. Aiguader, 88\nPRBB Building\n08003 Barcelona, Spain'\n\norg_welcome_url = 'https://ega-archive.org/'\n\norg_contact_url = 'mailto:beacon.ega@crg.eu'\n\norg_logo_url = 'https://legacy.ega-archive.org/images/logo.png'\n\norg_info = ''\n\ndescription = \"This Beacon is based on synthetic data hosted at the EGA. It contains information from one sample from EGAD00001008392\"\n\nversion = 'v2.0'\n\nwelcome_url = 'https://ega-archive.org/datasets/EGAD00001008392'\n\nalternative_url = 'https://beacon-apis-demo.ega-archive.org/api/'\n\ncreate_datetime = '2025-05-12T12:00:00.000000'\n\nupdate_datetime = ''","convert-all":"docker exec -it ri-tools python convert_csvTObff.py","mongoexport-genomicVariations":'docker exec ri-tools-mongo mongoexport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --collection genomicVariations > ri-tools/output_docs/genomicVariations.json',"vcf-file-path":"./beacon2-pi-api/ri-tools/files/vcf/files_to_read/case1C_subset.vcf.gz","vcf-import-output":"files/vcf/files_to_read/case1C_subset.vcf.gz\n100% | 5/5 [00:00<00:00, 6603.12it/s]\nSuccessfully inserted 5 records into beacon\nA total of 5 variants were processed\nA total of 0 variants were skipped","genomic-variations":"docker exec -it ri-tools python genomicVariations_vcf.py","phenopacket-output":"The mandatory fields for BioSamples were not present in the phenopacket\xa0\nMandatory properties for Individual schema (id and sex) present in the phenopacket.\n-> Creating Individuals schema ...\n    - diseases added to Individuals\n    - phenotypicFeatures added to Individuals\n-> Creating Individuals schema - DONE\n+ BFFs Individuals JSON saved in:\xa0\xa0\n/beacon2-pi-api/ri-tools/phenopackets-to-BFF/case1c.json","phenopacket-to-bff-case1c-command":"docker exec phenopackets-to-BFF python working-w-phenopackets.py /usr/src/app/examples/case1c.json EGAD00001008392","cloning-repository":"git clone https://github.com/EGA-archive/beacon2-pi-api.git","compress-vcf":"bgzip Case1C_subset.vcf","docker-compose":"docker compose up -d --build",conversionConfigRD:"#### Input and Output files config parameters ####\ncsv_folder = './csv/use_case_RD/'\noutput_docs_folder='./output_docs/use_case_RD_output/'\n\n#### VCF Conversion config parameters ####\nallele_counts=False # Variable still in test, leave it as False for now.\nreference_genome='GRCh37' # Choose one between NCBI36, GRCh37, GRCh38\ndatasetId='EGAD00001008392'\ncase_level_data=False\nexact_heterozygosity=False\nnum_rows=5\nverbosity=False # This variable, if True, will make the program run slower but give logs about all the skipped variants and the reason why.","vcf-snippet":'##fileformat=VCFv4.2\n##ALT=<ID=NON_REF,Description="Represents any possible alternative allele at this location">\n##FILTER=<ID=LowQual,Description="Low quality">\n##FORMAT=<ID=AD,Number=R,Type=Integer,Description="Allelic depths for the ref and alt alleles in the order listed">\n##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Approximate read depth (reads with MQ=255 or with bad mates are filtered)">\n##FORMAT=<ID=GQ,Number=1,Type=Integer,Description="Genotype Quality">\n##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">\n##FORMAT=<ID=MIN_DP,Number=1,Type=Integer,Description="Minimum DP observed within the GVCF block">\n##FORMAT=<ID=MQ0,Number=1,Type=Integer,Description="Number of Mapping Quality Zero Reads per sample">\n##FORMAT=<ID=PGT,Number=1,Type=String,Description="Physical phasing haplotype information, describing how the alternate alleles are phased in relation to one another">\n##FORMAT=<ID=PID,Number=1,Type=String,Description="Physical phasing ID information, where each unique ID within a given sample (but not across samples) connects records within a phasing group">\n##FORMAT=<ID=PL,Number=G,Type=Integer,Description="Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification">\n##FORMAT=<ID=SB,Number=4,Type=Integer,Description="Per-sample component statistics which comprise the Fisher\'s Exact Test to detect strand bias.">\n##GATKCommandLine.HaplotypeCaller=<ID=HaplotypeCaller,Version=3.6-0-g89b7209,Date="Thu Mar 23 15:53:02 CET 2017",Epoch=1490280782918,CommandLineOptions="analysis_type=HaplotypeCaller input_file=[/scratch/production/slaurie/PlatinumTrioForRD-Connect/HapMapTrio/gVCF/Case1C/E012882.bam] showFullBamList=false read_buffer_size=null read_filter=[BadCigar] disable_read_filter=[] intervals=[21] excludeIntervals=null interval_set_rule=UNION interval_merging=ALL interval_padding=0 reference_sequence=/project/production/Indexes/samtools/hsapiens.hs37d5.fasta nonDeterministicRandomSeed=false disableDithering=false maxRuntime=-1 maxRuntimeUnits=MINUTES downsampling_type=NONE downsample_to_fraction=null downsample_to_coverage=null baq=OFF baqGapOpenPenalty=40.0 refactor_NDN_cigar_string=false fix_misencoded_quality_scores=false allow_potentially_misencoded_quality_scores=false useOriginalQualities=false defaultBaseQualities=-1 performanceLog=null BQSR=null quantize_quals=0 static_quantized_quals=null round_down_quantized=false disable_indel_quals=false emit_original_quals=false preserve_qscores_less_than=6 globalQScorePrior=-1.0 validation_strictness=SILENT remove_program_records=false keep_program_records=false sample_rename_mapping_file=null unsafe=null disable_auto_index_creation_and_locking_when_reading_rods=false no_cmdline_in_header=false sites_only=false never_trim_vcf_format_field=true bcf=false bam_compression=null simplifyBAM=false disable_bam_indexing=false generate_md5=false num_threads=1 num_cpu_threads_per_data_thread=1 num_io_threads=0 monitorThreadEfficiency=false num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false generateShadowBCF=false variant_index_type=LINEAR variant_index_parameter=128000 reference_window_stop=0 phone_home= gatk_key=null tag=NA logging_level=INFO log_to_file=null help=false version=false likelihoodCalculationEngine=PairHMM heterogeneousKmerSizeResolution=COMBO_MIN dbsnp=(RodBinding name= source=UNBOUND) dontTrimActiveRegions=false maxDiscARExtension=25 maxGGAARExtension=300 paddingAroundIndels=150 paddingAroundSNPs=20 comp=[] annotation=[AlleleBalance, BaseCounts, BaseQualityRankSumTest, ChromosomeCounts, ClippingRankSumTest, Coverage, DepthPerAlleleBySample, DepthPerSampleHC, FisherStrand, GCContent, HaplotypeScore, HardyWeinberg, HomopolymerRun, ClippingRankSumTest, LikelihoodRankSumTest, LowMQ, MappingQualityRankSumTest, MappingQualityZero, MappingQualityZeroBySample, NBaseCount, QualByDepth, RMSMappingQuality, ReadPosRankSumTest, StrandBiasBySample, StrandOddsRatio, VariantType, StrandBiasBySample] excludeAnnotation=[ChromosomeCounts, FisherStrand, StrandOddsRatio, QualByDepth] group=[StandardAnnotation, StandardHCAnnotation] debug=false useFilteredReadsForAnnotations=false emitRefConfidence=GVCF bamOutput=null bamWriterType=CALLED_HAPLOTYPES emitDroppedReads=false disableOptimizations=false annotateNDA=false heterozygosity=0.001 indel_heterozygosity=1.25E-4 standard_min_confidence_threshold_for_calling=-0.0 standard_min_confidence_threshold_for_emitting=-0.0 max_alternate_alleles=6 max_num_PL_values=100 input_prior=[] sample_ploidy=2 genotyping_mode=DISCOVERY alleles=(RodBinding name= source=UNBOUND) contamination_fraction_to_filter=0.0 contamination_fraction_per_sample_file=null p_nonref_model=null exactcallslog=null output_mode=EMIT_VARIANTS_ONLY allSitePLs=true gcpHMM=10 pair_hmm_implementation=VECTOR_LOGLESS_CACHING pair_hmm_sub_implementation=ENABLE_ALL always_load_vector_logless_PairHMM_lib=false phredScaledGlobalReadMismappingRate=45 noFpga=false sample_name=null kmerSize=[10, 25] dontIncreaseKmerSizesForCycles=false allowNonUniqueKmersInRef=false numPruningSamples=1 recoverDanglingHeads=false doNotRecoverDanglingBranches=false minDanglingBranchLength=4 consensus=false maxNumHaplotypesInPopulation=128 errorCorrectKmers=false minPruning=2 debugGraphTransformations=false allowCyclesInKmerGraphToGeneratePaths=false graphOutput=null kmerLengthForReadErrorCorrection=25 minObservationsForKmerToBeSolid=20 GVCFGQBands=[20, 25, 30, 35, 40, 45, 50, 70, 90, 99] indelSizeToEliminateInRefModel=10 min_base_quality_score=10 includeUmappedReads=false useAllelesTrigger=false doNotRunPhysicalPhasing=false keepRG=null justDetermineActiveRegions=false dontGenotype=false dontUseSoftClippedBases=false captureAssemblyFailureBAM=false errorCorrectReads=false pcr_indel_model=CONSERVATIVE maxReadsInRegionPerSample=10000 minReadsPerAlignmentStart=10 mergeVariantsViaLD=false activityProfileOut=null activeRegionOut=null activeRegionIn=null activeRegionExtension=null forceActive=false activeRegionMaxSize=null bandPassSigma=null maxProbPropagationDistance=50 activeProbabilityThreshold=0.002 min_mapping_quality_score=20 filter_reads_with_N_cigar=false filter_mismatching_base_and_quals=false filter_bases_not_stored=false">\n##GVCFBlock0-20=minGQ=0(inclusive),maxGQ=20(exclusive)\n##GVCFBlock20-25=minGQ=20(inclusive),maxGQ=25(exclusive)\n##GVCFBlock25-30=minGQ=25(inclusive),maxGQ=30(exclusive)\n##GVCFBlock30-35=minGQ=30(inclusive),maxGQ=35(exclusive)\n##GVCFBlock35-40=minGQ=35(inclusive),maxGQ=40(exclusive)\n##GVCFBlock40-45=minGQ=40(inclusive),maxGQ=45(exclusive)\n##GVCFBlock45-50=minGQ=45(inclusive),maxGQ=50(exclusive)\n##GVCFBlock50-70=minGQ=50(inclusive),maxGQ=70(exclusive)\n##GVCFBlock70-90=minGQ=70(inclusive),maxGQ=90(exclusive)\n##GVCFBlock90-99=minGQ=90(inclusive),maxGQ=99(exclusive)\n##GVCFBlock99-2147483647=minGQ=99(inclusive),maxGQ=2147483647(exclusive)\n##INFO=<ID=ABHet,Number=1,Type=Float,Description="Allele Balance for heterozygous calls (ref/(ref+alt))">\n##INFO=<ID=ABHom,Number=1,Type=Float,Description="Allele Balance for homozygous calls (A/(A+O)) where A is the allele (ref or alt) and O is anything other">\n##INFO=<ID=BaseCounts,Number=4,Type=Integer,Description="Counts of each base">\n##INFO=<ID=BaseQRankSum,Number=1,Type=Float,Description="Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities">\n##INFO=<ID=ClippingRankSum,Number=1,Type=Float,Description="Z-score From Wilcoxon rank sum test of Alt vs. Ref number of hard clipped bases">\n##INFO=<ID=DP,Number=1,Type=Integer,Description="Approximate read depth; some reads may have been filtered">\n##INFO=<ID=DS,Number=0,Type=Flag,Description="Were any of the samples downsampled?">\n##INFO=<ID=END,Number=1,Type=Integer,Description="Stop position of the interval">\n##INFO=<ID=ExcessHet,Number=1,Type=Float,Description="Phred-scaled p-value for exact test of excess heterozygosity">\n##INFO=<ID=GC,Number=1,Type=Float,Description="GC content around the variant (see docs for window size details)">\n##INFO=<ID=HRun,Number=1,Type=Integer,Description="Largest Contiguous Homopolymer Run of Variant Allele In Either Direction">\n##INFO=<ID=HW,Number=1,Type=Float,Description="Phred-scaled p-value for Hardy-Weinberg violation">\n##INFO=<ID=HaplotypeScore,Number=1,Type=Float,Description="Consistency of the site with at most two segregating haplotypes">\n##INFO=<ID=InbreedingCoeff,Number=1,Type=Float,Description="Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation">\n##INFO=<ID=LikelihoodRankSum,Number=1,Type=Float,Description="Z-score from Wilcoxon rank sum test of Alt Vs. Ref haplotype likelihoods">\n##INFO=<ID=LowMQ,Number=3,Type=Float,Description="3-tuple: <fraction of reads with MQ=0>,<fraction of reads with MQ<=10>,<total number of reads>">\n##INFO=<ID=MLEAC,Number=A,Type=Integer,Description="Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed">\n##INFO=<ID=MLEAF,Number=A,Type=Float,Description="Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed">\n##INFO=<ID=MQ,Number=1,Type=Float,Description="RMS Mapping Quality">\n##INFO=<ID=MQ0,Number=1,Type=Integer,Description="Total Mapping Quality Zero Reads">\n##INFO=<ID=MQRankSum,Number=1,Type=Float,Description="Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities">\n##INFO=<ID=OND,Number=1,Type=Float,Description="Overall non-diploid ratio (alleles/(alleles+non-alleles))">\n##INFO=<ID=PercentNBase,Number=1,Type=Float,Description="Percentage of N bases in the pileup">\n##INFO=<ID=RAW_MQ,Number=1,Type=Float,Description="Raw data for RMS Mapping Quality">\n##INFO=<ID=ReadPosRankSum,Number=1,Type=Float,Description="Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias">\n##INFO=<ID=VariantType,Number=1,Type=String,Description="Variant type description">\n##contig=<ID=21,length=48129895>\n##reference=file:///project/production/Indexes/samtools/hsapiens.hs37d5.fasta\n#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tCase1C\n21\t9411635\t.\tC\tT\t49.77\t.\tBaseQRankSum=0.420;ClippingRankSum=0.000;DP=21;ExcessHet=3.0103;LikelihoodRankSum=-0.537;MLEAC=1,0;MLEAF=0.500,0.00;MQ0=0;MQRankSum=-1.786;RAW_MQ=42918.00;ReadPosRankSum=1.162\tGT:AD:DP:GQ:PL:SB\t0/1:17,4,0:21:78:78,0,544,129,556,685:8,9,4,0\n21\t9411732\t.\tT\tG\t823.77\t.\tDP=25;ExcessHet=3.0103;MLEAC=2,0;MLEAF=1.00,0.00;MQ0=0;RAW_MQ=67833.00\tGT:AD:DP:GQ:PL:SB\t1/1:0,25,0:25:75:852,75,0,852,75,852:0,0,14,11\n21\t9411931\t.\tG\tA\t104.77\t.\tBaseQRankSum=-0.873;ClippingRankSum=0.000;DP=27;ExcessHet=3.0103;LikelihoodRankSum=-0.375;MLEAC=1,0;MLEAF=0.500,0.00;MQ0=0;MQRankSum=-3.169;RAW_MQ=65235.00;ReadPosRankSum=0.657\tGT:AD:DP:GQ:PGT:PID:PL:SB\t0/1:22,5,0:27:99:0|1:9411931_G_A:133,0,874,199,889,1087:11,11,0,5\n21\t9411949\t.\tT\tC\t55.77\t.\tBaseQRankSum=1.914;ClippingRankSum=0.000;DP=29;ExcessHet=3.0103;LikelihoodRankSum=-0.063;MLEAC=1,0;MLEAF=0.500,0.00;MQ0=0;MQRankSum=-3.040;RAW_MQ=73991.00;ReadPosRankSum=-1.787\tGT:AD:DP:GQ:PGT:PID:PL:SB\t0/1:25,4,0:29:84:0|1:9411931_G_A:84,0,1132,168,1144,1312:12,13,0,4\n21\t9411965\t.\tT\tC\t0.01\t.\tBaseQRankSum=0.337;ClippingRankSum=0.000;DP=30;ExcessHet=3.0103;LikelihoodRankSum=-1.995;MLEAC=0,0;MLEAF=0.00,0.00;MQ0=0;MQRankSum=-2.589;RAW_MQ=90291.00;ReadPosRankSum=-2.080\tGT:AD:DP:GQ:PGT:PID:PL:SB\t0/0:28,2,0:30:0:0|1:9411931_G_A:0,0,1105,84,1111,1195:11,17,0,2',"phenopacket-snippet":'{\n        "id": "P0007498",\n        "subject": {\n          "id": "P0007498",\n          "dateOfBirth": "2013-01-01T00:00:00Z",\n          "sex": "MALE"\n        },\n        "phenotypicFeatures": [\n          {\n            "excluded": true,\n            "type": {\n              "id": "HP:0001249",\n              "label": "Intellectual disability"\n            }\n          },\n          {\n            "type": {\n              "id": "HP:0000467",\n              "label": "Neck muscle weakness"\n            }\n          },\n          {\n            "type": {\n              "id": "HP:0001252",\n              "label": "Muscular hypotonia"\n            }\n          },\n          {\n            "type": {\n              "id": "HP:0001319",\n              "label": "Neonatal hypotonia"\n            }\n          },\n          {\n            "type": {\n              "id": "HP:0001374",\n              "label": "Congenital hip dislocation"\n            }\n          },\n          {\n            "type": {\n              "id": "HP:0002540",\n              "label": "Inability to walk"\n            }\n          },\n          {\n            "type": {\n              "id": "HP:0002783",\n              "label": "Recurrent lower respiratory tract infections"\n            }\n          },\n          {\n            "type": {\n              "id": "HP:0002804",\n              "label": "Arthrogryposis multiplex congenita"\n            }\n          },\n          {\n            "type": {\n              "id": "HP:0003202",\n              "label": "Skeletal muscle atrophy"\n            }\n          },\n          {\n            "type": {\n              "id": "HP:0005684",\n              "label": "Distal arthrogryposis"\n            }\n          },\n          {\n            "type": {\n              "id": "HP:0030319",\n              "label": "Weakness of facial musculature"\n            }\n          }\n        ],\n        "diseases": [\n          {\n            "term": {\n              "id": "Orphanet:97245",\n              "label": "Congenital myopathy"\n            }\n          },\n          {\n            "term": {\n              "id": "OMIM:117000",\n              "label": "CENTRAL CORE DISEASE OF MUSCLE"\n            }\n          }\n        ],\n        "meta_data": {\n          "created": "2021-04-21T09:32:38.488Z",\n          "resources": [\n            {\n              "id": "hp",\n              "name": "Human Phenotype Ontology",\n              "url": "http://purl.obolibrary.org/obo/hp.owl",\n              "version": "2020-12-07",\n              "namespacePrefix": "HP",\n              "iriPrefix": "http://purl.obolibrary.org/obo/HP_"\n            },\n            {\n              "id": "orphanet",\n              "name": "Orphanet Rare Disease Ontology",\n              "url": "http://orpha.net/ontology/ORDO_en_3.1.owl",\n              "version": "3.1",\n              "namespacePrefix": "Orphanet",\n              "iriPrefix": "http://www.orpha.net/ORDO/Orphanet_"\n            },\n            {\n              "id": "hgnc",\n              "name": "HUGO Gene Nomenclature Committee",\n              "url": "https://www.genenames.org",\n              "version": "2021-01-13",\n              "namespacePrefix": "HGNC",\n              "iriPrefix": "https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/"\n            },\n            {\n              "id": "mim",\n              "name": "Online Mendelian Inheritance in Man",\n              "url": "https://omim.org/",\n              "version": "2021-01-21",\n              "namespacePrefix": "OMIM",\n              "iriPrefix": "https://omim.org/entry/"\n            }\n          ]\n        }\n      }',"docker-ps":"docker ps","img-ri-tools":"ghcr.io/ega-archive/beacon2-ri-tools-v2:latest","name-ri-tools":"ri-tools","img-phenopackets":"ghcr.io/ega-archive/phenopackets-to-bff:latest","name-phenopackets":"phenopackets-to-BFF","img-mongo-express":"mongo-express","name-mongo-express":"beacon2-pi-api-mongo-express-1","img-postgres":"ghcr.io/ega-archive/beacon2-ri-postgres-v2:latest","name-postgres":"idp-db","img-beaconprod":"beacon2-pi-api-beaconprod","name-beaconprod":"beaconprod","edit-ri-tools":"#### Input and Output files config parameters ####\n\n      csv_folder = './csv/examples/'\n      output_docs_folder='./output_docs/'\n      \n      #### VCF Conversion config parameters ####\n      \n      allele_counts=False\n      reference_genome='GRCh37' # Choose one between NCBI36, GRCh37, GRCh38\n      datasetId='COVID_pop11_fin_2'\n      case_level_data=False\n      exact_heterozygosity=False\n      num_rows=1500000\n      verbosity=False","connections-mongo-reinde-py":"docker exec beaconprod python /beacon/connections/mongo/reindex.py","python-convert-csvTObff-py":"docker exec -it ri-tools python convert_csvTObff.py","python-genomicVariations-vcf-py":"docker exec -it ri-tools python genomicVariations_vcf.py","mongo-export-snippet":'docker exec ri-tools mongoexport --jsonArray --uri "mongodb://\nroot:example@127.0.0.1:27017/beacon?authSource=admin" --collection\ngenomicVariations > ./beacon2-pi-api/ri-tools/output_docs//\ngenomicVariations.json',"execute-start-script":"bash mongostart.sh","connections-mongo-data":"cd ./beacon2-pi-api/beacon/connections/mongo/data","bulk-mongoimport-snippet":'docker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/datasets.json --collection datasets\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/individuals.json --collection individuals\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/cohorts.json --collection cohorts\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/analyses.json --collection analyses\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/biosamples.json --collection biosamples\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/genomicVariations.json --collection genomicVariations\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/runs.json --collection runs\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/targets.json --collection targets\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/caseLevelData.json --collection caseLevelData'};t[e]&&navigator.clipboard.writeText(t[e]).then((()=>{n((n=>({...n,[e]:!0}))),setTimeout((()=>{n((n=>({...n,[e]:!1})))}),1500)})).catch(console.error)};return(0,we.jsxs)("div",{className:"apiConfigContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/use-case:-rare-diseases-data",className:"no-undeline",children:"Tutorials"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/use-case:-rare-diseases-data",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:"Use Case: Rare Diseases Data "})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsxs)("h3",{children:["Use Case : Inserting Rare Disease Synthetic Dataset into a beacon."," "]}),(0,we.jsx)("h1",{id:"introduction",children:"Introduction"}),(0,we.jsx)("p",{children:"For this example we are going to use the synthetic dataset, hosted at the European Genome-Phenome Archive, EGAD00001008392."}),(0,we.jsxs)("p",{children:["The Rare Disease Synthetic Dataset (",(0,we.jsx)("a",{href:"https://ega-archive.org/datasets/EGAD00001008392",children:"EGAD00001008392"}),"), contains public human datasets for the study of rare diseases. The use of public human genomic background combined with the in-silico insertion of real disease-causing variants enable a representative dataset for testing purposes without facing ethical and legal issues associated with the use of human sensitive data. This project aims to help develop technical implementations for rare disease data integration, analysis, discovery, and federated access."]}),(0,we.jsx)("p",{children:"This dataset contains 18 samples and different data types (FASTQs, BAMs and VCFs). To simplify the example we\u2019ll only insert data from the sample EGAN00003364605, case1_index, from which we have the VCFs for the 24 chromosomes, a BAM and its index and a phenopacket with phenotypic and clinical information."}),(0,we.jsx)("p",{children:"For this example we\u2019ll include into our beacon, the variants from chromosome 21 VCF, the phenopacket and all the rest of information available using CSV templates."}),(0,we.jsxs)("table",{className:"dockerTableSimple",children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Object Type"})}),(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Identifier"})})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Rare Disease Synthetic Dataset"}),(0,we.jsx)("td",{children:(0,we.jsx)("a",{href:"https://metadata.ega-archive.org/datasets/EGAD00001008392",target:"_blank",children:"EGAD00001008392"})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"case1_index sample"}),(0,we.jsx)("td",{children:(0,we.jsx)("a",{href:"https://metadata.ega-archive.org/samples/EGAN00003364605",target:"_blank",children:"EGAN00003364605"})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"case1_index fastq"}),(0,we.jsx)("td",{children:(0,we.jsx)("a",{href:"https://metadata.ega-archive.org/runs/EGAR00003021166",target:"_blank",children:"EGAR00003021166"})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Chromosome 21 VCF"}),(0,we.jsx)("td",{children:(0,we.jsx)("a",{href:"https://metadata.ega-archive.org/analyses/EGAZ00001744017",target:"_blank",children:"EGAZ00001744017"})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Phenopacket"}),(0,we.jsx)("td",{children:(0,we.jsx)("a",{href:"https://metadata.ega-archive.org/analyses/EGAZ00001744003",target:"_blank",children:"EGAZ00001744003"})})]})]})]}),(0,we.jsx)("h2",{className:"h2-less-margin",children:"Table 1. EGA identifiers for all the synthetic data used.\xa0"}),(0,we.jsxs)("p",{children:["For further information about how to use the public metadata API please visit the"," ",(0,we.jsx)("a",{href:"https://ega-archive.org/discovery/metadata/public-metadata-api/",children:"EGA documentation"}),"."]}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsxs)("div",{children:["Before continuing with this tutorial, please make sure you've reviewed"," ",(0,we.jsx)("a",{href:"https://b2ri-documentation-demo.ega-archive.org/create-your-beacon",children:"10 steps to create your first beacon"}),". In that tutorial you'll find important information that will help you follow this use case theoretically and technically."]})]}),(0,we.jsx)("h1",{id:"dataset-model-creation\xa0",children:"Dataset model creation\xa0"}),(0,we.jsx)("p",{children:"Dataset model is described as a group of variants or individuals that share something in common, in this case that they belong to the same repository."}),(0,we.jsxs)("p",{children:["Let\u2019s start checking which information we\u2019ve got and which one can be included in the"," ",(0,we.jsx)("a",{href:"https://docs.genomebeacons.org/schemas-md/datasets_defaultSchema/",children:"dataset model"}),"."]}),(0,we.jsx)("p",{children:"EGA metadata about EGAD00001008392:"}),(0,we.jsxs)("table",{className:"dockerTableSimple",children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Metadata field"})}),(0,we.jsxs)("th",{children:[" ",(0,we.jsx)("b",{children:"Content"})]})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"stable_id"}),(0,we.jsx)("td",{children:"EGAD00001008392"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Title"}),(0,we.jsx)("td",{children:"Rare Disease Synthetic Dataset"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Description"}),(0,we.jsx)("td",{children:"The purpose of this project is to provide public human datasets for the study of rare diseases. The use of public human genomic background combined with the in-silico insertion of real disease-causing variants enable a representative dataset for testing purposes without facing ethical and legal issues associated with the use of human sensitive data. This project aims to help develop technical implementations for rare disease data integration, analysis, discovery, and federated access."})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Dataset types"}),(0,we.jsx)("td",{children:'{"Genomic variant calling", "Phenotype information", "Whole genome sequencing"}'})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Technologies"}),(0,we.jsx)("td",{children:'{"Illumina HiSeq 2000"}'})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Number of samples "}),(0,we.jsx)("td",{children:"18"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Created at"}),(0,we.jsx)("td",{children:"2021-12-01 11:53:53.000 +0100"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Edited at"}),(0,we.jsx)("td",{children:"2025-04-14 23:18:43.985 +0200"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Use conditions"}),(0,we.jsx)("td",{children:"Data may be used freely for non-commercial research and educational activities."})]})]})]}),(0,we.jsxs)("p",{children:["Now, let\u2019s take a look at the"," ",(0,we.jsx)("a",{href:"https://docs.genomebeacons.org/schemas-md/datasets_defaultSchema/",children:"dataset beacon model"}),".\xa0"]}),(0,we.jsx)("p",{children:"The mapping between our data and the beacon model would look like:\xa0"}),(0,we.jsxs)("table",{className:"dockerTableSimple",children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:(0,we.jsxs)("b",{children:["Dataset Beacon",(0,we.jsx)("br",{}),"Model Term"]})}),(0,we.jsxs)("th",{children:[" ",(0,we.jsxs)("b",{children:["Property ",(0,we.jsx)("br",{}),"Type"]})]}),(0,we.jsx)("th",{children:(0,we.jsxs)("b",{children:["EGA Metadata ",(0,we.jsx)("br",{}),"field"]})}),(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Content"})})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"id"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:"stable_id"}),(0,we.jsx)("td",{children:"EGAD00001008392"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Name"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:"Title"}),(0,we.jsx)("td",{children:"Rare Disease Synthetic Dataset"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Description"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:"Description"}),(0,we.jsx)("td",{children:"The purpose of this project is to provide public human datasets for the study of rare diseases. The use of public human genomic background combined with the in-silico insertion of real disease-causing variants enable a representative dataset for testing purposes without facing ethical and legal issues associated with the use of human sensitive data. This project aims to help develop technical implementations for rare disease data integration, analysis, discovery, and federated access."})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"X"}),(0,we.jsx)("td",{}),(0,we.jsx)("td",{children:"Dataset types"}),(0,we.jsx)("td",{children:'{"Genomic variant calling", "Phenotype information", "Whole genome sequencing"}'})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"X"}),(0,we.jsx)("td",{}),(0,we.jsx)("td",{children:"Technologies"}),(0,we.jsx)("td",{children:'{"Illumina HiSeq 2000"}'})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"X"}),(0,we.jsx)("td",{}),(0,we.jsx)("td",{children:"Number of samples "}),(0,we.jsx)("td",{children:"18"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"createDateTime"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:"Created at"}),(0,we.jsx)("td",{children:"2021-12-01 11:53:53.000 +0100"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"updateDateTime"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:"Edited at"}),(0,we.jsx)("td",{children:"2025-04-14 23:18:43.985 +0200"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"dataUseConditions\xa0"}),(0,we.jsx)("td",{children:"duoDataUse"}),(0,we.jsx)("td",{children:"Use conditions"}),(0,we.jsx)("td",{children:"Data may be used freely for non-commercial research and educational activities."})]})]})]}),(0,we.jsxs)("p",{children:["The first thing we should check it\u2019s if we have the mandatory information for the Dataset model, which is id and name. As we have, we can now"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/blob/main/csv/templates/datasets.csv",children:"fill in the CSV"})," ","with all the information.\xa0"]}),(0,we.jsxs)("p",{children:["We can see that most of the metadata and properties have a direct match, except the dataUseConditions property.\xa0",(0,we.jsx)("br",{})," In our metadata we have string but the property type is"," ",(0,we.jsx)("a",{href:"https://docs.genomebeacons.org/schemas-md/obj/duoDataUse/",children:"duoDataUse"}),", so we\u2019ll have to find the"," ",(0,we.jsx)("a",{href:"https://ega-archive.org/access/data-access-committee/data-use-ontology/",children:"DUO code"})," ","that means Data may be used freely for non-commercial research and educational activities. We\u2019ll choose:"]}),(0,we.jsxs)("table",{className:"dockerTableSimple",children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Term"})}),(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Label"})}),(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Definition"})}),(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Version"})})]})}),(0,we.jsx)("tbody",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"DUO:0000018"}),(0,we.jsx)("td",{children:"not for profit, non commercial use only"}),(0,we.jsx)("td",{children:"This data use modifier indicates that use of the data is limited to not-for-profit organizations and not-for-profit use, non-commercial use."}),(0,we.jsx)("td",{children:"2021-02-23"})]})})]}),(0,we.jsxs)("p",{children:["In the beacon model the term duoDataUse has four different properties:\xa0"," ",(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"id"}),(0,we.jsx)("li",{children:"label"}),(0,we.jsx)("li",{children:"description"}),(0,we.jsx)("li",{children:"modifiers"}),(0,we.jsx)("li",{children:"version"})]})]}),(0,we.jsxs)("p",{children:["To correctly include our information to the CSV we\u2019ll need to use the columns:\xa0"," ",(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"dataUseConditions|duoDataUse|id"}),(0,we.jsx)("li",{children:"dataUseConditions|duoDataUse|label"}),(0,we.jsx)("li",{children:"dataUseConditions|duoDataUse|description"}),(0,we.jsx)("li",{children:"dataUseConditions|duoDataUse|version"})]})]}),(0,we.jsxs)("p",{children:[" ","Take a look to our final"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/blob/main/csv/examples/Rare-diseases/datasets.csv",children:"CSV"}),"."]}),(0,we.jsx)("h1",{id:"Analysis-model-creation",children:"Analysis model creation"}),(0,we.jsx)("p",{children:"Now, let\u2019s focus on the information we\u2019ve got about EGAZ00001744017."}),(0,we.jsxs)("table",{className:"dockerTableSimple",children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Metadata field"})}),(0,we.jsxs)("th",{children:[" ",(0,we.jsx)("b",{children:"Content"})]})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"stable_id"}),(0,we.jsx)("td",{children:"EGAZ00001744017"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Title"}),(0,we.jsx)("td",{children:"case1_index VCF chr21"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Description"}),(0,we.jsx)("td",{children:"variants were called using GATK3.6"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Analysis type"}),(0,we.jsx)("td",{children:"SEQUENCE VARIATION"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Analysis file type"}),(0,we.jsx)("td",{children:"VCF"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Experiment type"}),(0,we.jsx)("td",{children:'{"Whole genome sequencing"}'})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Genome"}),(0,we.jsx)("td",{children:"GRCh37"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Chromosomes"}),(0,we.jsx)("td",{children:"CM000683.1"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"creation date"}),(0,we.jsx)("td",{children:"2021-10-22"})]})]})]}),(0,we.jsxs)("p",{children:["And let\u2019s try to match as much as possible to the"," ",(0,we.jsx)("a",{href:"https://docs.genomebeacons.org/schemas-md/analyses_defaultSchema/",children:"analysis model"}),":"]}),(0,we.jsxs)("table",{className:"dockerTableSimple",children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:(0,we.jsxs)("b",{children:["Dataset Beacon",(0,we.jsx)("br",{}),"Model Term"]})}),(0,we.jsxs)("th",{children:[" ",(0,we.jsxs)("b",{children:["Property ",(0,we.jsx)("br",{}),"Type"]})]}),(0,we.jsx)("th",{children:(0,we.jsxs)("b",{children:["EGA Metadata ",(0,we.jsx)("br",{}),"field"]})}),(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Content"})})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"id"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:"stable_id"}),(0,we.jsx)("td",{children:"EGAZ00001744017"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"variantCaller"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:"Description"}),(0,we.jsx)("td",{children:"variants were called using GATK3.6"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"runId"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:"run"}),(0,we.jsx)("td",{children:"EGAR00003021166"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"analysisDate"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:"creation date"}),(0,we.jsx)("td",{children:"2021-10-22"})]})]})]}),(0,we.jsxs)("p",{children:["Let\u2019s remember that to include Analysis model into beacon the required properties are:\xa0",(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"id"}),(0,we.jsx)("li",{children:"analysisDate"}),(0,we.jsx)("li",{children:"pipelineName"})]})]}),(0,we.jsx)("p",{children:"Unfortunately, we don\u2019t have any information about the pipelineName used, so we won\u2019t be able to fill in the analysis model."}),(0,we.jsx)("h1",{id:"run-model-creation",children:"Run model creation"}),(0,we.jsxs)("p",{children:["This is the information we have available regarding the procedures used for sequencing a biosample."," "]}),(0,we.jsxs)("table",{className:"dockerTableSimple",children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Metadata field"})}),(0,we.jsxs)("th",{children:[" ",(0,we.jsx)("b",{children:"Content"})]})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"biosampleID"}),(0,we.jsx)("td",{children:"SAMEA10450712"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Experiment type"}),(0,we.jsx)("td",{children:"Whole genome sequencing"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"run creation date"}),(0,we.jsx)("td",{children:"2021-10-22"})]})]})]}),(0,we.jsxs)("p",{children:["To fill in"," ",(0,we.jsxs)("a",{href:"https://docs.genomebeacons.org/schemas-md/runs_defaultSchema/",children:[" ","Run model"]})," ","we need at least the following properties:",(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"id"}),(0,we.jsx)("li",{children:"biosampleId"}),(0,we.jsx)("li",{children:"runDate"})]})]}),(0,we.jsxs)("p",{children:["It seems we\u2019ll have enough data to include Runs into our beacon instance:"," "]}),(0,we.jsxs)("table",{className:"dockerTableSimple",children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:(0,we.jsxs)("b",{children:["Dataset Beacon",(0,we.jsx)("br",{}),"Model Term"]})}),(0,we.jsxs)("th",{children:[" ",(0,we.jsxs)("b",{children:["Property ",(0,we.jsx)("br",{}),"Type"]})]}),(0,we.jsx)("th",{children:(0,we.jsxs)("b",{children:["EGA Metadata ",(0,we.jsx)("br",{}),"field"]})}),(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Content"})})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"id"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:"stable_id"}),(0,we.jsx)("td",{children:"EGAR00003021166"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"biosampleId"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:"biosampleID"}),(0,we.jsx)("td",{children:"SAMEA10450712"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"runDate"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:"run creation"}),(0,we.jsx)("td",{children:"2021-10-22"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"libraryStrategy"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:"Experiment type"}),(0,we.jsx)("td",{children:"Whole genome sequencing"})]})]})]}),(0,we.jsx)("p",{children:"But there are other terms from Runs that we can fill with the metadata we have available now, for example, from Dataset:"}),(0,we.jsxs)("table",{className:"dockerTableSimple",children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:(0,we.jsxs)("b",{children:["Dataset Beacon",(0,we.jsx)("br",{}),"Model Term"]})}),(0,we.jsxs)("th",{children:[" ",(0,we.jsxs)("b",{children:["Property ",(0,we.jsx)("br",{}),"Type"]})]}),(0,we.jsx)("th",{children:(0,we.jsxs)("b",{children:["EGA Metadata ",(0,we.jsx)("br",{}),"field"]})}),(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Content"})})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"platform"}),(0,we.jsx)("td",{children:"string"}),(0,we.jsx)("td",{children:"Technology"}),(0,we.jsx)("td",{children:"Illumina"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"platformModel"}),(0,we.jsx)("td",{children:"id,label"}),(0,we.jsx)("td",{children:"Technology"}),(0,we.jsx)("td",{children:"Illumina HiSeq 2000"})]})]})]}),(0,we.jsxs)("p",{children:["Take a look at the"," ",(0,we.jsxs)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/blob/main/csv/examples/Rare-diseases/runs.csv",children:[" ","CSV"]})," ","with all this information."]}),(0,we.jsx)("h1",{id:"individual-model-from-phenopacket",children:"Individual model from phenopacket"}),(0,we.jsxs)("p",{children:["In"," ",(0,we.jsxs)("a",{href:"https://metadata.ega-archive.org/analyses/EGAZ00001744003",children:[" ","EGAZ00001744003"]})," ","we have a phenopacket with the following information:\xa0",(0,we.jsx)("br",{})," ","Please save it to a file named"," ",(0,we.jsxs)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/blob/main/phenopackets-to-BFF/rare-disease/case1c.json",children:[" ","Case1C.json."]})]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'{\n  "id": "P0007498",\n  "subject": {\n    "id": "P0007498",\n    "dateOfBirth": "2013-01-01T00:00:00Z",\n    "sex": "MALE"\n  },\n  "phenotypicFeatures": [\n    {\n      "excluded": true,\n      "type": {\n        "id": "HP:0001249",\n        "label": "Intellectual disability"\n      }\n    },\n    {\n      "type": {\n        "id": "HP:0000467",\n        "label": "Neck muscle weakness"\n      }\n    },\n    {\n      "type": {\n        "id": "HP:0001252",\n        "label": "Muscular hypotonia"\n      }\n    },\n    {\n      "type": {\n        "id": "HP:0001319",\n        "label": "Neonatal hypotonia"\n      }\n    },\n    {\n      "type": {\n        "id": "HP:0001374",\n        "label": "Congenital hip dislocation"\n      }\n    },\n    {\n      "type": {\n        "id": "HP:0002540",\n        "label": "Inability to walk"\n      }\n    },\n    {\n      "type": {\n        "id": "HP:0002783",\n        "label": "Recurrent lower respiratory tract infections"\n      }\n    },\n    {\n      "type": {\n        "id": "HP:0002804",\n        "label": "Arthrogryposis multiplex congenita"\n      }\n    },\n    {\n      "type": {\n        "id": "HP:0003202",\n        "label": "Skeletal muscle atrophy"\n      }\n    },\n    {\n      "type": {\n        "id": "HP:0005684",\n        "label": "Distal arthrogryposis"\n      }\n    },\n    {\n      "type": {\n        "id": "HP:0030319",\n        "label": "Weakness of facial musculature"\n      }\n    }\n  ],\n  "diseases": [\n    {\n      "term": {\n        "id": "Orphanet:97245",\n        "label": "Congenital myopathy"\n      }\n    },\n    {\n      "term": {\n        "id": "OMIM:117000",\n        "label": "CENTRAL CORE DISEASE OF MUSCLE"\n      }\n    }\n  ],\n  "meta_data": {\n    "created": "2021-04-21T09:32:38.488Z",\n    "resources": [\n      {\n        "id": "hp",\n        "name": "Human Phenotype Ontology",\n        "url": "http://purl.obolibrary.org/obo/hp.owl",\n        "version": "2020-12-07",\n        "namespacePrefix": "HP",\n        "iriPrefix": "http://purl.obolibrary.org/obo/HP_"\n      },\n      {\n        "id": "orphanet",\n        "name": "Orphanet Rare Disease Ontology",\n        "url": "http://orpha.net/ontology/ORDO_en_3.1.owl",\n        "version": "3.1",\n        "namespacePrefix": "Orphanet",\n        "iriPrefix": "http://www.orpha.net/ORDO/Orphanet_"\n      },\n      {\n        "id": "hgnc",\n        "name": "HUGO Gene Nomenclature Committee",\n        "url": "https://www.genenames.org",\n        "version": "2021-01-13",\n        "namespacePrefix": "HGNC",\n        "iriPrefix": "https://www.genenames.org/data/gene-symbol-report/#!/hgnc_id/"\n      },\n      {\n        "id": "mim",\n        "name": "Online Mendelian Inheritance in Man",\n        "url": "https://omim.org/",\n        "version": "2021-01-21",\n        "namespacePrefix": "OMIM",\n        "iriPrefix": "https://omim.org/entry/"\n      }\n    ]\n  }\n}'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("phenopacket-snippet"),children:e["phenopacket-snippet"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol copySymbol-custom",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"Phenopackets v2 and the beacon schemas are only compatible for the biosamples and individual models.\xa0"}),(0,we.jsx)("p",{children:"In this example, this phenopacket contains only information for individuals, because the three mandatory terms for biosamples (id, biosampleStatus, sampleOriginType) are not present."}),(0,we.jsx)("h1",{id:"genomicVariations-model-from-VCF",children:"genomicVariations model from VCF"}),(0,we.jsxs)("p",{children:["Below, you\u2019ll find a subset of 5 variants from"," ",(0,we.jsx)("a",{href:"https://metadata.ega-archive.org/analyses/EGAZ00001744017",children:"EGAZ00001744017."}),(0,we.jsx)("br",{}),"Please, save the following information into a"," ",(0,we.jsx)("em",{children:"Case1C_subset.vcf "}),"file:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'##fileformat=VCFv4.2\n##ALT=<ID=NON_REF,Description="Represents any possible alternative allele at this location">\n##FILTER=<ID=LowQual,Description="Low quality">\n##FORMAT=<ID=AD,Number=R,Type=Integer,Description="Allelic depths for the ref and alt alleles in the order listed">\n##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Approximate read depth (reads with MQ=255 or with bad mates are filtered)">\n##FORMAT=<ID=GQ,Number=1,Type=Integer,Description="Genotype Quality">\n##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">\n##FORMAT=<ID=MIN_DP,Number=1,Type=Integer,Description="Minimum DP observed within the GVCF block">\n##FORMAT=<ID=MQ0,Number=1,Type=Integer,Description="Number of Mapping Quality Zero Reads per sample">\n##FORMAT=<ID=PGT,Number=1,Type=String,Description="Physical phasing haplotype information, describing how the alternate alleles are phased in relation to one another">\n##FORMAT=<ID=PID,Number=1,Type=String,Description="Physical phasing ID information, where each unique ID within a given sample (but not across samples) connects records within a phasing group">\n##FORMAT=<ID=PL,Number=G,Type=Integer,Description="Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification">\n##FORMAT=<ID=SB,Number=4,Type=Integer,Description="Per-sample component statistics which comprise the Fisher\'s Exact Test to detect strand bias.">\n##GATKCommandLine.HaplotypeCaller=<ID=HaplotypeCaller,Version=3.6-0-g89b7209,Date="Thu Mar 23 15:53:02 CET 2017",Epoch=1490280782918,CommandLineOptions="analysis_type=HaplotypeCaller input_file=[/scratch/production/slaurie/PlatinumTrioForRD-Connect/HapMapTrio/gVCF/Case1C/E012882.bam] showFullBamList=false read_buffer_size=null read_filter=[BadCigar] disable_read_filter=[] intervals=[21] excludeIntervals=null interval_set_rule=UNION interval_merging=ALL interval_padding=0 reference_sequence=/project/production/Indexes/samtools/hsapiens.hs37d5.fasta nonDeterministicRandomSeed=false disableDithering=false maxRuntime=-1 maxRuntimeUnits=MINUTES downsampling_type=NONE downsample_to_fraction=null downsample_to_coverage=null baq=OFF baqGapOpenPenalty=40.0 refactor_NDN_cigar_string=false fix_misencoded_quality_scores=false allow_potentially_misencoded_quality_scores=false useOriginalQualities=false defaultBaseQualities=-1 performanceLog=null BQSR=null quantize_quals=0 static_quantized_quals=null round_down_quantized=false disable_indel_quals=false emit_original_quals=false preserve_qscores_less_than=6 globalQScorePrior=-1.0 validation_strictness=SILENT remove_program_records=false keep_program_records=false sample_rename_mapping_file=null unsafe=null disable_auto_index_creation_and_locking_when_reading_rods=false no_cmdline_in_header=false sites_only=false never_trim_vcf_format_field=true bcf=false bam_compression=null simplifyBAM=false disable_bam_indexing=false generate_md5=false num_threads=1 num_cpu_threads_per_data_thread=1 num_io_threads=0 monitorThreadEfficiency=false num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false generateShadowBCF=false variant_index_type=LINEAR variant_index_parameter=128000 reference_window_stop=0 phone_home= gatk_key=null tag=NA logging_level=INFO log_to_file=null help=false version=false likelihoodCalculationEngine=PairHMM heterogeneousKmerSizeResolution=COMBO_MIN dbsnp=(RodBinding name= source=UNBOUND) dontTrimActiveRegions=false maxDiscARExtension=25 maxGGAARExtension=300 paddingAroundIndels=150 paddingAroundSNPs=20 comp=[] annotation=[AlleleBalance, BaseCounts, BaseQualityRankSumTest, ChromosomeCounts, ClippingRankSumTest, Coverage, DepthPerAlleleBySample, DepthPerSampleHC, FisherStrand, GCContent, HaplotypeScore, HardyWeinberg, HomopolymerRun, ClippingRankSumTest, LikelihoodRankSumTest, LowMQ, MappingQualityRankSumTest, MappingQualityZero, MappingQualityZeroBySample, NBaseCount, QualByDepth, RMSMappingQuality, ReadPosRankSumTest, StrandBiasBySample, StrandOddsRatio, VariantType, StrandBiasBySample] excludeAnnotation=[ChromosomeCounts, FisherStrand, StrandOddsRatio, QualByDepth] group=[StandardAnnotation, StandardHCAnnotation] debug=false useFilteredReadsForAnnotations=false emitRefConfidence=GVCF bamOutput=null bamWriterType=CALLED_HAPLOTYPES emitDroppedReads=false disableOptimizations=false annotateNDA=false heterozygosity=0.001 indel_heterozygosity=1.25E-4 standard_min_confidence_threshold_for_calling=-0.0 standard_min_confidence_threshold_for_emitting=-0.0 max_alternate_alleles=6 max_num_PL_values=100 input_prior=[] sample_ploidy=2 genotyping_mode=DISCOVERY alleles=(RodBinding name= source=UNBOUND) contamination_fraction_to_filter=0.0 contamination_fraction_per_sample_file=null p_nonref_model=null exactcallslog=null output_mode=EMIT_VARIANTS_ONLY allSitePLs=true gcpHMM=10 pair_hmm_implementation=VECTOR_LOGLESS_CACHING pair_hmm_sub_implementation=ENABLE_ALL always_load_vector_logless_PairHMM_lib=false phredScaledGlobalReadMismappingRate=45 noFpga=false sample_name=null kmerSize=[10, 25] dontIncreaseKmerSizesForCycles=false allowNonUniqueKmersInRef=false numPruningSamples=1 recoverDanglingHeads=false doNotRecoverDanglingBranches=false minDanglingBranchLength=4 consensus=false maxNumHaplotypesInPopulation=128 errorCorrectKmers=false minPruning=2 debugGraphTransformations=false allowCyclesInKmerGraphToGeneratePaths=false graphOutput=null kmerLengthForReadErrorCorrection=25 minObservationsForKmerToBeSolid=20 GVCFGQBands=[20, 25, 30, 35, 40, 45, 50, 70, 90, 99] indelSizeToEliminateInRefModel=10 min_base_quality_score=10 includeUmappedReads=false useAllelesTrigger=false doNotRunPhysicalPhasing=false keepRG=null justDetermineActiveRegions=false dontGenotype=false dontUseSoftClippedBases=false captureAssemblyFailureBAM=false errorCorrectReads=false pcr_indel_model=CONSERVATIVE maxReadsInRegionPerSample=10000 minReadsPerAlignmentStart=10 mergeVariantsViaLD=false activityProfileOut=null activeRegionOut=null activeRegionIn=null activeRegionExtension=null forceActive=false activeRegionMaxSize=null bandPassSigma=null maxProbPropagationDistance=50 activeProbabilityThreshold=0.002 min_mapping_quality_score=20 filter_reads_with_N_cigar=false filter_mismatching_base_and_quals=false filter_bases_not_stored=false">\n##GVCFBlock0-20=minGQ=0(inclusive),maxGQ=20(exclusive)\n##GVCFBlock20-25=minGQ=20(inclusive),maxGQ=25(exclusive)\n##GVCFBlock25-30=minGQ=25(inclusive),maxGQ=30(exclusive)\n##GVCFBlock30-35=minGQ=30(inclusive),maxGQ=35(exclusive)\n##GVCFBlock35-40=minGQ=35(inclusive),maxGQ=40(exclusive)\n##GVCFBlock40-45=minGQ=40(inclusive),maxGQ=45(exclusive)\n##GVCFBlock45-50=minGQ=45(inclusive),maxGQ=50(exclusive)\n##GVCFBlock50-70=minGQ=50(inclusive),maxGQ=70(exclusive)\n##GVCFBlock70-90=minGQ=70(inclusive),maxGQ=90(exclusive)\n##GVCFBlock90-99=minGQ=90(inclusive),maxGQ=99(exclusive)\n##GVCFBlock99-2147483647=minGQ=99(inclusive),maxGQ=2147483647(exclusive)\n##INFO=<ID=ABHet,Number=1,Type=Float,Description="Allele Balance for heterozygous calls (ref/(ref+alt))">\n##INFO=<ID=ABHom,Number=1,Type=Float,Description="Allele Balance for homozygous calls (A/(A+O)) where A is the allele (ref or alt) and O is anything other">\n##INFO=<ID=BaseCounts,Number=4,Type=Integer,Description="Counts of each base">\n##INFO=<ID=BaseQRankSum,Number=1,Type=Float,Description="Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities">\n##INFO=<ID=ClippingRankSum,Number=1,Type=Float,Description="Z-score From Wilcoxon rank sum test of Alt vs. Ref number of hard clipped bases">\n##INFO=<ID=DP,Number=1,Type=Integer,Description="Approximate read depth; some reads may have been filtered">\n##INFO=<ID=DS,Number=0,Type=Flag,Description="Were any of the samples downsampled?">\n##INFO=<ID=END,Number=1,Type=Integer,Description="Stop position of the interval">\n##INFO=<ID=ExcessHet,Number=1,Type=Float,Description="Phred-scaled p-value for exact test of excess heterozygosity">\n##INFO=<ID=GC,Number=1,Type=Float,Description="GC content around the variant (see docs for window size details)">\n##INFO=<ID=HRun,Number=1,Type=Integer,Description="Largest Contiguous Homopolymer Run of Variant Allele In Either Direction">\n##INFO=<ID=HW,Number=1,Type=Float,Description="Phred-scaled p-value for Hardy-Weinberg violation">\n##INFO=<ID=HaplotypeScore,Number=1,Type=Float,Description="Consistency of the site with at most two segregating haplotypes">\n##INFO=<ID=InbreedingCoeff,Number=1,Type=Float,Description="Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation">\n##INFO=<ID=LikelihoodRankSum,Number=1,Type=Float,Description="Z-score from Wilcoxon rank sum test of Alt Vs. Ref haplotype likelihoods">\n##INFO=<ID=LowMQ,Number=3,Type=Float,Description="3-tuple: <fraction of reads with MQ=0>,<fraction of reads with MQ<=10>,<total number of reads>">\n##INFO=<ID=MLEAC,Number=A,Type=Integer,Description="Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed">\n##INFO=<ID=MLEAF,Number=A,Type=Float,Description="Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed">\n##INFO=<ID=MQ,Number=1,Type=Float,Description="RMS Mapping Quality">\n##INFO=<ID=MQ0,Number=1,Type=Integer,Description="Total Mapping Quality Zero Reads">\n##INFO=<ID=MQRankSum,Number=1,Type=Float,Description="Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities">\n##INFO=<ID=OND,Number=1,Type=Float,Description="Overall non-diploid ratio (alleles/(alleles+non-alleles))">\n##INFO=<ID=PercentNBase,Number=1,Type=Float,Description="Percentage of N bases in the pileup">\n##INFO=<ID=RAW_MQ,Number=1,Type=Float,Description="Raw data for RMS Mapping Quality">\n##INFO=<ID=ReadPosRankSum,Number=1,Type=Float,Description="Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias">\n##INFO=<ID=VariantType,Number=1,Type=String,Description="Variant type description">\n##contig=<ID=21,length=48129895>\n##reference=file:///project/production/Indexes/samtools/hsapiens.hs37d5.fasta\n#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tCase1C\n21\t9411635\t.\tC\tT\t49.77\t.\tBaseQRankSum=0.420;ClippingRankSum=0.000;DP=21;ExcessHet=3.0103;LikelihoodRankSum=-0.537;MLEAC=1,0;MLEAF=0.500,0.00;MQ0=0;MQRankSum=-1.786;RAW_MQ=42918.00;ReadPosRankSum=1.162\tGT:AD:DP:GQ:PL:SB\t0/1:17,4,0:21:78:78,0,544,129,556,685:8,9,4,0\n21\t9411732\t.\tT\tG\t823.77\t.\tDP=25;ExcessHet=3.0103;MLEAC=2,0;MLEAF=1.00,0.00;MQ0=0;RAW_MQ=67833.00\tGT:AD:DP:GQ:PL:SB\t1/1:0,25,0:25:75:852,75,0,852,75,852:0,0,14,11\n21\t9411931\t.\tG\tA\t104.77\t.\tBaseQRankSum=-0.873;ClippingRankSum=0.000;DP=27;ExcessHet=3.0103;LikelihoodRankSum=-0.375;MLEAC=1,0;MLEAF=0.500,0.00;MQ0=0;MQRankSum=-3.169;RAW_MQ=65235.00;ReadPosRankSum=0.657\tGT:AD:DP:GQ:PGT:PID:PL:SB\t0/1:22,5,0:27:99:0|1:9411931_G_A:133,0,874,199,889,1087:11,11,0,5\n21\t9411949\t.\tT\tC\t55.77\t.\tBaseQRankSum=1.914;ClippingRankSum=0.000;DP=29;ExcessHet=3.0103;LikelihoodRankSum=-0.063;MLEAC=1,0;MLEAF=0.500,0.00;MQ0=0;MQRankSum=-3.040;RAW_MQ=73991.00;ReadPosRankSum=-1.787\tGT:AD:DP:GQ:PGT:PID:PL:SB\t0/1:25,4,0:29:84:0|1:9411931_G_A:84,0,1132,168,1144,1312:12,13,0,4\n21\t9411965\t.\tT\tC\t0.01\t.\tBaseQRankSum=0.337;ClippingRankSum=0.000;DP=30;ExcessHet=3.0103;LikelihoodRankSum=-1.995;MLEAC=0,0;MLEAF=0.00,0.00;MQ0=0;MQRankSum=-2.589;RAW_MQ=90291.00;ReadPosRankSum=-2.080\tGT:AD:DP:GQ:PGT:PID:PL:SB\t0/0:28,2,0:30:0:0|1:9411931_G_A:0,0,1105,84,1111,1195:11,17,0,2\n'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("vcf-snippet"),children:e["vcf-snippet"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol copySymbol-custom",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"And then, compress it:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"bgzip Case1C_subset.vcf"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("compress-vcf"),children:e["compress-vcf"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h1",{id:"conversion-of-CSV-penopacket-and-VCF-to-BFF",children:"Conversion of CSV, Penopacket and VCF to BFF"}),(0,we.jsxs)("p",{children:["At this point we have information ready to create 4 beacon models:\xa0",(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"dataset.csv"}),(0,we.jsx)("li",{children:"run.csv"}),(0,we.jsx)("li",{children:"phenopacket v2"}),(0,we.jsx)("li",{children:"VCF for genomicVariations"})]})]}),(0,we.jsx)("p",{children:"Now, let\u2019s convert them to Beacon Friendly Format (BFF)."}),(0,we.jsx)("h6",{className:"sub-title",children:"Installing Beacon 2 Production Implementation API and beacon2-ri-tools-v2"}),(0,we.jsxs)("p",{children:["Before starting, you should have installed:",(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:(0,we.jsx)("a",{href:"https://docs.docker.com/engine/install/",target:"_blank",rel:"noopener noreferrer",children:"Docker"})}),(0,we.jsx)("li",{children:(0,we.jsx)("a",{href:"https://docs.docker.com/compose/install/",target:"_blank",rel:"noopener noreferrer",children:"Docker Compose"})})]})]}),(0,we.jsx)("p",{children:"First, let\u2019s clone the beacon production implementation repository in your system:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["git clone"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-pi-api.git",target:"_blank",rel:"noopener noreferrer",children:"https://github.com/EGA-archive/beacon2-pi-api.git"})]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("cloning-repository"),children:e["cloning-repository"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{className:"paragraph-ports",children:"Make sure the next list of ports are free of use in your system:"}),(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"27017 \u2192 mongo"}),(0,we.jsx)("li",{children:"5050 \u2192 beacon"})]}),(0,we.jsx)("p",{children:"And now let\u2019s build all the necessary containers:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker compose up -d --build\xa0"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("docker-compose"),children:e["docker-compose"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{children:["By cloning and building the Beacon 2 Production Implementation repository we have downloaded the tools that will deploy our beacon and the"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2",children:"beacon2-ri-tools-v2"}),", a toolkit that will allow you to populate your own beacon in the easiest way possible, by using CSVs and directly reading variant information from a VCF."]}),(0,we.jsx)("p",{children:"Let\u2019s check if the building was correct. Run the following command:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker ps"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("docker-ps"),children:e["docker-ps"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"This command will show all the running containers in your computer. You should have built:\xa0"}),(0,we.jsxs)("table",{className:"dockerTable",children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:"Image"}),(0,we.jsx)("th",{children:"Names"})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"ghcr.io/ega-archive/beacon2-ri-tools-v2:latest"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("img-ri-tools"),children:e["img-ri-tools"]?"Copied!":(0,we.jsx)("img",{src:ql,alt:"Copy",className:"copySymbol"})})]})})}),(0,we.jsx)("td",{children:(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"ri-tools"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("name-ri-tools"),children:e["name-ri-tools"]?"Copied!":(0,we.jsx)("img",{src:ql,alt:"Copy",className:"copySymbol"})})]})})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"ghcr.io/ega-archive/phenopackets-to-bff:latest"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("img-phenopackets"),children:e["img-phenopackets"]?"Copied!":(0,we.jsx)("img",{src:ql,alt:"Copy",className:"copySymbol"})})]})})}),(0,we.jsx)("td",{children:(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"phenopackets-to-BFF"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("name-phenopackets"),children:e["name-phenopackets"]?"Copied!":(0,we.jsx)("img",{src:ql,alt:"Copy",className:"copySymbol"})})]})})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"mongo-express"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("img-mongo-express"),children:e["img-mongo-express"]?"Copied!":(0,we.jsx)("img",{src:ql,alt:"Copy",className:"copySymbol"})})]})})}),(0,we.jsx)("td",{children:(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"beacon2-pi-api-mongo-express-1"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("name-mongo-express"),children:e["name-mongo-express"]?"Copied!":(0,we.jsx)("img",{src:ql,alt:"Copy",className:"copySymbol"})})]})})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"ghcr.io/ega-archive/beacon2-ri-postgres-v2:latest"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("img-postgres"),children:e["img-postgres"]?"Copied!":(0,we.jsx)("img",{src:ql,alt:"Copy",className:"copySymbol"})})]})})}),(0,we.jsx)("td",{children:(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"idp-db"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("name-postgres"),children:e["name-postgres"]?"Copied!":(0,we.jsx)("img",{src:ql,alt:"Copy",className:"copySymbol"})})]})})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"beacon2-pi-api-beaconprod"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("img-beaconprod"),children:e["img-beaconprod"]?"Copied!":(0,we.jsx)("img",{src:ql,alt:"Copy",className:"copySymbol"})})]})})}),(0,we.jsx)("td",{children:(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"beaconprod"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("name-beaconprod"),children:e["name-beaconprod"]?"Copied!":(0,we.jsx)("img",{src:ql,alt:"Copy",className:"copySymbol"})})]})})})]})]})]}),(0,we.jsx)("h6",{className:"sub-title",children:"Setting up conf.py"}),(0,we.jsx)("p",{children:"This will be the content that you find in beacon2-ri-tools-v2/conf/conf.py:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"#### Input and Output files config parameters ####\ncsv_folder = './csv/use_case_RD/'\noutput_docs_folder='./output_docs/use_case_RD_output/'\n\n#### VCF Conversion config parameters ####\nallele_counts=False # Variable still in test, leave it as False for now.\nreference_genome='GRCh37' # Choose one between NCBI36, GRCh37, GRCh38\ndatasetId='EGAD00001008392'\ncase_level_data=False\nexact_heterozygosity=False\nnum_rows=5\nverbosity=False # This variable, if True, will make the program run slower but give logs about all the skipped variants and the reason why."}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("conversionConfigRD"),children:e.conversionConfigRD?"Copied!":(0,we.jsx)("img",{className:"copySymbol copySymbol-custom",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"Let\u2019s start at the beginning of the file with the parameters for the CSV conversion."}),(0,we.jsx)("ul",{children:(0,we.jsxs)("li",{children:["Create a folder named ",(0,we.jsx)("em",{children:"use_case_RD"})," in ",(0,we.jsx)("em",{children:"./csv/"}),"\xa0 and move the ",(0,we.jsx)("em",{children:"dataset.csv"})," and\xa0 ",(0,we.jsx)("em",{children:"run.csv"})," there, this will be the CSV converted to BFF.\xa0 Please, make sure to end the path with a slash (",(0,we.jsx)("em",{children:"/"}),").\xa0"]})}),(0,we.jsx)("ul",{children:(0,we.jsxs)("li",{children:["Create another folder where the output BFF will be saved. Let\u2019s name it ",(0,we.jsx)("em",{children:"use_case_RD_output"})," inside the folder",(0,we.jsx)("em",{children:"./output_docs/"}),".\xa0"]})}),(0,we.jsx)("p",{children:"Let\u2019s continue with the parameters for the VCF.\xa0"}),(0,we.jsx)("ul",{children:(0,we.jsx)("li",{children:"From the header of the VCF we know that the variants belong to GRCh37.\xa0"})}),(0,we.jsx)("ul",{children:(0,we.jsxs)("li",{children:[(0,we.jsx)("em",{children:"datasetId"})," must match the ID in the ",(0,we.jsx)("em",{children:"dataset.csv"})," ","for the mapping of the variant and metadata to be correct. So, in our case we need to fill in datasetID as EGAD00001008392.\xa0"]})}),(0,we.jsx)("ul",{children:(0,we.jsxs)("li",{children:["The ",(0,we.jsx)("em",{children:"case_level_data"})," is a boolean parameter (True or False) which will relate your variants to the samples they belong to. As we don\u2019t have the biosamples schema, let\u2019s set it as False.\xa0"]})}),(0,we.jsx)("ul",{children:(0,we.jsxs)("li",{children:["The ",(0,we.jsx)("em",{children:"exact_heterozygosity"})," is a boolean parameter (True or False) that, in case case_level_data is True, then, it will classify the biosamples as being heterozygous for either the reference or the alternate allele. As in our case case_level_data is false, let\u2019s set exact_heterozygosity as False too.\xa0"]})}),(0,we.jsx)("ul",{children:(0,we.jsxs)("li",{children:["Fill in ",(0,we.jsx)("em",{children:"num_rows"})," with a 5, as we only have 5 variants in our VCF."]})}),(0,we.jsx)("h6",{className:"sub-title",children:"Conversion from CSV"}),(0,we.jsxs)("p",{children:["As we have already moved our CSVs to the ",(0,we.jsx)("em",{children:"./csv/use_case_RD/"})," ","we only need to run:\xa0"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker exec -it ri-tools python convert_csvTObff.py"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("convert-all"),children:e["convert-all"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{children:["In the folder ",(0,we.jsx)("em",{children:"./output_docs/use_case_RD_output/"})," will find two json. These files contain our run and dataset information in Beacon Friendly Format.\xa0"]}),(0,we.jsx)("h6",{className:"sub-title",children:"Conversion from Phenopacket"}),(0,we.jsxs)("p",{children:["Let\u2019s save the ",(0,we.jsx)("em",{children:"Case1C_phenopacket.json"})," in",(0,we.jsx)("em",{children:"./beacon2-ri-tools-v2/phenopackets-to-BFF"})," folder and then run:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker exec phenopackets-to-BFF python working-w-phenopackets.py /usr/src/app/examples/case1c.json EGAD00001008392"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("phenopacket-to-bff-case1c-command"),children:e["phenopacket-to-bff-case1c-command"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:"Bear in mind that the path cannot be changed. Replace only the name of the phenopacket, once it has been copied in phenopackets-to-BFF folder."})]}),(0,we.jsx)("p",{children:"Make sure to add the same datasetID as in the dataset model. This will allow the correct link between the phenopackets information and the dataset it belongs to."}),(0,we.jsx)("p",{children:"If everything works as expected your terminal output should look like:\xa0"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"The mandatory fields for BioSamples were not present in the phenopacket\xa0\nMandatory properties for Individual schema (id and sex) present in the phenopacket.\n-> Creating Individuals schema ...\n    - diseases added to Individuals\n    - phenotypicFeatures added to Individuals\n-> Creating Individuals schema - DONE\n+ BFFs Individuals JSON saved in:\xa0\xa0\n/beacon2-pi-api/ri-tools/phenopackets-to-BFF/case1c.json"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("phenopacket-output"),children:e["phenopacket-output"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol copySymbol-custom",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{children:["You\u2019ll find your phenopacket in \xa0",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/tree/main/phenopackets-to-BFF",children:"phenopackets-to-BFF."})]}),(0,we.jsx)("h6",{className:"sub-title",children:"Conversion from VCF"}),(0,we.jsx)("p",{children:"First, let\u2019s move the VCF to the correct folder:\xa0"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"./beacon2-pi-api/ri-tools/files/vcf/files_to_read/case1C_subset.vcf.gz"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("vcf-file-path"),children:e["vcf-file-path"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"And now, run the conversion command from VCF to BFF:\xa0"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker exec -it ri-tools python genomicVariations_vcf.py"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("genomic-variations"),children:e["genomic-variations"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"If everything ran correctly, this will be your output:\xa0"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"files/vcf/files_to_read/case1C_subset.vcf.gz\n100% | 5/5 [00:00<00:00, 6603.12it/s]\nSuccessfully inserted 5 records into beacon\nA total of 5 variants were processed\nA total of 0 variants were skipped"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("vcf-import-output"),children:e["vcf-import-output"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:"If the conversion of the variants is not working as expected modify the conf.py and set verbosity = True, this will give you information about which variants are being included and why."})]}),(0,we.jsx)("p",{children:"The VCF was converted into BFF and saved in a MongoDB. Let\u2019s extract the variants to insert them into our beacon:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'docker exec ri-tools-mongo mongoexport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --collection genomicVariations > ri-tools/output_docs/genomicVariations.json'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("mongoexport-genomicVariations"),children:e["mongoexport-genomicVariations"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"Modify the output path so it works on your computer. In this case, we\u2019ve saved it with the BFFs converted from CSV."}),(0,we.jsx)("h1",{id:"deploying-a-beacon-2-production-implementation-api",children:"Deploying a Beacon 2 production implementation API"}),(0,we.jsxs)("p",{children:["Now that we have all the data ready let\u2019s start with the deployment.\xa0",(0,we.jsx)("br",{}),"First, we\u2019ll make three important steps for the configuration of the beacon:\xa0",(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"Create our own landing page"}),(0,we.jsx)("li",{children:"Add the right permission for our dataset\xa0"}),(0,we.jsx)("li",{children:"Move our BFF to the correct folder"})]})]}),(0,we.jsxs)("p",{children:["To know about further settings, please visit the"," ",(0,we.jsxs)("a",{href:"https://b2ri-documentation-demo.ega-archive.org/configuration",children:[" ","Configuration section."]})]}),(0,we.jsx)("h6",{className:"sub-title",children:"Personalize info entry point"}),(0,we.jsxs)("p",{children:["The info entry point will be the landing page of your beacon, found at ",(0,we.jsx)("a",{href:"http://localhost:5050/api",children:"http://localhost:5050/api"}),"."]}),(0,we.jsxs)("p",{children:["To modify the information shown you\u2019ll need to update the ./beacon2-pi-api/beacon/conf/conf.py ",(0,we.jsx)("br",{}),"Our conf.py would look something like:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"beacon_id = 'org.ega-archive.beacon-ri-demo'\n\nbeacon_name = 'Beacon Reference Implementation: Rare Diseases Use Case\u2019\n\napi_version = 'v2.0.0'\n\nuri = 'https://beacon-apis-demo.ega-archive.org/api/'\n\norg_id = 'EGA'\n\norg_name = 'European Genome-Phenome Archive (EGA)'\n\norg_description = 'The European Genome-phenome Archive (EGA) is a service for permanent archiving and sharing of all types of personally identifiable genetic and phenotypic data resulting from biomedical research projects.'\n\norg_adress = 'C/ Dr. Aiguader, 88\nPRBB Building\n08003 Barcelona, Spain'\n\norg_welcome_url = 'https://ega-archive.org/'\n\norg_contact_url = 'mailto:beacon.ega@crg.eu'\n\norg_logo_url = 'https://legacy.ega-archive.org/images/logo.png'\n\norg_info = ''\n\ndescription = \"This Beacon is based on synthetic data hosted at the EGA. It contains information from one sample from EGAD00001008392\"\n\nversion = 'v2.0'\n\nwelcome_url = 'https://ega-archive.org/datasets/EGAD00001008392'\n\nalternative_url = 'https://beacon-apis-demo.ega-archive.org/api/'\n\ncreate_datetime = '2025-05-12T12:00:00.000000'\n\nupdate_datetime = ''"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("beacon-ri-metadata"),children:e["beacon-ri-metadata"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol copySymbol-custom",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h6",{className:"sub-title",children:"Managing dataset permissions"}),(0,we.jsxs)("p",{children:["There are 3 possible levels of beacon security for a dataset: public, registered and controlled.",(0,we.jsxs)("ul",{children:[(0,we.jsxs)("li",{children:["A ",(0,we.jsx)("b",{children:"public dataset"})," is a dataset that will be returned in a beacon query without an authentication token."]}),(0,we.jsxs)("li",{children:["A ",(0,we.jsx)("b",{children:"registered dataset"})," is a dataset that will be shown after a user sends a valid token (in other words, is logged in)."]}),(0,we.jsxs)("li",{children:["A ",(0,we.jsx)("b",{children:"controlled dataset"})," is a dataset that needs a user to send a valid token for authentication and the user needs to be allowed to query that dataset."]})]})]}),(0,we.jsx)("p",{children:"As our data is synthetic and has no control access demand, we\u2019ll add our dataset as a public dataset into the beacon.\xa0"}),(0,we.jsx)("p",{children:"For this, first we\u2019ll need to edit the .yml inside ./beacon2-pi-api/beacon/permissions/datasets. As we have decided that EGAD00001008392 will be of public access let\u2019s edit public_datasets.yml:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"public_datasets:\n- EGAD00001008392"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("public-datasets"),children:e["public-datasets"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h6",{className:"sub-title",children:"Move the BFF for the insertion"}),(0,we.jsxs)("p",{children:["Let\u2019s prepare the data for the insertion. Edit the following Makefile: ",(0,we.jsx)("br",{}),(0,we.jsx)("em",{children:"/beacon/connections/mongo/Makefile"})]}),(0,we.jsx)("p",{children:"And point the paths to your BFF files."}),(0,we.jsxs)("p",{className:"wider-note",children:[(0,we.jsx)("img",{className:"note-symbol-wider",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsxs)("div",{children:["Bear in mind that if you don\u2019t modify the commands in the Makefile the files must be named as:\xa0"," ",(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"analyses.json"}),(0,we.jsx)("li",{children:"biosamples.json"}),(0,we.jsx)("li",{children:"genomicVariations.json"}),(0,we.jsx)("li",{children:"individuals.json"}),(0,we.jsx)("li",{children:"runs.json"}),(0,we.jsx)("li",{children:"cohorts.json"}),(0,we.jsx)("li",{children:"datasets.json"})]})]})]}),(0,we.jsx)("h6",{className:"sub-title",children:"Deploy a beacon instance"}),(0,we.jsx)("p",{children:"Finally, simply run:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"bash mongostart.sh"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("bash-mongostart"),children:e["bash-mongostart"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{children:["If the operation is successful, you will have your beacon up and running at"," ",(0,we.jsx)("a",{href:"http://localhost:5050/api",target:"_blank",rel:"noopener noreferrer",children:"http://localhost:5050/api"}),"."]}),(0,we.jsx)("img",{className:"relationship-elements",alt:"Relationship-elements",src:"/beaconInstanceSchema.svg"}),(0,we.jsx)("h6",{className:"sub-title",children:"Filtering Terms"}),(0,we.jsx)("p",{children:"Filtering terms are metadata fields that allow users to query your Beacon more precisely \u2014 for example, by filtering results based on sequencing strategy (libraryStrategy), tissue type, disease, or other structured attributes. These terms often rely on ontologies to ensure consistent vocabulary across datasets."}),(0,we.jsx)("p",{children:"As we have some ontology fields, let\u2019s create the filtering terms for our beacon. We\u2019ll populate the /filteringTerms endpoint automatically by running:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker exec beaconprod python beacon/connections/mongo/extract_filtering_terms.py"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("extract-filtering-terms"),children:e["extract-filtering-terms"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsxs)("div",{children:["\xa0If you want to add new filtering terms manually or enhance ontology filtering with descendants and semantic similarities, please visit the"," ",(0,we.jsx)("a",{href:"https://b2ri-documentation-demo.ega-archive.org/filtering-terms",children:"Filtering Terms"})," ","section."]})]}),(0,we.jsx)("h1",{children:"Acknowledgments"}),(0,we.jsxs)("p",{children:["We acknowledge the use of data from the"," ",(0,we.jsx)("a",{href:"https://platform.rd-connect.eu/#/",children:"RD-Connect Genome-Phenome Analysis Platform"}),". This work was supported by the European Commission H2020 projects EJP-RD (grant number 825575) and B1MG (grant number 951724), as well as the Generalitat de Catalunya VEIS project (grant number 001-P-001647)."]}),(0,we.jsx)("br",{}),(0,we.jsx)("br",{})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]})};const gc=t.p+"static/media/pin.c509a3a2c6533bfcfaf0c4b2ff6fc801.svg",yc=()=>{const[e,n]=(0,a.useState)({}),t=e=>{const t={"cloning-repository":"git clone https://github.com/EGA-archive/beacon2-pi-api.git","docker-compose":"docker compose up -d --build\xa0","docker-ps":"docker ps","img-ri-tools":"ghcr.io/ega-archive/beacon2-ri-tools-v2:latest","name-ri-tools":"ri-tools","img-phenopackets":"ghcr.io/ega-archive/phenopackets-to-bff:latest","name-phenopackets":"phenopackets-to-BFF","img-mongo-express":"mongo-express","name-mongo-express":"beacon2-pi-api-mongo-express-1","img-postgres":"ghcr.io/ega-archive/beacon2-ri-postgres-v2:latest","name-postgres":"idp-db","img-beaconprod":"beacon2-pi-api-beaconprod","name-beaconprod":"beaconprod","edit-ri-tools":"#### Input and Output files config parameters ####\n\n      csv_folder = './csv/examples/'\n      output_docs_folder='./output_docs/'\n      \n      #### VCF Conversion config parameters ####\n      \n      allele_counts=False\n      reference_genome='GRCh37' # Choose one between NCBI36, GRCh37, GRCh38\n      datasetId='COVID_pop11_fin_2'\n      case_level_data=False\n      exact_heterozygosity=False\n      num_rows=1500000\n      verbosity=False","connections-mongo-reinde-py":"docker exec beaconprod python /beacon/connections/mongo/reindex.py","python-convert-csvTObff-py":"docker exec -it ri-tools python convert_csvTObff.py","python-genomicVariations-vcf-py":"docker exec -it ri-tools python genomicVariations_vcf.py","mongo-export-snippet":'docker exec ri-tools mongoexport --jsonArray --uri "mongodb://\nroot:example@127.0.0.1:27017/beacon?authSource=admin" --collection\ngenomicVariations > ./beacon2-pi-api/ri-tools/output_docs//\ngenomicVariations.json',"execute-start-script":"bash mongostart.sh","connections-mongo-data":"cd ./beacon2-pi-api/beacon/connections/mongo/data","bulk-mongoimport-snippet":'docker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/datasets.json --collection datasets\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/individuals.json --collection individuals\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/cohorts.json --collection cohorts\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/analyses.json --collection analyses\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/biosamples.json --collection biosamples\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/genomicVariations.json --collection genomicVariations\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/runs.json --collection runs\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/targets.json --collection targets\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/caseLevelData.json --collection caseLevelData'};t[e]&&navigator.clipboard.writeText(t[e]).then((()=>{n((n=>({...n,[e]:!0}))),setTimeout((()=>{n((n=>({...n,[e]:!1})))}),1500)})).catch(console.error)};return(0,we.jsxs)("div",{className:"apiConfigContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/create-your-beacon",className:"no-undeline",children:"Tutorials"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/create-your-beacon",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:"Create your Beacon"})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h3",{children:"10 steps to create your first beacon "}),(0,we.jsx)("h1",{children:"Introduction"}),(0,we.jsxs)("p",{children:["This guide will walk you through setting up a GA4GH Beacon v2 instance using the ",(0,we.jsx)("b",{children:"Production Implementation API"}),". You\u2019ll deploy your own Beacon, load example data in"," ",(0,we.jsx)("b",{children:"Beacon Friendly Format (BFF)"}),", and explore how to query it through the API."]}),(0,we.jsx)("p",{children:"As defined by the Global Alliance for Genomics and Health (GA4GH), a Beacon v2 is an open standard for the discovery of genomic (and phenotypic and clinical) data in biomedical research and clinical applications. Its main goal is to facilitate the discovery - and potential retrieval - of genomic variants and biomedical data from different organisational and geographic locations."}),(0,we.jsx)("p",{children:"If you are here it means you are interested in creating your own beacon!"}),(0,we.jsx)("br",{}),(0,we.jsxs)("div",{className:"stepsSection",children:[(0,we.jsx)("img",{src:gc,alt:"pin",className:"pin-logo"}),(0,we.jsxs)("div",{children:[(0,we.jsx)("p",{className:"stepsList",children:"This will be the steps you'll need to follow:"}),(0,we.jsxs)("ol",{children:[(0,we.jsx)("li",{children:"Installing Beacon 2 Production Implementation API"}),(0,we.jsx)("li",{children:"Converting VCF to Genomic Variations Model"}),(0,we.jsx)("li",{children:"Filling in the CSV Templates"}),(0,we.jsx)("li",{children:"Configuring conf.py"}),(0,we.jsx)("li",{children:"Converting CSV/VCF to BFF"}),(0,we.jsx)("li",{children:"Deploying your beacon instance"})]})]})]}),(0,we.jsx)("h1",{id:"installing-beacon-2-production-implementation-api-and-beacon2-ri-tools-v2",children:"1. Installing Beacon 2 Production Implementation API and beacon2-ri-tools-v2"}),(0,we.jsx)("div",{className:"stepsSection",children:(0,we.jsxs)("div",{children:[(0,we.jsx)("p",{className:"stepsList",children:"Before getting started, make sure you have the following:"}),(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:(0,we.jsx)("a",{href:"https://docs.docker.com/engine/install/",target:"_blank",rel:"noopener noreferrer",children:"Docker"})}),(0,we.jsx)("li",{children:(0,we.jsx)("a",{href:"https://docs.docker.com/compose/install/",target:"_blank",rel:"noopener noreferrer",children:"Docker Compose"})}),(0,we.jsx)("li",{children:"git"})]})]})}),(0,we.jsx)("p",{children:"This tutorial uses Beacon Friendly Format (BFF) files \u2013 standardized JSON documents that describe the beacon models. To create the BFF in the easiest way possible (directly from CSV, VCF or phenopacket) we\u2019ll use the Beacon2 ri tools v2, included in the beacon2-pi-api repository.\xa0"}),(0,we.jsx)("p",{children:"First, let\u2019s clone the beacon production implementation repository in your system:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsxs)("code",{children:["git clone"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-pi-api.git",target:"_blank",rel:"noopener noreferrer",children:"https://github.com/EGA-archive/beacon2-pi-api.git"})]}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("cloning-repository"),children:e["cloning-repository"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:"Ensure port 21017 and 5050 are free for MongoDb and Beacon."})]}),(0,we.jsx)("p",{children:"And now let\u2019s build all the necessary containers:\xa0"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker compose up -d --build\xa0"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("docker-compose"),children:e["docker-compose"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"Let\u2019s check if the building was correct. Run the following command:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker ps"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("docker-ps"),children:e["docker-ps"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"This command will show all the running containers in your computer. You should have built:\xa0"}),(0,we.jsxs)("table",{className:"dockerTable",children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:"Image"}),(0,we.jsx)("th",{children:"Names"})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"ghcr.io/ega-archive/beacon2-ri-tools-v2:latest"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("img-ri-tools"),children:e["img-ri-tools"]?"Copied!":(0,we.jsx)("img",{src:ql,alt:"Copy",className:"copySymbol"})})]})})}),(0,we.jsx)("td",{children:(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"ri-tools"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("name-ri-tools"),children:e["name-ri-tools"]?"Copied!":(0,we.jsx)("img",{src:ql,alt:"Copy",className:"copySymbol"})})]})})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"ghcr.io/ega-archive/phenopackets-to-bff:latest"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("img-phenopackets"),children:e["img-phenopackets"]?"Copied!":(0,we.jsx)("img",{src:ql,alt:"Copy",className:"copySymbol"})})]})})}),(0,we.jsx)("td",{children:(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"phenopackets-to-BFF"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("name-phenopackets"),children:e["name-phenopackets"]?"Copied!":(0,we.jsx)("img",{src:ql,alt:"Copy",className:"copySymbol"})})]})})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"mongo-express"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("img-mongo-express"),children:e["img-mongo-express"]?"Copied!":(0,we.jsx)("img",{src:ql,alt:"Copy",className:"copySymbol"})})]})})}),(0,we.jsx)("td",{children:(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"beacon2-pi-api-mongo-express-1"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("name-mongo-express"),children:e["name-mongo-express"]?"Copied!":(0,we.jsx)("img",{src:ql,alt:"Copy",className:"copySymbol"})})]})})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"ghcr.io/ega-archive/beacon2-ri-postgres-v2:latest"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("img-postgres"),children:e["img-postgres"]?"Copied!":(0,we.jsx)("img",{src:ql,alt:"Copy",className:"copySymbol"})})]})})}),(0,we.jsx)("td",{children:(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"idp-db"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("name-postgres"),children:e["name-postgres"]?"Copied!":(0,we.jsx)("img",{src:ql,alt:"Copy",className:"copySymbol"})})]})})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"beacon2-pi-api-beaconprod"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("img-beaconprod"),children:e["img-beaconprod"]?"Copied!":(0,we.jsx)("img",{src:ql,alt:"Copy",className:"copySymbol"})})]})})}),(0,we.jsx)("td",{children:(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"beaconprod"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("name-beaconprod"),children:e["name-beaconprod"]?"Copied!":(0,we.jsx)("img",{src:ql,alt:"Copy",className:"copySymbol"})})]})})})]})]})]}),(0,we.jsx)("h1",{id:"understanding-beacon-data-models",children:"2. Understanding Beacon Data Models"}),(0,we.jsx)("p",{children:"There are several ways of populating a beacon. To introduce information into a beacon you need to populate one, some, or all of the beacon model entities."}),(0,we.jsx)("p",{children:"Let\u2019s recapitulate about what model entities the beacon allows:\xa0"}),(0,we.jsxs)("ul",{children:[(0,we.jsxs)("li",{children:["Collections (",(0,we.jsx)("strong",{children:"Datasets"})," and"," ",(0,we.jsx)("strong",{children:"Cohorts"}),"): groupings of variants or individuals that share something in common: e.g., who belong to the same repository (datasets) or study populations (cohorts)."]}),(0,we.jsxs)("li",{children:[(0,we.jsx)("strong",{children:"Genomic variations"}),": unique genomic alterations, e.g., position in a genome, sequence alterations, type, etc. This model can be filled in directly with a VCF or creating a CSV."]}),(0,we.jsxs)("li",{children:[(0,we.jsx)("strong",{children:"Individuals"}),": either patients or healthy controls whose details (including phenotypic and clinical) are stored in the repository."]}),(0,we.jsxs)("li",{children:[(0,we.jsx)("strong",{children:"Biosamples"}),": samples taken from individuals, including details of procedures, dates and times."]}),(0,we.jsxs)("li",{children:[(0,we.jsx)("strong",{children:"Analyses & Runs"}),": details on (a) procedures used for sequencing a biosample (runs), and (b) bioinformatic procedures to identify variants (analyses)."]})]}),(0,we.jsxs)("p",{className:"wider-note",children:[(0,we.jsx)("img",{className:"note-symbol-wider",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsxs)("div",{children:["Note that none of these models are mandatory for a beacon to be functional, but the dataset model is mandatory for the API to be queryable. All the beacons will respond with boolean, count and full response. But the response type, hit, miss, all and none \xa0will be available if the dataset model was included in the beacon. This is what these responses mean:",(0,we.jsxs)("ul",{children:[(0,we.jsxs)("li",{children:[(0,we.jsx)("b",{children:"Hit"}),": it returns the positive results of the query per dataset.\xa0"]}),(0,we.jsxs)("li",{children:[(0,we.jsx)("b",{children:"Miss"}),": it returns the datasets where the query did not have a hit.\xa0"]}),(0,we.jsxs)("li",{children:[" ",(0,we.jsx)("b",{children:"All"}),": it returns all the datasets in the beacon.\xa0"]}),(0,we.jsxs)("li",{children:[" ",(0,we.jsx)("b",{children:"None"}),": aggregated results (not splitted by dataset)"]})]}),(0,we.jsx)("p",{children:"If you are interested in using them, make sure to include the datasets model into your beacon instance."})]})]}),(0,we.jsx)("h1",{id:"converting-vcf-to-genomic-variations-model-beacon2-ri-tools-v2",children:"3. Converting VCF to Genomic Variations Model beacon2-ri-tools-v2"}),(0,we.jsxs)("p",{children:["The most simple way of converting variant information into the genomicVariations model is directly reading a VCF into Beacon Friendly Format (BFF). To convert directly from a VCF, copy all the files into the"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/tree/main/files/vcf/files_to_read",target:"_blank",rel:"noopener noreferrer",children:"files_to_read"})," ","folder and follow the instructions below to correctly fill in the conf.py."]}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:"Bear in mind that if your variant information is not in a VCF format you can also fill in the Genomic Variations CSV."})]}),(0,we.jsx)("h1",{id:"filling-in-the-csv-templates",children:"4. Filling in the CSV Templates"}),(0,we.jsx)("p",{children:"Now, let\u2019s see how to fill in the CSV that will be converted by the beacon r.i tools into the models."}),(0,we.jsxs)("p",{children:["Here you\u2019ll find the"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/tree/main/csv/templates",target:"_blank",rel:"noopener noreferrer",children:"CSV templates"})," ","for all the models."]}),(0,we.jsx)("p",{children:"To correctly fill in all the properties you have two sources of information:"}),(0,we.jsxs)("ul",{children:[(0,we.jsxs)("li",{children:["The"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/tree/main/ref_schemas",target:"_blank",rel:"noopener noreferrer",children:"JSON references schemas"})," ","which will be really useful if you are used to working with JSON schemas."]}),(0,we.jsxs)("li",{children:["The GA4GH Beacon v2 Documentation:",(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:(0,we.jsx)("a",{href:"https://docs.genomebeacons.org/schemas-md/analyses_defaultSchema/",target:"_blank",rel:"noopener noreferrer",children:"Cohorts"})}),(0,we.jsx)("li",{children:(0,we.jsx)("a",{href:"https://docs.genomebeacons.org/schemas-md/datasets_defaultSchema/",target:"_blank",rel:"noopener noreferrer",children:"Datasets"})}),(0,we.jsx)("li",{children:(0,we.jsx)("a",{href:"https://docs.genomebeacons.org/schemas-md/biosamples_defaultSchema/",target:"_blank",rel:"noopener noreferrer",children:"Biosamples"})}),(0,we.jsx)("li",{children:(0,we.jsx)("a",{href:"https://docs.genomebeacons.org/schemas-md/genomicVariations_defaultSchema/",target:"_blank",rel:"noopener noreferrer",children:"genomicVariations"})}),(0,we.jsx)("li",{children:(0,we.jsx)("a",{href:"https://docs.genomebeacons.org/schemas-md/individuals_defaultSchema/",target:"_blank",rel:"noopener noreferrer",children:"Individuals"})}),(0,we.jsx)("li",{children:(0,we.jsx)("a",{href:"https://docs.genomebeacons.org/schemas-md/runs_defaultSchema/",target:"_blank",rel:"noopener noreferrer",children:"Runs"})}),(0,we.jsx)("li",{children:(0,we.jsx)("a",{href:"https://docs.genomebeacons.org/schemas-md/analyses_defaultSchema/",target:"_blank",rel:"noopener noreferrer",children:"Analyses"})})]})]})]}),(0,we.jsxs)("p",{children:[" ","Beacon v2 aims to be a flexible tool, thus, again not all the fields are mandatory to be filled in. These are the fields that must be included in each model if you decide to include it:"]}),(0,we.jsxs)("h2",{className:"h2-less-margin",children:["Model \u2192 ",(0,we.jsx)("strong",{children:"Required fields"})]}),(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"Analyses \u2192 id, analysisDate, pipelineName"}),(0,we.jsx)("li",{children:"Biosamples \u2192 id, biosampleStatus, sampleOriginType"}),(0,we.jsx)("li",{children:"Cohorts \u2192 id, name, cohortType"}),(0,we.jsx)("li",{children:"Datasets \u2192 id, name"}),(0,we.jsx)("li",{children:"Individuals \u2192 id, sex"}),(0,we.jsx)("li",{children:"Runs \u2192 id, biosampleId, runDate"}),(0,we.jsx)("li",{children:"genomic Variations \u2192 variantInternalId, variation"})]}),(0,we.jsx)("h1",{id:"csv-formatting-tips-and-rules",children:"5. CSV Formatting Tips and Rules"}),(0,we.jsx)("p",{children:"Here are some things to consider during your CSV creation:"}),(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"Only \u201ccorrectly spelled\u201d fields (aka columns) will be inserted into the beacon. You can delete unused columns of your CSV, but don\u2019t modify the headers of the used ones."}),(0,we.jsx)("li",{children:"Every new row will be appended to the final output as a new and independent document. e.g. If in dataset.csv you have two rows, you will be creating two independent datasets."}),(0,we.jsxs)("li",{children:["To add multiple information in the same row separate it by a pipe"," ","'|'. For instance, in genomicVariations, if several samples have the same variant you will need to fill in caseLevelData properties as below. Note that you can leave the pipe empty or write the condition multiple times. e.g:",(0,we.jsx)("br",{}),(0,we.jsx)("br",{}),(0,we.jsx)("code",{className:"examples",children:"caseLevelData | biosampleId \xa0\xa0\xa0\xa0 caseLevelData | phenotypicEffects | conditionId"}),(0,we.jsx)("br",{}),(0,we.jsx)("code",{className:"examples",children:"SAMPLE2 | SAMPLE3 | SAMPLE4 \xa0\xa0\xa0\xa0 Alzheimer | |"}),(0,we.jsx)("br",{}),(0,we.jsx)("br",{}),"In this case, this would mean that the variant in that row was found in sample 2, 3, and 4 and it\u2019s associated to Alzheimer."]}),(0,we.jsxs)("li",{children:["The info field for each collection is very generic and can be filled with different data. You will need to fill the column data directly with JSON-type data. e.g:",(0,we.jsx)("br",{}),(0,we.jsx)("br",{}),(0,we.jsx)("code",{className:"examples",children:'{"info": "This dataset contains synthetic metadata"}'})]}),(0,we.jsx)("br",{}),(0,we.jsx)("li",{children:"The id field in Biosamples must match the sample IDs in the genomicVariations/VCF model. If these IDs don\u2019t match, the mapping with the variants and the biosamples won\u2019t be correctly performed."}),(0,we.jsxs)("li",{children:["The id in Individuals can match or not the ID in"," ","genomicVariations/VCF. If it doesn\u2019t match, for the mapping between individuals and variants to be correctly performed, the field individualsId must be filled in the Biosamples model, mapping the IDs in Individuals and the IDs in genomicVariations/VCF."]}),(0,we.jsx)("li",{children:"The datasetId in Datasets must match the datasetId field in the conf.py (this requirement applies only in the Beacon Production Implementation environment)."})]}),(0,we.jsx)("h1",{id:"configuring-conf-py",children:"6. Configuring conf.py"}),(0,we.jsx)("p",{children:"Now that you\u2019ve already gathered all the necessary information, let\u2019s convert it into Beacon Friendly Format (BFF) using the beacon r.i tools."}),(0,we.jsx)("p",{children:"First, let\u2019s edit the ./beacon2-pi-api/ri-tools/conf/conf.py:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"#### Input and Output files config parameters ####\n\ncsv_folder = './csv/examples/'\noutput_docs_folder='./output_docs/'\n\n#### VCF Conversion config parameters ####\n\nallele_counts=False\nreference_genome='GRCh37' # Choose one between NCBI36, GRCh37, GRCh38\ndatasetId='COVID_pop11_fin_2'\ncase_level_data=False\nexact_heterozygosity=False\nnum_rows=1500000\nverbosity=False"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("edit-ri-tools"),children:e["edit-ri-tools"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol copySymbol-custom",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("h4",{children:"How to fill in config.py?"}),(0,we.jsxs)("div",{children:[(0,we.jsxs)("p",{children:["The ",(0,we.jsx)("code",{children:"csv_folder"})," variable sets the path of a folder containing all the CSVs that will be converted to BFF."," ",(0,we.jsxs)("strong",{children:["Please, don\u2019t modify the path ",(0,we.jsx)("code",{children:"./csv/"}),"."]}),(0,we.jsx)("br",{}),"The ",(0,we.jsx)("code",{children:"output_docs_folder"})," variable sets the folder where your final ",(0,we.jsx)("code",{children:".json"})," files (BFF format) will be saved once execution of beacon tools finishes. This folder should always be located within the folder ",(0,we.jsx)("code",{children:"'output_docs'"}),", but subdirectories can be created inside it.",(0,we.jsx)("em",{children:" e.g"})," ",(0,we.jsx)("code",{children:"output_docs_folder='./output_docs/test1'"}),(0,we.jsx)("br",{}),"If you\u2019ve chosen to read variants directly from a VCF instead of filling in the genomicVariations CSV you\u2019ll need to fill in the VCF Conversion config parameters:"]}),(0,we.jsxs)("ul",{children:[(0,we.jsxs)("li",{children:["The ",(0,we.jsx)("code",{children:"reference_genome"})," is the reference genome the tool will use to map the position of the chromosomes. ",(0,we.jsx)("br",{}),"Make sure to select the same version as the one used to generate your data."]}),(0,we.jsxs)("li",{children:["The ",(0,we.jsx)("code",{children:"datasetId"})," needs to match the id of your"," ","datasets.csv or datasets.json file. ",(0,we.jsx)("br",{}),"This will add a datasetId field in every record to match the record with the dataset it belongs to."]}),(0,we.jsxs)("li",{children:["The ",(0,we.jsx)("code",{children:"case_level_data"})," is a boolean parameter ( True or False) which will relate your variants to the samples they belong to. ",(0,we.jsx)("br",{}),"As we don\u2019t have the biosamples schema, this mapping won\u2019t work. Let\u2019s set it up as False."]}),(0,we.jsxs)("li",{children:["The ",(0,we.jsx)("code",{children:"exact_heterozygosity"})," is a boolean parameter ( True or False) that, in case case_level_data is True, then it will classify the biosamples as being heterozygous for either the reference or the alternate allele. ",(0,we.jsx)("br",{}),"In our case, as case_level_data is False, let\u2019s set this parameter as False too."]}),(0,we.jsxs)("li",{children:["The ",(0,we.jsx)("code",{children:"num_rows"})," are the approximate calculation you expect for the total of variants in each VCF there are. ",(0,we.jsx)("br",{}),"Make sure this is greater than the total variants expected."]}),(0,we.jsxs)("li",{children:["The ",(0,we.jsx)("code",{children:"verbosity"})," will give streaming logs with the reason why a variant has been skipped to be inserted. ",(0,we.jsx)("br",{}),"Recommendation is to leave this as False."]})]})]}),(0,we.jsx)("h1",{id:"converting-csv-vcf-to-bff",children:"7. Converting CSV/VCF to BFF"}),(0,we.jsxs)("p",{children:["Time for the conversion!\xa0",(0,we.jsx)("br",{})," Before preceding ensure that:\xa0",(0,we.jsxs)("ul",{children:[(0,we.jsxs)("li",{children:["All the CSVs that you\u2019ve created have been saved in the csv folder stated in the conf.py,"," ",(0,we.jsxs)("em",{children:["e.g ./beacon2-pi-api/ri-tools/conf/conf.py/csv/your/path.\xa0"," "]})]}),(0,we.jsxs)("li",{children:["All the VCFs to be converted are saved in",(0,we.jsx)("em",{children:"./beacon2-pi-api/ri-tools/files/vcf/files_to_read"})]})]})]}),(0,we.jsx)("p",{children:"If this is done, then run:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker exec -it ri-tools python convert_csvTObff.py"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("python-convert-csvTObff-py"),children:e["python-convert-csvTObff-py"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{children:["The Beacon Friendly Format JSONs from CSVs will be generated in the output_docs folder, with the name of the collection followed by .json extension, e.g. biosamples.json. ",(0,we.jsx)("br",{}),"If you have VCFs to convert, then run too:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker exec -it ri-tools python genomicVariations_vcf.py"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("python-genomicVariations-vcf-py"),children:e["python-genomicVariations-vcf-py"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"And export it from the mongoDB to a JSON BFF with:\xa0"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:'docker exec ri-tools mongoexport --jsonArray --uri "mongodb://\nroot:example@127.0.0.1:27017/beacon?authSource=admin" --collection\ngenomicVariations > ./beacon2-pi-api/ri-tools/output_docs//\ngenomicVariations.json'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("mongo-export-snippet"),children:e["mongo-export-snippet"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol copySymbol-custom",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{children:["All these BFF JSONs will be used to populate a mongoDB for beacon usage.",(0,we.jsx)("br",{}),"Before moving to the actual installation and population of the beacon, ensure all the following steps have been accomplished:",(0,we.jsx)("br",{}),(0,we.jsxs)("ol",{children:[(0,we.jsxs)("li",{children:["If reading from VCFs, move the necessary files to files_to_read folder"," "]}),(0,we.jsx)("li",{children:"Fill in CSV templates "}),(0,we.jsx)("li",{children:"Set the correct paths in conf.py "}),(0,we.jsx)("li",{children:"Run: docker exec -it ri-tools python convert_csvTObff.py"}),(0,we.jsxs)("li",{children:[" ","If reading from a VCF, run: docker exec -it ri-tools python genomicVariations_vcf.py"]}),(0,we.jsx)("li",{children:"Find your BFFs in the output_docs"})," "]})]}),(0,we.jsx)("h1",{id:"deploying-a-beacon-2-production-implementation-api",children:"8. Deploying a Beacon 2 Production Implementation API"}),(0,we.jsxs)("p",{children:["To deploy your first beacon we\u2019ll use the"," ",(0,we.jsx)("a",{href:"https://b2ri-documentation-demo.ega-archive.org/pi-automated-deployment",target:"_blank",rel:"noopener noreferrer",children:"automated deployment"})," ","that relies on the beacon2-pi-api repository that we cloned at the beginning of this tutorial.",(0,we.jsx)("br",{}),"By executing the automated deployment we\u2019ll be performing two key actions:",(0,we.jsxs)("ol",{children:[(0,we.jsx)("li",{children:"The deployment of your beacon instance"}),(0,we.jsx)("li",{children:"The initial insertion of data"})]})]}),(0,we.jsxs)("p",{children:["By default, the data saved in the directory",(0,we.jsx)("em",{children:" ./beacon2-pi-api/beacon/connections/mongo/data/test"})," will be inserted.",(0,we.jsx)("br",{}),(0,we.jsx)("u",{children:"Want to insert your own data?"}),(0,we.jsx)("br",{}),(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"Move your BFFs file to the /data/test directory."}),(0,we.jsx)("li",{children:"Update the path in the /beacon/connections/mongo/Makefile document."})]})]}),(0,we.jsx)("p",{children:"From the root of the repository, ./beacon2-pi-api, run:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"bash mongostart.sh"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("execute-start-script"),children:e["execute-start-script"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{children:["If the operation is successful, you will have a beacon up and running at"," ",(0,we.jsx)("a",{href:"http://localhost:5050/api",target:"_blank",rel:"noopener noreferrer",children:"http://localhost:5050/api"}),"."]}),(0,we.jsx)("p",{children:"This is what you should be seeing:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker exec -it ri-tools python genomicVariations_vcf.py"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("python-genomicVariations-vcf-py"),children:e["python-genomicVariations-vcf-py"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("img",{className:"relationship-elements",alt:"Relationship-elements",src:"/apiSchema.svg"}),(0,we.jsxs)("p",{children:["When searching"," ",(0,we.jsx)("a",{href:"http://localhost:5050/api",target:"_blank",rel:"noopener noreferrer",children:"http://localhost:5050/api"})," ","we land at the ",(0,we.jsx)("code",{children:"/info"})," entrypoint of the API. To make your own landing page please visit Editing your beacon information webpage and modify the required files before deploying."]}),(0,we.jsx)("h1",{id:"reviewing-inserted-entries",children:"9. Reviewing Inserted Entries"}),(0,we.jsxs)("p",{children:["Now, let\u2019s move through your beacon and see the data you have inserted. If you did not modify the content of the"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-pi-api/tree/main/beacon/connections/mongo/data",target:"_blank",rel:"noopener noreferrer",children:"data folder"})," ","the test BFFs will have been included."]}),(0,we.jsx)("p",{children:"To visit the inserted data you need to move through the different beacon endpoints:"}),(0,we.jsxs)("div",{children:[(0,we.jsx)("li",{children:(0,we.jsx)("a",{href:"http://localhost:5050/api/datasets",target:"_blank",rel:"noopener noreferrer",children:"http://localhost:5050/api/datasets"})}),(0,we.jsx)("li",{children:(0,we.jsx)("a",{href:"http://localhost:5050/api/analyses",target:"_blank",rel:"noopener noreferrer",children:"http://localhost:5050/api/analyses"})}),(0,we.jsx)("li",{children:(0,we.jsx)("a",{href:"http://localhost:5050/api/runs",target:"_blank",rel:"noopener noreferrer",children:"http://localhost:5050/api/runs"})}),(0,we.jsx)("li",{children:(0,we.jsx)("a",{href:"http://localhost:5050/api/cohorts",target:"_blank",rel:"noopener noreferrer",children:"http://localhost:5050/api/cohorts"})}),(0,we.jsx)("li",{children:(0,we.jsx)("a",{href:"http://localhost:5050/api/biosamples",target:"_blank",rel:"noopener noreferrer",children:"http://localhost:5050/api/biosamples"})}),(0,we.jsx)("li",{children:(0,we.jsx)("a",{href:"http://localhost:5050/api/individuals",target:"_blank",rel:"noopener noreferrer",children:"http://localhost:5050/api/individuals"})}),(0,we.jsx)("li",{children:(0,we.jsx)("a",{href:"http://localhost:5050/api/g_variants",target:"_blank",rel:"noopener noreferrer",children:"http://localhost:5050/api/g_variants"})})]}),(0,we.jsx)("h1",{id:"loading-more-data",children:"10. Loading more data"}),(0,we.jsx)("p",{children:"From the cloned repository move to:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"cd ./beacon2-pi-api/beacon/connections/mongo/data"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("connections-mongo-data"),children:e["connections-mongo-data"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{children:["And copy all the BFFs that you want to insert.",(0,we.jsx)("br",{})," Then, run:"]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{id:"bulk-mongoimport-snippet",children:'docker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/datasets.json --collection datasets\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/individuals.json --collection individuals\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/cohorts.json --collection cohorts\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/analyses.json --collection analyses\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/biosamples.json --collection biosamples\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/genomicVariations.json --collection genomicVariations\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/runs.json --collection runs\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/targets.json --collection targets\ndocker exec mongoprod mongoimport --jsonArray --uri "mongodb://root:example@127.0.0.1:27017/beacon?authSource=admin" --file /data/caseLevelData.json --collection caseLevelData'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("bulk-mongoimport-snippet"),children:e["bulk-mongoimport-snippet"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol copySymbol-custom",src:ql,alt:"Copy"})})]})}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:"Note that you will only need to run the command for the collections that you have new data.\xa0"})]}),(0,we.jsx)("p",{children:"Each time data is imported into the beacon, indexes need to be created for the queries to run smoothly. Do it by running:"}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{children:"docker exec beaconprod python /beacon/connections/mongo/reindex.py"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>t("connections-mongo-reinde-py"),children:e["connections-mongo-reinde-py"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("br",{}),(0,we.jsx)("br",{})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]})},xc=()=>{const{pathname:e}=ee();return(0,a.useEffect)((()=>{setTimeout((()=>{document.documentElement.scrollTop=0,document.body.scrollTop=0,window.scrollTo({top:0,behavior:"smooth"})}),50)}),[e]),null};const bc=t.p+"static/media/copyIconError.76ff6e4a483cf216799ca31967473d21.svg",vc=()=>{const[e,n]=(0,a.useState)({}),t=(e,t)=>{navigator.clipboard.writeText(e).then((()=>{n((e=>({...e,[t]:!0}))),setTimeout((()=>{n((e=>({...e,[t]:!1})))}),1500)})).catch((e=>console.error("Failed to copy text: ",e)))};return(0,we.jsxs)("div",{className:"configFileToolsContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/configuration-file",className:"no-undeline",children:"Beacon RI Tools v2"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/configuration-file",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:"Common Errors"})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h3",{children:"Beacon RI Tools v2"}),(0,we.jsx)("h1",{children:"Common Errors"}),(0,we.jsx)("h2",{id:"value-error",children:"Case 1: Value Error"}),(0,we.jsx)("div",{className:"codeSnippet error",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{id:"biosamples-error",children:'0%                                                \n | 0/20 [00:00<?, ?it/s]Traceback (most recent call last):\n  File "/usr/src/app/biosamples_csv.py", line 358, in <module>\n    dict_generado, total_i = generate(dict_properties, list_of_headers)\n                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File "/usr/src/app/biosamples_csv.py", line 343, in generate\n    Biosamples(**definitivedict)\n  File "/usr/src/app/validators/biosamples.py", line 171, in __init__\n    super().__init__(**data)\n  File "/usr/local/lib/python3.11/site-packages/pydantic/main.py", line 171, in __init__\n    self.__pydantic_validator__.validate_python(data, self_instance=self)\npydantic_core._pydantic_core.ValidationError: 1 validation error for Biosamples\nhistologicalDiagnosis.id\n  Value error, id must be CURIE, e.g. NCIT:C42331 [type=value_error, input_value=\'NCIT_C4194\', input_type=str]\n    For further information visit https://errors.pydantic.dev/2.6/v/value_error'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>{var e;const n=null===(e=document.getElementById("biosamples-error"))||void 0===e?void 0:e.innerText;n&&t(n,"biosamples-error")},children:e["biosamples-error"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol copySymbol-custom",src:bc,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"This error is triggered during validation of the Biosamples model. It specifically points to the id property within the histologicalDiagnosis field."}),(0,we.jsxs)("p",{children:["Let\u2019s break it down:",(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"histologicalDiagnosis is the term being validated."}),(0,we.jsx)("li",{children:".id is the property within it that\u2019s incorrect."})]})]}),(0,we.jsx)("p",{children:"The error says that the value 'NCIT_C4194' is not a valid CURIE (Compact URI Expression). A CURIE must follow the format: PREFIX:ID."}),(0,we.jsx)("h6",{className:"underline",children:"How to fix it"}),(0,we.jsx)("p",{children:"In the CSV search for the field histologicalDiagnosis|id and ensure it\u2019s filled in with the correct CURIE structure.\xa0"}),(0,we.jsx)("h2",{id:"required-field-missing",children:"Case 2: Required field missing"}),(0,we.jsx)("div",{className:"codeSnippet error",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{id:"datasets-error",children:'Traceback (most recent call last):\n  File "/usr/src/app/datasets_csv.py", line 417, in <module>\n    dict_generado, total_i = generate(dict_properties,list_of_headers)\n                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File "/usr/src/app/datasets_csv.py", line 401, in generate\n    Datasets(**definitivedict)\n  File "/usr/src/app/validators/datasets.py", line 61, in __init__\n    super().__init__(**data)\n  File "/usr/local/lib/python3.11/site-packages/pydantic/main.py", line 171, in __init__\n    self.__pydantic_validator__.validate_python(data, self_instance=self)\npydantic_core._pydantic_core.ValidationError: 1 validation error for Datasets\ndataUseConditions.duoDataUse.description\n  Field required [type=missing, input_value={\'id\': \'DUO:0000019,DUO:0...rsion\': \'7-1-19,7-1-19\'}, input_type=dict]\n    For further information visit https://errors.pydantic.dev/2.6/v/missing\n\n0%'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>{var e;const n=null===(e=document.getElementById("datasets-error"))||void 0===e?void 0:e.innerText;n&&t(n,"datasets-error")},children:e["datasets-error"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol copySymbol-custom",src:bc,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"This validation error occurs when you're providing values for the dataUseConditions.duoDataUse object but omitting one of its required properties: description."}),(0,we.jsx)("p",{children:"In the JSON Schema for the duoDataUse object, the description field is mandatory. This field should provide a human-readable explanation of the data use conditions associated with the DUO (Data Use Ontology) terms provided in id."}),(0,we.jsx)("p",{children:"Because dataUseConditions was included in your dataset entry, the nested structure under duoDataUse must include all required fields\u2014including description."}),(0,we.jsx)("h6",{className:"underline",children:"How to fix it"}),(0,we.jsxs)("table",{className:"dockerTableSimple",children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Property"})}),(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Value"})})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"dataUseConditions | duoDataUse | id"}),(0,we.jsx)("td",{children:"DUO:0000019"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"dataUseConditions | duoDataUse | version"}),(0,we.jsx)("td",{children:"7-1-19"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"dataUseConditions | duoDataUse | description"}),(0,we.jsx)("td",{children:"Health/Medical/Biomedical use only"})]})]})]}),(0,we.jsx)("p",{children:"This will satisfy the schema and resolve the error."}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:"Reminder: Even if duoDataUse appears to be optional overall, any time you provide an object, you must also provide its required fields, as defined in the Beacon schema."})]}),(0,we.jsx)("h2",{id:"json-decode-error",children:"Case 3: JSONDecodeError"}),(0,we.jsx)("div",{className:"codeSnippet error",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{id:"jsondecode-error",children:"json.decoder.JSONDecodeError: Expecting property name enclosed in double quotes: line 1 column 54 (char 53)"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>{var e;const n=null===(e=document.getElementById("jsondecode-error"))||void 0===e?void 0:e.innerText;n&&t(n,"jsondecode-error")},children:e["jsondecode-error"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol copySymbol-custom",src:bc,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"This error is not directly caused by the Beacon RI tools but instead comes from Python\u2019s built-in json library, which fails to parse a malformed JSON string. It typically occurs during the conversion of CSV fields into JSON structures."}),(0,we.jsx)("p",{children:"The message Expecting property name enclosed in double quotes suggests that the parser encountered invalid syntax\u2014usually due to improper formatting of keys, quotes, or punctuation."}),(0,we.jsxs)("p",{children:["Some common causes are:\xa0",(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:'Wrong quotation marks: All JSON keys and string values must use standard double quotes ("). Smart quotes or typographic quotes (like \u201c or \u2018) will trigger this error.'}),(0,we.jsx)("li",{children:"Malformed fields where JSON is expected: info and distribution fields must be filled in as JSON type data. If it contains plain text the conversion will fail. Ensure that info and distribution fields have this structure:"})]})]}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{id:"method-2-env",children:'{"info": "Your information "}\n{"datatypes": {"OGMS:0000015": 20, "OMIABIS:0001032": 20, \u201cOGMS:0000073\u201d:20}}'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>{var e;const n=null===(e=document.getElementById("method-2-env"))||void 0===e?void 0:e.innerText;n&&t(n,"method-2-env")},children:e["method-2-env"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol copySymbol-custom",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"Now, let\u2019s take a look to the headers of the CSV for measurementValue and how we have filled them in:"}),(0,we.jsx)("h6",{className:"underline",children:"How to fix it"}),(0,we.jsx)("p",{children:"Open the CSV file in plain text editor and check if all the double quotes are the same and correct."}),(0,we.jsx)("h2",{id:"oneof-incorrect-use",children:"Case 4: Incorrect use of OneOf"}),(0,we.jsxs)("p",{children:["It\u2019s caused by incorrectly mixing two alternative JSON structures defined, in our example, under the ",(0,we.jsx)("em",{children:"measurementValue"})," ","property. Simplified error:"]}),(0,we.jsx)("p",{children:"Simplified error:"}),(0,we.jsx)("div",{className:"codeSnippet error",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{id:"complex-measurement-error",children:"measurements.measurementValue.Quantity.unit\n  Field required [type=missing]\n\nmeasurements.measurementValue.Quantity.id\n  Extra inputs are not permitted\n\nmeasurements.measurementValue.Quantity.label\n  Extra inputs are not permitted\n\nmeasurements.measurementValue.OntologyTerm.value\n  Extra inputs are not permitted"}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>{var e;const n=null===(e=document.getElementById("complex-measurement-error"))||void 0===e?void 0:e.innerText;n&&t(n,"complex-measurement-error")},children:e["complex-measurement-error"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol copySymbol-custom",src:bc,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"The root of the problem is how the measurementValue property is defined in the JSON schema."}),(0,we.jsx)("div",{className:"codeSnippet",children:(0,we.jsxs)("pre",{children:[(0,we.jsx)("code",{id:"measurement-value-schema",children:'"measurementValue": {\n  "description": "The result of the measurement",\n  "oneOf": [\n    { "$ref": "./value.json" },\n    { "$ref": "./complexValue.json" }\n  ]\n}'}),(0,we.jsx)("button",{className:"copyButtonCode",onClick:()=>{var e;const n=null===(e=document.getElementById("measurement-value-schema"))||void 0===e?void 0:e.innerText;n&&t(n,"measurement-value-schema")},children:e["measurement-value-schema"]?"Copied!":(0,we.jsx)("img",{className:"copySymbol copySymbol-custom",src:ql,alt:"Copy"})})]})}),(0,we.jsx)("p",{children:"The key detail here is oneOf. This means: Only one of these two structures (value.json or complexValue.json) can be used \u2014 not both."}),(0,we.jsxs)("p",{children:["If you choose to use ",(0,we.jsx)("em",{children:"value.json"}),", the structure looks like this (",(0,we.jsx)("em",{children:"quantity.json"}),"):"]}),(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"referenceRange"}),(0,we.jsx)("li",{children:"unit [REQUIRED]"}),(0,we.jsx)("li",{children:"value [REQUIRED]"})]}),(0,we.jsxs)("p",{children:["Now, let\u2019s take a look to the structure of"," ",(0,we.jsx)("em",{children:"complexValue.json"}),":"]}),(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"Quantity \u2013 referenceRange, unit, value [REQUIRED]"}),(0,we.jsx)("li",{children:"QuantityType \u2013 ontologyTerm [REQUIRED]"})]}),(0,we.jsxs)("p",{children:["Now, let\u2019s take a look to the headers of the CSV for"," ",(0,we.jsx)("em",{children:"measurementValue"})," and how we have filled them in:"]}),(0,we.jsxs)("table",{className:"dockerTableSimple dockerTableLong",children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Property"})}),(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Value"})}),(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Schema"})})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"measurements | measurementValue | id"}),(0,we.jsx)("td",{children:"LOINC:26515-7"}),(0,we.jsx)("td",{children:"value.json"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"measurements | measurementValue | label"}),(0,we.jsx)("td",{children:"Platelets [#/volume] in Blood"}),(0,we.jsx)("td",{children:"value.json"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"measurements | measurementValue | referenceRange | high"}),(0,we.jsx)("td",{}),(0,we.jsx)("td",{children:"value.json"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"measurements | measurementValue | referenceRange | low"}),(0,we.jsx)("td",{children:"10"}),(0,we.jsx)("td",{children:"value.json"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"measurements | measurementValue | referenceRange | unit | id"}),(0,we.jsx)("td",{children:"NCIT:C8253"}),(0,we.jsx)("td",{children:"value.json"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"measurements | measurementValue | referenceRange | unit | label"}),(0,we.jsx)("td",{children:"Milligram"}),(0,we.jsx)("td",{children:"value.json"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"measurements | measurementValue | typedQuantities | quantity | referenceRange | high"}),(0,we.jsx)("td",{}),(0,we.jsx)("td",{children:"complexValue.json"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"measurements | measurementValue | typedQuantities | quantity | referenceRange | low"}),(0,we.jsx)("td",{}),(0,we.jsx)("td",{children:"complexValue.json"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"measurements | measurementValue | typedQuantities | quantity | referenceRange | unit"}),(0,we.jsx)("td",{}),(0,we.jsx)("td",{children:"complexValue.json"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"measurements | measurementValue | typedQuantities | quantity | unit | id"}),(0,we.jsx)("td",{children:"-"}),(0,we.jsx)("td",{children:"complexValue.json"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"measurements | measurementValue | typedQuantities | quantity | unit | label"}),(0,we.jsx)("td",{}),(0,we.jsx)("td",{children:"complexValue.json"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"measurements | measurementValue | typedQuantities | quantity | value"}),(0,we.jsx)("td",{}),(0,we.jsx)("td",{children:"complexValue.json"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"measurements | measurementValue | unit | id"}),(0,we.jsx)("td",{}),(0,we.jsx)("td",{children:"complexValue.json"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"measurements | measurementValue | unit | label"}),(0,we.jsx)("td",{}),(0,we.jsx)("td",{children:"complexValue.json"})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"measurements | measurementValue | value"}),(0,we.jsx)("td",{children:"5"}),(0,we.jsx)("td",{children:"complexValue.json"})]})]})]}),(0,we.jsx)("p",{children:"Since measurementValue only accepts one schema (value.json or complexValue.json), combining fields from both leads to:"}),(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"Extra input errors \u2192 You added fields from the wrong schema"}),(0,we.jsx)("li",{children:"Missing required field \u2192 You didn't complete the schema you intended to use"})]}),(0,we.jsx)("h6",{className:"underline",children:"How to fix it"}),(0,we.jsxs)("p",{children:["Decide whether you're using:",(0,we.jsxs)("ul",{children:[(0,we.jsx)("li",{children:"Simple Quantity (value.json) \u2192 Use only value and unit"}),(0,we.jsx)("li",{children:"Complex Quantity (complexValue.json) \u2192 Use only typedQuantities, quantityType, and related nested fields"})]})]}),(0,we.jsx)("p",{children:"Then, ensure all required fields for that choice are included, and remove any fields from the other structure."}),(0,we.jsx)("br",{}),(0,we.jsx)("br",{})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]})};const jc=t.p+"static/media/medal.b5262ca9ed35624405d4c8935b51512e.svg",Cc=()=>{const[e,n]=(0,a.useState)({});return(0,we.jsxs)("div",{className:"configFileToolsContainer",children:[(0,we.jsxs)("h2",{className:"user-path",children:[(0,we.jsx)("a",{href:"/",className:"no-undeline",children:"Documentation"}),(0,we.jsx)("img",{src:"arrow-right-light.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/test-data",className:"no-undeline",children:"Beacon RI Tools v2"}),(0,we.jsx)("img",{src:"arrow-right-bold.svg",alt:"arrow right",className:"arrow-icon"}),(0,we.jsx)("a",{href:"/test-data",className:"no-undeline",children:(0,we.jsx)("span",{className:"user-path-title",children:"Test Data"})})]}),(0,we.jsxs)("div",{className:"contentWrapper",children:[(0,we.jsxs)("div",{className:"contentColumn",children:[(0,we.jsx)("h3",{children:"Beacon RI Tools v2"}),(0,we.jsx)("h1",{children:"Test Data"}),(0,we.jsx)("p",{children:"This section contains test datasets you can use to explore and validate the functionalities of the Beacon v2 Implementation."}),(0,we.jsx)("h1",{id:"dataset-1-cineca-synthetic-cohort-europe-uk1",children:"Dataset 1: CINECA Synthetic Cohort EUROPE UK1"}),(0,we.jsxs)("ul",{children:[(0,we.jsxs)("li",{children:[(0,we.jsx)("strong",{children:"Dataset name: "}),"CINECA_synthetic_cohort_EUROPE_UK1"]}),(0,we.jsxs)("li",{children:[(0,we.jsx)("strong",{children:"Origin: "}),"Derived from"," ",(0,we.jsx)("a",{href:"https://ega-archive.org/datasets/EGAD00001006673",target:"_blank",rel:"noopener noreferrer",children:"EGAD00001006673"})," ","(EGA) and a"," ",(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools/blob/main/CINECA_synthetic_cohort_EUROPE_UK1/uk1.tsv",target:"_blank",rel:"noopener noreferrer",children:"raw file"})," ","with phenoclinic data gently donated to us from our CINECA-EU partners."]}),(0,we.jsxs)("li",{children:[(0,we.jsx)("strong",{children:"Study size: "}),"2504 samples with genetic data (low coverage WGS) based on 1000 Genomes data (phase3 and Geuvadis), and 76 synthetic subject attributes and phenotypic data derived from UKBiobank."]})]}),(0,we.jsxs)("p",{className:"wider-note cite-box",children:[(0,we.jsx)("img",{className:"note-symbol-wider",src:jc,alt:"Note symbol"}),(0,we.jsx)("div",{children:"In support of FAIRness in data sharing, this dataset is made freely available under the Creative Commons Licence (CC-BY). Please ensure this preamble is included with this dataset and that the CINECA project (funding: EC H2020 grant 825775) is acknowledged."})]}),(0,we.jsx)("h6",{className:"underline",children:"Included files"}),(0,we.jsxs)("table",{className:"dockerTableSimple dockerTableLong",children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"File/Folder"})}),(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Description"})}),(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Where to find"})})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"uk1.tsv"}),(0,we.jsx)("td",{children:"Raw metadata and phenotypic data for 2,504 synthetic individuals"}),(0,we.jsx)("td",{children:(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/blob/main/csv/examples/CINECA_synthetic_cohort_EUROPE_UK1/uk1.tsv",target:"_blank",rel:"noopener noreferrer",children:"uk1.tsv"})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsxs)("td",{children:["chr22.Test.1000G.phas",(0,we.jsx)("br",{}),"e3.joint.vcf.gz"]}),(0,we.jsx)("td",{children:"VCF file consisting of WGS for chr22 \u2013 input used for genomicVariations"}),(0,we.jsxs)("td",{children:["$ wget ftp://",(0,we.jsx)("br",{}),"FTPuser:FTPusersPassword@xfer13",(0,we.jsx)("br",{}),".crg.eu:221/external_files/",(0,we.jsx)("br",{}),"CINECA_synthetic_cohort_EUROPE_",(0,we.jsx)("br",{}),"UK1/vcf/",(0,we.jsx)("br",{}),"chr22.Test.1000G.phase3.joint.vcf",(0,we.jsx)("br",{}),".gz"]})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"CSV models"}),(0,we.jsx)("td",{children:"CSV filled in with the data and metadata from the dataset files, ready to convert to Beacon Friendly Format (BFF)"}),(0,we.jsx)("td",{children:(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/tree/main/csv/examples/CINECA_synthetic_cohort_EUROPE_UK1",target:"_blank",rel:"noopener noreferrer",children:"Input folder"})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"BFF output"}),(0,we.jsx)("td",{children:"Output BFF files obtained by converting the CSV models using the beacon2 ri tools v2"}),(0,we.jsx)("td",{children:(0,we.jsxs)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/tree/main/output_docs/CINECA_synthetic_cohort_EUROPE_UK1",target:"_blank",rel:"noopener noreferrer",children:["CINECA_synthetic_cohort_EUROPE_",(0,we.jsx)("br",{}),"UK1"]})})]})]})]}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:(0,we.jsxs)("p",{className:"note-paragraph",children:["\xa0If you want to download the full original dataset metadata please follow the"," ",(0,we.jsx)("a",{href:"https://ega-archive.org/access/download/files/live-outbox/",children:"Live Distribution"})," ","or the"," ",(0,we.jsx)("a",{href:"https://ega-archive.org/access/download/files/pyega3/",children:"pyEGA3"})," ","instructions. Notice that these datasets are free access."]})})]}),(0,we.jsx)("h1",{id:"dataset-2-rare-disease-synthetic-dataset",children:"Dataset 2: Rare Disease Synthetic Dataset"}),(0,we.jsxs)("ul",{children:[(0,we.jsxs)("li",{children:[(0,we.jsx)("b",{children:"Dataset name:"})," Rare Disease Synthetic Dataset"]}),(0,we.jsxs)("li",{children:[(0,we.jsx)("b",{children:"Origin:"})," Derived from"," ",(0,we.jsx)("a",{href:"https://ega-archive.org/datasets/EGAD00001008392",target:"_blank",rel:"noopener noreferrer",children:"EGAD00001008392"})," ","(EGA)."]}),(0,we.jsxs)("li",{children:[(0,we.jsx)("b",{children:"Study size:"})," 18 whole genome samples (6 index cases with their parents) created with public human genomic background combined with in-silico insertion of real disease-causing variants."]})]}),(0,we.jsxs)("p",{className:"wider-note cite-box",children:[(0,we.jsx)("img",{className:"note-symbol-wider",src:jc,alt:"Note symbol"}),(0,we.jsx)("div",{children:"When using this dataset, please acknowledge the following contributors and funding sources: the RD-Connect GPAP (https://platform.rd-connect.eu/), EC H2020 project EJP-RD (grant # 825575), EC H2020 project B1MG (grant # 951724) and Generalitat de Catalunya VEIS project (grant # 001-P-001647)."})]}),(0,we.jsxs)("table",{className:"dockerTableSimple",children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"File/Folder"})}),(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Description"})}),(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Where to find"})})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:(0,we.jsx)("a",{href:"https://metadata.ega-archive.org/datasets/EGAD00001008392",target:"_blank",children:"EGAD00001008392"})}),(0,we.jsx)("td",{children:"Rare Disease Synthetic Dataset"}),(0,we.jsxs)("td",{children:["See Use Case: Inserting Rare Disease Synthetic Dataset into a beacon \u2192"," ",(0,we.jsxs)("a",{href:"https://b2ri-documentation-demo.ega-archive.org/use-case:-rare-diseases-data#dataset-model-creation%C2%A0",children:[" ","Dataset model creation section"]})]})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:(0,we.jsx)("a",{href:"https://metadata.ega-archive.org/samples/EGAN00003364605",target:"_blank",children:"EGAN00003364605"})}),(0,we.jsx)("td",{children:"case1_index \u2013 Male sample with Central core disease."}),(0,we.jsxs)("td",{children:["See Use Case: Inserting Rare Disease Synthetic Dataset into a beacon \u2192"," ",(0,we.jsxs)("a",{href:"https://b2ri-documentation-demo.ega-archive.org/use-case:-rare-diseases-data#dataset-model-creation%C2%A0",children:[" ","Dataset model creation section"]})]})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:(0,we.jsx)("a",{href:"https://metadata.ega-archive.org/runs/EGAR00003021166",target:"_blank",children:"EGAR00003021166"})}),(0,we.jsx)("td",{children:"case1_index fastq"}),(0,we.jsxs)("td",{children:["See Use Case: Inserting Rare Disease Synthetic Dataset into a beacon \u2192"," ",(0,we.jsx)("a",{href:"https://b2ri-documentation-demo.ega-archive.org/use-case:-rare-diseases-data#run-model-creation",children:"Run model from VCF section"})]})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:(0,we.jsx)("a",{href:"https://metadata.ega-archive.org/analyses/EGAZ00001744017",target:"_blank",children:"EGAZ00001744017"})}),(0,we.jsx)("td",{children:"Information about the creation of the chromosome 21 VCF"}),(0,we.jsxs)("td",{children:["See Use Case: Inserting Rare Disease Synthetic Dataset into a beacon \u2192"," ",(0,we.jsx)("a",{href:"https://b2ri-documentation-demo.ega-archive.org/use-case:-rare-diseases-data#Analysis-model-creation",children:"Analysis model creation section"})]})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:(0,we.jsx)("a",{href:"https://metadata.ega-archive.org/analyses/EGAZ00001744003",target:"_blank",children:"EGAZ00001744003"})}),(0,we.jsx)("td",{children:"Phenopacket with individual information from case1_index sample"}),(0,we.jsx)("td",{children:(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/tree/main/phenopackets-to-BFF/rare-disease",target:"_blank",children:"phenopacket"})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:(0,we.jsx)("a",{href:"https://metadata.ega-archive.org/files/EGAF00005572743",target:"_blank",children:"EGAF00005572743"})}),(0,we.jsx)("td",{children:"5 variants subset of the chromosome 21 VCF from case1_index"}),(0,we.jsxs)("td",{children:["See Use Case: Inserting Rare Disease Synthetic Dataset into a beacon \u2192"," ",(0,we.jsx)("a",{href:"https://b2ri-documentation-demo.ega-archive.org/use-case:-rare-diseases-data#genomicVariations-model-from-VCF",children:"genomicVariations model from VCF creation section"})]})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Rare disease use case folder"}),(0,we.jsx)("td",{children:"Folder with the CSVs for dataset and run creation, phenopacket for individuals model and the subset VCF for genomicVariations."}),(0,we.jsx)("td",{children:(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/tree/main/csv/examples/Rare-diseases",target:"_blank",children:"Input folder"})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"Rare disease use case output folder"}),(0,we.jsx)("td",{children:"Output BFF files obtained by converting the CSV models, phenopacket and VCF using the beacon2 ri tools v2"}),(0,we.jsx)("td",{children:(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/tree/main/output_docs/Rare-diseases",target:"_blank",children:"Output folder"})})]})]})]}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:(0,we.jsxs)("p",{className:"note-paragraph",children:["\xa0If you want to download the full original dataset metadata please follow the"," ",(0,we.jsx)("a",{href:"https://ega-archive.org/access/download/files/live-outbox/",children:"Live Distribution"})," ","or the"," ",(0,we.jsx)("a",{href:"https://ega-archive.org/access/download/files/pyega3/",children:"pyEGA3"})," ","instructions. Notice that these datasets are free access."]})})]}),(0,we.jsx)("h1",{id:"dataset-3-example-csv-all-fields",children:"Dataset 3: Example CSV with all the fields filled in"}),(0,we.jsxs)("ul",{children:[(0,we.jsxs)("li",{children:[(0,we.jsx)("b",{children:"Dataset name:"})," Dataset with synthetic data for beacon testing"]}),(0,we.jsxs)("li",{children:[(0,we.jsx)("b",{children:"Origin:"})," All the metadata inserted is synthetic."]}),(0,we.jsxs)("li",{children:[(0,we.jsx)("b",{children:"Study size:"})," 20 female samples with breast carcinoma."]})]}),(0,we.jsx)("h6",{className:"underline",children:"Included files"}),(0,we.jsxs)("table",{className:"dockerTableSimple dockerTableLong",children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"File/Folder"})}),(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Description"})}),(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Where to find"})})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"CSV folder"}),(0,we.jsx)("td",{children:"Example CSVs with all the possible fields filled in for analysis, datasets, cohorts, individuals, runs and biosamples model."}),(0,we.jsx)("td",{children:(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/tree/main/csv/examples/all-fields",target:"_blank",rel:"noopener noreferrer",children:"Input folder"})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"BFF output"}),(0,we.jsx)("td",{children:"Folder with the BFFs obtained by converting the VCF and the CSV using beacon2-ri-tools-v2"}),(0,we.jsx)("td",{children:(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/tree/main/output_docs/all-fields",target:"_blank",rel:"noopener noreferrer",children:"Output folder"})})]})]})]}),(0,we.jsx)("h1",{id:"dataset-4-example-csv-mandatory-fields",children:"Dataset 4: Example CSV with only the mandatory fields filled in"}),(0,we.jsx)("h6",{className:"underline",children:"Included files"}),(0,we.jsxs)("table",{className:"dockerTableSimple dockerTableLong",children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"File/Folder"})}),(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Description"})}),(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Where to find"})})]})}),(0,we.jsx)("tbody",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"CSV folder"}),(0,we.jsx)("td",{children:"Example CSVs with all the possible fields filled in for analysis, datasets, cohorts, individuals, runs and biosamples model."}),(0,we.jsx)("td",{children:(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/tree/main/csv/examples/mandatory-fields",target:"_blank",rel:"noopener noreferrer",children:"Input folder"})})]})})]}),(0,we.jsx)("h1",{id:"dataset-5-genomicvariations-csvs",children:"Dataset 5: genomicVariations CSVs"}),(0,we.jsxs)("p",{children:["You can load genomicVariations data into Beacon using either a VCF file or a CSV. Here you\u2019ll find two genomicVariations CSVs examples, one representing variants with caseLevelData information and the second one with variantLevelData.",(0,we.jsxs)("ul",{children:[(0,we.jsxs)("li",{children:[(0,we.jsx)("b",{children:"caseLevelData:"})," Describes which biosamples carry which variants. This is sample-level information based on individual analyses."]}),(0,we.jsxs)("li",{children:[(0,we.jsx)("b",{children:"variantLevelData:"})," Describes the variants themselves, without linking them to specific individuals. This is aggregate-level information, such as genomic position, alleles, frequency or phenotype associated."]})]})]}),(0,we.jsx)("h6",{className:"underline",children:"Included files"}),(0,we.jsxs)("table",{className:"dockerTableSimple dockerTableLong",children:[(0,we.jsx)("thead",{children:(0,we.jsxs)("tr",{children:[(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"File/Folder"})}),(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Description"})}),(0,we.jsx)("th",{children:(0,we.jsx)("b",{children:"Where to find"})})]})}),(0,we.jsxs)("tbody",{children:[(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"caseLevelData folder"}),(0,we.jsx)("td",{children:"genomicVariations CSV with the mandatory fields of the model and the caseLevelData properties."}),(0,we.jsx)("td",{children:(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/tree/main/csv/examples/genomicVariations/caseLevelData",target:"_blank",rel:"noopener noreferrer",children:"caseLevelData"})})]}),(0,we.jsxs)("tr",{children:[(0,we.jsx)("td",{children:"variantLevelData folder"}),(0,we.jsx)("td",{children:"genomicVariations CSV with the mandatory fields of the model and the variantLevelData properties."}),(0,we.jsx)("td",{children:(0,we.jsx)("a",{href:"https://github.com/EGA-archive/beacon2-ri-tools-v2/tree/main/csv/examples/genomicVariations/variantLevelData",target:"_blank",rel:"noopener noreferrer",children:"variantLevelData"})})]})]})]}),(0,we.jsxs)("p",{className:"note",children:[(0,we.jsx)("img",{className:"note-symbol",src:"https://github.com/costero-e/beacon-ri-documentation/blob/main/docu-app/public/note-symbol.png?raw=true",alt:"Note symbol"}),(0,we.jsx)("div",{children:(0,we.jsx)("p",{className:"note-paragraph",children:"Note: You can include either or both blocks in your Beacon instance. Combining both gives you a richer model that supports individual-level queries and population-level insights."})})]}),(0,we.jsx)("br",{}),(0,we.jsx)("br",{})]}),(0,we.jsx)("div",{className:"sidebarColumn",children:(0,we.jsx)(Hl,{})})]})]})};const wc=function(){const[e,n]=(0,a.useState)(""),[t,o]=(0,a.useState)([]),[i,r]=(0,a.useState)(!1);return(0,we.jsx)(ke,{children:(0,we.jsxs)(ve,{children:[(0,we.jsx)(xc,{}),(0,we.jsx)(Ql,{}),(0,we.jsxs)("div",{className:"appContainer",children:[(0,we.jsx)(_e,{onSearch:e=>{n(e);o([])}}),(0,we.jsxs)("div",{className:"contentContainer",children:[(0,we.jsx)(Gl,{menuItems:["Introduction","Beacon 2 PI API","Beacon 2 RI API","Beacon RI Tools v2","Beacon UI","Beacon Network UI","Tutorials","Resources"],subMenuItems:{"Beacon 2 PI API":["Automated Deployment","Manual Deployment","Filtering Terms","Configuration","Querying the API"],"Beacon 2 RI API":["Automated Deployment","Manual Deployment","Data Linking","Configuration","Querying the API"],"Beacon RI Tools v2":["Starting Guide","Configuration File","Conversion from CSV to BFF","Conversion from VCF to BFF","Conversion from Phenopackets to BFF","Common Errors","Test Data"],"Beacon UI":["Deployment","Configuration UI","Querying the UI"],"Beacon Network UI":["Network Deployment","Network Configuration UI","Network Querying the UI"],Tutorials:["Create your Beacon","Use Case: Rare Diseases Data"]},isSubmenuOpen:i,toggleSubmenu:()=>{r(!i)}}),(0,we.jsx)("div",{className:"contentContainer ".concat(i?"withSubmenuOpen":"withSubmenuClosed"),children:(0,we.jsxs)(ye,{children:[(0,we.jsx)(fe,{path:"/",element:(0,we.jsx)(Vl,{})}),(0,we.jsx)(fe,{path:"/pi-automated-deployment",element:(0,we.jsx)(uc,{})}),(0,we.jsx)(fe,{path:"/pi-manual-deployment",element:(0,we.jsx)(cc,{})}),(0,we.jsx)(fe,{path:"/filtering-terms",element:(0,we.jsx)(dc,{})}),(0,we.jsx)(fe,{path:"/configuration",element:(0,we.jsx)(sc,{})}),(0,we.jsx)(fe,{path:"/pi-querying-the-api",element:(0,we.jsx)(lc,{})}),(0,we.jsx)(fe,{path:"/automated-deployment",element:(0,we.jsx)(Wl,{})}),(0,we.jsx)(fe,{path:"/manual-deployment",element:(0,we.jsx)(zl,{})}),(0,we.jsx)(fe,{path:"/data-linking",element:(0,we.jsx)(Kl,{})}),(0,we.jsx)(fe,{path:"/api-configuration",element:(0,we.jsx)($l,{})}),(0,we.jsx)(fe,{path:"/querying-the-api",element:(0,we.jsx)(Yl,{})}),(0,we.jsx)(fe,{path:"/starting-guide",element:(0,we.jsx)(Xl,{})}),(0,we.jsx)(fe,{path:"/configuration-file",element:(0,we.jsx)(Zl,{})}),(0,we.jsx)(fe,{path:"/conversion-from-csv-to-bff",element:(0,we.jsx)(Jl,{})}),(0,we.jsx)(fe,{path:"/conversion-from-vcf-to-bff",element:(0,we.jsx)(ec,{})}),(0,we.jsx)(fe,{path:"/conversion-from-phenopackets-to-bff",element:(0,we.jsx)(hc,{})}),(0,we.jsx)(fe,{path:"/common-errors",element:(0,we.jsx)(vc,{})}),(0,we.jsx)(fe,{path:"/test-data",element:(0,we.jsx)(Cc,{})}),(0,we.jsx)(fe,{path:"/deployment",element:(0,we.jsx)(nc,{})}),(0,we.jsx)(fe,{path:"/configuration-ui",element:(0,we.jsx)(tc,{})}),(0,we.jsx)(fe,{path:"/querying-the-ui",element:(0,we.jsx)(oc,{})}),(0,we.jsx)(fe,{path:"/network-deployment",element:(0,we.jsx)(ac,{})}),(0,we.jsx)(fe,{path:"/network-configuration-ui",element:(0,we.jsx)(ic,{})}),(0,we.jsx)(fe,{path:"/network-querying-the-ui",element:(0,we.jsx)(rc,{})}),(0,we.jsx)(fe,{path:"/create-your-beacon",element:(0,we.jsx)(yc,{})}),(0,we.jsx)(fe,{path:"/use-case:-rare-diseases-data",element:(0,we.jsx)(fc,{})}),(0,we.jsx)(fe,{path:"/resources",element:(0,we.jsx)(mc,{})})]})})]})]})]})})},Sc=e=>{e&&e instanceof Function&&t.e(510).then(t.bind(t,510)).then((n=>{let{getCLS:t,getFID:o,getFCP:a,getLCP:i,getTTFB:r}=n;t(e),o(e),a(e),i(e),r(e)}))};r.createRoot(document.getElementById("root")).render((0,we.jsx)(wc,{})),Sc()})();
//# sourceMappingURL=main.53955aac.js.map